{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Numpy, Scipy, and Matplotlib\n",
    "\n",
    "<a href=\"http://python.org\"> Python </a> is a high-level, general-purpose programming language and it offers several tools that are broadly used by the scientific community.\n",
    "\n",
    "## Wait, is not Python an interpreted language? Is it efficient enough for heavy numeric computations?\n",
    "\n",
    "<a href=\"http://numpy.scipy.org/\">NumPy</a> is a large library of numerical and mathematical functions which  provides the support of highly optimized multidimensional arrays. \n",
    "\n",
    "<a href=\"http://www.scipy.org/\">SciPy</a> uses those arrays to provide scientific and technical computing which contatins modules for optimization, linear algebra and FFT. \n",
    "\n",
    "<a href=\"http://matplotlib.org/\">Matplotlib</a> is probably the most convenient and feature-rich library to plot high-quality graphs using Python.\n",
    "\n",
    "## There is more ... \n",
    "\n",
    "<a href=\"http://ipython.org/\">IPython</a> is an interactive interface, letting you quickly process data and test ideas. \n",
    "\n",
    "The (IPython) Jupyter **notebook** works in your web browser, allowing you to document your computations in an easily reproducible form, it is an interactive computational environment which you can combine code execution, rich text, mathematics, plots and rich media.\n",
    "\n",
    "## Why is it so popular?\n",
    "\n",
    "Perfect language for quick prototyping: it is interpreted and dynamically typed; provides high-level programming with nice readability.\n",
    "\n",
    "Python incorporates documentation directly into the language itself. Moreover, since there exists a large community of users and developers, it is easy to find help, recipes, and code snippets on-line.\n",
    "\n",
    "It is Free software / Open source.\n",
    "\n",
    "It runs natively on Windows, Mac OS, linux, and others.\n",
    "\n",
    "See ... <a href=\"http://www.xkcd.com/353/\">Programming is fun again!</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Installing Python, IPython, Numpy, Scipy, and Matplotlib\n",
    "\n",
    "<a href=\"http://continuum.io/downloads.html\">Anaconda</a>: A free Python distribution which supports Linux, Windows and Mac.\n",
    "\n",
    "Anaconda includes Python and analytic Python packages such as NumPy, SciPy, Matplotlib, and IPython.\n",
    "\n",
    "After downloading the installer of **Python 2.7**.\n",
    "\n",
    "## Linux\n",
    "\n",
    "In your terminal window execute for Python 2.7:\n",
    "<code> bash Anaconda2-4.3.0-Linux-x86_64.sh </code>\n",
    "\n",
    "\n",
    "## Windows\n",
    "\n",
    "Double-click the .exe file to install Anaconda and follow the instructions on the screen.\n",
    "\n",
    "## Mac\n",
    "\n",
    "- Double click the .pkg file and follow the instructions on the screen.\n",
    "- Command-Line Installs:\n",
    "After downloading the installer, in the shell execute for Python 2.7: \n",
    "\n",
    "<code> bash Anaconda2-4.3.0-MacOSX-x86_64.sh </code>\n",
    "\n",
    "NOTE: You should type \"bash\", regardless of whether or not you are actually using the bash shell.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using the Jupyter notebook\n",
    "\n",
    "The Jupyter notebook allows you to include text, code, and plots in the same document.\n",
    "\n",
    "This makes it ideal for example to write up a report about a python project and share it with others.\n",
    "\n",
    "**Starting up**\n",
    "The Jupyter Notebook App can be launched by clicking on the Jupyter Notebook icon installed by Anaconda in the start menu (Windows) or by typing in a terminal:\n",
    "\n",
    " jupyter notebook \n",
    "\n",
    "**First steps**\n",
    "At first glance, a notebook looks like a fairly typical application - it has a menubar (File, Edit, View, etc.) and a tool bar with icons. Below this, you will see an empty cell, in which you can type any Python code. You can write several lines of code, and once it is ready to run, you can press Shift+Enter and it will get executed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the pattern analysis and recognition Hello World!\n"
     ]
    }
   ],
   "source": [
    "print \"This is the pattern analysis and recognition Hello World!\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then click on that cell, change the Python code, and press Shift+Enter again to re-execute the code.\n",
    "\n",
    "You can insert/copy/paste cells at any point of your notebook using the \"Insert\" and \"Edit\" menu entries.\n",
    "\n",
    "There are basically two types of cells: code cells and text cells. You can change the type of a cell by using the \"Cell\" menu item. Text cells can be just raw text or formatted using Markdown syntax. Markdown sytanx allow you to render text with format (e.g. headings, bold text, lists, etc.)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Tips**\n",
    "\n",
    "Save often! There is currently no auto-save in the notebook, so you will lose your changes if you close the browser window. You can save your notebook using the \"File\" menuitem or just pressing Ctr+S keys.\n",
    "\n",
    "If you need more info about the Jupyter notebook check the oficial documentation <a href=\"http://ipython.org/notebook.html\">here</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.11.3\n"
     ]
    }
   ],
   "source": [
    "# Import NumPy\n",
    "import numpy as np\n",
    "\n",
    "print(np.version.full_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 1 2 3 4 5 6 7 8]\n"
     ]
    }
   ],
   "source": [
    "# Create a vector (1D array)\n",
    "A = np.array([0,1,2,3,4,5,6,7,8])\n",
    "\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on built-in function array in module numpy.core.multiarray:\n",
      "\n",
      "array(...)\n",
      "    array(object, dtype=None, copy=True, order=None, subok=False, ndmin=0)\n",
      "    \n",
      "    Create an array.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    object : array_like\n",
      "        An array, any object exposing the array interface, an\n",
      "        object whose __array__ method returns an array, or any\n",
      "        (nested) sequence.\n",
      "    dtype : data-type, optional\n",
      "        The desired data-type for the array.  If not given, then\n",
      "        the type will be determined as the minimum type required\n",
      "        to hold the objects in the sequence.  This argument can only\n",
      "        be used to 'upcast' the array.  For downcasting, use the\n",
      "        .astype(t) method.\n",
      "    copy : bool, optional\n",
      "        If true (default), then the object is copied.  Otherwise, a copy\n",
      "        will only be made if __array__ returns a copy, if obj is a\n",
      "        nested sequence, or if a copy is needed to satisfy any of the other\n",
      "        requirements (`dtype`, `order`, etc.).\n",
      "    order : {'C', 'F', 'A'}, optional\n",
      "        Specify the order of the array.  If order is 'C', then the array\n",
      "        will be in C-contiguous order (last-index varies the fastest).\n",
      "        If order is 'F', then the returned array will be in\n",
      "        Fortran-contiguous order (first-index varies the fastest).\n",
      "        If order is 'A' (default), then the returned array may be\n",
      "        in any order (either C-, Fortran-contiguous, or even discontiguous),\n",
      "        unless a copy is required, in which case it will be C-contiguous.\n",
      "    subok : bool, optional\n",
      "        If True, then sub-classes will be passed-through, otherwise\n",
      "        the returned array will be forced to be a base-class array (default).\n",
      "    ndmin : int, optional\n",
      "        Specifies the minimum number of dimensions that the resulting\n",
      "        array should have.  Ones will be pre-pended to the shape as\n",
      "        needed to meet this requirement.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    out : ndarray\n",
      "        An array object satisfying the specified requirements.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    empty, empty_like, zeros, zeros_like, ones, ones_like, fill\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> np.array([1, 2, 3])\n",
      "    array([1, 2, 3])\n",
      "    \n",
      "    Upcasting:\n",
      "    \n",
      "    >>> np.array([1, 2, 3.0])\n",
      "    array([ 1.,  2.,  3.])\n",
      "    \n",
      "    More than one dimension:\n",
      "    \n",
      "    >>> np.array([[1, 2], [3, 4]])\n",
      "    array([[1, 2],\n",
      "           [3, 4]])\n",
      "    \n",
      "    Minimum dimensions 2:\n",
      "    \n",
      "    >>> np.array([1, 2, 3], ndmin=2)\n",
      "    array([[1, 2, 3]])\n",
      "    \n",
      "    Type provided:\n",
      "    \n",
      "    >>> np.array([1, 2, 3], dtype=complex)\n",
      "    array([ 1.+0.j,  2.+0.j,  3.+0.j])\n",
      "    \n",
      "    Data-type consisting of more than one element:\n",
      "    \n",
      "    >>> x = np.array([(1,2),(3,4)],dtype=[('a','<i4'),('b','<i4')])\n",
      "    >>> x['a']\n",
      "    array([1, 3])\n",
      "    \n",
      "    Creating an array from sub-classes:\n",
      "    \n",
      "    >>> np.array(np.mat('1 2; 3 4'))\n",
      "    array([[1, 2],\n",
      "           [3, 4]])\n",
      "    \n",
      "    >>> np.array(np.mat('1 2; 3 4'), subok=True)\n",
      "    matrix([[1, 2],\n",
      "            [3, 4]])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Remember we can always get help from a Python object/function this way:\n",
    "help(np.array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the number of dimensions of an array\n",
    "A.ndim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9L,)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get it's shape\n",
    "A.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0 1 2]\n",
      " [3 4 5]\n",
      " [6 7 8]]\n"
     ]
    }
   ],
   "source": [
    "# Create a two dimensional array (3x3 matrix)\n",
    "B = np.array([[0,1,2],[3,4,5],[6,7,8]])\n",
    "\n",
    "print(B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, (3L, 3L))\n"
     ]
    }
   ],
   "source": [
    "print(B.ndim, B.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other ways to create arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1.,  1.,  1.])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create array (1D) filled with ones\n",
    "np.ones(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.,  0.,  0.],\n",
       "       [ 0.,  0.,  0.],\n",
       "       [ 0.,  0.,  0.]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create array (with shape (3,3)) filled with zeros\n",
    "np.zeros([3,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 1.,  0.,  0.],\n",
       "       [ 0.,  1.,  0.],\n",
       "       [ 0.,  0.,  1.]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identity matrix (2D)\n",
    "np.eye(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.88945588,  0.83393596,  0.43754882,  0.31495735],\n",
       "       [ 0.22267123,  0.87174515,  0.02436845,  0.41625465],\n",
       "       [ 0.89969582,  0.13522232,  0.60048985,  0.58233258],\n",
       "       [ 0.50469275,  0.45851164,  0.15729314,  0.30126492]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a random array with shape (4,4)\n",
    "np.random.rand(4,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 26  41 218 146]\n",
      " [142 194  26  16]\n",
      " [ 49 168 138 183]\n",
      " [144 224 164 125]]\n",
      "[[207 121 210 170]\n",
      " [217 227 181 189]\n",
      " [239 214 168 162]\n",
      " [250 181 130 147]]\n"
     ]
    }
   ],
   "source": [
    "# Create a random array of integers \n",
    "A = np.random.randint(255,size=(4,4))  # from 0-255\n",
    "B = np.random.randint(100,255,size=(4,4)) # from 100-255\n",
    "\n",
    "print (A)\n",
    "print (B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24\n",
      " 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49\n",
      " 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74\n",
      " 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99]\n",
      "[100 103 106 109 112 115 118 121 124 127 130 133 136 139 142 145 148 151\n",
      " 154 157 160 163 166 169 172 175 178 181 184 187 190 193 196 199]\n",
      "[ 100.  103.  106.  109.  112.  115.  118.  121.  124.  127.  130.  133.\n",
      "  136.  139.  142.  145.  148.  151.  154.  157.  160.  163.  166.  169.\n",
      "  172.  175.  178.  181.  184.  187.  190.  193.  196.  199.]\n"
     ]
    }
   ],
   "source": [
    "# Create an array with evenly spaced values within a given interval\n",
    "\n",
    "A = np.arange(100) # from 0 to 99, step=1\n",
    "\n",
    "B = np.arange(100,200,3) # from 100 to 199, step=3\n",
    "\n",
    "C = np.arange(100,200,3, dtype = np.float64) # from 100 to 199, step=3, data type = float64\n",
    "\n",
    "print(A)\n",
    "print(B)\n",
    "print(C) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reshape matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 1 2 3 4 5 6 7 8]\n",
      "[[0 1 2]\n",
      " [3 4 5]\n",
      " [6 7 8]]\n"
     ]
    }
   ],
   "source": [
    "A = np.array([0,1,2,3,4,5,6,7,8])\n",
    "print(A)\n",
    "\n",
    "# Reshape A to be a 3x3 matrix\n",
    "B = np.reshape(A,[3,3])\n",
    "print(B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "total size of new array must be unchanged",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-efd5a144d851>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[1;31m# New shape must agree in number of elements\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[1;31m# This will rise an error because A do not have enough elements to fill a 4x4 array\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mC\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mA\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32mD:\\Users\\byron\\Anaconda2\\lib\\site-packages\\numpy\\core\\fromnumeric.pyc\u001b[0m in \u001b[0;36mreshape\u001b[0;34m(a, newshape, order)\u001b[0m\n\u001b[1;32m    222\u001b[0m     \u001b[1;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    223\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0m_wrapit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'reshape'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnewshape\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0morder\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0morder\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m--> 224\u001b[0;31m     \u001b[1;32mreturn\u001b[0m \u001b[0mreshape\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnewshape\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0morder\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0morder\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    225\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    226\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: total size of new array must be unchanged"
     ]
    }
   ],
   "source": [
    "# New shape must agree in number of elements\n",
    "# This will rise an error because A do not have enough elements to fill a 4x4 array\n",
    "C = np.reshape(A,[4,4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Append rows/cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.98518129  0.60898041  0.40566328  0.01670487]\n",
      " [ 0.6450601   0.85978905  0.66902865  0.61851746]\n",
      " [ 0.7956447   0.10383916  0.73235019  0.7352059 ]\n",
      " [ 0.42563489  0.68353725  0.20302183  0.48926061]]\n",
      "[[ 0.]\n",
      " [ 0.]\n",
      " [ 0.]\n",
      " [ 0.]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[ 0.98518129,  0.60898041,  0.40566328,  0.01670487,  0.        ],\n",
       "       [ 0.6450601 ,  0.85978905,  0.66902865,  0.61851746,  0.        ],\n",
       "       [ 0.7956447 ,  0.10383916,  0.73235019,  0.7352059 ,  0.        ],\n",
       "       [ 0.42563489,  0.68353725,  0.20302183,  0.48926061,  0.        ]])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = np.random.rand(4,4)\n",
    "print(A)\n",
    "\n",
    "B = np.zeros([4,1]) # column vector\n",
    "print(B)\n",
    "\n",
    "# Append column vector (axis = 1)\n",
    "np.append(A,B,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.  0.  0.  0.]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[ 0.98518129,  0.60898041,  0.40566328,  0.01670487],\n",
       "       [ 0.6450601 ,  0.85978905,  0.66902865,  0.61851746],\n",
       "       [ 0.7956447 ,  0.10383916,  0.73235019,  0.7352059 ],\n",
       "       [ 0.42563489,  0.68353725,  0.20302183,  0.48926061],\n",
       "       [ 0.        ,  0.        ,  0.        ,  0.        ]])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B = np.zeros([1,4]) # row vector\n",
    "print(B)\n",
    "\n",
    "# Append row (axis=0)\n",
    "np.append(A,B,axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing and slicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0  1  2  3]\n",
      " [ 4  5  6  7]\n",
      " [ 8  9 10 11]]\n",
      "0\n",
      "1\n",
      "7\n"
     ]
    }
   ],
   "source": [
    "A = np.array([[0,1,2,3],[4,5,6,7],[8,9,10,11]])\n",
    "print(A)\n",
    "\n",
    "# You can access the value of a certain element by indexing with its row and column\n",
    "print(A[0,0]) # e.g. first row first column [0,0]\n",
    "print(A[0,1])\n",
    "\n",
    "print(A[1,3]) # second row, forth column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 1]\n",
      "[0 1 2 3]\n"
     ]
    }
   ],
   "source": [
    "# You can also get a \"slice\" of a row \n",
    "\n",
    "print(A[0,0:2]) # first row, elements from 0 to 2 \n",
    "\n",
    "print(A[0,:]) # or whole row using \":\" as an index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1 5]\n",
      "[1 5 9]\n"
     ]
    }
   ],
   "source": [
    "# The same for columns\n",
    "print(A[0:2,1]) # second column, first two elements\n",
    "print(A[:,1]) # second column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 8  9 10 11]\n",
      "[[ 8  9 10 11]]\n",
      "[[ 8]\n",
      " [ 9]\n",
      " [10]\n",
      " [11]]\n"
     ]
    }
   ],
   "source": [
    "# In the previous examples when you access a slice of a row/column you get a 1D vector\n",
    "# You can use NumPy broadcasting if you want to get a row/column array\n",
    "print(A[2,:])\n",
    "print(A[2,:][None,:])\n",
    "print(A[2,:][:,None])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 1  2]\n",
      " [ 5  6]\n",
      " [ 9 10]]\n"
     ]
    }
   ],
   "source": [
    "# You can also access a sub-matrix of the array\n",
    "print(A[0:3,1:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0 1 2 3]\n",
      " [3 3 3 3]\n",
      " [3 3 3 3]]\n"
     ]
    }
   ],
   "source": [
    "# You can also index with conditions. \n",
    "# For example A[A>3] refers to all elements in A whose values are bigger that 3\n",
    "\n",
    "A[A>3] = 3 # This way I can, for example, \"clip\" array A to a max. value of 3\n",
    "\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Operations with arrays "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 2 3 4]]\n",
      "[[1]\n",
      " [2]\n",
      " [3]\n",
      " [4]]\n",
      "[[1]\n",
      " [2]\n",
      " [3]\n",
      " [4]]\n"
     ]
    }
   ],
   "source": [
    "# Transposed array\n",
    "A = np.array([[1,2,3,4]])\n",
    "print(A)\n",
    "\n",
    "\n",
    "print(A.transpose())\n",
    "print(A.T) # shortcut for transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 8  9 10 11]]\n",
      "[[ 3  6  9 12]]\n",
      "[[ 1  4  9 16]]\n"
     ]
    }
   ],
   "source": [
    "# Operations with scalars \n",
    "\n",
    "print(A+7)  # addition\n",
    "print(A*3) # multiplication\n",
    "print(A**2) # exponentiation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.  1.  2.  3.]]\n",
      "[[ 2.  3.  4.  5.]]\n",
      "[[ 1.  2.  3.  4.]]\n"
     ]
    }
   ],
   "source": [
    "# Element-wise operations (the shape of the operands must agree)\n",
    "B = np.ones(A.shape)\n",
    "\n",
    "print(A-B)\n",
    "print(A+B)\n",
    "print(A*B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[30]])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Matrix (dot) product\n",
    "np.dot(A,A.T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[30]])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A.dot(A.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sum of array elements'\n",
    "np.sum(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4, 1)\n"
     ]
    }
   ],
   "source": [
    "# Max./min. value\n",
    "print(np.max(A), np.min(A))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing numpy dot product efficiency\n",
    "\n",
    "Let's check the difference of using numpy built-in operations or standard python code. Remember to use numpy built-in functionality always when posible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.05165280196\n",
      "0.0888759571863\n",
      "0.0242252138965\n"
     ]
    }
   ],
   "source": [
    "import timeit #Tool for measuring execution time of small code snippets.\n",
    "\n",
    "# In standard Python the time for adding the results of 1000 multiplications is:\n",
    "print(timeit.timeit('sum(x*x for x in xrange(1000))',number=10000))\n",
    "# Using NumPy array element-wise multiplication and array sum:\n",
    "print(timeit.timeit('np.sum(a*a)',setup=\"import numpy as np; a=np.arange(1000)\", number=10000))\n",
    "# Using NumPy dot product:\n",
    "print(timeit.timeit('a.dot(a)',setup=\"import numpy as np; a=np.arange(1000)\", number=10000))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional information \n",
    "Additional information about the Numpy arrays and some exercises can be found <a href=\"https://scipy-lectures.github.io/intro/numpy/array_object.html\">here</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='red'>Exercises</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first time you open the notebook always include the modules that are going to be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Import NumPy\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Familiarize yourself with the different functions available to create numpy arrays.\n",
    "\n",
    "* Create several arrays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A = [10 11 12 13 14 15 16 17 18 19]\n",
      "B =\n",
      "[[1 2]\n",
      " [3 4]]\n",
      "C =\n",
      "[[1, 2, 3] [4, 5] [6, 7, 8]]\n"
     ]
    }
   ],
   "source": [
    "# Creating a vector (1D array)\n",
    "A = np.array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])\n",
    "print \"A =\", A\n",
    "\n",
    "# Create a two dimensional array (2x2 matrix)\n",
    "B = np.array([[1, 2], [3, 4]])\n",
    "print(\"B =\")\n",
    "print(B)\n",
    "\n",
    "# Create a two dimensional array that is not a matrix\n",
    "C = np.array([[1, 2, 3], [4, 5], [6, 7, 8]])\n",
    "print(\"C =\")\n",
    "print(C)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Use the functions **len()**, **numpy.shape()** on these arrays. How do they relate to each other? And to the **ndim** attribute of the arrays?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function **len()** returns the number of columns of array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of the first dimension (columns) for A = 10\n",
      "Size of the first dimension (columns) for B = 2\n",
      "Size of the first dimension (columns) for C = 3\n"
     ]
    }
   ],
   "source": [
    "m = \"Size of the first dimension (columns) for A =\"\n",
    "print m, len(A)\n",
    "m = \"Size of the first dimension (columns) for B =\"\n",
    "print m, len(B)\n",
    "m = \"Size of the first dimension (columns) for C =\"\n",
    "print m, len(C)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function **numpy.shape()** returns the number of dimensions and the longitude (number of values) on each dimension of the array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of A = (10L,)\n",
      "Shape of B = (2L, 2L)\n",
      "Shape of C = (3L,)\n"
     ]
    }
   ],
   "source": [
    "m = \"Shape of A =\"\n",
    "print m, A.shape\n",
    "m = \"Shape of B =\"\n",
    "print m, B.shape\n",
    "m = \"Shape of C =\"\n",
    "print m, C.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous code tells that **A** is a one dimensional array (vector) with 10 values, while **B** is a two dimensional array with 2 values in each dimension (a matrix). However **C** is an interesting case, **numpy.shape()** tells that **C** is a one dimensional array (vector) with three values, but it doesn't tell that each value is itself a new array (vector) of 3, 2, and 3 values respectively.\n",
    "\n",
    "It can be said that the **len()** and the **numpy.shape()** function are related in that both can return the longitude of the array, but the NumPy function provides more information about the array, even though sometimes is not enough to completely describe the array as can be seen with **C**.\n",
    "\n",
    "The function **ndim()** returns the dimensions of the array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensions of A = 1\n",
      "Dimensions of B = 2\n",
      "Dimensions of C = 1\n"
     ]
    }
   ],
   "source": [
    "m = \"Dimensions of A =\"\n",
    "print m, A.ndim\n",
    "m = \"Dimensions of B =\"\n",
    "print m, B.ndim\n",
    "m = \"Dimensions of C =\"\n",
    "print m, C.ndim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous code tells that **A** is a one dimensional array (vector), while **B** is a two dimensional array (a matrix). Again **C** is an interesting case, **numpy.ndim()** tells that **C** is a one dimensional array (vector), but it doesn't tell that each value is itself a new array (vector).\n",
    "\n",
    "If **ndim** is compared with **len** and **shape**, it can be seen that it is rather a specific function as the information provided is not related to longitude, but instead counts the number of values that the **shape** function returns."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* There are even more ways to create numpy arrays. Use the **help()** function to find out about **np.linspace** and **np.diag**. How do they relate to **np.arange** and **np.eye** respectively?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function linspace in module numpy.core.function_base:\n",
      "\n",
      "linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None)\n",
      "    Return evenly spaced numbers over a specified interval.\n",
      "    \n",
      "    Returns `num` evenly spaced samples, calculated over the\n",
      "    interval [`start`, `stop`].\n",
      "    \n",
      "    The endpoint of the interval can optionally be excluded.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    start : scalar\n",
      "        The starting value of the sequence.\n",
      "    stop : scalar\n",
      "        The end value of the sequence, unless `endpoint` is set to False.\n",
      "        In that case, the sequence consists of all but the last of ``num + 1``\n",
      "        evenly spaced samples, so that `stop` is excluded.  Note that the step\n",
      "        size changes when `endpoint` is False.\n",
      "    num : int, optional\n",
      "        Number of samples to generate. Default is 50. Must be non-negative.\n",
      "    endpoint : bool, optional\n",
      "        If True, `stop` is the last sample. Otherwise, it is not included.\n",
      "        Default is True.\n",
      "    retstep : bool, optional\n",
      "        If True, return (`samples`, `step`), where `step` is the spacing\n",
      "        between samples.\n",
      "    dtype : dtype, optional\n",
      "        The type of the output array.  If `dtype` is not given, infer the data\n",
      "        type from the other input arguments.\n",
      "    \n",
      "        .. versionadded:: 1.9.0\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    samples : ndarray\n",
      "        There are `num` equally spaced samples in the closed interval\n",
      "        ``[start, stop]`` or the half-open interval ``[start, stop)``\n",
      "        (depending on whether `endpoint` is True or False).\n",
      "    step : float\n",
      "        Only returned if `retstep` is True\n",
      "    \n",
      "        Size of spacing between samples.\n",
      "    \n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    arange : Similar to `linspace`, but uses a step size (instead of the\n",
      "             number of samples).\n",
      "    logspace : Samples uniformly distributed in log space.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> np.linspace(2.0, 3.0, num=5)\n",
      "        array([ 2.  ,  2.25,  2.5 ,  2.75,  3.  ])\n",
      "    >>> np.linspace(2.0, 3.0, num=5, endpoint=False)\n",
      "        array([ 2. ,  2.2,  2.4,  2.6,  2.8])\n",
      "    >>> np.linspace(2.0, 3.0, num=5, retstep=True)\n",
      "        (array([ 2.  ,  2.25,  2.5 ,  2.75,  3.  ]), 0.25)\n",
      "    \n",
      "    Graphical illustration:\n",
      "    \n",
      "    >>> import matplotlib.pyplot as plt\n",
      "    >>> N = 8\n",
      "    >>> y = np.zeros(N)\n",
      "    >>> x1 = np.linspace(0, 10, N, endpoint=True)\n",
      "    >>> x2 = np.linspace(0, 10, N, endpoint=False)\n",
      "    >>> plt.plot(x1, y, 'o')\n",
      "    [<matplotlib.lines.Line2D object at 0x...>]\n",
      "    >>> plt.plot(x2, y + 0.5, 'o')\n",
      "    [<matplotlib.lines.Line2D object at 0x...>]\n",
      "    >>> plt.ylim([-0.5, 1])\n",
      "    (-0.5, 1)\n",
      "    >>> plt.show()\n",
      "\n",
      "Help on function diag in module numpy.lib.twodim_base:\n",
      "\n",
      "diag(v, k=0)\n",
      "    Extract a diagonal or construct a diagonal array.\n",
      "    \n",
      "    See the more detailed documentation for ``numpy.diagonal`` if you use this\n",
      "    function to extract a diagonal and wish to write to the resulting array;\n",
      "    whether it returns a copy or a view depends on what version of numpy you\n",
      "    are using.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    v : array_like\n",
      "        If `v` is a 2-D array, return a copy of its `k`-th diagonal.\n",
      "        If `v` is a 1-D array, return a 2-D array with `v` on the `k`-th\n",
      "        diagonal.\n",
      "    k : int, optional\n",
      "        Diagonal in question. The default is 0. Use `k>0` for diagonals\n",
      "        above the main diagonal, and `k<0` for diagonals below the main\n",
      "        diagonal.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    out : ndarray\n",
      "        The extracted diagonal or constructed diagonal array.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    diagonal : Return specified diagonals.\n",
      "    diagflat : Create a 2-D array with the flattened input as a diagonal.\n",
      "    trace : Sum along diagonals.\n",
      "    triu : Upper triangle of an array.\n",
      "    tril : Lower triangle of an array.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> x = np.arange(9).reshape((3,3))\n",
      "    >>> x\n",
      "    array([[0, 1, 2],\n",
      "           [3, 4, 5],\n",
      "           [6, 7, 8]])\n",
      "    \n",
      "    >>> np.diag(x)\n",
      "    array([0, 4, 8])\n",
      "    >>> np.diag(x, k=1)\n",
      "    array([1, 5])\n",
      "    >>> np.diag(x, k=-1)\n",
      "    array([3, 7])\n",
      "    \n",
      "    >>> np.diag(np.diag(x))\n",
      "    array([[0, 0, 0],\n",
      "           [0, 4, 0],\n",
      "           [0, 0, 8]])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(np.linspace)\n",
    "help(np.diag)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**np.linspace** creates a linear space with evenly spaced numbers over a specified interval. It relates to **np.arange** in that both can create a vector with the same characteristics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 30.  31.  32.  33.  34.  35.  36.  37.  38.  39.]\n",
      "[ 40.          40.20408163  40.40816327  40.6122449   40.81632653\n",
      "  41.02040816  41.2244898   41.42857143  41.63265306  41.83673469\n",
      "  42.04081633  42.24489796  42.44897959  42.65306122  42.85714286\n",
      "  43.06122449  43.26530612  43.46938776  43.67346939  43.87755102\n",
      "  44.08163265  44.28571429  44.48979592  44.69387755  44.89795918\n",
      "  45.10204082  45.30612245  45.51020408  45.71428571  45.91836735\n",
      "  46.12244898  46.32653061  46.53061224  46.73469388  46.93877551\n",
      "  47.14285714  47.34693878  47.55102041  47.75510204  47.95918367\n",
      "  48.16326531  48.36734694  48.57142857  48.7755102   48.97959184\n",
      "  49.18367347  49.3877551   49.59183673  49.79591837  50.        ]\n"
     ]
    }
   ],
   "source": [
    "print(np.linspace(30, 39, 10))\n",
    "print(np.linspace(40, 50))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**np.diag** extracts the diagonal or constructs a diagonal array. With the **k** paramater, all the diagonals can be chosen being **k=0** the main diagonal. It relates to **np.eye** in that both can create a diagonal matrix, however, **np.diag** can also create a diagonal with other values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x =\n",
      "[[0 1 2]\n",
      " [3 4 5]\n",
      " [6 7 8]]\n",
      "np.diag(x) returns: [0 4 8]\n",
      "np.diag(x, k=1) returns: [1 5]\n",
      "np.diag(x, k=-1) returns: [3 7]\n",
      "np.diag([1, 2, 3]) returns:\n",
      "[[1 0 0]\n",
      " [0 2 0]\n",
      " [0 0 3]]\n"
     ]
    }
   ],
   "source": [
    "x = np.arange(9).reshape((3, 3))\n",
    "print \"x =\"\n",
    "print(x)\n",
    "print \"np.diag(x) returns:\", np.diag(x)\n",
    "print \"np.diag(x, k=1) returns:\", np.diag(x, k=1)\n",
    "print \"np.diag(x, k=-1) returns:\", np.diag(x, k=-1)\n",
    "print \"np.diag([1, 2, 3]) returns:\"\n",
    "print(np.diag([1, 2, 3]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* What is the output of the **dtype** attribute of numpy arrays? Is it possible to specify a data type when we create an array? Check the **np.array** documentation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(dtype('int32'), dtype('float64'))"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = np.array([1, 2, 3])\n",
    "B = np.array([1.0, 2.0, 3.0])\n",
    "A.dtype,B.dtype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output of **dtype** is the data type of the array. It can be given as an **np.array** parameter to stablish the desired data-type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 4.  5.  6.]\n"
     ]
    }
   ],
   "source": [
    "C = np.array([4, 5, 6], dtype=float)\n",
    "print(C)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the **numpy.append** function array dimensions must agree. The following code will rise an error.  Write code to reshape B and append it to A."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function append in module numpy.lib.function_base:\n",
      "\n",
      "append(arr, values, axis=None)\n",
      "    Append values to the end of an array.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    arr : array_like\n",
      "        Values are appended to a copy of this array.\n",
      "    values : array_like\n",
      "        These values are appended to a copy of `arr`.  It must be of the\n",
      "        correct shape (the same shape as `arr`, excluding `axis`).  If\n",
      "        `axis` is not specified, `values` can be any shape and will be\n",
      "        flattened before use.\n",
      "    axis : int, optional\n",
      "        The axis along which `values` are appended.  If `axis` is not\n",
      "        given, both `arr` and `values` are flattened before use.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    append : ndarray\n",
      "        A copy of `arr` with `values` appended to `axis`.  Note that\n",
      "        `append` does not occur in-place: a new array is allocated and\n",
      "        filled.  If `axis` is None, `out` is a flattened array.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    insert : Insert elements into an array.\n",
      "    delete : Delete elements from an array.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> np.append([1, 2, 3], [[4, 5, 6], [7, 8, 9]])\n",
      "    array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
      "    \n",
      "    When `axis` is specified, `values` must have the correct shape.\n",
      "    \n",
      "    >>> np.append([[1, 2, 3], [4, 5, 6]], [[7, 8, 9]], axis=0)\n",
      "    array([[1, 2, 3],\n",
      "           [4, 5, 6],\n",
      "           [7, 8, 9]])\n",
      "    >>> np.append([[1, 2, 3], [4, 5, 6]], [7, 8, 9], axis=0)\n",
      "    Traceback (most recent call last):\n",
      "    ...\n",
      "    ValueError: arrays must have same number of dimensions\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(np.append)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 8.  6.  5.  4.  0.]\n",
      " [ 9.  4.  4.  2.  0.]\n",
      " [ 2.  9.  5.  0.  0.]\n",
      " [ 7.  4.  4.  3.  0.]]\n"
     ]
    }
   ],
   "source": [
    "A = np.random.randint(10,size=(4,4))\n",
    "B = np.zeros([2,2]).reshape([4,1])\n",
    "# Append column\n",
    "A = np.append(A,B,axis=1)\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Same as before, the following code rise an error. Write code to append a slice of vector C with the correct dimension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 8.  6.  5.  4.  0.]\n",
      " [ 9.  4.  4.  2.  0.]\n",
      " [ 2.  9.  5.  0.  0.]\n",
      " [ 7.  4.  4.  3.  0.]\n",
      " [ 0.  0.  0.  0.  0.]]\n",
      "[[ 1.  0.  9.  6.  0.]\n",
      " [ 4.  1.  5.  2.  0.]\n",
      " [ 7.  4.  1.  4.  0.]\n",
      " [ 7.  2.  2.  0.  0.]]\n"
     ]
    }
   ],
   "source": [
    "C = np.zeros([5,1]).reshape([1,5])\n",
    "# Append row\n",
    "print(np.append(A,C,axis=0))\n",
    "\n",
    "# Or\n",
    "A = np.random.randint(10,size=(4,4))\n",
    "C = np.zeros([4,1])\n",
    "# Append column\n",
    "print(np.append(A,C,axis=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to SciPy\n",
    "\n",
    "Scipy offers a variety of numerical algorithms that build on top of the efficient NumPy arrays. The diverse algorithms are grouped into different modules. To cite some of them:\n",
    "\n",
    "* **scipy.stats** Statistical functions\n",
    "\n",
    "* **scipy.cluster** Hierarchical clustering (cluster.hierarchy), Vector quantization / K-Means (cluster.vq)\n",
    "\n",
    "* **scipy.optimize** Optimization algorithms\n",
    "\n",
    "Appart of these modules there exist multiple scikits, scientific toolboxes build around SciPy. You can find a list of available scikits <a href=\"https://scikits.appspot.com/scikits\">here</a>. From wich the most related to Pattern Recognition is scikit-learn.\n",
    "\n",
    "* **scikit-learn** Machine Learning in Python: Simple and efficient tools for data mining and data analysis.\n",
    "\n",
    "Apart from scikit-learn, another popular scikit is scikit-image.\n",
    "\n",
    "Since all these modules are very specialized we are not going into details in this introduction. As always, remember you can get help on specific modules with **help()**. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scipy.stats in scipy:\n",
      "\n",
      "NAME\n",
      "    scipy.stats\n",
      "\n",
      "FILE\n",
      "    d:\\users\\byron\\anaconda2\\lib\\site-packages\\scipy\\stats\\__init__.py\n",
      "\n",
      "DESCRIPTION\n",
      "    ==========================================\n",
      "    Statistical functions (:mod:`scipy.stats`)\n",
      "    ==========================================\n",
      "    \n",
      "    .. module:: scipy.stats\n",
      "    \n",
      "    This module contains a large number of probability distributions as\n",
      "    well as a growing library of statistical functions.\n",
      "    \n",
      "    Each univariate distribution is an instance of a subclass of `rv_continuous`\n",
      "    (`rv_discrete` for discrete distributions):\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       rv_continuous\n",
      "       rv_discrete\n",
      "    \n",
      "    Continuous distributions\n",
      "    ========================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       alpha             -- Alpha\n",
      "       anglit            -- Anglit\n",
      "       arcsine           -- Arcsine\n",
      "       beta              -- Beta\n",
      "       betaprime         -- Beta Prime\n",
      "       bradford          -- Bradford\n",
      "       burr              -- Burr (Type III)\n",
      "       burr12            -- Burr (Type XII)\n",
      "       cauchy            -- Cauchy\n",
      "       chi               -- Chi\n",
      "       chi2              -- Chi-squared\n",
      "       cosine            -- Cosine\n",
      "       dgamma            -- Double Gamma\n",
      "       dweibull          -- Double Weibull\n",
      "       erlang            -- Erlang\n",
      "       expon             -- Exponential\n",
      "       exponnorm         -- Exponentially Modified Normal\n",
      "       exponweib         -- Exponentiated Weibull\n",
      "       exponpow          -- Exponential Power\n",
      "       f                 -- F (Snecdor F)\n",
      "       fatiguelife       -- Fatigue Life (Birnbaum-Saunders)\n",
      "       fisk              -- Fisk\n",
      "       foldcauchy        -- Folded Cauchy\n",
      "       foldnorm          -- Folded Normal\n",
      "       frechet_r         -- Frechet Right Sided, Extreme Value Type II (Extreme LB) or weibull_min\n",
      "       frechet_l         -- Frechet Left Sided, Weibull_max\n",
      "       genlogistic       -- Generalized Logistic\n",
      "       gennorm           -- Generalized normal\n",
      "       genpareto         -- Generalized Pareto\n",
      "       genexpon          -- Generalized Exponential\n",
      "       genextreme        -- Generalized Extreme Value\n",
      "       gausshyper        -- Gauss Hypergeometric\n",
      "       gamma             -- Gamma\n",
      "       gengamma          -- Generalized gamma\n",
      "       genhalflogistic   -- Generalized Half Logistic\n",
      "       gilbrat           -- Gilbrat\n",
      "       gompertz          -- Gompertz (Truncated Gumbel)\n",
      "       gumbel_r          -- Right Sided Gumbel, Log-Weibull, Fisher-Tippett, Extreme Value Type I\n",
      "       gumbel_l          -- Left Sided Gumbel, etc.\n",
      "       halfcauchy        -- Half Cauchy\n",
      "       halflogistic      -- Half Logistic\n",
      "       halfnorm          -- Half Normal\n",
      "       halfgennorm       -- Generalized Half Normal\n",
      "       hypsecant         -- Hyperbolic Secant\n",
      "       invgamma          -- Inverse Gamma\n",
      "       invgauss          -- Inverse Gaussian\n",
      "       invweibull        -- Inverse Weibull\n",
      "       johnsonsb         -- Johnson SB\n",
      "       johnsonsu         -- Johnson SU\n",
      "       kappa4            -- Kappa 4 parameter\n",
      "       kappa3            -- Kappa 3 parameter\n",
      "       ksone             -- Kolmogorov-Smirnov one-sided (no stats)\n",
      "       kstwobign         -- Kolmogorov-Smirnov two-sided test for Large N (no stats)\n",
      "       laplace           -- Laplace\n",
      "       levy              -- Levy\n",
      "       levy_l\n",
      "       levy_stable\n",
      "       logistic          -- Logistic\n",
      "       loggamma          -- Log-Gamma\n",
      "       loglaplace        -- Log-Laplace (Log Double Exponential)\n",
      "       lognorm           -- Log-Normal\n",
      "       lomax             -- Lomax (Pareto of the second kind)\n",
      "       maxwell           -- Maxwell\n",
      "       mielke            -- Mielke's Beta-Kappa\n",
      "       nakagami          -- Nakagami\n",
      "       ncx2              -- Non-central chi-squared\n",
      "       ncf               -- Non-central F\n",
      "       nct               -- Non-central Student's T\n",
      "       norm              -- Normal (Gaussian)\n",
      "       pareto            -- Pareto\n",
      "       pearson3          -- Pearson type III\n",
      "       powerlaw          -- Power-function\n",
      "       powerlognorm      -- Power log normal\n",
      "       powernorm         -- Power normal\n",
      "       rdist             -- R-distribution\n",
      "       reciprocal        -- Reciprocal\n",
      "       rayleigh          -- Rayleigh\n",
      "       rice              -- Rice\n",
      "       recipinvgauss     -- Reciprocal Inverse Gaussian\n",
      "       semicircular      -- Semicircular\n",
      "       skewnorm          -- Skew normal\n",
      "       t                 -- Student's T\n",
      "       trapz              -- Trapezoidal\n",
      "       triang            -- Triangular\n",
      "       truncexpon        -- Truncated Exponential\n",
      "       truncnorm         -- Truncated Normal\n",
      "       tukeylambda       -- Tukey-Lambda\n",
      "       uniform           -- Uniform\n",
      "       vonmises          -- Von-Mises (Circular)\n",
      "       vonmises_line     -- Von-Mises (Line)\n",
      "       wald              -- Wald\n",
      "       weibull_min       -- Minimum Weibull (see Frechet)\n",
      "       weibull_max       -- Maximum Weibull (see Frechet)\n",
      "       wrapcauchy        -- Wrapped Cauchy\n",
      "    \n",
      "    Multivariate distributions\n",
      "    ==========================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       multivariate_normal   -- Multivariate normal distribution\n",
      "       matrix_normal         -- Matrix normal distribution\n",
      "       dirichlet             -- Dirichlet\n",
      "       wishart               -- Wishart\n",
      "       invwishart            -- Inverse Wishart\n",
      "       special_ortho_group   -- SO(N) group\n",
      "       ortho_group           -- O(N) group\n",
      "       random_correlation    -- random correlation matrices\n",
      "    \n",
      "    Discrete distributions\n",
      "    ======================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       bernoulli         -- Bernoulli\n",
      "       binom             -- Binomial\n",
      "       boltzmann         -- Boltzmann (Truncated Discrete Exponential)\n",
      "       dlaplace          -- Discrete Laplacian\n",
      "       geom              -- Geometric\n",
      "       hypergeom         -- Hypergeometric\n",
      "       logser            -- Logarithmic (Log-Series, Series)\n",
      "       nbinom            -- Negative Binomial\n",
      "       planck            -- Planck (Discrete Exponential)\n",
      "       poisson           -- Poisson\n",
      "       randint           -- Discrete Uniform\n",
      "       skellam           -- Skellam\n",
      "       zipf              -- Zipf\n",
      "    \n",
      "    Statistical functions\n",
      "    =====================\n",
      "    \n",
      "    Several of these functions have a similar version in scipy.stats.mstats\n",
      "    which work for masked arrays.\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       describe          -- Descriptive statistics\n",
      "       gmean             -- Geometric mean\n",
      "       hmean             -- Harmonic mean\n",
      "       kurtosis          -- Fisher or Pearson kurtosis\n",
      "       kurtosistest      --\n",
      "       mode              -- Modal value\n",
      "       moment            -- Central moment\n",
      "       normaltest        --\n",
      "       skew              -- Skewness\n",
      "       skewtest          --\n",
      "       kstat             --\n",
      "       kstatvar          --\n",
      "       tmean             -- Truncated arithmetic mean\n",
      "       tvar              -- Truncated variance\n",
      "       tmin              --\n",
      "       tmax              --\n",
      "       tstd              --\n",
      "       tsem              --\n",
      "       variation         -- Coefficient of variation\n",
      "       find_repeats\n",
      "       trim_mean\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       cumfreq\n",
      "       histogram2\n",
      "       histogram\n",
      "       itemfreq\n",
      "       percentileofscore\n",
      "       scoreatpercentile\n",
      "       relfreq\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       binned_statistic     -- Compute a binned statistic for a set of data.\n",
      "       binned_statistic_2d  -- Compute a 2-D binned statistic for a set of data.\n",
      "       binned_statistic_dd  -- Compute a d-D binned statistic for a set of data.\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       obrientransform\n",
      "       signaltonoise\n",
      "       bayes_mvs\n",
      "       mvsdist\n",
      "       sem\n",
      "       zmap\n",
      "       zscore\n",
      "       iqr\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       sigmaclip\n",
      "       threshold\n",
      "       trimboth\n",
      "       trim1\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       f_oneway\n",
      "       pearsonr\n",
      "       spearmanr\n",
      "       pointbiserialr\n",
      "       kendalltau\n",
      "       linregress\n",
      "       theilslopes\n",
      "       f_value\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       ttest_1samp\n",
      "       ttest_ind\n",
      "       ttest_ind_from_stats\n",
      "       ttest_rel\n",
      "       kstest\n",
      "       chisquare\n",
      "       power_divergence\n",
      "       ks_2samp\n",
      "       mannwhitneyu\n",
      "       tiecorrect\n",
      "       rankdata\n",
      "       ranksums\n",
      "       wilcoxon\n",
      "       kruskal\n",
      "       friedmanchisquare\n",
      "       combine_pvalues\n",
      "       ss\n",
      "       square_of_sums\n",
      "       jarque_bera\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       ansari\n",
      "       bartlett\n",
      "       levene\n",
      "       shapiro\n",
      "       anderson\n",
      "       anderson_ksamp\n",
      "       binom_test\n",
      "       fligner\n",
      "       median_test\n",
      "       mood\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       boxcox\n",
      "       boxcox_normmax\n",
      "       boxcox_llf\n",
      "    \n",
      "       entropy\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       chisqprob\n",
      "       betai\n",
      "    \n",
      "    Circular statistical functions\n",
      "    ==============================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       circmean\n",
      "       circvar\n",
      "       circstd\n",
      "    \n",
      "    Contingency table functions\n",
      "    ===========================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       chi2_contingency\n",
      "       contingency.expected_freq\n",
      "       contingency.margins\n",
      "       fisher_exact\n",
      "    \n",
      "    Plot-tests\n",
      "    ==========\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       ppcc_max\n",
      "       ppcc_plot\n",
      "       probplot\n",
      "       boxcox_normplot\n",
      "    \n",
      "    \n",
      "    Masked statistics functions\n",
      "    ===========================\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       stats.mstats\n",
      "    \n",
      "    \n",
      "    Univariate and multivariate kernel density estimation (:mod:`scipy.stats.kde`)\n",
      "    ==============================================================================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       gaussian_kde\n",
      "    \n",
      "    For many more stat related functions install the software R and the\n",
      "    interface package rpy.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _binned_statistic\n",
      "    _constants\n",
      "    _continuous_distns\n",
      "    _discrete_distns\n",
      "    _distn_infrastructure\n",
      "    _distr_params\n",
      "    _multivariate\n",
      "    _stats\n",
      "    _stats_mstats_common\n",
      "    _tukeylambda_stats\n",
      "    contingency\n",
      "    distributions\n",
      "    kde\n",
      "    morestats\n",
      "    mstats\n",
      "    mstats_basic\n",
      "    mstats_extras\n",
      "    mvn\n",
      "    setup\n",
      "    statlib\n",
      "    stats\n",
      "    vonmises\n",
      "\n",
      "CLASSES\n",
      "    __builtin__.object\n",
      "        scipy.stats.kde.gaussian_kde\n",
      "    scipy.stats._distn_infrastructure.rv_generic(__builtin__.object)\n",
      "        scipy.stats._distn_infrastructure.rv_continuous\n",
      "        scipy.stats._distn_infrastructure.rv_discrete\n",
      "    \n",
      "    class gaussian_kde(__builtin__.object)\n",
      "     |  Representation of a kernel-density estimate using Gaussian kernels.\n",
      "     |  \n",
      "     |  Kernel density estimation is a way to estimate the probability density\n",
      "     |  function (PDF) of a random variable in a non-parametric way.\n",
      "     |  `gaussian_kde` works for both uni-variate and multi-variate data.   It\n",
      "     |  includes automatic bandwidth determination.  The estimation works best for\n",
      "     |  a unimodal distribution; bimodal or multi-modal distributions tend to be\n",
      "     |  oversmoothed.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  dataset : array_like\n",
      "     |      Datapoints to estimate from. In case of univariate data this is a 1-D\n",
      "     |      array, otherwise a 2-D array with shape (# of dims, # of data).\n",
      "     |  bw_method : str, scalar or callable, optional\n",
      "     |      The method used to calculate the estimator bandwidth.  This can be\n",
      "     |      'scott', 'silverman', a scalar constant or a callable.  If a scalar,\n",
      "     |      this will be used directly as `kde.factor`.  If a callable, it should\n",
      "     |      take a `gaussian_kde` instance as only parameter and return a scalar.\n",
      "     |      If None (default), 'scott' is used.  See Notes for more details.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  dataset : ndarray\n",
      "     |      The dataset with which `gaussian_kde` was initialized.\n",
      "     |  d : int\n",
      "     |      Number of dimensions.\n",
      "     |  n : int\n",
      "     |      Number of datapoints.\n",
      "     |  factor : float\n",
      "     |      The bandwidth factor, obtained from `kde.covariance_factor`, with which\n",
      "     |      the covariance matrix is multiplied.\n",
      "     |  covariance : ndarray\n",
      "     |      The covariance matrix of `dataset`, scaled by the calculated bandwidth\n",
      "     |      (`kde.factor`).\n",
      "     |  inv_cov : ndarray\n",
      "     |      The inverse of `covariance`.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  evaluate\n",
      "     |  __call__\n",
      "     |  integrate_gaussian\n",
      "     |  integrate_box_1d\n",
      "     |  integrate_box\n",
      "     |  integrate_kde\n",
      "     |  pdf\n",
      "     |  logpdf\n",
      "     |  resample\n",
      "     |  set_bandwidth\n",
      "     |  covariance_factor\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Bandwidth selection strongly influences the estimate obtained from the KDE\n",
      "     |  (much more so than the actual shape of the kernel).  Bandwidth selection\n",
      "     |  can be done by a \"rule of thumb\", by cross-validation, by \"plug-in\n",
      "     |  methods\" or by other means; see [3]_, [4]_ for reviews.  `gaussian_kde`\n",
      "     |  uses a rule of thumb, the default is Scott's Rule.\n",
      "     |  \n",
      "     |  Scott's Rule [1]_, implemented as `scotts_factor`, is::\n",
      "     |  \n",
      "     |      n**(-1./(d+4)),\n",
      "     |  \n",
      "     |  with ``n`` the number of data points and ``d`` the number of dimensions.\n",
      "     |  Silverman's Rule [2]_, implemented as `silverman_factor`, is::\n",
      "     |  \n",
      "     |      (n * (d + 2) / 4.)**(-1. / (d + 4)).\n",
      "     |  \n",
      "     |  Good general descriptions of kernel density estimation can be found in [1]_\n",
      "     |  and [2]_, the mathematics for this multi-dimensional implementation can be\n",
      "     |  found in [1]_.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] D.W. Scott, \"Multivariate Density Estimation: Theory, Practice, and\n",
      "     |         Visualization\", John Wiley & Sons, New York, Chicester, 1992.\n",
      "     |  .. [2] B.W. Silverman, \"Density Estimation for Statistics and Data\n",
      "     |         Analysis\", Vol. 26, Monographs on Statistics and Applied Probability,\n",
      "     |         Chapman and Hall, London, 1986.\n",
      "     |  .. [3] B.A. Turlach, \"Bandwidth Selection in Kernel Density Estimation: A\n",
      "     |         Review\", CORE and Institut de Statistique, Vol. 19, pp. 1-33, 1993.\n",
      "     |  .. [4] D.M. Bashtannyk and R.J. Hyndman, \"Bandwidth selection for kernel\n",
      "     |         conditional density estimation\", Computational Statistics & Data\n",
      "     |         Analysis, Vol. 36, pp. 279-298, 2001.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Generate some random two-dimensional data:\n",
      "     |  \n",
      "     |  >>> from scipy import stats\n",
      "     |  >>> def measure(n):\n",
      "     |  ...     \"Measurement model, return two coupled measurements.\"\n",
      "     |  ...     m1 = np.random.normal(size=n)\n",
      "     |  ...     m2 = np.random.normal(scale=0.5, size=n)\n",
      "     |  ...     return m1+m2, m1-m2\n",
      "     |  \n",
      "     |  >>> m1, m2 = measure(2000)\n",
      "     |  >>> xmin = m1.min()\n",
      "     |  >>> xmax = m1.max()\n",
      "     |  >>> ymin = m2.min()\n",
      "     |  >>> ymax = m2.max()\n",
      "     |  \n",
      "     |  Perform a kernel density estimate on the data:\n",
      "     |  \n",
      "     |  >>> X, Y = np.mgrid[xmin:xmax:100j, ymin:ymax:100j]\n",
      "     |  >>> positions = np.vstack([X.ravel(), Y.ravel()])\n",
      "     |  >>> values = np.vstack([m1, m2])\n",
      "     |  >>> kernel = stats.gaussian_kde(values)\n",
      "     |  >>> Z = np.reshape(kernel(positions).T, X.shape)\n",
      "     |  \n",
      "     |  Plot the results:\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig, ax = plt.subplots()\n",
      "     |  >>> ax.imshow(np.rot90(Z), cmap=plt.cm.gist_earth_r,\n",
      "     |  ...           extent=[xmin, xmax, ymin, ymax])\n",
      "     |  >>> ax.plot(m1, m2, 'k.', markersize=2)\n",
      "     |  >>> ax.set_xlim([xmin, xmax])\n",
      "     |  >>> ax.set_ylim([ymin, ymax])\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__ = evaluate(self, points)\n",
      "     |  \n",
      "     |  __init__(self, dataset, bw_method=None)\n",
      "     |  \n",
      "     |  covariance_factor = scotts_factor(self)\n",
      "     |  \n",
      "     |  evaluate(self, points)\n",
      "     |      Evaluate the estimated pdf on a set of points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      points : (# of dimensions, # of points)-array\n",
      "     |          Alternatively, a (# of dimensions,) vector can be passed in and\n",
      "     |          treated as a single point.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      values : (# of points,)-array\n",
      "     |          The values at each point.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError : if the dimensionality of the input points is different than\n",
      "     |                   the dimensionality of the KDE.\n",
      "     |  \n",
      "     |  integrate_box(self, low_bounds, high_bounds, maxpts=None)\n",
      "     |      Computes the integral of a pdf over a rectangular interval.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      low_bounds : array_like\n",
      "     |          A 1-D array containing the lower bounds of integration.\n",
      "     |      high_bounds : array_like\n",
      "     |          A 1-D array containing the upper bounds of integration.\n",
      "     |      maxpts : int, optional\n",
      "     |          The maximum number of points to use for integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      value : scalar\n",
      "     |          The result of the integral.\n",
      "     |  \n",
      "     |  integrate_box_1d(self, low, high)\n",
      "     |      Computes the integral of a 1D pdf between two bounds.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      low : scalar\n",
      "     |          Lower bound of integration.\n",
      "     |      high : scalar\n",
      "     |          Upper bound of integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      value : scalar\n",
      "     |          The result of the integral.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError\n",
      "     |          If the KDE is over more than one dimension.\n",
      "     |  \n",
      "     |  integrate_gaussian(self, mean, cov)\n",
      "     |      Multiply estimated density by a multivariate Gaussian and integrate\n",
      "     |      over the whole space.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      mean : aray_like\n",
      "     |          A 1-D array, specifying the mean of the Gaussian.\n",
      "     |      cov : array_like\n",
      "     |          A 2-D array, specifying the covariance matrix of the Gaussian.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      result : scalar\n",
      "     |          The value of the integral.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError\n",
      "     |          If the mean or covariance of the input Gaussian differs from\n",
      "     |          the KDE's dimensionality.\n",
      "     |  \n",
      "     |  integrate_kde(self, other)\n",
      "     |      Computes the integral of the product of this  kernel density estimate\n",
      "     |      with another.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      other : gaussian_kde instance\n",
      "     |          The other kde.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      value : scalar\n",
      "     |          The result of the integral.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError\n",
      "     |          If the KDEs have different dimensionality.\n",
      "     |  \n",
      "     |  logpdf(self, x)\n",
      "     |      Evaluate the log of the estimated pdf on a provided set of points.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      See `gaussian_kde.evaluate` for more details; this method simply\n",
      "     |      returns ``np.log(gaussian_kde.evaluate(x))``.\n",
      "     |  \n",
      "     |  pdf(self, x)\n",
      "     |      Evaluate the estimated pdf on a provided set of points.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is an alias for `gaussian_kde.evaluate`.  See the ``evaluate``\n",
      "     |      docstring for more details.\n",
      "     |  \n",
      "     |  resample(self, size=None)\n",
      "     |      Randomly sample a dataset from the estimated pdf.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      size : int, optional\n",
      "     |          The number of samples to draw.  If not provided, then the size is\n",
      "     |          the same as the underlying dataset.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      resample : (self.d, `size`) ndarray\n",
      "     |          The sampled dataset.\n",
      "     |  \n",
      "     |  scotts_factor(self)\n",
      "     |      Computes the coefficient (`kde.factor`) that\n",
      "     |      multiplies the data covariance matrix to obtain the kernel covariance\n",
      "     |      matrix. The default is `scotts_factor`.  A subclass can overwrite this\n",
      "     |      method to provide a different method, or set it through a call to\n",
      "     |      `kde.set_bandwidth`.\n",
      "     |  \n",
      "     |  set_bandwidth(self, bw_method=None)\n",
      "     |      Compute the estimator bandwidth with given method.\n",
      "     |      \n",
      "     |      The new bandwidth calculated after a call to `set_bandwidth` is used\n",
      "     |      for subsequent evaluations of the estimated density.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bw_method : str, scalar or callable, optional\n",
      "     |          The method used to calculate the estimator bandwidth.  This can be\n",
      "     |          'scott', 'silverman', a scalar constant or a callable.  If a\n",
      "     |          scalar, this will be used directly as `kde.factor`.  If a callable,\n",
      "     |          it should take a `gaussian_kde` instance as only parameter and\n",
      "     |          return a scalar.  If None (default), nothing happens; the current\n",
      "     |          `kde.covariance_factor` method is kept.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      .. versionadded:: 0.11\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> import scipy.stats as stats\n",
      "     |      >>> x1 = np.array([-7, -5, 1, 4, 5.])\n",
      "     |      >>> kde = stats.gaussian_kde(x1)\n",
      "     |      >>> xs = np.linspace(-10, 10, num=50)\n",
      "     |      >>> y1 = kde(xs)\n",
      "     |      >>> kde.set_bandwidth(bw_method='silverman')\n",
      "     |      >>> y2 = kde(xs)\n",
      "     |      >>> kde.set_bandwidth(bw_method=kde.factor / 3.)\n",
      "     |      >>> y3 = kde(xs)\n",
      "     |      \n",
      "     |      >>> import matplotlib.pyplot as plt\n",
      "     |      >>> fig, ax = plt.subplots()\n",
      "     |      >>> ax.plot(x1, np.ones(x1.shape) / (4. * x1.size), 'bo',\n",
      "     |      ...         label='Data points (rescaled)')\n",
      "     |      >>> ax.plot(xs, y1, label='Scott (default)')\n",
      "     |      >>> ax.plot(xs, y2, label='Silverman')\n",
      "     |      >>> ax.plot(xs, y3, label='Const (1/3 * Silverman)')\n",
      "     |      >>> ax.legend()\n",
      "     |      >>> plt.show()\n",
      "     |  \n",
      "     |  silverman_factor(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class rv_continuous(rv_generic)\n",
      "     |  A generic continuous random variable class meant for subclassing.\n",
      "     |  \n",
      "     |  `rv_continuous` is a base class to construct specific distribution classes\n",
      "     |  and instances for continuous random variables. It cannot be used\n",
      "     |  directly as a distribution.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  momtype : int, optional\n",
      "     |      The type of generic moment calculation to use: 0 for pdf, 1 (default)\n",
      "     |      for ppf.\n",
      "     |  a : float, optional\n",
      "     |      Lower bound of the support of the distribution, default is minus\n",
      "     |      infinity.\n",
      "     |  b : float, optional\n",
      "     |      Upper bound of the support of the distribution, default is plus\n",
      "     |      infinity.\n",
      "     |  xtol : float, optional\n",
      "     |      The tolerance for fixed point calculation for generic ppf.\n",
      "     |  badvalue : float, optional\n",
      "     |      The value in a result arrays that indicates a value that for which\n",
      "     |      some argument restriction is violated, default is np.nan.\n",
      "     |  name : str, optional\n",
      "     |      The name of the instance. This string is used to construct the default\n",
      "     |      example for distributions.\n",
      "     |  longname : str, optional\n",
      "     |      This string is used as part of the first line of the docstring returned\n",
      "     |      when a subclass has no docstring of its own. Note: `longname` exists\n",
      "     |      for backwards compatibility, do not use for new subclasses.\n",
      "     |  shapes : str, optional\n",
      "     |      The shape of the distribution. For example ``\"m, n\"`` for a\n",
      "     |      distribution that takes two integers as the two shape arguments for all\n",
      "     |      its methods. If not provided, shape parameters will be inferred from\n",
      "     |      the signature of the private methods, ``_pdf`` and ``_cdf`` of the\n",
      "     |      instance.\n",
      "     |  extradoc :  str, optional, deprecated\n",
      "     |      This string is used as the last part of the docstring returned when a\n",
      "     |      subclass has no docstring of its own. Note: `extradoc` exists for\n",
      "     |      backwards compatibility, do not use for new subclasses.\n",
      "     |  seed : None or int or ``numpy.random.RandomState`` instance, optional\n",
      "     |      This parameter defines the RandomState object to use for drawing\n",
      "     |      random variates.\n",
      "     |      If None (or np.random), the global np.random state is used.\n",
      "     |      If integer, it is used to seed the local RandomState instance.\n",
      "     |      Default is None.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  rvs\n",
      "     |  pdf\n",
      "     |  logpdf\n",
      "     |  cdf\n",
      "     |  logcdf\n",
      "     |  sf\n",
      "     |  logsf\n",
      "     |  ppf\n",
      "     |  isf\n",
      "     |  moment\n",
      "     |  stats\n",
      "     |  entropy\n",
      "     |  expect\n",
      "     |  median\n",
      "     |  mean\n",
      "     |  std\n",
      "     |  var\n",
      "     |  interval\n",
      "     |  __call__\n",
      "     |  fit\n",
      "     |  fit_loc_scale\n",
      "     |  nnlf\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Public methods of an instance of a distribution class (e.g., ``pdf``,\n",
      "     |  ``cdf``) check their arguments and pass valid arguments to private,\n",
      "     |  computational methods (``_pdf``, ``_cdf``). For ``pdf(x)``, ``x`` is valid\n",
      "     |  if it is within the support of a distribution, ``self.a <= x <= self.b``.\n",
      "     |  Whether a shape parameter is valid is decided by an ``_argcheck`` method\n",
      "     |  (which defaults to checking that its arguments are strictly positive.)\n",
      "     |  \n",
      "     |  **Subclassing**\n",
      "     |  \n",
      "     |  New random variables can be defined by subclassing the `rv_continuous` class\n",
      "     |  and re-defining at least the ``_pdf`` or the ``_cdf`` method (normalized\n",
      "     |  to location 0 and scale 1).\n",
      "     |  \n",
      "     |  If positive argument checking is not correct for your RV\n",
      "     |  then you will also need to re-define the ``_argcheck`` method.\n",
      "     |  \n",
      "     |  Correct, but potentially slow defaults exist for the remaining\n",
      "     |  methods but for speed and/or accuracy you can over-ride::\n",
      "     |  \n",
      "     |    _logpdf, _cdf, _logcdf, _ppf, _rvs, _isf, _sf, _logsf\n",
      "     |  \n",
      "     |  Rarely would you override ``_isf``, ``_sf`` or ``_logsf``, but you could.\n",
      "     |  \n",
      "     |  **Methods that can be overwritten by subclasses**\n",
      "     |  ::\n",
      "     |  \n",
      "     |    _rvs\n",
      "     |    _pdf\n",
      "     |    _cdf\n",
      "     |    _sf\n",
      "     |    _ppf\n",
      "     |    _isf\n",
      "     |    _stats\n",
      "     |    _munp\n",
      "     |    _entropy\n",
      "     |    _argcheck\n",
      "     |  \n",
      "     |  There are additional (internal and private) generic methods that can\n",
      "     |  be useful for cross-checking and for debugging, but might work in all\n",
      "     |  cases when directly called.\n",
      "     |  \n",
      "     |  A note on ``shapes``: subclasses need not specify them explicitly. In this\n",
      "     |  case, `shapes` will be automatically deduced from the signatures of the\n",
      "     |  overridden methods (`pdf`, `cdf` etc).\n",
      "     |  If, for some reason, you prefer to avoid relying on introspection, you can\n",
      "     |  specify ``shapes`` explicitly as an argument to the instance constructor.\n",
      "     |  \n",
      "     |  \n",
      "     |  **Frozen Distributions**\n",
      "     |  \n",
      "     |  Normally, you must provide shape parameters (and, optionally, location and\n",
      "     |  scale parameters to each call of a method of a distribution.\n",
      "     |  \n",
      "     |  Alternatively, the object may be called (as a function) to fix the shape,\n",
      "     |  location, and scale parameters returning a \"frozen\" continuous RV object:\n",
      "     |  \n",
      "     |  rv = generic(<shape(s)>, loc=0, scale=1)\n",
      "     |      frozen RV object with the same methods but holding the given shape,\n",
      "     |      location, and scale fixed\n",
      "     |  \n",
      "     |  **Statistics**\n",
      "     |  \n",
      "     |  Statistics are computed using numerical integration by default.\n",
      "     |  For speed you can redefine this using ``_stats``:\n",
      "     |  \n",
      "     |   - take shape parameters and return mu, mu2, g1, g2\n",
      "     |   - If you can't compute one of these, return it as None\n",
      "     |   - Can also be defined with a keyword argument ``moments``, which is a\n",
      "     |     string composed of \"m\", \"v\", \"s\", and/or \"k\".\n",
      "     |     Only the components appearing in string should be computed and\n",
      "     |     returned in the order \"m\", \"v\", \"s\", or \"k\"  with missing values\n",
      "     |     returned as None.\n",
      "     |  \n",
      "     |  Alternatively, you can override ``_munp``, which takes ``n`` and shape\n",
      "     |  parameters and returns the n-th non-central moment of the distribution.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  To create a new Gaussian distribution, we would do the following:\n",
      "     |  \n",
      "     |  >>> from scipy.stats import rv_continuous\n",
      "     |  >>> class gaussian_gen(rv_continuous):\n",
      "     |  ...     \"Gaussian distribution\"\n",
      "     |  ...     def _pdf(self, x):\n",
      "     |  ...         return np.exp(-x**2 / 2.) / np.sqrt(2.0 * np.pi)\n",
      "     |  >>> gaussian = gaussian_gen(name='gaussian')\n",
      "     |  \n",
      "     |  ``scipy.stats`` distributions are *instances*, so here we subclass\n",
      "     |  `rv_continuous` and create an instance. With this, we now have\n",
      "     |  a fully functional distribution with all relevant methods automagically\n",
      "     |  generated by the framework.\n",
      "     |  \n",
      "     |  Note that above we defined a standard normal distribution, with zero mean\n",
      "     |  and unit variance. Shifting and scaling of the distribution can be done\n",
      "     |  by using ``loc`` and ``scale`` parameters: ``gaussian.pdf(x, loc, scale)``\n",
      "     |  essentially computes ``y = (x - loc) / scale`` and\n",
      "     |  ``gaussian._pdf(y) / scale``.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_continuous\n",
      "     |      rv_generic\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, momtype=1, a=None, b=None, xtol=1e-14, badvalue=None, name=None, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  cdf(self, x, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `x`\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, scale=1, lb=None, ub=None, conditional=False, **kwds)\n",
      "     |      Calculate expected value of a function with respect to the\n",
      "     |      distribution.\n",
      "     |      \n",
      "     |      The expected value of a function ``f(x)`` with respect to a\n",
      "     |      distribution ``dist`` is defined as::\n",
      "     |      \n",
      "     |                  ubound\n",
      "     |          E[x] = Integral(f(x) * dist.pdf(x))\n",
      "     |                  lbound\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which integral is calculated. Takes only one argument.\n",
      "     |          The default is the identity mapping f(x) = x.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : float, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      lb, ub : scalar, optional\n",
      "     |          Lower and upper bound for integration. Default is set to the\n",
      "     |          support of the distribution.\n",
      "     |      conditional : bool, optional\n",
      "     |          If True, the integral is corrected by the conditional probability\n",
      "     |          of the integration interval.  The return value is the expectation\n",
      "     |          of the function, conditional on being in the given interval.\n",
      "     |          Default is False.\n",
      "     |      \n",
      "     |      Additional keyword arguments are passed to the integration routine.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          The calculated expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The integration behavior of this function is inherited from\n",
      "     |      `integrate.quad`.\n",
      "     |  \n",
      "     |  fit(self, data, *args, **kwds)\n",
      "     |      Return MLEs for shape, location, and scale parameters from data.\n",
      "     |      \n",
      "     |      MLE stands for Maximum Likelihood Estimate.  Starting estimates for\n",
      "     |      the fit are given by input arguments; for any arguments not provided\n",
      "     |      with starting estimates, ``self._fitstart(data)`` is called to generate\n",
      "     |      such.\n",
      "     |      \n",
      "     |      One can hold some parameters fixed to specific values by passing in\n",
      "     |      keyword arguments ``f0``, ``f1``, ..., ``fn`` (for shape parameters)\n",
      "     |      and ``floc`` and ``fscale`` (for location and scale parameters,\n",
      "     |      respectively).\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to use in calculating the MLEs.\n",
      "     |      args : floats, optional\n",
      "     |          Starting value(s) for any shape-characterizing arguments (those not\n",
      "     |          provided will be determined by a call to ``_fitstart(data)``).\n",
      "     |          No default value.\n",
      "     |      kwds : floats, optional\n",
      "     |          Starting values for the location and scale parameters; no default.\n",
      "     |          Special keyword arguments are recognized as holding certain\n",
      "     |          parameters fixed:\n",
      "     |      \n",
      "     |          - f0...fn : hold respective shape parameters fixed.\n",
      "     |            Alternatively, shape parameters to fix can be specified by name.\n",
      "     |            For example, if ``self.shapes == \"a, b\"``, ``fa``and ``fix_a``\n",
      "     |            are equivalent to ``f0``, and ``fb`` and ``fix_b`` are\n",
      "     |            equivalent to ``f1``.\n",
      "     |      \n",
      "     |          - floc : hold location parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - fscale : hold scale parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - optimizer : The optimizer to use.  The optimizer must take ``func``,\n",
      "     |            and starting position as the first two arguments,\n",
      "     |            plus ``args`` (for extra arguments to pass to the\n",
      "     |            function to be optimized) and ``disp=0`` to suppress\n",
      "     |            output as keyword arguments.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      shape, loc, scale : tuple of floats\n",
      "     |          MLEs for any shape statistics, followed by those for location and\n",
      "     |          scale.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This fit is computed by maximizing a log-likelihood function, with\n",
      "     |      penalty applied for samples outside of range of the distribution. The\n",
      "     |      returned answer is not guaranteed to be the globally optimal MLE, it\n",
      "     |      may only be locally optimal, or the optimization may fail altogether.\n",
      "     |      \n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Generate some data to fit: draw random variates from the `beta`\n",
      "     |      distribution\n",
      "     |      \n",
      "     |      >>> from scipy.stats import beta\n",
      "     |      >>> a, b = 1., 2.\n",
      "     |      >>> x = beta.rvs(a, b, size=1000)\n",
      "     |      \n",
      "     |      Now we can fit all four parameters (``a``, ``b``, ``loc`` and ``scale``):\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x)\n",
      "     |      \n",
      "     |      We can also use some prior knowledge about the dataset: let's keep\n",
      "     |      ``loc`` and ``scale`` fixed:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, floc=0, fscale=1)\n",
      "     |      >>> loc1, scale1\n",
      "     |      (0, 1)\n",
      "     |      \n",
      "     |      We can also keep shape parameters fixed by using ``f``-keywords. To\n",
      "     |      keep the zero-th shape parameter ``a`` equal 1, use ``f0=1`` or,\n",
      "     |      equivalently, ``fa=1``:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, fa=1, floc=0, fscale=1)\n",
      "     |      >>> a1\n",
      "     |      1\n",
      "     |  \n",
      "     |  fit_loc_scale(self, data, *args)\n",
      "     |      Estimate loc and scale parameters from data using 1st and 2nd moments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to fit.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Lhat : float\n",
      "     |          Estimated location parameter for the data.\n",
      "     |      Shat : float\n",
      "     |          Estimated scale parameter for the data.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          upper tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability q.\n",
      "     |  \n",
      "     |  logcdf(self, x, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at x\n",
      "     |  \n",
      "     |  logpdf(self, x, *args, **kwds)\n",
      "     |      Log of the probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      This uses a more numerically accurate calculation if available.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpdf : array_like\n",
      "     |          Log of the probability density function evaluated at x\n",
      "     |  \n",
      "     |  logsf(self, x, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as (1 - `cdf`),\n",
      "     |      evaluated at `x`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `x`.\n",
      "     |  \n",
      "     |  nnlf(self, theta, x)\n",
      "     |      Return negative loglikelihood function.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is ``-sum(log pdf(x, theta), axis=0)`` where `theta` are the\n",
      "     |      parameters (including loc and scale).\n",
      "     |  \n",
      "     |  pdf(self, x, *args, **kwds)\n",
      "     |      Probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pdf : ndarray\n",
      "     |          Probability density function evaluated at x\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          lower tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : array_like\n",
      "     |          quantile corresponding to the lower tail probability q.\n",
      "     |  \n",
      "     |  sf(self, x, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at x\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      stats.distributions.rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwds)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (default is 1).\n",
      "     |      random_state : None or int or ``np.random.RandomState`` instance, optional\n",
      "     |          If int or RandomState, use it for drawing the random variates.\n",
      "     |          If None, rely on ``self.random_state``.\n",
      "     |          Default is None.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the RandomState object for generating random variates.\n",
      "     |      \n",
      "     |      This can be either None or an existing RandomState object.\n",
      "     |      \n",
      "     |      If None (or np.random), use the RandomState singleton used by np.random.\n",
      "     |      If already a RandomState instance, use it.\n",
      "     |      If an int, use a new RandomState instance seeded with seed.\n",
      "    \n",
      "    class rv_discrete(rv_generic)\n",
      "     |  A generic discrete random variable class meant for subclassing.\n",
      "     |  \n",
      "     |  `rv_discrete` is a base class to construct specific distribution classes\n",
      "     |  and instances for discrete random variables. It can also be used\n",
      "     |  to construct an arbitrary distribution defined by a list of support\n",
      "     |  points and corresponding probabilities.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  a : float, optional\n",
      "     |      Lower bound of the support of the distribution, default: 0\n",
      "     |  b : float, optional\n",
      "     |      Upper bound of the support of the distribution, default: plus infinity\n",
      "     |  moment_tol : float, optional\n",
      "     |      The tolerance for the generic calculation of moments.\n",
      "     |  values : tuple of two array_like, optional\n",
      "     |      ``(xk, pk)`` where ``xk`` are integers with non-zero\n",
      "     |      probabilities ``pk``  with ``sum(pk) = 1``.\n",
      "     |  inc : integer, optional\n",
      "     |      Increment for the support of the distribution.\n",
      "     |      Default is 1. (other values have not been tested)\n",
      "     |  badvalue : float, optional\n",
      "     |      The value in a result arrays that indicates a value that for which\n",
      "     |      some argument restriction is violated, default is np.nan.\n",
      "     |  name : str, optional\n",
      "     |      The name of the instance. This string is used to construct the default\n",
      "     |      example for distributions.\n",
      "     |  longname : str, optional\n",
      "     |      This string is used as part of the first line of the docstring returned\n",
      "     |      when a subclass has no docstring of its own. Note: `longname` exists\n",
      "     |      for backwards compatibility, do not use for new subclasses.\n",
      "     |  shapes : str, optional\n",
      "     |      The shape of the distribution. For example \"m, n\" for a distribution\n",
      "     |      that takes two integers as the two shape arguments for all its methods\n",
      "     |      If not provided, shape parameters will be inferred from\n",
      "     |      the signatures of the private methods, ``_pmf`` and ``_cdf`` of\n",
      "     |      the instance.\n",
      "     |  extradoc :  str, optional\n",
      "     |      This string is used as the last part of the docstring returned when a\n",
      "     |      subclass has no docstring of its own. Note: `extradoc` exists for\n",
      "     |      backwards compatibility, do not use for new subclasses.\n",
      "     |  seed : None or int or ``numpy.random.RandomState`` instance, optional\n",
      "     |      This parameter defines the RandomState object to use for drawing\n",
      "     |      random variates.\n",
      "     |      If None, the global np.random state is used.\n",
      "     |      If integer, it is used to seed the local RandomState instance.\n",
      "     |      Default is None.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  rvs\n",
      "     |  pmf\n",
      "     |  logpmf\n",
      "     |  cdf\n",
      "     |  logcdf\n",
      "     |  sf\n",
      "     |  logsf\n",
      "     |  ppf\n",
      "     |  isf\n",
      "     |  moment\n",
      "     |  stats\n",
      "     |  entropy\n",
      "     |  expect\n",
      "     |  median\n",
      "     |  mean\n",
      "     |  std\n",
      "     |  var\n",
      "     |  interval\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  \n",
      "     |  This class is similar to `rv_continuous`, the main differences being:\n",
      "     |  \n",
      "     |  - the support of the distribution is a set of integers\n",
      "     |  - instead of the probability density function, ``pdf`` (and the\n",
      "     |    corresponding private ``_pdf``), this class defines the\n",
      "     |    *probability mass function*, `pmf` (and the corresponding\n",
      "     |    private ``_pmf``.)\n",
      "     |  - scale parameter is not defined.\n",
      "     |  \n",
      "     |  To create a new discrete distribution, we would do the following:\n",
      "     |  \n",
      "     |  >>> from scipy.stats import rv_discrete\n",
      "     |  >>> class poisson_gen(rv_discrete):\n",
      "     |  ...     \"Poisson distribution\"\n",
      "     |  ...     def _pmf(self, k, mu):\n",
      "     |  ...         return exp(-mu) * mu**k / factorial(k)\n",
      "     |  \n",
      "     |  and create an instance::\n",
      "     |  \n",
      "     |  >>> poisson = poisson_gen(name=\"poisson\")\n",
      "     |  \n",
      "     |  Note that above we defined the Poisson distribution in the standard form.\n",
      "     |  Shifting the distribution can be done by providing the ``loc`` parameter\n",
      "     |  to the methods of the instance. For example, ``poisson.pmf(x, mu, loc)``\n",
      "     |  delegates the work to ``poisson._pmf(x-loc, mu)``.\n",
      "     |  \n",
      "     |  **Discrete distributions from a list of probabilities**\n",
      "     |  \n",
      "     |  Alternatively, you can construct an arbitrary discrete rv defined\n",
      "     |  on a finite set of values ``xk`` with ``Prob{X=xk} = pk`` by using the\n",
      "     |  ``values`` keyword argument to the `rv_discrete` constructor.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  \n",
      "     |  Custom made discrete distribution:\n",
      "     |  \n",
      "     |  >>> from scipy import stats\n",
      "     |  >>> xk = np.arange(7)\n",
      "     |  >>> pk = (0.1, 0.2, 0.3, 0.1, 0.1, 0.0, 0.2)\n",
      "     |  >>> custm = stats.rv_discrete(name='custm', values=(xk, pk))\n",
      "     |  >>>\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig, ax = plt.subplots(1, 1)\n",
      "     |  >>> ax.plot(xk, custm.pmf(xk), 'ro', ms=12, mec='r')\n",
      "     |  >>> ax.vlines(xk, 0, custm.pmf(xk), colors='r', lw=4)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Random number generation:\n",
      "     |  \n",
      "     |  >>> R = custm.rvs(size=100)\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_discrete\n",
      "     |      rv_generic\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  cdf(self, k, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like, int\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `k`.\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, lb=None, ub=None, conditional=False, maxcount=1000, tolerance=1e-10, chunksize=32)\n",
      "     |      Calculate expected value of a function with respect to the distribution\n",
      "     |      for discrete distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which the expectation value is calculated.\n",
      "     |          Takes only one argument.\n",
      "     |          The default is the identity mapping f(k) = k.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter.\n",
      "     |          Default is 0.\n",
      "     |      lb, ub : int, optional\n",
      "     |          Lower and upper bound for the summation, default is set to the\n",
      "     |          support of the distribution, inclusive (``ul <= k <= ub``).\n",
      "     |      conditional : bool, optional\n",
      "     |          If true then the expectation is corrected by the conditional\n",
      "     |          probability of the summation interval. The return value is the\n",
      "     |          expectation of the function, `func`, conditional on being in\n",
      "     |          the given interval (k such that ``ul <= k <= ub``).\n",
      "     |          Default is False.\n",
      "     |      maxcount : int, optional\n",
      "     |          Maximal number of terms to evaluate (to avoid an endless loop for\n",
      "     |          an infinite sum). Default is 1000.\n",
      "     |      tolerance : float, optional\n",
      "     |          Absolute tolerance for the summation. Default is 1e-10.\n",
      "     |      chunksize : int, optional\n",
      "     |          Iterate over the support of a distributions in chunks of this size.\n",
      "     |          Default is 32.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          Expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      For heavy-tailed distributions, the expected value may or may not exist,\n",
      "     |      depending on the function, `func`. If it does exist, but the sum converges\n",
      "     |      slowly, the accuracy of the result may be rather low. For instance, for\n",
      "     |      ``zipf(4)``, accuracy for mean, variance in example is only 1e-5.\n",
      "     |      increasing `maxcount` and/or `chunksize` may improve the result, but may also\n",
      "     |      make zipf very slow.\n",
      "     |      \n",
      "     |      The function is not vectorized.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          Upper tail probability.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      k : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability, q.\n",
      "     |  \n",
      "     |  logcdf(self, k, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like, int\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at k.\n",
      "     |  \n",
      "     |  logpmf(self, k, *args, **kwds)\n",
      "     |      Log of the probability mass function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter. Default is 0.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpmf : array_like\n",
      "     |          Log of the probability mass function evaluated at k.\n",
      "     |  \n",
      "     |  logsf(self, k, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as 1 - `cdf`,\n",
      "     |      evaluated at `k`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `k`.\n",
      "     |  \n",
      "     |  pmf(self, k, *args, **kwds)\n",
      "     |      Probability mass function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pmf : array_like\n",
      "     |          Probability mass function evaluated at k\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          Lower tail probability.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      k : array_like\n",
      "     |          Quantile corresponding to the lower tail probability, q.\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwargs)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (Default is 1).  Note that `size`\n",
      "     |          has to be given as keyword, not as positional argument.\n",
      "     |      random_state : None or int or ``np.random.RandomState`` instance, optional\n",
      "     |          If int or RandomState, use it for drawing the random variates.\n",
      "     |          If None, rely on ``self.random_state``.\n",
      "     |          Default is None.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  sf(self, k, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at k.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods defined here:\n",
      "     |  \n",
      "     |  __new__(cls, a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  return_integers\n",
      "     |      `return_integers` is deprecated!\n",
      "     |      `return_integers` attribute is not used anywhere any  longer and is deprecated in scipy 0.18.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      stats.distributions.rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the RandomState object for generating random variates.\n",
      "     |      \n",
      "     |      This can be either None or an existing RandomState object.\n",
      "     |      \n",
      "     |      If None (or np.random), use the RandomState singleton used by np.random.\n",
      "     |      If already a RandomState instance, use it.\n",
      "     |      If an int, use a new RandomState instance seeded with seed.\n",
      "\n",
      "FUNCTIONS\n",
      "    anderson(x, dist='norm')\n",
      "        Anderson-Darling test for data coming from a particular distribution\n",
      "        \n",
      "        The Anderson-Darling test is a modification of the Kolmogorov-\n",
      "        Smirnov test `kstest` for the null hypothesis that a sample is\n",
      "        drawn from a population that follows a particular distribution.\n",
      "        For the Anderson-Darling test, the critical values depend on\n",
      "        which distribution is being tested against.  This function works\n",
      "        for normal, exponential, logistic, or Gumbel (Extreme Value\n",
      "        Type I) distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            array of sample data\n",
      "        dist : {'norm','expon','logistic','gumbel','extreme1'}, optional\n",
      "            the type of distribution to test against.  The default is 'norm'\n",
      "            and 'extreme1' is a synonym for 'gumbel'\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The Anderson-Darling test statistic\n",
      "        critical_values : list\n",
      "            The critical values for this distribution\n",
      "        significance_level : list\n",
      "            The significance levels for the corresponding critical values\n",
      "            in percents.  The function returns critical values for a\n",
      "            differing set of significance levels depending on the\n",
      "            distribution that is being tested against.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Critical values provided are for the following significance levels:\n",
      "        \n",
      "        normal/exponenential\n",
      "            15%, 10%, 5%, 2.5%, 1%\n",
      "        logistic\n",
      "            25%, 10%, 5%, 2.5%, 1%, 0.5%\n",
      "        Gumbel\n",
      "            25%, 10%, 5%, 2.5%, 1%\n",
      "        \n",
      "        If A2 is larger than these critical values then for the corresponding\n",
      "        significance level, the null hypothesis that the data come from the\n",
      "        chosen distribution can be rejected.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://www.itl.nist.gov/div898/handbook/prc/section2/prc213.htm\n",
      "        .. [2] Stephens, M. A. (1974). EDF Statistics for Goodness of Fit and\n",
      "               Some Comparisons, Journal of the American Statistical Association,\n",
      "               Vol. 69, pp. 730-737.\n",
      "        .. [3] Stephens, M. A. (1976). Asymptotic Results for Goodness-of-Fit\n",
      "               Statistics with Unknown Parameters, Annals of Statistics, Vol. 4,\n",
      "               pp. 357-369.\n",
      "        .. [4] Stephens, M. A. (1977). Goodness of Fit for the Extreme Value\n",
      "               Distribution, Biometrika, Vol. 64, pp. 583-588.\n",
      "        .. [5] Stephens, M. A. (1977). Goodness of Fit with Special Reference\n",
      "               to Tests for Exponentiality , Technical Report No. 262,\n",
      "               Department of Statistics, Stanford University, Stanford, CA.\n",
      "        .. [6] Stephens, M. A. (1979). Tests of Fit for the Logistic Distribution\n",
      "               Based on the Empirical Distribution Function, Biometrika, Vol. 66,\n",
      "               pp. 591-595.\n",
      "    \n",
      "    anderson_ksamp(samples, midrank=True)\n",
      "        The Anderson-Darling test for k-samples.\n",
      "        \n",
      "        The k-sample Anderson-Darling test is a modification of the\n",
      "        one-sample Anderson-Darling test. It tests the null hypothesis\n",
      "        that k-samples are drawn from the same population without having\n",
      "        to specify the distribution function of that population. The\n",
      "        critical values depend on the number of samples.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : sequence of 1-D array_like\n",
      "            Array of sample data in arrays.\n",
      "        midrank : bool, optional\n",
      "            Type of Anderson-Darling test which is computed. Default\n",
      "            (True) is the midrank test applicable to continuous and\n",
      "            discrete populations. If False, the right side empirical\n",
      "            distribution is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            Normalized k-sample Anderson-Darling test statistic.\n",
      "        critical_values : array\n",
      "            The critical values for significance levels 25%, 10%, 5%, 2.5%, 1%.\n",
      "        significance_level : float\n",
      "            An approximate significance level at which the null hypothesis for the\n",
      "            provided samples can be rejected.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        ValueError\n",
      "            If less than 2 samples are provided, a sample is empty, or no\n",
      "            distinct observations are in the samples.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ks_2samp : 2 sample Kolmogorov-Smirnov test\n",
      "        anderson : 1 sample Anderson-Darling test\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        [1]_ Defines three versions of the k-sample Anderson-Darling test:\n",
      "        one for continuous distributions and two for discrete\n",
      "        distributions, in which ties between samples may occur. The\n",
      "        default of this routine is to compute the version based on the\n",
      "        midrank empirical distribution function. This test is applicable\n",
      "        to continuous and discrete data. If midrank is set to False, the\n",
      "        right side empirical distribution is used for a test for discrete\n",
      "        data. According to [1]_, the two discrete test statistics differ\n",
      "        only slightly if a few collisions due to round-off errors occur in\n",
      "        the test not adjusted for ties between samples.\n",
      "        \n",
      "        .. versionadded:: 0.14.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Scholz, F. W and Stephens, M. A. (1987), K-Sample\n",
      "               Anderson-Darling Tests, Journal of the American Statistical\n",
      "               Association, Vol. 82, pp. 918-924.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(314159)\n",
      "        \n",
      "        The null hypothesis that the two random samples come from the same\n",
      "        distribution can be rejected at the 5% level because the returned\n",
      "        test value is greater than the critical value for 5% (1.961) but\n",
      "        not at the 2.5% level. The interpolation gives an approximate\n",
      "        significance level of 3.1%:\n",
      "        \n",
      "        >>> stats.anderson_ksamp([np.random.normal(size=50),\n",
      "        ... np.random.normal(loc=0.5, size=30)])\n",
      "        (2.4615796189876105,\n",
      "          array([ 0.325,  1.226,  1.961,  2.718,  3.752]),\n",
      "          0.03134990135800783)\n",
      "        \n",
      "        \n",
      "        The null hypothesis cannot be rejected for three samples from an\n",
      "        identical distribution. The approximate p-value (87%) has to be\n",
      "        computed by extrapolation and may not be very accurate:\n",
      "        \n",
      "        >>> stats.anderson_ksamp([np.random.normal(size=50),\n",
      "        ... np.random.normal(size=30), np.random.normal(size=20)])\n",
      "        (-0.73091722665244196,\n",
      "          array([ 0.44925884,  1.3052767 ,  1.9434184 ,  2.57696569,  3.41634856]),\n",
      "          0.8789283903979661)\n",
      "    \n",
      "    ansari(x, y)\n",
      "        Perform the Ansari-Bradley test for equal scale parameters\n",
      "        \n",
      "        The Ansari-Bradley test is a non-parametric test for the equality\n",
      "        of the scale parameter of the distributions from which two\n",
      "        samples were drawn.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            arrays of sample data\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The Ansari-Bradley test statistic\n",
      "        pvalue : float\n",
      "            The p-value of the hypothesis test\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fligner : A non-parametric test for the equality of k variances\n",
      "        mood : A non-parametric test for the equality of two scale parameters\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The p-value given is exact when the sample sizes are both less than\n",
      "        55 and there are no ties, otherwise a normal approximation for the\n",
      "        p-value is used.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Sprent, Peter and N.C. Smeeton.  Applied nonparametric statistical\n",
      "               methods.  3rd ed. Chapman and Hall/CRC. 2001.  Section 5.8.2.\n",
      "    \n",
      "    bartlett(*args)\n",
      "        Perform Bartlett's test for equal variances\n",
      "        \n",
      "        Bartlett's test tests the null hypothesis that all input samples\n",
      "        are from populations with equal variances.  For samples\n",
      "        from significantly non-normal populations, Levene's test\n",
      "        `levene` is more robust.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2,... : array_like\n",
      "            arrays of sample data.  May be different lengths.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The p-value of the test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fligner : A non-parametric test for the equality of k variances\n",
      "        levene : A robust parametric test for equality of k variances\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Conover et al. (1981) examine many of the existing parametric and\n",
      "        nonparametric tests by extensive simulations and they conclude that the\n",
      "        tests proposed by Fligner and Killeen (1976) and Levene (1960) appear to be\n",
      "        superior in terms of robustness of departures from normality and power [3]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1]  http://www.itl.nist.gov/div898/handbook/eda/section3/eda357.htm\n",
      "        \n",
      "        .. [2]  Snedecor, George W. and Cochran, William G. (1989), Statistical\n",
      "                  Methods, Eighth Edition, Iowa State University Press.\n",
      "        \n",
      "        .. [3] Park, C. and Lindsay, B. G. (1999). Robust Scale Estimation and\n",
      "               Hypothesis Testing based on Quadratic Inference Function. Technical\n",
      "               Report #99-03, Center for Likelihood Studies, Pennsylvania State\n",
      "               University.\n",
      "        \n",
      "        .. [4] Bartlett, M. S. (1937). Properties of Sufficiency and Statistical\n",
      "               Tests. Proceedings of the Royal Society of London. Series A,\n",
      "               Mathematical and Physical Sciences, Vol. 160, No.901, pp. 268-282.\n",
      "    \n",
      "    bayes_mvs(data, alpha=0.9)\n",
      "        Bayesian confidence intervals for the mean, var, and std.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input data, if multi-dimensional it is flattened to 1-D by `bayes_mvs`.\n",
      "            Requires 2 or more data points.\n",
      "        alpha : float, optional\n",
      "            Probability that the returned confidence interval contains\n",
      "            the true parameter.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mean_cntr, var_cntr, std_cntr : tuple\n",
      "            The three results are for the mean, variance and standard deviation,\n",
      "            respectively.  Each result is a tuple of the form::\n",
      "        \n",
      "                (center, (lower, upper))\n",
      "        \n",
      "            with `center` the mean of the conditional pdf of the value given the\n",
      "            data, and `(lower, upper)` a confidence interval, centered on the\n",
      "            median, containing the estimate to a probability ``alpha``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        mvsdist\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Each tuple of mean, variance, and standard deviation estimates represent\n",
      "        the (center, (lower, upper)) with center the mean of the conditional pdf\n",
      "        of the value given the data and (lower, upper) is a confidence interval\n",
      "        centered on the median, containing the estimate to a probability\n",
      "        ``alpha``.\n",
      "        \n",
      "        Converts data to 1-D and assumes all data has the same mean and variance.\n",
      "        Uses Jeffrey's prior for variance and std.\n",
      "        \n",
      "        Equivalent to ``tuple((x.mean(), x.interval(alpha)) for x in mvsdist(dat))``\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        T.E. Oliphant, \"A Bayesian perspective on estimating mean, variance, and\n",
      "        standard-deviation from data\", http://scholarsarchive.byu.edu/facpub/278,\n",
      "        2006.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First a basic example to demonstrate the outputs:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> data = [6, 9, 12, 7, 8, 8, 13]\n",
      "        >>> mean, var, std = stats.bayes_mvs(data)\n",
      "        >>> mean\n",
      "        Mean(statistic=9.0, minmax=(7.1036502226125329, 10.896349777387467))\n",
      "        >>> var\n",
      "        Variance(statistic=10.0, minmax=(3.176724206..., 24.45910382...))\n",
      "        >>> std\n",
      "        Std_dev(statistic=2.9724954732045084, minmax=(1.7823367265645143, 4.9456146050146295))\n",
      "        \n",
      "        Now we generate some normally distributed random data, and get estimates of\n",
      "        mean and standard deviation with 95% confidence intervals for those\n",
      "        estimates:\n",
      "        \n",
      "        >>> n_samples = 100000\n",
      "        >>> data = stats.norm.rvs(size=n_samples)\n",
      "        >>> res_mean, res_var, res_std = stats.bayes_mvs(data, alpha=0.95)\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.hist(data, bins=100, normed=True, label='Histogram of data')\n",
      "        >>> ax.vlines(res_mean.statistic, 0, 0.5, colors='r', label='Estimated mean')\n",
      "        >>> ax.axvspan(res_mean.minmax[0],res_mean.minmax[1], facecolor='r',\n",
      "        ...            alpha=0.2, label=r'Estimated mean (95% limits)')\n",
      "        >>> ax.vlines(res_std.statistic, 0, 0.5, colors='g', label='Estimated scale')\n",
      "        >>> ax.axvspan(res_std.minmax[0],res_std.minmax[1], facecolor='g', alpha=0.2,\n",
      "        ...            label=r'Estimated scale (95% limits)')\n",
      "        \n",
      "        >>> ax.legend(fontsize=10)\n",
      "        >>> ax.set_xlim([-4, 4])\n",
      "        >>> ax.set_ylim([0, 0.5])\n",
      "        >>> plt.show()\n",
      "    \n",
      "    betai(*args, **kwds)\n",
      "        `betai` is deprecated!\n",
      "        stats.betai is deprecated in scipy 0.17.0; use special.betainc instead\n",
      "        \n",
      "        \n",
      "            Returns the incomplete beta function.\n",
      "        \n",
      "            I_x(a,b) = 1/B(a,b)*(Integral(0,x) of t^(a-1)(1-t)^(b-1) dt)\n",
      "        \n",
      "            where a,b>0 and B(a,b) = G(a)*G(b)/(G(a+b)) where G(a) is the gamma\n",
      "            function of a.\n",
      "        \n",
      "            The standard broadcasting rules apply to a, b, and x.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            a : array_like or float > 0\n",
      "        \n",
      "            b : array_like or float > 0\n",
      "        \n",
      "            x : array_like or float\n",
      "                x will be clipped to be no greater than 1.0 .\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            betai : ndarray\n",
      "                Incomplete beta function.\n",
      "    \n",
      "    binned_statistic(x, values, statistic='mean', bins=10, range=None)\n",
      "        Compute a binned statistic for one or more sets of data.\n",
      "        \n",
      "        This is a generalization of a histogram function.  A histogram divides\n",
      "        the space into bins, and returns the count of the number of points in\n",
      "        each bin.  This function allows the computation of the sum, mean, median,\n",
      "        or other statistic of the values (or set of values) within each bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array_like\n",
      "            A sequence of values to be binned.\n",
      "        values : (N,) array_like or list of (N,) array_like\n",
      "            The data on which the statistic will be computed.  This must be\n",
      "            the same shape as `x`, or a set of sequences - each the same shape as\n",
      "            `x`.  If `values` is a set of sequences, the statistic will be computed\n",
      "            on each independently.\n",
      "        statistic : string or callable, optional\n",
      "            The statistic to compute (default is 'mean').\n",
      "            The following statistics are available:\n",
      "        \n",
      "              * 'mean' : compute the mean of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'median' : compute the median of values for points within each\n",
      "                bin. Empty bins will be represented by NaN.\n",
      "              * 'count' : compute the count of points within each bin.  This is\n",
      "                identical to an unweighted histogram.  `values` array is not\n",
      "                referenced.\n",
      "              * 'sum' : compute the sum of values for points within each bin.\n",
      "                This is identical to a weighted histogram.\n",
      "              * function : a user-defined function which takes a 1D array of\n",
      "                values, and outputs a single numerical statistic. This function\n",
      "                will be called on the values in each bin.  Empty bins will be\n",
      "                represented by function([]), or NaN if this returns an error.\n",
      "        \n",
      "        bins : int or sequence of scalars, optional\n",
      "            If `bins` is an int, it defines the number of equal-width bins in the\n",
      "            given range (10 by default).  If `bins` is a sequence, it defines the\n",
      "            bin edges, including the rightmost edge, allowing for non-uniform bin\n",
      "            widths.  Values in `x` that are smaller than lowest bin edge are\n",
      "            assigned to bin number 0, values beyond the highest bin are assigned to\n",
      "            ``bins[-1]``.  If the bin edges are specified, the number of bins will\n",
      "            be, (nx = len(bins)-1).\n",
      "        range : (float, float) or [(float, float)], optional\n",
      "            The lower and upper range of the bins.  If not provided, range\n",
      "            is simply ``(x.min(), x.max())``.  Values outside the range are\n",
      "            ignored.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : array\n",
      "            The values of the selected statistic in each bin.\n",
      "        bin_edges : array of dtype float\n",
      "            Return the bin edges ``(length(statistic)+1)``.\n",
      "        binnumber: 1-D ndarray of ints\n",
      "            Indices of the bins (corresponding to `bin_edges`) in which each value\n",
      "            of `x` belongs.  Same length as `values`.  A binnumber of `i` means the\n",
      "            corresponding value is between (bin_edges[i-1], bin_edges[i]).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.digitize, numpy.histogram, binned_statistic_2d, binned_statistic_dd\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        All but the last (righthand-most) bin is half-open.  In other words, if\n",
      "        `bins` is ``[1, 2, 3, 4]``, then the first bin is ``[1, 2)`` (including 1,\n",
      "        but excluding 2) and the second ``[2, 3)``.  The last bin, however, is\n",
      "        ``[3, 4]``, which *includes* 4.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        First some basic examples:\n",
      "        \n",
      "        Create two evenly spaced bins in the range of the given sample, and sum the\n",
      "        corresponding values in each of those bins:\n",
      "        \n",
      "        >>> values = [1.0, 1.0, 2.0, 1.5, 3.0]\n",
      "        >>> stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)\n",
      "        (array([ 4. ,  4.5]), array([ 1.,  4.,  7.]), array([1, 1, 1, 2, 2]))\n",
      "        \n",
      "        Multiple arrays of values can also be passed.  The statistic is calculated\n",
      "        on each set independently:\n",
      "        \n",
      "        >>> values = [[1.0, 1.0, 2.0, 1.5, 3.0], [2.0, 2.0, 4.0, 3.0, 6.0]]\n",
      "        >>> stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)\n",
      "        (array([[ 4. ,  4.5], [ 8. ,  9. ]]), array([ 1.,  4.,  7.]),\n",
      "            array([1, 1, 1, 2, 2]))\n",
      "        \n",
      "        >>> stats.binned_statistic([1, 2, 1, 2, 4], np.arange(5), statistic='mean',\n",
      "        ...                        bins=3)\n",
      "        (array([ 1.,  2.,  4.]), array([ 1.,  2.,  3.,  4.]),\n",
      "            array([1, 2, 1, 2, 3]))\n",
      "        \n",
      "        As a second example, we now generate some random data of sailing boat speed\n",
      "        as a function of wind speed, and then determine how fast our boat is for\n",
      "        certain wind speeds:\n",
      "        \n",
      "        >>> windspeed = 8 * np.random.rand(500)\n",
      "        >>> boatspeed = .3 * windspeed**.5 + .2 * np.random.rand(500)\n",
      "        >>> bin_means, bin_edges, binnumber = stats.binned_statistic(windspeed,\n",
      "        ...                 boatspeed, statistic='median', bins=[1,2,3,4,5,6,7])\n",
      "        >>> plt.figure()\n",
      "        >>> plt.plot(windspeed, boatspeed, 'b.', label='raw data')\n",
      "        >>> plt.hlines(bin_means, bin_edges[:-1], bin_edges[1:], colors='g', lw=5,\n",
      "        ...            label='binned statistic of data')\n",
      "        >>> plt.legend()\n",
      "        \n",
      "        Now we can use ``binnumber`` to select all datapoints with a windspeed\n",
      "        below 1:\n",
      "        \n",
      "        >>> low_boatspeed = boatspeed[binnumber == 0]\n",
      "        \n",
      "        As a final example, we will use ``bin_edges`` and ``binnumber`` to make a\n",
      "        plot of a distribution that shows the mean and distribution around that\n",
      "        mean per bin, on top of a regular histogram and the probability\n",
      "        distribution function:\n",
      "        \n",
      "        >>> x = np.linspace(0, 5, num=500)\n",
      "        >>> x_pdf = stats.maxwell.pdf(x)\n",
      "        >>> samples = stats.maxwell.rvs(size=10000)\n",
      "        \n",
      "        >>> bin_means, bin_edges, binnumber = stats.binned_statistic(x, x_pdf,\n",
      "        ...         statistic='mean', bins=25)\n",
      "        >>> bin_width = (bin_edges[1] - bin_edges[0])\n",
      "        >>> bin_centers = bin_edges[1:] - bin_width/2\n",
      "        \n",
      "        >>> plt.figure()\n",
      "        >>> plt.hist(samples, bins=50, normed=True, histtype='stepfilled',\n",
      "        ...          alpha=0.2, label='histogram of data')\n",
      "        >>> plt.plot(x, x_pdf, 'r-', label='analytical pdf')\n",
      "        >>> plt.hlines(bin_means, bin_edges[:-1], bin_edges[1:], colors='g', lw=2,\n",
      "        ...            label='binned statistic of data')\n",
      "        >>> plt.plot((binnumber - 0.5) * bin_width, x_pdf, 'g.', alpha=0.5)\n",
      "        >>> plt.legend(fontsize=10)\n",
      "        >>> plt.show()\n",
      "    \n",
      "    binned_statistic_2d(x, y, values, statistic='mean', bins=10, range=None, expand_binnumbers=False)\n",
      "        Compute a bidimensional binned statistic for one or more sets of data.\n",
      "        \n",
      "        This is a generalization of a histogram2d function.  A histogram divides\n",
      "        the space into bins, and returns the count of the number of points in\n",
      "        each bin.  This function allows the computation of the sum, mean, median,\n",
      "        or other statistic of the values (or set of values) within each bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array_like\n",
      "            A sequence of values to be binned along the first dimension.\n",
      "        y : (N,) array_like\n",
      "            A sequence of values to be binned along the second dimension.\n",
      "        values : (N,) array_like or list of (N,) array_like\n",
      "            The data on which the statistic will be computed.  This must be\n",
      "            the same shape as `x`, or a list of sequences - each with the same\n",
      "            shape as `x`.  If `values` is such a list, the statistic will be\n",
      "            computed on each independently.\n",
      "        statistic : string or callable, optional\n",
      "            The statistic to compute (default is 'mean').\n",
      "            The following statistics are available:\n",
      "        \n",
      "              * 'mean' : compute the mean of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'median' : compute the median of values for points within each\n",
      "                bin. Empty bins will be represented by NaN.\n",
      "              * 'count' : compute the count of points within each bin.  This is\n",
      "                identical to an unweighted histogram.  `values` array is not\n",
      "                referenced.\n",
      "              * 'sum' : compute the sum of values for points within each bin.\n",
      "                This is identical to a weighted histogram.\n",
      "              * function : a user-defined function which takes a 1D array of\n",
      "                values, and outputs a single numerical statistic. This function\n",
      "                will be called on the values in each bin.  Empty bins will be\n",
      "                represented by function([]), or NaN if this returns an error.\n",
      "        \n",
      "        bins : int or [int, int] or array_like or [array, array], optional\n",
      "            The bin specification:\n",
      "        \n",
      "              * the number of bins for the two dimensions (nx = ny = bins),\n",
      "              * the number of bins in each dimension (nx, ny = bins),\n",
      "              * the bin edges for the two dimensions (x_edge = y_edge = bins),\n",
      "              * the bin edges in each dimension (x_edge, y_edge = bins).\n",
      "        \n",
      "            If the bin edges are specified, the number of bins will be,\n",
      "            (nx = len(x_edge)-1, ny = len(y_edge)-1).\n",
      "        \n",
      "        range : (2,2) array_like, optional\n",
      "            The leftmost and rightmost edges of the bins along each dimension\n",
      "            (if not specified explicitly in the `bins` parameters):\n",
      "            [[xmin, xmax], [ymin, ymax]]. All values outside of this range will be\n",
      "            considered outliers and not tallied in the histogram.\n",
      "        expand_binnumbers : bool, optional\n",
      "            'False' (default): the returned `binnumber` is a shape (N,) array of\n",
      "            linearized bin indices.\n",
      "            'True': the returned `binnumber` is 'unraveled' into a shape (2,N)\n",
      "            ndarray, where each row gives the bin numbers in the corresponding\n",
      "            dimension.\n",
      "            See the `binnumber` returned value, and the `Examples` section.\n",
      "        \n",
      "            .. versionadded:: 0.17.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : (nx, ny) ndarray\n",
      "            The values of the selected statistic in each two-dimensional bin.\n",
      "        x_edge : (nx + 1) ndarray\n",
      "            The bin edges along the first dimension.\n",
      "        y_edge : (ny + 1) ndarray\n",
      "            The bin edges along the second dimension.\n",
      "        binnumber : (N,) array of ints or (2,N) ndarray of ints\n",
      "            This assigns to each element of `sample` an integer that represents the\n",
      "            bin in which this observation falls.  The representation depends on the\n",
      "            `expand_binnumbers` argument.  See `Notes` for details.\n",
      "        \n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.digitize, numpy.histogram2d, binned_statistic, binned_statistic_dd\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Binedges:\n",
      "        All but the last (righthand-most) bin is half-open.  In other words, if\n",
      "        `bins` is ``[1, 2, 3, 4]``, then the first bin is ``[1, 2)`` (including 1,\n",
      "        but excluding 2) and the second ``[2, 3)``.  The last bin, however, is\n",
      "        ``[3, 4]``, which *includes* 4.\n",
      "        \n",
      "        `binnumber`:\n",
      "        This returned argument assigns to each element of `sample` an integer that\n",
      "        represents the bin in which it belongs.  The representation depends on the\n",
      "        `expand_binnumbers` argument. If 'False' (default): The returned\n",
      "        `binnumber` is a shape (N,) array of linearized indices mapping each\n",
      "        element of `sample` to its corresponding bin (using row-major ordering).\n",
      "        If 'True': The returned `binnumber` is a shape (2,N) ndarray where\n",
      "        each row indicates bin placements for each dimension respectively.  In each\n",
      "        dimension, a binnumber of `i` means the corresponding value is between\n",
      "        (D_edge[i-1], D_edge[i]), where 'D' is either 'x' or 'y'.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        \n",
      "        Calculate the counts with explicit bin-edges:\n",
      "        \n",
      "        >>> x = [0.1, 0.1, 0.1, 0.6]\n",
      "        >>> y = [2.1, 2.6, 2.1, 2.1]\n",
      "        >>> binx = [0.0, 0.5, 1.0]\n",
      "        >>> biny = [2.0, 2.5, 3.0]\n",
      "        >>> ret = stats.binned_statistic_2d(x, y, None, 'count', bins=[binx,biny])\n",
      "        >>> ret.statistic\n",
      "        array([[ 2.,  1.],\n",
      "               [ 1.,  0.]])\n",
      "        \n",
      "        The bin in which each sample is placed is given by the `binnumber`\n",
      "        returned parameter.  By default, these are the linearized bin indices:\n",
      "        \n",
      "        >>> ret.binnumber\n",
      "        array([5, 6, 5, 9])\n",
      "        \n",
      "        The bin indices can also be expanded into separate entries for each\n",
      "        dimension using the `expand_binnumbers` parameter:\n",
      "        \n",
      "        >>> ret = stats.binned_statistic_2d(x, y, None, 'count', bins=[binx,biny],\n",
      "        ...                                 expand_binnumbers=True)\n",
      "        >>> ret.binnumber\n",
      "        array([[1, 1, 1, 2],\n",
      "               [1, 2, 1, 1]])\n",
      "        \n",
      "        Which shows that the first three elements belong in the xbin 1, and the\n",
      "        fourth into xbin 2; and so on for y.\n",
      "    \n",
      "    binned_statistic_dd(sample, values, statistic='mean', bins=10, range=None, expand_binnumbers=False)\n",
      "        Compute a multidimensional binned statistic for a set of data.\n",
      "        \n",
      "        This is a generalization of a histogramdd function.  A histogram divides\n",
      "        the space into bins, and returns the count of the number of points in\n",
      "        each bin.  This function allows the computation of the sum, mean, median,\n",
      "        or other statistic of the values within each bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample : array_like\n",
      "            Data to histogram passed as a sequence of D arrays of length N, or\n",
      "            as an (N,D) array.\n",
      "        values : (N,) array_like or list of (N,) array_like\n",
      "            The data on which the statistic will be computed.  This must be\n",
      "            the same shape as `x`, or a list of sequences - each with the same\n",
      "            shape as `x`.  If `values` is such a list, the statistic will be\n",
      "            computed on each independently.\n",
      "        statistic : string or callable, optional\n",
      "            The statistic to compute (default is 'mean').\n",
      "            The following statistics are available:\n",
      "        \n",
      "              * 'mean' : compute the mean of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'median' : compute the median of values for points within each\n",
      "                bin. Empty bins will be represented by NaN.\n",
      "              * 'count' : compute the count of points within each bin.  This is\n",
      "                identical to an unweighted histogram.  `values` array is not\n",
      "                referenced.\n",
      "              * 'sum' : compute the sum of values for points within each bin.\n",
      "                This is identical to a weighted histogram.\n",
      "              * function : a user-defined function which takes a 1D array of\n",
      "                values, and outputs a single numerical statistic. This function\n",
      "                will be called on the values in each bin.  Empty bins will be\n",
      "                represented by function([]), or NaN if this returns an error.\n",
      "        \n",
      "        bins : sequence or int, optional\n",
      "            The bin specification must be in one of the following forms:\n",
      "        \n",
      "              * A sequence of arrays describing the bin edges along each dimension.\n",
      "              * The number of bins for each dimension (nx, ny, ... = bins).\n",
      "              * The number of bins for all dimensions (nx = ny = ... = bins).\n",
      "        \n",
      "        range : sequence, optional\n",
      "            A sequence of lower and upper bin edges to be used if the edges are\n",
      "            not given explicitely in `bins`. Defaults to the minimum and maximum\n",
      "            values along each dimension.\n",
      "        expand_binnumbers : bool, optional\n",
      "            'False' (default): the returned `binnumber` is a shape (N,) array of\n",
      "            linearized bin indices.\n",
      "            'True': the returned `binnumber` is 'unraveled' into a shape (D,N)\n",
      "            ndarray, where each row gives the bin numbers in the corresponding\n",
      "            dimension.\n",
      "            See the `binnumber` returned value, and the `Examples` section of\n",
      "            `binned_statistic_2d`.\n",
      "        \n",
      "            .. versionadded:: 0.17.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : ndarray, shape(nx1, nx2, nx3,...)\n",
      "            The values of the selected statistic in each two-dimensional bin.\n",
      "        bin_edges : list of ndarrays\n",
      "            A list of D arrays describing the (nxi + 1) bin edges for each\n",
      "            dimension.\n",
      "        binnumber : (N,) array of ints or (D,N) ndarray of ints\n",
      "            This assigns to each element of `sample` an integer that represents the\n",
      "            bin in which this observation falls.  The representation depends on the\n",
      "            `expand_binnumbers` argument.  See `Notes` for details.\n",
      "        \n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.digitize, numpy.histogramdd, binned_statistic, binned_statistic_2d\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Binedges:\n",
      "        All but the last (righthand-most) bin is half-open in each dimension.  In\n",
      "        other words, if `bins` is ``[1, 2, 3, 4]``, then the first bin is\n",
      "        ``[1, 2)`` (including 1, but excluding 2) and the second ``[2, 3)``.  The\n",
      "        last bin, however, is ``[3, 4]``, which *includes* 4.\n",
      "        \n",
      "        `binnumber`:\n",
      "        This returned argument assigns to each element of `sample` an integer that\n",
      "        represents the bin in which it belongs.  The representation depends on the\n",
      "        `expand_binnumbers` argument. If 'False' (default): The returned\n",
      "        `binnumber` is a shape (N,) array of linearized indices mapping each\n",
      "        element of `sample` to its corresponding bin (using row-major ordering).\n",
      "        If 'True': The returned `binnumber` is a shape (D,N) ndarray where\n",
      "        each row indicates bin placements for each dimension respectively.  In each\n",
      "        dimension, a binnumber of `i` means the corresponding value is between\n",
      "        (bin_edges[D][i-1], bin_edges[D][i]), for each dimension 'D'.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "    \n",
      "    binom_test(x, n=None, p=0.5, alternative='two-sided')\n",
      "        Perform a test that the probability of success is p.\n",
      "        \n",
      "        This is an exact, two-sided test of the null hypothesis\n",
      "        that the probability of success in a Bernoulli experiment\n",
      "        is `p`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : integer or array_like\n",
      "            the number of successes, or if x has length 2, it is the\n",
      "            number of successes and the number of failures.\n",
      "        n : integer\n",
      "            the number of trials.  This is ignored if x gives both the\n",
      "            number of successes and failures\n",
      "        p : float, optional\n",
      "            The hypothesized probability of success.  0 <= p <= 1. The\n",
      "            default value is p = 0.5\n",
      "        alternative : {'two-sided', 'greater', 'less'}, optional\n",
      "            Indicates the alternative hypothesis. The default value is\n",
      "            'two-sided'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        p-value : float\n",
      "            The p-value of the hypothesis test\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://en.wikipedia.org/wiki/Binomial_test\n",
      "    \n",
      "    boxcox(x, lmbda=None, alpha=None)\n",
      "        Return a positive dataset transformed by a Box-Cox power transformation.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : ndarray\n",
      "            Input array.  Should be 1-dimensional.\n",
      "        lmbda : {None, scalar}, optional\n",
      "            If `lmbda` is not None, do the transformation for that value.\n",
      "        \n",
      "            If `lmbda` is None, find the lambda that maximizes the log-likelihood\n",
      "            function and return it as the second output argument.\n",
      "        alpha : {None, float}, optional\n",
      "            If ``alpha`` is not None, return the ``100 * (1-alpha)%`` confidence\n",
      "            interval for `lmbda` as the third output argument.\n",
      "            Must be between 0.0 and 1.0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        boxcox : ndarray\n",
      "            Box-Cox power transformed array.\n",
      "        maxlog : float, optional\n",
      "            If the `lmbda` parameter is None, the second returned argument is\n",
      "            the lambda that maximizes the log-likelihood function.\n",
      "        (min_ci, max_ci) : tuple of float, optional\n",
      "            If `lmbda` parameter is None and ``alpha`` is not None, this returned\n",
      "            tuple of floats represents the minimum and maximum confidence limits\n",
      "            given ``alpha``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        probplot, boxcox_normplot, boxcox_normmax, boxcox_llf\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The Box-Cox transform is given by::\n",
      "        \n",
      "            y = (x**lmbda - 1) / lmbda,  for lmbda > 0\n",
      "                log(x),                  for lmbda = 0\n",
      "        \n",
      "        `boxcox` requires the input data to be positive.  Sometimes a Box-Cox\n",
      "        transformation provides a shift parameter to achieve this; `boxcox` does\n",
      "        not.  Such a shift parameter is equivalent to adding a positive constant to\n",
      "        `x` before calling `boxcox`.\n",
      "        \n",
      "        The confidence limits returned when ``alpha`` is provided give the interval\n",
      "        where:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            llf(\\hat{\\lambda}) - llf(\\lambda) < \\frac{1}{2}\\chi^2(1 - \\alpha, 1),\n",
      "        \n",
      "        with ``llf`` the log-likelihood function and :math:`\\chi^2` the chi-squared\n",
      "        function.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        G.E.P. Box and D.R. Cox, \"An Analysis of Transformations\", Journal of the\n",
      "        Royal Statistical Society B, 26, 211-252 (1964).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        We generate some random variates from a non-normal distribution and make a\n",
      "        probability plot for it, to show it is non-normal in the tails:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax1 = fig.add_subplot(211)\n",
      "        >>> x = stats.loggamma.rvs(5, size=500) + 5\n",
      "        >>> prob = stats.probplot(x, dist=stats.norm, plot=ax1)\n",
      "        >>> ax1.set_xlabel('')\n",
      "        >>> ax1.set_title('Probplot against normal distribution')\n",
      "        \n",
      "        We now use `boxcox` to transform the data so it's closest to normal:\n",
      "        \n",
      "        >>> ax2 = fig.add_subplot(212)\n",
      "        >>> xt, _ = stats.boxcox(x)\n",
      "        >>> prob = stats.probplot(xt, dist=stats.norm, plot=ax2)\n",
      "        >>> ax2.set_title('Probplot after Box-Cox transformation')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    boxcox_llf(lmb, data)\n",
      "        The boxcox log-likelihood function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        lmb : scalar\n",
      "            Parameter for Box-Cox transformation.  See `boxcox` for details.\n",
      "        data : array_like\n",
      "            Data to calculate Box-Cox log-likelihood for.  If `data` is\n",
      "            multi-dimensional, the log-likelihood is calculated along the first\n",
      "            axis.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        llf : float or ndarray\n",
      "            Box-Cox log-likelihood of `data` given `lmb`.  A float for 1-D `data`,\n",
      "            an array otherwise.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        boxcox, probplot, boxcox_normplot, boxcox_normmax\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The Box-Cox log-likelihood function is defined here as\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            llf = (\\lambda - 1) \\sum_i(\\log(x_i)) -\n",
      "                  N/2 \\log(\\sum_i (y_i - \\bar{y})^2 / N),\n",
      "        \n",
      "        where ``y`` is the Box-Cox transformed input data ``x``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
      "        >>> np.random.seed(1245)\n",
      "        \n",
      "        Generate some random variates and calculate Box-Cox log-likelihood values\n",
      "        for them for a range of ``lmbda`` values:\n",
      "        \n",
      "        >>> x = stats.loggamma.rvs(5, loc=10, size=1000)\n",
      "        >>> lmbdas = np.linspace(-2, 10)\n",
      "        >>> llf = np.zeros(lmbdas.shape, dtype=float)\n",
      "        >>> for ii, lmbda in enumerate(lmbdas):\n",
      "        ...     llf[ii] = stats.boxcox_llf(lmbda, x)\n",
      "        \n",
      "        Also find the optimal lmbda value with `boxcox`:\n",
      "        \n",
      "        >>> x_most_normal, lmbda_optimal = stats.boxcox(x)\n",
      "        \n",
      "        Plot the log-likelihood as function of lmbda.  Add the optimal lmbda as a\n",
      "        horizontal line to check that that's really the optimum:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(lmbdas, llf, 'b.-')\n",
      "        >>> ax.axhline(stats.boxcox_llf(lmbda_optimal, x), color='r')\n",
      "        >>> ax.set_xlabel('lmbda parameter')\n",
      "        >>> ax.set_ylabel('Box-Cox log-likelihood')\n",
      "        \n",
      "        Now add some probability plots to show that where the log-likelihood is\n",
      "        maximized the data transformed with `boxcox` looks closest to normal:\n",
      "        \n",
      "        >>> locs = [3, 10, 4]  # 'lower left', 'center', 'lower right'\n",
      "        >>> for lmbda, loc in zip([-1, lmbda_optimal, 9], locs):\n",
      "        ...     xt = stats.boxcox(x, lmbda=lmbda)\n",
      "        ...     (osm, osr), (slope, intercept, r_sq) = stats.probplot(xt)\n",
      "        ...     ax_inset = inset_axes(ax, width=\"20%\", height=\"20%\", loc=loc)\n",
      "        ...     ax_inset.plot(osm, osr, 'c.', osm, slope*osm + intercept, 'k-')\n",
      "        ...     ax_inset.set_xticklabels([])\n",
      "        ...     ax_inset.set_yticklabels([])\n",
      "        ...     ax_inset.set_title('$\\lambda=%1.2f$' % lmbda)\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    boxcox_normmax(x, brack=(-2.0, 2.0), method='pearsonr')\n",
      "        Compute optimal Box-Cox transform parameter for input data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        brack : 2-tuple, optional\n",
      "            The starting interval for a downhill bracket search with\n",
      "            `optimize.brent`.  Note that this is in most cases not critical; the\n",
      "            final result is allowed to be outside this bracket.\n",
      "        method : str, optional\n",
      "            The method to determine the optimal transform parameter (`boxcox`\n",
      "            ``lmbda`` parameter). Options are:\n",
      "        \n",
      "            'pearsonr'  (default)\n",
      "                Maximizes the Pearson correlation coefficient between\n",
      "                ``y = boxcox(x)`` and the expected values for ``y`` if `x` would be\n",
      "                normally-distributed.\n",
      "        \n",
      "            'mle'\n",
      "                Minimizes the log-likelihood `boxcox_llf`.  This is the method used\n",
      "                in `boxcox`.\n",
      "        \n",
      "            'all'\n",
      "                Use all optimization methods available, and return all results.\n",
      "                Useful to compare different methods.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        maxlog : float or ndarray\n",
      "            The optimal transform parameter found.  An array instead of a scalar\n",
      "            for ``method='all'``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        boxcox, boxcox_llf, boxcox_normplot\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> np.random.seed(1234)  # make this example reproducible\n",
      "        \n",
      "        Generate some data and determine optimal ``lmbda`` in various ways:\n",
      "        \n",
      "        >>> x = stats.loggamma.rvs(5, size=30) + 5\n",
      "        >>> y, lmax_mle = stats.boxcox(x)\n",
      "        >>> lmax_pearsonr = stats.boxcox_normmax(x)\n",
      "        \n",
      "        >>> lmax_mle\n",
      "        7.177...\n",
      "        >>> lmax_pearsonr\n",
      "        7.916...\n",
      "        >>> stats.boxcox_normmax(x, method='all')\n",
      "        array([ 7.91667384,  7.17718692])\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> prob = stats.boxcox_normplot(x, -10, 10, plot=ax)\n",
      "        >>> ax.axvline(lmax_mle, color='r')\n",
      "        >>> ax.axvline(lmax_pearsonr, color='g', ls='--')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    boxcox_normplot(x, la, lb, plot=None, N=80)\n",
      "        Compute parameters for a Box-Cox normality plot, optionally show it.\n",
      "        \n",
      "        A Box-Cox normality plot shows graphically what the best transformation\n",
      "        parameter is to use in `boxcox` to obtain a distribution that is close\n",
      "        to normal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        la, lb : scalar\n",
      "            The lower and upper bounds for the ``lmbda`` values to pass to `boxcox`\n",
      "            for Box-Cox transformations.  These are also the limits of the\n",
      "            horizontal axis of the plot if that is generated.\n",
      "        plot : object, optional\n",
      "            If given, plots the quantiles and least squares fit.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        N : int, optional\n",
      "            Number of points on the horizontal axis (equally distributed from\n",
      "            `la` to `lb`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lmbdas : ndarray\n",
      "            The ``lmbda`` values for which a Box-Cox transform was done.\n",
      "        ppcc : ndarray\n",
      "            Probability Plot Correlelation Coefficient, as obtained from `probplot`\n",
      "            when fitting the Box-Cox transformed input `x` against a normal\n",
      "            distribution.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        probplot, boxcox, boxcox_normmax, boxcox_llf, ppcc_max\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Even if `plot` is given, the figure is not shown or saved by\n",
      "        `boxcox_normplot`; ``plt.show()`` or ``plt.savefig('figname.png')``\n",
      "        should be used after calling `probplot`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        Generate some non-normally distributed data, and create a Box-Cox plot:\n",
      "        \n",
      "        >>> x = stats.loggamma.rvs(5, size=500) + 5\n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> prob = stats.boxcox_normplot(x, -20, 20, plot=ax)\n",
      "        \n",
      "        Determine and plot the optimal ``lmbda`` to transform ``x`` and plot it in\n",
      "        the same plot:\n",
      "        \n",
      "        >>> _, maxlog = stats.boxcox(x)\n",
      "        >>> ax.axvline(maxlog, color='r')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    chi2_contingency(observed, correction=True, lambda_=None)\n",
      "        Chi-square test of independence of variables in a contingency table.\n",
      "        \n",
      "        This function computes the chi-square statistic and p-value for the\n",
      "        hypothesis test of independence of the observed frequencies in the\n",
      "        contingency table [1]_ `observed`.  The expected frequencies are computed\n",
      "        based on the marginal sums under the assumption of independence; see\n",
      "        `scipy.stats.contingency.expected_freq`.  The number of degrees of\n",
      "        freedom is (expressed using numpy functions and attributes)::\n",
      "        \n",
      "            dof = observed.size - sum(observed.shape) + observed.ndim - 1\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        observed : array_like\n",
      "            The contingency table. The table contains the observed frequencies\n",
      "            (i.e. number of occurrences) in each category.  In the two-dimensional\n",
      "            case, the table is often described as an \"R x C table\".\n",
      "        correction : bool, optional\n",
      "            If True, *and* the degrees of freedom is 1, apply Yates' correction\n",
      "            for continuity.  The effect of the correction is to adjust each\n",
      "            observed value by 0.5 towards the corresponding expected value.\n",
      "        lambda_ : float or str, optional.\n",
      "            By default, the statistic computed in this test is Pearson's\n",
      "            chi-squared statistic [2]_.  `lambda_` allows a statistic from the\n",
      "            Cressie-Read power divergence family [3]_ to be used instead.  See\n",
      "            `power_divergence` for details.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        chi2 : float\n",
      "            The test statistic.\n",
      "        p : float\n",
      "            The p-value of the test\n",
      "        dof : int\n",
      "            Degrees of freedom\n",
      "        expected : ndarray, same shape as `observed`\n",
      "            The expected frequencies, based on the marginal sums of the table.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        contingency.expected_freq\n",
      "        fisher_exact\n",
      "        chisquare\n",
      "        power_divergence\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        An often quoted guideline for the validity of this calculation is that\n",
      "        the test should be used only if the observed and expected frequency in\n",
      "        each cell is at least 5.\n",
      "        \n",
      "        This is a test for the independence of different categories of a\n",
      "        population. The test is only meaningful when the dimension of\n",
      "        `observed` is two or more.  Applying the test to a one-dimensional\n",
      "        table will always result in `expected` equal to `observed` and a\n",
      "        chi-square statistic equal to 0.\n",
      "        \n",
      "        This function does not handle masked arrays, because the calculation\n",
      "        does not make sense with missing values.\n",
      "        \n",
      "        Like stats.chisquare, this function computes a chi-square statistic;\n",
      "        the convenience this function provides is to figure out the expected\n",
      "        frequencies and degrees of freedom from the given contingency table.\n",
      "        If these were already known, and if the Yates' correction was not\n",
      "        required, one could use stats.chisquare.  That is, if one calls::\n",
      "        \n",
      "            chi2, p, dof, ex = chi2_contingency(obs, correction=False)\n",
      "        \n",
      "        then the following is true::\n",
      "        \n",
      "            (chi2, p) == stats.chisquare(obs.ravel(), f_exp=ex.ravel(),\n",
      "                                         ddof=obs.size - 1 - dof)\n",
      "        \n",
      "        The `lambda_` argument was added in version 0.13.0 of scipy.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Contingency table\", http://en.wikipedia.org/wiki/Contingency_table\n",
      "        .. [2] \"Pearson's chi-squared test\",\n",
      "               http://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test\n",
      "        .. [3] Cressie, N. and Read, T. R. C., \"Multinomial Goodness-of-Fit\n",
      "               Tests\", J. Royal Stat. Soc. Series B, Vol. 46, No. 3 (1984),\n",
      "               pp. 440-464.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        A two-way example (2 x 3):\n",
      "        \n",
      "        >>> from scipy.stats import chi2_contingency\n",
      "        >>> obs = np.array([[10, 10, 20], [20, 20, 20]])\n",
      "        >>> chi2_contingency(obs)\n",
      "        (2.7777777777777777,\n",
      "         0.24935220877729619,\n",
      "         2,\n",
      "         array([[ 12.,  12.,  16.],\n",
      "                [ 18.,  18.,  24.]]))\n",
      "        \n",
      "        Perform the test using the log-likelihood ratio (i.e. the \"G-test\")\n",
      "        instead of Pearson's chi-squared statistic.\n",
      "        \n",
      "        >>> g, p, dof, expctd = chi2_contingency(obs, lambda_=\"log-likelihood\")\n",
      "        >>> g, p\n",
      "        (2.7688587616781319, 0.25046668010954165)\n",
      "        \n",
      "        A four-way example (2 x 2 x 2 x 2):\n",
      "        \n",
      "        >>> obs = np.array(\n",
      "        ...     [[[[12, 17],\n",
      "        ...        [11, 16]],\n",
      "        ...       [[11, 12],\n",
      "        ...        [15, 16]]],\n",
      "        ...      [[[23, 15],\n",
      "        ...        [30, 22]],\n",
      "        ...       [[14, 17],\n",
      "        ...        [15, 16]]]])\n",
      "        >>> chi2_contingency(obs)\n",
      "        (8.7584514426741897,\n",
      "         0.64417725029295503,\n",
      "         11,\n",
      "         array([[[[ 14.15462386,  14.15462386],\n",
      "                  [ 16.49423111,  16.49423111]],\n",
      "                 [[ 11.2461395 ,  11.2461395 ],\n",
      "                  [ 13.10500554,  13.10500554]]],\n",
      "                [[[ 19.5591166 ,  19.5591166 ],\n",
      "                  [ 22.79202844,  22.79202844]],\n",
      "                 [[ 15.54012004,  15.54012004],\n",
      "                  [ 18.10873492,  18.10873492]]]]))\n",
      "    \n",
      "    chisqprob(*args, **kwds)\n",
      "        `chisqprob` is deprecated!\n",
      "        stats.chisqprob is deprecated in scipy 0.17.0; use stats.distributions.chi2.sf instead.\n",
      "        \n",
      "        \n",
      "            Probability value (1-tail) for the Chi^2 probability distribution.\n",
      "        \n",
      "            Broadcasting rules apply.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            chisq : array_like or float > 0\n",
      "        \n",
      "            df : array_like or float, probably int >= 1\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            chisqprob : ndarray\n",
      "                The area from `chisq` to infinity under the Chi^2 probability\n",
      "                distribution with degrees of freedom `df`.\n",
      "    \n",
      "    chisquare(f_obs, f_exp=None, ddof=0, axis=0)\n",
      "        Calculates a one-way chi square test.\n",
      "        \n",
      "        The chi square test tests the null hypothesis that the categorical data\n",
      "        has the given frequencies.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f_obs : array_like\n",
      "            Observed frequencies in each category.\n",
      "        f_exp : array_like, optional\n",
      "            Expected frequencies in each category.  By default the categories are\n",
      "            assumed to be equally likely.\n",
      "        ddof : int, optional\n",
      "            \"Delta degrees of freedom\": adjustment to the degrees of freedom\n",
      "            for the p-value.  The p-value is computed using a chi-squared\n",
      "            distribution with ``k - 1 - ddof`` degrees of freedom, where `k`\n",
      "            is the number of observed frequencies.  The default value of `ddof`\n",
      "            is 0.\n",
      "        axis : int or None, optional\n",
      "            The axis of the broadcast result of `f_obs` and `f_exp` along which to\n",
      "            apply the test.  If axis is None, all values in `f_obs` are treated\n",
      "            as a single data set.  Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        chisq : float or ndarray\n",
      "            The chi-squared test statistic.  The value is a float if `axis` is\n",
      "            None or `f_obs` and `f_exp` are 1-D.\n",
      "        p : float or ndarray\n",
      "            The p-value of the test.  The value is a float if `ddof` and the\n",
      "            return value `chisq` are scalars.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        power_divergence\n",
      "        mstats.chisquare\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This test is invalid when the observed or expected frequencies in each\n",
      "        category are too small.  A typical rule is that all of the observed\n",
      "        and expected frequencies should be at least 5.\n",
      "        \n",
      "        The default degrees of freedom, k-1, are for the case when no parameters\n",
      "        of the distribution are estimated. If p parameters are estimated by\n",
      "        efficient maximum likelihood then the correct degrees of freedom are\n",
      "        k-1-p. If the parameters are estimated in a different way, then the\n",
      "        dof can be between k-1-p and k-1. However, it is also possible that\n",
      "        the asymptotic distribution is not a chisquare, in which case this\n",
      "        test is not appropriate.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
      "               Statistics\". Chapter 8. http://faculty.vassar.edu/lowry/ch8pt1.html\n",
      "        .. [2] \"Chi-squared test\", http://en.wikipedia.org/wiki/Chi-squared_test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        When just `f_obs` is given, it is assumed that the expected frequencies\n",
      "        are uniform and given by the mean of the observed frequencies.\n",
      "        \n",
      "        >>> from scipy.stats import chisquare\n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12])\n",
      "        (2.0, 0.84914503608460956)\n",
      "        \n",
      "        With `f_exp` the expected frequencies can be given.\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12], f_exp=[16, 16, 16, 16, 16, 8])\n",
      "        (3.5, 0.62338762774958223)\n",
      "        \n",
      "        When `f_obs` is 2-D, by default the test is applied to each column.\n",
      "        \n",
      "        >>> obs = np.array([[16, 18, 16, 14, 12, 12], [32, 24, 16, 28, 20, 24]]).T\n",
      "        >>> obs.shape\n",
      "        (6, 2)\n",
      "        >>> chisquare(obs)\n",
      "        (array([ 2.        ,  6.66666667]), array([ 0.84914504,  0.24663415]))\n",
      "        \n",
      "        By setting ``axis=None``, the test is applied to all data in the array,\n",
      "        which is equivalent to applying the test to the flattened array.\n",
      "        \n",
      "        >>> chisquare(obs, axis=None)\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        >>> chisquare(obs.ravel())\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        \n",
      "        `ddof` is the change to make to the default degrees of freedom.\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12], ddof=1)\n",
      "        (2.0, 0.73575888234288467)\n",
      "        \n",
      "        The calculation of the p-values is done by broadcasting the\n",
      "        chi-squared statistic with `ddof`.\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12], ddof=[0,1,2])\n",
      "        (2.0, array([ 0.84914504,  0.73575888,  0.5724067 ]))\n",
      "        \n",
      "        `f_obs` and `f_exp` are also broadcast.  In the following, `f_obs` has\n",
      "        shape (6,) and `f_exp` has shape (2, 6), so the result of broadcasting\n",
      "        `f_obs` and `f_exp` has shape (2, 6).  To compute the desired chi-squared\n",
      "        statistics, we use ``axis=1``:\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12],\n",
      "        ...           f_exp=[[16, 16, 16, 16, 16, 8], [8, 20, 20, 16, 12, 12]],\n",
      "        ...           axis=1)\n",
      "        (array([ 3.5 ,  9.25]), array([ 0.62338763,  0.09949846]))\n",
      "    \n",
      "    circmean(samples, high=6.283185307179586, low=0, axis=None)\n",
      "        Compute the circular mean for samples in a range.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : array_like\n",
      "            Input array.\n",
      "        high : float or int, optional\n",
      "            High boundary for circular mean range.  Default is ``2*pi``.\n",
      "        low : float or int, optional\n",
      "            Low boundary for circular mean range.  Default is 0.\n",
      "        axis : int, optional\n",
      "            Axis along which means are computed.  The default is to compute\n",
      "            the mean of the flattened array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        circmean : float\n",
      "            Circular mean.\n",
      "    \n",
      "    circstd(samples, high=6.283185307179586, low=0, axis=None)\n",
      "        Compute the circular standard deviation for samples assumed to be in the\n",
      "        range [low to high].\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : array_like\n",
      "            Input array.\n",
      "        low : float or int, optional\n",
      "            Low boundary for circular standard deviation range.  Default is 0.\n",
      "        high : float or int, optional\n",
      "            High boundary for circular standard deviation range.\n",
      "            Default is ``2*pi``.\n",
      "        axis : int, optional\n",
      "            Axis along which standard deviations are computed.  The default is\n",
      "            to compute the standard deviation of the flattened array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        circstd : float\n",
      "            Circular standard deviation.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This uses a definition of circular standard deviation that in the limit of\n",
      "        small angles returns a number close to the 'linear' standard deviation.\n",
      "    \n",
      "    circvar(samples, high=6.283185307179586, low=0, axis=None)\n",
      "        Compute the circular variance for samples assumed to be in a range\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : array_like\n",
      "            Input array.\n",
      "        low : float or int, optional\n",
      "            Low boundary for circular variance range.  Default is 0.\n",
      "        high : float or int, optional\n",
      "            High boundary for circular variance range.  Default is ``2*pi``.\n",
      "        axis : int, optional\n",
      "            Axis along which variances are computed.  The default is to compute\n",
      "            the variance of the flattened array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        circvar : float\n",
      "            Circular variance.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This uses a definition of circular variance that in the limit of small\n",
      "        angles returns a number close to the 'linear' variance.\n",
      "    \n",
      "    combine_pvalues(pvalues, method='fisher', weights=None)\n",
      "        Methods for combining the p-values of independent tests bearing upon the\n",
      "        same hypothesis.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pvalues : array_like, 1-D\n",
      "            Array of p-values assumed to come from independent tests.\n",
      "        method : {'fisher', 'stouffer'}, optional\n",
      "            Name of method to use to combine p-values. The following methods are\n",
      "            available:\n",
      "        \n",
      "            - \"fisher\": Fisher's method (Fisher's combined probability test),\n",
      "              the default.\n",
      "            - \"stouffer\": Stouffer's Z-score method.\n",
      "        weights : array_like, 1-D, optional\n",
      "            Optional array of weights used only for Stouffer's Z-score method.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic: float\n",
      "            The statistic calculated by the specified method:\n",
      "            - \"fisher\": The chi-squared statistic\n",
      "            - \"stouffer\": The Z-score\n",
      "        pval: float\n",
      "            The combined p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Fisher's method (also known as Fisher's combined probability test) [1]_ uses\n",
      "        a chi-squared statistic to compute a combined p-value. The closely related\n",
      "        Stouffer's Z-score method [2]_ uses Z-scores rather than p-values. The\n",
      "        advantage of Stouffer's method is that it is straightforward to introduce\n",
      "        weights, which can make Stouffer's method more powerful than Fisher's\n",
      "        method when the p-values are from studies of different size [3]_ [4]_.\n",
      "        \n",
      "        Fisher's method may be extended to combine p-values from dependent tests\n",
      "        [5]_. Extensions such as Brown's method and Kost's method are not currently\n",
      "        implemented.\n",
      "        \n",
      "        .. versionadded:: 0.15.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Fisher%27s_method\n",
      "        .. [2] http://en.wikipedia.org/wiki/Fisher's_method#Relation_to_Stouffer.27s_Z-score_method\n",
      "        .. [3] Whitlock, M. C. \"Combining probability from independent tests: the\n",
      "               weighted Z-method is superior to Fisher's approach.\" Journal of\n",
      "               Evolutionary Biology 18, no. 5 (2005): 1368-1373.\n",
      "        .. [4] Zaykin, Dmitri V. \"Optimally weighted Z-test is a powerful method\n",
      "               for combining probabilities in meta-analysis.\" Journal of\n",
      "               Evolutionary Biology 24, no. 8 (2011): 1836-1841.\n",
      "        .. [5] https://en.wikipedia.org/wiki/Extensions_of_Fisher%27s_method\n",
      "    \n",
      "    cumfreq(a, numbins=10, defaultreallimits=None, weights=None)\n",
      "        Returns a cumulative frequency histogram, using the histogram function.\n",
      "        \n",
      "        A cumulative histogram is a mapping that counts the cumulative number of\n",
      "        observations in all of the bins up to the specified bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        numbins : int, optional\n",
      "            The number of bins to use for the histogram. Default is 10.\n",
      "        defaultreallimits : tuple (lower, upper), optional\n",
      "            The lower and upper values for the range of the histogram.\n",
      "            If no value is given, a range slightly larger than the range of the\n",
      "            values in `a` is used. Specifically ``(a.min() - s, a.max() + s)``,\n",
      "            where ``s = (1/2)(a.max() - a.min()) / (numbins - 1)``.\n",
      "        weights : array_like, optional\n",
      "            The weights for each value in `a`. Default is None, which gives each\n",
      "            value a weight of 1.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        cumcount : ndarray\n",
      "            Binned values of cumulative frequency.\n",
      "        lowerlimit : float\n",
      "            Lower real limit\n",
      "        binsize : float\n",
      "            Width of each bin.\n",
      "        extrapoints : int\n",
      "            Extra points.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy import stats\n",
      "        >>> x = [1, 4, 2, 1, 3, 1]\n",
      "        >>> res = stats.cumfreq(x, numbins=4, defaultreallimits=(1.5, 5))\n",
      "        >>> res.cumcount\n",
      "        array([ 1.,  2.,  3.,  3.])\n",
      "        >>> res.extrapoints\n",
      "        3\n",
      "        \n",
      "        Create a normal distribution with 1000 random values\n",
      "        \n",
      "        >>> rng = np.random.RandomState(seed=12345)\n",
      "        >>> samples = stats.norm.rvs(size=1000, random_state=rng)\n",
      "        \n",
      "        Calculate cumulative frequencies\n",
      "        \n",
      "        >>> res = stats.cumfreq(samples, numbins=25)\n",
      "        \n",
      "        Calculate space of values for x\n",
      "        \n",
      "        >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.cumcount.size,\n",
      "        ...                                  res.cumcount.size)\n",
      "        \n",
      "        Plot histogram and cumulative histogram\n",
      "        \n",
      "        >>> fig = plt.figure(figsize=(10, 4))\n",
      "        >>> ax1 = fig.add_subplot(1, 2, 1)\n",
      "        >>> ax2 = fig.add_subplot(1, 2, 2)\n",
      "        >>> ax1.hist(samples, bins=25)\n",
      "        >>> ax1.set_title('Histogram')\n",
      "        >>> ax2.bar(x, res.cumcount, width=res.binsize)\n",
      "        >>> ax2.set_title('Cumulative histogram')\n",
      "        >>> ax2.set_xlim([x.min(), x.max()])\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    describe(a, axis=0, ddof=1, bias=True, nan_policy='propagate')\n",
      "        Computes several descriptive statistics of the passed array.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "           Input data.\n",
      "        axis : int or None, optional\n",
      "           Axis along which statistics are calculated. Default is 0.\n",
      "           If None, compute over the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom (only for variance).  Default is 1.\n",
      "        bias : bool, optional\n",
      "            If False, then the skewness and kurtosis calculations are corrected for\n",
      "            statistical bias.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        nobs : int\n",
      "           Number of observations (length of data along `axis`).\n",
      "        minmax: tuple of ndarrays or floats\n",
      "           Minimum and maximum value of data array.\n",
      "        mean : ndarray or float\n",
      "           Arithmetic mean of data along axis.\n",
      "        variance : ndarray or float\n",
      "           Unbiased variance of the data along axis, denominator is number of\n",
      "           observations minus one.\n",
      "        skewness : ndarray or float\n",
      "           Skewness, based on moment calculations with denominator equal to\n",
      "           the number of observations, i.e. no degrees of freedom correction.\n",
      "        kurtosis : ndarray or float\n",
      "           Kurtosis (Fisher).  The kurtosis is normalized so that it is\n",
      "           zero for the normal distribution.  No degrees of freedom are used.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        skew, kurtosis\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(10)\n",
      "        >>> stats.describe(a)\n",
      "        DescribeResult(nobs=10, minmax=(0, 9), mean=4.5, variance=9.1666666666666661,\n",
      "                       skewness=0.0, kurtosis=-1.2242424242424244)\n",
      "        >>> b = [[1, 2], [3, 4]]\n",
      "        >>> stats.describe(b)\n",
      "        DescribeResult(nobs=2, minmax=(array([1, 2]), array([3, 4])),\n",
      "                       mean=array([ 2., 3.]), variance=array([ 2., 2.]),\n",
      "                       skewness=array([ 0., 0.]), kurtosis=array([-2., -2.]))\n",
      "    \n",
      "    entropy(pk, qk=None, base=None)\n",
      "        Calculate the entropy of a distribution for given probability values.\n",
      "        \n",
      "        If only probabilities `pk` are given, the entropy is calculated as\n",
      "        ``S = -sum(pk * log(pk), axis=0)``.\n",
      "        \n",
      "        If `qk` is not None, then compute the Kullback-Leibler divergence\n",
      "        ``S = sum(pk * log(pk / qk), axis=0)``.\n",
      "        \n",
      "        This routine will normalize `pk` and `qk` if they don't sum to 1.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pk : sequence\n",
      "            Defines the (discrete) distribution. ``pk[i]`` is the (possibly\n",
      "            unnormalized) probability of event ``i``.\n",
      "        qk : sequence, optional\n",
      "            Sequence against which the relative entropy is computed. Should be in\n",
      "            the same format as `pk`.\n",
      "        base : float, optional\n",
      "            The logarithmic base to use, defaults to ``e`` (natural logarithm).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        S : float\n",
      "            The calculated entropy.\n",
      "    \n",
      "    f_oneway(*args)\n",
      "        Performs a 1-way ANOVA.\n",
      "        \n",
      "        The one-way ANOVA tests the null hypothesis that two or more groups have\n",
      "        the same population mean.  The test is applied to samples from two or\n",
      "        more groups, possibly with differing sizes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The sample measurements for each group.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed F-value of the test.\n",
      "        pvalue : float\n",
      "            The associated p-value from the F-distribution.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The ANOVA test has important assumptions that must be satisfied in order\n",
      "        for the associated p-value to be valid.\n",
      "        \n",
      "        1. The samples are independent.\n",
      "        2. Each sample is from a normally distributed population.\n",
      "        3. The population standard deviations of the groups are all equal.  This\n",
      "           property is known as homoscedasticity.\n",
      "        \n",
      "        If these assumptions are not true for a given set of data, it may still be\n",
      "        possible to use the Kruskal-Wallis H-test (`scipy.stats.kruskal`) although\n",
      "        with some loss of power.\n",
      "        \n",
      "        The algorithm is from Heiman[2], pp.394-7.\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
      "               Statistics\". Chapter 14.\n",
      "               http://faculty.vassar.edu/lowry/ch14pt1.html\n",
      "        \n",
      "        .. [2] Heiman, G.W.  Research Methods in Statistics. 2002.\n",
      "        \n",
      "        .. [3] McDonald, G. H. \"Handbook of Biological Statistics\", One-way ANOVA.\n",
      "               http://www.biostathandbook.com/onewayanova.html\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import scipy.stats as stats\n",
      "        \n",
      "        [3]_ Here are some data on a shell measurement (the length of the anterior\n",
      "        adductor muscle scar, standardized by dividing by length) in the mussel\n",
      "        Mytilus trossulus from five locations: Tillamook, Oregon; Newport, Oregon;\n",
      "        Petersburg, Alaska; Magadan, Russia; and Tvarminne, Finland, taken from a\n",
      "        much larger data set used in McDonald et al. (1991).\n",
      "        \n",
      "        >>> tillamook = [0.0571, 0.0813, 0.0831, 0.0976, 0.0817, 0.0859, 0.0735,\n",
      "        ...              0.0659, 0.0923, 0.0836]\n",
      "        >>> newport = [0.0873, 0.0662, 0.0672, 0.0819, 0.0749, 0.0649, 0.0835,\n",
      "        ...            0.0725]\n",
      "        >>> petersburg = [0.0974, 0.1352, 0.0817, 0.1016, 0.0968, 0.1064, 0.105]\n",
      "        >>> magadan = [0.1033, 0.0915, 0.0781, 0.0685, 0.0677, 0.0697, 0.0764,\n",
      "        ...            0.0689]\n",
      "        >>> tvarminne = [0.0703, 0.1026, 0.0956, 0.0973, 0.1039, 0.1045]\n",
      "        >>> stats.f_oneway(tillamook, newport, petersburg, magadan, tvarminne)\n",
      "        (7.1210194716424473, 0.00028122423145345439)\n",
      "    \n",
      "    f_value(*args, **kwds)\n",
      "        `f_value` is deprecated!\n",
      "        stats.f_value deprecated in scipy 0.17.0\n",
      "        \n",
      "        \n",
      "            Returns an F-statistic for a restricted vs. unrestricted model.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            ER : float\n",
      "                 `ER` is the sum of squared residuals for the restricted model\n",
      "                  or null hypothesis\n",
      "        \n",
      "            EF : float\n",
      "                 `EF` is the sum of squared residuals for the unrestricted model\n",
      "                  or alternate hypothesis\n",
      "        \n",
      "            dfR : int\n",
      "                  `dfR` is the degrees of freedom in the restricted model\n",
      "        \n",
      "            dfF : int\n",
      "                  `dfF` is the degrees of freedom in the unrestricted model\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            F-statistic : float\n",
      "    \n",
      "    f_value_multivariate(*args, **kwds)\n",
      "        `f_value_multivariate` is deprecated!\n",
      "        stats.f_value_multivariate deprecated in scipy 0.17.0\n",
      "        \n",
      "        \n",
      "            Returns a multivariate F-statistic.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            ER : ndarray\n",
      "                Error associated with the null hypothesis (the Restricted model).\n",
      "                From a multivariate F calculation.\n",
      "            EF : ndarray\n",
      "                Error associated with the alternate hypothesis (the Full model)\n",
      "                From a multivariate F calculation.\n",
      "            dfnum : int\n",
      "                Degrees of freedom the Restricted model.\n",
      "            dfden : int\n",
      "                Degrees of freedom associated with the Restricted model.\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            fstat : float\n",
      "                The computed F-statistic.\n",
      "    \n",
      "    f_value_wilks_lambda(*args, **kwds)\n",
      "        `f_value_wilks_lambda` is deprecated!\n",
      "        stats.f_value_wilks_lambda deprecated in scipy 0.17.0\n",
      "        \n",
      "        Calculation of Wilks lambda F-statistic for multivarite data, per\n",
      "            Maxwell & Delaney p.657.\n",
      "    \n",
      "    fastsort(*args, **kwds)\n",
      "        `fastsort` is deprecated!\n",
      "        scipy.stats.fastsort is deprecated in scipy 0.16.0\n",
      "        \n",
      "        \n",
      "            Sort an array and provide the argsort.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            a : array_like\n",
      "                Input array.\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            fastsort : ndarray of type int\n",
      "                sorted indices into the original array\n",
      "    \n",
      "    find_repeats(arr)\n",
      "        Find repeats and repeat counts.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        arr : array_like\n",
      "            Input array. This is cast to float64.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        values : ndarray\n",
      "            The unique values from the (flattened) input that are repeated.\n",
      "        \n",
      "        counts : ndarray\n",
      "            Number of times the corresponding 'value' is repeated.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        In numpy >= 1.9 `numpy.unique` provides similar functionality. The main\n",
      "        difference is that `find_repeats` only returns repeated values.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.find_repeats([2, 1, 2, 3, 2, 2, 5])\n",
      "        RepeatedResults(values=array([ 2.]), counts=array([4]))\n",
      "        \n",
      "        >>> stats.find_repeats([[10, 20, 1, 2], [5, 5, 4, 4]])\n",
      "        RepeatedResults(values=array([ 4.,  5.]), counts=array([2, 2]))\n",
      "    \n",
      "    fisher_exact(table, alternative='two-sided')\n",
      "        Performs a Fisher exact test on a 2x2 contingency table.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        table : array_like of ints\n",
      "            A 2x2 contingency table.  Elements should be non-negative integers.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Which alternative hypothesis to the null hypothesis the test uses.\n",
      "            Default is 'two-sided'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        oddsratio : float\n",
      "            This is prior odds ratio and not a posterior estimate.\n",
      "        p_value : float\n",
      "            P-value, the probability of obtaining a distribution at least as\n",
      "            extreme as the one that was actually observed, assuming that the\n",
      "            null hypothesis is true.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        chi2_contingency : Chi-square test of independence of variables in a\n",
      "            contingency table.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The calculated odds ratio is different from the one R uses. This scipy\n",
      "        implementation returns the (more common) \"unconditional Maximum\n",
      "        Likelihood Estimate\", while R uses the \"conditional Maximum Likelihood\n",
      "        Estimate\".\n",
      "        \n",
      "        For tables with large numbers, the (inexact) chi-square test implemented\n",
      "        in the function `chi2_contingency` can also be used.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Say we spend a few days counting whales and sharks in the Atlantic and\n",
      "        Indian oceans. In the Atlantic ocean we find 8 whales and 1 shark, in the\n",
      "        Indian ocean 2 whales and 5 sharks. Then our contingency table is::\n",
      "        \n",
      "                    Atlantic  Indian\n",
      "            whales     8        2\n",
      "            sharks     1        5\n",
      "        \n",
      "        We use this table to find the p-value:\n",
      "        \n",
      "        >>> import scipy.stats as stats\n",
      "        >>> oddsratio, pvalue = stats.fisher_exact([[8, 2], [1, 5]])\n",
      "        >>> pvalue\n",
      "        0.0349...\n",
      "        \n",
      "        The probability that we would observe this or an even more imbalanced ratio\n",
      "        by chance is about 3.5%.  A commonly used significance level is 5%--if we\n",
      "        adopt that, we can therefore conclude that our observed imbalance is\n",
      "        statistically significant; whales prefer the Atlantic while sharks prefer\n",
      "        the Indian ocean.\n",
      "    \n",
      "    fligner(*args, **kwds)\n",
      "        Perform Fligner-Killeen test for equality of variance.\n",
      "        \n",
      "        Fligner's test tests the null hypothesis that all input samples\n",
      "        are from populations with equal variances.  Fligner-Killeen's test is\n",
      "        distribution free when populations are identical [2]_.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            Arrays of sample data.  Need not be the same length.\n",
      "        center : {'mean', 'median', 'trimmed'}, optional\n",
      "            Keyword argument controlling which function of the data is used in\n",
      "            computing the test statistic.  The default is 'median'.\n",
      "        proportiontocut : float, optional\n",
      "            When `center` is 'trimmed', this gives the proportion of data points\n",
      "            to cut from each end. (See `scipy.stats.trim_mean`.)\n",
      "            Default is 0.05.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        bartlett : A parametric test for equality of k variances in normal samples\n",
      "        levene : A robust parametric test for equality of k variances\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        As with Levene's test there are three variants of Fligner's test that\n",
      "        differ by the measure of central tendency used in the test.  See `levene`\n",
      "        for more information.\n",
      "        \n",
      "        Conover et al. (1981) examine many of the existing parametric and\n",
      "        nonparametric tests by extensive simulations and they conclude that the\n",
      "        tests proposed by Fligner and Killeen (1976) and Levene (1960) appear to be\n",
      "        superior in terms of robustness of departures from normality and power [3]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://www.stat.psu.edu/~bgl/center/tr/TR993.ps\n",
      "        \n",
      "        .. [2] Fligner, M.A. and Killeen, T.J. (1976). Distribution-free two-sample\n",
      "               tests for scale. 'Journal of the American Statistical Association.'\n",
      "               71(353), 210-213.\n",
      "        \n",
      "        .. [3] Park, C. and Lindsay, B. G. (1999). Robust Scale Estimation and\n",
      "               Hypothesis Testing based on Quadratic Inference Function. Technical\n",
      "               Report #99-03, Center for Likelihood Studies, Pennsylvania State\n",
      "               University.\n",
      "        \n",
      "        .. [4] Conover, W. J., Johnson, M. E. and Johnson M. M. (1981). A\n",
      "               comparative study of tests for homogeneity of variances, with\n",
      "               applications to the outer continental shelf biding data.\n",
      "               Technometrics, 23(4), 351-361.\n",
      "    \n",
      "    friedmanchisquare(*args)\n",
      "        Computes the Friedman test for repeated measurements\n",
      "        \n",
      "        The Friedman test tests the null hypothesis that repeated measurements of\n",
      "        the same individuals have the same distribution.  It is often used\n",
      "        to test for consistency among measurements obtained in different ways.\n",
      "        For example, if two measurement techniques are used on the same set of\n",
      "        individuals, the Friedman test can be used to determine if the two\n",
      "        measurement techniques are consistent.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        measurements1, measurements2, measurements3... : array_like\n",
      "            Arrays of measurements.  All of the arrays must have the same number\n",
      "            of elements.  At least 3 sets of measurements must be given.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            the test statistic, correcting for ties\n",
      "        pvalue : float\n",
      "            the associated p-value assuming that the test statistic has a chi\n",
      "            squared distribution\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Due to the assumption that the test statistic has a chi squared\n",
      "        distribution, the p-value is only reliable for n > 10 and more than\n",
      "        6 repeated measurements.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://en.wikipedia.org/wiki/Friedman_test\n",
      "    \n",
      "    gmean(a, axis=0, dtype=None)\n",
      "        Compute the geometric mean along the specified axis.\n",
      "        \n",
      "        Returns the geometric average of the array elements.\n",
      "        That is:  n-th root of (x1 * x2 * ... * xn)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array or object that can be converted to an array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the geometric mean is computed. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        dtype : dtype, optional\n",
      "            Type of the returned array and of the accumulator in which the\n",
      "            elements are summed. If dtype is not specified, it defaults to the\n",
      "            dtype of a, unless a has an integer dtype with a precision less than\n",
      "            that of the default platform integer. In that case, the default\n",
      "            platform integer is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        gmean : ndarray\n",
      "            see dtype parameter above\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.mean : Arithmetic average\n",
      "        numpy.average : Weighted average\n",
      "        hmean : Harmonic mean\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The geometric average is computed over a single dimension of the input\n",
      "        array, axis=0 by default, or all values in the array if axis=None.\n",
      "        float64 intermediate and return values are used for integer inputs.\n",
      "        \n",
      "        Use masked arrays to ignore any non-finite values in the input or that\n",
      "        arise in the calculations such as Not a Number and infinity because masked\n",
      "        arrays automatically mask any non-finite values.\n",
      "    \n",
      "    histogram(*args, **kwds)\n",
      "        `histogram` is deprecated!\n",
      "        scipy.stats.histogram is deprecated in scipy 0.17.0; use np.histogram instead\n",
      "    \n",
      "    histogram2(*args, **kwds)\n",
      "        `histogram2` is deprecated!\n",
      "        scipy.stats.histogram2 is deprecated in scipy 0.16.0; use np.histogram2d instead\n",
      "        \n",
      "        \n",
      "            Compute histogram using divisions in bins.\n",
      "        \n",
      "            Count the number of times values from array `a` fall into\n",
      "            numerical ranges defined by `bins`.  Range x is given by\n",
      "            bins[x] <= range_x < bins[x+1] where x =0,N and N is the\n",
      "            length of the `bins` array.  The last range is given by\n",
      "            bins[N] <= range_N < infinity.  Values less than bins[0] are\n",
      "            not included in the histogram.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            a : array_like of rank 1\n",
      "                The array of values to be assigned into bins\n",
      "            bins : array_like of rank 1\n",
      "                Defines the ranges of values to use during histogramming.\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            histogram2 : ndarray of rank 1\n",
      "                Each value represents the occurrences for a given bin (range) of\n",
      "                values.\n",
      "    \n",
      "    hmean(a, axis=0, dtype=None)\n",
      "        Calculates the harmonic mean along the specified axis.\n",
      "        \n",
      "        That is:  n / (1/x1 + 1/x2 + ... + 1/xn)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array, masked array or object that can be converted to an array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the harmonic mean is computed. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        dtype : dtype, optional\n",
      "            Type of the returned array and of the accumulator in which the\n",
      "            elements are summed. If `dtype` is not specified, it defaults to the\n",
      "            dtype of `a`, unless `a` has an integer `dtype` with a precision less\n",
      "            than that of the default platform integer. In that case, the default\n",
      "            platform integer is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        hmean : ndarray\n",
      "            see `dtype` parameter above\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.mean : Arithmetic average\n",
      "        numpy.average : Weighted average\n",
      "        gmean : Geometric mean\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The harmonic mean is computed over a single dimension of the input\n",
      "        array, axis=0 by default, or all values in the array if axis=None.\n",
      "        float64 intermediate and return values are used for integer inputs.\n",
      "        \n",
      "        Use masked arrays to ignore any non-finite values in the input or that\n",
      "        arise in the calculations such as Not a Number and infinity.\n",
      "    \n",
      "    iqr(x, axis=None, rng=(25, 75), scale='raw', nan_policy='propagate', interpolation='linear', keepdims=False)\n",
      "        Compute the interquartile range of the data along the specified\n",
      "        axis.\n",
      "        \n",
      "        The interquartile range (IQR) is the difference between the 75th and\n",
      "        25th percentile of the data. It is a measure of the dispersion\n",
      "        similar to standard deviation or variance, but is much more robust\n",
      "        against outliers [2]_.\n",
      "        \n",
      "        The ``rng`` parameter allows this function to compute other\n",
      "        percentile ranges than the actual IQR. For example, setting\n",
      "        ``rng=(0, 100)`` is equivalent to `numpy.ptp`.\n",
      "        \n",
      "        The IQR of an empty array is `np.nan`.\n",
      "        \n",
      "        .. versionadded:: 0.18.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array or object that can be converted to an array.\n",
      "        axis : int or sequence of int, optional\n",
      "            Axis along which the range is computed. The default is to\n",
      "            compute the IQR for the entire array.\n",
      "        rng : Two-element sequence containing floats in range of [0,100] optional\n",
      "            Percentiles over which to compute the range. Each must be\n",
      "            between 0 and 100, inclusive. The default is the true IQR:\n",
      "            `(25, 75)`. The order of the elements is not important.\n",
      "        scale : scalar or str, optional\n",
      "            The numerical value of scale will be divided out of the final\n",
      "            result. The following string values are recognized:\n",
      "        \n",
      "              'raw' : No scaling, just return the raw IQR.\n",
      "              'normal' : Scale by :math:`2 \\sqrt{2} erf^{-1}(\\frac{1}{2}) \\approx 1.349`.\n",
      "        \n",
      "            The default is 'raw'. Array-like scale is also allowed, as long\n",
      "            as it broadcasts correctly to the output such that\n",
      "            ``out / scale`` is a valid operation. The output dimensions\n",
      "            depend on the input array, `x`, the `axis` argument, and the\n",
      "            `keepdims` flag.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate'\n",
      "            returns nan, 'raise' throws an error, 'omit' performs the\n",
      "            calculations ignoring nan values. Default is 'propagate'.\n",
      "        interpolation : {'linear', 'lower', 'higher', 'midpoint', 'nearest'}, optional\n",
      "            Specifies the interpolation method to use when the percentile\n",
      "            boundaries lie between two data points `i` and `j`:\n",
      "        \n",
      "              * 'linear' : `i + (j - i) * fraction`, where `fraction` is the\n",
      "                  fractional part of the index surrounded by `i` and `j`.\n",
      "              * 'lower' : `i`.\n",
      "              * 'higher' : `j`.\n",
      "              * 'nearest' : `i` or `j` whichever is nearest.\n",
      "              * 'midpoint' : `(i + j) / 2`.\n",
      "        \n",
      "            Default is 'linear'.\n",
      "        keepdims : bool, optional\n",
      "            If this is set to `True`, the reduced axes are left in the\n",
      "            result as dimensions with size one. With this option, the result\n",
      "            will broadcast correctly against the original array `x`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        iqr : scalar or ndarray\n",
      "            If ``axis=None``, a scalar is returned. If the input contains\n",
      "            integers or floats of smaller precision than ``np.float64``, then the\n",
      "            output data-type is ``np.float64``. Otherwise, the output data-type is\n",
      "            the same as that of the input.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.std, numpy.var\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import iqr\n",
      "        >>> x = np.array([[10, 7, 4], [3, 2, 1]])\n",
      "        >>> x\n",
      "        array([[10,  7,  4],\n",
      "               [ 3,  2,  1]])\n",
      "        >>> iqr(x)\n",
      "        4.0\n",
      "        >>> iqr(x, axis=0)\n",
      "        array([ 3.5,  2.5,  1.5])\n",
      "        >>> iqr(x, axis=1)\n",
      "        array([ 3.,  1.])\n",
      "        >>> iqr(x, axis=1, keepdims=True)\n",
      "        array([[ 3.],\n",
      "               [ 1.]])\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function is heavily dependent on the version of `numpy` that is\n",
      "        installed. Versions greater than 1.11.0b3 are highly recommended, as they\n",
      "        include a number of enhancements and fixes to `numpy.percentile` and\n",
      "        `numpy.nanpercentile` that affect the operation of this function. The\n",
      "        following modifications apply:\n",
      "        \n",
      "        Below 1.10.0 : `nan_policy` is poorly defined.\n",
      "            The default behavior of `numpy.percentile` is used for 'propagate'. This\n",
      "            is a hybrid of 'omit' and 'propagate' that mostly yields a skewed\n",
      "            version of 'omit' since NaNs are sorted to the end of the data. A\n",
      "            warning is raised if there are NaNs in the data.\n",
      "        Below 1.9.0: `numpy.nanpercentile` does not exist.\n",
      "            This means that `numpy.percentile` is used regardless of `nan_policy`\n",
      "            and a warning is issued. See previous item for a description of the\n",
      "            behavior.\n",
      "        Below 1.9.0: `keepdims` and `interpolation` are not supported.\n",
      "            The keywords get ignored with a warning if supplied with non-default\n",
      "            values. However, multiple axes are still supported.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Interquartile range\" https://en.wikipedia.org/wiki/Interquartile_range\n",
      "        .. [2] \"Robust measures of scale\" https://en.wikipedia.org/wiki/Robust_measures_of_scale\n",
      "        .. [3] \"Quantile\" https://en.wikipedia.org/wiki/Quantile\n",
      "    \n",
      "    itemfreq(a)\n",
      "        Returns a 2-D array of item frequencies.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : (N,) array_like\n",
      "            Input array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        itemfreq : (K, 2) ndarray\n",
      "            A 2-D frequency table.  Column 1 contains sorted, unique values from\n",
      "            `a`, column 2 contains their respective counts.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.array([1, 1, 5, 0, 1, 2, 2, 0, 1, 4])\n",
      "        >>> stats.itemfreq(a)\n",
      "        array([[ 0.,  2.],\n",
      "               [ 1.,  4.],\n",
      "               [ 2.,  2.],\n",
      "               [ 4.,  1.],\n",
      "               [ 5.,  1.]])\n",
      "        >>> np.bincount(a)\n",
      "        array([2, 4, 2, 0, 1, 1])\n",
      "        \n",
      "        >>> stats.itemfreq(a/10.)\n",
      "        array([[ 0. ,  2. ],\n",
      "               [ 0.1,  4. ],\n",
      "               [ 0.2,  2. ],\n",
      "               [ 0.4,  1. ],\n",
      "               [ 0.5,  1. ]])\n",
      "    \n",
      "    jarque_bera(x)\n",
      "        Perform the Jarque-Bera goodness of fit test on sample data.\n",
      "        \n",
      "        The Jarque-Bera test tests whether the sample data has the skewness and\n",
      "        kurtosis matching a normal distribution.\n",
      "        \n",
      "        Note that this test only works for a large enough number of data samples\n",
      "        (>2000) as the test statistic asymptotically has a Chi-squared distribution\n",
      "        with 2 degrees of freedom.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Observations of a random variable.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        jb_value : float\n",
      "            The test statistic.\n",
      "        p : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Jarque, C. and Bera, A. (1980) \"Efficient tests for normality,\n",
      "               homoscedasticity and serial independence of regression residuals\",\n",
      "               6 Econometric Letters 255-259.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(987654321)\n",
      "        >>> x = np.random.normal(0, 1, 100000)\n",
      "        >>> y = np.random.rayleigh(1, 100000)\n",
      "        >>> stats.jarque_bera(x)\n",
      "        (4.7165707989581342, 0.09458225503041906)\n",
      "        >>> stats.jarque_bera(y)\n",
      "        (6713.7098548143422, 0.0)\n",
      "    \n",
      "    kendalltau(x, y, initial_lexsort=None, nan_policy='propagate')\n",
      "        Calculates Kendall's tau, a correlation measure for ordinal data.\n",
      "        \n",
      "        Kendall's tau is a measure of the correspondence between two rankings.\n",
      "        Values close to 1 indicate strong agreement, values close to -1 indicate\n",
      "        strong disagreement.  This is the tau-b version of Kendall's tau which\n",
      "        accounts for ties.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Arrays of rankings, of the same shape. If arrays are not 1-D, they will\n",
      "            be flattened to 1-D.\n",
      "        initial_lexsort : bool, optional\n",
      "            Whether to use lexsort or quicksort as the sorting method for the\n",
      "            initial sort of the inputs. Default is lexsort (True), for which\n",
      "            `kendalltau` is of complexity O(n log(n)). If False, the complexity is\n",
      "            O(n^2), but with a smaller pre-factor (so quicksort may be faster for\n",
      "            small arrays).\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float\n",
      "           The tau statistic.\n",
      "        pvalue : float\n",
      "           The two-sided p-value for a hypothesis test whose null hypothesis is\n",
      "           an absence of association, tau = 0.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        spearmanr : Calculates a Spearman rank-order correlation coefficient.\n",
      "        theilslopes : Computes the Theil-Sen estimator for a set of points (x, y).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The definition of Kendall's tau that is used is::\n",
      "        \n",
      "          tau = (P - Q) / sqrt((P + Q + T) * (P + Q + U))\n",
      "        \n",
      "        where P is the number of concordant pairs, Q the number of discordant\n",
      "        pairs, T the number of ties only in `x`, and U the number of ties only in\n",
      "        `y`.  If a tie occurs for the same pair in both `x` and `y`, it is not\n",
      "        added to either T or U.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        W.R. Knight, \"A Computer Method for Calculating Kendall's Tau with\n",
      "        Ungrouped Data\", Journal of the American Statistical Association, Vol. 61,\n",
      "        No. 314, Part 1, pp. 436-439, 1966.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x1 = [12, 2, 1, 12, 2]\n",
      "        >>> x2 = [1, 4, 7, 1, 0]\n",
      "        >>> tau, p_value = stats.kendalltau(x1, x2)\n",
      "        >>> tau\n",
      "        -0.47140452079103173\n",
      "        >>> p_value\n",
      "        0.24821309157521476\n",
      "    \n",
      "    kruskal(*args, **kwargs)\n",
      "        Compute the Kruskal-Wallis H-test for independent samples\n",
      "        \n",
      "        The Kruskal-Wallis H-test tests the null hypothesis that the population\n",
      "        median of all of the groups are equal.  It is a non-parametric version of\n",
      "        ANOVA.  The test works on 2 or more independent samples, which may have\n",
      "        different sizes.  Note that rejecting the null hypothesis does not\n",
      "        indicate which of the groups differs.  Post-hoc comparisons between\n",
      "        groups are required to determine which groups are different.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "           Two or more arrays with the sample measurements can be given as\n",
      "           arguments.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "           The Kruskal-Wallis H statistic, corrected for ties\n",
      "        pvalue : float\n",
      "           The p-value for the test using the assumption that H has a chi\n",
      "           square distribution\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        f_oneway : 1-way ANOVA\n",
      "        mannwhitneyu : Mann-Whitney rank test on two samples.\n",
      "        friedmanchisquare : Friedman test for repeated measurements\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Due to the assumption that H has a chi square distribution, the number\n",
      "        of samples in each group must not be too small.  A typical rule is\n",
      "        that each sample must have at least 5 measurements.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] W. H. Kruskal & W. W. Wallis, \"Use of Ranks in\n",
      "           One-Criterion Variance Analysis\", Journal of the American Statistical\n",
      "           Association, Vol. 47, Issue 260, pp. 583-621, 1952.\n",
      "        .. [2] http://en.wikipedia.org/wiki/Kruskal-Wallis_one-way_analysis_of_variance\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = [1, 3, 5, 7, 9]\n",
      "        >>> y = [2, 4, 6, 8, 10]\n",
      "        >>> stats.kruskal(x, y)\n",
      "        KruskalResult(statistic=0.27272727272727337, pvalue=0.60150813444058948)\n",
      "        \n",
      "        >>> x = [1, 1, 1]\n",
      "        >>> y = [2, 2, 2]\n",
      "        >>> z = [2, 2]\n",
      "        >>> stats.kruskal(x, y, z)\n",
      "        KruskalResult(statistic=7.0, pvalue=0.030197383422318501)\n",
      "    \n",
      "    ks_2samp(data1, data2)\n",
      "        Computes the Kolmogorov-Smirnov statistic on 2 samples.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that 2 independent samples\n",
      "        are drawn from the same continuous distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data1, data2 : sequence of 1-D ndarrays\n",
      "            two arrays of sample observations assumed to be drawn from a continuous\n",
      "            distribution, sample sizes can be different\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            KS statistic\n",
      "        pvalue : float\n",
      "            two-tailed p-value\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This tests whether 2 samples are drawn from the same distribution. Note\n",
      "        that, like in the case of the one-sample K-S test, the distribution is\n",
      "        assumed to be continuous.\n",
      "        \n",
      "        This is the two-sided test, one-sided tests are not implemented.\n",
      "        The test uses the two-sided asymptotic Kolmogorov-Smirnov distribution.\n",
      "        \n",
      "        If the K-S statistic is small or the p-value is high, then we cannot\n",
      "        reject the hypothesis that the distributions of the two samples\n",
      "        are the same.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(12345678)  #fix random seed to get the same result\n",
      "        >>> n1 = 200  # size of first sample\n",
      "        >>> n2 = 300  # size of second sample\n",
      "        \n",
      "        For a different distribution, we can reject the null hypothesis since the\n",
      "        pvalue is below 1%:\n",
      "        \n",
      "        >>> rvs1 = stats.norm.rvs(size=n1, loc=0., scale=1)\n",
      "        >>> rvs2 = stats.norm.rvs(size=n2, loc=0.5, scale=1.5)\n",
      "        >>> stats.ks_2samp(rvs1, rvs2)\n",
      "        (0.20833333333333337, 4.6674975515806989e-005)\n",
      "        \n",
      "        For a slightly different distribution, we cannot reject the null hypothesis\n",
      "        at a 10% or lower alpha since the p-value at 0.144 is higher than 10%\n",
      "        \n",
      "        >>> rvs3 = stats.norm.rvs(size=n2, loc=0.01, scale=1.0)\n",
      "        >>> stats.ks_2samp(rvs1, rvs3)\n",
      "        (0.10333333333333333, 0.14498781825751686)\n",
      "        \n",
      "        For an identical distribution, we cannot reject the null hypothesis since\n",
      "        the p-value is high, 41%:\n",
      "        \n",
      "        >>> rvs4 = stats.norm.rvs(size=n2, loc=0.0, scale=1.0)\n",
      "        >>> stats.ks_2samp(rvs1, rvs4)\n",
      "        (0.07999999999999996, 0.41126949729859719)\n",
      "    \n",
      "    kstat(data, n=2)\n",
      "        Return the nth k-statistic (1<=n<=4 so far).\n",
      "        \n",
      "        The nth k-statistic k_n is the unique symmetric unbiased estimator of the\n",
      "        nth cumulant kappa_n.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input array. Note that n-D input gets flattened.\n",
      "        n : int, {1, 2, 3, 4}, optional\n",
      "            Default is equal to 2.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        kstat : float\n",
      "            The nth k-statistic.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstatvar: Returns an unbiased estimator of the variance of the k-statistic.\n",
      "        moment: Returns the n-th central moment about the mean for a sample.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For a sample size n, the first few k-statistics are given by:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            k_{1} = \\mu\n",
      "            k_{2} = \\frac{n}{n-1} m_{2}\n",
      "            k_{3} = \\frac{ n^{2} } {(n-1) (n-2)} m_{3}\n",
      "            k_{4} = \\frac{ n^{2} [(n + 1)m_{4} - 3(n - 1) m^2_{2}]} {(n-1) (n-2) (n-3)}\n",
      "        \n",
      "        where ``:math:\\mu`` is the sample mean, ``:math:m_2`` is the sample\n",
      "        variance, and ``:math:m_i`` is the i-th sample central moment.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        http://mathworld.wolfram.com/k-Statistic.html\n",
      "        \n",
      "        http://mathworld.wolfram.com/Cumulant.html\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rndm = np.random.RandomState(1234)\n",
      "        \n",
      "        As sample size increases, n-th moment and n-th k-statistic converge to the\n",
      "        same number (although they aren't identical). In the case of the normal\n",
      "        distribution, they converge to zero.\n",
      "        \n",
      "        >>> for n in [2, 3, 4, 5, 6, 7]:\n",
      "        ...     x = rndm.normal(size=10**n)\n",
      "        ...     m, k = stats.moment(x, 3), stats.kstat(x, 3)\n",
      "        ...     print(\"%.3g %.3g %.3g\" % (m, k, m-k))\n",
      "        -0.631 -0.651 0.0194\n",
      "        0.0282 0.0283 -8.49e-05\n",
      "        -0.0454 -0.0454 1.36e-05\n",
      "        7.53e-05 7.53e-05 -2.26e-09\n",
      "        0.00166 0.00166 -4.99e-09\n",
      "        -2.88e-06 -2.88e-06 8.63e-13\n",
      "    \n",
      "    kstatvar(data, n=2)\n",
      "        Returns an unbiased estimator of the variance of the k-statistic.\n",
      "        \n",
      "        See `kstat` for more details of the k-statistic.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input array. Note that n-D input gets flattened.\n",
      "        n : int, {1, 2}, optional\n",
      "            Default is equal to 2.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        kstatvar : float\n",
      "            The nth k-statistic variance.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstat: Returns the n-th k-statistic.\n",
      "        moment: Returns the n-th central moment about the mean for a sample.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The variances of the first few k-statistics are given by:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            var(k_{1}) = \\frac{\\kappa^2}{n}\n",
      "            var(k_{2}) = \\frac{\\kappa^4}{n} + \\frac{2\\kappa^2_{2}}{n - 1}\n",
      "            var(k_{3}) = \\frac{\\kappa^6}{n} + \\frac{9 \\kappa_2 \\kappa_4}{n - 1} +\n",
      "                         \\frac{9 \\kappa^2_{3}}{n - 1} +\n",
      "                         \\frac{6 n \\kappa^3_{2}}{(n-1) (n-2)}\n",
      "            var(k_{4}) = \\frac{\\kappa^8}{n} + \\frac{16 \\kappa_2 \\kappa_6}{n - 1} +\n",
      "                         \\frac{48 \\kappa_{3} \\kappa_5}{n - 1} +\n",
      "                         \\frac{34 \\kappa^2_{4}}{n-1} + \\frac{72 n \\kappa^2_{2} \\kappa_4}{(n - 1) (n - 2)} +\n",
      "                         \\frac{144 n \\kappa_{2} \\kappa^2_{3}}{(n - 1) (n - 2)} +\n",
      "                         \\frac{24 (n + 1) n \\kappa^4_{2}}{(n - 1) (n - 2) (n - 3)}\n",
      "    \n",
      "    kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx')\n",
      "        Perform the Kolmogorov-Smirnov test for goodness of fit.\n",
      "        \n",
      "        This performs a test of the distribution G(x) of an observed\n",
      "        random variable against a given distribution F(x). Under the null\n",
      "        hypothesis the two distributions are identical, G(x)=F(x). The\n",
      "        alternative hypothesis can be either 'two-sided' (default), 'less'\n",
      "        or 'greater'. The KS test is only valid for continuous distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rvs : str, array or callable\n",
      "            If a string, it should be the name of a distribution in `scipy.stats`.\n",
      "            If an array, it should be a 1-D array of observations of random\n",
      "            variables.\n",
      "            If a callable, it should be a function to generate random variables;\n",
      "            it is required to have a keyword argument `size`.\n",
      "        cdf : str or callable\n",
      "            If a string, it should be the name of a distribution in `scipy.stats`.\n",
      "            If `rvs` is a string then `cdf` can be False or the same as `rvs`.\n",
      "            If a callable, that callable is used to calculate the cdf.\n",
      "        args : tuple, sequence, optional\n",
      "            Distribution parameters, used if `rvs` or `cdf` are strings.\n",
      "        N : int, optional\n",
      "            Sample size if `rvs` is string or callable.  Default is 20.\n",
      "        alternative : {'two-sided', 'less','greater'}, optional\n",
      "            Defines the alternative hypothesis (see explanation above).\n",
      "            Default is 'two-sided'.\n",
      "        mode : 'approx' (default) or 'asymp', optional\n",
      "            Defines the distribution used for calculating the p-value.\n",
      "        \n",
      "              - 'approx' : use approximation to exact distribution of test statistic\n",
      "              - 'asymp' : use asymptotic distribution of test statistic\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            KS test statistic, either D, D+ or D-.\n",
      "        pvalue :  float\n",
      "            One-tailed or two-tailed p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        In the one-sided test, the alternative is that the empirical\n",
      "        cumulative distribution function of the random variable is \"less\"\n",
      "        or \"greater\" than the cumulative distribution function F(x) of the\n",
      "        hypothesis, ``G(x)<=F(x)``, resp. ``G(x)>=F(x)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        \n",
      "        >>> x = np.linspace(-15, 15, 9)\n",
      "        >>> stats.kstest(x, 'norm')\n",
      "        (0.44435602715924361, 0.038850142705171065)\n",
      "        \n",
      "        >>> np.random.seed(987654321) # set random seed to get the same result\n",
      "        >>> stats.kstest('norm', False, N=100)\n",
      "        (0.058352892479417884, 0.88531190944151261)\n",
      "        \n",
      "        The above lines are equivalent to:\n",
      "        \n",
      "        >>> np.random.seed(987654321)\n",
      "        >>> stats.kstest(stats.norm.rvs(size=100), 'norm')\n",
      "        (0.058352892479417884, 0.88531190944151261)\n",
      "        \n",
      "        *Test against one-sided alternative hypothesis*\n",
      "        \n",
      "        Shift distribution to larger values, so that ``cdf_dgp(x) < norm.cdf(x)``:\n",
      "        \n",
      "        >>> np.random.seed(987654321)\n",
      "        >>> x = stats.norm.rvs(loc=0.2, size=100)\n",
      "        >>> stats.kstest(x,'norm', alternative = 'less')\n",
      "        (0.12464329735846891, 0.040989164077641749)\n",
      "        \n",
      "        Reject equal distribution against alternative hypothesis: less\n",
      "        \n",
      "        >>> stats.kstest(x,'norm', alternative = 'greater')\n",
      "        (0.0072115233216311081, 0.98531158590396395)\n",
      "        \n",
      "        Don't reject equal distribution against alternative hypothesis: greater\n",
      "        \n",
      "        >>> stats.kstest(x,'norm', mode='asymp')\n",
      "        (0.12464329735846891, 0.08944488871182088)\n",
      "        \n",
      "        *Testing t distributed random variables against normal distribution*\n",
      "        \n",
      "        With 100 degrees of freedom the t distribution looks close to the normal\n",
      "        distribution, and the K-S test does not reject the hypothesis that the\n",
      "        sample came from the normal distribution:\n",
      "        \n",
      "        >>> np.random.seed(987654321)\n",
      "        >>> stats.kstest(stats.t.rvs(100,size=100),'norm')\n",
      "        (0.072018929165471257, 0.67630062862479168)\n",
      "        \n",
      "        With 3 degrees of freedom the t distribution looks sufficiently different\n",
      "        from the normal distribution, that we can reject the hypothesis that the\n",
      "        sample came from the normal distribution at the 10% level:\n",
      "        \n",
      "        >>> np.random.seed(987654321)\n",
      "        >>> stats.kstest(stats.t.rvs(3,size=100),'norm')\n",
      "        (0.131016895759829, 0.058826222555312224)\n",
      "    \n",
      "    kurtosis(a, axis=0, fisher=True, bias=True, nan_policy='propagate')\n",
      "        Computes the kurtosis (Fisher or Pearson) of a dataset.\n",
      "        \n",
      "        Kurtosis is the fourth central moment divided by the square of the\n",
      "        variance. If Fisher's definition is used, then 3.0 is subtracted from\n",
      "        the result to give 0.0 for a normal distribution.\n",
      "        \n",
      "        If bias is False then the kurtosis is calculated using k statistics to\n",
      "        eliminate bias coming from biased moment estimators\n",
      "        \n",
      "        Use `kurtosistest` to see if result is close enough to normal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array\n",
      "            data for which the kurtosis is calculated\n",
      "        axis : int or None, optional\n",
      "            Axis along which the kurtosis is calculated. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        fisher : bool, optional\n",
      "            If True, Fisher's definition is used (normal ==> 0.0). If False,\n",
      "            Pearson's definition is used (normal ==> 3.0).\n",
      "        bias : bool, optional\n",
      "            If False, then the calculations are corrected for statistical bias.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        kurtosis : array\n",
      "            The kurtosis of values along an axis. If all values are equal,\n",
      "            return -3 for Fisher's definition and 0 for Pearson's definition.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "    \n",
      "    kurtosistest(a, axis=0, nan_policy='propagate')\n",
      "        Tests whether a dataset has normal kurtosis\n",
      "        \n",
      "        This function tests the null hypothesis that the kurtosis\n",
      "        of the population from which the sample was drawn is that\n",
      "        of the normal distribution: ``kurtosis = 3(n-1)/(n+1)``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array\n",
      "            array of the sample data\n",
      "        axis : int or None, optional\n",
      "           Axis along which to compute test. Default is 0. If None,\n",
      "           compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed z-score for this test.\n",
      "        pvalue : float\n",
      "            The 2-sided p-value for the hypothesis test\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Valid only for n>20.  The Z-score is set to 0 for bad entries.\n",
      "        This function uses the method described in [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] see e.g. F. J. Anscombe, W. J. Glynn, \"Distribution of the kurtosis\n",
      "           statistic b2 for normal samples\", Biometrika, vol. 70, pp. 227-234, 1983.\n",
      "    \n",
      "    levene(*args, **kwds)\n",
      "        Perform Levene test for equal variances.\n",
      "        \n",
      "        The Levene test tests the null hypothesis that all input samples\n",
      "        are from populations with equal variances.  Levene's test is an\n",
      "        alternative to Bartlett's test `bartlett` in the case where\n",
      "        there are significant deviations from normality.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The sample data, possibly with different lengths\n",
      "        center : {'mean', 'median', 'trimmed'}, optional\n",
      "            Which function of the data to use in the test.  The default\n",
      "            is 'median'.\n",
      "        proportiontocut : float, optional\n",
      "            When `center` is 'trimmed', this gives the proportion of data points\n",
      "            to cut from each end. (See `scipy.stats.trim_mean`.)\n",
      "            Default is 0.05.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The p-value for the test.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Three variations of Levene's test are possible.  The possibilities\n",
      "        and their recommended usages are:\n",
      "        \n",
      "          * 'median' : Recommended for skewed (non-normal) distributions>\n",
      "          * 'mean' : Recommended for symmetric, moderate-tailed distributions.\n",
      "          * 'trimmed' : Recommended for heavy-tailed distributions.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1]  http://www.itl.nist.gov/div898/handbook/eda/section3/eda35a.htm\n",
      "        .. [2]   Levene, H. (1960). In Contributions to Probability and Statistics:\n",
      "                   Essays in Honor of Harold Hotelling, I. Olkin et al. eds.,\n",
      "                   Stanford University Press, pp. 278-292.\n",
      "        .. [3]  Brown, M. B. and Forsythe, A. B. (1974), Journal of the American\n",
      "                  Statistical Association, 69, 364-367\n",
      "    \n",
      "    linregress(x, y=None)\n",
      "        Calculate a linear least-squares regression for two sets of measurements.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Two sets of measurements.  Both arrays should have the same length.\n",
      "            If only x is given (and y=None), then it must be a two-dimensional\n",
      "            array where one dimension has length 2.  The two sets of measurements\n",
      "            are then found by splitting the array along the length-2 dimension.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        slope : float\n",
      "            slope of the regression line\n",
      "        intercept : float\n",
      "            intercept of the regression line\n",
      "        rvalue : float\n",
      "            correlation coefficient\n",
      "        pvalue : float\n",
      "            two-sided p-value for a hypothesis test whose null hypothesis is\n",
      "            that the slope is zero.\n",
      "        stderr : float\n",
      "            Standard error of the estimated gradient.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        optimize.curve_fit : Use non-linear least squares to fit a function to data.\n",
      "        optimize.leastsq : Minimize the sum of squares of a set of equations.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(12345678)\n",
      "        >>> x = np.random.random(10)\n",
      "        >>> y = np.random.random(10)\n",
      "        >>> slope, intercept, r_value, p_value, std_err = stats.linregress(x,y)\n",
      "        \n",
      "        # To get coefficient of determination (r_squared)\n",
      "        \n",
      "        >>> print(\"r-squared:\", r_value**2)\n",
      "        ('r-squared:', 0.080402268539028335)\n",
      "    \n",
      "    mannwhitneyu(x, y, use_continuity=True, alternative=None)\n",
      "        Computes the Mann-Whitney rank test on samples x and y.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Array of samples, should be one-dimensional.\n",
      "        use_continuity : bool, optional\n",
      "                Whether a continuity correction (1/2.) should be taken into\n",
      "                account. Default is True.\n",
      "        alternative : None (deprecated), 'less', 'two-sided', or 'greater'\n",
      "                Whether to get the p-value for the one-sided hypothesis ('less'\n",
      "                or 'greater') or for the two-sided hypothesis ('two-sided').\n",
      "                Defaults to None, which results in a p-value half the size of\n",
      "                the 'two-sided' p-value and a different U statistic. The\n",
      "                default behavior is not the same as using 'less' or 'greater':\n",
      "                it only exists for backward compatibility and is deprecated.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The Mann-Whitney U statistic, equal to min(U for x, U for y) if\n",
      "            `alternative` is equal to None (deprecated; exists for backward\n",
      "            compatibility), and U for y otherwise.\n",
      "        pvalue : float\n",
      "            p-value assuming an asymptotic normal distribution. One-sided or\n",
      "            two-sided, depending on the choice of `alternative`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Use only when the number of observation in each sample is > 20 and\n",
      "        you have 2 independent samples of ranks. Mann-Whitney U is\n",
      "        significant if the u-obtained is LESS THAN or equal to the critical\n",
      "        value of U.\n",
      "        \n",
      "        This test corrects for ties and by default uses a continuity correction.\n",
      "    \n",
      "    median_test(*args, **kwds)\n",
      "        Mood's median test.\n",
      "        \n",
      "        Test that two or more samples come from populations with the same median.\n",
      "        \n",
      "        Let ``n = len(args)`` be the number of samples.  The \"grand median\" of\n",
      "        all the data is computed, and a contingency table is formed by\n",
      "        classifying the values in each sample as being above or below the grand\n",
      "        median.  The contingency table, along with `correction` and `lambda_`,\n",
      "        are passed to `scipy.stats.chi2_contingency` to compute the test statistic\n",
      "        and p-value.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The set of samples.  There must be at least two samples.\n",
      "            Each sample must be a one-dimensional sequence containing at least\n",
      "            one value.  The samples are not required to have the same length.\n",
      "        ties : str, optional\n",
      "            Determines how values equal to the grand median are classified in\n",
      "            the contingency table.  The string must be one of::\n",
      "        \n",
      "                \"below\":\n",
      "                    Values equal to the grand median are counted as \"below\".\n",
      "                \"above\":\n",
      "                    Values equal to the grand median are counted as \"above\".\n",
      "                \"ignore\":\n",
      "                    Values equal to the grand median are not counted.\n",
      "        \n",
      "            The default is \"below\".\n",
      "        correction : bool, optional\n",
      "            If True, *and* there are just two samples, apply Yates' correction\n",
      "            for continuity when computing the test statistic associated with\n",
      "            the contingency table.  Default is True.\n",
      "        lambda_ : float or str, optional.\n",
      "            By default, the statistic computed in this test is Pearson's\n",
      "            chi-squared statistic.  `lambda_` allows a statistic from the\n",
      "            Cressie-Read power divergence family to be used instead.  See\n",
      "            `power_divergence` for details.\n",
      "            Default is 1 (Pearson's chi-squared statistic).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        stat : float\n",
      "            The test statistic.  The statistic that is returned is determined by\n",
      "            `lambda_`.  The default is Pearson's chi-squared statistic.\n",
      "        p : float\n",
      "            The p-value of the test.\n",
      "        m : float\n",
      "            The grand median.\n",
      "        table : ndarray\n",
      "            The contingency table.  The shape of the table is (2, n), where\n",
      "            n is the number of samples.  The first row holds the counts of the\n",
      "            values above the grand median, and the second row holds the counts\n",
      "            of the values below the grand median.  The table allows further\n",
      "            analysis with, for example, `scipy.stats.chi2_contingency`, or with\n",
      "            `scipy.stats.fisher_exact` if there are two samples, without having\n",
      "            to recompute the table.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kruskal : Compute the Kruskal-Wallis H-test for independent samples.\n",
      "        mannwhitneyu : Computes the Mann-Whitney rank test on samples x and y.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 0.15.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Mood, A. M., Introduction to the Theory of Statistics. McGraw-Hill\n",
      "            (1950), pp. 394-399.\n",
      "        .. [2] Zar, J. H., Biostatistical Analysis, 5th ed. Prentice Hall (2010).\n",
      "            See Sections 8.12 and 10.15.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        A biologist runs an experiment in which there are three groups of plants.\n",
      "        Group 1 has 16 plants, group 2 has 15 plants, and group 3 has 17 plants.\n",
      "        Each plant produces a number of seeds.  The seed counts for each group\n",
      "        are::\n",
      "        \n",
      "            Group 1: 10 14 14 18 20 22 24 25 31 31 32 39 43 43 48 49\n",
      "            Group 2: 28 30 31 33 34 35 36 40 44 55 57 61 91 92 99\n",
      "            Group 3:  0  3  9 22 23 25 25 33 34 34 40 45 46 48 62 67 84\n",
      "        \n",
      "        The following code applies Mood's median test to these samples.\n",
      "        \n",
      "        >>> g1 = [10, 14, 14, 18, 20, 22, 24, 25, 31, 31, 32, 39, 43, 43, 48, 49]\n",
      "        >>> g2 = [28, 30, 31, 33, 34, 35, 36, 40, 44, 55, 57, 61, 91, 92, 99]\n",
      "        >>> g3 = [0, 3, 9, 22, 23, 25, 25, 33, 34, 34, 40, 45, 46, 48, 62, 67, 84]\n",
      "        >>> from scipy.stats import median_test\n",
      "        >>> stat, p, med, tbl = median_test(g1, g2, g3)\n",
      "        \n",
      "        The median is\n",
      "        \n",
      "        >>> med\n",
      "        34.0\n",
      "        \n",
      "        and the contingency table is\n",
      "        \n",
      "        >>> tbl\n",
      "        array([[ 5, 10,  7],\n",
      "               [11,  5, 10]])\n",
      "        \n",
      "        `p` is too large to conclude that the medians are not the same:\n",
      "        \n",
      "        >>> p\n",
      "        0.12609082774093244\n",
      "        \n",
      "        The \"G-test\" can be performed by passing ``lambda_=\"log-likelihood\"`` to\n",
      "        `median_test`.\n",
      "        \n",
      "        >>> g, p, med, tbl = median_test(g1, g2, g3, lambda_=\"log-likelihood\")\n",
      "        >>> p\n",
      "        0.12224779737117837\n",
      "        \n",
      "        The median occurs several times in the data, so we'll get a different\n",
      "        result if, for example, ``ties=\"above\"`` is used:\n",
      "        \n",
      "        >>> stat, p, med, tbl = median_test(g1, g2, g3, ties=\"above\")\n",
      "        >>> p\n",
      "        0.063873276069553273\n",
      "        \n",
      "        >>> tbl\n",
      "        array([[ 5, 11,  9],\n",
      "               [11,  4,  8]])\n",
      "        \n",
      "        This example demonstrates that if the data set is not large and there\n",
      "        are values equal to the median, the p-value can be sensitive to the\n",
      "        choice of `ties`.\n",
      "    \n",
      "    mode(a, axis=0, nan_policy='propagate')\n",
      "        Returns an array of the modal (most common) value in the passed array.\n",
      "        \n",
      "        If there is more than one such value, only the first is returned.\n",
      "        The bin-count for the modal bins is also returned.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            n-dimensional array of which to find mode(s).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mode : ndarray\n",
      "            Array of modal values.\n",
      "        count : ndarray\n",
      "            Array of counts for each mode.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> a = np.array([[6, 8, 3, 0],\n",
      "        ...               [3, 2, 1, 7],\n",
      "        ...               [8, 1, 8, 4],\n",
      "        ...               [5, 3, 0, 5],\n",
      "        ...               [4, 7, 5, 9]])\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.mode(a)\n",
      "        (array([[3, 1, 0, 0]]), array([[1, 1, 1, 1]]))\n",
      "        \n",
      "        To get mode of whole array, specify ``axis=None``:\n",
      "        \n",
      "        >>> stats.mode(a, axis=None)\n",
      "        (array([3]), array([3]))\n",
      "    \n",
      "    moment(a, moment=1, axis=0, nan_policy='propagate')\n",
      "        Calculates the nth moment about the mean for a sample.\n",
      "        \n",
      "        A moment is a specific quantitative measure of the shape of a set of points.\n",
      "        It is often used to calculate coefficients of skewness and kurtosis due\n",
      "        to its close relationship with them.\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "           data\n",
      "        moment : int or array_like of ints, optional\n",
      "           order of central moment that is returned. Default is 1.\n",
      "        axis : int or None, optional\n",
      "           Axis along which the central moment is computed. Default is 0.\n",
      "           If None, compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        n-th central moment : ndarray or float\n",
      "           The appropriate moment along the given axis or over all values if axis\n",
      "           is None. The denominator for the moment calculation is the number of\n",
      "           observations, no degrees of freedom correction is done.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        kurtosis, skew, describe\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The k-th central moment of a data sample is:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            m_k = \\frac{1}{n} \\sum_{i = 1}^n (x_i - \\bar{x})^k\n",
      "        \n",
      "        Where n is the number of samples and x-bar is the mean. This function uses\n",
      "        exponentiation by squares [1]_ for efficiency.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://eli.thegreenplace.net/2009/03/21/efficient-integer-exponentiation-algorithms\n",
      "    \n",
      "    mood(x, y, axis=0)\n",
      "        Perform Mood's test for equal scale parameters.\n",
      "        \n",
      "        Mood's two-sample test for scale parameters is a non-parametric\n",
      "        test for the null hypothesis that two samples are drawn from the\n",
      "        same distribution with the same scale parameter.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Arrays of sample data.\n",
      "        axis : int, optional\n",
      "            The axis along which the samples are tested.  `x` and `y` can be of\n",
      "            different length along `axis`.\n",
      "            If `axis` is None, `x` and `y` are flattened and the test is done on\n",
      "            all values in the flattened arrays.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        z : scalar or ndarray\n",
      "            The z-score for the hypothesis test.  For 1-D inputs a scalar is\n",
      "            returned.\n",
      "        p-value : scalar ndarray\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fligner : A non-parametric test for the equality of k variances\n",
      "        ansari : A non-parametric test for the equality of 2 variances\n",
      "        bartlett : A parametric test for equality of k variances in normal samples\n",
      "        levene : A parametric test for equality of k variances\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The data are assumed to be drawn from probability distributions ``f(x)``\n",
      "        and ``f(x/s) / s`` respectively, for some probability density function f.\n",
      "        The null hypothesis is that ``s == 1``.\n",
      "        \n",
      "        For multi-dimensional arrays, if the inputs are of shapes\n",
      "        ``(n0, n1, n2, n3)``  and ``(n0, m1, n2, n3)``, then if ``axis=1``, the\n",
      "        resulting z and p values will have shape ``(n0, n2, n3)``.  Note that\n",
      "        ``n1`` and ``m1`` don't have to be equal, but the other dimensions do.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(1234)\n",
      "        >>> x2 = np.random.randn(2, 45, 6, 7)\n",
      "        >>> x1 = np.random.randn(2, 30, 6, 7)\n",
      "        >>> z, p = stats.mood(x1, x2, axis=1)\n",
      "        >>> p.shape\n",
      "        (2, 6, 7)\n",
      "        \n",
      "        Find the number of points where the difference in scale is not significant:\n",
      "        \n",
      "        >>> (p > 0.1).sum()\n",
      "        74\n",
      "        \n",
      "        Perform the test with different scales:\n",
      "        \n",
      "        >>> x1 = np.random.randn(2, 30)\n",
      "        >>> x2 = np.random.randn(2, 35) * 10.0\n",
      "        >>> stats.mood(x1, x2, axis=1)\n",
      "        (array([-5.7178125 , -5.25342163]), array([  1.07904114e-08,   1.49299218e-07]))\n",
      "    \n",
      "    mvsdist(data)\n",
      "        'Frozen' distributions for mean, variance, and standard deviation of data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input array. Converted to 1-D using ravel.\n",
      "            Requires 2 or more data-points.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mdist : \"frozen\" distribution object\n",
      "            Distribution object representing the mean of the data\n",
      "        vdist : \"frozen\" distribution object\n",
      "            Distribution object representing the variance of the data\n",
      "        sdist : \"frozen\" distribution object\n",
      "            Distribution object representing the standard deviation of the data\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        bayes_mvs\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The return values from ``bayes_mvs(data)`` is equivalent to\n",
      "        ``tuple((x.mean(), x.interval(0.90)) for x in mvsdist(data))``.\n",
      "        \n",
      "        In other words, calling ``<dist>.mean()`` and ``<dist>.interval(0.90)``\n",
      "        on the three distribution objects returned from this function will give\n",
      "        the same results that are returned from `bayes_mvs`.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        T.E. Oliphant, \"A Bayesian perspective on estimating mean, variance, and\n",
      "        standard-deviation from data\", http://scholarsarchive.byu.edu/facpub/278,\n",
      "        2006.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> data = [6, 9, 12, 7, 8, 8, 13]\n",
      "        >>> mean, var, std = stats.mvsdist(data)\n",
      "        \n",
      "        We now have frozen distribution objects \"mean\", \"var\" and \"std\" that we can\n",
      "        examine:\n",
      "        \n",
      "        >>> mean.mean()\n",
      "        9.0\n",
      "        >>> mean.interval(0.95)\n",
      "        (6.6120585482655692, 11.387941451734431)\n",
      "        >>> mean.std()\n",
      "        1.1952286093343936\n",
      "    \n",
      "    normaltest(a, axis=0, nan_policy='propagate')\n",
      "        Tests whether a sample differs from a normal distribution.\n",
      "        \n",
      "        This function tests the null hypothesis that a sample comes\n",
      "        from a normal distribution.  It is based on D'Agostino and\n",
      "        Pearson's [1]_, [2]_ test that combines skew and kurtosis to\n",
      "        produce an omnibus test of normality.\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            The array containing the data to be tested.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. Default is 0. If None,\n",
      "            compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            ``s^2 + k^2``, where ``s`` is the z-score returned by `skewtest` and\n",
      "            ``k`` is the z-score returned by `kurtosistest`.\n",
      "        pvalue : float or array\n",
      "           A 2-sided chi squared probability for the hypothesis test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] D'Agostino, R. B. (1971), \"An omnibus test of normality for\n",
      "               moderate and large sample size\", Biometrika, 58, 341-348\n",
      "        \n",
      "        .. [2] D'Agostino, R. and Pearson, E. S. (1973), \"Tests for departure from\n",
      "               normality\", Biometrika, 60, 613-622\n",
      "    \n",
      "    obrientransform(*args)\n",
      "        Computes the O'Brien transform on input data (any number of arrays).\n",
      "        \n",
      "        Used to test for homogeneity of variance prior to running one-way stats.\n",
      "        Each array in ``*args`` is one level of a factor.\n",
      "        If `f_oneway` is run on the transformed data and found significant,\n",
      "        the variances are unequal.  From Maxwell and Delaney [1]_, p.112.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        args : tuple of array_like\n",
      "            Any number of arrays.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        obrientransform : ndarray\n",
      "            Transformed data for use in an ANOVA.  The first dimension\n",
      "            of the result corresponds to the sequence of transformed\n",
      "            arrays.  If the arrays given are all 1-D of the same length,\n",
      "            the return value is a 2-D array; otherwise it is a 1-D array\n",
      "            of type object, with each element being an ndarray.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] S. E. Maxwell and H. D. Delaney, \"Designing Experiments and\n",
      "               Analyzing Data: A Model Comparison Perspective\", Wadsworth, 1990.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We'll test the following data sets for differences in their variance.\n",
      "        \n",
      "        >>> x = [10, 11, 13, 9, 7, 12, 12, 9, 10]\n",
      "        >>> y = [13, 21, 5, 10, 8, 14, 10, 12, 7, 15]\n",
      "        \n",
      "        Apply the O'Brien transform to the data.\n",
      "        \n",
      "        >>> from scipy.stats import obrientransform\n",
      "        >>> tx, ty = obrientransform(x, y)\n",
      "        \n",
      "        Use `scipy.stats.f_oneway` to apply a one-way ANOVA test to the\n",
      "        transformed data.\n",
      "        \n",
      "        >>> from scipy.stats import f_oneway\n",
      "        >>> F, p = f_oneway(tx, ty)\n",
      "        >>> p\n",
      "        0.1314139477040335\n",
      "        \n",
      "        If we require that ``p < 0.05`` for significance, we cannot conclude\n",
      "        that the variances are different.\n",
      "    \n",
      "    pdf_fromgamma(*args, **kwds)\n",
      "        `pdf_fromgamma` is deprecated!\n",
      "        scipy.stats.pdf_fromgamma is deprecated in scipy 0.16.0 in favour of statsmodels.distributions.ExpandedNormal.\n",
      "    \n",
      "    pearsonr(x, y)\n",
      "        Calculates a Pearson correlation coefficient and the p-value for testing\n",
      "        non-correlation.\n",
      "        \n",
      "        The Pearson correlation coefficient measures the linear relationship\n",
      "        between two datasets. Strictly speaking, Pearson's correlation requires\n",
      "        that each dataset be normally distributed, and not necessarily zero-mean.\n",
      "        Like other correlation coefficients, this one varies between -1 and +1\n",
      "        with 0 implying no correlation. Correlations of -1 or +1 imply an exact\n",
      "        linear relationship. Positive correlations imply that as x increases, so\n",
      "        does y. Negative correlations imply that as x increases, y decreases.\n",
      "        \n",
      "        The p-value roughly indicates the probability of an uncorrelated system\n",
      "        producing datasets that have a Pearson correlation at least as extreme\n",
      "        as the one computed from these datasets. The p-values are not entirely\n",
      "        reliable but are probably reasonable for datasets larger than 500 or so.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array_like\n",
      "            Input\n",
      "        y : (N,) array_like\n",
      "            Input\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        r : float\n",
      "            Pearson's correlation coefficient\n",
      "        p-value : float\n",
      "            2-tailed p-value\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        http://www.statsoft.com/textbook/glosp.html#Pearson%20Correlation\n",
      "    \n",
      "    percentileofscore(a, score, kind='rank')\n",
      "        The percentile rank of a score relative to a list of scores.\n",
      "        \n",
      "        A `percentileofscore` of, for example, 80% means that 80% of the\n",
      "        scores in `a` are below the given score. In the case of gaps or\n",
      "        ties, the exact definition depends on the optional keyword, `kind`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of scores to which `score` is compared.\n",
      "        score : int or float\n",
      "            Score that is compared to the elements in `a`.\n",
      "        kind : {'rank', 'weak', 'strict', 'mean'}, optional\n",
      "            This optional parameter specifies the interpretation of the\n",
      "            resulting score:\n",
      "        \n",
      "            - \"rank\": Average percentage ranking of score.  In case of\n",
      "                      multiple matches, average the percentage rankings of\n",
      "                      all matching scores.\n",
      "            - \"weak\": This kind corresponds to the definition of a cumulative\n",
      "                      distribution function.  A percentileofscore of 80%\n",
      "                      means that 80% of values are less than or equal\n",
      "                      to the provided score.\n",
      "            - \"strict\": Similar to \"weak\", except that only values that are\n",
      "                        strictly less than the given score are counted.\n",
      "            - \"mean\": The average of the \"weak\" and \"strict\" scores, often used in\n",
      "                      testing.  See\n",
      "        \n",
      "                      http://en.wikipedia.org/wiki/Percentile_rank\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        pcos : float\n",
      "            Percentile-position of score (0-100) relative to `a`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.percentile\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Three-quarters of the given values lie below a given score:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> stats.percentileofscore([1, 2, 3, 4], 3)\n",
      "        75.0\n",
      "        \n",
      "        With multiple matches, note how the scores of the two matches, 0.6\n",
      "        and 0.8 respectively, are averaged:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3)\n",
      "        70.0\n",
      "        \n",
      "        Only 2/5 values are strictly less than 3:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='strict')\n",
      "        40.0\n",
      "        \n",
      "        But 4/5 values are less than or equal to 3:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='weak')\n",
      "        80.0\n",
      "        \n",
      "        The average between the weak and the strict scores is\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='mean')\n",
      "        60.0\n",
      "    \n",
      "    pointbiserialr(x, y)\n",
      "        Calculates a point biserial correlation coefficient and its p-value.\n",
      "        \n",
      "        The point biserial correlation is used to measure the relationship\n",
      "        between a binary variable, x, and a continuous variable, y. Like other\n",
      "        correlation coefficients, this one varies between -1 and +1 with 0\n",
      "        implying no correlation. Correlations of -1 or +1 imply a determinative\n",
      "        relationship.\n",
      "        \n",
      "        This function uses a shortcut formula but produces the same result as\n",
      "        `pearsonr`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like of bools\n",
      "            Input array.\n",
      "        y : array_like\n",
      "            Input array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float\n",
      "            R value\n",
      "        pvalue : float\n",
      "            2-tailed p-value\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `pointbiserialr` uses a t-test with ``n-1`` degrees of freedom.\n",
      "        It is equivalent to `pearsonr.`\n",
      "        \n",
      "        The value of the point-biserial correlation can be calculated from:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            r_{pb} = \\frac{\\overline{Y_{1}} -\n",
      "                     \\overline{Y_{0}}}{s_{y}}\\sqrt{\\frac{N_{1} N_{2}}{N (N - 1))}}\n",
      "        \n",
      "        Where :math:`Y_{0}` and :math:`Y_{1}` are means of the metric\n",
      "        observations coded 0 and 1 respectively; :math:`N_{0}` and :math:`N_{1}`\n",
      "        are number of observations coded 0 and 1 respectively; :math:`N` is the\n",
      "        total number of observations and :math:`s_{y}` is the standard\n",
      "        deviation of all the metric observations.\n",
      "        \n",
      "        A value of :math:`r_{pb}` that is significantly different from zero is\n",
      "        completely equivalent to a significant difference in means between the two\n",
      "        groups. Thus, an independent groups t Test with :math:`N-2` degrees of\n",
      "        freedom may be used to test whether :math:`r_{pb}` is nonzero. The\n",
      "        relation between the t-statistic for comparing two independent groups and\n",
      "        :math:`r_{pb}` is given by:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            t = \\sqrt{N - 2}\\frac{r_{pb}}{\\sqrt{1 - r^{2}_{pb}}}\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J. Lev, \"The Point Biserial Coefficient of Correlation\", Ann. Math.\n",
      "               Statist., Vol. 20, no.1, pp. 125-126, 1949.\n",
      "        \n",
      "        .. [2] R.F. Tate, \"Correlation Between a Discrete and a Continuous\n",
      "               Variable. Point-Biserial Correlation.\", Ann. Math. Statist., Vol. 25,\n",
      "               np. 3, pp. 603-607, 1954.\n",
      "        \n",
      "        .. [3] http://onlinelibrary.wiley.com/doi/10.1002/9781118445112.stat06227/full\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.array([0, 0, 0, 1, 1, 1, 1])\n",
      "        >>> b = np.arange(7)\n",
      "        >>> stats.pointbiserialr(a, b)\n",
      "        (0.8660254037844386, 0.011724811003954652)\n",
      "        >>> stats.pearsonr(a, b)\n",
      "        (0.86602540378443871, 0.011724811003954626)\n",
      "        >>> np.corrcoef(a, b)\n",
      "        array([[ 1.       ,  0.8660254],\n",
      "               [ 0.8660254,  1.       ]])\n",
      "    \n",
      "    power_divergence(f_obs, f_exp=None, ddof=0, axis=0, lambda_=None)\n",
      "        Cressie-Read power divergence statistic and goodness of fit test.\n",
      "        \n",
      "        This function tests the null hypothesis that the categorical data\n",
      "        has the given frequencies, using the Cressie-Read power divergence\n",
      "        statistic.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f_obs : array_like\n",
      "            Observed frequencies in each category.\n",
      "        f_exp : array_like, optional\n",
      "            Expected frequencies in each category.  By default the categories are\n",
      "            assumed to be equally likely.\n",
      "        ddof : int, optional\n",
      "            \"Delta degrees of freedom\": adjustment to the degrees of freedom\n",
      "            for the p-value.  The p-value is computed using a chi-squared\n",
      "            distribution with ``k - 1 - ddof`` degrees of freedom, where `k`\n",
      "            is the number of observed frequencies.  The default value of `ddof`\n",
      "            is 0.\n",
      "        axis : int or None, optional\n",
      "            The axis of the broadcast result of `f_obs` and `f_exp` along which to\n",
      "            apply the test.  If axis is None, all values in `f_obs` are treated\n",
      "            as a single data set.  Default is 0.\n",
      "        lambda_ : float or str, optional\n",
      "            `lambda_` gives the power in the Cressie-Read power divergence\n",
      "            statistic.  The default is 1.  For convenience, `lambda_` may be\n",
      "            assigned one of the following strings, in which case the\n",
      "            corresponding numerical value is used::\n",
      "        \n",
      "                String              Value   Description\n",
      "                \"pearson\"             1     Pearson's chi-squared statistic.\n",
      "                                            In this case, the function is\n",
      "                                            equivalent to `stats.chisquare`.\n",
      "                \"log-likelihood\"      0     Log-likelihood ratio. Also known as\n",
      "                                            the G-test [3]_.\n",
      "                \"freeman-tukey\"      -1/2   Freeman-Tukey statistic.\n",
      "                \"mod-log-likelihood\" -1     Modified log-likelihood ratio.\n",
      "                \"neyman\"             -2     Neyman's statistic.\n",
      "                \"cressie-read\"        2/3   The power recommended in [5]_.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or ndarray\n",
      "            The Cressie-Read power divergence test statistic.  The value is\n",
      "            a float if `axis` is None or if` `f_obs` and `f_exp` are 1-D.\n",
      "        pvalue : float or ndarray\n",
      "            The p-value of the test.  The value is a float if `ddof` and the\n",
      "            return value `stat` are scalars.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        chisquare\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This test is invalid when the observed or expected frequencies in each\n",
      "        category are too small.  A typical rule is that all of the observed\n",
      "        and expected frequencies should be at least 5.\n",
      "        \n",
      "        When `lambda_` is less than zero, the formula for the statistic involves\n",
      "        dividing by `f_obs`, so a warning or error may be generated if any value\n",
      "        in `f_obs` is 0.\n",
      "        \n",
      "        Similarly, a warning or error may be generated if any value in `f_exp` is\n",
      "        zero when `lambda_` >= 0.\n",
      "        \n",
      "        The default degrees of freedom, k-1, are for the case when no parameters\n",
      "        of the distribution are estimated. If p parameters are estimated by\n",
      "        efficient maximum likelihood then the correct degrees of freedom are\n",
      "        k-1-p. If the parameters are estimated in a different way, then the\n",
      "        dof can be between k-1-p and k-1. However, it is also possible that\n",
      "        the asymptotic distribution is not a chisquare, in which case this\n",
      "        test is not appropriate.\n",
      "        \n",
      "        This function handles masked arrays.  If an element of `f_obs` or `f_exp`\n",
      "        is masked, then data at that position is ignored, and does not count\n",
      "        towards the size of the data set.\n",
      "        \n",
      "        .. versionadded:: 0.13.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
      "               Statistics\". Chapter 8. http://faculty.vassar.edu/lowry/ch8pt1.html\n",
      "        .. [2] \"Chi-squared test\", http://en.wikipedia.org/wiki/Chi-squared_test\n",
      "        .. [3] \"G-test\", http://en.wikipedia.org/wiki/G-test\n",
      "        .. [4] Sokal, R. R. and Rohlf, F. J. \"Biometry: the principles and\n",
      "               practice of statistics in biological research\", New York: Freeman\n",
      "               (1981)\n",
      "        .. [5] Cressie, N. and Read, T. R. C., \"Multinomial Goodness-of-Fit\n",
      "               Tests\", J. Royal Stat. Soc. Series B, Vol. 46, No. 3 (1984),\n",
      "               pp. 440-464.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        (See `chisquare` for more examples.)\n",
      "        \n",
      "        When just `f_obs` is given, it is assumed that the expected frequencies\n",
      "        are uniform and given by the mean of the observed frequencies.  Here we\n",
      "        perform a G-test (i.e. use the log-likelihood ratio statistic):\n",
      "        \n",
      "        >>> from scipy.stats import power_divergence\n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12], lambda_='log-likelihood')\n",
      "        (2.006573162632538, 0.84823476779463769)\n",
      "        \n",
      "        The expected frequencies can be given with the `f_exp` argument:\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12],\n",
      "        ...                  f_exp=[16, 16, 16, 16, 16, 8],\n",
      "        ...                  lambda_='log-likelihood')\n",
      "        (3.3281031458963746, 0.6495419288047497)\n",
      "        \n",
      "        When `f_obs` is 2-D, by default the test is applied to each column.\n",
      "        \n",
      "        >>> obs = np.array([[16, 18, 16, 14, 12, 12], [32, 24, 16, 28, 20, 24]]).T\n",
      "        >>> obs.shape\n",
      "        (6, 2)\n",
      "        >>> power_divergence(obs, lambda_=\"log-likelihood\")\n",
      "        (array([ 2.00657316,  6.77634498]), array([ 0.84823477,  0.23781225]))\n",
      "        \n",
      "        By setting ``axis=None``, the test is applied to all data in the array,\n",
      "        which is equivalent to applying the test to the flattened array.\n",
      "        \n",
      "        >>> power_divergence(obs, axis=None)\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        >>> power_divergence(obs.ravel())\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        \n",
      "        `ddof` is the change to make to the default degrees of freedom.\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12], ddof=1)\n",
      "        (2.0, 0.73575888234288467)\n",
      "        \n",
      "        The calculation of the p-values is done by broadcasting the\n",
      "        test statistic with `ddof`.\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12], ddof=[0,1,2])\n",
      "        (2.0, array([ 0.84914504,  0.73575888,  0.5724067 ]))\n",
      "        \n",
      "        `f_obs` and `f_exp` are also broadcast.  In the following, `f_obs` has\n",
      "        shape (6,) and `f_exp` has shape (2, 6), so the result of broadcasting\n",
      "        `f_obs` and `f_exp` has shape (2, 6).  To compute the desired chi-squared\n",
      "        statistics, we must use ``axis=1``:\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12],\n",
      "        ...                  f_exp=[[16, 16, 16, 16, 16, 8],\n",
      "        ...                         [8, 20, 20, 16, 12, 12]],\n",
      "        ...                  axis=1)\n",
      "        (array([ 3.5 ,  9.25]), array([ 0.62338763,  0.09949846]))\n",
      "    \n",
      "    ppcc_max(x, brack=(0.0, 1.0), dist='tukeylambda')\n",
      "        Calculate the shape parameter that maximizes the PPCC\n",
      "        \n",
      "        The probability plot correlation coefficient (PPCC) plot can be used to\n",
      "        determine the optimal shape parameter for a one-parameter family of\n",
      "        distributions.  ppcc_max returns the shape parameter that would maximize the\n",
      "        probability plot correlation coefficient for the given data to a\n",
      "        one-parameter family of distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        brack : tuple, optional\n",
      "            Triple (a,b,c) where (a<b<c). If bracket consists of two numbers (a, c)\n",
      "            then they are assumed to be a starting interval for a downhill bracket\n",
      "            search (see `scipy.optimize.brent`).\n",
      "        dist : str or stats.distributions instance, optional\n",
      "            Distribution or distribution function name.  Objects that look enough\n",
      "            like a stats.distributions instance (i.e. they have a ``ppf`` method)\n",
      "            are also accepted.  The default is ``'tukeylambda'``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        shape_value : float\n",
      "            The shape parameter at which the probability plot correlation\n",
      "            coefficient reaches its max value.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        ppcc_plot, probplot, boxcox\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The brack keyword serves as a starting point which is useful in corner\n",
      "        cases. One can use a plot to obtain a rough visual estimate of the location\n",
      "        for the maximum to start the search near it.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J.J. Filliben, \"The Probability Plot Correlation Coefficient Test for\n",
      "               Normality\", Technometrics, Vol. 17, pp. 111-117, 1975.\n",
      "        \n",
      "        .. [2] http://www.itl.nist.gov/div898/handbook/eda/section3/ppccplot.htm\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First we generate some random data from a Tukey-Lambda distribution,\n",
      "        with shape parameter -0.7:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> x = stats.tukeylambda.rvs(-0.7, loc=2, scale=0.5, size=10000,\n",
      "        ...                           random_state=1234567) + 1e4\n",
      "        \n",
      "        Now we explore this data with a PPCC plot as well as the related\n",
      "        probability plot and Box-Cox normplot.  A red line is drawn where we\n",
      "        expect the PPCC value to be maximal (at the shape parameter -0.7 used\n",
      "        above):\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> fig = plt.figure(figsize=(8, 6))\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> res = stats.ppcc_plot(x, -5, 5, plot=ax)\n",
      "        \n",
      "        We calculate the value where the shape should reach its maximum and a red\n",
      "        line is drawn there. The line should coincide with the highest point in the\n",
      "        ppcc_plot.\n",
      "        \n",
      "        >>> max = stats.ppcc_max(x)\n",
      "        >>> ax.vlines(max, 0, 1, colors='r', label='Expected shape value')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    ppcc_plot(x, a, b, dist='tukeylambda', plot=None, N=80)\n",
      "        Calculate and optionally plot probability plot correlation coefficient.\n",
      "        \n",
      "        The probability plot correlation coefficient (PPCC) plot can be used to\n",
      "        determine the optimal shape parameter for a one-parameter family of\n",
      "        distributions.  It cannot be used for distributions without shape parameters\n",
      "        (like the normal distribution) or with multiple shape parameters.\n",
      "        \n",
      "        By default a Tukey-Lambda distribution (`stats.tukeylambda`) is used. A\n",
      "        Tukey-Lambda PPCC plot interpolates from long-tailed to short-tailed\n",
      "        distributions via an approximately normal one, and is therefore particularly\n",
      "        useful in practice.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        a, b: scalar\n",
      "            Lower and upper bounds of the shape parameter to use.\n",
      "        dist : str or stats.distributions instance, optional\n",
      "            Distribution or distribution function name.  Objects that look enough\n",
      "            like a stats.distributions instance (i.e. they have a ``ppf`` method)\n",
      "            are also accepted.  The default is ``'tukeylambda'``.\n",
      "        plot : object, optional\n",
      "            If given, plots PPCC against the shape parameter.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        N : int, optional\n",
      "            Number of points on the horizontal axis (equally distributed from\n",
      "            `a` to `b`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        svals : ndarray\n",
      "            The shape values for which `ppcc` was calculated.\n",
      "        ppcc : ndarray\n",
      "            The calculated probability plot correlation coefficient values.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        ppcc_max, probplot, boxcox_normplot, tukeylambda\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        J.J. Filliben, \"The Probability Plot Correlation Coefficient Test for\n",
      "        Normality\", Technometrics, Vol. 17, pp. 111-117, 1975.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First we generate some random data from a Tukey-Lambda distribution,\n",
      "        with shape parameter -0.7:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> np.random.seed(1234567)\n",
      "        >>> x = stats.tukeylambda.rvs(-0.7, loc=2, scale=0.5, size=10000) + 1e4\n",
      "        \n",
      "        Now we explore this data with a PPCC plot as well as the related\n",
      "        probability plot and Box-Cox normplot.  A red line is drawn where we\n",
      "        expect the PPCC value to be maximal (at the shape parameter -0.7 used\n",
      "        above):\n",
      "        \n",
      "        >>> fig = plt.figure(figsize=(12, 4))\n",
      "        >>> ax1 = fig.add_subplot(131)\n",
      "        >>> ax2 = fig.add_subplot(132)\n",
      "        >>> ax3 = fig.add_subplot(133)\n",
      "        >>> res = stats.probplot(x, plot=ax1)\n",
      "        >>> res = stats.boxcox_normplot(x, -5, 5, plot=ax2)\n",
      "        >>> res = stats.ppcc_plot(x, -5, 5, plot=ax3)\n",
      "        >>> ax3.vlines(-0.7, 0, 1, colors='r', label='Expected shape value')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    probplot(x, sparams=(), dist='norm', fit=True, plot=None, rvalue=False)\n",
      "        Calculate quantiles for a probability plot, and optionally show the plot.\n",
      "        \n",
      "        Generates a probability plot of sample data against the quantiles of a\n",
      "        specified theoretical distribution (the normal distribution by default).\n",
      "        `probplot` optionally calculates a best-fit line for the data and plots the\n",
      "        results using Matplotlib or a given plot function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Sample/response data from which `probplot` creates the plot.\n",
      "        sparams : tuple, optional\n",
      "            Distribution-specific shape parameters (shape parameters plus location\n",
      "            and scale).\n",
      "        dist : str or stats.distributions instance, optional\n",
      "            Distribution or distribution function name. The default is 'norm' for a\n",
      "            normal probability plot.  Objects that look enough like a\n",
      "            stats.distributions instance (i.e. they have a ``ppf`` method) are also\n",
      "            accepted.\n",
      "        fit : bool, optional\n",
      "            Fit a least-squares regression (best-fit) line to the sample data if\n",
      "            True (default).\n",
      "        plot : object, optional\n",
      "            If given, plots the quantiles and least squares fit.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        (osm, osr) : tuple of ndarrays\n",
      "            Tuple of theoretical quantiles (osm, or order statistic medians) and\n",
      "            ordered responses (osr).  `osr` is simply sorted input `x`.\n",
      "            For details on how `osm` is calculated see the Notes section.\n",
      "        (slope, intercept, r) : tuple of floats, optional\n",
      "            Tuple  containing the result of the least-squares fit, if that is\n",
      "            performed by `probplot`. `r` is the square root of the coefficient of\n",
      "            determination.  If ``fit=False`` and ``plot=None``, this tuple is not\n",
      "            returned.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Even if `plot` is given, the figure is not shown or saved by `probplot`;\n",
      "        ``plt.show()`` or ``plt.savefig('figname.png')`` should be used after\n",
      "        calling `probplot`.\n",
      "        \n",
      "        `probplot` generates a probability plot, which should not be confused with\n",
      "        a Q-Q or a P-P plot.  Statsmodels has more extensive functionality of this\n",
      "        type, see ``statsmodels.api.ProbPlot``.\n",
      "        \n",
      "        The formula used for the theoretical quantiles (horizontal axis of the\n",
      "        probability plot) is Filliben's estimate::\n",
      "        \n",
      "            quantiles = dist.ppf(val), for\n",
      "        \n",
      "                    0.5**(1/n),                  for i = n\n",
      "              val = (i - 0.3175) / (n + 0.365),  for i = 2, ..., n-1\n",
      "                    1 - 0.5**(1/n),              for i = 1\n",
      "        \n",
      "        where ``i`` indicates the i-th ordered value and ``n`` is the total number\n",
      "        of values.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> nsample = 100\n",
      "        >>> np.random.seed(7654321)\n",
      "        \n",
      "        A t distribution with small degrees of freedom:\n",
      "        \n",
      "        >>> ax1 = plt.subplot(221)\n",
      "        >>> x = stats.t.rvs(3, size=nsample)\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        A t distribution with larger degrees of freedom:\n",
      "        \n",
      "        >>> ax2 = plt.subplot(222)\n",
      "        >>> x = stats.t.rvs(25, size=nsample)\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        A mixture of two normal distributions with broadcasting:\n",
      "        \n",
      "        >>> ax3 = plt.subplot(223)\n",
      "        >>> x = stats.norm.rvs(loc=[0,5], scale=[1,1.5],\n",
      "        ...                    size=(nsample//2,2)).ravel()\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        A standard normal distribution:\n",
      "        \n",
      "        >>> ax4 = plt.subplot(224)\n",
      "        >>> x = stats.norm.rvs(loc=0, scale=1, size=nsample)\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        Produce a new figure with a loggamma distribution, using the ``dist`` and\n",
      "        ``sparams`` keywords:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> x = stats.loggamma.rvs(c=2.5, size=500)\n",
      "        >>> res = stats.probplot(x, dist=stats.loggamma, sparams=(2.5,), plot=ax)\n",
      "        >>> ax.set_title(\"Probplot for loggamma dist with shape parameter 2.5\")\n",
      "        \n",
      "        Show the results with Matplotlib:\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    rankdata(a, method='average')\n",
      "        rankdata(a, method='average')\n",
      "        \n",
      "        Assign ranks to data, dealing with ties appropriately.\n",
      "        \n",
      "        Ranks begin at 1.  The `method` argument controls how ranks are assigned\n",
      "        to equal values.  See [1]_ for further discussion of ranking methods.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            The array of values to be ranked.  The array is first flattened.\n",
      "        method : str, optional\n",
      "            The method used to assign ranks to tied elements.\n",
      "            The options are 'average', 'min', 'max', 'dense' and 'ordinal'.\n",
      "        \n",
      "            'average':\n",
      "                The average of the ranks that would have been assigned to\n",
      "                all the tied values is assigned to each value.\n",
      "            'min':\n",
      "                The minimum of the ranks that would have been assigned to all\n",
      "                the tied values is assigned to each value.  (This is also\n",
      "                referred to as \"competition\" ranking.)\n",
      "            'max':\n",
      "                The maximum of the ranks that would have been assigned to all\n",
      "                the tied values is assigned to each value.\n",
      "            'dense':\n",
      "                Like 'min', but the rank of the next highest element is assigned\n",
      "                the rank immediately after those assigned to the tied elements.\n",
      "            'ordinal':\n",
      "                All values are given a distinct rank, corresponding to the order\n",
      "                that the values occur in `a`.\n",
      "        \n",
      "            The default is 'average'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        ranks : ndarray\n",
      "             An array of length equal to the size of `a`, containing rank\n",
      "             scores.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Ranking\", http://en.wikipedia.org/wiki/Ranking\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import rankdata\n",
      "        >>> rankdata([0, 2, 3, 2])\n",
      "        array([ 1. ,  2.5,  4. ,  2.5])\n",
      "        >>> rankdata([0, 2, 3, 2], method='min')\n",
      "        array([ 1,  2,  4,  2])\n",
      "        >>> rankdata([0, 2, 3, 2], method='max')\n",
      "        array([ 1,  3,  4,  3])\n",
      "        >>> rankdata([0, 2, 3, 2], method='dense')\n",
      "        array([ 1,  2,  3,  2])\n",
      "        >>> rankdata([0, 2, 3, 2], method='ordinal')\n",
      "        array([ 1,  2,  4,  3])\n",
      "    \n",
      "    ranksums(x, y)\n",
      "        Compute the Wilcoxon rank-sum statistic for two samples.\n",
      "        \n",
      "        The Wilcoxon rank-sum test tests the null hypothesis that two sets\n",
      "        of measurements are drawn from the same distribution.  The alternative\n",
      "        hypothesis is that values in one sample are more likely to be\n",
      "        larger than the values in the other sample.\n",
      "        \n",
      "        This test should be used to compare two samples from continuous\n",
      "        distributions.  It does not handle ties between measurements\n",
      "        in x and y.  For tie-handling and an optional continuity correction\n",
      "        see `scipy.stats.mannwhitneyu`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x,y : array_like\n",
      "            The data from the two samples\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic under the large-sample approximation that the\n",
      "            rank sum statistic is normally distributed\n",
      "        pvalue : float\n",
      "            The two-sided p-value of the test\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://en.wikipedia.org/wiki/Wilcoxon_rank-sum_test\n",
      "    \n",
      "    relfreq(a, numbins=10, defaultreallimits=None, weights=None)\n",
      "        Returns a relative frequency histogram, using the histogram function.\n",
      "        \n",
      "        A relative frequency  histogram is a mapping of the number of\n",
      "        observations in each of the bins relative to the total of observations.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        numbins : int, optional\n",
      "            The number of bins to use for the histogram. Default is 10.\n",
      "        defaultreallimits : tuple (lower, upper), optional\n",
      "            The lower and upper values for the range of the histogram.\n",
      "            If no value is given, a range slightly larger than the range of the\n",
      "            values in a is used. Specifically ``(a.min() - s, a.max() + s)``,\n",
      "            where ``s = (1/2)(a.max() - a.min()) / (numbins - 1)``.\n",
      "        weights : array_like, optional\n",
      "            The weights for each value in `a`. Default is None, which gives each\n",
      "            value a weight of 1.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        frequency : ndarray\n",
      "            Binned values of relative frequency.\n",
      "        lowerlimit : float\n",
      "            Lower real limit\n",
      "        binsize : float\n",
      "            Width of each bin.\n",
      "        extrapoints : int\n",
      "            Extra points.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.array([2, 4, 1, 2, 3, 2])\n",
      "        >>> res = stats.relfreq(a, numbins=4)\n",
      "        >>> res.frequency\n",
      "        array([ 0.16666667, 0.5       , 0.16666667,  0.16666667])\n",
      "        >>> np.sum(res.frequency)  # relative frequencies should add up to 1\n",
      "        1.0\n",
      "        \n",
      "        Create a normal distribution with 1000 random values\n",
      "        \n",
      "        >>> rng = np.random.RandomState(seed=12345)\n",
      "        >>> samples = stats.norm.rvs(size=1000, random_state=rng)\n",
      "        \n",
      "        Calculate relative frequencies\n",
      "        \n",
      "        >>> res = stats.relfreq(samples, numbins=25)\n",
      "        \n",
      "        Calculate space of values for x\n",
      "        \n",
      "        >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.frequency.size,\n",
      "        ...                                  res.frequency.size)\n",
      "        \n",
      "        Plot relative frequency histogram\n",
      "        \n",
      "        >>> fig = plt.figure(figsize=(5, 4))\n",
      "        >>> ax = fig.add_subplot(1, 1, 1)\n",
      "        >>> ax.bar(x, res.frequency, width=res.binsize)\n",
      "        >>> ax.set_title('Relative frequency histogram')\n",
      "        >>> ax.set_xlim([x.min(), x.max()])\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    scoreatpercentile(a, per, limit=(), interpolation_method='fraction', axis=None)\n",
      "        Calculate the score at a given percentile of the input sequence.\n",
      "        \n",
      "        For example, the score at `per=50` is the median. If the desired quantile\n",
      "        lies between two data points, we interpolate between them, according to\n",
      "        the value of `interpolation`. If the parameter `limit` is provided, it\n",
      "        should be a tuple (lower, upper) of two values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            A 1-D array of values from which to extract score.\n",
      "        per : array_like\n",
      "            Percentile(s) at which to extract score.  Values should be in range\n",
      "            [0,100].\n",
      "        limit : tuple, optional\n",
      "            Tuple of two scalars, the lower and upper limits within which to\n",
      "            compute the percentile. Values of `a` outside\n",
      "            this (closed) interval will be ignored.\n",
      "        interpolation_method : {'fraction', 'lower', 'higher'}, optional\n",
      "            This optional parameter specifies the interpolation method to use,\n",
      "            when the desired quantile lies between two data points `i` and `j`\n",
      "        \n",
      "              - fraction: ``i + (j - i) * fraction`` where ``fraction`` is the\n",
      "                fractional part of the index surrounded by ``i`` and ``j``.\n",
      "              - lower: ``i``.\n",
      "              - higher: ``j``.\n",
      "        \n",
      "        axis : int, optional\n",
      "            Axis along which the percentiles are computed. Default is None. If\n",
      "            None, compute over the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        score : float or ndarray\n",
      "            Score at percentile(s).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        percentileofscore, numpy.percentile\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function will become obsolete in the future.\n",
      "        For Numpy 1.9 and higher, `numpy.percentile` provides all the functionality\n",
      "        that `scoreatpercentile` provides.  And it's significantly faster.\n",
      "        Therefore it's recommended to use `numpy.percentile` for users that have\n",
      "        numpy >= 1.9.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(100)\n",
      "        >>> stats.scoreatpercentile(a, 50)\n",
      "        49.5\n",
      "    \n",
      "    sem(a, axis=0, ddof=1, nan_policy='propagate')\n",
      "        Calculates the standard error of the mean (or standard error of\n",
      "        measurement) of the values in the input array.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            An array containing the values for which the standard error is\n",
      "            returned.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees-of-freedom. How many degrees of freedom to adjust\n",
      "            for bias in limited samples relative to the population estimate\n",
      "            of variance. Defaults to 1.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        s : ndarray or float\n",
      "            The standard error of the mean in the sample(s), along the input axis.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The default value for `ddof` is different to the default (0) used by other\n",
      "        ddof containing routines, such as np.std and np.nanstd.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Find standard error along the first axis:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(20).reshape(5,4)\n",
      "        >>> stats.sem(a)\n",
      "        array([ 2.8284,  2.8284,  2.8284,  2.8284])\n",
      "        \n",
      "        Find standard error across the whole array, using n degrees of freedom:\n",
      "        \n",
      "        >>> stats.sem(a, axis=None, ddof=0)\n",
      "        1.2893796958227628\n",
      "    \n",
      "    shapiro(x, a=None, reta=False)\n",
      "        Perform the Shapiro-Wilk test for normality.\n",
      "        \n",
      "        The Shapiro-Wilk test tests the null hypothesis that the\n",
      "        data was drawn from a normal distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Array of sample data.\n",
      "        a : array_like, optional\n",
      "            Array of internal parameters used in the calculation.  If these\n",
      "            are not given, they will be computed internally.  If x has length\n",
      "            n, then a must have length n/2.\n",
      "        reta : bool, optional\n",
      "            Whether or not to return the internally computed a values.  The\n",
      "            default is False.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        W : float\n",
      "            The test statistic.\n",
      "        p-value : float\n",
      "            The p-value for the hypothesis test.\n",
      "        a : array_like, optional\n",
      "            If `reta` is True, then these are the internally computed \"a\"\n",
      "            values that may be passed into this function on future calls.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        anderson : The Anderson-Darling test for normality\n",
      "        kstest : The Kolmogorov-Smirnov test for goodness of fit.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The algorithm used is described in [4]_ but censoring parameters as\n",
      "        described are not implemented. For N > 5000 the W test statistic is accurate\n",
      "        but the p-value may not be.\n",
      "        \n",
      "        The chance of rejecting the null hypothesis when it is true is close to 5%\n",
      "        regardless of sample size.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://www.itl.nist.gov/div898/handbook/prc/section2/prc213.htm\n",
      "        .. [2] Shapiro, S. S. & Wilk, M.B (1965). An analysis of variance test for\n",
      "               normality (complete samples), Biometrika, Vol. 52, pp. 591-611.\n",
      "        .. [3] Razali, N. M. & Wah, Y. B. (2011) Power comparisons of Shapiro-Wilk,\n",
      "               Kolmogorov-Smirnov, Lilliefors and Anderson-Darling tests, Journal of\n",
      "               Statistical Modeling and Analytics, Vol. 2, pp. 21-33.\n",
      "        .. [4] ALGORITHM AS R94 APPL. STATIST. (1995) VOL. 44, NO. 4.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(12345678)\n",
      "        >>> x = stats.norm.rvs(loc=5, scale=3, size=100)\n",
      "        >>> stats.shapiro(x)\n",
      "        (0.9772805571556091, 0.08144091814756393)\n",
      "    \n",
      "    sigmaclip(a, low=4.0, high=4.0)\n",
      "        Iterative sigma-clipping of array elements.\n",
      "        \n",
      "        The output array contains only those elements of the input array `c`\n",
      "        that satisfy the conditions ::\n",
      "        \n",
      "            mean(c) - std(c)*low < c < mean(c) + std(c)*high\n",
      "        \n",
      "        Starting from the full sample, all elements outside the critical range are\n",
      "        removed. The iteration continues with a new critical range until no\n",
      "        elements are outside the range.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Data array, will be raveled if not 1-D.\n",
      "        low : float, optional\n",
      "            Lower bound factor of sigma clipping. Default is 4.\n",
      "        high : float, optional\n",
      "            Upper bound factor of sigma clipping. Default is 4.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        clipped : ndarray\n",
      "            Input array with clipped elements removed.\n",
      "        lower : float\n",
      "            Lower threshold value use for clipping.\n",
      "        upper : float\n",
      "            Upper threshold value use for clipping.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import sigmaclip\n",
      "        >>> a = np.concatenate((np.linspace(9.5, 10.5, 31),\n",
      "        ...                     np.linspace(0, 20, 5)))\n",
      "        >>> fact = 1.5\n",
      "        >>> c, low, upp = sigmaclip(a, fact, fact)\n",
      "        >>> c\n",
      "        array([  9.96666667,  10.        ,  10.03333333,  10.        ])\n",
      "        >>> c.var(), c.std()\n",
      "        (0.00055555555555555165, 0.023570226039551501)\n",
      "        >>> low, c.mean() - fact*c.std(), c.min()\n",
      "        (9.9646446609406727, 9.9646446609406727, 9.9666666666666668)\n",
      "        >>> upp, c.mean() + fact*c.std(), c.max()\n",
      "        (10.035355339059327, 10.035355339059327, 10.033333333333333)\n",
      "        \n",
      "        >>> a = np.concatenate((np.linspace(9.5, 10.5, 11),\n",
      "        ...                     np.linspace(-100, -50, 3)))\n",
      "        >>> c, low, upp = sigmaclip(a, 1.8, 1.8)\n",
      "        >>> (c == np.linspace(9.5, 10.5, 11)).all()\n",
      "        True\n",
      "    \n",
      "    signaltonoise(*args, **kwds)\n",
      "        `signaltonoise` is deprecated!\n",
      "        scipy.stats.signaltonoise is deprecated in scipy 0.16.0\n",
      "        \n",
      "        \n",
      "            The signal-to-noise ratio of the input data.\n",
      "        \n",
      "            Returns the signal-to-noise ratio of `a`, here defined as the mean\n",
      "            divided by the standard deviation.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            a : array_like\n",
      "                An array_like object containing the sample data.\n",
      "            axis : int or None, optional\n",
      "                Axis along which to operate. Default is 0. If None, compute over\n",
      "                the whole array `a`.\n",
      "            ddof : int, optional\n",
      "                Degrees of freedom correction for standard deviation. Default is 0.\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            s2n : ndarray\n",
      "                The mean to standard deviation ratio(s) along `axis`, or 0 where the\n",
      "                standard deviation is 0.\n",
      "    \n",
      "    skew(a, axis=0, bias=True, nan_policy='propagate')\n",
      "        Computes the skewness of a data set.\n",
      "        \n",
      "        For normally distributed data, the skewness should be about 0. A skewness\n",
      "        value > 0 means that there is more weight in the left tail of the\n",
      "        distribution. The function `skewtest` can be used to determine if the\n",
      "        skewness value is close enough to 0, statistically speaking.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : ndarray\n",
      "            data\n",
      "        axis : int or None, optional\n",
      "            Axis along which skewness is calculated. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        bias : bool, optional\n",
      "            If False, then the calculations are corrected for statistical bias.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        skewness : ndarray\n",
      "            The skewness of values along an axis, returning 0 where all values are\n",
      "            equal.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        \n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "           Section 2.2.24.1\n",
      "    \n",
      "    skewtest(a, axis=0, nan_policy='propagate')\n",
      "        Tests whether the skew is different from the normal distribution.\n",
      "        \n",
      "        This function tests the null hypothesis that the skewness of\n",
      "        the population that the sample was drawn from is the same\n",
      "        as that of a corresponding normal distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array\n",
      "            The data to be tested\n",
      "        axis : int or None, optional\n",
      "           Axis along which statistics are calculated. Default is 0.\n",
      "           If None, compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed z-score for this test.\n",
      "        pvalue : float\n",
      "            a 2-sided p-value for the hypothesis test\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The sample size must be at least 8.\n",
      "    \n",
      "    spearmanr(a, b=None, axis=0, nan_policy='propagate')\n",
      "        Calculates a Spearman rank-order correlation coefficient and the p-value\n",
      "        to test for non-correlation.\n",
      "        \n",
      "        The Spearman correlation is a nonparametric measure of the monotonicity\n",
      "        of the relationship between two datasets. Unlike the Pearson correlation,\n",
      "        the Spearman correlation does not assume that both datasets are normally\n",
      "        distributed. Like other correlation coefficients, this one varies\n",
      "        between -1 and +1 with 0 implying no correlation. Correlations of -1 or\n",
      "        +1 imply an exact monotonic relationship. Positive correlations imply that\n",
      "        as x increases, so does y. Negative correlations imply that as x\n",
      "        increases, y decreases.\n",
      "        \n",
      "        The p-value roughly indicates the probability of an uncorrelated system\n",
      "        producing datasets that have a Spearman correlation at least as extreme\n",
      "        as the one computed from these datasets. The p-values are not entirely\n",
      "        reliable but are probably reasonable for datasets larger than 500 or so.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : 1D or 2D array_like, b is optional\n",
      "            One or two 1-D or 2-D arrays containing multiple variables and\n",
      "            observations. When these are 1-D, each represents a vector of\n",
      "            observations of a single variable. For the behavior in the 2-D case,\n",
      "            see under ``axis``, below.\n",
      "            Both arrays need to have the same length in the ``axis`` dimension.\n",
      "        axis : int or None, optional\n",
      "            If axis=0 (default), then each column represents a variable, with\n",
      "            observations in the rows. If axis=1, the relationship is transposed:\n",
      "            each row represents a variable, while the columns contain observations.\n",
      "            If axis=None, then both arrays will be raveled.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float or ndarray (2-D square)\n",
      "            Spearman correlation matrix or correlation coefficient (if only 2\n",
      "            variables are given as parameters. Correlation matrix is square with\n",
      "            length equal to total number of variables (columns or rows) in a and b\n",
      "            combined.\n",
      "        pvalue : float\n",
      "            The two-sided p-value for a hypothesis test whose null hypothesis is\n",
      "            that two sets of data are uncorrelated, has same dimension as rho.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Changes in scipy 0.8.0: rewrite to add tie-handling, and axis.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        \n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "           Section  14.7\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.spearmanr([1,2,3,4,5], [5,6,7,8,7])\n",
      "        (0.82078268166812329, 0.088587005313543798)\n",
      "        >>> np.random.seed(1234321)\n",
      "        >>> x2n = np.random.randn(100, 2)\n",
      "        >>> y2n = np.random.randn(100, 2)\n",
      "        >>> stats.spearmanr(x2n)\n",
      "        (0.059969996999699973, 0.55338590803773591)\n",
      "        >>> stats.spearmanr(x2n[:,0], x2n[:,1])\n",
      "        (0.059969996999699973, 0.55338590803773591)\n",
      "        >>> rho, pval = stats.spearmanr(x2n, y2n)\n",
      "        >>> rho\n",
      "        array([[ 1.        ,  0.05997   ,  0.18569457,  0.06258626],\n",
      "               [ 0.05997   ,  1.        ,  0.110003  ,  0.02534653],\n",
      "               [ 0.18569457,  0.110003  ,  1.        ,  0.03488749],\n",
      "               [ 0.06258626,  0.02534653,  0.03488749,  1.        ]])\n",
      "        >>> pval\n",
      "        array([[ 0.        ,  0.55338591,  0.06435364,  0.53617935],\n",
      "               [ 0.55338591,  0.        ,  0.27592895,  0.80234077],\n",
      "               [ 0.06435364,  0.27592895,  0.        ,  0.73039992],\n",
      "               [ 0.53617935,  0.80234077,  0.73039992,  0.        ]])\n",
      "        >>> rho, pval = stats.spearmanr(x2n.T, y2n.T, axis=1)\n",
      "        >>> rho\n",
      "        array([[ 1.        ,  0.05997   ,  0.18569457,  0.06258626],\n",
      "               [ 0.05997   ,  1.        ,  0.110003  ,  0.02534653],\n",
      "               [ 0.18569457,  0.110003  ,  1.        ,  0.03488749],\n",
      "               [ 0.06258626,  0.02534653,  0.03488749,  1.        ]])\n",
      "        >>> stats.spearmanr(x2n, y2n, axis=None)\n",
      "        (0.10816770419260482, 0.1273562188027364)\n",
      "        >>> stats.spearmanr(x2n.ravel(), y2n.ravel())\n",
      "        (0.10816770419260482, 0.1273562188027364)\n",
      "        \n",
      "        >>> xint = np.random.randint(10, size=(100, 2))\n",
      "        >>> stats.spearmanr(xint)\n",
      "        (0.052760927029710199, 0.60213045837062351)\n",
      "    \n",
      "    square_of_sums(*args, **kwds)\n",
      "        `square_of_sums` is deprecated!\n",
      "        scipy.stats.square_of_sums is deprecated in scipy 0.17.0\n",
      "    \n",
      "    ss(*args, **kwds)\n",
      "        `ss` is deprecated!\n",
      "        scipy.stats.ss is deprecated in scipy 0.17.0\n",
      "    \n",
      "    theilslopes(y, x=None, alpha=0.95)\n",
      "        Computes the Theil-Sen estimator for a set of points (x, y).\n",
      "        \n",
      "        `theilslopes` implements a method for robust linear regression.  It\n",
      "        computes the slope as the median of all slopes between paired values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : array_like\n",
      "            Dependent variable.\n",
      "        x : array_like or None, optional\n",
      "            Independent variable. If None, use ``arange(len(y))`` instead.\n",
      "        alpha : float, optional\n",
      "            Confidence degree between 0 and 1. Default is 95% confidence.\n",
      "            Note that `alpha` is symmetric around 0.5, i.e. both 0.1 and 0.9 are\n",
      "            interpreted as \"find the 90% confidence interval\".\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        medslope : float\n",
      "            Theil slope.\n",
      "        medintercept : float\n",
      "            Intercept of the Theil line, as ``median(y) - medslope*median(x)``.\n",
      "        lo_slope : float\n",
      "            Lower bound of the confidence interval on `medslope`.\n",
      "        up_slope : float\n",
      "            Upper bound of the confidence interval on `medslope`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The implementation of `theilslopes` follows [1]_. The intercept is\n",
      "        not defined in [1]_, and here it is defined as ``median(y) -\n",
      "        medslope*median(x)``, which is given in [3]_. Other definitions of\n",
      "        the intercept exist in the literature. A confidence interval for\n",
      "        the intercept is not given as this question is not addressed in\n",
      "        [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] P.K. Sen, \"Estimates of the regression coefficient based on Kendall's tau\",\n",
      "               J. Am. Stat. Assoc., Vol. 63, pp. 1379-1389, 1968.\n",
      "        .. [2] H. Theil, \"A rank-invariant method of linear and polynomial\n",
      "               regression analysis I, II and III\",  Nederl. Akad. Wetensch., Proc.\n",
      "               53:, pp. 386-392, pp. 521-525, pp. 1397-1412, 1950.\n",
      "        .. [3] W.L. Conover, \"Practical nonparametric statistics\", 2nd ed.,\n",
      "               John Wiley and Sons, New York, pp. 493.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        >>> x = np.linspace(-5, 5, num=150)\n",
      "        >>> y = x + np.random.normal(size=x.size)\n",
      "        >>> y[11:15] += 10  # add outliers\n",
      "        >>> y[-5:] -= 7\n",
      "        \n",
      "        Compute the slope, intercept and 90% confidence interval.  For comparison,\n",
      "        also compute the least-squares fit with `linregress`:\n",
      "        \n",
      "        >>> res = stats.theilslopes(y, x, 0.90)\n",
      "        >>> lsq_res = stats.linregress(x, y)\n",
      "        \n",
      "        Plot the results. The Theil-Sen regression line is shown in red, with the\n",
      "        dashed red lines illustrating the confidence interval of the slope (note\n",
      "        that the dashed red lines are not the confidence interval of the regression\n",
      "        as the confidence interval of the intercept is not included). The green\n",
      "        line shows the least-squares fit for comparison.\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(x, y, 'b.')\n",
      "        >>> ax.plot(x, res[1] + res[0] * x, 'r-')\n",
      "        >>> ax.plot(x, res[1] + res[2] * x, 'r--')\n",
      "        >>> ax.plot(x, res[1] + res[3] * x, 'r--')\n",
      "        >>> ax.plot(x, lsq_res[1] + lsq_res[0] * x, 'g-')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    threshold(*args, **kwds)\n",
      "        `threshold` is deprecated!\n",
      "        stats.threshold is deprecated in scipy 0.17.0\n",
      "        \n",
      "        \n",
      "            Clip array to a given value.\n",
      "        \n",
      "            Similar to numpy.clip(), except that values less than `threshmin` or\n",
      "            greater than `threshmax` are replaced by `newval`, instead of by\n",
      "            `threshmin` and `threshmax` respectively.\n",
      "        \n",
      "            Parameters\n",
      "            ----------\n",
      "            a : array_like\n",
      "                Data to threshold.\n",
      "            threshmin : float, int or None, optional\n",
      "                Minimum threshold, defaults to None.\n",
      "            threshmax : float, int or None, optional\n",
      "                Maximum threshold, defaults to None.\n",
      "            newval : float or int, optional\n",
      "                Value to put in place of values in `a` outside of bounds.\n",
      "                Defaults to 0.\n",
      "        \n",
      "            Returns\n",
      "            -------\n",
      "            out : ndarray\n",
      "                The clipped input array, with values less than `threshmin` or\n",
      "                greater than `threshmax` replaced with `newval`.\n",
      "        \n",
      "            Examples\n",
      "            --------\n",
      "            >>> a = np.array([9, 9, 6, 3, 1, 6, 1, 0, 0, 8])\n",
      "            >>> from scipy import stats\n",
      "            >>> stats.threshold(a, threshmin=2, threshmax=8, newval=-1)\n",
      "            array([-1, -1,  6,  3, -1,  6, -1, -1, -1,  8])\n",
      "    \n",
      "    tiecorrect(rankvals)\n",
      "        Tie correction factor for ties in the Mann-Whitney U and\n",
      "        Kruskal-Wallis H tests.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rankvals : array_like\n",
      "            A 1-D sequence of ranks.  Typically this will be the array\n",
      "            returned by `stats.rankdata`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        factor : float\n",
      "            Correction factor for U or H.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rankdata : Assign ranks to the data\n",
      "        mannwhitneyu : Mann-Whitney rank test\n",
      "        kruskal : Kruskal-Wallis H test\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Siegel, S. (1956) Nonparametric Statistics for the Behavioral\n",
      "               Sciences.  New York: McGraw-Hill.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import tiecorrect, rankdata\n",
      "        >>> tiecorrect([1, 2.5, 2.5, 4])\n",
      "        0.9\n",
      "        >>> ranks = rankdata([1, 3, 2, 4, 5, 7, 2, 8, 4])\n",
      "        >>> ranks\n",
      "        array([ 1. ,  4. ,  2.5,  5.5,  7. ,  8. ,  2.5,  9. ,  5.5])\n",
      "        >>> tiecorrect(ranks)\n",
      "        0.9833333333333333\n",
      "    \n",
      "    tmax(a, upperlimit=None, axis=0, inclusive=True, nan_policy='propagate')\n",
      "        Compute the trimmed maximum\n",
      "        \n",
      "        This function computes the maximum value of an array along a given axis,\n",
      "        while ignoring values larger than a specified upper limit.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            array of values\n",
      "        upperlimit : None or float, optional\n",
      "            Values in the input array greater than the given limit will be ignored.\n",
      "            When upperlimit is None, then all values are used. The default value\n",
      "            is None.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        inclusive : {True, False}, optional\n",
      "            This flag determines whether values exactly equal to the upper limit\n",
      "            are included.  The default value is True.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tmax : float, int or ndarray\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tmax(x)\n",
      "        19\n",
      "        \n",
      "        >>> stats.tmax(x, 13)\n",
      "        13\n",
      "        \n",
      "        >>> stats.tmax(x, 13, inclusive=False)\n",
      "        12\n",
      "    \n",
      "    tmean(a, limits=None, inclusive=(True, True), axis=None)\n",
      "        Compute the trimmed mean.\n",
      "        \n",
      "        This function finds the arithmetic mean of given values, ignoring values\n",
      "        outside the given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored.  When limits is None (default), then all\n",
      "            values are used.  Either of the limit values in the tuple can also be\n",
      "            None representing a half-open interval.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. Default is None.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tmean : float\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        trim_mean : returns mean after trimming a proportion from both tails.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tmean(x)\n",
      "        9.5\n",
      "        >>> stats.tmean(x, (3,17))\n",
      "        10.0\n",
      "    \n",
      "    tmin(a, lowerlimit=None, axis=0, inclusive=True, nan_policy='propagate')\n",
      "        Compute the trimmed minimum\n",
      "        \n",
      "        This function finds the miminum value of an array `a` along the\n",
      "        specified axis, but only considering values greater than a specified\n",
      "        lower limit.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            array of values\n",
      "        lowerlimit : None or float, optional\n",
      "            Values in the input array less than the given limit will be ignored.\n",
      "            When lowerlimit is None, then all values are used. The default value\n",
      "            is None.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        inclusive : {True, False}, optional\n",
      "            This flag determines whether values exactly equal to the lower limit\n",
      "            are included.  The default value is True.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tmin : float, int or ndarray\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tmin(x)\n",
      "        0\n",
      "        \n",
      "        >>> stats.tmin(x, 13)\n",
      "        13\n",
      "        \n",
      "        >>> stats.tmin(x, 13, inclusive=False)\n",
      "        14\n",
      "    \n",
      "    trim1(a, proportiontocut, tail='right', axis=0)\n",
      "        Slices off a proportion from ONE end of the passed array distribution.\n",
      "        \n",
      "        If `proportiontocut` = 0.1, slices off 'leftmost' or 'rightmost'\n",
      "        10% of scores. The lowest or highest values are trimmed (depending on\n",
      "        the tail).\n",
      "        Slices off less if proportion results in a non-integer slice index\n",
      "        (i.e., conservatively slices off `proportiontocut` ).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array\n",
      "        proportiontocut : float\n",
      "            Fraction to cut off of 'left' or 'right' of distribution\n",
      "        tail : {'left', 'right'}, optional\n",
      "            Defaults to 'right'.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to trim data. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        trim1 : ndarray\n",
      "            Trimmed version of array `a`. The order of the trimmed content is\n",
      "            undefined.\n",
      "    \n",
      "    trim_mean(a, proportiontocut, axis=0)\n",
      "        Return mean of array after trimming distribution from both tails.\n",
      "        \n",
      "        If `proportiontocut` = 0.1, slices off 'leftmost' and 'rightmost' 10% of\n",
      "        scores. The input is sorted before slicing. Slices off less if proportion\n",
      "        results in a non-integer slice index (i.e., conservatively slices off\n",
      "        `proportiontocut` ).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array\n",
      "        proportiontocut : float\n",
      "            Fraction to cut off of both tails of the distribution\n",
      "        axis : int or None, optional\n",
      "            Axis along which the trimmed means are computed. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        trim_mean : ndarray\n",
      "            Mean of trimmed array.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        trimboth\n",
      "        tmean : compute the trimmed mean ignoring values outside given `limits`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.trim_mean(x, 0.1)\n",
      "        9.5\n",
      "        >>> x2 = x.reshape(5, 4)\n",
      "        >>> x2\n",
      "        array([[ 0,  1,  2,  3],\n",
      "               [ 4,  5,  6,  7],\n",
      "               [ 8,  9, 10, 11],\n",
      "               [12, 13, 14, 15],\n",
      "               [16, 17, 18, 19]])\n",
      "        >>> stats.trim_mean(x2, 0.25)\n",
      "        array([  8.,   9.,  10.,  11.])\n",
      "        >>> stats.trim_mean(x2, 0.25, axis=1)\n",
      "        array([  1.5,   5.5,   9.5,  13.5,  17.5])\n",
      "    \n",
      "    trimboth(a, proportiontocut, axis=0)\n",
      "        Slices off a proportion of items from both ends of an array.\n",
      "        \n",
      "        Slices off the passed proportion of items from both ends of the passed\n",
      "        array (i.e., with `proportiontocut` = 0.1, slices leftmost 10% **and**\n",
      "        rightmost 10% of scores). The trimmed values are the lowest and\n",
      "        highest ones.\n",
      "        Slices off less if proportion results in a non-integer slice index (i.e.,\n",
      "        conservatively slices off`proportiontocut`).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Data to trim.\n",
      "        proportiontocut : float\n",
      "            Proportion (in range 0-1) of total data set to trim of each end.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to trim data. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        out : ndarray\n",
      "            Trimmed version of array `a`. The order of the trimmed content\n",
      "            is undefined.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        trim_mean\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(20)\n",
      "        >>> b = stats.trimboth(a, 0.1)\n",
      "        >>> b.shape\n",
      "        (16,)\n",
      "    \n",
      "    tsem(a, limits=None, inclusive=(True, True), axis=0, ddof=1)\n",
      "        Compute the trimmed standard error of the mean.\n",
      "        \n",
      "        This function finds the standard error of the mean for given\n",
      "        values, ignoring values outside the given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            array of values\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored. When limits is None, then all values are\n",
      "            used. Either of the limit values in the tuple can also be None\n",
      "            representing a half-open interval.  The default value is None.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tsem : float\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `tsem` uses unbiased sample standard deviation, i.e. it uses a\n",
      "        correction factor ``n / (n - 1)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tsem(x)\n",
      "        1.3228756555322954\n",
      "        >>> stats.tsem(x, (3,17))\n",
      "        1.1547005383792515\n",
      "    \n",
      "    tstd(a, limits=None, inclusive=(True, True), axis=0, ddof=1)\n",
      "        Compute the trimmed sample standard deviation\n",
      "        \n",
      "        This function finds the sample standard deviation of given values,\n",
      "        ignoring values outside the given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            array of values\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored. When limits is None, then all values are\n",
      "            used. Either of the limit values in the tuple can also be None\n",
      "            representing a half-open interval.  The default value is None.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tstd : float\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `tstd` computes the unbiased sample standard deviation, i.e. it uses a\n",
      "        correction factor ``n / (n - 1)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tstd(x)\n",
      "        5.9160797830996161\n",
      "        >>> stats.tstd(x, (3,17))\n",
      "        4.4721359549995796\n",
      "    \n",
      "    ttest_1samp(a, popmean, axis=0, nan_policy='propagate')\n",
      "        Calculates the T-test for the mean of ONE group of scores.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that the expected value\n",
      "        (mean) of a sample of independent observations `a` is equal to the given\n",
      "        population mean, `popmean`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            sample observation\n",
      "        popmean : float or array_like\n",
      "            expected value in null hypothesis, if array_like than it must have the\n",
      "            same shape as `a` excluding the axis dimension\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. If None, compute over the whole\n",
      "            array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            t-statistic\n",
      "        pvalue : float or array\n",
      "            two-tailed p-value\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        \n",
      "        >>> np.random.seed(7654567)  # fix seed to get the same result\n",
      "        >>> rvs = stats.norm.rvs(loc=5, scale=10, size=(50,2))\n",
      "        \n",
      "        Test if mean of random sample is equal to true mean, and different mean.\n",
      "        We reject the null hypothesis in the second case and don't reject it in\n",
      "        the first case.\n",
      "        \n",
      "        >>> stats.ttest_1samp(rvs,5.0)\n",
      "        (array([-0.68014479, -0.04323899]), array([ 0.49961383,  0.96568674]))\n",
      "        >>> stats.ttest_1samp(rvs,0.0)\n",
      "        (array([ 2.77025808,  4.11038784]), array([ 0.00789095,  0.00014999]))\n",
      "        \n",
      "        Examples using axis and non-scalar dimension for population mean.\n",
      "        \n",
      "        >>> stats.ttest_1samp(rvs,[5.0,0.0])\n",
      "        (array([-0.68014479,  4.11038784]), array([  4.99613833e-01,   1.49986458e-04]))\n",
      "        >>> stats.ttest_1samp(rvs.T,[5.0,0.0],axis=1)\n",
      "        (array([-0.68014479,  4.11038784]), array([  4.99613833e-01,   1.49986458e-04]))\n",
      "        >>> stats.ttest_1samp(rvs,[[5.0],[0.0]])\n",
      "        (array([[-0.68014479, -0.04323899],\n",
      "               [ 2.77025808,  4.11038784]]), array([[  4.99613833e-01,   9.65686743e-01],\n",
      "               [  7.89094663e-03,   1.49986458e-04]]))\n",
      "    \n",
      "    ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate')\n",
      "        Calculates the T-test for the means of *two independent* samples of scores.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that 2 independent samples\n",
      "        have identical average (expected) values. This test assumes that the\n",
      "        populations have identical variances by default.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : array_like\n",
      "            The arrays must have the same shape, except in the dimension\n",
      "            corresponding to `axis` (the first, by default).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. If None, compute over the whole\n",
      "            arrays, `a`, and `b`.\n",
      "        equal_var : bool, optional\n",
      "            If True (default), perform a standard independent 2 sample test\n",
      "            that assumes equal population variances [1]_.\n",
      "            If False, perform Welch's t-test, which does not assume equal\n",
      "            population variance [2]_.\n",
      "        \n",
      "            .. versionadded:: 0.11.0\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            The calculated t-statistic.\n",
      "        pvalue : float or array\n",
      "            The two-tailed p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        We can use this test, if we observe two independent samples from\n",
      "        the same or different population, e.g. exam scores of boys and\n",
      "        girls or of two ethnic groups. The test measures whether the\n",
      "        average (expected) value differs significantly across samples. If\n",
      "        we observe a large p-value, for example larger than 0.05 or 0.1,\n",
      "        then we cannot reject the null hypothesis of identical average scores.\n",
      "        If the p-value is smaller than the threshold, e.g. 1%, 5% or 10%,\n",
      "        then we reject the null hypothesis of equal averages.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://en.wikipedia.org/wiki/T-test#Independent_two-sample_t-test\n",
      "        \n",
      "        .. [2] http://en.wikipedia.org/wiki/Welch%27s_t_test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(12345678)\n",
      "        \n",
      "        Test with sample with identical means:\n",
      "        \n",
      "        >>> rvs1 = stats.norm.rvs(loc=5,scale=10,size=500)\n",
      "        >>> rvs2 = stats.norm.rvs(loc=5,scale=10,size=500)\n",
      "        >>> stats.ttest_ind(rvs1,rvs2)\n",
      "        (0.26833823296239279, 0.78849443369564776)\n",
      "        >>> stats.ttest_ind(rvs1,rvs2, equal_var = False)\n",
      "        (0.26833823296239279, 0.78849452749500748)\n",
      "        \n",
      "        `ttest_ind` underestimates p for unequal variances:\n",
      "        \n",
      "        >>> rvs3 = stats.norm.rvs(loc=5, scale=20, size=500)\n",
      "        >>> stats.ttest_ind(rvs1, rvs3)\n",
      "        (-0.46580283298287162, 0.64145827413436174)\n",
      "        >>> stats.ttest_ind(rvs1, rvs3, equal_var = False)\n",
      "        (-0.46580283298287162, 0.64149646246569292)\n",
      "        \n",
      "        When n1 != n2, the equal variance t-statistic is no longer equal to the\n",
      "        unequal variance t-statistic:\n",
      "        \n",
      "        >>> rvs4 = stats.norm.rvs(loc=5, scale=20, size=100)\n",
      "        >>> stats.ttest_ind(rvs1, rvs4)\n",
      "        (-0.99882539442782481, 0.3182832709103896)\n",
      "        >>> stats.ttest_ind(rvs1, rvs4, equal_var = False)\n",
      "        (-0.69712570584654099, 0.48716927725402048)\n",
      "        \n",
      "        T-test with different means, variance, and n:\n",
      "        \n",
      "        >>> rvs5 = stats.norm.rvs(loc=8, scale=20, size=100)\n",
      "        >>> stats.ttest_ind(rvs1, rvs5)\n",
      "        (-1.4679669854490653, 0.14263895620529152)\n",
      "        >>> stats.ttest_ind(rvs1, rvs5, equal_var = False)\n",
      "        (-0.94365973617132992, 0.34744170334794122)\n",
      "    \n",
      "    ttest_ind_from_stats(mean1, std1, nobs1, mean2, std2, nobs2, equal_var=True)\n",
      "        T-test for means of two independent samples from descriptive statistics.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that 2 independent samples\n",
      "        have identical average (expected) values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        mean1 : array_like\n",
      "            The mean(s) of sample 1.\n",
      "        std1 : array_like\n",
      "            The standard deviation(s) of sample 1.\n",
      "        nobs1 : array_like\n",
      "            The number(s) of observations of sample 1.\n",
      "        mean2 : array_like\n",
      "            The mean(s) of sample 2\n",
      "        std2 : array_like\n",
      "            The standard deviations(s) of sample 2.\n",
      "        nobs2 : array_like\n",
      "            The number(s) of observations of sample 2.\n",
      "        equal_var : bool, optional\n",
      "            If True (default), perform a standard independent 2 sample test\n",
      "            that assumes equal population variances [1]_.\n",
      "            If False, perform Welch's t-test, which does not assume equal\n",
      "            population variance [2]_.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            The calculated t-statistics\n",
      "        pvalue : float or array\n",
      "            The two-tailed p-value.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        scipy.stats.ttest_ind\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        .. versionadded:: 0.16.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://en.wikipedia.org/wiki/T-test#Independent_two-sample_t-test\n",
      "        \n",
      "        .. [2] http://en.wikipedia.org/wiki/Welch%27s_t_test\n",
      "    \n",
      "    ttest_rel(a, b, axis=0, nan_policy='propagate')\n",
      "        Calculates the T-test on TWO RELATED samples of scores, a and b.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that 2 related or\n",
      "        repeated samples have identical average (expected) values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : array_like\n",
      "            The arrays must have the same shape.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. If None, compute over the whole\n",
      "            arrays, `a`, and `b`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            t-statistic\n",
      "        pvalue : float or array\n",
      "            two-tailed p-value\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Examples for the use are scores of the same set of student in\n",
      "        different exams, or repeated sampling from the same units. The\n",
      "        test measures whether the average score differs significantly\n",
      "        across samples (e.g. exams). If we observe a large p-value, for\n",
      "        example greater than 0.05 or 0.1 then we cannot reject the null\n",
      "        hypothesis of identical average scores. If the p-value is smaller\n",
      "        than the threshold, e.g. 1%, 5% or 10%, then we reject the null\n",
      "        hypothesis of equal averages. Small p-values are associated with\n",
      "        large t-statistics.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        http://en.wikipedia.org/wiki/T-test#Dependent_t-test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> np.random.seed(12345678) # fix random seed to get same numbers\n",
      "        \n",
      "        >>> rvs1 = stats.norm.rvs(loc=5,scale=10,size=500)\n",
      "        >>> rvs2 = (stats.norm.rvs(loc=5,scale=10,size=500) +\n",
      "        ...         stats.norm.rvs(scale=0.2,size=500))\n",
      "        >>> stats.ttest_rel(rvs1,rvs2)\n",
      "        (0.24101764965300962, 0.80964043445811562)\n",
      "        >>> rvs3 = (stats.norm.rvs(loc=8,scale=10,size=500) +\n",
      "        ...         stats.norm.rvs(scale=0.2,size=500))\n",
      "        >>> stats.ttest_rel(rvs1,rvs3)\n",
      "        (-3.9995108708727933, 7.3082402191726459e-005)\n",
      "    \n",
      "    tvar(a, limits=None, inclusive=(True, True), axis=0, ddof=1)\n",
      "        Compute the trimmed variance\n",
      "        \n",
      "        This function computes the sample variance of an array of values,\n",
      "        while ignoring values which are outside of given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored. When limits is None, then all values are\n",
      "            used. Either of the limit values in the tuple can also be None\n",
      "            representing a half-open interval.  The default value is None.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tvar : float\n",
      "            Trimmed variance.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `tvar` computes the unbiased sample variance, i.e. it uses a correction\n",
      "        factor ``n / (n - 1)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tvar(x)\n",
      "        35.0\n",
      "        >>> stats.tvar(x, (3,17))\n",
      "        20.0\n",
      "    \n",
      "    variation(a, axis=0, nan_policy='propagate')\n",
      "        Computes the coefficient of variation, the ratio of the biased standard\n",
      "        deviation to the mean.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to calculate the coefficient of variation. Default\n",
      "            is 0. If None, compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        variation : ndarray\n",
      "            The calculated variation along the requested axis.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "    \n",
      "    wilcoxon(x, y=None, zero_method='wilcox', correction=False)\n",
      "        Calculate the Wilcoxon signed-rank test.\n",
      "        \n",
      "        The Wilcoxon signed-rank test tests the null hypothesis that two\n",
      "        related paired samples come from the same distribution. In particular,\n",
      "        it tests whether the distribution of the differences x - y is symmetric\n",
      "        about zero. It is a non-parametric version of the paired T-test.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            The first set of measurements.\n",
      "        y : array_like, optional\n",
      "            The second set of measurements.  If `y` is not given, then the `x`\n",
      "            array is considered to be the differences between the two sets of\n",
      "            measurements.\n",
      "        zero_method : string, {\"pratt\", \"wilcox\", \"zsplit\"}, optional\n",
      "            \"pratt\":\n",
      "                Pratt treatment: includes zero-differences in the ranking process\n",
      "                (more conservative)\n",
      "            \"wilcox\":\n",
      "                Wilcox treatment: discards all zero-differences\n",
      "            \"zsplit\":\n",
      "                Zero rank split: just like Pratt, but spliting the zero rank\n",
      "                between positive and negative ones\n",
      "        correction : bool, optional\n",
      "            If True, apply continuity correction by adjusting the Wilcoxon rank\n",
      "            statistic by 0.5 towards the mean value when computing the\n",
      "            z-statistic.  Default is False.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The sum of the ranks of the differences above or below zero, whichever\n",
      "            is smaller.\n",
      "        pvalue : float\n",
      "            The two-sided p-value for the test.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Because the normal approximation is used for the calculations, the\n",
      "        samples used should be large.  A typical rule is to require that\n",
      "        n > 20.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] http://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test\n",
      "    \n",
      "    zmap(scores, compare, axis=0, ddof=0)\n",
      "        Calculates the relative z-scores.\n",
      "        \n",
      "        Returns an array of z-scores, i.e., scores that are standardized to\n",
      "        zero mean and unit variance, where mean and variance are calculated\n",
      "        from the comparison array.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        scores : array_like\n",
      "            The input for which z-scores are calculated.\n",
      "        compare : array_like\n",
      "            The input from which the mean and standard deviation of the\n",
      "            normalization are taken; assumed to have the same dimension as\n",
      "            `scores`.\n",
      "        axis : int or None, optional\n",
      "            Axis over which mean and variance of `compare` are calculated.\n",
      "            Default is 0. If None, compute over the whole array `scores`.\n",
      "        ddof : int, optional\n",
      "            Degrees of freedom correction in the calculation of the\n",
      "            standard deviation. Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        zscore : array_like\n",
      "            Z-scores, in the same shape as `scores`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function preserves ndarray subclasses, and works also with\n",
      "        matrices and masked arrays (it uses `asanyarray` instead of\n",
      "        `asarray` for parameters).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import zmap\n",
      "        >>> a = [0.5, 2.0, 2.5, 3]\n",
      "        >>> b = [0, 1, 2, 3, 4]\n",
      "        >>> zmap(a, b)\n",
      "        array([-1.06066017,  0.        ,  0.35355339,  0.70710678])\n",
      "    \n",
      "    zscore(a, axis=0, ddof=0)\n",
      "        Calculates the z score of each value in the sample, relative to the\n",
      "        sample mean and standard deviation.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            An array like object containing the sample data.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Degrees of freedom correction in the calculation of the\n",
      "            standard deviation. Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        zscore : array_like\n",
      "            The z-scores, standardized by mean and standard deviation of\n",
      "            input array `a`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function preserves ndarray subclasses, and works also with\n",
      "        matrices and masked arrays (it uses `asanyarray` instead of\n",
      "        `asarray` for parameters).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> a = np.array([ 0.7972,  0.0767,  0.4383,  0.7866,  0.8091,\n",
      "        ...                0.1954,  0.6307,  0.6599,  0.1065,  0.0508])\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.zscore(a)\n",
      "        array([ 1.1273, -1.247 , -0.0552,  1.0923,  1.1664, -0.8559,  0.5786,\n",
      "                0.6748, -1.1488, -1.3324])\n",
      "        \n",
      "        Computing along a specified axis, using n-1 degrees of freedom\n",
      "        (``ddof=1``) to calculate the standard deviation:\n",
      "        \n",
      "        >>> b = np.array([[ 0.3148,  0.0478,  0.6243,  0.4608],\n",
      "        ...               [ 0.7149,  0.0775,  0.6072,  0.9656],\n",
      "        ...               [ 0.6341,  0.1403,  0.9759,  0.4064],\n",
      "        ...               [ 0.5918,  0.6948,  0.904 ,  0.3721],\n",
      "        ...               [ 0.0921,  0.2481,  0.1188,  0.1366]])\n",
      "        >>> stats.zscore(b, axis=1, ddof=1)\n",
      "        array([[-0.19264823, -1.28415119,  1.07259584,  0.40420358],\n",
      "               [ 0.33048416, -1.37380874,  0.04251374,  1.00081084],\n",
      "               [ 0.26796377, -1.12598418,  1.23283094, -0.37481053],\n",
      "               [-0.22095197,  0.24468594,  1.19042819, -1.21416216],\n",
      "               [-0.82780366,  1.4457416 , -0.43867764, -0.1792603 ]])\n",
      "\n",
      "DATA\n",
      "    __all__ = ['absolute_import', 'alpha', 'anderson', 'anderson_ksamp', '...\n",
      "    absolute_import = _Feature((2, 5, 0, 'alpha', 1), (3, 0, 0, 'alpha', 0...\n",
      "    alpha = <scipy.stats._continuous_distns.alpha_gen object>\n",
      "    anglit = <scipy.stats._continuous_distns.anglit_gen object>\n",
      "    arcsine = <scipy.stats._continuous_distns.arcsine_gen object>\n",
      "    bernoulli = <scipy.stats._discrete_distns.bernoulli_gen object>\n",
      "    beta = <scipy.stats._continuous_distns.beta_gen object>\n",
      "    betaprime = <scipy.stats._continuous_distns.betaprime_gen object>\n",
      "    binom = <scipy.stats._discrete_distns.binom_gen object>\n",
      "    boltzmann = <scipy.stats._discrete_distns.boltzmann_gen object>\n",
      "    bradford = <scipy.stats._continuous_distns.bradford_gen object>\n",
      "    burr = <scipy.stats._continuous_distns.burr_gen object>\n",
      "    burr12 = <scipy.stats._continuous_distns.burr12_gen object>\n",
      "    cauchy = <scipy.stats._continuous_distns.cauchy_gen object>\n",
      "    chi = <scipy.stats._continuous_distns.chi_gen object>\n",
      "    chi2 = <scipy.stats._continuous_distns.chi2_gen object>\n",
      "    cosine = <scipy.stats._continuous_distns.cosine_gen object>\n",
      "    dgamma = <scipy.stats._continuous_distns.dgamma_gen object>\n",
      "    dirichlet = <scipy.stats._multivariate.dirichlet_gen object>\n",
      "    division = _Feature((2, 2, 0, 'alpha', 2), (3, 0, 0, 'alpha', 0), 8192...\n",
      "    dlaplace = <scipy.stats._discrete_distns.dlaplace_gen object>\n",
      "    dweibull = <scipy.stats._continuous_distns.dweibull_gen object>\n",
      "    erlang = <scipy.stats._continuous_distns.erlang_gen object>\n",
      "    expon = <scipy.stats._continuous_distns.expon_gen object>\n",
      "    exponnorm = <scipy.stats._continuous_distns.exponnorm_gen object>\n",
      "    exponpow = <scipy.stats._continuous_distns.exponpow_gen object>\n",
      "    exponweib = <scipy.stats._continuous_distns.exponweib_gen object>\n",
      "    f = <scipy.stats._continuous_distns.f_gen object>\n",
      "    fatiguelife = <scipy.stats._continuous_distns.fatiguelife_gen object>\n",
      "    fisk = <scipy.stats._continuous_distns.fisk_gen object>\n",
      "    foldcauchy = <scipy.stats._continuous_distns.foldcauchy_gen object>\n",
      "    foldnorm = <scipy.stats._continuous_distns.foldnorm_gen object>\n",
      "    frechet_l = <scipy.stats._continuous_distns.frechet_l_gen object>\n",
      "    frechet_r = <scipy.stats._continuous_distns.frechet_r_gen object>\n",
      "    gamma = <scipy.stats._continuous_distns.gamma_gen object>\n",
      "    gausshyper = <scipy.stats._continuous_distns.gausshyper_gen object>\n",
      "    genexpon = <scipy.stats._continuous_distns.genexpon_gen object>\n",
      "    genextreme = <scipy.stats._continuous_distns.genextreme_gen object>\n",
      "    gengamma = <scipy.stats._continuous_distns.gengamma_gen object>\n",
      "    genhalflogistic = <scipy.stats._continuous_distns.genhalflogistic_gen ...\n",
      "    genlogistic = <scipy.stats._continuous_distns.genlogistic_gen object>\n",
      "    gennorm = <scipy.stats._continuous_distns.gennorm_gen object>\n",
      "    genpareto = <scipy.stats._continuous_distns.genpareto_gen object>\n",
      "    geom = <scipy.stats._discrete_distns.geom_gen object>\n",
      "    gilbrat = <scipy.stats._continuous_distns.gilbrat_gen object>\n",
      "    gompertz = <scipy.stats._continuous_distns.gompertz_gen object>\n",
      "    gumbel_l = <scipy.stats._continuous_distns.gumbel_l_gen object>\n",
      "    gumbel_r = <scipy.stats._continuous_distns.gumbel_r_gen object>\n",
      "    halfcauchy = <scipy.stats._continuous_distns.halfcauchy_gen object>\n",
      "    halfgennorm = <scipy.stats._continuous_distns.halfgennorm_gen object>\n",
      "    halflogistic = <scipy.stats._continuous_distns.halflogistic_gen object...\n",
      "    halfnorm = <scipy.stats._continuous_distns.halfnorm_gen object>\n",
      "    hypergeom = <scipy.stats._discrete_distns.hypergeom_gen object>\n",
      "    hypsecant = <scipy.stats._continuous_distns.hypsecant_gen object>\n",
      "    invgamma = <scipy.stats._continuous_distns.invgamma_gen object>\n",
      "    invgauss = <scipy.stats._continuous_distns.invgauss_gen object>\n",
      "    invweibull = <scipy.stats._continuous_distns.invweibull_gen object>\n",
      "    invwishart = <scipy.stats._multivariate.invwishart_gen object>\n",
      "    johnsonsb = <scipy.stats._continuous_distns.johnsonsb_gen object>\n",
      "    johnsonsu = <scipy.stats._continuous_distns.johnsonsu_gen object>\n",
      "    kappa3 = <scipy.stats._continuous_distns.kappa3_gen object>\n",
      "    kappa4 = <scipy.stats._continuous_distns.kappa4_gen object>\n",
      "    ksone = <scipy.stats._continuous_distns.ksone_gen object>\n",
      "    kstwobign = <scipy.stats._continuous_distns.kstwobign_gen object>\n",
      "    laplace = <scipy.stats._continuous_distns.laplace_gen object>\n",
      "    levy = <scipy.stats._continuous_distns.levy_gen object>\n",
      "    levy_l = <scipy.stats._continuous_distns.levy_l_gen object>\n",
      "    levy_stable = <scipy.stats._continuous_distns.levy_stable_gen object>\n",
      "    loggamma = <scipy.stats._continuous_distns.loggamma_gen object>\n",
      "    logistic = <scipy.stats._continuous_distns.logistic_gen object>\n",
      "    loglaplace = <scipy.stats._continuous_distns.loglaplace_gen object>\n",
      "    lognorm = <scipy.stats._continuous_distns.lognorm_gen object>\n",
      "    logser = <scipy.stats._discrete_distns.logser_gen object>\n",
      "    lomax = <scipy.stats._continuous_distns.lomax_gen object>\n",
      "    matrix_normal = <scipy.stats._multivariate.matrix_normal_gen object>\n",
      "    maxwell = <scipy.stats._continuous_distns.maxwell_gen object>\n",
      "    mielke = <scipy.stats._continuous_distns.mielke_gen object>\n",
      "    multivariate_normal = <scipy.stats._multivariate.multivariate_normal_g...\n",
      "    nakagami = <scipy.stats._continuous_distns.nakagami_gen object>\n",
      "    nbinom = <scipy.stats._discrete_distns.nbinom_gen object>\n",
      "    ncf = <scipy.stats._continuous_distns.ncf_gen object>\n",
      "    nct = <scipy.stats._continuous_distns.nct_gen object>\n",
      "    ncx2 = <scipy.stats._continuous_distns.ncx2_gen object>\n",
      "    norm = <scipy.stats._continuous_distns.norm_gen object>\n",
      "    ortho_group = <scipy.stats._multivariate.ortho_group_gen object>\n",
      "    pareto = <scipy.stats._continuous_distns.pareto_gen object>\n",
      "    pearson3 = <scipy.stats._continuous_distns.pearson3_gen object>\n",
      "    planck = <scipy.stats._discrete_distns.planck_gen object>\n",
      "    poisson = <scipy.stats._discrete_distns.poisson_gen object>\n",
      "    powerlaw = <scipy.stats._continuous_distns.powerlaw_gen object>\n",
      "    powerlognorm = <scipy.stats._continuous_distns.powerlognorm_gen object...\n",
      "    powernorm = <scipy.stats._continuous_distns.powernorm_gen object>\n",
      "    print_function = _Feature((2, 6, 0, 'alpha', 2), (3, 0, 0, 'alpha', 0)...\n",
      "    randint = <scipy.stats._discrete_distns.randint_gen object>\n",
      "    random_correlation = <scipy.stats._multivariate.random_correlation_gen...\n",
      "    rayleigh = <scipy.stats._continuous_distns.rayleigh_gen object>\n",
      "    rdist = <scipy.stats._continuous_distns.rdist_gen object>\n",
      "    recipinvgauss = <scipy.stats._continuous_distns.recipinvgauss_gen obje...\n",
      "    reciprocal = <scipy.stats._continuous_distns.reciprocal_gen object>\n",
      "    rice = <scipy.stats._continuous_distns.rice_gen object>\n",
      "    semicircular = <scipy.stats._continuous_distns.semicircular_gen object...\n",
      "    skellam = <scipy.stats._discrete_distns.skellam_gen object>\n",
      "    skewnorm = <scipy.stats._continuous_distns.skew_norm_gen object>\n",
      "    special_ortho_group = <scipy.stats._multivariate.special_ortho_group_g...\n",
      "    t = <scipy.stats._continuous_distns.t_gen object>\n",
      "    trapz = <scipy.stats._continuous_distns.trapz_gen object>\n",
      "    triang = <scipy.stats._continuous_distns.triang_gen object>\n",
      "    truncexpon = <scipy.stats._continuous_distns.truncexpon_gen object>\n",
      "    truncnorm = <scipy.stats._continuous_distns.truncnorm_gen object>\n",
      "    tukeylambda = <scipy.stats._continuous_distns.tukeylambda_gen object>\n",
      "    uniform = <scipy.stats._continuous_distns.uniform_gen object>\n",
      "    vonmises = <scipy.stats._continuous_distns.vonmises_gen object>\n",
      "    vonmises_line = <scipy.stats._continuous_distns.vonmises_gen object>\n",
      "    wald = <scipy.stats._continuous_distns.wald_gen object>\n",
      "    weibull_max = <scipy.stats._continuous_distns.frechet_l_gen object>\n",
      "    weibull_min = <scipy.stats._continuous_distns.frechet_r_gen object>\n",
      "    wishart = <scipy.stats._multivariate.wishart_gen object>\n",
      "    wrapcauchy = <scipy.stats._continuous_distns.wrapcauchy_gen object>\n",
      "    zipf = <scipy.stats._discrete_distns.zipf_gen object>\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as sp\n",
    "help(sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module scipy.stats.distributions in scipy.stats:\n",
      "\n",
      "NAME\n",
      "    scipy.stats.distributions\n",
      "\n",
      "FILE\n",
      "    d:\\users\\byron\\anaconda2\\lib\\site-packages\\scipy\\stats\\distributions.py\n",
      "\n",
      "DESCRIPTION\n",
      "    # Author:  Travis Oliphant  2002-2011 with contributions from\n",
      "    #          SciPy Developers 2004-2011\n",
      "    #\n",
      "    # NOTE: To look at history using `git blame`, use `git blame -M -C -C`\n",
      "    #       instead of `git blame -Lxxx,+x`.\n",
      "    #\n",
      "\n",
      "CLASSES\n",
      "    scipy.stats._distn_infrastructure.rv_generic(__builtin__.object)\n",
      "        scipy.stats._distn_infrastructure.rv_continuous\n",
      "        scipy.stats._distn_infrastructure.rv_discrete\n",
      "    \n",
      "    class rv_continuous(rv_generic)\n",
      "     |  A generic continuous random variable class meant for subclassing.\n",
      "     |  \n",
      "     |  `rv_continuous` is a base class to construct specific distribution classes\n",
      "     |  and instances for continuous random variables. It cannot be used\n",
      "     |  directly as a distribution.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  momtype : int, optional\n",
      "     |      The type of generic moment calculation to use: 0 for pdf, 1 (default)\n",
      "     |      for ppf.\n",
      "     |  a : float, optional\n",
      "     |      Lower bound of the support of the distribution, default is minus\n",
      "     |      infinity.\n",
      "     |  b : float, optional\n",
      "     |      Upper bound of the support of the distribution, default is plus\n",
      "     |      infinity.\n",
      "     |  xtol : float, optional\n",
      "     |      The tolerance for fixed point calculation for generic ppf.\n",
      "     |  badvalue : float, optional\n",
      "     |      The value in a result arrays that indicates a value that for which\n",
      "     |      some argument restriction is violated, default is np.nan.\n",
      "     |  name : str, optional\n",
      "     |      The name of the instance. This string is used to construct the default\n",
      "     |      example for distributions.\n",
      "     |  longname : str, optional\n",
      "     |      This string is used as part of the first line of the docstring returned\n",
      "     |      when a subclass has no docstring of its own. Note: `longname` exists\n",
      "     |      for backwards compatibility, do not use for new subclasses.\n",
      "     |  shapes : str, optional\n",
      "     |      The shape of the distribution. For example ``\"m, n\"`` for a\n",
      "     |      distribution that takes two integers as the two shape arguments for all\n",
      "     |      its methods. If not provided, shape parameters will be inferred from\n",
      "     |      the signature of the private methods, ``_pdf`` and ``_cdf`` of the\n",
      "     |      instance.\n",
      "     |  extradoc :  str, optional, deprecated\n",
      "     |      This string is used as the last part of the docstring returned when a\n",
      "     |      subclass has no docstring of its own. Note: `extradoc` exists for\n",
      "     |      backwards compatibility, do not use for new subclasses.\n",
      "     |  seed : None or int or ``numpy.random.RandomState`` instance, optional\n",
      "     |      This parameter defines the RandomState object to use for drawing\n",
      "     |      random variates.\n",
      "     |      If None (or np.random), the global np.random state is used.\n",
      "     |      If integer, it is used to seed the local RandomState instance.\n",
      "     |      Default is None.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  rvs\n",
      "     |  pdf\n",
      "     |  logpdf\n",
      "     |  cdf\n",
      "     |  logcdf\n",
      "     |  sf\n",
      "     |  logsf\n",
      "     |  ppf\n",
      "     |  isf\n",
      "     |  moment\n",
      "     |  stats\n",
      "     |  entropy\n",
      "     |  expect\n",
      "     |  median\n",
      "     |  mean\n",
      "     |  std\n",
      "     |  var\n",
      "     |  interval\n",
      "     |  __call__\n",
      "     |  fit\n",
      "     |  fit_loc_scale\n",
      "     |  nnlf\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Public methods of an instance of a distribution class (e.g., ``pdf``,\n",
      "     |  ``cdf``) check their arguments and pass valid arguments to private,\n",
      "     |  computational methods (``_pdf``, ``_cdf``). For ``pdf(x)``, ``x`` is valid\n",
      "     |  if it is within the support of a distribution, ``self.a <= x <= self.b``.\n",
      "     |  Whether a shape parameter is valid is decided by an ``_argcheck`` method\n",
      "     |  (which defaults to checking that its arguments are strictly positive.)\n",
      "     |  \n",
      "     |  **Subclassing**\n",
      "     |  \n",
      "     |  New random variables can be defined by subclassing the `rv_continuous` class\n",
      "     |  and re-defining at least the ``_pdf`` or the ``_cdf`` method (normalized\n",
      "     |  to location 0 and scale 1).\n",
      "     |  \n",
      "     |  If positive argument checking is not correct for your RV\n",
      "     |  then you will also need to re-define the ``_argcheck`` method.\n",
      "     |  \n",
      "     |  Correct, but potentially slow defaults exist for the remaining\n",
      "     |  methods but for speed and/or accuracy you can over-ride::\n",
      "     |  \n",
      "     |    _logpdf, _cdf, _logcdf, _ppf, _rvs, _isf, _sf, _logsf\n",
      "     |  \n",
      "     |  Rarely would you override ``_isf``, ``_sf`` or ``_logsf``, but you could.\n",
      "     |  \n",
      "     |  **Methods that can be overwritten by subclasses**\n",
      "     |  ::\n",
      "     |  \n",
      "     |    _rvs\n",
      "     |    _pdf\n",
      "     |    _cdf\n",
      "     |    _sf\n",
      "     |    _ppf\n",
      "     |    _isf\n",
      "     |    _stats\n",
      "     |    _munp\n",
      "     |    _entropy\n",
      "     |    _argcheck\n",
      "     |  \n",
      "     |  There are additional (internal and private) generic methods that can\n",
      "     |  be useful for cross-checking and for debugging, but might work in all\n",
      "     |  cases when directly called.\n",
      "     |  \n",
      "     |  A note on ``shapes``: subclasses need not specify them explicitly. In this\n",
      "     |  case, `shapes` will be automatically deduced from the signatures of the\n",
      "     |  overridden methods (`pdf`, `cdf` etc).\n",
      "     |  If, for some reason, you prefer to avoid relying on introspection, you can\n",
      "     |  specify ``shapes`` explicitly as an argument to the instance constructor.\n",
      "     |  \n",
      "     |  \n",
      "     |  **Frozen Distributions**\n",
      "     |  \n",
      "     |  Normally, you must provide shape parameters (and, optionally, location and\n",
      "     |  scale parameters to each call of a method of a distribution.\n",
      "     |  \n",
      "     |  Alternatively, the object may be called (as a function) to fix the shape,\n",
      "     |  location, and scale parameters returning a \"frozen\" continuous RV object:\n",
      "     |  \n",
      "     |  rv = generic(<shape(s)>, loc=0, scale=1)\n",
      "     |      frozen RV object with the same methods but holding the given shape,\n",
      "     |      location, and scale fixed\n",
      "     |  \n",
      "     |  **Statistics**\n",
      "     |  \n",
      "     |  Statistics are computed using numerical integration by default.\n",
      "     |  For speed you can redefine this using ``_stats``:\n",
      "     |  \n",
      "     |   - take shape parameters and return mu, mu2, g1, g2\n",
      "     |   - If you can't compute one of these, return it as None\n",
      "     |   - Can also be defined with a keyword argument ``moments``, which is a\n",
      "     |     string composed of \"m\", \"v\", \"s\", and/or \"k\".\n",
      "     |     Only the components appearing in string should be computed and\n",
      "     |     returned in the order \"m\", \"v\", \"s\", or \"k\"  with missing values\n",
      "     |     returned as None.\n",
      "     |  \n",
      "     |  Alternatively, you can override ``_munp``, which takes ``n`` and shape\n",
      "     |  parameters and returns the n-th non-central moment of the distribution.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  To create a new Gaussian distribution, we would do the following:\n",
      "     |  \n",
      "     |  >>> from scipy.stats import rv_continuous\n",
      "     |  >>> class gaussian_gen(rv_continuous):\n",
      "     |  ...     \"Gaussian distribution\"\n",
      "     |  ...     def _pdf(self, x):\n",
      "     |  ...         return np.exp(-x**2 / 2.) / np.sqrt(2.0 * np.pi)\n",
      "     |  >>> gaussian = gaussian_gen(name='gaussian')\n",
      "     |  \n",
      "     |  ``scipy.stats`` distributions are *instances*, so here we subclass\n",
      "     |  `rv_continuous` and create an instance. With this, we now have\n",
      "     |  a fully functional distribution with all relevant methods automagically\n",
      "     |  generated by the framework.\n",
      "     |  \n",
      "     |  Note that above we defined a standard normal distribution, with zero mean\n",
      "     |  and unit variance. Shifting and scaling of the distribution can be done\n",
      "     |  by using ``loc`` and ``scale`` parameters: ``gaussian.pdf(x, loc, scale)``\n",
      "     |  essentially computes ``y = (x - loc) / scale`` and\n",
      "     |  ``gaussian._pdf(y) / scale``.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_continuous\n",
      "     |      rv_generic\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, momtype=1, a=None, b=None, xtol=1e-14, badvalue=None, name=None, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  cdf(self, x, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `x`\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, scale=1, lb=None, ub=None, conditional=False, **kwds)\n",
      "     |      Calculate expected value of a function with respect to the\n",
      "     |      distribution.\n",
      "     |      \n",
      "     |      The expected value of a function ``f(x)`` with respect to a\n",
      "     |      distribution ``dist`` is defined as::\n",
      "     |      \n",
      "     |                  ubound\n",
      "     |          E[x] = Integral(f(x) * dist.pdf(x))\n",
      "     |                  lbound\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which integral is calculated. Takes only one argument.\n",
      "     |          The default is the identity mapping f(x) = x.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : float, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      lb, ub : scalar, optional\n",
      "     |          Lower and upper bound for integration. Default is set to the\n",
      "     |          support of the distribution.\n",
      "     |      conditional : bool, optional\n",
      "     |          If True, the integral is corrected by the conditional probability\n",
      "     |          of the integration interval.  The return value is the expectation\n",
      "     |          of the function, conditional on being in the given interval.\n",
      "     |          Default is False.\n",
      "     |      \n",
      "     |      Additional keyword arguments are passed to the integration routine.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          The calculated expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The integration behavior of this function is inherited from\n",
      "     |      `integrate.quad`.\n",
      "     |  \n",
      "     |  fit(self, data, *args, **kwds)\n",
      "     |      Return MLEs for shape, location, and scale parameters from data.\n",
      "     |      \n",
      "     |      MLE stands for Maximum Likelihood Estimate.  Starting estimates for\n",
      "     |      the fit are given by input arguments; for any arguments not provided\n",
      "     |      with starting estimates, ``self._fitstart(data)`` is called to generate\n",
      "     |      such.\n",
      "     |      \n",
      "     |      One can hold some parameters fixed to specific values by passing in\n",
      "     |      keyword arguments ``f0``, ``f1``, ..., ``fn`` (for shape parameters)\n",
      "     |      and ``floc`` and ``fscale`` (for location and scale parameters,\n",
      "     |      respectively).\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to use in calculating the MLEs.\n",
      "     |      args : floats, optional\n",
      "     |          Starting value(s) for any shape-characterizing arguments (those not\n",
      "     |          provided will be determined by a call to ``_fitstart(data)``).\n",
      "     |          No default value.\n",
      "     |      kwds : floats, optional\n",
      "     |          Starting values for the location and scale parameters; no default.\n",
      "     |          Special keyword arguments are recognized as holding certain\n",
      "     |          parameters fixed:\n",
      "     |      \n",
      "     |          - f0...fn : hold respective shape parameters fixed.\n",
      "     |            Alternatively, shape parameters to fix can be specified by name.\n",
      "     |            For example, if ``self.shapes == \"a, b\"``, ``fa``and ``fix_a``\n",
      "     |            are equivalent to ``f0``, and ``fb`` and ``fix_b`` are\n",
      "     |            equivalent to ``f1``.\n",
      "     |      \n",
      "     |          - floc : hold location parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - fscale : hold scale parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - optimizer : The optimizer to use.  The optimizer must take ``func``,\n",
      "     |            and starting position as the first two arguments,\n",
      "     |            plus ``args`` (for extra arguments to pass to the\n",
      "     |            function to be optimized) and ``disp=0`` to suppress\n",
      "     |            output as keyword arguments.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      shape, loc, scale : tuple of floats\n",
      "     |          MLEs for any shape statistics, followed by those for location and\n",
      "     |          scale.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This fit is computed by maximizing a log-likelihood function, with\n",
      "     |      penalty applied for samples outside of range of the distribution. The\n",
      "     |      returned answer is not guaranteed to be the globally optimal MLE, it\n",
      "     |      may only be locally optimal, or the optimization may fail altogether.\n",
      "     |      \n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Generate some data to fit: draw random variates from the `beta`\n",
      "     |      distribution\n",
      "     |      \n",
      "     |      >>> from scipy.stats import beta\n",
      "     |      >>> a, b = 1., 2.\n",
      "     |      >>> x = beta.rvs(a, b, size=1000)\n",
      "     |      \n",
      "     |      Now we can fit all four parameters (``a``, ``b``, ``loc`` and ``scale``):\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x)\n",
      "     |      \n",
      "     |      We can also use some prior knowledge about the dataset: let's keep\n",
      "     |      ``loc`` and ``scale`` fixed:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, floc=0, fscale=1)\n",
      "     |      >>> loc1, scale1\n",
      "     |      (0, 1)\n",
      "     |      \n",
      "     |      We can also keep shape parameters fixed by using ``f``-keywords. To\n",
      "     |      keep the zero-th shape parameter ``a`` equal 1, use ``f0=1`` or,\n",
      "     |      equivalently, ``fa=1``:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, fa=1, floc=0, fscale=1)\n",
      "     |      >>> a1\n",
      "     |      1\n",
      "     |  \n",
      "     |  fit_loc_scale(self, data, *args)\n",
      "     |      Estimate loc and scale parameters from data using 1st and 2nd moments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to fit.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Lhat : float\n",
      "     |          Estimated location parameter for the data.\n",
      "     |      Shat : float\n",
      "     |          Estimated scale parameter for the data.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          upper tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability q.\n",
      "     |  \n",
      "     |  logcdf(self, x, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at x\n",
      "     |  \n",
      "     |  logpdf(self, x, *args, **kwds)\n",
      "     |      Log of the probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      This uses a more numerically accurate calculation if available.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpdf : array_like\n",
      "     |          Log of the probability density function evaluated at x\n",
      "     |  \n",
      "     |  logsf(self, x, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as (1 - `cdf`),\n",
      "     |      evaluated at `x`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `x`.\n",
      "     |  \n",
      "     |  nnlf(self, theta, x)\n",
      "     |      Return negative loglikelihood function.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is ``-sum(log pdf(x, theta), axis=0)`` where `theta` are the\n",
      "     |      parameters (including loc and scale).\n",
      "     |  \n",
      "     |  pdf(self, x, *args, **kwds)\n",
      "     |      Probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pdf : ndarray\n",
      "     |          Probability density function evaluated at x\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          lower tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : array_like\n",
      "     |          quantile corresponding to the lower tail probability q.\n",
      "     |  \n",
      "     |  sf(self, x, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at x\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      stats.distributions.rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwds)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (default is 1).\n",
      "     |      random_state : None or int or ``np.random.RandomState`` instance, optional\n",
      "     |          If int or RandomState, use it for drawing the random variates.\n",
      "     |          If None, rely on ``self.random_state``.\n",
      "     |          Default is None.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the RandomState object for generating random variates.\n",
      "     |      \n",
      "     |      This can be either None or an existing RandomState object.\n",
      "     |      \n",
      "     |      If None (or np.random), use the RandomState singleton used by np.random.\n",
      "     |      If already a RandomState instance, use it.\n",
      "     |      If an int, use a new RandomState instance seeded with seed.\n",
      "    \n",
      "    class rv_discrete(rv_generic)\n",
      "     |  A generic discrete random variable class meant for subclassing.\n",
      "     |  \n",
      "     |  `rv_discrete` is a base class to construct specific distribution classes\n",
      "     |  and instances for discrete random variables. It can also be used\n",
      "     |  to construct an arbitrary distribution defined by a list of support\n",
      "     |  points and corresponding probabilities.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  a : float, optional\n",
      "     |      Lower bound of the support of the distribution, default: 0\n",
      "     |  b : float, optional\n",
      "     |      Upper bound of the support of the distribution, default: plus infinity\n",
      "     |  moment_tol : float, optional\n",
      "     |      The tolerance for the generic calculation of moments.\n",
      "     |  values : tuple of two array_like, optional\n",
      "     |      ``(xk, pk)`` where ``xk`` are integers with non-zero\n",
      "     |      probabilities ``pk``  with ``sum(pk) = 1``.\n",
      "     |  inc : integer, optional\n",
      "     |      Increment for the support of the distribution.\n",
      "     |      Default is 1. (other values have not been tested)\n",
      "     |  badvalue : float, optional\n",
      "     |      The value in a result arrays that indicates a value that for which\n",
      "     |      some argument restriction is violated, default is np.nan.\n",
      "     |  name : str, optional\n",
      "     |      The name of the instance. This string is used to construct the default\n",
      "     |      example for distributions.\n",
      "     |  longname : str, optional\n",
      "     |      This string is used as part of the first line of the docstring returned\n",
      "     |      when a subclass has no docstring of its own. Note: `longname` exists\n",
      "     |      for backwards compatibility, do not use for new subclasses.\n",
      "     |  shapes : str, optional\n",
      "     |      The shape of the distribution. For example \"m, n\" for a distribution\n",
      "     |      that takes two integers as the two shape arguments for all its methods\n",
      "     |      If not provided, shape parameters will be inferred from\n",
      "     |      the signatures of the private methods, ``_pmf`` and ``_cdf`` of\n",
      "     |      the instance.\n",
      "     |  extradoc :  str, optional\n",
      "     |      This string is used as the last part of the docstring returned when a\n",
      "     |      subclass has no docstring of its own. Note: `extradoc` exists for\n",
      "     |      backwards compatibility, do not use for new subclasses.\n",
      "     |  seed : None or int or ``numpy.random.RandomState`` instance, optional\n",
      "     |      This parameter defines the RandomState object to use for drawing\n",
      "     |      random variates.\n",
      "     |      If None, the global np.random state is used.\n",
      "     |      If integer, it is used to seed the local RandomState instance.\n",
      "     |      Default is None.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  rvs\n",
      "     |  pmf\n",
      "     |  logpmf\n",
      "     |  cdf\n",
      "     |  logcdf\n",
      "     |  sf\n",
      "     |  logsf\n",
      "     |  ppf\n",
      "     |  isf\n",
      "     |  moment\n",
      "     |  stats\n",
      "     |  entropy\n",
      "     |  expect\n",
      "     |  median\n",
      "     |  mean\n",
      "     |  std\n",
      "     |  var\n",
      "     |  interval\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  \n",
      "     |  This class is similar to `rv_continuous`, the main differences being:\n",
      "     |  \n",
      "     |  - the support of the distribution is a set of integers\n",
      "     |  - instead of the probability density function, ``pdf`` (and the\n",
      "     |    corresponding private ``_pdf``), this class defines the\n",
      "     |    *probability mass function*, `pmf` (and the corresponding\n",
      "     |    private ``_pmf``.)\n",
      "     |  - scale parameter is not defined.\n",
      "     |  \n",
      "     |  To create a new discrete distribution, we would do the following:\n",
      "     |  \n",
      "     |  >>> from scipy.stats import rv_discrete\n",
      "     |  >>> class poisson_gen(rv_discrete):\n",
      "     |  ...     \"Poisson distribution\"\n",
      "     |  ...     def _pmf(self, k, mu):\n",
      "     |  ...         return exp(-mu) * mu**k / factorial(k)\n",
      "     |  \n",
      "     |  and create an instance::\n",
      "     |  \n",
      "     |  >>> poisson = poisson_gen(name=\"poisson\")\n",
      "     |  \n",
      "     |  Note that above we defined the Poisson distribution in the standard form.\n",
      "     |  Shifting the distribution can be done by providing the ``loc`` parameter\n",
      "     |  to the methods of the instance. For example, ``poisson.pmf(x, mu, loc)``\n",
      "     |  delegates the work to ``poisson._pmf(x-loc, mu)``.\n",
      "     |  \n",
      "     |  **Discrete distributions from a list of probabilities**\n",
      "     |  \n",
      "     |  Alternatively, you can construct an arbitrary discrete rv defined\n",
      "     |  on a finite set of values ``xk`` with ``Prob{X=xk} = pk`` by using the\n",
      "     |  ``values`` keyword argument to the `rv_discrete` constructor.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  \n",
      "     |  Custom made discrete distribution:\n",
      "     |  \n",
      "     |  >>> from scipy import stats\n",
      "     |  >>> xk = np.arange(7)\n",
      "     |  >>> pk = (0.1, 0.2, 0.3, 0.1, 0.1, 0.0, 0.2)\n",
      "     |  >>> custm = stats.rv_discrete(name='custm', values=(xk, pk))\n",
      "     |  >>>\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig, ax = plt.subplots(1, 1)\n",
      "     |  >>> ax.plot(xk, custm.pmf(xk), 'ro', ms=12, mec='r')\n",
      "     |  >>> ax.vlines(xk, 0, custm.pmf(xk), colors='r', lw=4)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Random number generation:\n",
      "     |  \n",
      "     |  >>> R = custm.rvs(size=100)\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_discrete\n",
      "     |      rv_generic\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  cdf(self, k, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like, int\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `k`.\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, lb=None, ub=None, conditional=False, maxcount=1000, tolerance=1e-10, chunksize=32)\n",
      "     |      Calculate expected value of a function with respect to the distribution\n",
      "     |      for discrete distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which the expectation value is calculated.\n",
      "     |          Takes only one argument.\n",
      "     |          The default is the identity mapping f(k) = k.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter.\n",
      "     |          Default is 0.\n",
      "     |      lb, ub : int, optional\n",
      "     |          Lower and upper bound for the summation, default is set to the\n",
      "     |          support of the distribution, inclusive (``ul <= k <= ub``).\n",
      "     |      conditional : bool, optional\n",
      "     |          If true then the expectation is corrected by the conditional\n",
      "     |          probability of the summation interval. The return value is the\n",
      "     |          expectation of the function, `func`, conditional on being in\n",
      "     |          the given interval (k such that ``ul <= k <= ub``).\n",
      "     |          Default is False.\n",
      "     |      maxcount : int, optional\n",
      "     |          Maximal number of terms to evaluate (to avoid an endless loop for\n",
      "     |          an infinite sum). Default is 1000.\n",
      "     |      tolerance : float, optional\n",
      "     |          Absolute tolerance for the summation. Default is 1e-10.\n",
      "     |      chunksize : int, optional\n",
      "     |          Iterate over the support of a distributions in chunks of this size.\n",
      "     |          Default is 32.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          Expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      For heavy-tailed distributions, the expected value may or may not exist,\n",
      "     |      depending on the function, `func`. If it does exist, but the sum converges\n",
      "     |      slowly, the accuracy of the result may be rather low. For instance, for\n",
      "     |      ``zipf(4)``, accuracy for mean, variance in example is only 1e-5.\n",
      "     |      increasing `maxcount` and/or `chunksize` may improve the result, but may also\n",
      "     |      make zipf very slow.\n",
      "     |      \n",
      "     |      The function is not vectorized.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          Upper tail probability.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      k : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability, q.\n",
      "     |  \n",
      "     |  logcdf(self, k, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like, int\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at k.\n",
      "     |  \n",
      "     |  logpmf(self, k, *args, **kwds)\n",
      "     |      Log of the probability mass function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter. Default is 0.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpmf : array_like\n",
      "     |          Log of the probability mass function evaluated at k.\n",
      "     |  \n",
      "     |  logsf(self, k, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as 1 - `cdf`,\n",
      "     |      evaluated at `k`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `k`.\n",
      "     |  \n",
      "     |  pmf(self, k, *args, **kwds)\n",
      "     |      Probability mass function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pmf : array_like\n",
      "     |          Probability mass function evaluated at k\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          Lower tail probability.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      k : array_like\n",
      "     |          Quantile corresponding to the lower tail probability, q.\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwargs)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (Default is 1).  Note that `size`\n",
      "     |          has to be given as keyword, not as positional argument.\n",
      "     |      random_state : None or int or ``np.random.RandomState`` instance, optional\n",
      "     |          If int or RandomState, use it for drawing the random variates.\n",
      "     |          If None, rely on ``self.random_state``.\n",
      "     |          Default is None.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  sf(self, k, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at k.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods defined here:\n",
      "     |  \n",
      "     |  __new__(cls, a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  return_integers\n",
      "     |      `return_integers` is deprecated!\n",
      "     |      `return_integers` attribute is not used anywhere any  longer and is deprecated in scipy 0.18.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      stats.distributions.rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the RandomState object for generating random variates.\n",
      "     |      \n",
      "     |      This can be either None or an existing RandomState object.\n",
      "     |      \n",
      "     |      If None (or np.random), use the RandomState singleton used by np.random.\n",
      "     |      If already a RandomState instance, use it.\n",
      "     |      If an int, use a new RandomState instance seeded with seed.\n",
      "\n",
      "FUNCTIONS\n",
      "    entropy(pk, qk=None, base=None)\n",
      "        Calculate the entropy of a distribution for given probability values.\n",
      "        \n",
      "        If only probabilities `pk` are given, the entropy is calculated as\n",
      "        ``S = -sum(pk * log(pk), axis=0)``.\n",
      "        \n",
      "        If `qk` is not None, then compute the Kullback-Leibler divergence\n",
      "        ``S = sum(pk * log(pk / qk), axis=0)``.\n",
      "        \n",
      "        This routine will normalize `pk` and `qk` if they don't sum to 1.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pk : sequence\n",
      "            Defines the (discrete) distribution. ``pk[i]`` is the (possibly\n",
      "            unnormalized) probability of event ``i``.\n",
      "        qk : sequence, optional\n",
      "            Sequence against which the relative entropy is computed. Should be in\n",
      "            the same format as `pk`.\n",
      "        base : float, optional\n",
      "            The logarithmic base to use, defaults to ``e`` (natural logarithm).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        S : float\n",
      "            The calculated entropy.\n",
      "\n",
      "DATA\n",
      "    __all__ = ['entropy', 'rv_discrete', 'rv_continuous', 'genhalflogistic...\n",
      "    alpha = <scipy.stats._continuous_distns.alpha_gen object>\n",
      "    anglit = <scipy.stats._continuous_distns.anglit_gen object>\n",
      "    arcsine = <scipy.stats._continuous_distns.arcsine_gen object>\n",
      "    bernoulli = <scipy.stats._discrete_distns.bernoulli_gen object>\n",
      "    beta = <scipy.stats._continuous_distns.beta_gen object>\n",
      "    betaprime = <scipy.stats._continuous_distns.betaprime_gen object>\n",
      "    binom = <scipy.stats._discrete_distns.binom_gen object>\n",
      "    boltzmann = <scipy.stats._discrete_distns.boltzmann_gen object>\n",
      "    bradford = <scipy.stats._continuous_distns.bradford_gen object>\n",
      "    burr = <scipy.stats._continuous_distns.burr_gen object>\n",
      "    burr12 = <scipy.stats._continuous_distns.burr12_gen object>\n",
      "    cauchy = <scipy.stats._continuous_distns.cauchy_gen object>\n",
      "    chi = <scipy.stats._continuous_distns.chi_gen object>\n",
      "    chi2 = <scipy.stats._continuous_distns.chi2_gen object>\n",
      "    cosine = <scipy.stats._continuous_distns.cosine_gen object>\n",
      "    dgamma = <scipy.stats._continuous_distns.dgamma_gen object>\n",
      "    dlaplace = <scipy.stats._discrete_distns.dlaplace_gen object>\n",
      "    dweibull = <scipy.stats._continuous_distns.dweibull_gen object>\n",
      "    erlang = <scipy.stats._continuous_distns.erlang_gen object>\n",
      "    expon = <scipy.stats._continuous_distns.expon_gen object>\n",
      "    exponnorm = <scipy.stats._continuous_distns.exponnorm_gen object>\n",
      "    exponpow = <scipy.stats._continuous_distns.exponpow_gen object>\n",
      "    exponweib = <scipy.stats._continuous_distns.exponweib_gen object>\n",
      "    f = <scipy.stats._continuous_distns.f_gen object>\n",
      "    fatiguelife = <scipy.stats._continuous_distns.fatiguelife_gen object>\n",
      "    fisk = <scipy.stats._continuous_distns.fisk_gen object>\n",
      "    foldcauchy = <scipy.stats._continuous_distns.foldcauchy_gen object>\n",
      "    foldnorm = <scipy.stats._continuous_distns.foldnorm_gen object>\n",
      "    frechet_l = <scipy.stats._continuous_distns.frechet_l_gen object>\n",
      "    frechet_r = <scipy.stats._continuous_distns.frechet_r_gen object>\n",
      "    gamma = <scipy.stats._continuous_distns.gamma_gen object>\n",
      "    gausshyper = <scipy.stats._continuous_distns.gausshyper_gen object>\n",
      "    genexpon = <scipy.stats._continuous_distns.genexpon_gen object>\n",
      "    genextreme = <scipy.stats._continuous_distns.genextreme_gen object>\n",
      "    gengamma = <scipy.stats._continuous_distns.gengamma_gen object>\n",
      "    genhalflogistic = <scipy.stats._continuous_distns.genhalflogistic_gen ...\n",
      "    genlogistic = <scipy.stats._continuous_distns.genlogistic_gen object>\n",
      "    gennorm = <scipy.stats._continuous_distns.gennorm_gen object>\n",
      "    genpareto = <scipy.stats._continuous_distns.genpareto_gen object>\n",
      "    geom = <scipy.stats._discrete_distns.geom_gen object>\n",
      "    gilbrat = <scipy.stats._continuous_distns.gilbrat_gen object>\n",
      "    gompertz = <scipy.stats._continuous_distns.gompertz_gen object>\n",
      "    gumbel_l = <scipy.stats._continuous_distns.gumbel_l_gen object>\n",
      "    gumbel_r = <scipy.stats._continuous_distns.gumbel_r_gen object>\n",
      "    halfcauchy = <scipy.stats._continuous_distns.halfcauchy_gen object>\n",
      "    halfgennorm = <scipy.stats._continuous_distns.halfgennorm_gen object>\n",
      "    halflogistic = <scipy.stats._continuous_distns.halflogistic_gen object...\n",
      "    halfnorm = <scipy.stats._continuous_distns.halfnorm_gen object>\n",
      "    hypergeom = <scipy.stats._discrete_distns.hypergeom_gen object>\n",
      "    hypsecant = <scipy.stats._continuous_distns.hypsecant_gen object>\n",
      "    invgamma = <scipy.stats._continuous_distns.invgamma_gen object>\n",
      "    invgauss = <scipy.stats._continuous_distns.invgauss_gen object>\n",
      "    invweibull = <scipy.stats._continuous_distns.invweibull_gen object>\n",
      "    johnsonsb = <scipy.stats._continuous_distns.johnsonsb_gen object>\n",
      "    johnsonsu = <scipy.stats._continuous_distns.johnsonsu_gen object>\n",
      "    kappa3 = <scipy.stats._continuous_distns.kappa3_gen object>\n",
      "    kappa4 = <scipy.stats._continuous_distns.kappa4_gen object>\n",
      "    ksone = <scipy.stats._continuous_distns.ksone_gen object>\n",
      "    kstwobign = <scipy.stats._continuous_distns.kstwobign_gen object>\n",
      "    laplace = <scipy.stats._continuous_distns.laplace_gen object>\n",
      "    levy = <scipy.stats._continuous_distns.levy_gen object>\n",
      "    levy_l = <scipy.stats._continuous_distns.levy_l_gen object>\n",
      "    levy_stable = <scipy.stats._continuous_distns.levy_stable_gen object>\n",
      "    loggamma = <scipy.stats._continuous_distns.loggamma_gen object>\n",
      "    logistic = <scipy.stats._continuous_distns.logistic_gen object>\n",
      "    loglaplace = <scipy.stats._continuous_distns.loglaplace_gen object>\n",
      "    lognorm = <scipy.stats._continuous_distns.lognorm_gen object>\n",
      "    logser = <scipy.stats._discrete_distns.logser_gen object>\n",
      "    lomax = <scipy.stats._continuous_distns.lomax_gen object>\n",
      "    maxwell = <scipy.stats._continuous_distns.maxwell_gen object>\n",
      "    mielke = <scipy.stats._continuous_distns.mielke_gen object>\n",
      "    nakagami = <scipy.stats._continuous_distns.nakagami_gen object>\n",
      "    nbinom = <scipy.stats._discrete_distns.nbinom_gen object>\n",
      "    ncf = <scipy.stats._continuous_distns.ncf_gen object>\n",
      "    nct = <scipy.stats._continuous_distns.nct_gen object>\n",
      "    ncx2 = <scipy.stats._continuous_distns.ncx2_gen object>\n",
      "    norm = <scipy.stats._continuous_distns.norm_gen object>\n",
      "    pareto = <scipy.stats._continuous_distns.pareto_gen object>\n",
      "    pearson3 = <scipy.stats._continuous_distns.pearson3_gen object>\n",
      "    planck = <scipy.stats._discrete_distns.planck_gen object>\n",
      "    poisson = <scipy.stats._discrete_distns.poisson_gen object>\n",
      "    powerlaw = <scipy.stats._continuous_distns.powerlaw_gen object>\n",
      "    powerlognorm = <scipy.stats._continuous_distns.powerlognorm_gen object...\n",
      "    powernorm = <scipy.stats._continuous_distns.powernorm_gen object>\n",
      "    randint = <scipy.stats._discrete_distns.randint_gen object>\n",
      "    rayleigh = <scipy.stats._continuous_distns.rayleigh_gen object>\n",
      "    rdist = <scipy.stats._continuous_distns.rdist_gen object>\n",
      "    recipinvgauss = <scipy.stats._continuous_distns.recipinvgauss_gen obje...\n",
      "    reciprocal = <scipy.stats._continuous_distns.reciprocal_gen object>\n",
      "    rice = <scipy.stats._continuous_distns.rice_gen object>\n",
      "    semicircular = <scipy.stats._continuous_distns.semicircular_gen object...\n",
      "    skellam = <scipy.stats._discrete_distns.skellam_gen object>\n",
      "    skewnorm = <scipy.stats._continuous_distns.skew_norm_gen object>\n",
      "    t = <scipy.stats._continuous_distns.t_gen object>\n",
      "    trapz = <scipy.stats._continuous_distns.trapz_gen object>\n",
      "    triang = <scipy.stats._continuous_distns.triang_gen object>\n",
      "    truncexpon = <scipy.stats._continuous_distns.truncexpon_gen object>\n",
      "    truncnorm = <scipy.stats._continuous_distns.truncnorm_gen object>\n",
      "    tukeylambda = <scipy.stats._continuous_distns.tukeylambda_gen object>\n",
      "    uniform = <scipy.stats._continuous_distns.uniform_gen object>\n",
      "    vonmises = <scipy.stats._continuous_distns.vonmises_gen object>\n",
      "    vonmises_line = <scipy.stats._continuous_distns.vonmises_gen object>\n",
      "    wald = <scipy.stats._continuous_distns.wald_gen object>\n",
      "    weibull_max = <scipy.stats._continuous_distns.frechet_l_gen object>\n",
      "    weibull_min = <scipy.stats._continuous_distns.frechet_r_gen object>\n",
      "    wrapcauchy = <scipy.stats._continuous_distns.wrapcauchy_gen object>\n",
      "    zipf = <scipy.stats._discrete_distns.zipf_gen object>\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(sp.distributions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package sklearn:\n",
      "\n",
      "NAME\n",
      "    sklearn\n",
      "\n",
      "FILE\n",
      "    d:\\users\\byron\\anaconda2\\lib\\site-packages\\sklearn\\__init__.py\n",
      "\n",
      "DESCRIPTION\n",
      "    Machine learning module for Python\n",
      "    ==================================\n",
      "    \n",
      "    sklearn is a Python module integrating classical machine\n",
      "    learning algorithms in the tightly-knit world of scientific Python\n",
      "    packages (numpy, scipy, matplotlib).\n",
      "    \n",
      "    It aims to provide simple and efficient solutions to learning problems\n",
      "    that are accessible to everybody and reusable in various contexts:\n",
      "    machine-learning as a versatile tool for science and engineering.\n",
      "    \n",
      "    See http://scikit-learn.org for complete documentation.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    __check_build (package)\n",
      "    _build_utils (package)\n",
      "    _isotonic\n",
      "    base\n",
      "    calibration\n",
      "    cluster (package)\n",
      "    covariance (package)\n",
      "    cross_decomposition (package)\n",
      "    cross_validation\n",
      "    datasets (package)\n",
      "    decomposition (package)\n",
      "    discriminant_analysis\n",
      "    dummy\n",
      "    ensemble (package)\n",
      "    exceptions\n",
      "    externals (package)\n",
      "    feature_extraction (package)\n",
      "    feature_selection (package)\n",
      "    gaussian_process (package)\n",
      "    grid_search\n",
      "    isotonic\n",
      "    kernel_approximation\n",
      "    kernel_ridge\n",
      "    lda\n",
      "    learning_curve\n",
      "    linear_model (package)\n",
      "    manifold (package)\n",
      "    metrics (package)\n",
      "    mixture (package)\n",
      "    model_selection (package)\n",
      "    multiclass\n",
      "    multioutput\n",
      "    naive_bayes\n",
      "    neighbors (package)\n",
      "    neural_network (package)\n",
      "    pipeline\n",
      "    preprocessing (package)\n",
      "    qda\n",
      "    random_projection\n",
      "    semi_supervised (package)\n",
      "    setup\n",
      "    svm (package)\n",
      "    tests (package)\n",
      "    tree (package)\n",
      "    utils (package)\n",
      "\n",
      "FUNCTIONS\n",
      "    clone(estimator, safe=True)\n",
      "        Constructs a new estimator with the same parameters.\n",
      "        \n",
      "        Clone does a deep copy of the model in an estimator\n",
      "        without actually copying attached data. It yields a new estimator\n",
      "        with the same parameters that has not been fit on any data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        estimator: estimator object, or list, tuple or set of objects\n",
      "            The estimator or group of estimators to be cloned\n",
      "        \n",
      "        safe: boolean, optional\n",
      "            If safe is false, clone will fall back to a deepcopy on objects\n",
      "            that are not estimators.\n",
      "\n",
      "DATA\n",
      "    __SKLEARN_SETUP__ = False\n",
      "    __all__ = ['calibration', 'cluster', 'covariance', 'cross_decompositio...\n",
      "    __version__ = '0.18.1'\n",
      "\n",
      "VERSION\n",
      "    0.18.1\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sklearn as skl\n",
    "help(skl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package sklearn.datasets in sklearn:\n",
      "\n",
      "NAME\n",
      "    sklearn.datasets\n",
      "\n",
      "FILE\n",
      "    d:\\users\\byron\\anaconda2\\lib\\site-packages\\sklearn\\datasets\\__init__.py\n",
      "\n",
      "DESCRIPTION\n",
      "    The :mod:`sklearn.datasets` module includes utilities to load datasets,\n",
      "    including methods to load and fetch popular reference datasets. It also\n",
      "    features some artificial data generators.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _svmlight_format\n",
      "    base\n",
      "    california_housing\n",
      "    covtype\n",
      "    kddcup99\n",
      "    lfw\n",
      "    mlcomp\n",
      "    mldata\n",
      "    olivetti_faces\n",
      "    rcv1\n",
      "    samples_generator\n",
      "    setup\n",
      "    species_distributions\n",
      "    svmlight_format\n",
      "    tests (package)\n",
      "    twenty_newsgroups\n",
      "\n",
      "FUNCTIONS\n",
      "    clear_data_home(data_home=None)\n",
      "        Delete all the content of the data home cache.\n",
      "    \n",
      "    dump_svmlight_file(X, y, f, zero_based=True, comment=None, query_id=None, multilabel=False)\n",
      "        Dump the dataset in svmlight / libsvm file format.\n",
      "        \n",
      "        This format is a text-based format, with one sample per line. It does\n",
      "        not store zero valued features hence is suitable for sparse dataset.\n",
      "        \n",
      "        The first element of each line can be used to store a target variable\n",
      "        to predict.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X : {array-like, sparse matrix}, shape = [n_samples, n_features]\n",
      "            Training vectors, where n_samples is the number of samples and\n",
      "            n_features is the number of features.\n",
      "        \n",
      "        y : {array-like, sparse matrix}, shape = [n_samples (, n_labels)]\n",
      "            Target values. Class labels must be an\n",
      "            integer or float, or array-like objects of integer or float for\n",
      "            multilabel classifications.\n",
      "        \n",
      "        f : string or file-like in binary mode\n",
      "            If string, specifies the path that will contain the data.\n",
      "            If file-like, data will be written to f. f should be opened in binary\n",
      "            mode.\n",
      "        \n",
      "        zero_based : boolean, optional\n",
      "            Whether column indices should be written zero-based (True) or one-based\n",
      "            (False).\n",
      "        \n",
      "        comment : string, optional\n",
      "            Comment to insert at the top of the file. This should be either a\n",
      "            Unicode string, which will be encoded as UTF-8, or an ASCII byte\n",
      "            string.\n",
      "            If a comment is given, then it will be preceded by one that identifies\n",
      "            the file as having been dumped by scikit-learn. Note that not all\n",
      "            tools grok comments in SVMlight files.\n",
      "        \n",
      "        query_id : array-like, shape = [n_samples]\n",
      "            Array containing pairwise preference constraints (qid in svmlight\n",
      "            format).\n",
      "        \n",
      "        multilabel : boolean, optional\n",
      "            Samples may have several labels each (see\n",
      "            http://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/multilabel.html)\n",
      "        \n",
      "            .. versionadded:: 0.17\n",
      "               parameter *multilabel* to support multilabel datasets.\n",
      "    \n",
      "    fetch_20newsgroups(data_home=None, subset='train', categories=None, shuffle=True, random_state=42, remove=(), download_if_missing=True)\n",
      "        Load the filenames and data from the 20 newsgroups dataset.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <20newsgroups>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        subset : 'train' or 'test', 'all', optional\n",
      "            Select the dataset to load: 'train' for the training set, 'test'\n",
      "            for the test set, 'all' for both, with shuffled ordering.\n",
      "        \n",
      "        data_home : optional, default: None\n",
      "            Specify a download and cache folder for the datasets. If None,\n",
      "            all scikit-learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        categories : None or collection of string or unicode\n",
      "            If None (default), load all the categories.\n",
      "            If not None, list of category names to load (other categories\n",
      "            ignored).\n",
      "        \n",
      "        shuffle : bool, optional\n",
      "            Whether or not to shuffle the data: might be important for models that\n",
      "            make the assumption that the samples are independent and identically\n",
      "            distributed (i.i.d.), such as stochastic gradient descent.\n",
      "        \n",
      "        random_state : numpy random number generator or seed integer\n",
      "            Used to shuffle the dataset.\n",
      "        \n",
      "        download_if_missing : optional, True by default\n",
      "            If False, raise an IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        remove : tuple\n",
      "            May contain any subset of ('headers', 'footers', 'quotes'). Each of\n",
      "            these are kinds of text that will be detected and removed from the\n",
      "            newsgroup posts, preventing classifiers from overfitting on\n",
      "            metadata.\n",
      "        \n",
      "            'headers' removes newsgroup headers, 'footers' removes blocks at the\n",
      "            ends of posts that look like signatures, and 'quotes' removes lines\n",
      "            that appear to be quoting another post.\n",
      "        \n",
      "            'headers' follows an exact standard; the other filters are not always\n",
      "            correct.\n",
      "    \n",
      "    fetch_20newsgroups_vectorized(subset='train', remove=(), data_home=None)\n",
      "        Load the 20 newsgroups dataset and transform it into tf-idf vectors.\n",
      "        \n",
      "        This is a convenience function; the tf-idf transformation is done using the\n",
      "        default settings for `sklearn.feature_extraction.text.Vectorizer`. For more\n",
      "        advanced usage (stopword filtering, n-gram extraction, etc.), combine\n",
      "        fetch_20newsgroups with a custom `Vectorizer` or `CountVectorizer`.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <20newsgroups>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        \n",
      "        subset : 'train' or 'test', 'all', optional\n",
      "            Select the dataset to load: 'train' for the training set, 'test'\n",
      "            for the test set, 'all' for both, with shuffled ordering.\n",
      "        \n",
      "        data_home : optional, default: None\n",
      "            Specify an download and cache folder for the datasets. If None,\n",
      "            all scikit-learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        remove : tuple\n",
      "            May contain any subset of ('headers', 'footers', 'quotes'). Each of\n",
      "            these are kinds of text that will be detected and removed from the\n",
      "            newsgroup posts, preventing classifiers from overfitting on\n",
      "            metadata.\n",
      "        \n",
      "            'headers' removes newsgroup headers, 'footers' removes blocks at the\n",
      "            ends of posts that look like signatures, and 'quotes' removes lines\n",
      "            that appear to be quoting another post.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        \n",
      "        bunch : Bunch object\n",
      "            bunch.data: sparse matrix, shape [n_samples, n_features]\n",
      "            bunch.target: array, shape [n_samples]\n",
      "            bunch.target_names: list, length [n_classes]\n",
      "    \n",
      "    fetch_california_housing(data_home=None, download_if_missing=True)\n",
      "        Loader for the California housing dataset from StatLib.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data_home : optional, default: None\n",
      "            Specify another download and cache folder for the datasets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        download_if_missing: optional, True by default\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        dataset : dict-like object with the following attributes:\n",
      "        \n",
      "        dataset.data : ndarray, shape [20640, 8]\n",
      "            Each row corresponding to the 8 feature values in order.\n",
      "        \n",
      "        dataset.target : numpy array of shape (20640,)\n",
      "            Each value corresponds to the average house value in units of 100,000.\n",
      "        \n",
      "        dataset.feature_names : array of length 8\n",
      "            Array of ordered feature names used in the dataset.\n",
      "        \n",
      "        dataset.DESCR : string\n",
      "            Description of the California housing dataset.\n",
      "        \n",
      "        Notes\n",
      "        ------\n",
      "        \n",
      "        This dataset consists of 20,640 samples and 9 features.\n",
      "    \n",
      "    fetch_covtype(data_home=None, download_if_missing=True, random_state=None, shuffle=False)\n",
      "        Load the covertype dataset, downloading it if necessary.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data_home : string, optional\n",
      "            Specify another download and cache folder for the datasets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        download_if_missing : boolean, default=True\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            Random state for shuffling the dataset.\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        shuffle : bool, default=False\n",
      "            Whether to shuffle dataset.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        dataset : dict-like object with the following attributes:\n",
      "        \n",
      "        dataset.data : numpy array of shape (581012, 54)\n",
      "            Each row corresponds to the 54 features in the dataset.\n",
      "        \n",
      "        dataset.target : numpy array of shape (581012,)\n",
      "            Each value corresponds to one of the 7 forest covertypes with values\n",
      "            ranging between 1 to 7.\n",
      "        \n",
      "        dataset.DESCR : string\n",
      "            Description of the forest covertype dataset.\n",
      "    \n",
      "    fetch_kddcup99(subset=None, shuffle=False, random_state=None, percent10=True, download_if_missing=True)\n",
      "        Load and return the kddcup 99 dataset (classification).\n",
      "        \n",
      "        The KDD Cup '99 dataset was created by processing the tcpdump portions\n",
      "        of the 1998 DARPA Intrusion Detection System (IDS) Evaluation dataset,\n",
      "        created by MIT Lincoln Lab [1] . The artificial data was generated using\n",
      "        a closed network and hand-injected attacks to produce a large number of\n",
      "        different types of attack with normal activity in the background.\n",
      "        As the initial goal was to produce a large training set for supervised\n",
      "        learning algorithms, there is a large proportion (80.1%) of abnormal\n",
      "        data which is unrealistic in real world, and inappropriate for unsupervised\n",
      "        anomaly detection which aims at detecting 'abnormal' data, ie\n",
      "        \n",
      "        1) qualitatively different from normal data.\n",
      "        \n",
      "        2) in large minority among the observations.\n",
      "        \n",
      "        We thus transform the KDD Data set into two different data sets: SA and SF.\n",
      "        \n",
      "        - SA is obtained by simply selecting all the normal data, and a small\n",
      "          proportion of abnormal data to gives an anomaly proportion of 1%.\n",
      "        \n",
      "        - SF is obtained as in [2]\n",
      "          by simply picking up the data whose attribute logged_in is positive, thus\n",
      "          focusing on the intrusion attack, which gives a proportion of 0.3% of\n",
      "          attack.\n",
      "        \n",
      "        - http and smtp are two subsets of SF corresponding with third feature\n",
      "          equal to 'http' (resp. to 'smtp')\n",
      "        \n",
      "        \n",
      "        General KDD structure :\n",
      "        \n",
      "        ================      ==========================================\n",
      "        Samples total         4898431\n",
      "        Dimensionality        41\n",
      "        Features              discrete (int) or continuous (float)\n",
      "        Targets               str, 'normal.' or name of the anomaly type\n",
      "        ================      ==========================================\n",
      "        \n",
      "        SA structure :\n",
      "        \n",
      "        ================      ==========================================\n",
      "        Samples total         976158\n",
      "        Dimensionality        41\n",
      "        Features              discrete (int) or continuous (float)\n",
      "        Targets               str, 'normal.' or name of the anomaly type\n",
      "        ================      ==========================================\n",
      "        \n",
      "        SF structure :\n",
      "        \n",
      "        ================      ==========================================\n",
      "        Samples total         699691\n",
      "        Dimensionality        4\n",
      "        Features              discrete (int) or continuous (float)\n",
      "        Targets               str, 'normal.' or name of the anomaly type\n",
      "        ================      ==========================================\n",
      "        \n",
      "        http structure :\n",
      "        \n",
      "        ================      ==========================================\n",
      "        Samples total         619052\n",
      "        Dimensionality        3\n",
      "        Features              discrete (int) or continuous (float)\n",
      "        Targets               str, 'normal.' or name of the anomaly type\n",
      "        ================      ==========================================\n",
      "        \n",
      "        smtp structure :\n",
      "        \n",
      "        ================      ==========================================\n",
      "        Samples total         95373\n",
      "        Dimensionality        3\n",
      "        Features              discrete (int) or continuous (float)\n",
      "        Targets               str, 'normal.' or name of the anomaly type\n",
      "        ================      ==========================================\n",
      "        \n",
      "        .. versionadded:: 0.18\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        subset : None, 'SA', 'SF', 'http', 'smtp'\n",
      "            To return the corresponding classical subsets of kddcup 99.\n",
      "            If None, return the entire kddcup 99 dataset.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            Random state for shuffling the dataset.\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        shuffle : bool, default=False\n",
      "            Whether to shuffle dataset.\n",
      "        \n",
      "        percent10 : bool, default=False\n",
      "            Whether to load only 10 percent of the data.\n",
      "        \n",
      "        download_if_missing : bool, default=True\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn and 'target', the regression target for each\n",
      "            sample.\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Analysis and Results of the 1999 DARPA Off-Line Intrusion\n",
      "               Detection Evaluation Richard Lippmann, Joshua W. Haines,\n",
      "               David J. Fried, Jonathan Korba, Kumar Das\n",
      "        \n",
      "        .. [2] A Geometric Framework for Unsupervised Anomaly Detection: Detecting\n",
      "               Intrusions in Unlabeled Data (2002) by Eleazar Eskin, Andrew Arnold,\n",
      "               Michael Prerau, Leonid Portnoy, Sal Stolfo\n",
      "    \n",
      "    fetch_lfw_pairs(subset='train', data_home=None, funneled=True, resize=0.5, color=False, slice_=(slice(70, 195, None), slice(78, 172, None)), download_if_missing=True)\n",
      "        Loader for the Labeled Faces in the Wild (LFW) pairs dataset\n",
      "        \n",
      "        This dataset is a collection of JPEG pictures of famous people\n",
      "        collected on the internet, all details are available on the\n",
      "        official website:\n",
      "        \n",
      "            http://vis-www.cs.umass.edu/lfw/\n",
      "        \n",
      "        Each picture is centered on a single face. Each pixel of each channel\n",
      "        (color in RGB) is encoded by a float in range 0.0 - 1.0.\n",
      "        \n",
      "        The task is called Face Verification: given a pair of two pictures,\n",
      "        a binary classifier must predict whether the two images are from\n",
      "        the same person.\n",
      "        \n",
      "        In the official `README.txt`_ this task is described as the\n",
      "        \"Restricted\" task.  As I am not sure as to implement the\n",
      "        \"Unrestricted\" variant correctly, I left it as unsupported for now.\n",
      "        \n",
      "          .. _`README.txt`: http://vis-www.cs.umass.edu/lfw/README.txt\n",
      "        \n",
      "        The original images are 250 x 250 pixels, but the default slice and resize\n",
      "        arguments reduce them to 62 x 74.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <labeled_faces_in_the_wild>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        subset : optional, default: 'train'\n",
      "            Select the dataset to load: 'train' for the development training\n",
      "            set, 'test' for the development test set, and '10_folds' for the\n",
      "            official evaluation set that is meant to be used with a 10-folds\n",
      "            cross validation.\n",
      "        \n",
      "        data_home : optional, default: None\n",
      "            Specify another download and cache folder for the datasets. By\n",
      "            default all scikit learn data is stored in '~/scikit_learn_data'\n",
      "            subfolders.\n",
      "        \n",
      "        funneled : boolean, optional, default: True\n",
      "            Download and use the funneled variant of the dataset.\n",
      "        \n",
      "        resize : float, optional, default 0.5\n",
      "            Ratio used to resize the each face picture.\n",
      "        \n",
      "        color : boolean, optional, default False\n",
      "            Keep the 3 RGB channels instead of averaging them to a single\n",
      "            gray level channel. If color is True the shape of the data has\n",
      "            one more dimension than the shape with color = False.\n",
      "        \n",
      "        slice_ : optional\n",
      "            Provide a custom 2D slice (height, width) to extract the\n",
      "            'interesting' part of the jpeg files and avoid use statistical\n",
      "            correlation from the background\n",
      "        \n",
      "        download_if_missing : optional, True by default\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        The data is returned as a Bunch object with the following attributes:\n",
      "        \n",
      "        data : numpy array of shape (2200, 5828). Shape depends on ``subset``.\n",
      "            Each row corresponds to 2 ravel'd face images of original size 62 x 47\n",
      "            pixels. Changing the ``slice_``, ``resize`` or ``subset`` parameters\n",
      "            will change the shape of the output.\n",
      "        \n",
      "        pairs : numpy array of shape (2200, 2, 62, 47). Shape depends on\n",
      "                ``subset``.\n",
      "            Each row has 2 face images corresponding to same or different person\n",
      "            from the dataset containing 5749 people. Changing the ``slice_``,\n",
      "            ``resize`` or ``subset`` parameters will change the shape of the\n",
      "            output.\n",
      "        \n",
      "        target : numpy array of shape (2200,). Shape depends on ``subset``.\n",
      "            Labels associated to each pair of images. The two label values being\n",
      "            different persons or the same person.\n",
      "        \n",
      "        DESCR : string\n",
      "            Description of the Labeled Faces in the Wild (LFW) dataset.\n",
      "    \n",
      "    fetch_lfw_people(data_home=None, funneled=True, resize=0.5, min_faces_per_person=0, color=False, slice_=(slice(70, 195, None), slice(78, 172, None)), download_if_missing=True)\n",
      "        Loader for the Labeled Faces in the Wild (LFW) people dataset\n",
      "        \n",
      "        This dataset is a collection of JPEG pictures of famous people\n",
      "        collected on the internet, all details are available on the\n",
      "        official website:\n",
      "        \n",
      "            http://vis-www.cs.umass.edu/lfw/\n",
      "        \n",
      "        Each picture is centered on a single face. Each pixel of each channel\n",
      "        (color in RGB) is encoded by a float in range 0.0 - 1.0.\n",
      "        \n",
      "        The task is called Face Recognition (or Identification): given the\n",
      "        picture of a face, find the name of the person given a training set\n",
      "        (gallery).\n",
      "        \n",
      "        The original images are 250 x 250 pixels, but the default slice and resize\n",
      "        arguments reduce them to 62 x 74.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data_home : optional, default: None\n",
      "            Specify another download and cache folder for the datasets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        funneled : boolean, optional, default: True\n",
      "            Download and use the funneled variant of the dataset.\n",
      "        \n",
      "        resize : float, optional, default 0.5\n",
      "            Ratio used to resize the each face picture.\n",
      "        \n",
      "        min_faces_per_person : int, optional, default None\n",
      "            The extracted dataset will only retain pictures of people that have at\n",
      "            least `min_faces_per_person` different pictures.\n",
      "        \n",
      "        color : boolean, optional, default False\n",
      "            Keep the 3 RGB channels instead of averaging them to a single\n",
      "            gray level channel. If color is True the shape of the data has\n",
      "            one more dimension than the shape with color = False.\n",
      "        \n",
      "        slice_ : optional\n",
      "            Provide a custom 2D slice (height, width) to extract the\n",
      "            'interesting' part of the jpeg files and avoid use statistical\n",
      "            correlation from the background\n",
      "        \n",
      "        download_if_missing : optional, True by default\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        dataset : dict-like object with the following attributes:\n",
      "        \n",
      "        dataset.data : numpy array of shape (13233, 2914)\n",
      "            Each row corresponds to a ravelled face image of original size 62 x 47\n",
      "            pixels. Changing the ``slice_`` or resize parameters will change the\n",
      "            shape of the output.\n",
      "        \n",
      "        dataset.images : numpy array of shape (13233, 62, 47)\n",
      "            Each row is a face image corresponding to one of the 5749 people in\n",
      "            the dataset. Changing the ``slice_`` or resize parameters will change\n",
      "            the shape of the output.\n",
      "        \n",
      "        dataset.target : numpy array of shape (13233,)\n",
      "            Labels associated to each face image. Those labels range from 0-5748\n",
      "            and correspond to the person IDs.\n",
      "        \n",
      "        dataset.DESCR : string\n",
      "            Description of the Labeled Faces in the Wild (LFW) dataset.\n",
      "    \n",
      "    fetch_mldata(dataname, target_name='label', data_name='data', transpose_data=True, data_home=None)\n",
      "        Fetch an mldata.org data set\n",
      "        \n",
      "        If the file does not exist yet, it is downloaded from mldata.org .\n",
      "        \n",
      "        mldata.org does not have an enforced convention for storing data or\n",
      "        naming the columns in a data set. The default behavior of this function\n",
      "        works well with the most common cases:\n",
      "        \n",
      "          1) data values are stored in the column 'data', and target values in the\n",
      "             column 'label'\n",
      "          2) alternatively, the first column stores target values, and the second\n",
      "             data values\n",
      "          3) the data array is stored as `n_features x n_samples` , and thus needs\n",
      "             to be transposed to match the `sklearn` standard\n",
      "        \n",
      "        Keyword arguments allow to adapt these defaults to specific data sets\n",
      "        (see parameters `target_name`, `data_name`, `transpose_data`, and\n",
      "        the examples below).\n",
      "        \n",
      "        mldata.org data sets may have multiple columns, which are stored in the\n",
      "        Bunch object with their original name.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        \n",
      "        dataname :\n",
      "            Name of the data set on mldata.org,\n",
      "            e.g.: \"leukemia\", \"Whistler Daily Snowfall\", etc.\n",
      "            The raw name is automatically converted to a mldata.org URL .\n",
      "        \n",
      "        target_name : optional, default: 'label'\n",
      "            Name or index of the column containing the target values.\n",
      "        \n",
      "        data_name : optional, default: 'data'\n",
      "            Name or index of the column containing the data.\n",
      "        \n",
      "        transpose_data : optional, default: True\n",
      "            If True, transpose the downloaded data array.\n",
      "        \n",
      "        data_home : optional, default: None\n",
      "            Specify another download and cache folder for the data sets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        \n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn, 'target', the classification labels,\n",
      "            'DESCR', the full description of the dataset, and\n",
      "            'COL_NAMES', the original names of the dataset columns.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Load the 'iris' dataset from mldata.org:\n",
      "        \n",
      "        >>> from sklearn.datasets.mldata import fetch_mldata\n",
      "        >>> import tempfile\n",
      "        >>> test_data_home = tempfile.mkdtemp()\n",
      "        \n",
      "        >>> iris = fetch_mldata('iris', data_home=test_data_home)\n",
      "        >>> iris.target.shape\n",
      "        (150,)\n",
      "        >>> iris.data.shape\n",
      "        (150, 4)\n",
      "        \n",
      "        Load the 'leukemia' dataset from mldata.org, which needs to be transposed\n",
      "        to respects the scikit-learn axes convention:\n",
      "        \n",
      "        >>> leuk = fetch_mldata('leukemia', transpose_data=True,\n",
      "        ...                     data_home=test_data_home)\n",
      "        >>> leuk.data.shape\n",
      "        (72, 7129)\n",
      "        \n",
      "        Load an alternative 'iris' dataset, which has different names for the\n",
      "        columns:\n",
      "        \n",
      "        >>> iris2 = fetch_mldata('datasets-UCI iris', target_name=1,\n",
      "        ...                      data_name=0, data_home=test_data_home)\n",
      "        >>> iris3 = fetch_mldata('datasets-UCI iris',\n",
      "        ...                      target_name='class', data_name='double0',\n",
      "        ...                      data_home=test_data_home)\n",
      "        \n",
      "        >>> import shutil\n",
      "        >>> shutil.rmtree(test_data_home)\n",
      "    \n",
      "    fetch_olivetti_faces(data_home=None, shuffle=False, random_state=0, download_if_missing=True)\n",
      "        Loader for the Olivetti faces data-set from AT&T.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <olivetti_faces>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data_home : optional, default: None\n",
      "            Specify another download and cache folder for the datasets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        shuffle : boolean, optional\n",
      "            If True the order of the dataset is shuffled to avoid having\n",
      "            images of the same person grouped.\n",
      "        \n",
      "        download_if_missing : optional, True by default\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        random_state : optional, integer or RandomState object\n",
      "            The seed or the random number generator used to shuffle the\n",
      "            data.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        An object with the following attributes:\n",
      "        \n",
      "        data : numpy array of shape (400, 4096)\n",
      "            Each row corresponds to a ravelled face image of original size 64 x 64 pixels.\n",
      "        \n",
      "        images : numpy array of shape (400, 64, 64)\n",
      "            Each row is a face image corresponding to one of the 40 subjects of the dataset.\n",
      "        \n",
      "        target : numpy array of shape (400, )\n",
      "            Labels associated to each face image. Those labels are ranging from\n",
      "            0-39 and correspond to the Subject IDs.\n",
      "        \n",
      "        DESCR : string\n",
      "            Description of the modified Olivetti Faces Dataset.\n",
      "        \n",
      "        Notes\n",
      "        ------\n",
      "        \n",
      "        This dataset consists of 10 pictures each of 40 individuals. The original\n",
      "        database was available from (now defunct)\n",
      "        \n",
      "            http://www.cl.cam.ac.uk/research/dtg/attarchive/facedatabase.html\n",
      "        \n",
      "        The version retrieved here comes in MATLAB format from the personal\n",
      "        web page of Sam Roweis:\n",
      "        \n",
      "            http://www.cs.nyu.edu/~roweis/\n",
      "    \n",
      "    fetch_rcv1(data_home=None, subset='all', download_if_missing=True, random_state=None, shuffle=False)\n",
      "        Load the RCV1 multilabel dataset, downloading it if necessary.\n",
      "        \n",
      "        Version: RCV1-v2, vectors, full sets, topics multilabels.\n",
      "        \n",
      "        ==============     =====================\n",
      "        Classes                              103\n",
      "        Samples total                     804414\n",
      "        Dimensionality                     47236\n",
      "        Features           real, between 0 and 1\n",
      "        ==============     =====================\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        .. versionadded:: 0.17\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data_home : string, optional\n",
      "            Specify another download and cache folder for the datasets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        subset: string, 'train', 'test', or 'all', default='all'\n",
      "            Select the dataset to load: 'train' for the training set\n",
      "            (23149 samples), 'test' for the test set (781265 samples),\n",
      "            'all' for both, with the training samples first if shuffle is False.\n",
      "            This follows the official LYRL2004 chronological split.\n",
      "        \n",
      "        download_if_missing : boolean, default=True\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            Random state for shuffling the dataset.\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        shuffle : bool, default=False\n",
      "            Whether to shuffle dataset.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        dataset : dict-like object with the following attributes:\n",
      "        \n",
      "        dataset.data : scipy csr array, dtype np.float64, shape (804414, 47236)\n",
      "            The array has 0.16% of non zero values.\n",
      "        \n",
      "        dataset.target : scipy csr array, dtype np.uint8, shape (804414, 103)\n",
      "            Each sample has a value of 1 in its categories, and 0 in others.\n",
      "            The array has 3.15% of non zero values.\n",
      "        \n",
      "        dataset.sample_id : numpy array, dtype np.uint32, shape (804414,)\n",
      "            Identification number of each sample, as ordered in dataset.data.\n",
      "        \n",
      "        dataset.target_names : numpy array, dtype object, length (103)\n",
      "            Names of each target (RCV1 topics), as ordered in dataset.target.\n",
      "        \n",
      "        dataset.DESCR : string\n",
      "            Description of the RCV1 dataset.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Lewis, D. D., Yang, Y., Rose, T. G., & Li, F. (2004). RCV1: A new\n",
      "        benchmark collection for text categorization research. The Journal of\n",
      "        Machine Learning Research, 5, 361-397.\n",
      "    \n",
      "    fetch_species_distributions(data_home=None, download_if_missing=True)\n",
      "        Loader for species distribution dataset from Phillips et. al. (2006)\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data_home : optional, default: None\n",
      "            Specify another download and cache folder for the datasets. By default\n",
      "            all scikit learn data is stored in '~/scikit_learn_data' subfolders.\n",
      "        \n",
      "        download_if_missing : optional, True by default\n",
      "            If False, raise a IOError if the data is not locally available\n",
      "            instead of trying to download the data from the source site.\n",
      "        \n",
      "        Returns\n",
      "        --------\n",
      "        The data is returned as a Bunch object with the following attributes:\n",
      "        \n",
      "        coverages : array, shape = [14, 1592, 1212]\n",
      "            These represent the 14 features measured at each point of the map grid.\n",
      "            The latitude/longitude values for the grid are discussed below.\n",
      "            Missing data is represented by the value -9999.\n",
      "        \n",
      "        train : record array, shape = (1623,)\n",
      "            The training points for the data.  Each point has three fields:\n",
      "        \n",
      "            - train['species'] is the species name\n",
      "            - train['dd long'] is the longitude, in degrees\n",
      "            - train['dd lat'] is the latitude, in degrees\n",
      "        \n",
      "        test : record array, shape = (619,)\n",
      "            The test points for the data.  Same format as the training data.\n",
      "        \n",
      "        Nx, Ny : integers\n",
      "            The number of longitudes (x) and latitudes (y) in the grid\n",
      "        \n",
      "        x_left_lower_corner, y_left_lower_corner : floats\n",
      "            The (x,y) position of the lower-left corner, in degrees\n",
      "        \n",
      "        grid_size : float\n",
      "            The spacing between points of the grid, in degrees\n",
      "        \n",
      "        Notes\n",
      "        ------\n",
      "        \n",
      "        This dataset represents the geographic distribution of species.\n",
      "        The dataset is provided by Phillips et. al. (2006).\n",
      "        \n",
      "        The two species are:\n",
      "        \n",
      "        - `\"Bradypus variegatus\"\n",
      "          <http://www.iucnredlist.org/details/3038/0>`_ ,\n",
      "          the Brown-throated Sloth.\n",
      "        \n",
      "        - `\"Microryzomys minutus\"\n",
      "          <http://www.iucnredlist.org/details/13408/0>`_ ,\n",
      "          also known as the Forest Small Rice Rat, a rodent that lives in Peru,\n",
      "          Colombia, Ecuador, Peru, and Venezuela.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        \n",
      "        * `\"Maximum entropy modeling of species geographic distributions\"\n",
      "          <http://www.cs.princeton.edu/~schapire/papers/ecolmod.pdf>`_\n",
      "          S. J. Phillips, R. P. Anderson, R. E. Schapire - Ecological Modelling,\n",
      "          190:231-259, 2006.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        * See examples/applications/plot_species_distribution_modeling.py\n",
      "          for an example of using this dataset with scikit-learn\n",
      "    \n",
      "    get_data_home(data_home=None)\n",
      "        Return the path of the scikit-learn data dir.\n",
      "        \n",
      "        This folder is used by some large dataset loaders to avoid\n",
      "        downloading the data several times.\n",
      "        \n",
      "        By default the data dir is set to a folder named 'scikit_learn_data'\n",
      "        in the user home folder.\n",
      "        \n",
      "        Alternatively, it can be set by the 'SCIKIT_LEARN_DATA' environment\n",
      "        variable or programmatically by giving an explicit folder path. The\n",
      "        '~' symbol is expanded to the user home folder.\n",
      "        \n",
      "        If the folder does not already exist, it is automatically created.\n",
      "    \n",
      "    load_boston(return_X_y=False)\n",
      "        Load and return the boston house-prices dataset (regression).\n",
      "        \n",
      "        ==============     ==============\n",
      "        Samples total                 506\n",
      "        Dimensionality                 13\n",
      "        Features           real, positive\n",
      "        Targets             real 5. - 50.\n",
      "        ==============     ==============\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        return_X_y : boolean, default=False.\n",
      "            If True, returns ``(data, target)`` instead of a Bunch object.\n",
      "            See below for more information about the `data` and `target` object.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn, 'target', the regression targets,\n",
      "            and 'DESCR', the full description of the dataset.\n",
      "        \n",
      "        (data, target) : tuple if ``return_X_y`` is True\n",
      "        \n",
      "            .. versionadded:: 0.18    \n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from sklearn.datasets import load_boston\n",
      "        >>> boston = load_boston()\n",
      "        >>> print(boston.data.shape)\n",
      "        (506, 13)\n",
      "    \n",
      "    load_breast_cancer(return_X_y=False)\n",
      "        Load and return the breast cancer wisconsin dataset (classification).\n",
      "        \n",
      "        The breast cancer dataset is a classic and very easy binary classification\n",
      "        dataset.\n",
      "        \n",
      "        =================   ==============\n",
      "        Classes                          2\n",
      "        Samples per class    212(M),357(B)\n",
      "        Samples total                  569\n",
      "        Dimensionality                  30\n",
      "        Features            real, positive\n",
      "        =================   ==============\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        return_X_y : boolean, default=False\n",
      "            If True, returns ``(data, target)`` instead of a Bunch object.\n",
      "            See below for more information about the `data` and `target` object.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn, 'target', the classification labels,\n",
      "            'target_names', the meaning of the labels, 'feature_names', the\n",
      "            meaning of the features, and 'DESCR', the\n",
      "            full description of the dataset.\n",
      "        \n",
      "        (data, target) : tuple if ``return_X_y`` is True\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        The copy of UCI ML Breast Cancer Wisconsin (Diagnostic) dataset is\n",
      "        downloaded from:\n",
      "        https://goo.gl/U2Uwz2\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Let's say you are interested in the samples 10, 50, and 85, and want to\n",
      "        know their class name.\n",
      "        \n",
      "        >>> from sklearn.datasets import load_breast_cancer\n",
      "        >>> data = load_breast_cancer()\n",
      "        >>> data.target[[10, 50, 85]]\n",
      "        array([0, 1, 0])\n",
      "        >>> list(data.target_names)\n",
      "        ['malignant', 'benign']\n",
      "    \n",
      "    load_diabetes(return_X_y=False)\n",
      "        Load and return the diabetes dataset (regression).\n",
      "        \n",
      "        ==============      ==================\n",
      "        Samples total       442\n",
      "        Dimensionality      10\n",
      "        Features            real, -.2 < x < .2\n",
      "        Targets             integer 25 - 346\n",
      "        ==============      ==================\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        return_X_y : boolean, default=False.\n",
      "            If True, returns ``(data, target)`` instead of a Bunch object.\n",
      "            See below for more information about the `data` and `target` object.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn and 'target', the regression target for each\n",
      "            sample.\n",
      "        \n",
      "        (data, target) : tuple if ``return_X_y`` is True\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "    \n",
      "    load_digits(n_class=10, return_X_y=False)\n",
      "        Load and return the digits dataset (classification).\n",
      "        \n",
      "        Each datapoint is a 8x8 image of a digit.\n",
      "        \n",
      "        =================   ==============\n",
      "        Classes                         10\n",
      "        Samples per class             ~180\n",
      "        Samples total                 1797\n",
      "        Dimensionality                  64\n",
      "        Features             integers 0-16\n",
      "        =================   ==============\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_class : integer, between 0 and 10, optional (default=10)\n",
      "            The number of classes to return.\n",
      "        \n",
      "        return_X_y : boolean, default=False.\n",
      "            If True, returns ``(data, target)`` instead of a Bunch object.\n",
      "            See below for more information about the `data` and `target` object.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn, 'images', the images corresponding\n",
      "            to each sample, 'target', the classification labels for each\n",
      "            sample, 'target_names', the meaning of the labels, and 'DESCR',\n",
      "            the full description of the dataset.\n",
      "        \n",
      "        (data, target) : tuple if ``return_X_y`` is True\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        To load the data and visualize the images::\n",
      "        \n",
      "            >>> from sklearn.datasets import load_digits\n",
      "            >>> digits = load_digits()\n",
      "            >>> print(digits.data.shape)\n",
      "            (1797, 64)\n",
      "            >>> import matplotlib.pyplot as plt #doctest: +SKIP\n",
      "            >>> plt.gray() #doctest: +SKIP\n",
      "            >>> plt.matshow(digits.images[0]) #doctest: +SKIP\n",
      "            >>> plt.show() #doctest: +SKIP\n",
      "    \n",
      "    load_files(container_path, description=None, categories=None, load_content=True, shuffle=True, encoding=None, decode_error='strict', random_state=0)\n",
      "        Load text files with categories as subfolder names.\n",
      "        \n",
      "        Individual samples are assumed to be files stored a two levels folder\n",
      "        structure such as the following:\n",
      "        \n",
      "            container_folder/\n",
      "                category_1_folder/\n",
      "                    file_1.txt\n",
      "                    file_2.txt\n",
      "                    ...\n",
      "                    file_42.txt\n",
      "                category_2_folder/\n",
      "                    file_43.txt\n",
      "                    file_44.txt\n",
      "                    ...\n",
      "        \n",
      "        The folder names are used as supervised signal label names. The\n",
      "        individual file names are not important.\n",
      "        \n",
      "        This function does not try to extract features into a numpy array or\n",
      "        scipy sparse matrix. In addition, if load_content is false it\n",
      "        does not try to load the files in memory.\n",
      "        \n",
      "        To use text files in a scikit-learn classification or clustering\n",
      "        algorithm, you will need to use the `sklearn.feature_extraction.text`\n",
      "        module to build a feature extraction transformer that suits your\n",
      "        problem.\n",
      "        \n",
      "        If you set load_content=True, you should also specify the encoding of\n",
      "        the text using the 'encoding' parameter. For many modern text files,\n",
      "        'utf-8' will be the correct encoding. If you leave encoding equal to None,\n",
      "        then the content will be made of bytes instead of Unicode, and you will\n",
      "        not be able to use most functions in `sklearn.feature_extraction.text`.\n",
      "        \n",
      "        Similar feature extractors should be built for other kind of unstructured\n",
      "        data input such as images, audio, video, ...\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        container_path : string or unicode\n",
      "            Path to the main folder holding one subfolder per category\n",
      "        \n",
      "        description: string or unicode, optional (default=None)\n",
      "            A paragraph describing the characteristic of the dataset: its source,\n",
      "            reference, etc.\n",
      "        \n",
      "        categories : A collection of strings or None, optional (default=None)\n",
      "            If None (default), load all the categories.\n",
      "            If not None, list of category names to load (other categories ignored).\n",
      "        \n",
      "        load_content : boolean, optional (default=True)\n",
      "            Whether to load or not the content of the different files. If\n",
      "            true a 'data' attribute containing the text information is present\n",
      "            in the data structure returned. If not, a filenames attribute\n",
      "            gives the path to the files.\n",
      "        \n",
      "        encoding : string or None (default is None)\n",
      "            If None, do not try to decode the content of the files (e.g. for\n",
      "            images or other non-text content).\n",
      "            If not None, encoding to use to decode text files to Unicode if\n",
      "            load_content is True.\n",
      "        \n",
      "        decode_error: {'strict', 'ignore', 'replace'}, optional\n",
      "            Instruction on what to do if a byte sequence is given to analyze that\n",
      "            contains characters not of the given `encoding`. Passed as keyword\n",
      "            argument 'errors' to bytes.decode.\n",
      "        \n",
      "        shuffle : bool, optional (default=True)\n",
      "            Whether or not to shuffle the data: might be important for models that\n",
      "            make the assumption that the samples are independent and identically\n",
      "            distributed (i.i.d.), such as stochastic gradient descent.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=0)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are: either\n",
      "            data, the raw text data to learn, or 'filenames', the files\n",
      "            holding it, 'target', the classification labels (integer index),\n",
      "            'target_names', the meaning of the labels, and 'DESCR', the full\n",
      "            description of the dataset.\n",
      "    \n",
      "    load_iris(return_X_y=False)\n",
      "        Load and return the iris dataset (classification).\n",
      "        \n",
      "        The iris dataset is a classic and very easy multi-class classification\n",
      "        dataset.\n",
      "        \n",
      "        =================   ==============\n",
      "        Classes                          3\n",
      "        Samples per class               50\n",
      "        Samples total                  150\n",
      "        Dimensionality                   4\n",
      "        Features            real, positive\n",
      "        =================   ==============\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        return_X_y : boolean, default=False.\n",
      "            If True, returns ``(data, target)`` instead of a Bunch object.\n",
      "            See below for more information about the `data` and `target` object.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'data', the data to learn, 'target', the classification labels,\n",
      "            'target_names', the meaning of the labels, 'feature_names', the\n",
      "            meaning of the features, and 'DESCR', the\n",
      "            full description of the dataset.\n",
      "        \n",
      "        (data, target) : tuple if ``return_X_y`` is True\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Let's say you are interested in the samples 10, 25, and 50, and want to\n",
      "        know their class name.\n",
      "        \n",
      "        >>> from sklearn.datasets import load_iris\n",
      "        >>> data = load_iris()\n",
      "        >>> data.target[[10, 25, 50]]\n",
      "        array([0, 0, 1])\n",
      "        >>> list(data.target_names)\n",
      "        ['setosa', 'versicolor', 'virginica']\n",
      "    \n",
      "    load_lfw_pairs(*args, **kwargs)\n",
      "        DEPRECATED: Function 'load_lfw_pairs' has been deprecated in 0.17 and will be removed in 0.19.Use fetch_lfw_pairs(download_if_missing=False) instead.\n",
      "        \n",
      "        \n",
      "        Alias for fetch_lfw_pairs(download_if_missing=False)\n",
      "        \n",
      "        .. deprecated:: 0.17\n",
      "            This function will be removed in 0.19.\n",
      "            Use :func:`sklearn.datasets.fetch_lfw_pairs` with parameter\n",
      "            ``download_if_missing=False`` instead.\n",
      "        \n",
      "        Check fetch_lfw_pairs.__doc__ for the documentation and parameter list.\n",
      "    \n",
      "    load_lfw_people(*args, **kwargs)\n",
      "        DEPRECATED: Function 'load_lfw_people' has been deprecated in 0.17 and will be removed in 0.19.Use fetch_lfw_people(download_if_missing=False) instead.\n",
      "        \n",
      "        \n",
      "        Alias for fetch_lfw_people(download_if_missing=False)\n",
      "        \n",
      "        .. deprecated:: 0.17\n",
      "            This function will be removed in 0.19.\n",
      "            Use :func:`sklearn.datasets.fetch_lfw_people` with parameter\n",
      "            ``download_if_missing=False`` instead.\n",
      "        \n",
      "        Check fetch_lfw_people.__doc__ for the documentation and parameter list.\n",
      "    \n",
      "    load_linnerud(return_X_y=False)\n",
      "        Load and return the linnerud dataset (multivariate regression).\n",
      "        \n",
      "        Samples total: 20\n",
      "        Dimensionality: 3 for both data and targets\n",
      "        Features: integer\n",
      "        Targets: integer\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        return_X_y : boolean, default=False.\n",
      "            If True, returns ``(data, target)`` instead of a Bunch object.\n",
      "            See below for more information about the `data` and `target` object.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are: 'data' and\n",
      "            'targets', the two multivariate datasets, with 'data' corresponding to\n",
      "            the exercise and 'targets' corresponding to the physiological\n",
      "            measurements, as well as 'feature_names' and 'target_names'.\n",
      "        \n",
      "        (data, target) : tuple if ``return_X_y`` is True\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "    \n",
      "    load_mlcomp(name_or_id, set_='raw', mlcomp_root=None, **kwargs)\n",
      "        Load a datasets as downloaded from http://mlcomp.org\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        \n",
      "        name_or_id : the integer id or the string name metadata of the MLComp\n",
      "                     dataset to load\n",
      "        \n",
      "        set_ : select the portion to load: 'train', 'test' or 'raw'\n",
      "        \n",
      "        mlcomp_root : the filesystem path to the root folder where MLComp datasets\n",
      "                      are stored, if mlcomp_root is None, the MLCOMP_DATASETS_HOME\n",
      "                      environment variable is looked up instead.\n",
      "        \n",
      "        **kwargs : domain specific kwargs to be passed to the dataset loader.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <datasets>`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        \n",
      "        data : Bunch\n",
      "            Dictionary-like object, the interesting attributes are:\n",
      "            'filenames', the files holding the raw to learn, 'target', the\n",
      "            classification labels (integer index), 'target_names',\n",
      "            the meaning of the labels, and 'DESCR', the full description of the\n",
      "            dataset.\n",
      "        \n",
      "        Note on the lookup process: depending on the type of name_or_id,\n",
      "        will choose between integer id lookup or metadata name lookup by\n",
      "        looking at the unzipped archives and metadata file.\n",
      "        \n",
      "        TODO: implement zip dataset loading too\n",
      "    \n",
      "    load_sample_image(image_name)\n",
      "        Load the numpy array of a single sample image\n",
      "        \n",
      "        Parameters\n",
      "        -----------\n",
      "        image_name: {`china.jpg`, `flower.jpg`}\n",
      "            The name of the sample image loaded\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        img: 3D array\n",
      "            The image as a numpy array: height x width x color\n",
      "        \n",
      "        Examples\n",
      "        ---------\n",
      "        \n",
      "        >>> from sklearn.datasets import load_sample_image\n",
      "        >>> china = load_sample_image('china.jpg')   # doctest: +SKIP\n",
      "        >>> china.dtype                              # doctest: +SKIP\n",
      "        dtype('uint8')\n",
      "        >>> china.shape                              # doctest: +SKIP\n",
      "        (427, 640, 3)\n",
      "        >>> flower = load_sample_image('flower.jpg') # doctest: +SKIP\n",
      "        >>> flower.dtype                             # doctest: +SKIP\n",
      "        dtype('uint8')\n",
      "        >>> flower.shape                             # doctest: +SKIP\n",
      "        (427, 640, 3)\n",
      "    \n",
      "    load_sample_images()\n",
      "        Load sample images for image manipulation.\n",
      "        Loads both, ``china`` and ``flower``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : Bunch\n",
      "            Dictionary-like object with the following attributes :\n",
      "            'images', the two sample images, 'filenames', the file\n",
      "            names for the images, and 'DESCR'\n",
      "            the full description of the dataset.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        To load the data and visualize the images:\n",
      "        \n",
      "        >>> from sklearn.datasets import load_sample_images\n",
      "        >>> dataset = load_sample_images()     #doctest: +SKIP\n",
      "        >>> len(dataset.images)                #doctest: +SKIP\n",
      "        2\n",
      "        >>> first_img_data = dataset.images[0] #doctest: +SKIP\n",
      "        >>> first_img_data.shape               #doctest: +SKIP\n",
      "        (427, 640, 3)\n",
      "        >>> first_img_data.dtype               #doctest: +SKIP\n",
      "        dtype('uint8')\n",
      "    \n",
      "    load_svmlight_file(f, n_features=None, dtype=<type 'numpy.float64'>, multilabel=False, zero_based='auto', query_id=False)\n",
      "        Load datasets in the svmlight / libsvm format into sparse CSR matrix\n",
      "        \n",
      "        This format is a text-based format, with one sample per line. It does\n",
      "        not store zero valued features hence is suitable for sparse dataset.\n",
      "        \n",
      "        The first element of each line can be used to store a target variable\n",
      "        to predict.\n",
      "        \n",
      "        This format is used as the default format for both svmlight and the\n",
      "        libsvm command line programs.\n",
      "        \n",
      "        Parsing a text based source can be expensive. When working on\n",
      "        repeatedly on the same dataset, it is recommended to wrap this\n",
      "        loader with joblib.Memory.cache to store a memmapped backup of the\n",
      "        CSR results of the first call and benefit from the near instantaneous\n",
      "        loading of memmapped structures for the subsequent calls.\n",
      "        \n",
      "        In case the file contains a pairwise preference constraint (known\n",
      "        as \"qid\" in the svmlight format) these are ignored unless the\n",
      "        query_id parameter is set to True. These pairwise preference\n",
      "        constraints can be used to constraint the combination of samples\n",
      "        when using pairwise loss functions (as is the case in some\n",
      "        learning to rank problems) so that only pairs with the same\n",
      "        query_id value are considered.\n",
      "        \n",
      "        This implementation is written in Cython and is reasonably fast.\n",
      "        However, a faster API-compatible loader is also available at:\n",
      "        \n",
      "          https://github.com/mblondel/svmlight-loader\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : {str, file-like, int}\n",
      "            (Path to) a file to load. If a path ends in \".gz\" or \".bz2\", it will\n",
      "            be uncompressed on the fly. If an integer is passed, it is assumed to\n",
      "            be a file descriptor. A file-like or file descriptor will not be closed\n",
      "            by this function. A file-like object must be opened in binary mode.\n",
      "        \n",
      "        n_features : int or None\n",
      "            The number of features to use. If None, it will be inferred. This\n",
      "            argument is useful to load several files that are subsets of a\n",
      "            bigger sliced dataset: each subset might not have examples of\n",
      "            every feature, hence the inferred shape might vary from one\n",
      "            slice to another.\n",
      "        \n",
      "        multilabel : boolean, optional, default False\n",
      "            Samples may have several labels each (see\n",
      "            http://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/multilabel.html)\n",
      "        \n",
      "        zero_based : boolean or \"auto\", optional, default \"auto\"\n",
      "            Whether column indices in f are zero-based (True) or one-based\n",
      "            (False). If column indices are one-based, they are transformed to\n",
      "            zero-based to match Python/NumPy conventions.\n",
      "            If set to \"auto\", a heuristic check is applied to determine this from\n",
      "            the file contents. Both kinds of files occur \"in the wild\", but they\n",
      "            are unfortunately not self-identifying. Using \"auto\" or True should\n",
      "            always be safe.\n",
      "        \n",
      "        query_id : boolean, default False\n",
      "            If True, will return the query_id array for each file.\n",
      "        \n",
      "        dtype : numpy data type, default np.float64\n",
      "            Data type of dataset to be loaded. This will be the data type of the\n",
      "            output numpy arrays ``X`` and ``y``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : scipy.sparse matrix of shape (n_samples, n_features)\n",
      "        \n",
      "        y : ndarray of shape (n_samples,), or, in the multilabel a list of\n",
      "            tuples of length n_samples.\n",
      "        \n",
      "        query_id : array of shape (n_samples,)\n",
      "           query_id for each sample. Only returned when query_id is set to\n",
      "           True.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        load_svmlight_files: similar function for loading multiple files in this\n",
      "        format, enforcing the same number of features/columns on all of them.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        To use joblib.Memory to cache the svmlight file::\n",
      "        \n",
      "            from sklearn.externals.joblib import Memory\n",
      "            from sklearn.datasets import load_svmlight_file\n",
      "            mem = Memory(\"./mycache\")\n",
      "        \n",
      "            @mem.cache\n",
      "            def get_data():\n",
      "                data = load_svmlight_file(\"mysvmlightfile\")\n",
      "                return data[0], data[1]\n",
      "        \n",
      "            X, y = get_data()\n",
      "    \n",
      "    load_svmlight_files(files, n_features=None, dtype=<type 'numpy.float64'>, multilabel=False, zero_based='auto', query_id=False)\n",
      "        Load dataset from multiple files in SVMlight format\n",
      "        \n",
      "        This function is equivalent to mapping load_svmlight_file over a list of\n",
      "        files, except that the results are concatenated into a single, flat list\n",
      "        and the samples vectors are constrained to all have the same number of\n",
      "        features.\n",
      "        \n",
      "        In case the file contains a pairwise preference constraint (known\n",
      "        as \"qid\" in the svmlight format) these are ignored unless the\n",
      "        query_id parameter is set to True. These pairwise preference\n",
      "        constraints can be used to constraint the combination of samples\n",
      "        when using pairwise loss functions (as is the case in some\n",
      "        learning to rank problems) so that only pairs with the same\n",
      "        query_id value are considered.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        files : iterable over {str, file-like, int}\n",
      "            (Paths of) files to load. If a path ends in \".gz\" or \".bz2\", it will\n",
      "            be uncompressed on the fly. If an integer is passed, it is assumed to\n",
      "            be a file descriptor. File-likes and file descriptors will not be\n",
      "            closed by this function. File-like objects must be opened in binary\n",
      "            mode.\n",
      "        \n",
      "        n_features : int or None\n",
      "            The number of features to use. If None, it will be inferred from the\n",
      "            maximum column index occurring in any of the files.\n",
      "        \n",
      "            This can be set to a higher value than the actual number of features\n",
      "            in any of the input files, but setting it to a lower value will cause\n",
      "            an exception to be raised.\n",
      "        \n",
      "        multilabel : boolean, optional\n",
      "            Samples may have several labels each (see\n",
      "            http://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/multilabel.html)\n",
      "        \n",
      "        zero_based : boolean or \"auto\", optional\n",
      "            Whether column indices in f are zero-based (True) or one-based\n",
      "            (False). If column indices are one-based, they are transformed to\n",
      "            zero-based to match Python/NumPy conventions.\n",
      "            If set to \"auto\", a heuristic check is applied to determine this from\n",
      "            the file contents. Both kinds of files occur \"in the wild\", but they\n",
      "            are unfortunately not self-identifying. Using \"auto\" or True should\n",
      "            always be safe.\n",
      "        \n",
      "        query_id : boolean, defaults to False\n",
      "            If True, will return the query_id array for each file.\n",
      "        \n",
      "        dtype : numpy data type, default np.float64\n",
      "            Data type of dataset to be loaded. This will be the data type of the\n",
      "            output numpy arrays ``X`` and ``y``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        [X1, y1, ..., Xn, yn]\n",
      "        where each (Xi, yi) pair is the result from load_svmlight_file(files[i]).\n",
      "        \n",
      "        If query_id is set to True, this will return instead [X1, y1, q1,\n",
      "        ..., Xn, yn, qn] where (Xi, yi, qi) is the result from\n",
      "        load_svmlight_file(files[i])\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        When fitting a model to a matrix X_train and evaluating it against a\n",
      "        matrix X_test, it is essential that X_train and X_test have the same\n",
      "        number of features (X_train.shape[1] == X_test.shape[1]). This may not\n",
      "        be the case if you load the files individually with load_svmlight_file.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        load_svmlight_file\n",
      "    \n",
      "    make_biclusters(shape, n_clusters, noise=0.0, minval=10, maxval=100, shuffle=True, random_state=None)\n",
      "        Generate an array with constant block diagonal structure for\n",
      "        biclustering.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        shape : iterable (n_rows, n_cols)\n",
      "            The shape of the result.\n",
      "        \n",
      "        n_clusters : integer\n",
      "            The number of biclusters.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise.\n",
      "        \n",
      "        minval : int, optional (default=10)\n",
      "            Minimum value of a bicluster.\n",
      "        \n",
      "        maxval : int, optional (default=100)\n",
      "            Maximum value of a bicluster.\n",
      "        \n",
      "        shuffle : boolean, optional (default=True)\n",
      "            Shuffle the samples.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape `shape`\n",
      "            The generated array.\n",
      "        \n",
      "        rows : array of shape (n_clusters, X.shape[0],)\n",
      "            The indicators for cluster membership of each row.\n",
      "        \n",
      "        cols : array of shape (n_clusters, X.shape[1],)\n",
      "            The indicators for cluster membership of each column.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        \n",
      "        .. [1] Dhillon, I. S. (2001, August). Co-clustering documents and\n",
      "            words using bipartite spectral graph partitioning. In Proceedings\n",
      "            of the seventh ACM SIGKDD international conference on Knowledge\n",
      "            discovery and data mining (pp. 269-274). ACM.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_checkerboard\n",
      "    \n",
      "    make_blobs(n_samples=100, n_features=2, centers=3, cluster_std=1.0, center_box=(-10.0, 10.0), shuffle=True, random_state=None)\n",
      "        Generate isotropic Gaussian blobs for clustering.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The total number of points equally divided among clusters.\n",
      "        \n",
      "        n_features : int, optional (default=2)\n",
      "            The number of features for each sample.\n",
      "        \n",
      "        centers : int or array of shape [n_centers, n_features], optional\n",
      "            (default=3)\n",
      "            The number of centers to generate, or the fixed center locations.\n",
      "        \n",
      "        cluster_std : float or sequence of floats, optional (default=1.0)\n",
      "            The standard deviation of the clusters.\n",
      "        \n",
      "        center_box : pair of floats (min, max), optional (default=(-10.0, 10.0))\n",
      "            The bounding box for each cluster center when centers are\n",
      "            generated at random.\n",
      "        \n",
      "        shuffle : boolean, optional (default=True)\n",
      "            Shuffle the samples.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The generated samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The integer labels for cluster membership of each sample.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from sklearn.datasets.samples_generator import make_blobs\n",
      "        >>> X, y = make_blobs(n_samples=10, centers=3, n_features=2,\n",
      "        ...                   random_state=0)\n",
      "        >>> print(X.shape)\n",
      "        (10, 2)\n",
      "        >>> y\n",
      "        array([0, 0, 1, 0, 2, 2, 2, 1, 1, 0])\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_classification: a more intricate variant\n",
      "    \n",
      "    make_checkerboard(shape, n_clusters, noise=0.0, minval=10, maxval=100, shuffle=True, random_state=None)\n",
      "        Generate an array with block checkerboard structure for\n",
      "        biclustering.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        shape : iterable (n_rows, n_cols)\n",
      "            The shape of the result.\n",
      "        \n",
      "        n_clusters : integer or iterable (n_row_clusters, n_column_clusters)\n",
      "            The number of row and column clusters.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise.\n",
      "        \n",
      "        minval : int, optional (default=10)\n",
      "            Minimum value of a bicluster.\n",
      "        \n",
      "        maxval : int, optional (default=100)\n",
      "            Maximum value of a bicluster.\n",
      "        \n",
      "        shuffle : boolean, optional (default=True)\n",
      "            Shuffle the samples.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape `shape`\n",
      "            The generated array.\n",
      "        \n",
      "        rows : array of shape (n_clusters, X.shape[0],)\n",
      "            The indicators for cluster membership of each row.\n",
      "        \n",
      "        cols : array of shape (n_clusters, X.shape[1],)\n",
      "            The indicators for cluster membership of each column.\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        \n",
      "        .. [1] Kluger, Y., Basri, R., Chang, J. T., & Gerstein, M. (2003).\n",
      "            Spectral biclustering of microarray data: coclustering genes\n",
      "            and conditions. Genome research, 13(4), 703-716.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_biclusters\n",
      "    \n",
      "    make_circles(n_samples=100, shuffle=True, noise=None, random_state=None, factor=0.8)\n",
      "        Make a large circle containing a smaller circle in 2d.\n",
      "        \n",
      "        A simple toy dataset to visualize clustering and classification\n",
      "        algorithms.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The total number of points generated.\n",
      "        \n",
      "        shuffle: bool, optional (default=True)\n",
      "            Whether to shuffle the samples.\n",
      "        \n",
      "        noise : double or None (default=None)\n",
      "            Standard deviation of Gaussian noise added to the data.\n",
      "        \n",
      "        factor : double < 1 (default=.8)\n",
      "            Scale factor between inner and outer circle.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 2]\n",
      "            The generated samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The integer labels (0 or 1) for class membership of each sample.\n",
      "    \n",
      "    make_classification(n_samples=100, n_features=20, n_informative=2, n_redundant=2, n_repeated=0, n_classes=2, n_clusters_per_class=2, weights=None, flip_y=0.01, class_sep=1.0, hypercube=True, shift=0.0, scale=1.0, shuffle=True, random_state=None)\n",
      "        Generate a random n-class classification problem.\n",
      "        \n",
      "        This initially creates clusters of points normally distributed (std=1)\n",
      "        about vertices of a `2 * class_sep`-sided hypercube, and assigns an equal\n",
      "        number of clusters to each class. It introduces interdependence between\n",
      "        these features and adds various types of further noise to the data.\n",
      "        \n",
      "        Prior to shuffling, `X` stacks a number of these primary \"informative\"\n",
      "        features, \"redundant\" linear combinations of these, \"repeated\" duplicates\n",
      "        of sampled features, and arbitrary noise for and remaining features.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        n_features : int, optional (default=20)\n",
      "            The total number of features. These comprise `n_informative`\n",
      "            informative features, `n_redundant` redundant features, `n_repeated`\n",
      "            duplicated features and `n_features-n_informative-n_redundant-\n",
      "            n_repeated` useless features drawn at random.\n",
      "        \n",
      "        n_informative : int, optional (default=2)\n",
      "            The number of informative features. Each class is composed of a number\n",
      "            of gaussian clusters each located around the vertices of a hypercube\n",
      "            in a subspace of dimension `n_informative`. For each cluster,\n",
      "            informative features are drawn independently from  N(0, 1) and then\n",
      "            randomly linearly combined within each cluster in order to add\n",
      "            covariance. The clusters are then placed on the vertices of the\n",
      "            hypercube.\n",
      "        \n",
      "        n_redundant : int, optional (default=2)\n",
      "            The number of redundant features. These features are generated as\n",
      "            random linear combinations of the informative features.\n",
      "        \n",
      "        n_repeated : int, optional (default=0)\n",
      "            The number of duplicated features, drawn randomly from the informative\n",
      "            and the redundant features.\n",
      "        \n",
      "        n_classes : int, optional (default=2)\n",
      "            The number of classes (or labels) of the classification problem.\n",
      "        \n",
      "        n_clusters_per_class : int, optional (default=2)\n",
      "            The number of clusters per class.\n",
      "        \n",
      "        weights : list of floats or None (default=None)\n",
      "            The proportions of samples assigned to each class. If None, then\n",
      "            classes are balanced. Note that if `len(weights) == n_classes - 1`,\n",
      "            then the last class weight is automatically inferred.\n",
      "            More than `n_samples` samples may be returned if the sum of `weights`\n",
      "            exceeds 1.\n",
      "        \n",
      "        flip_y : float, optional (default=0.01)\n",
      "            The fraction of samples whose class are randomly exchanged.\n",
      "        \n",
      "        class_sep : float, optional (default=1.0)\n",
      "            The factor multiplying the hypercube dimension.\n",
      "        \n",
      "        hypercube : boolean, optional (default=True)\n",
      "            If True, the clusters are put on the vertices of a hypercube. If\n",
      "            False, the clusters are put on the vertices of a random polytope.\n",
      "        \n",
      "        shift : float, array of shape [n_features] or None, optional (default=0.0)\n",
      "            Shift features by the specified value. If None, then features\n",
      "            are shifted by a random value drawn in [-class_sep, class_sep].\n",
      "        \n",
      "        scale : float, array of shape [n_features] or None, optional (default=1.0)\n",
      "            Multiply features by the specified value. If None, then features\n",
      "            are scaled by a random value drawn in [1, 100]. Note that scaling\n",
      "            happens after shifting.\n",
      "        \n",
      "        shuffle : boolean, optional (default=True)\n",
      "            Shuffle the samples and the features.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The generated samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The integer labels for class membership of each sample.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The algorithm is adapted from Guyon [1] and was designed to generate\n",
      "        the \"Madelon\" dataset.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] I. Guyon, \"Design of experiments for the NIPS 2003 variable\n",
      "               selection benchmark\", 2003.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_blobs: simplified variant\n",
      "        make_multilabel_classification: unrelated generator for multilabel tasks\n",
      "    \n",
      "    make_friedman1(n_samples=100, n_features=10, noise=0.0, random_state=None)\n",
      "        Generate the \"Friedman \\#1\" regression problem\n",
      "        \n",
      "        This dataset is described in Friedman [1] and Breiman [2].\n",
      "        \n",
      "        Inputs `X` are independent features uniformly distributed on the interval\n",
      "        [0, 1]. The output `y` is created according to the formula::\n",
      "        \n",
      "            y(X) = 10 * sin(pi * X[:, 0] * X[:, 1]) + 20 * (X[:, 2] - 0.5) ** 2 + 10 * X[:, 3] + 5 * X[:, 4] + noise * N(0, 1).\n",
      "        \n",
      "        Out of the `n_features` features, only 5 are actually used to compute\n",
      "        `y`. The remaining features are independent of `y`.\n",
      "        \n",
      "        The number of features has to be >= 5.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        n_features : int, optional (default=10)\n",
      "            The number of features. Should be at least 5.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise applied to the output.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The input samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The output values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J. Friedman, \"Multivariate adaptive regression splines\", The Annals\n",
      "               of Statistics 19 (1), pages 1-67, 1991.\n",
      "        \n",
      "        .. [2] L. Breiman, \"Bagging predictors\", Machine Learning 24,\n",
      "               pages 123-140, 1996.\n",
      "    \n",
      "    make_friedman2(n_samples=100, noise=0.0, random_state=None)\n",
      "        Generate the \"Friedman \\#2\" regression problem\n",
      "        \n",
      "        This dataset is described in Friedman [1] and Breiman [2].\n",
      "        \n",
      "        Inputs `X` are 4 independent features uniformly distributed on the\n",
      "        intervals::\n",
      "        \n",
      "            0 <= X[:, 0] <= 100,\n",
      "            40 * pi <= X[:, 1] <= 560 * pi,\n",
      "            0 <= X[:, 2] <= 1,\n",
      "            1 <= X[:, 3] <= 11.\n",
      "        \n",
      "        The output `y` is created according to the formula::\n",
      "        \n",
      "            y(X) = (X[:, 0] ** 2 + (X[:, 1] * X[:, 2]  - 1 / (X[:, 1] * X[:, 3])) ** 2) ** 0.5 + noise * N(0, 1).\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise applied to the output.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 4]\n",
      "            The input samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The output values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J. Friedman, \"Multivariate adaptive regression splines\", The Annals\n",
      "               of Statistics 19 (1), pages 1-67, 1991.\n",
      "        \n",
      "        .. [2] L. Breiman, \"Bagging predictors\", Machine Learning 24,\n",
      "               pages 123-140, 1996.\n",
      "    \n",
      "    make_friedman3(n_samples=100, noise=0.0, random_state=None)\n",
      "        Generate the \"Friedman \\#3\" regression problem\n",
      "        \n",
      "        This dataset is described in Friedman [1] and Breiman [2].\n",
      "        \n",
      "        Inputs `X` are 4 independent features uniformly distributed on the\n",
      "        intervals::\n",
      "        \n",
      "            0 <= X[:, 0] <= 100,\n",
      "            40 * pi <= X[:, 1] <= 560 * pi,\n",
      "            0 <= X[:, 2] <= 1,\n",
      "            1 <= X[:, 3] <= 11.\n",
      "        \n",
      "        The output `y` is created according to the formula::\n",
      "        \n",
      "            y(X) = arctan((X[:, 1] * X[:, 2] - 1 / (X[:, 1] * X[:, 3])) / X[:, 0]) + noise * N(0, 1).\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise applied to the output.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 4]\n",
      "            The input samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The output values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J. Friedman, \"Multivariate adaptive regression splines\", The Annals\n",
      "               of Statistics 19 (1), pages 1-67, 1991.\n",
      "        \n",
      "        .. [2] L. Breiman, \"Bagging predictors\", Machine Learning 24,\n",
      "               pages 123-140, 1996.\n",
      "    \n",
      "    make_gaussian_quantiles(mean=None, cov=1.0, n_samples=100, n_features=2, n_classes=3, shuffle=True, random_state=None)\n",
      "        Generate isotropic Gaussian and label samples by quantile\n",
      "        \n",
      "        This classification dataset is constructed by taking a multi-dimensional\n",
      "        standard normal distribution and defining classes separated by nested\n",
      "        concentric multi-dimensional spheres such that roughly equal numbers of\n",
      "        samples are in each class (quantiles of the :math:`\\chi^2` distribution).\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        mean : array of shape [n_features], optional (default=None)\n",
      "            The mean of the multi-dimensional normal distribution.\n",
      "            If None then use the origin (0, 0, ...).\n",
      "        \n",
      "        cov : float, optional (default=1.)\n",
      "            The covariance matrix will be this value times the unit matrix. This\n",
      "            dataset only produces symmetric normal distributions.\n",
      "        \n",
      "        n_samples : int, optional (default=100)\n",
      "            The total number of points equally divided among classes.\n",
      "        \n",
      "        n_features : int, optional (default=2)\n",
      "            The number of features for each sample.\n",
      "        \n",
      "        n_classes : int, optional (default=3)\n",
      "            The number of classes\n",
      "        \n",
      "        shuffle : boolean, optional (default=True)\n",
      "            Shuffle the samples.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The generated samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The integer labels for quantile membership of each sample.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The dataset is from Zhu et al [1].\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J. Zhu, H. Zou, S. Rosset, T. Hastie, \"Multi-class AdaBoost\", 2009.\n",
      "    \n",
      "    make_hastie_10_2(n_samples=12000, random_state=None)\n",
      "        Generates data for binary classification used in\n",
      "        Hastie et al. 2009, Example 10.2.\n",
      "        \n",
      "        The ten features are standard independent Gaussian and\n",
      "        the target ``y`` is defined by::\n",
      "        \n",
      "          y[i] = 1 if np.sum(X[i] ** 2) > 9.34 else -1\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=12000)\n",
      "            The number of samples.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 10]\n",
      "            The input samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The output values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] T. Hastie, R. Tibshirani and J. Friedman, \"Elements of Statistical\n",
      "               Learning Ed. 2\", Springer, 2009.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_gaussian_quantiles: a generalization of this dataset approach\n",
      "    \n",
      "    make_low_rank_matrix(n_samples=100, n_features=100, effective_rank=10, tail_strength=0.5, random_state=None)\n",
      "        Generate a mostly low rank matrix with bell-shaped singular values\n",
      "        \n",
      "        Most of the variance can be explained by a bell-shaped curve of width\n",
      "        effective_rank: the low rank part of the singular values profile is::\n",
      "        \n",
      "            (1 - tail_strength) * exp(-1.0 * (i / effective_rank) ** 2)\n",
      "        \n",
      "        The remaining singular values' tail is fat, decreasing as::\n",
      "        \n",
      "            tail_strength * exp(-0.1 * i / effective_rank).\n",
      "        \n",
      "        The low rank part of the profile can be considered the structured\n",
      "        signal part of the data while the tail can be considered the noisy\n",
      "        part of the data that cannot be summarized by a low number of linear\n",
      "        components (singular vectors).\n",
      "        \n",
      "        This kind of singular profiles is often seen in practice, for instance:\n",
      "         - gray level pictures of faces\n",
      "         - TF-IDF vectors of text documents crawled from the web\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        n_features : int, optional (default=100)\n",
      "            The number of features.\n",
      "        \n",
      "        effective_rank : int, optional (default=10)\n",
      "            The approximate number of singular vectors required to explain most of\n",
      "            the data by linear combinations.\n",
      "        \n",
      "        tail_strength : float between 0.0 and 1.0, optional (default=0.5)\n",
      "            The relative importance of the fat noisy tail of the singular values\n",
      "            profile.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The matrix.\n",
      "    \n",
      "    make_moons(n_samples=100, shuffle=True, noise=None, random_state=None)\n",
      "        Make two interleaving half circles\n",
      "        \n",
      "        A simple toy dataset to visualize clustering and classification\n",
      "        algorithms. Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The total number of points generated.\n",
      "        \n",
      "        shuffle : bool, optional (default=True)\n",
      "            Whether to shuffle the samples.\n",
      "        \n",
      "        noise : double or None (default=None)\n",
      "            Standard deviation of Gaussian noise added to the data.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 2]\n",
      "            The generated samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The integer labels (0 or 1) for class membership of each sample.\n",
      "    \n",
      "    make_multilabel_classification(n_samples=100, n_features=20, n_classes=5, n_labels=2, length=50, allow_unlabeled=True, sparse=False, return_indicator='dense', return_distributions=False, random_state=None)\n",
      "        Generate a random multilabel classification problem.\n",
      "        \n",
      "        For each sample, the generative process is:\n",
      "            - pick the number of labels: n ~ Poisson(n_labels)\n",
      "            - n times, choose a class c: c ~ Multinomial(theta)\n",
      "            - pick the document length: k ~ Poisson(length)\n",
      "            - k times, choose a word: w ~ Multinomial(theta_c)\n",
      "        \n",
      "        In the above process, rejection sampling is used to make sure that\n",
      "        n is never zero or more than `n_classes`, and that the document length\n",
      "        is never zero. Likewise, we reject classes which have already been chosen.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        n_features : int, optional (default=20)\n",
      "            The total number of features.\n",
      "        \n",
      "        n_classes : int, optional (default=5)\n",
      "            The number of classes of the classification problem.\n",
      "        \n",
      "        n_labels : int, optional (default=2)\n",
      "            The average number of labels per instance. More precisely, the number\n",
      "            of labels per sample is drawn from a Poisson distribution with\n",
      "            ``n_labels`` as its expected value, but samples are bounded (using\n",
      "            rejection sampling) by ``n_classes``, and must be nonzero if\n",
      "            ``allow_unlabeled`` is False.\n",
      "        \n",
      "        length : int, optional (default=50)\n",
      "            The sum of the features (number of words if documents) is drawn from\n",
      "            a Poisson distribution with this expected value.\n",
      "        \n",
      "        allow_unlabeled : bool, optional (default=True)\n",
      "            If ``True``, some instances might not belong to any class.\n",
      "        \n",
      "        sparse : bool, optional (default=False)\n",
      "            If ``True``, return a sparse feature matrix\n",
      "        \n",
      "            .. versionadded:: 0.17\n",
      "               parameter to allow *sparse* output.\n",
      "        \n",
      "        return_indicator : 'dense' (default) | 'sparse' | False\n",
      "            If ``dense`` return ``Y`` in the dense binary indicator format. If\n",
      "            ``'sparse'`` return ``Y`` in the sparse binary indicator format.\n",
      "            ``False`` returns a list of lists of labels.\n",
      "        \n",
      "        return_distributions : bool, optional (default=False)\n",
      "            If ``True``, return the prior class probability and conditional\n",
      "            probabilities of features given classes, from which the data was\n",
      "            drawn.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The generated samples.\n",
      "        \n",
      "        Y : array or sparse CSR matrix of shape [n_samples, n_classes]\n",
      "            The label sets.\n",
      "        \n",
      "        p_c : array, shape [n_classes]\n",
      "            The probability of each class being drawn. Only returned if\n",
      "            ``return_distributions=True``.\n",
      "        \n",
      "        p_w_c : array, shape [n_features, n_classes]\n",
      "            The probability of each feature being drawn given each class.\n",
      "            Only returned if ``return_distributions=True``.\n",
      "    \n",
      "    make_regression(n_samples=100, n_features=100, n_informative=10, n_targets=1, bias=0.0, effective_rank=None, tail_strength=0.5, noise=0.0, shuffle=True, coef=False, random_state=None)\n",
      "        Generate a random regression problem.\n",
      "        \n",
      "        The input set can either be well conditioned (by default) or have a low\n",
      "        rank-fat tail singular profile. See :func:`make_low_rank_matrix` for\n",
      "        more details.\n",
      "        \n",
      "        The output is generated by applying a (potentially biased) random linear\n",
      "        regression model with `n_informative` nonzero regressors to the previously\n",
      "        generated input and some gaussian centered noise with some adjustable\n",
      "        scale.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        n_features : int, optional (default=100)\n",
      "            The number of features.\n",
      "        \n",
      "        n_informative : int, optional (default=10)\n",
      "            The number of informative features, i.e., the number of features used\n",
      "            to build the linear model used to generate the output.\n",
      "        \n",
      "        n_targets : int, optional (default=1)\n",
      "            The number of regression targets, i.e., the dimension of the y output\n",
      "            vector associated with a sample. By default, the output is a scalar.\n",
      "        \n",
      "        bias : float, optional (default=0.0)\n",
      "            The bias term in the underlying linear model.\n",
      "        \n",
      "        effective_rank : int or None, optional (default=None)\n",
      "            if not None:\n",
      "                The approximate number of singular vectors required to explain most\n",
      "                of the input data by linear combinations. Using this kind of\n",
      "                singular spectrum in the input allows the generator to reproduce\n",
      "                the correlations often observed in practice.\n",
      "            if None:\n",
      "                The input set is well conditioned, centered and gaussian with\n",
      "                unit variance.\n",
      "        \n",
      "        tail_strength : float between 0.0 and 1.0, optional (default=0.5)\n",
      "            The relative importance of the fat noisy tail of the singular values\n",
      "            profile if `effective_rank` is not None.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise applied to the output.\n",
      "        \n",
      "        shuffle : boolean, optional (default=True)\n",
      "            Shuffle the samples and the features.\n",
      "        \n",
      "        coef : boolean, optional (default=False)\n",
      "            If True, the coefficients of the underlying linear model are returned.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The input samples.\n",
      "        \n",
      "        y : array of shape [n_samples] or [n_samples, n_targets]\n",
      "            The output values.\n",
      "        \n",
      "        coef : array of shape [n_features] or [n_features, n_targets], optional\n",
      "            The coefficient of the underlying linear model. It is returned only if\n",
      "            coef is True.\n",
      "    \n",
      "    make_s_curve(n_samples=100, noise=0.0, random_state=None)\n",
      "        Generate an S curve dataset.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of sample points on the S curve.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 3]\n",
      "            The points.\n",
      "        \n",
      "        t : array of shape [n_samples]\n",
      "            The univariate position of the sample according to the main dimension\n",
      "            of the points in the manifold.\n",
      "    \n",
      "    make_sparse_coded_signal(n_samples, n_components, n_features, n_nonzero_coefs, random_state=None)\n",
      "        Generate a signal as a sparse combination of dictionary elements.\n",
      "        \n",
      "        Returns a matrix Y = DX, such as D is (n_features, n_components),\n",
      "        X is (n_components, n_samples) and each column of X has exactly\n",
      "        n_nonzero_coefs non-zero elements.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int\n",
      "            number of samples to generate\n",
      "        \n",
      "        n_components:  int,\n",
      "            number of components in the dictionary\n",
      "        \n",
      "        n_features : int\n",
      "            number of features of the dataset to generate\n",
      "        \n",
      "        n_nonzero_coefs : int\n",
      "            number of active (non-zero) coefficients in each sample\n",
      "        \n",
      "        random_state : int or RandomState instance, optional (default=None)\n",
      "            seed used by the pseudo random number generator\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        data : array of shape [n_features, n_samples]\n",
      "            The encoded signal (Y).\n",
      "        \n",
      "        dictionary : array of shape [n_features, n_components]\n",
      "            The dictionary with normalized components (D).\n",
      "        \n",
      "        code : array of shape [n_components, n_samples]\n",
      "            The sparse code such that each column of this matrix has exactly\n",
      "            n_nonzero_coefs non-zero items (X).\n",
      "    \n",
      "    make_sparse_spd_matrix(dim=1, alpha=0.95, norm_diag=False, smallest_coef=0.1, largest_coef=0.9, random_state=None)\n",
      "        Generate a sparse symmetric definite positive matrix.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        dim : integer, optional (default=1)\n",
      "            The size of the random matrix to generate.\n",
      "        \n",
      "        alpha : float between 0 and 1, optional (default=0.95)\n",
      "            The probability that a coefficient is zero (see notes). Larger values \n",
      "            enforce more sparsity.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        largest_coef : float between 0 and 1, optional (default=0.9)\n",
      "            The value of the largest coefficient.\n",
      "        \n",
      "        smallest_coef : float between 0 and 1, optional (default=0.1)\n",
      "            The value of the smallest coefficient.\n",
      "        \n",
      "        norm_diag : boolean, optional (default=False)\n",
      "            Whether to normalize the output matrix to make the leading diagonal\n",
      "            elements all 1\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        prec : sparse matrix of shape (dim, dim)\n",
      "            The generated matrix.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The sparsity is actually imposed on the cholesky factor of the matrix.\n",
      "        Thus alpha does not translate directly into the filling fraction of\n",
      "        the matrix itself.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_spd_matrix\n",
      "    \n",
      "    make_sparse_uncorrelated(n_samples=100, n_features=10, random_state=None)\n",
      "        Generate a random regression problem with sparse uncorrelated design\n",
      "        \n",
      "        This dataset is described in Celeux et al [1]. as::\n",
      "        \n",
      "            X ~ N(0, 1)\n",
      "            y(X) = X[:, 0] + 2 * X[:, 1] - 2 * X[:, 2] - 1.5 * X[:, 3]\n",
      "        \n",
      "        Only the first 4 features are informative. The remaining features are\n",
      "        useless.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of samples.\n",
      "        \n",
      "        n_features : int, optional (default=10)\n",
      "            The number of features.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, n_features]\n",
      "            The input samples.\n",
      "        \n",
      "        y : array of shape [n_samples]\n",
      "            The output values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] G. Celeux, M. El Anbari, J.-M. Marin, C. P. Robert,\n",
      "               \"Regularization in regression: comparing Bayesian and frequentist\n",
      "               methods in a poorly informative situation\", 2009.\n",
      "    \n",
      "    make_spd_matrix(n_dim, random_state=None)\n",
      "        Generate a random symmetric, positive-definite matrix.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_dim : int\n",
      "            The matrix dimension.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_dim, n_dim]\n",
      "            The random symmetric, positive-definite matrix.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        make_sparse_spd_matrix\n",
      "    \n",
      "    make_swiss_roll(n_samples=100, noise=0.0, random_state=None)\n",
      "        Generate a swiss roll dataset.\n",
      "        \n",
      "        Read more in the :ref:`User Guide <sample_generators>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n_samples : int, optional (default=100)\n",
      "            The number of sample points on the S curve.\n",
      "        \n",
      "        noise : float, optional (default=0.0)\n",
      "            The standard deviation of the gaussian noise.\n",
      "        \n",
      "        random_state : int, RandomState instance or None, optional (default=None)\n",
      "            If int, random_state is the seed used by the random number generator;\n",
      "            If RandomState instance, random_state is the random number generator;\n",
      "            If None, the random number generator is the RandomState instance used\n",
      "            by `np.random`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        X : array of shape [n_samples, 3]\n",
      "            The points.\n",
      "        \n",
      "        t : array of shape [n_samples]\n",
      "            The univariate position of the sample according to the main dimension\n",
      "            of the points in the manifold.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The algorithm is from Marsland [1].\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] S. Marsland, \"Machine Learning: An Algorithmic Perspective\",\n",
      "               Chapter 10, 2009.\n",
      "               http://seat.massey.ac.nz/personal/s.r.marsland/Code/10/lle.py\n",
      "    \n",
      "    mldata_filename(dataname)\n",
      "        Convert a raw name for a data set in a mldata.org filename.\n",
      "\n",
      "DATA\n",
      "    __all__ = ['clear_data_home', 'dump_svmlight_file', 'fetch_20newsgroup...\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sklearn.datasets as datasets\n",
    "help(datasets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Matplotlib\n",
    "\n",
    "\n",
    "\n",
    "When working with ipython notebooks the use of '--pylab inline' is important, so that plots are displayed in the notebook and not in a new window. Alternatively you can also use the following code inside a notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x82ad160>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYYAAAEICAYAAABbOlNNAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl4VOX5//H3nRUICRASQiDsJIEAAhIWEWVHQC1q1UKr\ntSu1ikvVWu2m3a2tta5YFCsWFbVuiCgqioKyhX0nCwHClrAlEBKy3b8/MvQX8k3IMpM5M5n7dV1z\n5cw5z5n5TMS585zleURVMcYYY84JcjqAMcYY32KFwRhjzHmsMBhjjDmPFQZjjDHnscJgjDHmPFYY\njDHGnMcKgzH1ICK/FJEXGtA+RUTSRETq0fYtEZniXkJjPEfsPgZjPE9E3gLeVNUF9Wg7DJitqkOa\nPpkxdbPCYIyHiUg8sA3opKrF9dwnHZihqmlNGs6YerBDScZUIyK/EJEDInJKRHaJyHgReVhE5ru2\ndxcRFZFbRGSfiBwVkV9VeYmJwPpzRUFEeonIcRG52PW8k4jkiciYKvssA670zic05sKsMBhThYgk\nA7OAoaoaCVwBZNfSfBSQDIwHfisifV3rBwC7zjVS1UzgF8B8EWkF/BuYp6rLqrzWDmCg5z6JMY1n\nhcGY85UD4UCKiISqarbri70mv1PVIlXdBGzi/3+xtwVOVW2oqs8DGcBqIB6o2sPA1b6thz6DMW6x\nwmBMFaqaAdwNPAzkisgCEelUS/PDVZbPAK1dyyeAyBraPw/0B55S1bPVtkUCJxub2xhPssJgTDWq\n+qqqjgK6AQr8tYEvsRlIqrpCRFoD/wTmAg+LSHS1ffpS2eswxnFWGIypQkSSRWSciIQDxUARUNHA\nl/kEuFhEWlRZ9wSQpqo/Aj4Anqu2z2jgw0bGNsajrDAYc75w4BHgKJWHijoADzbkBVT1CPAZMA1A\nRKYBk4GfuprcQ2Xh+I5r+1DgtKqu8cQHMMZddh+DMU1ARFKAecAwreN/MtfNcHNVdbFXwhlTBysM\nxhhjzmOHkowxxpzHCoMxxpjzWGEwxhhznhCnAzRGTEyMdu/e3ekYxhjjV9atW3dUVWPraueXhaF7\n9+6kpdkglMYY0xAisrc+7exQkjHGmPNYYTDGGHMeKwzGGGPOY4XBGGPMeawwGGOMOY9HCoOIvCgi\nuSKytZbtIiJPikiGiGw+N8Wha9tk1/SJGSLygCfyGGOMaTxP9RheonL0yNpMARJdj5nAbAARCQae\ncW1PAWa4Bh8zxhjjEI/cx6CqX4pI9ws0mQa87BplcpWItBWReKA7kKGqWQAissDVdrsncpmGKS4t\nJyP3NNnHCjlRWMLJM6UAhIcG0To8lC7RLekWHUFCu5YEBYnDaY0xTcVbN7h1BvZXeZ7jWlfT+uE1\nvYCIzKSyt0HXrl2bJmWAKSuvYM2e43yxO48v04+y63ABFfUYbLdtq1CGdG3HyN4xTOnfkU5tWzZ9\nWGOM1/jNnc+qOgeYA5Cammpjhbth77FCXluzn3c25HCk4CyhwcKQbu24fWxvkjtG0iu2Ne1bh9G2\nZRgiUFJWwcmiUvYfP0P20UI27DvJ2uzjLN2Zyx8Wbefirm25aUQ3rrwonvCQYKc/njHGTd4qDAeA\nLlWeJ7jWhday3jSBnYcLePbzTBZtPoiIMDY5loeuTmB0UiwR4bX/UwgNDiIiPITObVsyomd7pg+r\n7LHtOVrI4i2HeHt9Dve8sYk/L97JD0f14Hsju9MyzAqEMf7KYxP1uM4xLFLV/jVsuxKYBUyl8lDR\nk6o6TERCgN3AeCoLwlrg26q67ULvlZqaqjZWUv3lnirmbx/t4r/rc2gVGsxNl3TjB5f2IC6qRd07\n14Oqsjz9KHNX7OGL3Xl0iAznrgmJTB/alWA7F2GMzxCRdaqaWlc7j/QYROQ1YAwQIyI5wENU9gZQ\n1eeAxVQWhQzgDPB917YyEZkFLAGCgRfrKgqm/ioqlJe+zuaxj3dRUl7Bjy/ryW1jetG2VZhH30dE\nuDwplsuTYlmbfZy/friTX72zlTfScnjkugH0jY/y6PsZY5qWX07taT2Guu09VsjP39zMmuzjjEmO\n5aGr+9EjJsIr762qLNx0kN+/v538olJuH9ubO8b1JiTY7qc0xkle7TEY37J4yyF+/uYmgkT42/UX\ncf2QBES8d0hHRJg2qDOjk2L53fvbeWJpOiszj/H49EF0tiuYjPF59idcM1JaXsEfF23ntlfWk9Qx\nkiU/u5wbUrt4tShU1bZVGI9/axCPf2sg2w7mc+WTy1mZecyRLMaY+rPC0Ezknynl5rmreWHFHr43\nsjuvz7zEZ+4vuHZwAovuvIyY1uHcPHc1r67e53QkY8wFWGFoBg6eLOL6575m3d4TPP6tgTz8jX6E\nhfjWf9oeMRG8fdtIRiXG8Mt3tvD797dTUZ+76YwxXudb3x6mwXYcKuDaZ7/icH4x834wjGsHJzgd\nqVZRLUKZe8tQvn9pd178ag/3/XcTZeUVTscyxlRjJ5/92NYD+XznhdW0CA3izZ9eQp+Ovn9ZaHCQ\n8NurUohuFcZjn+ym8GwZT84YbHdMG+NDrMfgp84VhdbhIfz31pF+URTOERHuGJ/IQ1ensGTbEWa+\nvI6zZeVOxzLGuFhh8ENVi8KCmSPoEt3K6UiN8v1Le/DXbw7gi9153PHqBkrtsJIxPsEKg5/JzDvN\nd19c4/dF4ZxvDe3Kw1en8PH2I9z35ibK7YS0MY6zcwx+5EhBMd+duwYB5v9ouN8XhXO+d2kPzpSW\n8+hHu2gdHsIfr+nv2L0XxhgrDH4jv6iUW15cw8kzJSyYeYnXhrfwltvG9Ca/qJR/fZFFl+hW3Dq6\nl9ORjAlYVhj8QGl5BT+dv47MvNP8+3vDGJDQxulITeIXV/Th4MliHvlwJ53btuTqgZ2cjmRMQLLC\n4Af+sGg7X2ce4+83DGRUYozTcZpMUFDl2E6H84u4941NxLdpQWr3aKdjGRNw7OSzj5u/ai8vr9zL\nzMt7cv0Q3715zVNahAYz5+ZUOrdrya3z13M4v9jpSMYEHCsMPmxl5jEeXriNscmx/GJyH6fjeE27\niDDm3DyEopIybp1v9zgY421WGHzUwZNF3PbKOrrHRPDkjMEBNxNaYlwkj904kI37T/LwQpu7yRhv\n8khhEJHJIrJLRDJE5IEatv9cRDa6HltFpFxEol3bskVki2ubzb5D5cnmWa+up7RcmXPzECJbhDod\nyRGT+8dz+9hevLZmv43IaowXuV0YRCQYeAaYAqQAM0QkpWobVf2bqg5S1UHAg8AXqnq8SpOxru11\nziwUCB79aCfr953kkW8OoGdsa6fjOOqeicmMTorl4YXb2Hog3+k4xgQET/QYhgEZqpqlqiXAAmDa\nBdrPAF7zwPs2Sx9vO8zzy/fw3Uu6cdVFdrlmcJDw+LcG0S4ilDte20Dh2TKnIxnT7HmiMHQG9ld5\nnuNa93+ISCtgMvBWldUKfCoi60RkZm1vIiIzRSRNRNLy8vI8ENv37D9+hvve3MSAzm341ZV9nY7j\nM6Ijwnhi+mD2Hivkt+/Z+QZjmpq3Tz5fDXxV7TDSKNchpinA7SJyeU07quocVU1V1dTY2FhvZPWq\n8grlnjc2ogrPfPtiG4a6mhE92zNrXCJvrc/hnQ05TscxplnzRGE4AHSp8jzBta4m06l2GElVD7h+\n5gLvUHloKuDM+TKLtdkn+N20fnRt3zzGQPK0O8f1Zlj3aH79zlayjxY6HceYZssThWEtkCgiPUQk\njMov/4XVG4lIG2A08F6VdREiEnluGZgEbPVAJr+y9UA+//hkF1cOiOfawTUehTNASHAQ/5w+iOAg\n4V4bidWYJuN2YVDVMmAWsATYAbyhqttE5FYRubVK02uBj1W16p96ccAKEdkErAE+UNWP3M3kT4pL\ny/nZ6xtp1yrMRhWth05tW/K7af1Yt/cEc1dkOR3HmGbJI2MlqepiYHG1dc9Ve/4S8FK1dVnAQE9k\n8Fd/W7KL9NzTzPvBMNpFhDkdxy9cM6gzH245zN8/3s3Y5A4kxkU6HcmYZsXufHbQqqxjzF1ReWnq\n6KTmd0K9qYgIf7p2ABFhwdz75ibKbOY3YzzKCoNDikvLeeCtzXSNbsWDU+zS1IaKjQznT9cOYHNO\nPrOXZTodx5hmxQqDQx7/dDfZx87wyHUDaBlml6Y2xtQB8Vw9sBNPLE1nx6ECp+MY02xYYXDA5pyT\nPP9lFtOHdmFk7+Y7v4I3/P4b/YhqGcqDb2+xq5SM8RArDF5WWl7B/f/dTEzrcB6caoeQ3NUuIozf\nXpXCxv0nmb9qr9NxjGkWrDB42Zwvs9h5+BR/uKY/bVoG5qipnjZtUCcuS4zhb0t2cSi/yOk4xvg9\nKwxetOdoIU98ms6VA+K5ol9Hp+M0GyLCn64ZQFlFBQ/ZWErGuM0Kg5eoKr99byvhIUE8dHVK3TuY\nBunavhV3T0ji4+1H+GjrYafjGOPXrDB4yYdbD7M8/Sj3TEqiQ1QLp+M0Sz8c1YO+8VE8tHArp214\nbmMazQqDFxSeLeMPi7bTNz6Km0d0czpOsxUaHMSfr+3PkYKzPLU03ek4xvgtKwxe8ORn6RzKL+aP\n1/QjJNh+5U1pcNd23JiawNwVe8jIPe10HGP8kn1LNbH0I6eYu3wPNwxJYEi3aKfjBIT7J/ehZVgw\nDy/chqrd22BMQ1lhaEKqym/e20pEeAgPTOnjdJyAEdM6nHsnJrEi4yhLttmJaGMaygpDE3p/8yFW\nZR3n51ck0751uNNxAspNI7rRp2Mkf1i0g6KScqfjGONXrDA0kaKSch5ZvIP+naOYMayr03ECTkhw\nEL/7Rj8OnCxi9rIMp+MY41esMDSR55dncTC/mN9cmUJwkE2+44ThPdvzjYGdeO7LLPYdO+N0HGP8\nhkcKg4hMFpFdIpIhIg/UsH2MiOSLyEbX47f13dcfHSkoZvayTKb078jwnu2djhPQfjm1LyFBwl8+\n3OF0FGP8htuFQUSCgWeAKUAKMENEarq1d7mqDnI9ft/Aff3Kox/torxCbZ4FH9CxTQt+cnkvPtx6\nmDV7jjsdxxi/4IkewzAgQ1WzVLUEWABM88K+PmlLTj5vrc/h+6O607V9K6fjGODHl/egY1QL/vTB\ndipsaG5j6uSJwtAZ2F/leY5rXXUjRWSziHwoIv0auC8iMlNE0kQkLS8vzwOxPU9V+f2ibcS0DmPW\n2N5OxzEurcJCuO+KZDbl5PP+5oNOxzHG53nr5PN6oKuqXgQ8Bbzb0BdQ1TmqmqqqqbGxvjk/8odb\nD7M2+wT3TEwmsoUNqe1LrhvcmX6donj0o10Ul9rlq8ZciCcKwwGgS5XnCa51/6OqBap62rW8GAgV\nkZj67OsvikvL+fPiHfTpGMm3hnapewfjVUFBwq+u7MuBk0W8+NUep+MY49M8URjWAoki0kNEwoDp\nwMKqDUSko4iIa3mY632P1Wdff/HyymxyThTxm6vs8lRfNbJXDBP6xvHs55kcPX3W6TjG+Cy3C4Oq\nlgGzgCXADuANVd0mIreKyK2uZtcDW0VkE/AkMF0r1bivu5m8Lf9MKc98nsmY5FgutTmcfdqDU/tQ\nXFrO45/sdjqKMT4rxBMv4jo8tLjauueqLD8NPF3fff3N7C8yKSgu5f4rbDwkX9crtjXfGd6V/6za\ny/dGdicxLtLpSMb4HLvz2U2H8ov491d7uHZQZ1I6RTkdx9TDXROSiAgL4e8f73I6ijE+yQqDm574\nNB1V+NnEJKejmHqKjghj5uU9WbLtCOv3nXA6jjE+xwqDGzJyT/FG2n5uGtGNLtF2M5s/+cGoHsS0\nDuOvH+60ORuMqcYKgxse/WgXrcJCmDXObmbzNxHhIdwxLpHVe47zZfpRp+MY41OsMDTSur3H+Xj7\nEW4d3ZPoiDCn45hGmDGsKwntWvLoRzttqAxjqrDC0AiqyiMf7iQ2MpwfjOrhdBzTSGEhQdw7KYlt\nBwv4YMshp+MY4zOsMDTC0h25rM0+wd0TEmkV5pErfo1Dpg3sTJ+OkTz28S5KyyucjmOMT7DC0EAV\nFcrfluyiR0wEN6ba0Bf+LihIuH9yMtnHzvBG2v66dzAmAFhhaKBFWw6x68gp7p6QSGiw/fqag7HJ\nHRjavR1PfJpu80MbgxWGBikrr+Cfn+4mOS6Sqy/q5HQc4yEiwv2T+5B76iz//toG2DPGCkMDvLvx\nIFl5hfxsYhJBNlBeszK0ezTj+nTgX19kcaq41Ok4xjjKCkM9lZRV8MTS3fTvHMUV/eKcjmOawD0T\nk8gvKuXFFdlORzHGUVYY6unNdfvZf7yIeycm4xpB3DQz/Tu3YVJKHC+syCL/jPUaTOCywlAPxaXl\nPP1ZBhd3bcuYZN+cPc54xs8mJnGquIwXVmQ5HcUYx1hhqIfX1uzjUH4x902y3kJz1zc+iisviufF\nFXs4XljidBxjHOGRwiAik0Vkl4hkiMgDNWz/johsFpEtIvK1iAyssi3btX6jiKR5Io8nFZWU88zn\nmYzoGc1Im4QnINw9PpEzpeX868tMp6MY4wi3C4OIBAPPAFOAFGCGiKRUa7YHGK2qA4A/AHOqbR+r\nqoNUNdXdPJ728spsjp4+y72Tkp2OYrwkMS6SaQM78fLXe8k7ZVOAmsDjiR7DMCBDVbNUtQRYAEyr\n2kBVv1bVcwPfrwISPPC+Te5UcSnPfZHJ6KRYhnaPdjqO8aI7xydytqyc576wXoMJPJ4oDJ2BqmMJ\n5LjW1eaHwIdVnivwqYisE5GZte0kIjNFJE1E0vLy8twKXF///iqbE2dKuXeSTcITaHrGtua6ixOY\nv2ovRwqKnY5jjFd59eSziIylsjD8osrqUao6iMpDUbeLyOU17auqc1Q1VVVTY2Ob/sqg/KJSnl+e\nxcSUOC5KaNvk72d8z53jEimvUJ79PMPpKMZ4lScKwwGg6mhyCa515xGRi4AXgGmqeuzcelU94PqZ\nC7xD5aEpx730VTanisu4e0Ki01GMQ7q2b8UNqQm8tmY/B04WOR3HGK/xRGFYCySKSA8RCQOmAwur\nNhCRrsDbwM2qurvK+ggRiTy3DEwCtnogk1sKikuZu6Kyt9CvUxun4xgHzRqXiKI8/Zn1GkzgcLsw\nqGoZMAtYAuwA3lDVbSJyq4jc6mr2W6A98Gy1y1LjgBUisglYA3ygqh+5m8ld877KpqC4jDvHWW8h\n0HVu25LpQ7vyZtp+ck6ccTqOMV4h/jgRempqqqalNc0tD6eKSxn1189J7daOud8b2iTvYfzLofwi\nRj+6jOtTE/jztQOcjmNMo4nIuvrcFmB3Plfz8sq95BeVcpedWzAu8W1ackNqAm+m7eegnWswAcAK\nQxWFZ8t4YXkWY5Nj7Uokc57bxvYGYPYyu6/BNH9WGKp4eeVeTpwp5c7x1lsw5+vctiXXD0ng9bX7\nOZxv9zWY5s0Kg8uZkjKeX57F5UmxDO7azuk4xgfdNqY3Fap2N7Rp9qwwuMxftZfjhSXcZb0FU4su\n0a247uLOvLpmn90NbZo1KwxUjqA658ssLkuMYUg36y2Y2t0+tjflFcq/vrD5GkzzZYUBeGX1Xo6e\nLrFzC6ZO3dpHcM2gzryyei+5p6zXYJqngC8MxaXlPPdFFiN7tbcRVE29zBrXm9LyCp7/0noNpnkK\n+MLw6up9HD191s4tmHrrERPBtEGd+c+qvRw9bfM1mOYnoAtDZW8hk+E9ohnes73TcYwfmTWuNyVl\nFTy/3HoNpvkJ6MKwYM0+ck+dtbucTYP1im3N1QM78Z+Ve21uaNPsBGxhKC4tZ/YXmQzrHs0l1lsw\njXDHuN4UlZZbr8E0OwFbGN5M28+Rgsregog4Hcf4od4dIrlyQDwvf53NCes1mGYkIAvD2bJynl2W\nyZBu7RjZy3oLpvHuGJdIYUk5c1fscTqKMR4TkIXhzbQcDuUXc9d46y0Y9yR3jGTqgI689HU2+WdK\nnY5jjEcEXGEoKatg9rJMBndty2WJMU7HMc3AHeMSOX22jLlfWa/BNA8eKQwiMllEdolIhog8UMN2\nEZEnXds3i8jF9d3X095an8OBk0XWWzAe0zc+iiv6xfHvr/aQX2S9BuP/3C4MIhIMPANMAVKAGSKS\nUq3ZFCDR9ZgJzG7Avh5TWl7BM59nMDChDaOTYpvqbUwAumNcIqeKy3jpq2ynoxjjNk/0GIYBGaqa\npaolwAJgWrU204CXtdIqoK2IxNdzX495e30OOSeK7Eok43H9O7dhQt845q7IoqDYeg3G88rKK/jP\nymwKz5Y1+Xt5ojB0BvZXeZ7jWlefNvXZFwARmSkiaSKSlpeX16igeafOktqtHWOTOzRqf2Mu5K7x\niRQUlzHPeg2mCby/+SC/eW8bKzKONvl7+c3JZ1Wdo6qpqpoaG9u4w0CzxiXy+k8usd6CaRIDEtow\nvk8HXlixh9Ne+KvOBI7yCuXpzzLo0zGSiX3jmvz9PFEYDgBdqjxPcK2rT5v67OtRwUFWFEzTuXN8\nIvlFpcz7OtvpKKYZ+WDLITLzCrlzfCJBXvgO80RhWAskikgPEQkDpgMLq7VZCHzXdXXSCCBfVQ/V\nc19j/MbALm0ZkxzLC8uzvHIs2DR/FRXKU0vTSYprzeR+Hb3ynm4XBlUtA2YBS4AdwBuquk1EbhWR\nW13NFgNZQAbwPHDbhfZ1N5MxTrpzfCInzpTyn1V7nY5imoEPtx4mPfc0d4zzTm8BQFTVK2/kSamp\nqZqWluZ0DGNqdfPc1Ww/WMDyX4ylVViI03GMn6qoUKY+uZzS8go+/tlotw+Fi8g6VU2tq53fnHw2\nxp/cPSGRY4UlzLdeg3HDx9sPs/PwKe4Yl+jV86NWGIxpAkO6RTOqdwxzvsyiqKTc6TjGD6kqTyzN\noGdMBFcP7OTV97bCYEwTuWtCIkdPl/DKaus1mIb7ZPsRdhwq4Paxvb1+NaUVBmOayFDXJFD/+jKL\n4lLrNZj6U1We/Cydbu1bMW2Qd3sLYIXBmCZ114RE8k6d5dXV+5yOYvzIZztz2XqgsrcQEuz9r2kr\nDMY0oRE92zO8RzTPfZFpvQZTL6rKk0vT6RLdkmsH1zhCUJOzwmBME7trQiK5p87y+tr9dTc2AW/Z\n7jw25eRz+5jehDrQWwArDMY0uUt6tmdo93bMXpbJ2TLrNZjaqSpPfJpO57Ytue7iBMdyWGEwpomJ\nCHeNT+JwQTFvWK/BXMDy9KNs3H+S28b2IizEua9nKwzGeMGlvdszpFs7nrVeg6lF5X0L6XRq04Lr\nhzjXWwArDMZ4RWWvIZFD+cX8d12O03GMD/o68xjr9p7gp2N6ER4S7GgWKwzGeMlliTEM6tKWZz/P\npKSswuk4xsc8sTSdjlEtuHFol7obNzErDMZ4iYhw14REDpws4u311msw/9/KzGOs2XOcW0f3dLy3\nAFYYjPGqMUmxDExow9OfZ1Babr0GU+nJpel0iAxn+rCuTkcBrDAY41Uiwp3jE8k5UcQ765t0skLj\nJ1ZmHmNl1jF+MroXLUKd7y2AFQZjvG5cnw4M6Gy9BlN5JdLjn+wmLiqc7wz3jd4CuFkYRCRaRD4R\nkXTXz3Y1tOkiIp+LyHYR2SYid1XZ9rCIHBCRja7HVHfyGOMPzvUa9h0/w3sbDzodxzjoq4xjrMk+\nzu1je/tMbwHc7zE8ACxV1URgqet5dWXAvaqaAowAbheRlCrbH1fVQa7HYjfzGOMXJvTtQEp8FE9/\nlk6Z9RoCkqryj0920alNC77lA1ciVeVuYZgGzHMtzwOuqd5AVQ+p6nrX8ikq53Z2ZmQoY3zEuSuU\nso+d4Z0Ndq4hEC3bncf6fSeZNS7RJ65EqsrdwhCnqodcy4eBuAs1FpHuwGBgdZXVd4jIZhF5saZD\nUVX2nSkiaSKSlpeX52ZsY5w3KSWO/p2jePKzdLuvIcCcO7eQ0K6l43c516TOwiAin4rI1hoe06q2\nU1UF9AKv0xp4C7hbVQtcq2cDPYFBwCHgsdr2V9U5qpqqqqmxsbF1fzJjfJyIcO+kZPYfL+LNdTaG\nUiBZuiOXzTn53Dku0dExkWoTUlcDVZ1Q2zYROSIi8ap6SETigdxa2oVSWRReUdW3q7z2kSptngcW\nNSS8Mf5uTFIsQ7q146mlGXzz4gSfOgFpmkZFhfKPT3bTrX0rrrvYN4+qu1uqFgK3uJZvAd6r3kBE\nBJgL7FDVf1TbFl/l6bXAVjfzGONXKnsNlSOv2ixvgeHj7YfZfqiAu8YnOjI7W324m+oRYKKIpAMT\nXM8RkU4icu4Ko0uBm4FxNVyW+qiIbBGRzcBY4Gdu5jHG74zsFcPIXu15dlkGZ0rKnI5jmlBFhfL4\nJ+n0jI3gGwO9P5dzfdV5KOlCVPUYML6G9QeBqa7lFYDUsv/N7ry/Mc3FvZOS+Obslcz7ei8/HdPL\n6TimiXyw5RC7jpziiemDfLa3AHbnszE+YUi3aMYmx/LcF5kUFJc6Hcc0gfIK5Z+f7iYprjVXXeS7\nvQWwwmCMz7hnYjL5RaW8uGKP01FME3h3wwEy8wq5a3wSwUE1HkTxGVYYjPERAxLaMLlfR+Yu38OJ\nwhKn4xgPOltWzuOf7qZ/5yim9O/odJw6WWEwxofcMymJ0yVlzFme5XQU40Gvrd5Hzoki7r+iD0E+\n3lsAKwzG+JSkuEimDezES19lk3uq2Ok4xgMKz5bx1GcZjOgZzWWJMU7HqRcrDMb4mLsmJFFSXsGz\nn2c6HcV4wIsr9nCssIT7J/eh8rYu32eFwRgf0yMmghtTu/DK6r3sO3bG6TjGDccLS5jzZRaTUuK4\nuGutQ8H5HCsMxviguyckEhwk/P3jXU5HMW6YvSyD0yVl3HdFstNRGsQKgzE+KC6qBT8a1ZOFmw6y\nJSff6TimEQ7lFzFv5V6uG5xAUlyk03EaxAqDMT5q5uietGsVyl8/2ul0FNMIT3yaDlrZ+/M3VhiM\n8VFRLUKZNS6RFRlHWZ5uc5D4k8y807yRtp9vD+9Kl+hWTsdpMCsMxviwm0Z0JaFdSx75cCcVFbVO\nd2J8zN8+2kWL0GBmjevtdJRGscJgjA8LDwnmvknJbDtYwPubDzodx9TD2uzjfLTtMLeO7kVM63Cn\n4zSKFQbmS3ofAAAROUlEQVRjfNw3BnYiJT6Kvy3ZxdmycqfjmAuoqFD++MEO4qLC+fFlPZ2O02hW\nGIzxcUFBwgNT+pBzoohXVtlkPr5s0ZZDbNp/kvsmJdMyzH9n43OrMIhItIh8IiLprp813sEhItmu\nCXk2ikhaQ/c3JtBdlhjDpb3b8+Rn6eSfsWG5fVFxaTl//XAnKfFRXHdxgtNx3OJuj+EBYKmqJgJL\nXc9rM1ZVB6lqaiP3NyZgiQi/mppCflEpTyxNdzqOqcG8r7M5cLKIX13Z1+eH1a6Lu4VhGjDPtTwP\nuMbL+xsTMFI6RTF9aBdeXplNZt5pp+OYKo4XlvD05xmM69OBS3v7x0B5F+JuYYhT1UOu5cNAXC3t\nFPhURNaJyMxG7I+IzBSRNBFJy8uza7pNYLpnYjItQoP58wc7nI5iqnhyaTpnSsr55dQ+TkfxiDoL\ng4h8KiJba3hMq9pOVZXKAlCTUao6CJgC3C4il1dvUMf+qOocVU1V1dTY2Ni6YhvTLMVGhjNrXG+W\n7szly932B5IvyMo7zfxVe5k+tAu9O/jX0Be1qbMwqOoEVe1fw+M94IiIxAO4fubW8hoHXD9zgXeA\nYa5N9drfGPP/ff/S7nSNbsUfP9hOWXmF03ECmqry+0XbaREazN0TkpyO4zHuHkpaCNziWr4FeK96\nAxGJEJHIc8vAJGBrffc3xpwvPCSYX07ty+4jp3ltjV2+6qSlO3JZtiuPuyckEhvpnzez1cTdwvAI\nMFFE0oEJrueISCcRWexqEwesEJFNwBrgA1X96EL7G2Mu7Ip+cYzoGc0/Ptltl686pLi0nN8v2k7v\nDq25ZWR3p+N4VIg7O6vqMWB8DesPAlNdy1nAwIbsb4y5MBHhN1elcNVTK/jn0t08dHU/pyMFnBeW\nZ7Hv+Bnm/3A4ocHN617h5vVpjAkg/Tq1Ycawrry8ci87DhU4HSegHDxZxDOfZzK5X0dG+ck8zg1h\nhcEYP/bzScm0aRnKr9/daqOvetGfFu+gQpVfXdnX6ShNwgqDMX6sXUQYD0zpw7q9J/jv+hyn4wSE\nrzOP8sHmQ/x0TC+/nGuhPqwwGOPnrr84gdRu7fjL4h2cKCxxOk6zVlJWwcMLt9G5bUtuHd3L6ThN\nxgqDMX4uKEj4wzX9KSgu49Elu5yO06w9vzyL3UdO8/tp/WgR6r+jp9bFCoMxzUDf+Ci+P7I7C9bu\nY8O+E07HaZayjxbyxNJ0pg7oyPi+tY7e0yxYYTCmmbh7YhIdIsP59btb7Y5oD1NVfv3uVsKDgwLi\n0mArDMY0E63DQ3jo6n5sO1jA3BV7nI7TrLy78QArMo5y/+Rk4qJaOB2nyVlhMKYZmdK/I5NS4vjH\nJ7vJsqG5PeJEYQl/WLSDQV3a8p3h3ZyO4xVWGIxpRkSEP17Tn/CQIB54a4vd2+ABf/xgBwVFpfzl\nugEE+fkEPPVlhcGYZqZDVAt+fVUKa7KP88rqvU7H8WtLdxzhrfU53Dq6F33jo5yO4zVWGIxphm4Y\nksBliTE88uFODpwscjqOXzp5poQH395Cn46R3DG+t9NxvMoKgzHNkIjw52sHoMCDb2+hch4s0xC/\ne387xwtL+PsNAwkPab73LNTECoMxzVSX6FY8MKUPX+7OY/5qm7ehIZZsO8w7Gw5w+9je9O/cxuk4\nXmeFwZhm7OYR3RidFMufPthORq5dpVQfxwtL+NU7W0iJj+L2sYF1COkcKwzGNGMiwt+uv4iWocHc\n/foGSsrsxrcLUVV++fYW8otKeezGgYSFBOZXpFufWkSiReQTEUl3/WxXQ5tkEdlY5VEgIne7tj0s\nIgeqbJvqTh5jzP/VIaoFf7nuIrYeKOCJpbudjuPTXl2zj4+2HebnVyQH1FVI1blbDh8AlqpqIrDU\n9fw8qrpLVQep6iBgCHAGeKdKk8fPbVfVxdX3N8a4b3L/jnwrtQvPLstkzZ7jTsfxSbuPnOL372/n\nssQYfjSqp9NxHOVuYZgGzHMtzwOuqaP9eCBTVe3iamO87LdXp9A1uhV3L9hgw3NXU1xazp2vbSCy\nRQiP3TgwYG5kq427hSFOVQ+5lg8DdQ05OB14rdq6O0Rks4i8WNOhqHNEZKaIpIlIWl5enhuRjQlM\nEeEhPDVjMEdPl/CzNzbaXdFV/OmDHew8fIq/3zCQDpHNfyykutRZGETkUxHZWsNjWtV2WnmhdK3/\n0kQkDPgG8GaV1bOBnsAg4BDwWG37q+ocVU1V1dTY2Ni6YhtjanBRQlt+c1Vflu3KY/YXmU7H8Qnv\nbjjAf1bt5ceX9WBMcgen4/iEkLoaqOqE2raJyBERiVfVQyISD+Re4KWmAOtV9UiV1/7fsog8Dyyq\nX2xjTGPdNKIba7NP8NjHuxjctS0jezW/yezra/vBAh54ezPDekRz/+Q+TsfxGe4eSloI3OJavgV4\n7wJtZ1DtMJKrmJxzLbDVzTzGmDqICH+5bgA9YiK487UNHM4vdjqSI/LPlHLr/HW0aRnK098eTGhw\nYF6aWhN3fxOPABNFJB2Y4HqOiHQSkf9dYSQiEcBE4O1q+z8qIltEZDMwFviZm3mMMfUQER7C7JuG\nUFRSzo9fTqOopNzpSF5VUaHc9foGDuUX8ex3hth5hWrcKgyqekxVx6tqoqpOUNXjrvUHVXVqlXaF\nqtpeVfOr7X+zqg5Q1YtU9RtVTmQbY5pYUlwkT84YzNaD+dz7ZmCdjP7rkp0s25XHb69KYUi3Wq95\nCVjWdzImgI3vG8cvp/Rl8ZbD/HNputNxvOLV1fv41xdZ3DSiKzeNCIyJdxqqzpPPxpjm7UeX9SA9\n9xRPLk2nZ0wE1wzu7HSkJvPl7jx+895WRifF8vDV/RAJ7PsVamOFwZgAVznr2wD2HjvDfW9uom2r\n0GZ52ea2g/nc9sp6Eju05ulvDybETjbXyn4zxhjCQoJ4/pZUkuIi+en89azbe8LpSB6VkXua785d\nQ1SLEF783lAiW4Q6HcmnWWEwxgAQ1SKUeT8YRlxUOD94aS27Dp9yOpJH5Jw4w81zVyMC8380nE5t\nWzodyedZYTDG/E9sZDj/+eFwWoQGMeP5Vew4VOB0JLccyi/iphdWU3i2jJd/MJyesa2djuQXrDAY\nY87TJboVC2ZeQlhwEN9+fhXbDubXvZMP2nfsDDc8t5Jjp0v49/eHkdIpcIfRbigrDMaY/6NHTASv\n/2QErcJC+Pbzq9mwz7/OOWTknuaGf33N6bNlvPLj4XavQgNZYTDG1Khb+wgWzBxBm5ahzHh+FR9v\nO+x0pHpZt/c4N/5rJeUVsGDmCC5KaOt0JL9jhcEYU6su0a14+7aRJHeM4ifz1/HSV3ucjnRB7244\nwIw5q4lqEcIbPxlBn452+KgxrDAYYy4opnU4C348gol943j4/e088NZmikt9a2yl8grlsY93cffr\nGxnctS3v3HapnWh2gxUGY0ydWoYFM/umIdw+thcL1u7nume/JvtoodOxAMgtKObmuat56rMMbkxN\n4D8/HE67iDCnY/k1KwzGmHoJDhJ+fkUf/v29oRzML+Kqp1bwxtr9VM7R5YyPtx1m6pPLWb/vBI9e\nfxF//eZFhIXY15q77DdojGmQsX06sOiOUaR0iuL+tzbz3RfXsP/4Ga9myDt1lttfXc/M/6wjpnU4\nC2eN4sbULjb2kYeIk9W+sVJTUzUtLc3pGMYEtIoK5ZU1+3hk8Q5KK5TvX9qd28f2JqoJh5soKinn\nxa/2MHtZJiVlFdw5vjc/Gd3LJtmpJxFZp6qpdbazwmCMcceh/CL+vmQ3b2/IoU3LUG65pDvfvaQb\n7VuHe+w9ThWXsmDNfuau2MPhgmImpsTxwJQ+9LITzA3ilcIgIjcADwN9gWGqWuO3tYhMBp4AgoEX\nVPXcTG/RwOtAdyAbuFFV67yTxgqDMb5n64F8nliazifbjxAeEsSVF8Vz7eDOjOwVQ3BQww/xVFQo\n6/adYOHGg7y74QCnzpYxvEc090xMYnjP9k3wCZo/bxWGvkAF8C/gvpoKg4gEA7upnNozB1gLzFDV\n7SLyKHBcVR8RkQeAdqr6i7re1wqDMb4rI/cUc1dks2jzQU4VlxEdEcYlvdpzSc/2pHSKoldsa9q0\n/L+Hm06fLSP7aCHbDxawKusYX2ce43BBMeEhQUzu35EfjuphN6u5yauHkkRkGbUXhkuAh1X1Ctfz\nBwFU9S8isgsYo6qHRCQeWKaqyXW9nxUGY3xfcWk5n+3M5dMdR/g6o/JL/pzW4SG0Dg+hZVgwxaXl\nFJ4to6C47H/boyPCGNEzmokpcUxM6UjrcJs6xhPqWxi88dvuDOyv8jwHGO5ajqsyz/NhIK62FxGR\nmcBMgK5duzZBTGOMJ7UIDWbqgHimDohHVdl3/AzpR06TkXea3IKznD5bSlFpBS1CgmgVFkxcmxb0\njImgd4fW9IptbVcYOajOwiAinwIda9j0K1V9z1NBVFVFpNbui6rOAeZAZY/BU+9rjGl6IkK39hF0\nax/BhNr//jM+os7CoKoT3HyPA0CXKs8TXOsAjohIfJVDSbluvpcxxhg3eePi37VAooj0EJEwYDqw\n0LVtIXCLa/kWwGM9EGOMMY3jVmEQkWtFJAe4BPhARJa41ncSkcUAqloGzAKWADuAN1R1m+slHgEm\nikg6MMH13BhjjIPsBjdjjAkQ9b0qye4jN8YYcx4rDMYYY85jhcEYY8x5rDAYY4w5j1+efBaRPGBv\nI3ePAY56MI4T/P0z+Ht+8P/PYPmd58Rn6KaqsXU18svC4A4RSavPWXlf5u+fwd/zg/9/BsvvPF/+\nDHYoyRhjzHmsMBhjjDlPIBaGOU4H8AB//wz+nh/8/zNYfuf57GcIuHMMxhhjLiwQewzGGGMuwAqD\nMcaY8wRUYRCRySKyS0QyXHNM+xUReVFEckVkq9NZGkNEuojI5yKyXUS2ichdTmdqCBFpISJrRGST\nK//vnM7UGCISLCIbRGSR01kaQ0SyRWSLiGwUEb8cTVNE2orIf0Vkp4jscE2B7DMC5hyDiAQDu4GJ\nVE4vuhaYoarbHQ3WACJyOXAaeFlV+zudp6FckzHFq+p6EYkE1gHX+Mt/A6mcazJCVU+LSCiwArhL\nVVc5HK1BROQeIBWIUtWrnM7TUCKSDaSqqt/e4CYi84DlqvqCa56aVqp60ulc5wRSj2EYkKGqWapa\nAiwApjmcqUFU9UvguNM5GktVD6nqetfyKSrn5+jsbKr600qnXU9DXQ+/+stKRBKAK4EXnM4SqESk\nDXA5MBdAVUt8qShAYBWGzsD+Ks9z8KMvpeZGRLoDg4HVziZpGNdhmI1UTkP7iar6VX7gn8D9QIXT\nQdygwKcisk5EZjodphF6AHnAv12H9F4QkQinQ1UVSIXB+AgRaQ28BdytqgVO52kIVS1X1UFUzl0+\nTET85pCeiFwF5KrqOqezuGmU67/BFOB21yFWfxICXAzMVtXBQCHgU+c8A6kwHAC6VHme4FpnvMh1\nbP4t4BVVfdvpPI3l6vp/Dkx2OksDXAp8w3WMfgEwTkTmOxup4VT1gOtnLvAOlYeJ/UkOkFOlt/lf\nKguFzwikwrAWSBSRHq6TPdOBhQ5nCiiuk7dzgR2q+g+n8zSUiMSKSFvXcksqL2TY6Wyq+lPVB1U1\nQVW7U/nv/zNVvcnhWA0iIhGuCxdwHX6ZBPjVVXqqehjYLyLJrlXjAZ+6ACPE6QDeoqplIjILWAIE\nAy+q6jaHYzWIiLwGjAFiRCQHeEhV5zqbqkEuBW4GtriO0wP8UlUXO5ipIeKBea4r3IKAN1TVLy/5\n9GNxwDuVf2MQAryqqh85G6lR7gBecf2RmgV83+E85wmYy1WNMcbUTyAdSjLGGFMPVhiMMcacxwqD\nMcaY81hhMMYYcx4rDMYYY85jhcEYY8x5rDAYY4w5z/8Dk+oDIbCmH/8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x81f4828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Plot a curve. Y as a function of X\n",
    "X = np.arange(0,2*np.pi,0.01)\n",
    "Y = sin(X)\n",
    "plt.plot(X,Y) # line plot\n",
    "plt.title('sin(x)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEICAYAAABWJCMKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3Xd0FNXbwPHvTa8EEiCUBAg9EGogFBERGyJFQJGqqIgN\nu/4UC4pdXxVFQVRElCYISFeahN4CJNRAKIEkQBqk9937/jGLhpK+2dlN7uecPWx2yn0mG+aZuXOL\nkFKiKIqiVD92egegKIqi6EMlAEVRlGpKJQBFUZRqSiUARVGUakolAEVRlGpKJQBFUZRqSiUAxSyE\nEO8IIWbqHYel2fJx23LsinkI1Q9AuZ4QIqPQj25ALmAw/fyklHK+5aOyLCHEncAsKWWTStr/dqAL\nUFDo49ullPsqqbxKPR7FNjnoHYBifaSUHlffCyGigfFSyo1FrS+EcJBSFhS1vKoyw3E/JaWcY654\nFKWsVBWQUmZCiA+FEIuEEAuFEOnAGNNncwqt86gQ4rwQIkkI8aYQIlYI0ce0zEUIMU0IcVEIESeE\n+EoI4WRadqcQIloI8T8hRKIQ4oIQ4uFC+60phJhnWhYthJgkhBCmZeOFEFtM+04RQpwSQnQTQjwu\nhIgRQsQLIcYU2peLqeyry2aYPvMCVgGNhBAZplfdUh53byHEbiFEqmm/Y8v4u20uhJDXfbZdCDHu\numOcajrGM0KIuwut6yOEmGP63V4RQiwt4XgKxz5ECHHUtN9/hBCtCi2LFUK8LIQ4bDq2hUIIZ9Oy\nukKItabtLgshtpblmBX9qASglNcQYAHgBSwqvEAI0Q6YBowAGgJ1gHqFVpmMVv3RHugE3AJMKrTc\nD3AFGgBPAd8LIWqYls1Aq5ZqCvQFHgceLrRtT2Af4AMsARYDHYDmwKPAdCGEm2nd/wMCTHG0AJoA\nb0kpU4GBwHkppYfplVCK4w4A1gJfmcrvBBy++a+vQnqa9usDTAV+LrRsAeAEtAHqAt+UcDxXYw8E\n5gLPoX1fG4GVQgjHQqsNB+5C+90HA1eT22vAGf77nt8236EqlUklAKW8tkspV0kpjVLK7OuWPQgs\nl1LulFLmcuMJYTTwnpQy0XQiep//TiYAOcCHUsp8KeVKtGcQLU0no+HAG1LKdCnlGbQTYOFtT0kp\n50opDWgn6EbAFCllrpRyrWmdpkIIO+AJ4EUp5RUpZRrwCVrSKu9xjwH+klIullIWSCmTpJThxexr\nhumqOUUIsbeEcgs7LaWcbTrGXwE/IURtIYQ/cAfwtOmY8qWUpb0aHwGslFL+I6XMBz5FS3LdCq3z\ntZTykpQyGVgNdDR9no+WrBtJKfPKUKaiM5UAlPKKKWZZg8LLpZSZwJXrlp8r9PM5tDuFq5JMJ7er\nsgAPtCta+xK2jS/0PhswmE5YhT/zQLtSdQYirp6E0U5qdYs5Lij+uP2B0yVsX9gzUsqapldIGba7\nVOh9lulfD1P5SaYr/rK65juRUhqBWK793V5f7tVnRZ+att0khDgthHitHOUrOlAJQCmv4pqPXUSr\nxgFACOEO1Cq0/ALQuNDPjYC4UpSZgNYaqTzbXi8eyANaFToJe0kpvUzLizq+4o47BmhWjlgKywQo\nVE0F11afFScGqF2ouqywkpr7XfOdmO6Q/CjF71ZKmSalfMnUwuh+4HUhxG2ljFnRkUoASmX4A7hf\nCNHd9HD3/euWLwQmm6ot6gDvAPNK2qmpamIJ8LEQwsNU5/5Saba9yb4MwCzgayFEHaHxK/RANR7t\nZOpZht3OA/oJIYYJIRxMx9ehjKFdMr3GCCHshRATuDbhFUlKGYNWdz/d9LDcUQjR27S4pONZDAwS\nQvQxVbW9BqQDe0oqVwgxUAjRzPQwPhUtSRtLE7OiL5UAFLOTUh5COzH/gXZlmWx65ZpWmQJEAEeA\nQ2gnmU9Kuftn0K7co4EtaHXgv5Uz1FfQqi72op241qM9DEZKeQRYCkSbqohKqhpCSnkW7WHr68Bl\n4ADQriwBSa1jzhPAm0AS2sPrEk/ChVxt5XQS7aT/nGm/xR6PlPIo8AjwPZAI9AMGmZJuSVoB/wAZ\nwA60B8/byhCzohPVEUypdKYqiRSgsekqVVEUK6DuAJRKIYQYJIRwE0J4AF8CB9TJX1Gsi1kSgBBi\nthAiQQhxpIjlfUydR8JNr8nmKFexakPQqn9i0drXj9Q1GkVRbmCWKiDTg6YM4DcpZdBNlvcBXpVS\nDqhwYYqiKIpZmOUOwNTx47I59qUoiqJYhiUHg+sphDiE1q74VVOrgxuYmr1NAHB1dQ329/cvV2FG\noxE7O9t7xJFpzCS5IJl69vVwsneySJnphnSuGK7g5+SHnRkfC9nqd3CVrccPNx5DgSzgQv4FvB28\n8bDzKGZL88k2ZpNYkEh9x/o4XjOyRMmq4ndQ2U6ePJkkpaxTqpWllGZ5odXzHiliWQ3Aw/S+PxBV\nmn0GBwfL8tq8eXO5t9XTB7s+kN3md5Ob/tlksTLnH5svg+YEycvZl826X1v9Dq6y9filvPEY1p5Z\nK4PmBMmjSUctFsPG6I0yaE6QjEyOLPO2VfE7qGxAmCzledsiaUlqPQUzTO/XAo5CiNqWKNvWRCRG\n0K52O+yEbV/1KNbpUOIhnO2daVGrhd6hKFbAImcZIUQ9Uy9BhBAhpnKTi9+q+snKz+LklZN0qFPW\nzqOKUjqHkw7TxqcNjnZlq4pRqiazPAMQQiwE+qB1NY8F3gUcAaSUM4EHgKeFEAVog3GNMN2qKIUc\nTT6KURppX6c9xlTVk14xr3xDPseTjzOidUkDnirVhVkSgJSy2DbeUsrvgO/MUVZVdijxEADta7cn\n/FRxowgrStmdvHKSPGMe7eqUaXQKpQpTFc1WJCIxgsY1GlPTpabeoShV0KGk/y4wFAVUArAaUkoO\nJR5S/zmVSnM48TA+Lj7Ud6+vdyiKlVAJwEpcyLxAck6yegCsVJqDCQfpUKcDpvYYiqISgLWISIgA\noH0ddQegmF9CVgKxGbF09u2sdyiKFVEJwEocSDiAm4Obap+tVIoD8QcA6OLbRedIFGuiEoCVCLsU\nRiffTjjYWXJ0DqW6CIsPw83BjVberfQORbEiKgFYgcs5lzmdelpdnSmV5kDCATrW7aguMJRrqARg\nBdTtuVKZUnNTiboSRee6qv5fuZZKAFYgLD4MF3sX2vq01TsUpQo6mHAQgGDfYJ0jUayNSgBWIOxS\nGB3qdsDRXo3PopjfgfgDONo5qh7Ayg1UAtBZam4qJ6+cVNU/SqXZe2kv7Wq3w9neWe9QFCujEoDO\nDsQfQCJVAlAqRaYhk2PJx+jeoLveoShWSCUAne2+uBtXB1fVAUypFFE5UUgkPer30DsUxQqpBKCz\nnRd2EuwbbLHpH5Xq5UTOCdwd3WlbWzUwUG6kEoCOLmZcJDotWl2dKZXmRM4Juvp2VRPAKDelEoCO\ndl3cBUDPBj11jkSpiuIy4kgsSFT1/0qRVALQ0a4Lu6jrWpdmNZvpHYpSBe25uAeAbvW66RyJYq1U\nAtCJURrZfXE33Rt0V8PzKpViR9wOatjXUBcYSpFUAtDJ8cvHSclNoUcDVf+vmF++MZ+dF3bS1rWt\nusBQiqQSgE62xmxFINQDYKVShCeEk5GfQVtX1fpHKZpKADoJjQ2lfZ32+Lj66B2KUgVtidmCo50j\nrVzU8M9K0VQC0EFCVgLHko/Rx7+P3qEoVdSW2C10rdcVFzsXvUNRrJhKADrYErsFgNv8btM5EqUq\nOp92nui0aHr79dY7FMXKqQSggy0xW2jo0ZDmNZvrHYpSBW2N3QpA74YqASjFUwnAwrILstl9cTe3\n+d2mWmcolWJL7BYCvALwr+GvdyiKlVMJwML2XNxDriGX2/xV9Y9iful56YTFh6mrf6VUVAKwsNCY\nUNwd3enq21XvUJQqaEvsFgqMBdzZ+E69Q1FsgFkSgBBithAiQQhxpIjlQggxTQhxSghxSAhRLScn\nLTAW8M/5f7i14a1q9i+lUmyI3kBd17pqeHGlVMx1BzAH6FfM8nuBFqbXBOB7M5VrU/Zd2seV3Cvc\n0+QevUNRqqDM/Ey2x23nzsZ3YifUzb1SMrP8lUgptwKXi1llMPCb1OwGagoh6puj7JtKisI5JwEy\nkyE/p9KKKav159bj6uBKr4a99A4FgJx8AxdTs0nJygcgMT2XnHyDzlEp5bUtdht5xjzubnK33qEo\n5WE0QNZluHwGkk9bpEgHi5QCDYGYQj/Hmj67WCml/dCbHvlZsNv0s6s3eAdArQBo0BH8QqB+B3C0\nXCeZAmMBG89tpI9fH1wcLNs5JzUrnz1nkzkcl8rJ+HSiEjK4mJJDtulk71jrJC714J6vtyIN7tR0\nc8TX04XmdT1o27AGbRt4Edy4Fh7OlvpzUcpj/bn11HatTcc6HfUORSmOoQAuHYJzOyHhGCRFQXIU\nZF/5bx33uvBaVKWHYnX/o4UQE9CqifD19SU0NLTM+6jT4lkKslJwcwB7QzYuOYm4ZMXjlrQVlyNL\nADAKR67Uak9S7W4k+4SQ51zLnIdxg8jsSFJyU2iY2bBUx5SRkVGuY78qLsPI3osFhCcaOJ9mRAJ2\nAuq6CRp62NGigcDDyREPR8FZ7DkoYXhLR/ILHEnJlVzOzmLvqQzWHNZytL2AlrXsaFfHnpB6DtR2\nLfnmsaLHoDdbij/XmMuW2C10c+/Gtq3b/v3cGo7hSJb2aHBf2D4uOpXtms8a4q+ojIwMtm9YjU/y\nXuok7qRmymEcDFrNRK5TLbLcGpJdM4RcX28KHNwpcHDHYO9KkgWO21IJIA4o3CjZz/TZDaSUPwI/\nAnTp0kX26dOnHMX1ITQ0lA432zYjAWL2YnduBz4n1uJzcgaImdD8LggeBy3uBnvz/1pCd4bidtmN\nJ+9+slR3AKGhoZT12NNy8lkSFsvisBgiL2UiBHRpXIshIbXp0dSHDv41cXG0v2G7BccTObgX3nio\nD7Vcrk2EqVn5HI5LZdupRLacSGTxiXQWn8ine1NvhnX2Y2CHBjfdZ3mPwZrYUvzro9eTF5PHuJ7j\nCKkf8u/n1nAMhnMGCIWuXbrSyrtsYxNZQ/zlJiWc3ULi2u+oczkMjPlQww86jYYmt0CjnjjXqI8z\nULmXn0WzVAJYCUwUQvwOdANSpZSVU/1TEo+6EDhAe93zMSQchyNL4OB8+H0k1GgIPZ+Hzg+Dk5tZ\nisw35rPp/CZu87+tUqp/Yi5nMWvbGZbsjyUzz0BH/5q8N7AN/dvVp26NipXn5eZIrxa16dWiNpPu\nDSTmchbLD8ax9EAsry05xKd/RfJIzyaM7d6YWu5qXmO9rD27Fh8XHzr7VssGdtalIA8iFsCu6ZB0\nkpoOnhAyAYKGQsNgsKIOoGZJAEKIhUAfoLYQIhZ4F3AEkFLOBNYC/YFTQBbwqDnKrTAhwLcN+E6G\nPpPg5DrYPQP+fh22/h/c8jyEPFnhZwW7LuwiJTfF7K1/LqXm8N3mKBbti0EgGNChPuN6NqG9X02z\nllOYv7cbz93Rgol9m7PrdDI/bTvDVxtO8n3oaZ64NYAnejfF00U1cbWk1NxUtsRuYWTrkTjYWV2t\nbvVhyIfwBbD1C0g9Dw06wf0z2ZXsQ+87rPPBvFn+WqSUI0tYLoFnzVFWpbF3/O/O4NxOLQFsmAz7\nZsFd70Ob+8uduVedXoWXs5fZemdm5xn4PvQUM7eewWiUPNTVn4l9m1Pfy9Us+y8NIQQ9m9emZ/Pa\nnIxP55tNUUz75xTz9pznub7NGdu9scViqe7WRa+jwFjAgKYD9A6l+jr9D/z1OiSd1K7yB0yF5neA\nEBit+BmGuly4mcY9YeyfcGYLrHsT/hgHjW+Bgd9A7RZl2lVaXhr/nP+HoS2GmqXz17qjl3h/1THi\nUrIZ1KEBr93TCn9v81RVlVdLX0+mj+rMk71T+PSvSKasOsYfYbE80NhAH10jqx5Wn1lNM69mBHoH\n6h1K9ZMSA3+/AZGrwbspjFgIre61qmqe4qjeIsVpehs8uVU78ccfhe9vgW1fabd6pbQhegN5xjwG\nNx9coVCuZObx7IIDPDl3Px7ODvw+oTvTRnbS/eRfWHu/mswf340ZozuTnJnLB7tzeG/lUbLzVN+C\nyhKTHsPBhIMMaDZADS5oSVJC2C8wowec3gx3vAvP7IbW/W3m5A/qDqBkdvZa66CW98LaV2HTFDi+\nEh6YrWX8Eqw8vZIArwDa+pR/ar5/IuN5felhUrLyeO2eVkzo3RRHe+vM3UII+rerz60tavPi7H+Y\nszOarVGJfPNQJ9r5eekdXpWz+sxq7fmPqv6xnLQLsPwZOLMZAnrDoO+glm1WeVrnWcQaefrCQ3Ph\nwTlaT72ZveHwkmI3iUmP4UDCAQY1G1Suq7PcAgPvLD/CY3PC8HF3YsWzvXj29uZWe/IvzNPFkbFt\nnJk/vhtZuQaGzNjBjNBTGI1S79CqDCklq0+vpku9LtRzr6d3ONXDqY0wsxfE7IX7voKHV9rsyR9U\nAii7tkPgqe1a66Glj8PK54ocbmLV6VXlvjpLzjYy/IfdzN19jvG9Algx8RbaNKhR0egt7pbmtfn7\nxVu5p209Pv/7BE/O209aTumr0JSihcWHcT79PEOaD9E7lKrPUACbPoB5w8DDFyaEQtfHbaq652ZU\nAiiPmo1g3Fq49RU48BvM6Q/pl65ZxWA0sCxqGT0a9Cjz1dn2qCTe3ZnN6YQMZo7pzNsD2uDscPPO\nVragppsT343qxOQBbfgnMoH7v9vByfh0vcOyeUtOLsHTyZO7Gt+ldyhVW/YVmDcUtn0BncbC+E1Q\np6XeUZmFSgDlZe8Ad0yGh+ZDQiT8eDvEHfh38fa47cRnxfNgywfLtNu5u8/x8Ow9eDkLVky8hX5B\nlTdmniUJIXisVwALxncjLaeA+6fvYOOxeL3DslkpOSlsOLeBAU0HWHxsqWol+TTMulNrGj54Ogz+\nzmwdRK2BSgAVFTgAHl8Pdg7wy71w9E9Auzqr7Vq71DN/GY2Sj9Yc453lR+jTqi7vdHelWR2Pyoxc\nF92a+rDm+V40q+PBhLlhzN0VrXdINmnVmVXkG/MZ1mKY3qFUXWe3wU99tTuAR1ZBpzF6R2R2KgGY\nQ70gmLAZ6neEPx7l0o6v2Bq3lSHNh+BoV3Lb/5x8A88uOMBP287ycI/G/Dg2GBcH265bLI5vDRcW\nPdmd21vV5Z0VR/lk7XH1cLgMpJQsPbmUdrXblXlsHaWUDi+Bufdr9f3jN0HjHnpHVClUAjAX99rw\n8HJodS9/7puKlEaGluLhXHpOPg/P3svfRy/xzoA2TBnUFgcbaOVTUW5ODvwwNpgx3Rvxw9YzvLQ4\nnHyDUe+wbEJ4YjinU0+rq//Ksm8WLB0P/t1g/AZtKPkqquqfaSzJ0ZWCB35hqXddemZl47f1KzAW\nfVK7kpnHmFl7OHDuCt+M6MTjvQKqVWceB3s7PhgcxGv3tGJF+AUmLjhAboHqNFaShccX4unoyb0B\n9+odStUipTYEzJpXoGU/GLMUXKp23xWVAMxsU9wW4mUewxveDmGzYdVzN00CCek5jPhxN8cvpTNz\nTDCDOjTQIVr9CSF49vbmTB7QhnVH45nw237Vc7gY8ZnxbDi3gSEthuDmWHUeRupOStj4HvzzIbQf\nofX5cbTc2Fp6UQnAzOYdm4e/pz+33fc93PY6HJx3QxK4evKPuZLFL+O6cmcbXx0jtg6P9Qrg06Ht\n2BqVyKNz9qokUIRFJxZhkAZGti52/EWlLKTUevjv+Bq6PA73f68NDlkNqARgRkeSjhCeGM7owNHY\n2zvA7W/CbW9oSWCllgSSM3IZ/dMeLqXm8OtjIdzSvLbeYVuNESGNmDq8I3vPXmbC3DA1P/F1cg25\nLDm5hD7+ffDz9NM7nKpBSu2qf/tUCH4U+n8BdtXntKjGAjKjecfn4e7ozuBmhQZ+u32S9u+WT8k1\nGHk4ZgTnL2fxy6Nd6drEW59Ardj9nRqSbzDy2pJDPDv/AN+PCcbJofr8hyzO2jNruZJ7hdGBo/UO\npeoI/UTr4NX5YW1oh2p08geVAMwmISuBdWfXMaL1CDycrmu/f/skcvPzcd75JcMMqTQb+zU9m6kr\n/6I82MWfnAIj7yw/wkuLwvlmRMdq0TKqOFJKFkQuoHnN5oTUCyl5A6Vk26fCls+09v0Dvql2J39Q\nCcBsrtbNjgocdcOynHwDD5++k/sMR3jMfg0kBEOrV3SI0naM7d6Y3HwDH645jrOjHV880AE7u+rT\nQup6YfFhRF6OZHKPydWqpVilOfCb9tA36AEY+G21PPmDegZgFln5WSw+sZg+/n3w9/S/ZpnBKHnx\n93D2nrtCrWFfQbvhsOl9rYWQUqzxtzblpTtbsuxAHJ+vO6F3OLr6+cjPeLt4M7DpQL1DsX2Ra2DV\nC9DsDu2BbzU9+YO6AzCLpVFLSclN4bGgx675XErJeyuP8vfRS0we0IZBHf2g3QzISYXVL4NrLW10\nUaVIz9/RnMSMHGZuOU29Gs6Mu6XqdsopSuTlSHbE7eCFzi+ocX8q6txOWPKYNl/v8N/AwUnviHRV\nfVOfmeQb8vn16K908e1Cx7odr1k2ffMp5u4+x5O3NeWxXqYTl72jNqdAo+6w7Ek4v9vyQdsQIQRT\nBgVxdxtfpqw+xl+HL+odksXNPjwbd0d3hrcarncoti3hOCwYoY3mO+oPcK56Y22VlUoAFbT6zGri\ns+IZ3278NZ8v2R/LF+tPMrRTQ16/p/W1Gzm5wYgF4OUHv4/SJphRimRvJ5g2shOdG9XihUXh7D17\nWe+QLCYmLYZ159YxvNVwajjZ3nwQViMjAeYP1zp3jVkG7j56R2QVVAKoAIPRwOwjswn0DqRng57/\nfr737GUmLTtEr+a1+eyB9jd/eOnmDaP/AGnU/jCzqs9JrTxcHO2Z9XAX/Gq5MmFuGNFJmXqHZBFz\njs7BXtgzNnCs3qHYrvxs7UIrKwlG/Q41/UvepppQCaACNp7fSHRaNI+3e/zflhnnk7N4cm4Y/t5u\nTB/VufjpG32aaXcCKedg0VgoyLNQ5LaplrsTv4zrCsD438Kq/MxiiVmJLD+1nMHNB1PHrY7e4dgm\no1Gbvzc2DIb+qNX9K/9SCaCcjNLIrMOzaFKjCXc2uhOAtJx8Hv91H0YJPz/SFS+3UnQnb9xTm1T6\n3HZY9bzWM1EpUmMfd74fHUx0UibPLThIQRUeQfTnIz9jkAYea/tYySsrNxf6CRxdBne+B4GqBdX1\nVAIop03nNxF5OZIJ7Sdgb2dPgcHIcwsOcjYpk+/HdCagtnvpd9bhIW3IiIiFsOu7ygu6iujRzIcP\n7g9iy8lEPl4bqXc4leJS5iX+OPEHg5sPxr+GqrIol0OLYevnWkevW17QOxqrpJqBloPBaGBG+AwC\nvALoH9AfgI/XRrLlZCIfDQkqXy/fPm9AwjHYMBl8gwDV2ac4I0MacTI+ndk7ztLC14ORIY30Dsms\nZh2ehVEamdB+gt6h2KYL4dr4W417wX1TbX7y9sqi7gDKYV30Ok6lnOKZDs9gb2fPnwdjmb3jLON6\nNmF0t8bl26kQWqeU2q1gyaO4ZKv5ckvyVv9AereswzvLjxAWXXUeol/MuMiyqGUMaTGEhh4N9Q7H\n9mQma8/U3Hy0JtfVvK1/cVQCKKMCYwHfR3xPi1otuLvJ3Ry7kMakZYcJCfDmrfsCK7ZzZw8YMR+k\nkaAjH0Ne9WjpUl4O9nZ8O7ITfrVceWb+ARLSc/QOySx+OvwTEskT7Z7QOxTbYyiAJY9CRrw2pr+H\nenheHLMkACFEPyHECSHEKSHEGzdZ3kcIkSqECDe9JpujXD2sObOG6LRonu3wLOnZBp6atx8vV0e+\nG9Wp+BY/peXTDIbNxj3zHKx4Vj0ULoGXqyPfjwkmLSefiQsO2vy0knEZcfwZ9SfDWgyjvkd9vcOx\nPZumwNktcN+X0DBY72isXoXPWEIIe2A6cC/QBhgphGhzk1W3SSk7ml7vV7RcPeQb85kZMZNA70D6\n+N3Oi4sOcjE1mxmjO1PX04xd9FvcyZmmY+Hon7DzW/Ptt4oKrF+DT4a2Y+/Zy3z+t20/FJ4RPgM7\nYXdDx0KlFI4sg53ToMtj0Fn1mygNc9wBhACnpJRnpJR5wO/A4BK2sUlLTi4hNiOWiZ0m8u3mU2w+\nkcjkAW0Ibmz+cf1j/IdC4CBtxEI1XESJhnTy45Eejflp21nWHLLN4SJOXD7BqtOrGBU4inru9fQO\nx7bEH4MVE8EvBPp9pnc0NsMcrYAaAjGFfo4Fut1kvZ5CiENAHPCqlPLozXYmhJgATADw9fUlNDS0\nXEFlZGSUe9ubyTZmMy1uGi2cW3BgZy7fHIjilgYO+OWcJTQ02mzlXJWRmck2nxF0cd6L3bxRhHWZ\nSr5T5UxQHZUWBcCOHTvwsDff+Cjm/g5KcqunZEdNO15edIDU86408KjY9Y2l458RPwNXO1cCUwPN\nVq6lj+FmjmQdAWBf2D4uOpUtOZcmfvuCbIL3v4IDjoT5P0Xe9p3lDbVSWMN3UCQpZYVewAPArEI/\njwW+u26dGoCH6X1/IKo0+w4ODpbltXnz5nJvezNf7/9aBs0JkqFn98sOU9bJfl9vldl5BWYto7B/\n478QLuX7daT87X4pDYZKKWv+sfkyaE6QvJx92az7Nfd3UBoXU7Jl8AfrZd8vNsuMnPwK7cuS8e+I\n2yGD5gTJOUfmmHW/enwH19sYvVEGzQmSkcmRZd62xPiNRimXTpDyXS8pT4eWL8BKZunvAAiTpTx/\nm6MKKA4o3FPFz/RZ4SSTJqXMML1fCzgKIWxmSqxLmZeYe2wu/Zv059u/csgvMDJjdGdcHO0rv/D6\nHeDez+D0P7Dty8ovz8bV83Jh2shOnEnKZPKKm95kWh2jNDJ1/1QaejRUk72XVfh8OPS71o+m6W16\nR2NzzJEA9gEthBABQggnYASwsvAKQoh6wjRYjhAixFRushnKtohvD36LlBKXzAGEnbvCx0Pbla2n\nb0UFj9Mmkgn9GM5ssVy5Nqpns9o817cFSw/EsuxArN7hlGjNmTVEXo7k+U7P42Sv2qyXWsJxWPMq\nBPSG3q/NQj0iAAAgAElEQVTpHY1NqnACkFIWABOBdcBxYLGU8qgQ4ikhxFOm1R4AjgghIoBpwAjT\nrYrVi7wcyarTq+hdbwi/bUvjoS7+DO5o4c45QsCAqeDTApaOh/RLli3fBj3ftzkhAd68vfwIpxMz\n9A6nSLmGXKYdnEYbnzb0C+indzi2Iy8TFj+i9Z0ZOgvsLHA3XgWZpR+AlHKtlLKllLKZlPIj02cz\npZQzTe+/k1K2lVJ2kFJ2l1Ja11OaIkgp+TLsSzycarB1b3ua1/HgvUFt9QnG2QOG/wq56fDnU9oo\nh0qRHOztmDaiE84Odjw7/wA5+Qa9Q7qpucfmcinzEq8Ev4KdUP0yS23ta5B0Eob+BJ6+ekdjs9Rf\nXDE2x2xm98XdeGb3IzPHkemjO+PqpOOVRt1A6PcJnNmsBo0rhXpeLnw5vAORl9L5aM1xvcO5QXxm\nPD8e+pG+/n0JqR+idzi2I3yhVvff+zVodrve0dg0lQCKkFOQw+f7PqeWgz8nTrZnyqC2tPT11Dss\n7XlA4EBtYvkLB/WOxur1be3L+F4BzN19zuqmk/xq/1cYjAZe66rqr0st+TSseUUb5K3PDYMOKGWk\nEkAR5hydQ1xGHBfO3MPgjv4M72IlQ/IKAQOngXsdWPI45Fpv/ba1+F+/1nTw8+J/Sw8RczlL73AA\n2B+/n7Vn1/Jo0KP4efrpHY5tMOTDsie0ebWH/qjq/c1AJYCbuJBxgZ8Oz8IxpyP1nNrx4f1B/874\nZRXcvLX/AJfPwN+v6x2N1XNysOPbkZ2REl5ZHIHBqG/7A4PRwCd7PqGeez0eb/e4rrHYlC2fQdx+\nGPgNeKlRUs1BJYCb+CLsCwoMktTYfnz9UEc8XUoxs5elBdwKt74CB+dpY6AoxWrk48aUQW3ZG32Z\nH7ae1jWWJSeXcOLKCV7t8iquDq66xmIzzu3U+sF0HANt79c7mipDJYDr7Lqwiw3nNpCd0Idnbu1C\nlybmH+fHbPq8AQ27wKoXIeW83tFYvaGdG3Jfu/p8tf4kh2NTdYkhJSeFb8O/JaReCHc3vluXGGxO\ndgosmwA1G8O9n+odTZWiEkAh+YZ8Ptj1MeT70Mp1AM/f0ULvkIpn7wjDZoE0wtIntLHQlSIJIfho\nSBA+Hk68sOgg2XmWbxr6Xfh3ZORl8HrI69ZVrWjN1r4KaRe0v3VnK2iIUYWoBFDI3GNzicmIxpA4\niG8e6mqe8f0rm3cADPgKYnbDjq/1jsbq1XRz4ssHO3ImMZOP11q2aeihxEMsPrGYka1H0rJWS4uW\nbavqxofC4T+gzyTw66J3OFWODZzhLCMuI45vD04nPz2Qt/s+QNM65hsVs9K1exDaDoHQT+BihN7R\nWL1eLWr/2zR0c2SCRcrMN+YzZdcU6rjVYWKniRYp0+ZdiablyR+gUQ+49WW9o6mSVAJA6/E7act7\n5Bugq+fjjAyxkiafpSUE3PcVuNXW6krzq8bUiJXp1Xta0bqeJ68tiSApI7fSy5t/bD4nr5zkzZA3\ncXe04DhStspogGVPau+H/KCafFYSlQCA1af/4mDSbhxS+/P1sD62WTfr5g2Dp0NiJPzzgd7RWD0X\nR3u+HtGRtOwC3lh6iMocmiouI44ZETPo49+Hvo36Vlo5VcrOaRCzm6gWT0KtxnpHU2VV+wSQlpfG\n+zs/wZDdkC/veQYfD2e9Qyq/FndCl8dh13Q4u03vaKxe63o1+F+/Vmw8nsDCvTElb1AOUko+2v0R\nAG+GvGmbFxeWFn8MNn8MgQOJ91VDPFemap8AJm3+lGxjKnfUeZY7AqvAJNx3f6A9GF7+DOSk6R2N\n1XvslgBuae7Dh2uOcT7Z/L2EN5zbwLa4bUzsOFFN8l4ahnxY/pTW2ue+qVr1plJpqnUC2BUXxtZL\nq3DJvo3PBt6rdzjm4eQOQ36EtFj4W42VUhI7O8HnD3TAXgheXRKB0Yy9hNPz0vl076cEegcyKnCU\n2fZbpW37SmvIMGAqeNTRO5oqr9omgHxDPq/88w7GfC+m3v06bk7mmB7ZSvh3hV4vayMmHl+tdzRW\nr2FNV94Z2Ia9Zy/zy85os+33mwPfkJyTzLs93sXBrgr9fVWWixGw9XMIegDaDNY7mmqh2iaAyVum\nkW6Mpbf3k9zarAoOxnXb61CvPax6HjIs09TRlj0Y7Mcdrevy+d+RZplAZt+lfSw6sYhRrUfRtrZO\nc0jYkoJc+PNpcPOB/v+ndzTVRrVMAAcvHWP1+bk453Zm6sAqemvu4KRNlpGbASufB9uYgE03Qgg+\nGdoOVyd7XlkcQYGh/BPuZBdk8+7Od/Hz8OO5Ts+ZMcoqbMtnkHBUG+nWzYqHX6liql0CKDAWMHHD\nG0iDC1/1fc8yE7vrpW5ruGMynPwLIn7XOxqrV7eGCx8MDiI8JoUftp4p936mH5xOTHoMU3pOwc3R\nzYwRVlFx+2H7VOg4GlqpaTEtqdolgHc2f0ea8Sy3+TxJ7+YBeodT+bo/rfWk/Ot1bTwVpVgDOzTg\nvnb1+XrjSWLSy34XEJEYwdzjcxnecria5as08nO0qh/P+tpsd4pFVasEEH7pBKtj5uCS15GvBjys\ndziWYWevdRAz5KmqoFL64P4gvFwd+fFQLnkFpU8CeYY8Ju+YTF23urwU/FIlRliFbP4Qkk7AoG/B\nxUvvaKqdapMADEYDz65/HWlwZuod7+PsUIWrfq7n0wzumgKnNmjzByjF8nZ34pOh7YlJN/LdP1Gl\n3m5mxEzOpJ7h3R7v4uFkQ2NJ6eX8btj5HQQ/Cs3v0DuaaqnaJIC3N88gTZ7m9joT6NW0GlT9XK/r\nE9o8quvehNRYvaOxene18eWWBg5MDz1NRExKiesfTz7O7COzGdRsEL0a9rJAhDYuLxOWPw01/bXO\ni4ouqkUCCL94ktXnZ+Oa354v+4/TOxx92NnB4O+0QbZWPqeqgkphVKATdT2deeWPCHLyi547IN+Y\nz+Sdk6nlUov/df2fBSO0YRunaFOaDp6hxvjXUZVPAFrVzySkdGTaXR/iVJ2qfq7nHaBVBZ3+Bw78\nqnc0Vs/dUfDZsPacSsjgy/UnilzvlyO/EHk5kre7v42Xs6rHLtHZrbD3B+j2lDa1qaKbKp8A3to0\nkzRO0rfueLo3roZVP9fr8jgE9IZ1b6lpJEuhd8s6jO7WiFnbz7Iv+vINy0+nnGZmxEz6NenHHY1U\nPXaJctNhxbPg3RTueFfvaKq9Kp0Awi+eYk3sbNwK2vJlv8f1Dsc62NnBoO+09yueBWP5OzxVF2/2\nD8Svliuv/hFBVt5/024ajAYm75iMh6MHk7pN0jFCG7L+bUiJgfu/ByfVR0JvVTYBGIwGnlk3CSkF\n39z1EY7VuernerUaw90farfi+2frHY3Vc3d24P8e6MC55Cw++yvy389/O/Ybh5IOManbJLxdVO/V\nEsXshv1zoOdEaNRd72gUzJQAhBD9hBAnhBCnhBA3DEEpNNNMyw8JITqbo9zi/Ba7nXQRyZ2+4+ne\nqFllF2d7gsdB09th/WTIStY7GqvXvakPj97ShF93nWPnqSTOpJ7hu4PfcWejO+nXRPVeLZXNn0Lt\nVnD723pHophUOAEIIeyB6cC9QBtgpBCizXWr3Qu0ML0mAN9XtNziHLxwhv2GlbgbAvmy34TKLMp2\nCaG1CrKzh0N/6B2NTfjfPa0JqO3Oq0vCeXPb27g6uvJW97fUJC+llZUMQ74HRxe9I1FMzHEHEAKc\nklKekVLmAb8D14/lOhj4TWp2AzWFEJUyO4bRaOTZdVp97LS7PsLevsrWclWclx/c8zFcPq13JDbB\n1cmeLx7sQJL9Bo4mH+bNkDep7Vpb77CsX9x+7d9OY6FhsL6x2IDFh7fz7qZfMVrg+Zw5BilvCBSe\nTy8W6FaKdRoCF6/fmRBiAtpdAr6+voSGhpYpmMTcTLILMmhr14+s03GEno4r0/bWIiMjo8zHXi7S\nj0z3RkAG+zcuw8HDfNVlFjuGSnKz+C/lX8Klznry0toSFeZI6PnQm25rLfT+Dhzy08g/9APUdiNM\ntONiGWPRO35zKMsxZBbk8mH0Z0iRT/BGX2o4Ve7dktXNUiGl/BH4EaBLly6yT58+Zd7H4IJ72Lp1\nC+XZ1lqEhoZaLP4LNR+GQzMITvidWv3XaS2FzMCSx1AZro/fYDTwyN+PUMPZHacro1l4yo5xA2/B\ny9VRvyBLoPt3sOQxNhlzADe6hPSglXerMm2ue/xmUJZjeHDxm0jHRJ5v838M6lr5z5bM8T89DvAv\n9LOf6bOyrmM2Tg4OONipVj+ldnUQrtgw2DNT31is2Lzj84hIjGBSt0l8/eCtJGbkMmXVUb3Dsl5H\nl8ORpdB2iN6R2ITFh7dzPGs1jRxv5wkLnPzBPAlgH9BCCBEghHACRgArr1tnJfCwqTVQdyBVSnlD\n9Y+is2Z3wKYpkFT6AdCqi7OpZ/n24Lf09e9L/4D+tPeryTN9mrHsQBwbjsXrHZ71yUiENS9D/Y4Q\nOFDvaKxeSnYWH++dgp3Ri9mD3rdYuRVOAFLKAmAisA44DiyWUh4VQjwlhHjKtNpa4AxwCvgJeKai\n5SqV4J6PwMFFG6TLWPTYN9WNwWjgnR3v4GzvzDs93vm31c9zfVsQWL8Gk5Yd5kpmns5RWhEpYc1L\nWq/fITNBzYdcoidWfozB4RIT203C16Omxco1S2WvlHKtlLKllLKZlPIj02czpZQzTe+llPJZ0/J2\nUsowc5SrmJlHXbjvS4jdBzun6R2N1Shc9VO41Y+Tgx1fPtiB1Ow8Jq9UVUH/OrwEjq+C29+CuoF6\nR2P1Fkbs4Hj2Sho73c6Erv0tWrZqI6lcK2gYBA6CzR9D/DG9o9FddGo03x78lj7+fbgv4L4blrdp\nUIPn+7ZgVcQF1h5WtZqkXYS1r4BfCPRU8yGX5EpWFp+GvYe90YvZA6dYvHyVAJRrCQEDpoJzDfjz\nSTDk6x2RbozS+G/Vz+Tuk4vs8PV0n2a09/Pi7eVHSMrItXCUVkRKWPU8FOSZqn5UQ4ySPL7iY4wO\nl3i+wyTqetSyePkqASg3cq8NA7+BS4dg6xd6R6Ob0PRQwhPDeSPkDeq41SlyPQd7rSooI7eAt/48\njKyucy0cnAtR67Uhx33U8CslmXdwOydzVxHg3IfHgy1b9XOVSgDKzQUOgPYPwdb/gwsH9Y7G4s6l\nnWN1ymr6+PVhQNMBJa7fwteTV+5qybqj8awIv2CBCK3MlXPw9yRocqs2+5xSrKTMTP5v//vYGz35\neaDlWv1cTyUApWj3fqY9GP7zacjP0TsaizFKI5N3TMZBODC5R9FVP9cbf2tTghvXYvKKI8SnVZ/f\nF0ajNrQ4AgZPN1tHwqps/PLPMDpe5KVOb1LH3fJVP1epb0opmmstbe6AxOMQ+rHe0VjMkpNLOJBw\ngKG1hhZb9XM9ezvBFw92IM9g5I2lh6pPVdC+nyB6m9aMuFZjvaOxevP27+ZU/koCXG5lXCd9qn6u\nUglAKV6LO6HzI7BjGpzfo3c0lS4+M56p+6fSrX43urlfP6RVyQJqu/N6v9ZsPpHIH2GxlRChlUk+\nDRveheZ3QeeH9Y7G6iVlZPP5/g+xl678dJ9+VT9XqQSglOyej6CmPyx/CvIy9Y6m0kgp+XDPhxQY\nC3i3+7vlHub5kR5N6N7Um/dXHyMuJdvMUVoRowH+fAocnGHQt1oLMqVYE5ZPQzqf45n2L+Prof9I\nsioBKCVz9tTqdi+fgY2Wb6tsKevPrSc0JpRnOz6Lfw3/kjcogp2d4P8e6ICUkteXVOGqoJ3fQuxe\n6P8F1KiU0d2rlPlh4ZzMX4yfc2cmdH5A73AAlQCU0groDd2egr0/wJktekdjdqm5qXy852Pa+LRh\nTJsxFd6fv7cbb94XyPZTSczbc94MEVqZ+GOw+SOt02A76ziZWbPE9Bw+C/sIOyGYee9HVjOJkEoA\nSund8S54N4MVEyEnTe9ozOqLsC9IzU1lSs8pOJhp7JpRIY24tUVtPll7nPPJWWbZp1Uw5GudBJ1r\naJ0GreRkZq2klExY9hPSNZJHA5+msZef3iH9SyUApfSc3OD+7yEtFta/pXc0ZrPrwi6Wn1rOuLbj\naO3d2mz7FULw2bD22AvBq0siMBqrSFXQ1i+0ToIDpmqdBpVibYxN4WTBPHydW/F813F6h3MNlQCU\nsmnUDXo+Dwd+gxN/6x1NheUU5PDB7g9oXKMxT3V4quQNyqhBTVcmD2zD3rOX+WVntNn3b3GxYVrn\nwPYPQZtBekdj9c4nZ7E85U/s7HOZfvcn2FvZ8BgqAShld/ub4Bukdf7JSNA7mgqZdXgWMekxvNP9\nHVwcKmf6vQeC/bijdV0+/zuSUwkZlVKGReRmwLInoEYD6P9/ekdj9QoMRp5csgi7GgcZ2eoRWnm3\n0DukG6gEoJSdgzMMm6WN975iojYImA2KTo1m9pHZ9A/oT7f6ZW/zX1pCCD4Z2g43J3teXHSQvILK\nn+y7UqybBJfPwpAf/ptFTinS9NBIzol5uOPNy12f1jucm1IJQCmfuoFw1/sQtQ7CftY7mjKTUvLR\nno9wtnfmta6vVXp5dWu48Omw9hyJS+OrDScrvTyzO75aq/br9SI0uUXvaKzeodgUvj/4C/bOCYyp\n82Cl3V1WlEoASvmFTIBmfWHd25BoWye1ddHr2H1xN891eu6aSV4q0z1t6zEypBE/bD3NztNJFinT\nLNIvwcrnoH4H6POm3tFYvew8A8/98Q+OPpvo3fB2gtyC9A6pSCoBKOVnZ6e1CnJ0hWXjtXHgbUBG\nXgaf7/ucQO9AHmr1kEXLfmdAIAE+7ry8KIKULBv4fUkJy5+B/GwYOgscnPSOyOp9vPY48Y6LcHaw\n550e1p0wVQJQKsaznjYMwMUImxkwbnr4dJKyk5jcY7LFW2W4OTnw9YiOJGXk8qYtzB2w9yc4vQnu\n/gDqtNQ7Gqu3+UQCC478haPnMZ7t9Az13OvpHVKxVAJQKi5wgDYQ2PavIXq73tEU68TlEyyIXMDw\nVsMJqq3PrXl7v5q8fHdL1h6+xJL9VjxgXEIkbHgHWtwNXcfrHY3VS8rI5bUlYXg2WEOAV1Oz9Civ\nbCoBKOZxzyfgHaANDpadonc0NyWl5LN9n1HDqQbPddJ3vtonezejW4A37608SnSSFQ6wV5CrVes5\neWjjQKnevsUyGiWvLI4g03UTBvtk3un+No52jnqHVSKVABTzcPaAoT9B2gVY/aJVNg3deH4j+y7t\nY2LHiXg569uM0d5OMPWhjtjbCV5cFE6+wcqahm6YDJcOa9V7HnX1jsbq/bz9LFvPnMK19hbuanwX\nXet11TukUlEJQDEfvy7Q9204+ifsn6N3NNfINeTyZdiXtKjVgmEth+kdDqD1Ev5oSDvCY1L4ZmOU\n3uH8J3IN7JmpDf7XWt8JS2xBREwKn6+LJKBFKEJIXunyit4hlZpKAIp53fKi1jT07zdwz4jWO5p/\nzT02l7iMOF7v+rrZBnszh4EdGvBAsB/TQ0+xPcoKmoamxGitfup30Pp5KMVKz8nnuYUH8a51kUR2\n8UjbR2jo0VDvsEpNJQDFvOzs/u0p2ubY/1nFBDKJWYn8eOhH+vr3rdQev+X1/uC2NK/jwYuLDpKg\n51zChgJYOl6b6OWBX7Qe30qRpJS8vfwIsVcy8G26jjqudRjfzrYelqsEoJifR10Y+iNuWXGwtvJ7\n2ZbkmwPfUGAs4NUur+odyk25OTkwfXRnMnILeOH3cAx6jRoa+jHE7IaBX4NPM31isCFL9seyIvwC\n9/W4xNn047zQ+QXcHN30DqtMVAJQKkfTPpxr/CCEz4eI33UL41jyMVacXsHYNmMrNMtXZWvp68n7\ng4PYdSaZaZt0eB5wejNs+wo6jVUTvJTCqYQMJq84SkhTd47m/E6QTxADmw3UO6wyq1ACEEJ4CyE2\nCCGiTP/WKmK9aCHEYSFEuBAirCJlKrbjXOMR0KgnrH4ZkvR5yPn1/q+p6VzTJm7NHwz2Y2jnhkz7\nJ4qdpyz4PCAjAZZNgDqt4N7PLVeujcrMLeDpeftxdbKnR6fjJGYn8GrXV7ETtnc9XdGI3wA2SSlb\nAJtMPxfldillRylllwqWqdgIaWevjRrq4AyLH4E8y86KtevCLnZd3MWE9hPwdPK0aNnlIYTgg8FB\nNK3tzvO/h5OQboHnAYYCWPIY5KZp9f5OtlWFYWlSSiYtO8zpxAw+eaApf0TN5Ta/2wj2DdY7tHKp\naAIYDPxqev8rcH8F96dUNV4Ntf4BCccs2j/AKI18feBrGrg3sPh4PxXh7uzAjNHBZOTm8+Lv4RRU\ndv+ATVMgehsM+Bp821RuWVXA3N3nWBlxgZfvasnhjD/JyM/ghc4v6B1WuVW0PZyvlPKi6f0lwLeI\n9SSwUQhhAH6QUv5Y1A6FEBOACQC+vr6EhoaWK7CMjIxyb2sNLBl/VJpWPbNjxw487D3Mtt//jsGB\nxk1GEHBoIVFZXsT53We2MoqyP3M/x5KPMdZnLDu37SzXPvT8GxrdyoGfjyTz/KyNDG9V/gHYijuG\n2ok7CTo6jbgG9xKVUh8q6ViPZB0BYF/YPi46XSxh7WtZ0//jUykGPtmTQ4c69tQzHOKjY/MJcQ8h\nLiKOOOKK3M6ajuEGUspiX8BG4MhNXoOBlOvWvVLEPhqa/q0LRAC9SypXSklwcLAsr82bN5d7W2tg\nyfjnH5svg+YEycvZl82632uOwWCQcv5wKad4S3lul1nLuV5eQZ68d+m9cuiKobLAUFDu/ej9N/Tm\nskOy8eur5eqIC+XeR5HHkHBCyo8aSPnTHVLm55Z7/6WxMXqjDJoTJCOTI8u8rd7fwVVJ6Tmy+8cb\n5S2fbpIpmXnyrW1vyc6/dZYX0kv+bix9DECYLMX5VUpZchWQlPJOKWXQTV4rgHghRH0A0783nR9Q\nShln+jcB+BMIKW/CUmzU1f4BXv7a84D0+EoramnUUmLSY3ih8wtWNwdrWbw7sC3BjWvx2pIITlxK\nN9+Oc9Nh0RhwcIEHf1VDPJfAYJS8uCic5Mw8Zo4JJjH3HKvOrGJk65HU96ivd3gVUtFnACuBR0zv\nHwFWXL+CEMJdCOF59T1wN9odhFLduNaEh+ZBTir8MQ4M+WYvIteQy0+Hf6JT3U7c2vBWs+/fkpwc\n7JgxujPuzg48OTeM1Gwz/L6k1KbxTI6CB3/RntEoxfp8XSTbopKYMqgtQQ29+D7ie1wdXG2iZVlJ\nKpoAPgXuEkJEAXeafkYI0UAIsda0ji+wXQgRAewF1kgp/65guYqtqhcEg6bB+Z2w/m2z735Z1DIS\nshJ4puMziCowgqVvDRe+H92Z2CvZvLQoHGNFO4ltnwrHlsOd70FAb3OEWKX9eTCWH7acYUz3RowM\naUTUlSg2nNvAqNajqOlSU+/wKqxCD4GllMnAHTf5/ALQ3/T+DNChIuUoVUz74XDhIOyeAXVaQ5dH\nzbLbXEMusw7PonPdznSrZ31DPpRXlybevDuoLe8sP8LXG0/y8t2tyrejyDWw6X0IGgY9nzdvkFVQ\neEwKry89TPem3rw7sC0APxz6ATdHNx5u87DO0ZmH7fVcUKqGuz6A5nfC2lfh7Faz7PLq1f/THZ+u\nElf/hY3p1oiHuvgz7Z9TrAgvusVJkS4dhqVPQINOanz/UohPy2HCb2HU9XRmxuhgHO3tOHXlFOuj\n11eZq39QCUDRi70DPDAbvJvBorGQfLpCu6uqV/9XCSH44P4gugV489ofh9h/7nLpN85IgIUjwcUL\nRi7U5nBWipSTb2DCb2Fk5BYw65EueLtrD8l/OPQDrg6uVebqH1QCUPTk4gWjfgdhBwseqtBMYlX5\n6v8qJwc7Zo4JpmEtVyb8tp/zySX3rBbGfK3FT2aSdvL3tO45avVmNEpeW3KIiNhUpj7Ukdb1agBw\nOuU066LXMSqw6lz9g0oAit68m2otg65El7tlUL4hn58P/1xlr/4Lq+XuxM+PdKHAKHns133FtwyS\nklYnvoOYPTBkJjToaLlAbdTn606wKuIC/+vXinva/pcsq+LVP6gEoFiDJrfAgKlwZjOsfL7Mw0Ws\nPbuW+Kx4xrcbX2Wv/gtrWseDmWOCOZecycQFB4qeTnLTFOrFh0Lfd6CtGqWlJHN3n2PmltOM7taI\np2/7bzjs2PRY1kWv46FWD1HL5abjXdoslQAU69B5LPSZBBELtJYqpWSURuYcnUPzms3p1bBXJQZo\nXXo08+GjIe3YFpXE/5YcurF56J4fYftU4hrcC7fazhSFetl4LJ53VxzhjtZ1mTKo7TUXEr8d+w07\nYcfowNE6Rlg5rGduPEW57XVIvwjbvwLP+tBtQombbI/bzqmUU3zc6+NqcfVf2PAu/iSk5fDF+pN4\nuzvx9n2B2u/g2Ar463/QegBRvo/RsJr9XsoqPCaF5xYepG0DL74d1QkH+/+ui6/kXOHPqD8Z0HQA\nvu5FDXVmu1QCUKyHEND/S63Vyl//02YWK6HqYvaR2dRzr0e/gH4WCtK6PHt7c5Iy8vh5+1lqezjz\ndJNLWnNP/xBtKO4de/QO0apFXkrjkdl7qe3pxM/juuDmdO0p8fcTv5NjyGFc23H6BFjJVBWQYl3s\nHWDYz9oJbNkTELWxyFUjEiPYH7+fsYFjcbRztGCQ1kMIweQBbRjUoQHr1q0mf+4DUKsxjPxdNfcs\nwdmkTMbM2ouLox3zH+9OXU+Xa5bnFOSw8PhCbvO7jWY1q+YUmSoBKNbHyQ1GLdJmqFo0Gs5suelq\nvx79FU8nT4a1HGbhAK2LnZ3gy1sF810+J67Ak3VdfgI3b73DsmoXUrIZM2sPRimZ93g3GvncOBHO\nilMruJJ7pcpe/YNKAIq1cq0FY1dArQBYOALO7bpm8cWMi2w6v4kHWz6Iu6O7TkFaiYTjOM4fgqtn\nTb7w/ZynV8Sx+tAFvaOyWonpuYyZtYe07Hx+eyyEFr43zhYnpWTe8XkE+QTZ7GxfpaESgGK93H3g\nkZOpZagAABy1SURBVJVQoyHMfxBi/5tOevHJxQA2NdtXpUiKgl8Hgb0Tdo+s5PPxA+jS2JsXfg9n\n7eGyTb5SHcSn5TDix11cTM1h9qNdCWroddP1dl/cTXRaNKMCR1XpxgUqASjWzaOulgTca8PcIXBu\nF7mGXJaeXEofvz408Gigd4T6uXQEfrkXkNrvyKcZbk4O/PJoVzr51+S5hQfZd6lA7yitxsXUbEb8\nuJtLqTn8+lgIXZsUXU22MHIh3i7e3NPkHgtGaHkqASjWr0YDGLcGPHxh7hDW7fmKK7lXGBk4Uu/I\n9BN3AObcB3aOMG6t9rzExN3ZgTmPhdDRvybfR+SydH+sjoFah9grWTz0w26S0nP57fEQQgKKPvnH\nZcSxJXYLw1oMw8m+ak+WY3PNQPPz84mNjSUnJ6fY9by8vDh+/LiFojK/4uJ3cXHBz88PR8dq1PLF\nqyE8+hdy7v0sOPIrTb0aVPlhH4p0bhcsGK5NsPPIKqjV5IZVPJwd+PWxEIZP28Arf0SQmp3PY70C\nLB+rFTgZn84js/eSmVvA3PHd6Ohf/Fg+i09o1YvDWw23RHi6srkEEBsbi6enJ02aNCm2bi49PR1P\nzxsf7tiKouKXUpKcnExsbCwBAdXsP7RHHQ4P/Jyjm57irbgoxP45ZptLwGYcWwHLJoCXHzy8stgZ\nvTycHXgp2IWlcTV4f/UxUrLyeOmullW6Tvt6+6Iv8/icfbg42vP7hB60aVCj2PVzCnJYFrWMvv59\nqede9QfOs7kqoJycHHx8fKrVH3FhQgh8fHxKvAOqqn4/uwZ3BzcG1ukKq1+EDe+CsYixcKoSKWHX\ndG0+5Xrt4LF1pZrO0dFO8N2oTv/OJfDakkPkFhgsELD+/j5ykdGz9lDb05mlT/cs8eQP8Hf036Tk\npjCydfWoXrS5OwCg2p78r6qux5+Wl8b6c//f3p3HVVXnjx9/fbiAoBjI5q6YKCJuiGsKqbS4MJV+\ny2q++U0zm6/ZaDNtls20WeM3s5qpySyzyZkcG1PTkuTnRpobooiKKLig7MgqoGyXz++PeyVNWS5c\nOPdcPs/Hg4eXe889533uxfP5nM/y/vw/HvB/gHbDF8KPL8DeD6HwAjzwKTi51L8TPTJWQdQrELMC\nAu+DaZ9ZNMnL0eDAkv8aSGcPFz7cnszFvCt8OiOkJs+9vZFSsvynsyyNOs2Q7h6senw4HRp4rt8m\nfUsv914M7zS8maO0Dbq7A1Bar8hzkZQby5naZ6ppxvCU900riyVsNI2GKUzVOkTrK82Ff001XfxH\nzYOHvmrUDF8hBM/e1Ze/PRrM0bRCHvj7XpKzi5shYG1drTDy+3/H8e7W00wZ2Jk1T45q8MX/XOE5\n4i/FM81/WqupZKkCwAqefPJJTp48We92H374IatXr65zm0ceeYTk5GRrhWZXNiRvoJ9nP/p79jc9\nIQSMmQ+PrIG8M7AiDM7u1DZIa0o/DCvuhIsHTcs4TnwHHJr2X/a+wV345qlRXKkwcv/f97Ixzn5G\nCKXmX+G/lu9jy/FMXprYj48eDcbV2dDg9288sxFH4UhE74hmjNK2qALAClauXEn//v3r3KaqqopV\nq1bx29/+ts7t5s6dy7vvvmvN8OzCqfxTJOYnMtV/6s21s35TYM4u02pX/5wGPy2Fah23c0sJh1bC\nqomm1dJmR0HwY1bbfXCPDvzw+7EM6OLOH76JZ+H6Y5RV6vjzAjbHZzD5r3tILbjCqseHM3dcb4tq\n8ZXVlWw+u5mwbmF4u3o3Y6S2RZd9ANe88X0CJzMu3/I1o9GIwdDw0v+a/l1u47XfBNX6emlpKdOn\nTyctLQ2j0cif/vQnli9fznvvvcewYcNwc3NjwYIF/PDDD7i6urJp0yY6duzIzp07GTp0KI6OjlRV\nVTF69GiWLl3KuHHjePnll3FwcODtt98mNDSUmTNnUlWlJvBcb0PyBpwdnJly+5Rbb+DtD09uh++f\nhV2L4cx2mLrctOKYnpTkwKZnIDkKeofDtM9NM6KtrJO7C2vmjOT9bUl8En2Wo6mFLJs+mKAut54Z\na6tKyqt4fXMC3x5OI7iHB397JJjunjfn9anP7rTd5JflM63PtGaI0napOwALbd26lS5duhAfH8+J\nEyeYOPHGNMSlpaWMGjWK+Ph4wsLC+PzzzwHYu3cvISGmnCKOjo784x//YO7cuWzfvp2tW7fy2muv\nAeDg4IC/vz/Hjx9v2ROzYeXGcrac20J4z3Dc29RxgXJuZ+ognfY55CTC8rFw6AuLVxjThJSmvoxP\nRsP5n2DSu/Df3zbLxf8aR4MDL07sx5ezhpNbUsH9H+/lw+1JVFTpY1TV3jO5TPnbHjYcSWP+BH/+\n87vRjbr4A3yX/B0+rj6M6TrGylHaNl3fAdRVU2+ueQADBw7kueee46WXXiIiIoLQ0NAbXnd2diYi\nwtSGGBISwrZt2wDIzMwkMDCwZrugoCBmzJhBREQE+/fvx9n5l44qX19fsrKyrB67Xv2U+hOXKy7z\ngH8DljUUAgZNh55jYNM82PJHOP4tTF4KnQY0f7CNUZACW56HM9ug82CYugJ8A+t9m7WMD/Bl2x/C\neP37BD7cnszWE1m8ef+AOmfLaim/tIK3tySy/kgafl5t+fecUYy8vfEFZd7VPPak7+HxoMdxdND1\nJdFiretsraBv374cOXKEyMhIXn31VcLDw2943cnJqabt0WAw1DTluLq63jR2//jx43h4eJCTk3PD\n82VlZbi42OmQxkaIPB+Jt6u3ZTN/3bvCjI0Q90/TXIEVYTBijmnVMVtJlVx2GfZ9ZPpxMMC9f4ER\nT5lGOLWwDu2c+esjwUwZ2JnXNycwfcV+7hvchZcn96Ozu22sK1BeZWT1vgt8vOsMpeVVPDPen2cm\n+OPiZHlT7/WiUqIwSiO/uf03VopUP1QBYKGMjAw8PT157LHH8PDwYOXKlQ16X2BgIGfOnKn5fcOG\nDeTn57N7924iIiKIiYnBw8M0RT0pKaneTuXWoqi8iN1pu3k44GEMDhb+RxcChv4P9IuAnW/BwRVw\ndA2MehpGPw0uGrV3V5bB4X/A7nfhSh4ETYN7FjdoYldzuyeoE6F9fFj+01lW/HSWqIQsfmteJN33\nNm0qJZXGar6Pz+D9bUmkFVwlrK8PiyYHEtDJOnf4kecj6duhL/4d/K2yPz1RBYCFjh8/zgsvvICD\ngwNOTk4sX76c559/vt73TZo0iRkzZgCQm5vLwoUL2bFjB927d+eZZ55hwYIFfPXVV2RnZ+Pq6krH\njva3/mhj7Li4g8rqyto7fxuirSdEfADD50D0O/DTEjj4KYTMhOFPgkd3q8Vbp9I80+iemM/gSi70\nCoO73oCuQ1vm+A3k6mzgj3f35aGQbny0M5nV+y+w5uBFHh7enf8Z3RN/35ZJsXKlooptFypZdCCa\n9MKrBHa+jX/OHkhoHx+rHSO1OJX4S/E8O/RZq+1TT5pUAAghHgJeBwKBEVLK2Fq2mwj8FTAAK6WU\nS5pyXC3de++93HvvjSlio6Ojax6XlJTUPH7wwQd58MEHAejZsydeXl4kJyfTp08fkpKSarabP39+\nzeM1a9bwu9/9rpmi15/Ic5H0vK0nQV619/c0WMf+8PC/ICMO9iyDfX8zNb/0mwyDHgb/u60/m9hY\nBeeiIf7fcOoHqCqDPvfCHb8Hv7GmuxQb1d2zLe8+OJh54/35eOcZ1saksnr/Be7o7cXDw7sTHtgR\ntzbWrUNKKTlysZB1san8cCyTkvIqhvXswJv3BzE+wBcHB+t+Xj+e/xGASb0mWXW/etHUb+8EMA1Y\nUdsGQggD8HfgbiANOCSE2CylrH/mlJ1ZsmQJmZmZ9OnTp9ZtPDw8mDFjBlevXm3ByGxTdmk2MVkx\n/O/g/7XuzMwuwaaCoPCiaZRQ3L8g8Xtocxv0nQi3jzPVzhurNA/O7YIzO0zDUUtzwMUDhvy3qR+i\nBTt4raGnVzuWPjSYlyb145tDqaw5eJEFa4/i7OjAnX19GB/gy4henvT2adeo76noaiWxKfnsPJXD\nrlM5ZBSV0dbZwOSBnQlwzGXO1Dua4axMhc2Wc1sY6ju01a4r0aQCQEqZCPXmphkBnJFSnjNvuxa4\nH2h1BUBAQAABAQF1bjNrVivLblmHrSlbkUgm95rcPAfw6AF3vwET/mQaenliAyT9CMdN6YBHtfGG\ntGDw7WdamrKtl+nHyRWqq0yTzcoK4XIGFGdCzinIOgZF5pQUrh3g9vEwYBr0uQcc2zTPebQQb7c2\nzBvvz9w7e3P4YgFbjmWy9UQW205mm193JrDzbfT2ceN2n3Z4tWuDu6sT54tLAYhPLSQjJ4esy2Vk\nFF4lJe8Kx9MKScm7AkBbZwNj/b35w919mTSwM25tHG+4u7a2pIIkzhWd49WRrzbbMWxdS/QBdAWu\nT9KSBrTSRO6KJbZd2EagZyB+7n7NeyCDI/iHm36qqyHnJKTsoehIJC7FWXB+NxjL696HcADP3tB9\nhKlfwW+s6U7D0o5rHXBwEAz382S4nyev/aY/53NLiTmfT0xKPsnZJayLTaW04peZxY7tE3HtBgs3\nHKO6PNe0DwGd3V0Z0PU2HhrWnSHdPRjm14E2ji33eUWlRGEQBu7xu6fFjmlr6i0AhBDbgVslxl4k\npdxk7YCEEE8BTwF07NjxphqAu7s7xcX1J7EyGo0N2s5W1Rd/WVmZ1WpHyZdNuYf27t2Lm8HNKvsE\nU39IY2Msqioi/lI8ER4RzVoLrF0gJT264+bmBtKIc0UhTpXFOFVexqG6AikMSOFAlWM7Kpw9qXB2\nB2G+eFUBZ0rgzB4N4r5RU74DS3QC7vMFfEHKNhSVS0oqobRSkljuxI4KmBXUhs7OLnRoI/BoIzA4\nCKAEKKEyLY39t0hL1Jzxb07fTO82vYk/EN8s+7+mpb6Dxqi3AJBS3tXEY6QD1w+z6GZ+rrbjfQZ8\nBjBs2DA5bty4G15PTExs0AQve10Q5hoXFxeCg4OtcqyMxAyIgTFjxtDBpYNV9gmmzvFff38NtfbU\nWkiHOXfO4XYPbdI5NCV+W2EL57DjAuyIhul3jSTAs+4m0F9rrvjPFZ4j+0I2s0NmM66f9fd/PVv4\nDmrTEqkgDgF9hBC9hBDOwCPA5hY4rqJj2y9up5d7L80u/op923FxBwATuk/QOBJtNakAEEJMFUKk\nAaOBLUKIKPPzXYQQkQBSyirgGSAKSAT+I6VMaFrY+hYXF8fs2bPr3GbFihWsWrWqhSKyLYVlhcRm\nxXJXj6befCrKrW2/uJ1B3oPo2K51z7dpUgEgpdwopewmpWwjpewopbzX/HyGlHLyddtFSin7Sil7\nSynfbmrQevfOO+/cMPb/VmbMmMFHH33UQhHZlui0aIzSSHjP8Po3VhQLZZZkcjLvJBN6tO7aP+h9\nJvCPCyHr1lkzXY1Vjcup0mkgTKp7ntrq1at57733EEIwaNAg3nrrLZ544glyc3Px8fHhyy+/pEeP\nHqxbt4433ngDg8GAu7s7u3fvpri4mGPHjjF48GAAFixYgJeXF3/+85+Jiori7bffJjo6mrZt2+Ln\n50dMTAwjRoyw/Dx0bMeFHXRu1/mXhV8UxYquNf+E91AVDH0XABpISEhg8eLF7Nu3D29vb/Lz83n8\n8cdrflatWsX8+fP57rvvePPNN4mKiqJr164UFhYCEBsby4ABv2Sl/Mtf/sLw4cMJDQ1l/vz5REZG\n4mBe9WnYsGHs2bOnVRUAVyqvsC9jH9MDpreaZfmUlrXj4g78Pfybf3ixDui7AKijpn61mUYB7dy5\nk4ceeghvb9OqQZ6enuzfv58NGzYApqabF198ETCNqpk5cybTp09n2jTTQhOZmZn4+PySy6Rt27Z8\n/vnnhIWF8cEHH9C7d++a13x9fTl16pTVz8GWHco6REV1BWHdmjATV1FqcbniMnE5ccwaoCZcgloQ\npll9+umnLF68mNTUVEJCQsjLy6s1LbSXlxcZGRk3PF9WVoarq22k4m0pe9L34OroSkjHEK1DUezQ\ngYwDGKWRsV3Hah2KTVAFgIUmTJjAunXryMvLAyA/P5877riDtWvXAvD111/XLBJz9uxZRo4cyZtv\nvomPjw+pqak3pYW+cOECy5YtIy4ujh9//JGDBw/WvJaUlHRDc5G9k1Lyc/rPjOw8EmeDc/1vUBQL\n/Zz+M+2d2jPYZ7DWodgEfTcBaSAoKIhFixZx5513YjAYCA4O5qOPPmLWrFksXbq0phMY4IUXXiA5\nORkpJeHh4QwePBghBEVFRRQXF+Pm5sbs2bN577336NKlC1988QUzZ87k0KFDgGlm7uuvv67h2bas\n85fPk16SzhMDntA6FMUOSSnZm76X0V1Gt7qVv2qjPoVGuNbhe72dO3fetN21foFfe+KJJ/jmm294\n8skn2b59e83zISEhNWsBx8bGEhQUhJdX860Ja2t+TvsZQN2eK80iqSCJnKs56u/rOqoJSANz586l\nTZu6M0Pm5eXx1ltvtVBEtmFP+h5ud7+91abmVZrXnnRTbiZVAPxCFQAacHFxqVkdrDYTJkzAz8+v\nZQKyAVcqr3A4+zChXUO1DkWxU3vS9tDPsx8+ba23opjeqQJAsQkHMw9SWV3J2G6qdqZY3+WKy8Rf\nile1/19RBYBiEw5kHsDF4MJQX9taH1exD4cyD2GURsZ0GaN1KDZFFQCKTYjJiiHYN1gN/1SaRUxW\nDK6Ormr456+oAkDRXO7VXM4UnmFE59aT8kJpWdcqGE4GJ61DsSmqALBQYWEhn3zySZP28eyzz7J7\n9+46t7nvvvsoKCho0nH0IjYrFoCRndRKoYr1XatgDO80XOtQbI4qACzU1AIgLy+PAwcOEBZWd66b\nhx9+uMkFjV4czDqIm5MbgV6BWoei2CFVwaidrieC/V/M/3Eq/9bJ0oxGIwaD5QtM9/Psx0sjXqr1\n9YULF3L27FmGDBnC+PHjOXbsGAUFBVRWVrJ48WLuv/9+UlJSmDRpEmPHjmXfvn107dqVTZs24erq\nyvr165k4cSIARUVFjBgxgs2bNxMQEMCjjz7KhAkTmDNnDpMnT2by5MksWrTI4nPQm5jMGIZ1HKZm\nZyrNIiYrRlUwaqHuACy0ZMkSevfuzdGjR1m6dCkbN27kyJEj7Nq1i+eeew4pJQDJycnMmzePhIQE\nPDw8WL9+PWBK7xASYkp05u7uzscff8zMmTNZu3YtBQUFzJkzB4AOHTpQXl5ek3PIXmWWZHKx+KJq\n/1eaTUxWDCEdQ1QF4xZ0/YnUVVNviUXhpZS88sor7N69GwcHB9LT08nOzgagV69eDBkyBDCleEhJ\nSQFuTgd99913s27dOubNm0d8fPwN+/f19SUjI8Ou00EcyjblPRrRSRUAivVll2Zz4fIFpvedrnUo\nNknXBYDWvv76ay5dusThw4dxcnLCz8+vJtXz9akeDAYDV69eBbgpHXR1dTWJiYm0bduWgoICunXr\nVvNaa0gHfST7CO2d2tOnQx+tQ1Hs0JGcIwCEdFLpxW9FNQFZqH379hQXFwOmNnxfX1+cnJzYtWsX\nFy5cqPf9v04H/cEHHxAYGMiaNWuYNWsWlZWVgOnuIisry+7TQRzNOcog30E4CPWnqFhfXE4cro6u\nBHQI0DoUm6T+11nIy8uLMWPGMGDAAI4ePUpsbCwDBw5k9erV9OvXr973T5kyhejoaABOnz7NypUr\nWbZsGaGhoYSFhbF48WIA4uLiGDVqFI6O9nuTVlRexNmis2r2r9JsjuYcZZD3INX+Xwv1qTTCmjVr\n6t3mxIkTNY+ff/75msehoaG8/PLLFBYWEhAQQGJiYs1r77//fs3jtWvX8vTTT1spYtsUf8nU5xHs\nG6xxJIo9Kq0s5XTBaeYMnKN1KDZL3QFoYNmyZVy8eLHObfr37094eHgLRaSNI9lHcBSODPBuPaue\nKS3n2KVjVMtqVcGog7oD0MDIkfVPSJk5c2bzB6KxuJw4Ar0CcXW0745uRRtHc44iEAzyGaR1KDZL\nl3cA18bat1b2cP6VxkoS8hIY4jtE61AUOxWXE0efDn1o79y8w8H1THcFgIuLC3l5eXZxEWwMKSV5\neXm4uLhoHUqTnMw/SbmxXN2eK83CWG3kWO4x9fdVD901AXXr1o20tDQuXbpU53ZlZWW6vkjWFb+L\ni8sN8wX06Pgl09rHKj2v0hzOFp2ltLJU/X3VQ3cFgJOTE7169ap3u+joaIKD9Vv66z3++pzIO4Gv\nqy++bX21DkWxQwm5CQBqgEE9mtQEJIR4SAiRIISoFkIMq2O7FCHEcSHEUSFEbFOOqdiHhNwE+nv3\n1zoMxU4l5CXg5uRGz9t6ah2KTWtqH8AJYBpQd3J7k/FSyiFSyloLCqV1uFxxmZTLKQzwUrUzpXmc\nyD1Bf6/+aoZ5PZr06UgpE6WUp60VjKItSct0rCfmmSa/qdvz1qWl/r4qjBWcLjhNkHdQixxPz4Q1\nRtMIIaKB56WUt2zeEUKcB4oAI7BCSvlZHft6CnjK/GsA0NgCxhvIbeR7bYHe4wf9n4Pe4wf9n4Pe\n44eWP4eeUkqf+jdrQCewEGI70OkWLy2SUm5qYEBjpZTpQghfYJsQ4pSU8pbNRubCodYCoqGEELF6\nbm7Se/yg/3PQe/yg/3PQe/xg2+dQbwEgpbyrqQeRUqab/80RQmwERtCwfgNFURSlmTR7D4kQop0Q\nov21x8A9mDqPFUVRFA01dRjoVCFEGjAa2CKEiDI/30UIEWnerCPwsxAiHogBtkgptzbluA3U5GYk\njek9ftD/Oeg9ftD/Oeg9frDhc7BKJ7CiKIqiP2qQrKIoSiulCgBFUZRWyu4KACHERCHEaSHEGSHE\nQq3jsZQQYpUQIkcIocuOciFEdyHELiHESXOakAVax2QpIYSLECJGCBFvPoc3tI6pMYQQBiFEnBDi\nB61jaQy9p5ARQngIIb4VQpwSQiQKIUZrHdOv2VUfgBDCACQBdwNpwCHgUSnlSU0Ds4AQIgwoAVZL\nKXU3VVYI0RnoLKU8Yh79dRh4QGffgQDaSSlLhBBOwM/AAinlAY1Ds4gQ4o/AMOA2KWWE1vFYSgiR\nAgyTUupyIpgQ4itgj5RypRDCGWgrpSzUOq7r2dsdwAjgjJTynJSyAlgL3K9xTBYxT5DL1zqOxpJS\nZkopj5gfFwOJQFdto7KMNCkx/+pk/tFVTUkI0Q2YAqzUOpbWSAjhDoQBXwBIKSts7eIP9lcAdAVS\nr/s9DZ1dfOyJEMIPCAYOahuJ5czNJ0eBHGCblFJv5/Ah8CJQrXUgTSCB7UKIw+YUMXrSC7gEfGlu\nhltpngdlU+ytAFBshBDCDVgPPCulvKx1PJaSUhqllEOAbsAIIYRumuOEEBFAjpTysNaxNNFY83cw\nCZhnbh7VC0dgKLBcShkMlAI21ydpbwVAOtD9ut+7mZ9TWpC53Xw98LWUcoPW8TSF+bZ9FzBR61gs\nMAa4z9yGvhaYIIT4l7YhWe76FDLAtRQyepEGpF135/gtpgLBpthbAXAI6COE6GXudHkE2KxxTK2K\nuQP1CyBRSvm+1vE0hhDCRwjhYX7simlQwSlto2o4KeXLUspuUko/TP8HdkopH9M4LIvoPYWMlDIL\nSBVCBJifCgdsbiCE7paErIuUskoI8QwQBRiAVVLKBI3DsogQ4t/AOMDbnGbjNSnlF9pGZZExwAzg\nuLkNHeAVKWVkHe+xNZ2Br8yjyhyA/0gpdTmUUsc6AhtN9QkcgTUtlELGmn4PfG2ujJ4DZmkcz03s\nahiooiiK0nD21gSkKIqiNJAqABRFUVopVQAoiqK0UqoAUBRFaaVUAaAoitJKqQJAURSllVIFgKIo\nSiv1/wFjdqawUMA80AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x8d5b9b0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot more than one curve and show legend\n",
    "plt.plot(X,sin(X),label='sin(x)')\n",
    "plt.plot(X,cos(X),label='cos(x)')\n",
    "plt.plot(X,tan(X),label='tan(x)')\n",
    "plt.title('Trigonometric Functions')\n",
    "\n",
    "plt.ylim(-1.5,1.5) # limit the range of y values\n",
    "plt.legend() # show the legend\n",
    "plt.grid() # show the grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAD8CAYAAACRkhiPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xd8XNWZ//HPGY16l1Ws3m0V44KNjQvG4I4NhjRCEpYa\nJ2xIyI8kG1IhyYbNQkI2C0mWngAhhCQQmnEBDLjgjpuaLatYvfc27fz+0NjIsoxlWZr6vF/Wy6Or\nuTPPnbkz33vPPfcepbVGCCGEdzI4uwAhhBDOIyEghBBeTEJACCG8mISAEEJ4MQkBIYTwYhICQgjh\nxSQEhBDCi0kICCGEF5MQEEIIL2Z05JMppZ4B1gKNWutpo5knOjpap6WlTWhdQgjhafbv39+stY45\n3/0cGgLAn4DHgOdGO0NaWhr79u2bsIKEEMITKaUqR3M/h4aA1vpDpVSaI59TDOo3W2ntMdHaY6Kl\nx0R7r4k+kxWT1YbJYmPAYgPAz8eAn3HwJ8DXQGSQH1HBgz+Tgv0J9PNx8pIIIcaTo/cERkUptR5Y\nD5CSkuLkatyH2WqjrKmHwroOypp6ONnaO/jT0ktLj2lcniMyyJeUqCCSo4JIiQoiPTqY3PgwsuNC\n8DdKQAjhblwyBLTWTwBPAMyZM0cuczoCrTXlzT3sq2hjf2UbBXUdHKvvxmQd3KI3KEiICCQlKogV\n+XEkRgQyKcT/9FZ9ZJAfQX4+p7f6/XwMKAUmy+Cegclqo89kpa3XTGuPibYeE809A9S09XGytZej\nNR1sPFqPxTb49hgNiqzYEPLiw7g0NZK56VFkxYRgMChnvkxCiPNwyRAQI6tq7eX9Y03sON7MvspW\nmrsHt+4jg3yZlhjObQvTyEsIIzc+jPToYHx9Lrzz14VszVusNipbeymq66SorpPC2k4+PN7MKx/X\nABAR5Muc1EjmZ0azZGoMGdHBKCWhIIQrkRBwYWarjb3lrWwtaWRrSROljd0AJEYEsjg7hjlpUcxN\njyQzJsQpX65GHwOZMSFkxoSwdnoCMLiHcrK1l70Vbewtb2VPRSvvFDXyizchOSqQJVNiuTonloVZ\n0fgZpYeyEM6mHDmojFLqr8ASIBpoAO7XWj/9afPMmTNHe1PvIIvVxu7yVt48XMvGo/W09Zrx8zEw\nLyOKK6fEcFVOrNttUZ/ag/mgpJGdJ1roNVkJDTCyPC+OtdPjWZQVI4EgxDhTSu3XWs857/1cfWQx\nbwgBrTWHqzv4x/5q3j5aR3O3iWA/H5blxXHNJfEsyoom2N8zdtoGLFZ2lrbw1pE6NhfU09lvITTA\nyKr8yXzhsmTmpEa6VcAJ4aokBNxAe6+JVz+u4W97qyiu7yLA18DSnMGt46tyYgnw9ezeNiaLjR0n\nmnnrcB1vH6mjx2QlPTqYz81O4rOXJjE5PMDZJQrhtiQEXJTWmj3lrfxl90k2FtRjstiYnhTOjZcl\nc+2MBMICfJ1dolP0mixsOFLP3/dVsbu8FYOCq3NiuWVBGouyomXvQIgLJCHgYgYsVt44VMcz28sp\nrOskLMDIZy5N4gtzkslLCHN2eS6lormHv++v4qU9VbT0mMiKDeGW+al85tIkj2kWE2KiSQi4iKau\nAV7YVclfdlfS3G0iOzaE2xelc/3MRDn79jwGLFbeOlzHn3ZWcLi6g9AAIzfOSeaOK9KJDw90dnlC\nuDQJASerbuvl8Q/K+Nu+KkwWG1fnxHLbQmnaGAutNR9XtfOnHRW8daQOg4IbZiXy9SszyYgJcXZ5\nQrgkCQEnOdHUzR/fP8G/Pq5BKfjMrCTWX5lBpnxZjYvqtl6e/LCMl/ZWYbLaWD1tMv++JItpieHO\nLk0IlyIh4GDF9Z08+l4pG47U4edj4Ka5KaxfnEFChDRbTISmrgGe3VHO8x9V0jVgYcnUGO5dPoXp\nSRHOLk0IlyAh4CDlzT38dssx3jhcS7Cfka9cnsodi9KJCfV3dmleobPfzPMfVfLktjLae82syIvj\n3hVTyJksB9uFd5MQmGC17X3877vH+fv+avx8DNy6MI2vLc4gIsjP2aV5pa5+M89sr+CpbWV0myys\nnZ7At5dlSzOc8FoSAhOkuXuA328t5S+7TgLwpXkp/PtVmcSGyolNrqC918ST28p4dkcF/WYrN8xK\n4jsrpkiznPA6EgLjrM9k5altZfzfByfoM1v53OwkvrU0m6TIIGeXJkbQ3D3AH98/wfO7KlHA7YvS\nuWtJpteejCe8j4TAOLHZNK9+XMPDm0qo7+xnVf5kvrdqqjQzuInqtl5+vamEfx2sJSrYj28vy+am\nuSljusy2EO5EQmAcfHSihV9uKORoTSczksL50Zo85qZHOaUWcXEOV7fz4IYidpW1khEdzPdX57Ai\nL07O2RAeS0LgIpxo6ua/NhTxTlEjCeEB/MeqHK6bkSCjZLk5rTXvFTfy4IYiTjT1MDctih+vzZVu\npcIjSQiMQUevmd++c4wXdlUS4OvDXUsyuWNRusdfzdPbWKw2/ravit9uOUZLj4kvzE7me6umEh0i\n3XqF55AQuAA2m+blfVU8tKmE9l4TX5ybwr3Lp8iXgofr7Dfz6LvHeXZHBYF+PtyzNJtbFqTJ8QLh\nESQERulgVTv3v3aUQ9UdzEmN5Gfr8slPkEsQeJPSxm5+/mYhHx5rIjMmmPuvzWfxlBhnlyXERZEQ\nOI/m7gEe2ljMy/uqiQ3154fX5LJuZoIcKPRSWmveLWrkF28VUtnSy7LcOH6yNpfUScHOLk2IMZEQ\nOAeL1cbzuyp5ZMsx+kxW7liUzjeXZhMi16kXDF6++pntFTz63nEsVs2dV6TzjauyZBwD4XYkBEbw\n0YkWHni9gJKGLq7Ijub+a/PJipX+/uJsDZ39/PfbxbzycQ2TwwL4wTWDPcRkT1G4i9GGgMOPgCml\nVimlSpRSpUqp+xzxnHUdfdz94gFuenIX3QMWHr95Ns/dPlcCQJxTXFgAj9w4k3/etYCYUH/ueekg\nNz6xi6K6TmeXJsS4cuiegFLKBzgGLAeqgb3ATVrrwnPNczF7AgMWK09tK+ex90qxac3Xr8zkriWZ\n0uVTXBCrTfO3vVU8vKmYjj4zX7k8lXuXT5GLBQqXNto9AUc3dM4FSrXWZQBKqZeAdcA5Q2CsthY3\n8rM3Cqho6WVlfhw/XpNHcpRc50dcOB+D4kvzUrjmksk8smXwPJI3DtXyvZU53HhZMj5yEqEYZx29\nZt4/1si6mYkT/lyObg5KBKqG/F5tn3YGpdR6pdQ+pdS+pqamC34Si9XGf75ViMGgeO72uTx+8xwJ\nAHHRIoL8+Pm6abz5zSvIjg3lh68e4frf72B/ZZuzSxMewmbTvLy3iqt+8z7fefkQte19E/6cjm4O\n+hywSmt9p/33m4F5Wuu7zzXPWJuDTrb0Mjk8AD+jnPgjxp/WmtcP1fLghiIaOgf4zKWJ3Lc6Ry4p\nLsbscHU7P32tgINV7eNyzpKrNgfVAMlDfk+yTxt3KZNky19MHKUU62Ymsiw3jse2lvLUtjI2FzRw\nz9Jsbl0oZx2L0WvpHuDhTSX8bV8Vk4L9+c3nZ3DDrESHXavM0XsCRgYPDC9l8Mt/L/AlrXXBueZx\n9qWkhRiN8uYefv5GAVtLBs86fuC6fK7IlrOOxblZrDZe3HOSX28qoddk5dYFadyzLJvQcRrzwiX3\nBLTWFqXU3cAmwAd45tMCQAh3kR4dzLO3zeXdogZ+/mYhNz+9h1X5k/nRmlw5HiXOsqe8lZ++dpTi\n+i4WZk3igWvzyY4LdUotXnWymBCO0G+28vT2T7om37Ukk69fKV2TBdR39PNfbxfx2sFaEiMC+fGa\nXFZNmzwhJyHKGcNCOFltex8PbijizcN1JEUG8uM1eazMl4FsvJHJYuOZHeX877vHsdg0X1+cwV1L\nsgj0m7gNAwkBIVyEXK7Ee526MOGDG4ooa+5hWW4cP12b55COKxICQriQ4RcuvG1hGt9aOn4HAYXr\nKazt5JcbCtlR2kJGTDA/WZvHVVNjHfb8EgJCuKDm7gEe3ljCy/uriA7x575VOQ7tDigmXmNXP49s\nPsbf9lURHujL/1s2hS/NS3F4t2EJASFc2KGqdn76egGHqtq5JDGcH6zOYUFWtLPLEhfhVIeAP2wt\nxWS1ccv8NL55dTbhQc7Z25MQEMLF2Wyafx2s4Tebj1HT3seVU2K4b3UOufFhzi5NXACbbfDs8Yc3\nlVDT3sfK/DjuW51LerRzBySSEBDCTfSbrTz/USWPbS2ls9/MDbMS+c6KqSRGBDq7NPEptNZsLWnk\noY0lFNd3kZ8Qxo/X5DE/c5KzSwMkBIRwOx29Zv7wfinP7qwA4LYFady1JFMuWe2C9la08tDGYvZW\ntJE6KYh7l0/h2ukJLnVsR0JACDdV097HI5uP8crH1YT4GbltUTp3LEonPFB6EjlbYW0nv95cwnvF\njcSG+vOtpdnceFmyS14rSkJACDdXXN/J7945zttH6wkNMHL7wnRulzBwiqM1HTz2XikbC+oJCzBy\n15Isbl2QNqEne10sCQEhPERhbSe/e/cYmwoaCAswcseiDG5blEaYnGMw4Q5WtfPou8d5t7iR0AAj\nty1M546F6U7r8XMhJASE8DAFtR387p3jbC5sIDTAyJfnpXLbwjTiwmQMg/G2r6KVR98r5YNjTYQH\n+nLnonRuWehewSshIISHOlrTwR8/OMHbR+rwMSiun5nI+sUZTrsKpaewWG1sKmjgqe1lfHyynahg\nP756RQY3z08lxN/RQ69cPAkBITzcyZZentpexsv7qug321iaE8sdV6QzP2OSXKTuAnQPWPjb3iqe\n3VFOdVsfqZOCuGNROp+bnUSQn/t9+Z8iISCEl2jtMfHcRxU891ElrT0msmJD+PK8FD5zaZIcRP4U\nxxq6eHH3Sf65v5quAQuXpUVy5xUZLMuNw8eFunqOlYSAEF6m32zlzcN1PL+rkkNV7QT6+rBuZgJf\nnpfKtMQw2Ttg8DXacKSOF3efZF9lG34+BlZfMpnbFqYzMznC2eWNKwkBIbzYkeoOXthVyWuHaug3\n25gSF8INs5K4flYC8eHedSayzabZV9nGqx/XsOFIHR19ZtKjg/nS3BQ+OzuJqGDPPBlPQkAIQUev\nmdcP1/LqgWoOnGxHKViQOYl1MxNZlhvnsV+AWmtKGrp47WAtrx+spaa9j0BfH1bkx3HjnGTmZ3r+\ncRMJASHEGcqbe3j14xr+9XENJ1t7MSi4LC2KlfmTWZEfR1Kke4+FbLHa2FvRxpbCBt4pauBkay8+\nBsUV2dFcPzOR5XlxBLthL5+xkhAQQoxIa83Rmk42F9azqaCeYw3dAORMDmVhVjQLsyYxN32Sy3eL\n1FpT1drHzhPN7DzRwofHm2jvNePnY2BB1iSW58WxIm8yMaH+zi7VKVwuBJRSnwceAHKBuVrrUX2z\nSwgIMbHKm3vYXFDPtuPN7KloxWSxYTQoZiRHcGlKBDOTI5mZEkFCeIBTm1BMFhvF9Z0cqu7g4Ml2\ndpW1UNPeB0B0iD+Ls6NZnhfH4ikxXrXFfy6uGAK5gA14HPiuhIAQrqffbGV/ZRvbS5vZVdZCQW0n\nJosNgNhQf3Liw8iODRn8iQshbVIwUcF+4xoOFquN6rY+TjR1c6Kpm7KmHorquyiq7cRkHawlMsiX\neemTWJA1iQWZk8iMCfH4Nv4LNdoQcFhcaq2LAHmjhHBhAb4+9iahwVHOTBYbRXWdHKxq51BVO8ca\nu/jL7hb6zbbT8/gZDcSHBxAfHkBcWADhgb6EBhgJDRj839dgAAWKwc+/1Waje8BK74CFbpOFrn4L\njZ0DNHb1U9/RT3P3ALYh26bRIX5kxYZw68I0pieFMyMpgqTIQPkuGScuuc+klFoPrAdISUlxcjVC\neC8/o4EZyRHMGNKH3mbT1LT3cayhi5OtvdR39FPb0U9dex8HTrbR1T/4xW61nb+Vwc9oINTfSEyo\nP3FhAeRMDiUuLIDkqCAyY0LIjAmW8RQm2LiGgFLqHWDyCH/6kdb6tdE+jtb6CeAJGGwOGqfyhBDj\nwGBQJEcFkRx17t5EWmv6zFa6+i2YrTZOtTprDQYDhPgbCfY3uuR1+L3NuIaA1nrZeD6eEMI9KaUI\n8jO69bV3vIXLv0P79+9vVkpVjnH2aKB5POtxMHevH9x/Gdy9fnD/ZXD3+sE5y5A6mjs5snfQDcCj\nQAzQDhzUWq+c4OfcN5qj467K3esH918Gd68f3H8Z3L1+cO1lcGTvoFeBVx31fEIIIc5PjsoIIYQX\n8/QQeMLZBVwkd68f3H8Z3L1+cP9lcPf6wYWXweWvHSSEEGLiePqegBBCiE8hISCEEF5MQkAIIbyY\nhIAQQngxCQEhhPBiEgJCCOHFHBoCSqlnlFKNSqmjjnxeIYQQI3PoeQJKqcVAN/Cc1nraaOaJjo7W\naWlpE1qXEEJ4mv379zdrrWPOdz+HXkVUa/2hUirtQuZJS0tDhpcUQogLM9qrL7v8paTH6nt/P0R9\nZz9+Pgb8jAZ8fQwE+fkQGexHVJAfUcF+RIX4kRQRSHJUEAG+Ps4uWQjh4aw2TV1HH1WtfTR3D9Da\nY6Klx0RrzwC9A1YGLDb7jxWTxcYLd86b8IF3XDIExmN4SYtN0z1gwWSxDf5YbfSarLT1mLCMMOxd\nXJg/KVFBZESHkJcQRm58GDnxoYQF+F7UsgghvI/ZaqOsqYfCug4Kazs51tBNVWsvVW29mK1nf/9E\nBPkS4m/E32jAz+hj/9+A1aaZ6O1Th187yN4c9OZojwnMmTNHj2dzkNaargELrd0mWnoGqG7r42RL\nLydbe6ls7aW0sZvWHtPp+6dOCmJ2aiSXpUVxWVoUmTHBMsC1EOIMrT0m9la0sre8lb2VbRTVdWKy\n2IDBcZSzYkJIiw4iJSqY1ElBJEcGERPqT1SwH5FBvhgnYGtfKbV/NGMYuOSewERSShEW4EtYgC9p\n0cHMHjb2jtaahs4Biuo6Kazr5HB1Ox+UNPHKgRoAooL9mJ85iSVTYrhyagyxoQFOWAohhDMNWKzs\nKW9la3ET2443cbyxGxj8wp+ZFMGtC9LIiw8jLyGMjOjgCfmSHy8ODQGl1F+BJUC0UqoauF9r/bQj\nazgfpRSTwwOYHB7AVTmxwGAwlDf3sK+ijd3lrWw73sRbh+sAyE8IY2lOLNdMj2dqXKjsJQjhoZq7\nB9hc0MB7xY3sPNFMr8mKn9HAvPQorp+VyNz0KKYnheNvdK/jiy5/Kenxbg4aD1prCus6eb+kiQ9K\nmthX2YpNQ2ZMMGsuiZdAEMJDtPea2FRQz5uH69h5ogWrTZMYEcjVObEsmRrD/MxJBPm5ZoPKaJuD\nJATGQVPXAJsK6nnrcB27y1uwaZgaF8rn5yRxw6xEJoX4O7tEIcQoma023i1q5O/7qvjgWBMWmyZ1\nUhBrp8ez5pIEcuPdYwNPQsBJmroG2FhQzysHqvn4ZDtGg2Jpbiyfn53MkqkxLt02KIQ3K23s5uV9\nVbxyoJrmbhNxYf6sm5nItdMTmJYY5hZf/ENJCLiA4w1d/H1/9emVKiE8gK/MT+WLl6UQFezn7PKE\n8HoWq43NhQ38aUcFeypaT2+03XhZMouz3XujTULAhZzavXx+VwU7SlvwNxpYNzOBWxakkZ8Q7uzy\nhPA67b0mXtpbxfMfVVLT3kdSZCBfuTyVz16aREyoZzTfSgi4qJL6Lv78UQWvHqihz2xlXnoUdy3J\n5MopMW63uymEu6lo7uGJbWW8cqCafrONyzOiuG1hOsty4/AxeNbnT0LAxXX0mnl5XxXP7CinrqOf\n/IQw7lqSyepp8R63MgrhbMX1nfxh6wnePFyL0cfA9TMTuG1hOrnxYc4ubcJICLgJk8XGvw7W8H/v\nn6CsuYf06GC+tjiDGy5NdLv+xkK4mo9PtvH7rSd4p6iBID8fvnJ5KncuSic2zPNP8pQQcDNWm2ZT\nQT1/eL+UozWdJIQH8K2l2Xx2dtKEX0BKCE9z4GQbj2w+xvbSZsIDfbltYRq3LkgjIsh7OmRICLgp\nrTUfHm/mkS3HOFTVTtqkIL69bArXzkiQZiIhzqOwtpPfbC7h3eJGJgX7sX5xBl++PJUQf9c8oWsi\nSQi4Oa017xY18pstxyiq6yQ7NoR7l09hZf5kDBIGQpzhRFM3j2w5xluH6wgLMPK1KzO5dUEawV74\n5X+KhICHsNk0GwvqeWTLMUobu5mWGMYPVueyMCva2aUJ4XS17X38dssx/nmgmgBfH25fmM5XF2cQ\nHiiXgJcQ8DBWm+a1gzU8suUY1W19XDU1hh9ck8uUuFBnlyaEw3UPWPjj+6U8ta0cDdx8eSp3Lckk\nWi7RcpqEgIcasFh5bmclj753nO4BCzdelsz/WzbFK3o7CGGx2nhpbxX/884xmrtNXD8zge+unEpS\nZJCzS3M5EgIerq3HxKPvlfL8rgp8fQysX5zBV6/I8Oo2UOG5tNZsLWnkwQ3FlDZ2Mzctih+tyWVG\ncoSzS3NZEgJeorKlh4c2lvDWkTpiQv35/qocPjMrUQ4eC49RUNvBgxuK2FHaQnp0MPetzmFFXpyc\nYX8eEgJeZn9lG794s5CDVe3MSI7gZ9flM1O2koQba+sx8evNJby45yThgb7cszSbL89Lxc8o582M\nhoSAF7LZNK9+XMOvNhbT1DXA52Yn8R+rpsoQmMKtWG2av+45ya83l9DVb+Hf5qfy7aVTCA+SHj8X\nQsYY9kIGg+Kzs5NYOW0yj71XytPby9h4tJ5vXp3FbQvTZQtKuLz9la389LUCCmo7uTwjigeuyydn\nsude38cVyJ6ABytv7uE/3yzk3eJGMqKD+cnavNPjJgvhShq7+vnV28W8cqCGyWEB/GhNLmunx0u7\n/0WQ5iBx2taSRn7xRiFlzT1cNTWGn16bT3p0sLPLEgKz1cafd1bwP+8cx2SxcecV6Xzjqizp5TYO\nXDYElFKrgN8BPsBTWutffdr9JQTGh8ky+GH73buDH7Y7rkjnbvmwCSfafryZB94ooLSxWzZOJoBL\nhoBSygc4BiwHqoG9wE1a68JzzSMhML4au/r577dL+OeBaiaHBfCDa3K4bkaC7HYLh6lp7+OXbxWy\n4Ug9KVFB3H9tHktz45xdlsdx1RCYDzygtV5p//0HAFrr/zrXPBICE2N/ZRsPvF7AkZoO5qZH8cC1\n+eQlyAE4MXH6zVae/LCM379fCsDdV2Vx5xUZBPjKuBkTwVV7ByUCVUN+rwbmObgGAcxOjeRf31jI\ny/uqeGhjMWsf3cZXLk/l3uVTvOqa62Linboi7s/fLORkay9rLonnh2tySYwIdHZpAhftIqqUWg+s\nB0hJSXFyNZ7Lx6C4aW4K10yL55EtJTy/q5I3DtXyvZU53HhZsoxfIC5aeXMPP3+jgK0lTWTHhvCX\nO+fJFXBdjDQHidMKazt54PUC9lS0Mi0xjJ9dN43ZqZHOLku4oZ4BC7/fOniVTz+jgW8vy+aWBWky\nSp4DueoxASODB4aXAjUMHhj+kta64FzzSAg4ltaa1w/V8uCGIho6B/jMpYnctzpHzjoWo6K15o3D\ndTz4VhH1nf189tIkvr9azlp3Bpc8JqC1tiil7gY2MdhF9JlPCwDheEop1s1MZFluHI9tLeWpbWVs\nLmjgnqWDW3Jy1rE4l+L6Tu5/rYDd5YN7kr//8qWyJ+kG5GQx8amGtulmxgTzwHX5XJEd4+yyhAvp\n6DPz2y3HeH5XJaEBRr63cipfvCxFjik5mUs2B42FhIBreLeogZ+/WUhlSy8r8+P48Zo8kqNkIA9v\nZrNp/rG/mv/eWExbr4kvz0vlOyukd5mrcMnmIOG+lubGsTArmqe3l/PYe6UsK/mAr1+ZyV1LMqWf\ntxc6VNXOT18v4FBVO3NSI3lu3VzyE8KdXZYYA9kTEBestr2PBzcU8ebhOhIjAvnJ2lxW5k+Ws469\nQH1HPw9tGrzQW0yoPz+8JofrZybKe++CpDlITLidJ5r52euFlDR0sSgrmgeuyyMrVga+90S9JguP\nf1DG4x+ewGaD2xel842rMgkNkGv8uyoJAeEQFquN53dV8siWY/SZrNy6II1vXp0tA4B4CJtN88rH\nNTy8qZiGzgHWTI/nvlU5cjzIDUgICIdq7h7g4Y0lvLy/irAAX+6+Koub56fK8QI3trushV+8VcjR\nmk5mJEfw07W5zE6NcnZZYpQkBIRTFNR28Ku3i9l2vJnEiEC+u3IK62bIwPfupLSxi4c3lbCpoIGE\n8AC+vzqHa6cnyHvoZiQEhFNtP97Mf71dREFtJ3nxYdy3OofFU+T8AldW097H/2w5xj8PVBPkZ+Rr\nizO484oMAv1kb84dSQgIp7PZNG8cruXhTSVUt/WxKCua/1g1lelJEc4uTQzR0j3A77ee4IVdlaDg\n3y5P5d+vyiIqWPr7uzMJAeEyBixWXth1ksfeO05br5mlObF8e9kULkmSfuXO1NFn5pnt5Ty1rYw+\ns5XPz07mnmXZJMglnj2ChIBwOV39Zv68s4Int5XT0WdmWW4c316WzbRECQNHau818cz2cp7dUUHX\ngIVrLpnMvcunkhUb4uzSxDiSEBAuq6vfzJ92VPDktjI6+y2syIvjW0slDCZaS/cAT20v57mdFfSY\nrKyeNpm7r86SM309lISAcHmd/Wae3V7BU9vL6Oq3sCgrmq9dmcGirGg5A3Uc1XX08cz2cl7YdZJ+\ni5U1l8TzzauzmTpZTuzzZBICwm109pt5cfdJntleTmPXAHnxYXztygyuuSReBiG5CEdrOnhyWxlv\nHa7DpjXXzUjg7quz5KxuLyEhINzOgMXKawdreeLDMkobu0mMCOSWBal8fnYykdJTZVSsNs3W4kae\n3FbG7vJWQvyNfPGyZG5ZkCZn+XoZCQHhtmw2zdaSRh7/sIw95a34GQ2snR7PVy5PZVZyhDQVjaCx\nq5+/76vmxd0nqWnvIzEikNsWpvGFy5IJk+v7eCUJAeERius7+cuuk7xyoJoek5X8hDC+NC+FtZck\neP31iaw2zc4Tzby4+yRbChuw2DQLMifx5XmprMyPwyhNaV5NQkB4lO4BC//6uIYXdlVSXN+Fn4+B\npbmx3DAfkCLpAAAPI0lEQVQrkSVTY71m2EutNUV1XfzrYA2vH6ylvrOfyCBfPjc7iZvmppARI908\nxSAJAeGRtNYcqenglQM1vHGolpYeE5FBvqy+JJ6V+ZOZnzHJ4wJBa82Jpm42FTTw+sFaShq6MBoU\nS6bGsG5mIsvz4uRCfeIsEgLC45mtNrYdb+KVAzW8V9xIr8lKqL+Rq3JiWZEfx+IpMW7bHm61afZX\ntrGlsJ53ihopb+4BYHZqJNfPSmTNJfFyWQfxqVwuBJRSnwceAHKBuVrrUX2zSwiI0eg3W9lR2szm\nggbeKWqgpceEj0ExPSmcRVnRLMiM5tLUCPyNrrnFrLXmeGM3O0ub+aishV1lrXT0mfH1UczPjGZ5\nXhzLcmOJD5dLOojRccUQyAVswOPAdyUExEQ5tRW97XgTO0qbOVTdgdWm8TMayE8IY2ZyBDOTI5iV\nHElyVKBTehu19Zg4XNPB4ap2DlW3c7CqneZuEwCJEYEszJrE4ikxXDklRkbvEmPiciFw+gmVeh8J\nAeFAnf1mdpe1sqe8hYNV7Ryp6aDfbAMg2M+HrNgQMmNDyI4NJW1SEJPDA0iICCQ6xB+fMV5DX2tN\nZ7+Fhs5+6jv6qWjp4URjNyeaejjR1E1dR//p+2bEBDMjKYLLM6JYkBkt/fnFuBhtCBgdUYwQzhQW\n4MvyvDiW58UBg8cSSuq7OFzdwbGGLkobu9lR2swrB2rOmM9oUESH+BMWaCQ0wJcQfyOhAUaMBoVS\nCgWgwGLV9AxY6DFZ6Bmw0tlvprFzgD6z9YzHC/E3khkTzPyMSWTHhTIjKZxpSeFue9xCeIZxDQGl\n1DvA5BH+9COt9WsX8DjrgfUAKSkp41SdEIN8fQxMSww/64J1HX1mqtt6qWvvp66zn/qOPho6B+ju\nt9A1YKat10RVay8Wm0aj0Rq0Bl8fRbC/kWA/I9EhfqRFBxMX6s/k8ABiwwKIC/UnLTqY2FB/OdFN\nuJxxDQGt9bJxepwngCdgsDloPB5TiPMJD/QlPDBcrqopvIpndagWQghxQRzZO+gG4FEgBmgHDmqt\nV45iviagcoxPGw00j3FeV+Du9YP7L4O71w/uvwzuXj84ZxlStdbnHdjb5U8WuxhKqX2jOTruqty9\nfnD/ZXD3+sH9l8Hd6wfXXgZpDhJCCC8mISCEEF7M00PgCWcXcJHcvX5w/2Vw9/rB/ZfB3esHF14G\njz4mIIQQ4tN5+p6AEEKITyEhIIQQXkxCQAghvJiEgBBCeDEJASGE8GISAkII4cUkBIQQwou5/KAy\n0dHROi0tzdllCCGEW9m/f3/zaC4g5/IhkJaWhgwvKYQQF0YpNaqrL0tzkADg5qd3c9MTu5xdhhDC\nwVx+T0A4xrbj7n65diHEWMiegBBCeDEJASGE8GISAkII4cUkBIQQwotJCAghhBeTEBBCCC923hBQ\nSj2jlGpUSh0dMi1KKbVFKXXc/n/kkL/9QClVqpQqUUqtHDJ9tlLqiP1v/6uUUuO/OEIIIS7EaPYE\n/gSsGjbtPuBdrXU28K79d5RSecAXgXz7PH9QSvnY5/kj8FUg2/4z/DGFEEI42HlDQGv9IdA6bPI6\n4M/2238Grh8y/SWt9YDWuhwoBeYqpeKBMK31Lj04qPFzQ+YRQgjhJGM9JhCnta6z364H4uy3E4Gq\nIfertk9LtN8ePn1ESqn1Sql9Sql9TU1NYyxRCCHE+Vz0gWH7lr0eh1qGPuYTWus5Wus5MTHnvQie\nEEKIMRprCDTYm3iw/99on14DJA+5X5J9Wo399vDpQgghnGisIfA6cIv99i3Aa0Omf1Ep5a+USmfw\nAPAee9NRp1LqcnuvoH8bMo8QQggnOe9VRJVSfwWWANFKqWrgfuBXwMtKqTuASuALAFrrAqXUy0Ah\nYAG+obW22h/q3xnsaRQIvG3/EUII4UTnDQGt9U3n+NPSc9z/l8AvR5i+D5h2QdUJIYSYUHLGsBBC\neDEJASGE8GISAkII4cUkBIQQwotJCAghhBeTEBBCCC8mISCEEF5MQkAIIbyYhIAQQngxCQHB4IVg\nhRDeSEJAYJMMEMJrSQgIbLInIITXkhAQWIfsCkjTkBDeRUJAnLEnIBkghHeREBBnHBOQpiEhvIuE\ngDijOUgOEgvhXSQExBnHAWRPQAjvIiEghu0JSAgI4U0kBMSwYwLOq0N4ruc/quAzf9jh7DLECM47\nxrDwfDZpDhIT7CevFTi7BHEOsicgzuwianNiIUIIh5MQEHJMQDiMnIzoeiQEBLYhW/9W+ZCKCWSR\ng04uR0JAyDEB4TAWq6xfrkZCQJyx9S8ZICaS2SYHnVyNhICQk8WEw5gtEgKuRkJAYB3yuZQmWzGR\n5JiA65EQEGceE5APqZhAZqvsCbgaCQEhXUSFw5jlwLDLkRAQZxwMlh0BMZEssifgciQExBm9g2RP\nQEwkk4SAy5EQEMNGFpMQEONr6Dol5wm4HgkBQXe/5fRtaQ4S421gSLdQi5wn4HIcHgJKqVVKqRKl\nVKlS6j5HP784296K1tO3pTlIjLdDVe2nb5sssn65GoeGgFLKB/g9sBrIA25SSuU5sgbxCatNs6e8\nlb/uqTpjmhDjQWvN8YYuHtpUcnqa7Am4HkePJzAXKNValwEopV4C1gGF4/1EWwob6OgzA4Mr4+mv\nNg2nftOa09NPbQBr9JDbn/xBD73PkMc78zHOvvzC8McbafrQecdc0wjPY7XZGLDYMNl/Buw/3QNm\nGjsHqOvop89sJSbUn++tnMrDm0rOumyE1pp+s42uATMWq8ZstWG2aiw22+nfbVpj0/ba7a/NWbft\n9WnsextDp+nBaZ8sz9nLMvw9G/5aD32sT+Y7+30f/voOn8YZdQxfhpHeq+F1jfB+6LPvc9Y6OOw9\nHen91HxSxOnlG+E+w9erkeo6cx0b/trpEZeHM+Yb+XWx2jTtvWbaek3UdfTT0WcmwNfArQvS+NPO\nihGPCZgsNrr6zZistjPWMbPVhsWm7RsmZ65jtiGvh23Icp9ajzh1e+g6d8YyDzXy34be7azPxajm\nOXtZR57n3LWtm5mIj0Gd83HGg6NDIBGoGvJ7NTBv+J2UUuuB9QApKSljeqKHNhZzvLF7TPO6A2Vf\nLxSg7L+oM6YrjD4KP6MBPx8D/r72/40+BPv7kJsQxpKpscxIDmdpbhy7y1oAONHUzeuHajlQ2UZF\nSy8dfSbp2+0gw9/T4e+n/d8Z05R9GkPmYfg0da71RJ1+XnX6/+HTPvkCUurc9zl1Lx+DIjzQl6TI\nIC5NjSQvPowV+XE0dg7wp50VNHcP8PgHJ9hxooUTjd209pjoM1vH6yX0ONdcEo+PwWdCn8MlRxbT\nWj8BPAEwZ86cMX0D/fn2uVismiHr8CcfniEr7fCVGmDI5+P0B+3UpHN94Z6+/7AP4Ej3HbGmT3me\n0487dMZxZrA/9j0vHcTXRzErOZKlObFEBvsRFmgk1N+Ir48Bo48BXx81eNsw+L/BoDCoM7+Uhn/5\nnPO2/bnVGa+fOv2FM/S1GfpacI4vwBHfT/Xp9/nki3PkL8Az5hn2Po+0Lg1/f0d6b89YJybwfXUV\nbT2De+Xf+8dhAHImh3JZWiTRIf6EB/oSGmDE39fn9Do1uK4pfH2Uff34ZB0zKE6/Xwb1yTplGLJe\nGIa8xqfW7TPeF858zUf6TA6/3/C3aeivZ/5ttPOoc0z/5Lafz8S32Ds6BGqA5CG/J9mnjbuEiMCJ\neFiPNXTF2/H9q4kNC3BeMcLj+Pp8soL99auXMz9zkhOrEUM5unfQXiBbKZWulPIDvgi87uAaxAhO\ntZnOSomQABDjznfIFq0EgGtx6J6A1tqilLob2AT4AM9orWUEahfQ0DkAwIykCCdXIjyRrwOaNcTY\nOPyYgNZ6A7DB0c8rPt21MxIoqe/iOyumOLsU4YFOdXBJmxTk3ELEWVzywLBwvBB/Iw9cl+/sMoSH\nign157srpnD9rERnlyKGkRAQQkw4pRR3X53t7DLECKShTgghvJhy9atGKqWagMoxzh4NNI9jOY7m\n7vWD+y+Du9cP7r8M7l4/OGcZUrXWMee7k8uHwMVQSu3TWs9xdh1j5e71g/svg7vXD+6/DO5eP7j2\nMkhzkBBCeDEJASGE8GKeHgJPOLuAi+Tu9YP7L4O71w/uvwzuXj+48DJ49DEBIYQQn87T9wSEEEJ8\nCo8MAXcfwlIp9YxSqlEpddTZtYyFUipZKbVVKVWolCpQSt3j7JoulFIqQCm1Ryl1yL4MP3N2TWOh\nlPJRSn2slHrT2bWMhVKqQil1RCl1UCm1z9n1XCilVIRS6h9KqWKlVJFSar6zaxrO45qD7ENYHgOW\nMzhozV7gJq31uI9eNlGUUouBbuA5rfU0Z9dzoZRS8UC81vqAUioU2A9c72bvgQKCtdbdSilfYDtw\nj9Z6l5NLuyBKqXuBOUCY1nqts+u5UEqpCmCO1totzxNQSv0Z2Ka1fsp+5eQgrXX7+eZzJE/cEzg9\nhKXW2gScGsLSbWitPwRaz3tHF6W1rtNaH7Df7gKKGBxVzm3oQaeGpvO1/7jVFpNSKglYAzzl7Fq8\nkVIqHFgMPA2gtTa5WgCAZ4bASENYutUXkCdRSqUBs4Ddzq3kwtmbUg4CjcAWrbW7LcP/AP8BuPPo\n7hp4Rym13z7srDtJB5qAZ+1Nck8ppYKdXdRwnhgCwkUopUKAfwLf1lp3OrueC6W1tmqtZzI4At5c\npZTbNM0ppdYCjVrr/c6u5SItsr8Hq4Fv2JtK3YURuBT4o9Z6FtADuNwxSk8MAYcNYSnOzd6O/k/g\nL1rrV5xdz8Ww78JvBVY5u5YLsBC4zt6m/hJwtVLqBeeWdOG01jX2/xuBVxls7nUX1UD1kD3IfzAY\nCi7FE0NAhrB0MvtB1aeBIq31I86uZyyUUjFKqQj77UAGOxoUO7eq0dNa/0BrnaS1TmPwM/Ce1vor\nTi7rgiilgu0dC7A3o6wA3KbHnNa6HqhSSk21T1oKuFznCI8bT8AThrBUSv0VWAJEK6Wqgfu11k87\nt6oLshC4GThib1MH+KF9VDl3EQ/82d7bzAC8rLV2y26WbiwOeHVwmwIj8KLWeqNzS7pg3wT+Yt8g\nLQNuc3I9Z/G4LqJCCCFGzxObg4QQQoyShIAQQngxCQEhhPBiEgJCCOHFJASEEMKLSQgIIYQXkxAQ\nQggvJiEghBBe7P8DNtz+ksfXUPUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x8ed4390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot in three subplots (sharing the x axis)\n",
    "f, (ax1, ax2, ax3) = plt.subplots(3,sharex=True,sharey=False)\n",
    "ax1.plot(X,sin(X));\n",
    "ax2.plot(X,cos(X));\n",
    "ax3.plot(X,tan(X));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEWCAYAAACJ0YulAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAG61JREFUeJzt3XuUXGWd7vHvQ4AEaC4SgkAnSDCxNYSAEYJ4iR3FQwho\nZgaEtIAiMBlmyFHPyFEcBiZcZkBlOOiAk4nACAIJWQoSNFzkQKMuFEICCSTQoQU5JAYRHC4tk0DI\n7/yxd28rnequ6qR276rO81mrV6r2fmvvpyvd/av3ffdFEYGZmRnAdkUHMDOz+uGiYGZmGRcFMzPL\nuCiYmVnGRcHMzDIuCmZmlnFRsEFP0p2SPj8A+zlAUkjavtb7lfRRSR0lz38r6ahabDvd3gpJrbXa\nnjUu+TwFqzVJM4D/BYwH/gQ8C1wP/HsM4h84SQeQfK87RMSGfrwugLER0dmP1/wWODMi7u1nTCR9\nH1gdEf/Y39fa4OeegtWUpK8A3wa+BewDvBM4C/gwsGOB0RpWd8/DbCC4KFjNSNoduAj4u4j4YUS8\nHolHI+LkiFiftjtW0qOSXpP0vKTZJdtolbS6x3azoRJJkyQ9kr7295KuSJcPk3SjpJclvSJpsaR3\npuvaJZ2ZPn63pPvSdi9JuknSHj32dY6k5ZJelXSLpGG9fL9DJF2ebucZ4Nge60v3O0bSA+k2X5J0\nS7r852nzZZK6JJ3U/R5I+pqkF4D/LPe+AIdLWinpvyT9Z3dOSadJ+mWPLJFmmAmcDHw13d8dZd7j\noZKulPS79OtKSUNL/38kfUXSi5LWSvpCLz8S1oBcFKyWjgSGArdXaPcn4HPAHiR/SP9W0l9UuY9v\nA9+OiN2AdwML0uWfB3YHRgHDSXon/13m9QIuBfYD3pe2n92jzYnAVGA0MAE4rZcsfw0cB7wfOAw4\noY/cFwP3AO8ARgL/BhARk9P1h0REU0Tckj7fB9gTeBcws5dtngwcTfI+vAeoOBwUEXOBm4Bvpvv7\nVJlm5wEfBA4FDgEm9dj2PiTvdTNwBnC1pHdU2rc1BhcFq6W9gJdKx9MlPZh+cv9vSZMBIqI9Ih6P\niI0RsRyYB3ysyn28BYyRtFdEdEXEr0uWDwfGRMTbEbEkIl7r+eKI6IyIn0XE+oj4A3BFmX1/JyJ+\nFxF/BO4g+eNYzonAlRHxfNr20gq53wXsFxHrIuKXfbQF2Aj8U5qzXHEDuKpk3/8MtFXYZrVOBi6K\niBfT9+hC4NSS9W+l69+KiEVAF9BSo31bwVwUrJZeBvYqHQOPiA9FxB7puu0AJB0h6X5Jf5D0Ksmn\n+r2q3McZJJ+Kn0qHiI5Ll/8AuBuYnw55fFPSDj1fLOmdkuZLWiPpNeDGMvt+oeTxG0BTL1n2A54v\nef5cH7m/StJLeTg90uf0PtoC/CEi1lVo03Pf+1VoX6392PR76bntl3tMpPf1HlmDcVGwWvoVsB6Y\nXqHdzcBCYFRE7A7MIfmDCcnQ0s7dDSUNAUZ0P4+IpyOiDdgb+AbwQ0m7pJ9aL4yIccCHSIZ1Pldm\n3/8CBHBwOgR1Ssm++2styfBTt/17axgRL0TEX0fEfsDfAN+VNKaPbVdzlFbPff8ufdzzPdynn9v+\nHUmvpty2bZBzUbCaiYhXSIYavivpBEm7StpO0qHALiVNdwX+GBHrJE0CPluybhUwLJ2M3oFkLHto\n90pJp0gaEREbgVfSxRslTZF0cFpEXiMZ4thYJuauJMMdr0pqBv73VnzLC4AvShqZjqmf21tDSZ+R\nNDJ9+l8kf5i78/0eOHAL9n92uu89SeYBuucjlgEHSTo0nXye3eN1lfY3D/hHSSMk7QVcQNKjsm2A\ni4LVVER8E/h7kuGS36df/wF8DXgwbfZ3wEWSXif5g7Og5PWvpuuvAdaQfOotPepmKrBCUhfJpPOM\ndMx9H+CHJAXhSeABkiGlni4EJgKvAj8Fbt2Kb/d7JENWy4ClFbZ1OPBQmnsh8KWIeCZdNxu4Pp17\nObEf+7+ZZPL6GeA3wCUAEbGK5Ciwe4GngZ7zF9cC49L9/bjMdi8BHgGWA4+n39sl/chlDcwnr5mZ\nWcY9BTMzy+RaFCRNldQhqVPSZuOt6Ykwr0p6LP26IM88ZmbWt9xOn08n/K4GPkkyJrxY0sKIWNmj\n6S8i4rjNNmBmZgMuz57CJKAzIp6JiDeB+VQ+VNHMzAqU54W2mtn05JrVwBFl2n1I0nKSI03OiYgV\nPRuk12uZCTBs2LAP7L9/r4eD142NGzey3Xb1P2XjnLXVCDkbISM4Z62tWrXqpYgYUald0VdfXArs\nHxFdkqYBPwbG9myUXq9lLkBLS0t0dHT0bFJ32tvbaW1tLTpGRc5ZW42QsxEygnPWmqS+zrjP5Fne\n1rDpGZcj02WZiHgtIrrSx4uAHdKTZczMrAB5FoXFwFhJoyXtCMwgOWknI2kfSUofT0rzvJxjJjMz\n60Nuw0cRsUHSLJIzPocA10XECklnpevnkFxq+G8lbSC5zPGMwXxnLjOzepfrnEI6JLSox7I5JY+v\nAq7KM4OZmVWv/qfMzcxswLgomJlZxkXBzMwyLgpmZpZxUTAzs4yLgpmZZVwUzMws46JgZmYZFwUz\nM8u4KFhN3XXXXbS0tDBmzBguu+yywnKcfvrp7L333owfP76wDGaNyEXBaubtt9/m7LPP5s4772Tl\nypXMmzePlSt73mhvYJx22mncddddhezbrJG5KBgAra2tPPXUUwC8/PLLW/QJ++GHH2bMmDEceOCB\n7LjjjsyYMYPbb7+96tcvW7aMyZMnM27cOLbbbjskccEFW3bb7smTJ7Pnnntu0WvNtmVF32TH6kRn\nZyfvec97AFi+fDkHH3zwJus/+tGP8vrrr2/2ussvv5yjjjoKgDVr1jBq1J9voTFy5Egeeuihqva/\nbt06TjrpJG644QYmTZrE+eefz7p167jwwgv7lcHMto6LgvHcc8/R3Nyc3VJw+fLlTJgwYZM2v/jF\nL3LNcO+99zJx4kQmTZoEwIQJE7jrrrtIb7cxIBnMzEXBSIZtSovAkiVLOOmkkzZpU82n9ObmZp5/\n/s+35V69ejXNzc1VZXjiiSc26Z0sXbqUiRMn9juDmW0dFwXjscceY926dQA8/fTT3H777VxyySWb\ntKnmU/rhhx/O008/zbPPPktzczPz58/n5ptvztZ/4hOf4IYbbihbKIYPH859990HwKpVq7j11lt5\n8MEH+53BzLaOJ5qNZcuWsXHjRg455BAuuugixo0bx/XXX9/v7Wy//fZcddVVHH300bzvfe/jxBNP\n5KCDDgJg48aNdHZ29jr529bWRldXF+PHj2fmzJnMmzeP4cOHb/H31NbWxpFHHklHRwcjR47k2muv\n3eJtmW1L3FMwli9fztKlS9l11123elvTpk1j2rRpmy1fuXIlxx9/PDvttFPZ1zU1NXHHHXds9f67\nzZs3r2bbMtuWuKewjXvjjTeQVJOC0Jfx48dzxRVX5LoPM9t6LgrbuJ133plVq1YVHcPM6oSLgpmZ\nZVwUzMws46JgZmYZFwUzM8u4KJiZWcZFwczMMi4KZmaWcVEwM7OMi4KZmWVcFMzMLOOiYGZmGRcF\nMzPLuCiYmVnGRcHMzDIuCmZmlsm1KEiaKqlDUqekc/tod7ikDZJOyDOPmZn1LbeiIGkIcDVwDDAO\naJM0rpd23wDuySuLmZlVJ8+ewiSgMyKeiYg3gfnA9DLt/ifwI+DFHLOYmVkVFBH5bDgZCpoaEWem\nz08FjoiIWSVtmoGbgSnAdcBPIuKHZbY1E5gJMGLEiA8sWLAgl8y11NXVRVNTU9ExKnLO/lnbtbbP\n9UPfHsr6Ies3W75v0755Req3enkvK3HO2poyZcqSiDisUrvtByJMH64EvhYRGyX12igi5gJzAVpa\nWqK1tXVg0m2F9vZ2nLN26iXn7PbZfa5v6Wqho6ljs+VtrW05Jeq/enkvK3HOYuRZFNYAo0qej0yX\nlToMmJ8WhL2AaZI2RMSPc8xlZma9yLMoLAbGShpNUgxmAJ8tbRARo7sfS/o+yfCRC4KZWUFyKwoR\nsUHSLOBuYAhwXUSskHRWun5OXvs2M7Mtk+ucQkQsAhb1WFa2GETEaXlmMTOzynxGs5mZZVwUzMws\nU/QhqWZ1p9Jhp2aDmXsKZmaWcVEwM7OMh4/MBkClIanZrX2vNxso7imYmVnGPQWzOtBXT8K9CBtI\n7imYmVnGRcHMzDIuCmZmlvGcgm2TfIKaWXnuKZiZWcZFwczMMi4KZmaWcVEwM7OMi4KZmWVcFMzM\nLOOiYGZmGRcFMzPL+OQ1G7R8gppZ/7mnYGZmGfcUzOqcL6ttA8k9BTMzy7gomJlZxkXBzMwyLgpm\nZpZxUTAzs4yLgpmZZVwUzMws46JgZmYZFwUzM8u4KJiZWcZFwczMMrkWBUlTJXVI6pR0bpn10yUt\nl/SYpEckfSTPPGZm1reqioKkWyUdK6nqIiJpCHA1cAwwDmiTNK5Hs/8LHBIRhwKnA9dUu30zM6u9\nav/Ifxf4LPC0pMsktVTxmklAZ0Q8ExFvAvOB6aUNIqIrIiJ9ugsQmJlZYfTnv8lVNJZ2B9qA84Dn\nge8BN0bEW2XangBMjYgz0+enAkdExKwe7f4SuBTYGzg2In5VZlszgZkAI0aM+MCCBQuqzlyUrq4u\nmpqaio5R0WDOubZrbU5pejf07aGsH7J+wPdbzr5N+5ZdPpj/z4vQKDmnTJmyJCIOq9Su6vspSBoO\nnAKcCjwK3AR8BPg80LplMSEibgNukzQZuBg4qkybucBcgJaWlmht3eLdDZj29nacs3Z6y9nn3dUK\n+D1t6Wqho6lj4HdcRltrW9nljf5/Xm8aJWe1qioKkm4DWoAfAJ+KiO6PYLdIeqSXl60BRpU8H5ku\nKysifi7pQEl7RcRL1eQyM7Paqran8L2IWFS6QNLQiFjfR3dkMTBW0miSYjCDZF6idBtjgN9EREia\nCAwFXu7Xd2BmZjVT7UTzJWWWbTb2XyoiNgCzgLuBJ4EFEbFC0lmSzkqbHQ88IekxkiOVTor+THKY\nmVlN9dlTkLQP0AzsJOn9gNJVuwE7V9p42rtY1GPZnJLH3wC+0c/MZmaWk0rDR0cDp5HMB1xRsvx1\n4B9yymRmZgXpsyhExPXA9ZKOj4gfDVAmMzMrSKXho1Mi4kbgAEl/33N9RFxR5mVmZtagKg0f7ZL+\nW/9nZpiZ2VarNHz0H+m/Fw5MHDMzK1Kl4aPv9LU+Ir5Y2zhmZlakSsNHSwYkhZmZ1YVqjj4yM7Nt\nRKXhoysj4suS7qDMZa0j4tO5JTMzswFXafjoB+m/l+cdxMzMildp+GhJ+u8DknYE3kvSY+hIb5xj\nZmaDSLWXzj4WmAP8huT6R6Ml/U1E3JlnODPbcr3da6Klq5obJ9q2qtpLZ/8rMCUiOgEkvRv4KeCi\nYGY2iFR76ezXuwtC6hmSi+KZmdkgUunoo79KHz4iaRGwgGRO4TMkN9ExM7NBpNLw0adKHv8e+Fj6\n+A/ATrkkMjOzwlQ6+ugLAxXErDez22fT0tXS68SpmdVOtUcfDQPOAA4ChnUvj4jTc8plZmYFqHai\n+QfAPiR3YnuA5E5snmg2Mxtkqi0KYyLifOBP6fWQjgWOyC+WmZkVodqi8Fb67yuSxgO7A3vnE8nM\nzIpS7clrcyW9AzgfWEhyJ7bzc0tlZmaFqKooRMQ16cMHgAPzi2NmZkWqavhI0nBJ/yZpqaQlkq6U\nNDzvcGZmNrCqnVOYD7wIHA+cALwE3JJXKDMzK0a1cwr7RsTFJc8vkXRSHoHMzKw41fYU7pE0Q9J2\n6deJwN15BjMzs4FX6YJ4r5NcAE/Al4Eb01XbAV3AObmmMzOzAVXp2ke7DlQQMzMrXrVzCkj6NDA5\nfdoeET/JJ5KZmRWl2gviXQYcDtyULvqSpA9HxNdzS2ZmuenrirOzW3tfZ4NftT2FacChEbERQNL1\nwKOAi4LVhC+LbVYfqj36CGCPkse71zqImZkVr9qewqXAo5LuJzkSaTJwbm6pzMysEBWLgiQBvwQ+\nSDKvAPC1iHghz2BmZjbwKg4fRUQAiyJibUQsTL+qKgiSpkrqkNQpabOehaSTJS2X9LikByUdsgXf\ng5mZ1Ui1cwpLJR1eudmfSRoCXA0cA4wD2iSN69HsWeBjEXEwcDEwtz/7MDOz2qp2TuEI4BRJvwX+\nRDKvEBExoY/XTAI6I+IZAEnzgenAyu4GEfFgSftfk9zm08zMCqJkdKhCI+ld5ZZHxHN9vOYEYGpE\nnJk+PxU4IiJm9dL+HOC93e17rJsJzAQYMWLEBxYsWFAxc9G6urpoamoqOkZF9ZJzbdfaPtcPfXso\n64esH6A0W64RclbKuG/TvgOYpnf18rNZSaPknDJlypKIOKxSu0rXPhoGnAWMAR4Hro2IDbWJuMl+\npgBnAB8ptz4i5pIOLbW0tERra2utI9Rce3s7zlm9SucptHS10NHUMTBhtkIj5KyUsa21bQDT9K5e\nfjYraZSc1ao0p3A9cBhJQTgG+Nd+bHsNMKrk+ch02SYkTQCuAaZHxMv92L6ZmdVYpTmFcekkMJKu\nBR7ux7YXA2MljSYpBjOAz5Y2kLQ/cCtwakSs6se2zcwsB5WKwlvdDyJiQ3LKQnXS9rNI7rswBLgu\nIlZIOitdPwe4ABgOfDfd9oZqxrzMzCwflYrCIZJeSx8L2Cl93n300W59vTgiFgGLeiybU/L4TGCz\niWUzMytGpfspDBmoIGZmVrz+XBDPzMwGORcFMzPLVH3nNTPbNvgGPNs29xTMzCzjomBmZhkPH9mA\n8S03zeqfewpmZpZxUTAzs4yLgpmZZVwUzMws46JgZmYZFwUzM8u4KJiZWcZFwczMMi4KZmaWcVEw\nM7OMi4KZmWVcFMzMLOML4llN+aJ3g5vvtTD4uadgZmYZFwUzM8u4KJiZWcZFwczMMi4KZmaWcVEw\nM7OMi4KZmWVcFMzMLOOiYGZmGRcFMzPLuCiYmVnGRcHMzDK+IJ71my96Z+X4YnmDg3sKZmaWybUo\nSJoqqUNSp6Rzy6x/r6RfSVov6Zw8s5iZWWW5DR9JGgJcDXwSWA0slrQwIlaWNPsj8EXgL/LKYWZm\n1cuzpzAJ6IyIZyLiTWA+ML20QUS8GBGLgbdyzGFmZlXKc6K5GXi+5Plq4Igt2ZCkmcBMgBEjRtDe\n3r7V4fLW1dU1aHO2dLXkE6YPQ98eWsh++6sRchaRcUt+Fwbz71A9a4ijjyJiLjAXoKWlJVpbW4sN\nVIX29nYGa84ijj5q6Wqho6ljwPfbX42Qs4iMba1t/X7NYP4dqmd5Dh+tAUaVPB+ZLjMzszqVZ1FY\nDIyVNFrSjsAMYGGO+zMzs62U2/BRRGyQNAu4GxgCXBcRKySdla6fI2kf4BFgN2CjpC8D4yLitbxy\nmZlZ73KdU4iIRcCiHsvmlDx+gWRYyczM6kBDTDTbwPJlLMy2Xb7MhZmZZdxTMLPc+WJ5jcM9BTMz\ny7gomJlZxkXBzMwyLgpmZpZxUTAzs4yLgpmZZVwUzMws46JgZmYZFwUzM8v4jOZtVPcZpi1dLb7W\nkRWqt5+/er+D3WDlnoKZmWVcFMzMLOOiYGZmGRcFMzPLuCiYmVnGRx8NYj6qyMz6y0XBzOqWb84z\n8Dx8ZGZmGRcFMzPLuCiYmVnGRcHMzDKeaDazhuRJ6Hy4p2BmZhn3FBqcz0Uws1pyT8HMzDIuCmZm\nlvHwUQPwEJFZ/1T6nfFEdO/cUzAzs4x7Cma2zfHhrL1zUagTHiIys3rg4SMzM8u4pzCA3Bswq3/b\n+tBSrj0FSVMldUjqlHRumfWS9J10/XJJE/PMY2ZmfcutpyBpCHA18ElgNbBY0sKIWFnS7BhgbPp1\nBPDv6b8Nq/tTRktXi3sGZoNMud/pan7XG6mHkefw0SSgMyKeAZA0H5gOlBaF6cANERHAryXtIWnf\niFhbqxD+w2xmRdvSv0NFFJM8i0Iz8HzJ89Vs3gso16YZ2KQoSJoJzEyfrpf0RG2j5mIv4KWiQ1TB\nOWurEXI2QkZwTi7kwlpurqWaRg0x0RwRc4G5AJIeiYjDCo5UkXPWlnPWTiNkBOesNUmPVNMuz4nm\nNcCokucj02X9bWNmZgMkz6KwGBgrabSkHYEZwMIebRYCn0uPQvog8Got5xPMzKx/chs+iogNkmYB\ndwNDgOsiYoWks9L1c4BFwDSgE3gD+EIVm56bU+Rac87acs7aaYSM4Jy1VlVOJQf+mJmZ+TIXZmZW\nwkXBzMwyDV0UJH1FUkjaq+gs5Ui6OL18x2OS7pG0X9GZypH0LUlPpVlvk7RH0Zl6kvQZSSskbZRU\nd4f/VbqkSz2QdJ2kF+v9PB9JoyTdL2ll+n/+paIzlSNpmKSHJS1Lc9b0pIJakjRE0qOSflKpbcMW\nBUmjgP8B/L+is/ThWxExISIOBX4CXFB0oF78DBgfEROAVcDXC85TzhPAXwE/LzpITyWXdDkGGAe0\nSRpXbKqyvg9MLTpEFTYAX4mIccAHgbPr9P1cD3w8Ig4BDgWmpkdR1qMvAU9W07BhiwLwf4CvAnU7\nUx4Rr5U83YU6zRoR90TEhvTpr0nOF6krEfFkRHQUnaMX2SVdIuJNoPuSLnUlIn4O/LHoHJVExNqI\nWJo+fp3kj1lzsak2F4mu9OkO6Vfd/Y5LGgkcC1xTTfuGLAqSpgNrImJZ0VkqkfTPkp4HTqZ+ewql\nTgfuLDpEg+ntci22lSQdALwfeKjYJOWlwzKPAS8CP4uIesx5JckH6I3VNK7by1xIuhfYp8yq84B/\nIBk6KlxfOSPi9og4DzhP0teBWcA/DWjAVKWcaZvzSLruNw1ktm7VZLRth6Qm4EfAl3v0uutGRLwN\nHJrOw90maXxE1M2cjaTjgBcjYomk1mpeU7dFISKOKrdc0sHAaGCZJEiGOpZKmhQRLwxgRKD3nGXc\nRHKyXiFFoVJOSacBxwGfiIJOXunHe1lvfLmWGpO0A0lBuCkibi06TyUR8Yqk+0nmbOqmKAAfBj4t\naRowDNhN0o0RcUpvL2i44aOIeDwi9o6IAyLiAJKu+sQiCkIlksaWPJ0OPFVUlr5ImkrSvfx0RLxR\ndJ4GVM0lXaxKSj7tXQs8GRFXFJ2nN5JGdB+pJ2knknvH1NXveER8PSJGpn8rZwD39VUQoAGLQoO5\nTNITkpaTDHfV5aF1wFXArsDP0sNn5xQdqCdJfylpNXAk8FNJdxedqVs6Sd99SZcngQURsaLYVJuT\nNA/4FdAiabWkM4rO1IsPA6cCH09/Hh9LP+nWm32B+9Pf78UkcwoVD/msd77MhZmZZdxTMDOzjIuC\nmZllXBTMzCzjomBmZhkXBTMzy7gomJlZxkXBzMwyLgpmW0nS4em9KIZJ2iW9tv74onOZbQmfvGZW\nA5IuIbm2zE7A6oi4tOBIZlvERcGsBtJrHi0G1gEfSq+eadZwPHxkVhvDgSaSa0gNKziL2RZzT8Gs\nBiQtJLnj2mhg34iYVXAksy1St/dTMGsUkj4HvBURN6f3a35Q0scj4r6is5n1l3sKZmaW8ZyCmZll\nXBTMzCzjomBmZhkXBTMzy7gomJlZxkXBzMwyLgpmZpb5/wGd1a3U0aFCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x8f7f1d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Draw 10000 samples from the standard normal distribution \n",
    "mu=0; sigma=1;\n",
    "p = np.random.normal(mu, sigma, size=100000)\n",
    "\n",
    "# plot the histogram of the sampled data\n",
    "plt.hist(p, bins=50, normed=1, facecolor='g', alpha=0.5, label='my data')\n",
    "\n",
    "plt.xlabel('x')\n",
    "plt.ylabel('Probability')\n",
    "plt.title('Gaussian distribution')\n",
    "plt.text(-2, 0.45, r'$\\mu=0,\\ \\sigma=1$')\n",
    "plt.xlim(-4, 4)\n",
    "plt.ylim(0, 0.5)\n",
    "plt.grid(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color='red'>Exercise</font>\n",
    "\n",
    "Read the documentation of the np.random package and check the available univariate distributions. You'll find several distributions apart from the **np.random.normal** we have used in the previous example. Select four of them, draw N samples, and plot their histograms using subplots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "help(np.random)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Linear Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((47L, 2L), (47L,))"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle # The pickle module implements a fundamental, but powerful algorithm for \n",
    "              # serializing and de-serializing a Python object structure. \n",
    "              # “Pickling” is the process whereby a Python object hierarchy is converted \n",
    "              # into a byte stream that we can use e.g. to save data/load into/from a file.\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Load data for the regression example\n",
    "with open('./P0data1.pkl','rb') as f:  # mode 'rb' means read and binary\n",
    " (X,y) = pickle.load(f)\n",
    "\n",
    "# Our dataset is composed by input data (X) and output data (y) pairs.\n",
    "X.shape, y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x85f8668>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAEWCAYAAACwtjr+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3X2cXVV97/HP1xBgACEJxBSGh6SShoJYInOR3qhFuU1o\nVUgjF8OtJUoKtvriaq2pibSieFtCsWKpQsUiBEQhIoQoYhqJVrQ8TUiUJ2OigDACiSQBgQgk/O4f\nex2y5zBz5pyZ2efx+369zuvss85++O0dOL/Za629liICMzOzIr2q0QGYmVn7c7IxM7PCOdmYmVnh\nnGzMzKxwTjZmZlY4JxszMyuck41Zm5P0ZknrGh2HdTYnG7McSQ9J+l9lZe+V9MNGxVQuxbND0jOS\nnpa0VtI7Bls/Im6NiGn1jNGsnJONWWu6LSL2AsYBlwFLJY0vX0nSLnWPzGwATjZmNZL0+5K+L2mr\npPsknZj77vuS/jL3+eW7ImUulLQx3ZHcI+l16bvdJH1G0i8lPSHp3yV1DRVLRLwEfBnoAl4r6ThJ\nj0r6mKTHgctLZbmYDpJ0vaRNkp6U9Pncd6dLekDSFkkrJB0yVOxm1XCyMauBpLHAN4H/BF4DnAVc\nLamaaqqZwFuA3wP2AU4BnkzfLU7lRwGHAt3AJ6qIZxfgL4FngPWp+HeACcAhwJll648BvgU8DExO\nx7kmfXcS8HFgDjARuBX4WhWxmw3JycbslZalu5atkrYCF+e+OxbYC1gcES9ExCqyH+9Tq9jvi8Cr\ngcMARcQDEfGYJJElhb+JiM0R8Rvgn4C5FfZ1bIrt8XTsP4uIp9J3LwHnRMTzEbGtbLtjgAOABRHx\nbET8NiJK7VF/BZyX4tqeYjgq3d0MGHsV52wGONmYDWR2RIwrvYAP5L47AHgkVV+VPEx2h1BRSkyf\nB74AbJR0qaS9ye4i9gBW5xLcd1L5YG5P8e0XEcdGxHdz322KiN8Ost1BwMMpmZQ7BPjXXAybAQHd\nFWI3q4qTjVltfgUcJCn//87BQF9afpYscZT8Tn7jiLgoIo4GDierkloA/BrYBhyRS3L7pA4Aw1Fp\nKPdHgIMH6TjwCPD+fKKNiK6I+O8KsZtVxcnGrDZ3AM8BfydprKTjgHeS2j2AtcAcSXtIOhSYX9pQ\n0v+Q9MbU7vMs8FvgpXSX9CXgQkmvSet2S5pVQPx3Ao8BiyXtKWl3STPSd/8OLJJ0RIphH0n/u1Ls\nBcRnbcrJxqwGEfECWXL5E7I7kouB0yLip2mVC4EXgCeAJcDVuc33JksqW8iq3p4ELkjffQzYANwu\n6Wngu8CoPxsTETtS/IcCvwQeBd6dvrsBOB+4JsVwbzrPoWI3G5I8eZqZmRXNdzZmZlY4JxszMyuc\nk42ZmRXOycbMzArnQfqS/fbbLyZPntzoMMzMWsrq1at/HRGVHkAGnGxeNnnyZHp7exsdhplZS5H0\ncDXruRrNzMwK52RjZmaFc7IxM7PCOdmYmVnhnGzMzKxwhSUbSdMkrc29npb0YUkTJK2UtD69j89t\ns0jSBknr8iPeSjo6TUO7QdJFabKp0lS616byOyRNzm0zLx1jvaR5RZ2nmVmrWramjxmLVzFl4U3M\nWLyKZWv6ht5omApLNhGxLiKOioijgKPJhmW/AVgI3BIRU4Fb0mckHU42M+ERwAnAxWkKW4BLgDOA\nqel1QiqfD2yJiEPJRts9P+1rAnAO8EaymQnPySc1M7NOt2xNH4uuv4e+rdsIoG/rNhZdf09hCade\n1WjHAz+PiIeBk8iGXie9z07LJwHXpKlsHyQbbv0YSfsDe0fE7ZENUX1l2TalfV0HHJ/uemYBK9MU\nu1uAlexMUGZmHe+CFevY9uKOfmXbXtzBBSvWFXK8eiWbucDX0vKk3NzljwOT0nI32UyBJY+msu60\nXF7eb5s0ze1TwL4V9tWPpDMl9Urq3bRp0/DOzMysBf1q67aaykeq8GQjaVfgRODr5d+lO5WGTagT\nEZdGRE9E9EycOORoC2ZmbeOAcV01lY9UPe5s/gS4OyKeSJ+fSFVjpPeNqbwPOCi33YGprC8tl5f3\n2ybNqb4P2QyCg+3LzMyABbOm0TV2TL+yrrFjWDBr1CeIBeqTbE5lZxUawHKg1DtsHnBjrnxu6mE2\nhawjwJ2pyu1pScem9pjTyrYp7etkYFW6W1oBzJQ0PnUMmJnKzMwMmD29m/PmHEn3uC4EdI/r4rw5\nRzJ7+itaHEZFoQNxStoT+GPg/bnixcBSSfPJ5jI/BSAi7pO0FLgf2A58MM2XDvAB4AqgC7g5vQAu\nA66StAHYTNY2RERslvRp4K603rkRsbmQkzQza1Gzp3cXllzKKbsRsJ6envCoz2ZmtZG0OiJ6hlrP\nIwiYmVnhnGzMzKxwTjZmZlY4JxszMyuck42ZmRXOycbMzArnZGNmZoVzsjEzs8I52ZiZWeGcbMzM\nrHBONmZmVjgnGzMzK5yTjZmZFc7JxszMCudkY2ZmhXOyMTOzwjnZmJlZ4ZxszMyscE42ZmZWOCcb\nMzMrnJONmZkVzsnGzMwK52RjZmaFc7IxM7PCOdmYmVnhnGzMzKxwTjZmZlY4JxszMytcoclG0jhJ\n10n6qaQHJP2hpAmSVkpan97H59ZfJGmDpHWSZuXKj5Z0T/ruIklK5btJujaV3yFpcm6beekY6yXN\nK/I8zcyssqLvbP4V+E5EHAb8AfAAsBC4JSKmArekz0g6HJgLHAGcAFwsaUzazyXAGcDU9Dohlc8H\ntkTEocCFwPlpXxOAc4A3AscA5+STmpkNbdmaPmYsXsWUhTcxY/Eqlq3pa3RI1sIKSzaS9gHeAlwG\nEBEvRMRW4CRgSVptCTA7LZ8EXBMRz0fEg8AG4BhJ+wN7R8TtERHAlWXblPZ1HXB8uuuZBayMiM0R\nsQVYyc4EZWZDWLamj0XX30Pf1m0E0Ld1G4uuv8cJx4atyDubKcAm4HJJayT9h6Q9gUkR8Vha53Fg\nUlruBh7Jbf9oKutOy+Xl/baJiO3AU8C+FfZlZlW4YMU6tr24o1/Zthd3cMGKdQ2KyFpdkclmF+AN\nwCURMR14llRlVpLuVKLAGCqSdKakXkm9mzZtalQYZk3nV1u31VRuNpQik82jwKMRcUf6fB1Z8nki\nVY2R3jem7/uAg3LbH5jK+tJyeXm/bSTtAuwDPFlhX/1ExKUR0RMRPRMnThzmaZq1nwPGddVUbjaU\nwpJNRDwOPCJpWio6HrgfWA6UeofNA25My8uBuamH2RSyjgB3piq3pyUdm9pjTivbprSvk4FV6W5p\nBTBT0vjUMWBmKjOzKiyYNY2usWP6lXWNHcOCWdMG2cKssl0K3v9ZwNWSdgV+AbyPLMEtlTQfeBg4\nBSAi7pO0lCwhbQc+GBGlSuMPAFcAXcDN6QVZ54OrJG0ANpP1ZiMiNkv6NHBXWu/ciNhc5ImatZPZ\n07MmzgtWrONXW7dxwLguFsya9nK5Wa2U3QhYT09P9Pb2NjoMs5axbE2fk5EhaXVE9Ay1XtF3NmbW\nhkpdo0s91kpdowEnHBuQk42Z1axS1+hOSTa+s6uNk42Z1azTu0b7zq52HojTzGrW6V2j/dBr7Zxs\nzKxmnd41utPv7IbDycbMajZ7ejfnzTmS7nFdCOge18V5c47smCqkTr+zGw632ZjZsMye3t0xyaXc\nglnT+rXZQGfd2Q2Hk42ZWY380GvtnGzMzIahk+/shsNtNmZmVjgnGzMzK5yr0czMmlw7jFbgZGNm\n1sTaZbQCV6OZmTWxdhmtwHc2ZvaydqiuaTftMlqB72zMDNhZXdO3dRvBzuqaZWteMaO61VG7jFbg\nZGNmQPtU17SbdhmHztVoZga0T3VNu2mX0QqcbMwMyKpl+gZILK1WXdOO2mG0AlejmRnQPtU11pyq\nurOR1AO8GTgA2AbcC6yMiC0FxmZmddQu1TXWnComG0nvA84CHgRWA+uA3YE3AR+TdC/wDxHxy6ID\nNbPitUN1jTWnoe5s9gBmRMSALYSSjgKmAk42ZmY2qIrJJiK+MMT3a0c3HDMza0dDdhCQ9BpJe6bl\nLklnS1osaf/iwzMzs3ZQTW+0a4B90/KngEOBLcBXiwrKzMzaS8VkI2ke8FrguLT8bqAXeBw4RNJp\nkl5ffJhmZtbKhuog8H3gWeAnZHc3TwDfBAR8MH3/VHHhmZlZOxiqg8DDkv4NWAG8BJwREb+UdDDw\npLs8m5lZNYZss4mIS8iq0l4bEd9KxU8Cpw61raSHJN0jaa2k3lQ2QdJKSevT+/jc+oskbZC0TtKs\nXPnRaT8bJF0kSal8N0nXpvI7JE3ObTMvHWN9qgI0a2rL1vQxY/Eqpiy8iRmLV3m0ZWsrQ7XZ7AUQ\nEc9ExHOl8oh4NiK25tep4K0RcVRE9KTPC4FbImIqcEv6jKTDgbnAEcAJwMWSSmNnXAKcQfZMz9T0\nPcB8YEtEHApcCJyf9jUBOAd4I3AMcE4+qZk1Gw/vb+1uqDubGyX9i6S3lLo/A0j6XUnzJa1g5w9/\ntU4ClqTlJcDsXPk1EfF8RDwIbACOSV2s946I2yMigCvLtint6zrg+HTXM4tsOJ3NaUidlcOI06xu\nPLy/tbuKySYijie7+3g/cJ+kpyQ9CXwF+B1gXkRcV2kXwHclrZZ0ZiqbFBGPpeXHgUlpuRt4JLft\no6msOy2Xl/fbJiK2k3VW2LfCvvqRdKakXkm9mzZtqnAaZsXy8P7W7oYciDMivg18e5j7f1NE9El6\nDbBS0k/L9h2SYpj7HrGIuBS4FKCnp6dhcZh5eH9rd4VOMRARfel9I3ADWfvJE6XRB9L7xrR6H3BQ\nbvMDU1lfWi4v77eNpF2Afcg6Lwy2L7Om5OH9rd0Vlmwk7Snp1aVlYCbZ1ATLgVLvsHnAjWl5OTA3\n9TCbQtYR4M5U5fa0pGNTe8xpZduU9nUysCq166wAZkoanzoGzExlZk1p9vRuzptzJN3juhDQPa6L\n8+Yc6RGYrW0UOVPnJOCG1Et5F+CrEfEdSXcBSyXNBx4GTgGIiPskLQXuB7YDH4yIUovpB4ArgC7g\n5vQCuAy4StIGYDNZbzYiYrOkTwN3pfXOjYjNBZ6r2Yh5eH9rZ8puBKpYUXoTMDUiLpc0Edgr9Rpr\nCz09PdHb29voMMzMWoqk1blHWwZV7Uyd5wA9wDTgcmAsWY+0GSMJ0qwZLFvT59kpzQpWbTXanwHT\ngbsBIuJXpfYYs1ZWepiy9IxL6WFKwAnHbBRVm2xeyHdTzj/gadbKKj1MWUSy8V2Udapqk81SSV8E\nxkk6Azgd+FJxYZnVRz0fpvRdlHWyqpJNRHxG0h8DT5O123wiIlYWGpnVXSf+1V3PhynrfRdl1kyq\nes4mPfdya0QsiIiPAj/Mj7Bsra9TB4Ks58OUHpLGOlm1D3V+nWw+m5IdqczaRKcOBFnPhykHu1vy\nkDTWCapts9klIl4ofYiIFyTtWlBM1gCd/Fd3vR6mXDBrWr82G/CQNNY5qr2z2STpxNIHSScBvy4m\nJGsE/9VdPA9JY52s2jubvwKulvR5QGTD959WWFRWd/6re3Cj2XHCQ9JYp6q2N9rPgWPzM3cWGpXV\nXekHsNN6ow3F3ZXNRkfFZCPpPRHxFUkfKSsHICI+W2BsVmf+q/uV3F3ZbHQMdWdTGinAQ9NYR+rk\njhNmo6lisomIL0oaAzwdERfWKSazpuEZNM1Gx5C90dKcMqfWIRazpuMZNM1GR7W90X6UeqJdCzxb\nKoyIuwuJyqxJuOOE2eioNtkcld7PzZUF8LbRDces+bjjhNnIVdv1+a1FB2JmZu2rYpuNpDdK+rGk\nZyTdJun36xWYmZm1j6E6CHwB+CiwL/BZ4HOFR2RmZm1nqGTzqohYGRHPR8TXgYn1CMrMzNrLUG02\n4yTNGexzRFxfTFhmZtZOhko2/wW8c5DPATjZmJnZkIYaQeB99QrEzMzaV7Xz2ZiZmQ2bk42ZmRXO\nycbMzApXVbKRtIekf5D0pfR5qqR3FBuamZm1i2rvbC4Hngf+MH3uA/5fNRtKGiNpjaRvpc8TJK2U\ntD69j8+tu0jSBknrJM3KlR8t6Z703UVKs7dJ2k3Stan8DkmTc9vMS8dYL2leledpbWjZmj5mLF7F\nlIU3MWPxKpat6Wt0SGYdp9pk89qI+GfgRYCIeA5Qldt+CHgg93khcEtETAVuSZ+RdDgwFzgCOAG4\nOM2lA3AJcAYwNb1OSOXzgS0RcShwIXB+2tcE4BzgjcAxwDn5pGadozStc9/WbQQ7p3V2wjGrr2qT\nzQuSusierUHSa8nudCqSdCDwduA/csUnAUvS8hJgdq78mjRawYPABuAYSfsDe0fE7RERwJVl25T2\ndR1wfLrrmQWsjIjNEbEFWMnOBGUdpNK0zmZWP9Umm3OA7wAHSbqa7I7k76rY7nNpvZdyZZMi4rG0\n/DgwKS13A4/k1ns0lXWn5fLyfttExHbgKbJx3AbbVz+SzpTUK6l306ZNVZyOtRpP62zWHKpKNhGx\nEpgDvBf4GtATEd+vtE3qQLAxIlZX2G+Q7pYaISIujYieiOiZONHDvrWjwaZv9rTOZvVVbW+0GcBv\nI+ImYBzwcUmHDLHZDOBESQ8B1wBvk/QV4IlUNUZ635jW7wMOym1/YCrrS8vl5f22kbQLsA/wZIV9\nWYfxtM5mzaHaarRLgOck/QHwEeDnZG0ng4qIRRFxYERMJmv4XxUR7wGWA6XeYfOAG9PycmBu6mE2\nhawjwJ2pyu1pScem9pjTyrYp7evkdIwAVgAzJY1PHQNmpjLrMLOnd3PenCPpHteFgO5xXZw350jP\nvGlWZ9VOC709IkLSScAXIuIySfOHeczFwNK0/cPAKQARcZ+kpcD9wHbggxFRatn9AHAF0AXcnF4A\nlwFXSdoAbCZLakTEZkmfBu5K650bEZuHGa+1OE/rbNZ4ym4EhlhJ+i+yDgLvA95CVvX144g4stjw\n6qenpyd6e3sbHYaZWUuRtDoieoZar9pqtHeTdXWeHxGPk7WBXDCC+MzMrINUVY2WEsxnc59/yRBt\nNmZmZiVVJRtJv2FnF+VdgbHAMxGxT1GBmZlZ+6j2zubVpeXUI+wk4NiigrLWtWxNHxesWMevtm7j\ngHFdLJg1rSMa5zv1vM2qVfMUA5FZRjYkjNnLOnUcsk49b7NaVFuNNif38VVAD/DbQiKyllVpHLJ2\n/iu/U8/brBbVPmfzztzyduAhsqo0s5d16jhknXreZrWots3mfUUHYq3vgHFd9A3wA9vu45B16nmb\n1aLasdEOlHSDpI3p9Y00fYDZyzp1HLJOPW+zWtQyU+dy4ID0+mYqM3tZp45D1qnnbVaLaoerWRsR\nRw1V1so8XE3jufuwWeupdriaajsIPCnpPWRz2QCcSjaUv9moKHUfLvXqKnUfBiomHCcos9ZQbTXa\n6WSjMz8OPEY2nL87DdioGc70zbU+37JsTR8zFq9iysKbmLF4lZ+DMaujanujPQycWHAs1sGG0324\nludbhnvnZGajo2KykfRvVJi2OSL+76hHZB1pON2Ha0lQfvDSrLGGqkbrBVan14m55dLLbFQMp/vw\nYIlooHI/eGnWWBXvbCJiSWlZ0ofzn81GU+nuopbG/gWzpvWrGoPBE5QfvDRrrGp7o0GF6jTrPEX0\nAiufvrnUoD/YMWpJUNUmJvduMytGLcnGDKhPY3u1xyhPUIOpJjG5E4FZcSo+1Fk2adoewHOlr8hm\nG9i72PDqxw91Vm/G4lUDVkl1j+viRwvf1jLHaIZjmrW6UXmoMz9pmllJPRrbG9Gg704EZsWpefI0\ns1p6gTXzMZrhmGadwsnGalaPUY4bMZKyR282K447CNiQBuqhdd6cIwvttTWcrtAj7Uk2nGOOlHu/\nWaeoatTnTuAOAgMr76EF2V/7zTaEfqvEmdeKMZuVq7aDgKvRrKLhDJDZCK0SZ14rxmw2XE42VlGr\n9NBqlTjzWjFms+FysrGKWqWHVqvEmdeKMZsNV2HJRtLuku6U9GNJ90n6VCqfIGmlpPXpfXxum0WS\nNkhaJ2lWrvxoSfek7y6SpFS+m6RrU/kdkibntpmXjrFe0ryizrOZjcb8LfXuoVVLzPl1n31+O2PH\nqG5xjobBru1bD5voeXes7RTZG+154G0R8YykscAPJd0MzAFuiYjFkhYCC4GPSTocmAscARwAfFfS\n70XEDuAS4AzgDuDbwAnAzcB8YEtEHCppLnA+8G5JE4BzgB6yERBWS1oeEVsKPN+6qaYH02gNvVLP\nHlq1xFy+7tZtLzL2VWL8HmPZ+tyLLdGza6Br+9bDJvKN1X0eMsfaTl16o0naA/gh8NfAlcBxEfGY\npP2B70fENEmLACLivLTNCuCTwEPA9yLisFR+atr+/aV1IuI2SbuQzSQ6kSxpHRcR70/bfDEd52sM\nolV6o1Xbg6kVh16pJeZWPL9qtOt5Wftqit5oksZIWgtsBFZGxB3ApIh4LK3yODApLXcDj+Q2fzSV\ndafl8vJ+20TEduApYN8K+yqP70xJvZJ6N23aNOzzrKdqezC1YuNzLTG34vlVo13Py6zQZBMROyLi\nKOBA4BhJryv7Pmjg1AURcWlE9EREz8SJExsVRk2q/TFqxcbnWmJuxfOrRruel1ldeqNFxFbge2Rt\nLU+k6jPS+8a0Wh9wUG6zA1NZX1ouL++3TapG2wd4ssK+Wl61P0atOPRKLTGP9PxGo/NEEVrx382s\nGkX2RpsoaVxa7gL+GPgpsBwo9Q6bB9yYlpcDc1MPsynAVODOVOX2tKRjUy+008q2Ke3rZGBVulta\nAcyUND71dpuZylpetT9Gs6d3c96cI+ke14XI6vyb8cn0/I/+BSvW8a6ju6uKeSTnV2r36tu6jWBn\nI3wzJJxW+Xczq1VhHQQkvR5YAowhS2pLI+JcSfsCS4GDgYeBUyJic9rmbOB0YDvw4Yi4OZX3AFcA\nXWS90M6KiJC0O3AVMB3YDMyNiF+kbU4HPp7C+ceIuLxSvK3SQQDaZzytRg3X4kZ4s9FTbQcBj42W\ntFKyaReN+tGfsvCmARsKBTy4+O2FHdesHTVFbzSzShrV88qN8Gb152RjDdOoH303wpvVn5ONNUyj\nfvTdCG9Wf548zRqmEZOV5Y/t5GJWP0421lD+0TfrDK5GMzOzwvnOpo20y/M3Q+mU8zRrJ042bWK0\nphRodp1ynmbtxsmmTVQaDbpVf4QHuoNpx/M06wRONm2iVYamr7YKbLA7mPJEU9Js52lm/TnZNJnh\ntkccMK5rwKFfqnlAsl5tILVUgQ12BzNGYscAQyz56X+z5ubeaE1kJKMRD/cByXqOgFztxG8w+J3K\njgg//W/WgpxsmkgtP8blhvtU/EiOWataqvoGu1MpnZef/jdrLa5GayIjbXcZzgOSg+27b+s2Zixe\nVbFqrdbqt1qq+hbMmjbg9AOlYzi5mLUW39k0kWoHphzNWSYHO6agYtXacKrfaqnqG+hO7V1Hd3PB\ninVNN7ummQ3NyaaJVPNjPNptLAMdU/CK+V7Kq9aGU/1Wa1Xf7Ond/Gjh23hw8dtZMGsa31jd15Sz\na5rZ0FyN1kSqGZhytJ8zGeiYA1V1Qf8qt+FW+Q23CszP15i1NiebJjPUj/FoPk9T3uZy4buPYvb0\n7kFn0MxXuY2kq/VwtMpzRGY2MFejtZjBfswDamrHqFQdV011Xr3novHsmmatzcmmxQz0I19SSzvG\nUNVSQ7Wt1HsCMs+uadbaXI3WYvJtLANVY1XbjjFUtVQ1bSv17ILcyInWzGzknGxaUOlHfsrCm17R\nawyqa8eod5vLaPDzNWaty9VoLWwk7RiuljKzenKyaWEjSRj1bnMxs87marQWNpx2DM9yaWaN4GTT\n4mppx/Asl2bWKK5G6yD1HOHZzCzPyaaD+Cl8M2sUJ5sO4qfwzaxRCks2kg6S9D1J90u6T9KHUvkE\nSSslrU/v43PbLJK0QdI6SbNy5UdLuid9d5EkpfLdJF2byu+QNDm3zbx0jPWS5hV1no1Wy3QDA/Ve\nGztGPPv8dg/bb2aFKvLOZjvwtxFxOHAs8EFJhwMLgVsiYipwS/pM+m4ucARwAnCxpNIv4yXAGcDU\n9Dohlc8HtkTEocCFwPlpXxOAc4A3AscA5+STWruodbqB8u7O4/cYCwFbt73oYfvNrFCFJZuIeCwi\n7k7LvwEeALqBk4AlabUlwOy0fBJwTUQ8HxEPAhuAYyTtD+wdEbdHRABXlm1T2td1wPHprmcWsDIi\nNkfEFmAlOxNU2xjunDKlOWL22HUXXnyp/xgE7jBgZkWoS5tNqt6aDtwBTIqIx9JXjwOT0nI38Ehu\ns0dTWXdaLi/vt01EbAeeAvatsK/yuM6U1Cupd9OmTcM8u8YZaYO/OwyYWb0Unmwk7QV8A/hwRDyd\n/y7dqQw0vFddRMSlEdETET0TJ05sVBjDNtIGf3cYMLN6KTTZSBpLlmiujojrU/ETqWqM9L4xlfcB\nB+U2PzCV9aXl8vJ+20jaBdgHeLLCvtrKSMc38/hoZlYvRfZGE3AZ8EBEfDb31XKg1DtsHnBjrnxu\n6mE2hawjwJ2pyu1pScemfZ5Wtk1pXycDq9Ld0gpgpqTxqWPAzFTWVkY6vpnHRzOzelH221zAjqU3\nAbcC9wAvpeKPk7XbLAUOBh4GTomIzWmbs4HTyXqyfTgibk7lPcAVQBdwM3BWRISk3YGryNqDNgNz\nI+IXaZvT0/EA/jEiLq8Ub09PT/T29tZ8nh5rzMw6maTVEdEz5HpFJZtWM5xkUz7WGGTVUL47MLNO\nUW2y8QgCI+CxxszMquNkMwLuOmxmVh1PMTACw5la2W08ZtaJfGczArV2Ha51eBkzs3bhZDMCtXYd\ndhuPmXUqV6ONUC0zZbqNx8w6le9s6sjDw5hZp3KyqSMPD2NmncrVaHVUqm5zbzQz6zRONnVWSxuP\nmVm7cDWamZkVzsnGzMwK52RjZmaFc7IxM7PCOdmYmVnhPJ9NImkT2WRu9bIf8Os6Hm+4HOfocpyj\ny3GOruHEeUhETBxqJSebBpHUW82EQ43mOEeX4xxdjnN0FRmnq9HMzKxwTjZmZlY4J5vGubTRAVTJ\ncY4uxzmxFugjAAAIdElEQVS6HOfoKixOt9mYmVnhfGdjZmaFc7IxM7PCOdmMEklflrRR0r25sgmS\nVkpan97H575bJGmDpHWSZuXKj5Z0T/ruIkmqQ5yflNQnaW16/WkTxHmQpO9Jul/SfZI+lMqb6ppW\niLOprqmk3SXdKenHKc5PpfJmu56DxdlU1zN3jDGS1kj6VvrcVNezQpz1v54R4dcovIC3AG8A7s2V\n/TOwMC0vBM5Py4cDPwZ2A6YAPwfGpO/uBI4FBNwM/Ekd4vwk8NEB1m1knPsDb0jLrwZ+luJpqmta\nIc6muqZpn3ul5bHAHelYzXY9B4uzqa5n7vgfAb4KfCt9bqrrWSHOul9P39mMkoj4AbC5rPgkYEla\nXgLMzpVfExHPR8SDwAbgGEn7A3tHxO2R/etemdumyDgH08g4H4uIu9Pyb4AHgG6a7JpWiHMwjYoz\nIuKZ9HFsegXNdz0Hi3MwDftvVNKBwNuB/yiLp2muZ4U4B1NYnE42xZoUEY+l5ceBSWm5G3gkt96j\nqaw7LZeX18NZkn6irJqtdOvfFHFKmgxMJ/srt2mvaVmc0GTXNFWlrAU2Aisjoimv5yBxQpNdT+Bz\nwN8BL+XKmu56DhIn1Pl6OtnUSfproFn7mV8C/C5wFPAY8C+NDWcnSXsB3wA+HBFP579rpms6QJxN\nd00jYkdEHAUcSPbX6uvKvm+K6zlInE11PSW9A9gYEasHW6cZrmeFOOt+PZ1sivVEuv0kvW9M5X3A\nQbn1DkxlfWm5vLxQEfFE+h/8JeBLwDHNEKeksWQ/4FdHxPWpuOmu6UBxNus1TbFtBb4HnEATXs+B\n4mzC6zkDOFHSQ8A1wNskfYXmu54DxtmQ6zmSRie/XtG4Npn+De8X0L+x8J/T8hH0b4T7BYM3wv1p\nHeLcP7f8N2R1tg2NM+33SuBzZeVNdU0rxNlU1xSYCIxLy13ArcA7mvB6DhZnU13PspiPY2fDe1Nd\nzwpx1v16jvoJdeoL+BrZ7eiLZPWZ84F9gVuA9cB3gQm59c8m6+mxjlyvDqAHuDd993nSKA8Fx3kV\ncA/wE2B52X+IjYrzTWRVED8B1qbXnzbbNa0QZ1NdU+D1wJoUz73AJ1J5s13PweJsqutZFvNx7PwR\nb6rrWSHOul9PD1djZmaFc5uNmZkVzsnGzMwK52RjZmaFc7IxM7PCOdmYmVnhnGys5Uh6puzzeyV9\nvgFxHCvpjjRq7gOSPpnKT5S0cBT2v39plN5mJenNaXTmtZK6hrH9x3PLu0r6gaRdRjdKawZONmbD\ntwQ4M7KhVV4HLAWIiOURsXgU9v8Rsqe766rGH/s/B86LiKMiYtswDvdysomIF8ieUXn3MPZjTc7J\nxtqKpMmSVqUBBm+RdHAqv0LSybn1nknv+6e/ptdKulfSm1P5TEm3Sbpb0tfT2GflXkP2gCyRDf1x\nf9r25Tut3HwhayVtk/RHkvZMgx/emeYYOWmQ03kX8J20nyPS+mvTuU1N5WdL+pmkH0r6mqSPpvLv\nS+pJy/ul4UpK1+fWdF53S/qfqfy4VL4cKJ3He3LH/KKkMWXX+i+BU4BPS7o6lS2QdFeK8VO5dV+x\nL0mLga5UdnVadRlZArN2U8STqn75VeQL2MHOp/XXAr8EPp+++yYwLy2fDixLy1cAJ+f28Ux6/1vg\n7LQ8hmxOmv2AHwB7pvKPkZ5kL4vjE8AW4Abg/cDuqfy9pXhy676TbOiVscA/Ae9J5ePI5sDZs2z9\nKcDq3Od/A/48Le9KNpTL0WRPge8B7E02HPxH0zrfB3rS8n7AQ2l5j1ycU4HetHwc8CwwJX3+/XQt\nx6bPFwOnDXANXr6uwEzgUrLhTF4FfIts/qRB91X6d8jtbwywqdH/jfk1+i/XjVor2hZZ1RWQ3UmQ\nDaUB8IfAnLR8FdlkVpXcBXw5Daa5LCLWSvojskmkfqRsMsJdgdvKN4yIc9Nf5DOB/wOcSvaj3U+6\nC7kAeGtEvChpJtngiB9Nq+wOHEw2F07J/sCm3OfbgLOVzU1yfUSsT3dhN0TEc+k4y4c4V8iS3ecl\nHUWWtH8v992dkc1hAnA8WTK7K12DLnYOKjmYmem1Jn3eiyyhvb7afUXEDkkvSHp1ZPMDWZtwsrFO\nsZ1UbSzpVWQJhIj4gaS3kE0udYWkz5LdrayMiFOH2mlE/By4RNKXgE2S9s1/n6rflgJnxM55TgS8\nKyLWVdj1NrIkVDrOVyXdkeL8tqT3V3u++f2QDbr4BPAH6fvf5r57Nh86sCQiFg1xnDyRtd98sV+h\ndFaN+9qtLC5rA26zsXbz38DctPznZFVXAA+R/XUNcCLZX/hIOgR4IiK+RDaT4RuA24EZkg5N6+wp\nKX8HQCp/u/TyPOxTye4Utpat9mXg8oi4NVe2gmziKqX9TB/gPH5GNjp36Vi/C/wiIi4CbiS7W/gB\nMFtSl6RXk1XVleTP9+Rc+T7AY5ENLf8XZNVWA7kFOFnSa9LxJ6RrVckK4PRS+5ak7rR9pX29mO4q\nS+e5L/DriHhxiGNZi3GysXZzFvA+ST8h+zH9UCr/EvBHkn5MVtVW+iv+OODHktaQ9YL614jYRNbu\n8rW0n9uAwwY41l8A65TNKnkVWZvKjtKX6Qf1ZLIf4FIngR7g02TJ7ieS7kuf+4mIZ4GflxIeWUP8\nvelYrwOujGw66mvJhoS/maxKsOQzwF+n89ovV34xMC9dh8PofzeTP/79wN8D/5muwUqyqr1BRcR/\nks1zf5uke4DrgFcPsa9L03UodRB4K3BTpeNYa/Koz2ZNStKfAUdHxN9Xuf4nyRrcP1NoYAWSdD3Z\nfDA/a3QsNrrcZmPWpCLihvI2oHYmaVeyThpONG3IdzZmZlY4t9mYmVnhnGzMzKxwTjZmZlY4Jxsz\nMyuck42ZmRXu/wOTNSx/5Uv7QAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x8848390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# This dataset is known as the \"house prices\" dataset\n",
    "# the task to be done is to predict the price of a house given some 'features' of the house\n",
    "# our input data (X) contains two features per sample (size of the house, and number of rooms)\n",
    "# the output data (y) contains the price of each sample.\n",
    "# for simplification we are going to use only one feature (the size of the house i.e. X[:,0])\n",
    "\n",
    "\n",
    "# Plot sample points.\n",
    "plt.plot(X[:,0],y,'o') # plot dots ('o') for each sample (house size, house price)\n",
    "plt.title('House Prices')\n",
    "plt.xlabel('House Size (square feet)')\n",
    "plt.ylabel('House Price ($)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to find a model to predict prices (y) for houses that are not in our dataset. In linear regression, the model (h) is a linear function of the input data (x, the house size):\n",
    "\n",
    "$h_\\theta(x) = \\theta_0 + \\theta_1 x$\n",
    "\n",
    "For this we need to choose the parameters $\\theta_i$ minimizing a cost function $J$ e.g. the average squared difference between the predictions ($h_\\theta$) and the real prices ($y$) in our training data:\n",
    "\n",
    "$\\hat{\\theta} = \\underset{\\theta}{\\text{minimize}} {1 \\over 2m} \\sum_{i=1}^m{(h_\\theta(x^{(i)}) - y^{(i)})^2}$\n",
    "\n",
    "where $(x^{(i)},y^{(i)})$ is the i-th training sample, and $m$ (=47) is the number of samples in our training set.\n",
    "\n",
    "This optimization problem can be solved in different ways, in this example we are going to implement the Gradient Descent algorithm. In pseudo-code the Gradient Descent algorithm is formalized as follows:\n",
    "\n",
    "repeat until convergence {\n",
    "\n",
    "$\\theta_j := \\theta_j - \\alpha \\frac{\\partial}{\\partial \\theta_j} J(\\theta)$ (for all j)\n",
    "\n",
    "}\n",
    "\n",
    "where $\\alpha$ is the learning rate, and the partial derivative (the gradient) of the cost function is given by:\n",
    "\n",
    "$\\frac{\\partial}{\\partial \\theta_j} J(\\theta) = {1 \\over m} \\sum_{i=1}^m{(h_\\theta(x^{(i)}) - y^{(i)}) x_j^{(i)}}$\n",
    "\n",
    "Let's see a first version Gradient Descent in python:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def GradientDescent(X,y,max_iterations=100, alpha=1):\n",
    "    m = X.shape[0] # number of samples\n",
    "\n",
    "    # y must be a column vector\n",
    "    y = y.reshape(m,1)\n",
    "    \n",
    "    #initialize the parameters to zero\n",
    "    theta = np.zeros(shape=(2,1)) # only two parameters in this particular case (\\theta_0 and \\theta_1)\n",
    "    \n",
    "    # Repeat for max_iterations (it would be nice to also check convergence...)\n",
    "    for iteration in range(max_iterations):\n",
    "        grad = np.dot(X.T , (np.dot(X,theta) - y)) / m;\n",
    "        theta = theta - alpha*grad\n",
    "    return theta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "theta result:\n",
      "[[ 340412.65957447]\n",
      " [ 106907.5627611 ]]\n"
     ]
    }
   ],
   "source": [
    "#prepare x with a column of ones (this is the x_0 for the bias term)\n",
    "x = np.ones(shape=(X.shape[0], 2))\n",
    "x[:,1] = X[:,0] #discard the 2nd feature (number of rooms)\n",
    "\n",
    "# Scale features and set them to zero mean (standarize)\n",
    "mu = np.mean(x,0)\n",
    "sigma = np.std(x,0,ddof=1)\n",
    "x[:,1] = (x[:,1] - mu[1]) / sigma[1]\n",
    "\n",
    "theta = GradientDescent(x,y)\n",
    "print 'theta result:'\n",
    "print theta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicted price: [ 461393.82683743]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x8aa0978>]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAEWCAYAAACwtjr+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XecVNX5x/HPlyIsKk2RUIUIYlBjQ8RgwwJYwRLFWNCo\nJIpGE8VAzC9YYkRNbLFijKISFY0CgogUjRUUBEVRBAtlBUGq9GX3+f1xz+iwbpndnbszu/u8X695\n7cy5ZZ5L4j577jn3OTIznHPOuTjVynQAzjnnqj9PNs4552LnycY551zsPNk455yLnScb55xzsfNk\n45xzLnaebJyr5iQdLmlepuNwNZsnG+eSSPpK0rGF2i6Q9GamYiosxJMvab2kdZJmSzqpuP3N7A0z\n61SZMTpXmCcb56qmd8xsJ6Ax8AgwSlKTwjtJqlPpkTlXBE82zpWRpJ9Jek3SGkkfSzoladtrki5O\n+vx9r0iROyUtDz2SOZL2CdvqSfq7pEWSvpH0oKSc0mIxswLg30AOsIekoyQtkfRHScuARxNtSTG1\nkfS8pBWSVkq6N2nbryV9Imm1pImSdi8tdudS4cnGuTKQVBd4EXgF2A24AhgpKZXbVD2BI4A9gUbA\nmcDKsG1YaN8f6AC0Av6SQjx1gIuB9cD80PwToCmwOzCg0P61gXHAQqBd+J6nw7Y+wJ+A04BmwBvA\nUynE7lypPNk492OjQ69ljaQ1wP1J27oBOwHDzGyrmU0l+uV9dgrnzQN2BvYCZGafmNlSSSJKCr83\ns1Vm9h3wN6BfCefqFmJbFr77VDNbG7YVAEPNbIuZbSp0XFegJTDIzDaY2WYzS4xH/Ra4JcS1LcSw\nf+jdFBl7CtfsHODJxrmi9DWzxokXcFnStpbA4nD7KmEhUQ+hRCEx3QvcByyXNFxSQ6JeRANgZlKC\nezm0F2daiG9XM+tmZpOTtq0ws83FHNcGWBiSSWG7A3cnxbAKENCqhNidS4knG+fK5mugjaTk/3ba\nArnh/QaixJHwk+SDzeweMzsI6Ex0S2oQ8C2wCdg7Kck1ChMAyqOkUu6LgbbFTBxYDPwmOdGaWY6Z\nvV1C7M6lxJONc2UzHdgIXCuprqSjgJMJ4x7AbOA0SQ0kdQAuShwo6WBJh4Rxnw3AZqAg9JIeBu6U\ntFvYt5WkXjHE/y6wFBgmaUdJ9SV1D9seBIZI2jvE0EjSL0uKPYb4XDXlyca5MjCzrUTJ5XiiHsn9\nwPlm9mnY5U5gK/ANMAIYmXR4Q6Kkspro1ttK4Paw7Y/AAmCapHXAZCDtz8aYWX6IvwOwCFgCnBW2\nvQDcCjwdYvgoXGdpsTtXKvniac455+LmPRvnnHOx82TjnHMudp5snHPOxc6TjXPOudh5kb5g1113\ntXbt2mU6DOecq1Jmzpz5rZmV9AAy4Mnme+3atWPGjBmZDsM556oUSQtT2c9voznnnIudJxvnnHOx\n82TjnHMudp5snHPOxc6TjXPOudh5snHOORc7TzbOOedi58nGOedqsk9fgvcfj/1rPNk451xNtH4F\nPHshPH02vP8EFMS7Fp5XEHDOuZrEDD4cBS//EbZugKP/DN2vglrx9j082TjnXE2xZjGM+z0smASt\nu0Kfe6FZ2heELZInG+ecq+4KCmDGIzD5+qhnc/xtcPDFUKt2pYXgycY556qzbxfA2Ctg0dvw0x5w\n8t3QZPdKD8OTjXPOVUf52+Cdf8Krt0Dd+tDnftj/VyBlJBxPNs45V90smwNjBsLSD2Cvk+DEf8DO\nP8loSJ5snHOuusjbDK/fDm/dBTlN4JcjYO++mY4K8GTjnHPVw6LpMPZy+PYz2O9X0OtmaNA001F9\nz5ONc85VZVvWw9SbYPpD0Kg1nPtf6HBspqP6EU82zjlXVS2YAi9eBWsXQ9dL4Ji/QL2dMx1VkTzZ\nOOdcVbNpNUy8DmaPhF06woUTYPdDMx1ViTzZOOdcVTJ3LLx0DWz4Fg77Axz5x2hqc5bzZOOcc1XB\nd99ESeaTsfCTfeGcZ6HFfpmOKmWebJxzLpuZwQdPwctDIG9TNC7zi99B7bqZjqxMYivzKamTpNlJ\nr3WSrpLUVNIkSfPDzyZJxwyRtEDSPEm9ktoPkjQnbLtHih6BlVRP0jOhfbqkdknH9A/fMV9S/7iu\n0znnijN6Vi7dh02l/eDxdB82ldGzcst2gtUL4cnTYPSl0Gwv+O2bcPjVVS7RQIzJxszmmdn+ZrY/\ncBCwEXgBGAxMMbOOwJTwGUmdgX7A3kBv4H5JiSpxDwCXAB3Dq3dovwhYbWYdgDuBW8O5mgJDgUOA\nrsDQ5KTmnHNxGz0rlyHPzyF3zSYMyF2ziSHPz0kt4RQUwPThcP+h0fMzx98eTQJotmfsccelshZP\nOwb43MwWAn2AEaF9BJB4vLUP8LSZbTGzL4EFQFdJLYCGZjbNzAx4vNAxiXM9BxwTej29gElmtsrM\nVgOT+CFBOedc7G6fOI9NefnbtW3Ky+f2ifNKPnDFZ/Do8TBhELQ9BAZOg0MGxL7eTNwqa8ymH/BU\neN/czJaG98uA5uF9K2Ba0jFLQlteeF+4PXHMYgAz2yZpLbBLcnsRx3xP0gBgAEDbtm3Lc13OOVek\nr9dsKlM7+Xnw1t3wv1uhbgPo+yDs1y9jhTPTLfZUKWkH4BTg2cLbQk/F4o6hOGY23My6mFmXZs2a\nZSoM51w11LJxTurtSz+Ah3tElQA6HQ+Xvwf7n11tEg1Uzm2044H3zeyb8PmbcGuM8HN5aM8F2iQd\n1zq05Yb3hdu3O0ZSHaARsLKEcznnXKUY1KsTOXW3X5wsp25tBvVKWhkzb3O0oNnwHrB+OZz1JJz5\nOOy0W+UGWwkqI9mczQ+30ADGAonZYf2BMUnt/cIMs/ZEEwHeDbfc1knqFsZjzi90TOJcZwBTQ29p\nItBTUpMwMaBnaHPOuUrR94BW3HLavrRqnIOAVo1zuOW0fel7QLijv/AdePAwePPOqBczcDr87OSM\nxhynWMdsJO0IHAf8Jql5GDBK0kXAQuBMADP7WNIoYC6wDRhoZonRtcuAx4AcYEJ4ATwCPCFpAbCK\naGwIM1sl6SbgvbDfjWa2KpaLdM65YvQ9oNUPySVhy3cw+QZ472Fo3BbOGw179MhMgJVIUUfAdenS\nxWbMmJHpMJxz1dn8yTDuKli7BLpdCj2ug3o7ZTqqCpE008y6lLafVxBwzrm4bVwFE/8UVQLYtRNc\n9Aq06ZrpqCqVJxvnnIuLGcwdE9U027QajrgWjrgG6tTLdGSVzpONc87F4btlMP5q+HQctNgfznsh\nKqBZQ3mycc65dDKL1pmZ+CfYtgWOuxG6DYTaNfvXbc2+euecS6fVX8GLV8IXr8Hu3eHke2DXDpmO\nKit4snHOuYoqyId3H4YpN4Bqw4l3wEEXVvl6ZunkycY55ypi+acw9gpY8i50OA5OvgsatS79uBrG\nk41zzpVHfh68eRe8fhvssBOc9jDs+8tqVc8snTzZOOdcWX09C8ZcDt98BHufGq03s5MX8y2JJxvn\nnEtV3iZ47RZ4+5+wU3Po9xTsdUKmo6oSPNk451wqvnoLxl4Oq76AA/tHU5pzGmc6qirDk41zzpVk\n8zqYPBRm/BuatIPzx8JPj8x0VFWOJxvnnCvOZ69EhTO/WwqHXg49/gQ77JjpqKokTzbOOVfYhpXw\n8mCYMwqa7RUtaNa61MLGrgSebJxzLsEMPn4eXroWNq+BIwfD4X+okYUz082TjXPOAaxbCuP/APNe\ngpYHQp+x0HzvTEdVbXiycc7VbGbw/uPwyv9B/lbo+VfodhnUqp3pyKoVTzbOuZpr1Rcw9nfw1RvQ\n7nA4+W7YZY9MR1UtebJxztU8Bfkw7QGY+leoXRdOuit6dsYLZ8bGk41zrmb5Zm70cGbuTNizd1Sh\nuVGrTEdV7Xmycc7VDNu2wpt3wOt/h/oN4fRHYJ/TvXBmJfFk45yr/pbMjHozy+dGlZl7D4Mdd810\nVDVKrDcoJTWW9JykTyV9IulQSU0lTZI0P/xskrT/EEkLJM2T1Cup/SBJc8K2e6ToTxFJ9SQ9E9qn\nS2qXdEz/8B3zJfWP8zqdc1lq60aYeB08cixsWgNnPwOn/8sTTQbEPRp2N/Cyme0F7Ad8AgwGpphZ\nR2BK+IykzkA/YG+gN3C/pMTcwweAS4CO4dU7tF8ErDazDsCdwK3hXE2BocAhQFdgaHJSc66mGT0r\nl+7DptJ+8Hi6D5vK6Fm5mQ4pfl++Dg8cCu/cGw3+D5wGnXqXfpyLRWzJRlIj4AjgEQAz22pma4A+\nwIiw2wigb3jfB3jazLaY2ZfAAqCrpBZAQzObZmYGPF7omMS5ngOOCb2eXsAkM1tlZquBSfyQoJyr\nUUbPymXI83PIXbMJA3LXbGLI83Oqb8LZvBZevBJGnAwI+o+LVs+s3yjTkdVocfZs2gMrgEclzZL0\nL0k7As3NbGnYZxnQPLxvBSxOOn5JaGsV3hdu3+4YM9sGrAV2KeFcztU4t0+cx6a8/O3aNuXlc/vE\neRmKKEbzJsB93aKHNH9xBVz6NrQ/PNNROeJNNnWAA4EHzOwAYAPhlllC6KlYjDGUSNIASTMkzVix\nYkWmwnAuVl+v2VSm9ippw7fw3EXwVD/IaQIXT44qAezQINORuSDOZLMEWGJm08Pn54iSzzfh1hjh\n5/KwPRdok3R869CWG94Xbt/uGEl1gEbAyhLOtR0zG25mXcysS7NmvqSrq55aNs4pU3uVYgYfPgv3\nHgxzx0CP62DAa9DqoExH5gqJLdmY2TJgsaROoekYYC4wFkjMDusPjAnvxwL9wgyz9kQTAd4Nt9zW\nSeoWxmPOL3RM4lxnAFNDb2ki0FNSkzAxoGdoc67GGdSrEzl1t6/zlVO3NoN6dSrmiCpi7RL4z1nw\n/MXQ9Kfw2zfgyGuhzg6ZjswVIe7nbK4ARkraAfgCuJAowY2SdBGwEDgTwMw+ljSKKCFtAwaaWeJG\n82XAY0AOMCG8IJp88ISkBcAqotlsmNkqSTcB74X9bjSzVXFeqHPZqu8B0XDl7RPn8fWaTbRsnMOg\nXp2+bx89K7fYbVmpoABmPgqThoLlQ69b4JDfeOHMLKeoI+C6dOliM2bMyHQYzlWqxEy15AkEOXVr\nc8tp+2Znwln5eVQ4c+Gb0P7IqHBm0/aZjqpGkzTTzEpdWc4rCDhXg5U0Uy2rkk3+Nph2H7z6N6hd\nD075JxxwHkhVr2dWQ3myca4GqxIz1ZZ9FJWa+XoWdDoRTvwHNGwB/LhnlniGCPCEk2W8nrZzNVhW\nz1TbtgWm3gzDj4Q1i+GMR6HfyO8TDdSwZ4iqOE82ztVgWTtTbfF78NAR8PptsM8ZcPl7sM9pP6rQ\nXCV6Zg7w22jO1WilzVSrdFs3RAuaTXsAGraCc56DjscVu3vLxjnkFpFYsqJn5rbjyca5Gq7vAa2y\nY3zji9eimWZrFsLBl8CxQ6HeziUeMqhXpyJn02W8Z+Z+xJONcy6zNq2BV/4Ms56ApnvAhRNg91+k\ndGjW9cxcsTzZOOcy59PxMO4PsGEFdL8KjhoMdct2CyxremauRCklG0ldgMOBlsAm4COiEv6rY4zN\nOVddrV8BEwbBxy9A833hV09DywMyHdV2/Pmd9Cox2Ui6kKjkzJfATGAeUB84DPijpI+A/zOzRXEH\n6pyrBszgw2fg5cHRZIAef4bDroLadTMd2Xb8+Z30K61n0wDobmZFziOUtD9RwUxPNs5loaz663zN\nYhj3e1gwCVp3hT73QrPsHMivMpUVqpASk42Z3VfK9tnpDcc5ly5Z89d5QQHMeAQmXx/1bI6/DQ6+\nOKsLZ/rzO+lX6kOdknYLK2wiKUfSdZKGJdakcc5lp6x4uv7b+fDYCfDSNdD6YLjsnSpRoTmrKytU\nUalUEHiaaKllgBuADsBq4D9xBeWcq7iM/nWevw3euAMe6A7L50Kf++G8F6DJ7vF/dxpkbWWFKqy0\nCQL9gT2Ao8LCZWcBtwHrgd0lnQ/MNrMPY4/UOVcmGXu6fumHUeHMpR/Az06GE/4BOzeP9zvTzJ/f\nSb/SJgi8BmwAPiTq3XwDvAgIGBi2r40vPOdceVX60/V5m6NaZm/eBQ12gTMfh8594vmuSuDP76RX\naRMEFkr6J9GSygXAJWa2SFJbYKVPeXYue1XqX+eLpke9mW8/g/1+Bb1uhgZN0/89rspKaaVOSTsB\nBWa2MXzeEahrZmtijq/S+EqdzpXDlvUw5UZ4dzg0ag0n3wUdjs10VK4SpWWlTkk7mdl6M1uf3G5m\nGwrvU/5QnXNV0udT4cUro+dnul4Cx/yl1MKZruYqbcxmjKTZwBhgZiLJSPop0AM4E3gYeC7WKJ1z\n2WPjqqhw5uyRsEvHUDjz0ExH5bJcaWM2x0g6AfgN0F1SE2AbUdma8UB/M1sWf5jOuawwdwyMvwY2\nroTDr4YjroW69TMdlasCSi3EaWYvAS9VQizOuWz13TfRg5mfjIWf/BzO/S+0+Hmmo3JViC8x4Jwr\nnhnM/g9MHBJNbT72ejj0Cqjtvzpc2aRSQaDcJH0laY6k2ZJmhLamkiZJmh9+Nknaf4ikBZLmSeqV\n1H5QOM8CSfeEB0yRVE/SM6F9uqR2Scf0D98xPzyc6lzWGD0rl+7DptJ+8Hi6D5vK6Fm5mQ7px1Yv\nhCdOhTGXwW6d4dK34LDfe6Jx5RJrsgl6mNn+SVPjBgNTzKwjMCV8RlJnoB+wN9AbuF9Sol7EA8Al\nRBWmO4btABcBq82sA3AncGs4V1NgKHAI0BUYmpzUnMukRIHM3DWbMH4okJk1CaegAKY/BPcfCkve\ngxP+Dhe8BLt2zHRkrgpLOdlIOiysb4OkZpLal/M7+wAjwvsRQN+k9qfNbIuZfQksALqGgp8NzWya\nRQ8FPV7omMS5ngOOCb2eXkSLu60KC7xN4ocE5VxGZUWBzOKs+Awe7Q0Tro1mmF32TjStuVZl/F3q\nqrNUV+ocCnQBOgGPAnWBJ4HupRxqwGRJ+cBDZjYcaG5mS8P2ZUCiaFIrYFrSsUtCW154X7g9ccxi\nADPbJmktUVmd79uLOCb5ugYAAwDatm1byqU4lx5ZWb4+Pw/euhv+dyvssCP0fRD26wfRHWvnKizV\nm6+nAgcA7wOY2deSUnl66zAzy5W0GzBJ0qfJG83MJJVewiAmIfkNh6iCQKbicDVLxgpkFufr2VGp\nmWVzoHNfOOF22Gm3zMTiqq1U+8Zbwy0sg+/L1ZTKzHLDz+XAC0TjJ98k1sIJP5eH3XOBNkmHtw5t\nueF94fbtjpFUB2gErCzhXM5lXNaUr8/bFC1o9vDRsH45nPUknDnCE42LRarJZpSkh4DGki4BJhNV\nDiiWpB0TvZ+QnHoCHwFjgcTssP5E1QkI7f3CDLP2RBMB3g233NZJ6hbGY84vdEziXGcAU0NSnAj0\nlNQkTAzoGdqcy7i+B7TiltP2pVXjHAS0apzDLaftW7kVhhe+Aw8eBm/eCfudDQOnR8sBOBeTlG6j\nmdnfJR0HrCMat/mLmU0q5bDmwAthlnId4D9m9rKk94iS10XAQqKSN5jZx5JGAXOJqhQMNLPEKOpl\nwGNADjAhvAAeAZ6QtABYRTSbDTNbJekm4L2w341mtiqVa3WuMmSsfP2W72DyDfDew9C4LZw3Gvbo\nUflxuBon1arP7YGlZrY5fM4hGuj/Kt7wKo9XfXYQTUuutgtmzZ8M466CtUug26XQ4zqot1Omo3JV\nXFqqPid5FvhF0uf80HZwOWJzLislnn9JTEtOPP8CVO2Es3EVTPwTfPAU7NoJLnoF2nTNdFSuhkk1\n2dQxs62JD2a2VdIOMcXkXEaU9PxLRZJNxnpLZjB3NLw0CDatjopmHnEN1KkX/3c7V0iqyWaFpFPM\nbCyApD7At/GF5aqqqnwbKo7nXzLWW/puGYy/Gj4dBy32h/NegJ/sG9/3OVeKVGej/Rb4k6RFkhYD\nfyRadsC572V9GZZSFPecS0Wef6mUagEjR0K7dtFT/rvvDjdfCvd1hQWT4bgb4eIpnmhcxqWUbMzs\nczPrBnQGfmZmvzCzBfGG5qqarC7DkoI4nn+JvVrAyJEwYAAsXBjdNlu0CG54EL5oApe+Dd2v9MKZ\nLiuUtiz0uWb2pKQ/FGoHwMzuiDE2V8VkZRmWMkjc1krnbcDYqwVcdx1s3Lh9Wx4wcQ0M3yM93+Fc\nGpT2J0+iUoAvLO5KlXVlWMoh3c+/DOrVabsxG0hztYBFi4puX7y46HbnMqS0ZaEfCmX+15nZnZUU\nk6uiYv/FmkHlnfgQR28JgG1b4a27oKFgbRHPynlhWZdlUlkWOl/S2UTrxThXrNh+sWZYRWeUpb1a\nQO77MPYK+OYj6H8YPPwebErqUTZoADffnL7vcy4NUh05fEvSvcAzwIZEo5m9H0tUrsrKWBmWGMX1\n/E2Zbd0Ir90C79wLOzWHfk/BXidA15HR2M2iRVGP5uab4ZxzKi8u51KQarLZP/y8ManNgKPTG45z\n2ScrJj589WbUm1n1BRzYH3reBPUbRdvOOceTi8t6qRbi9Ep9rsbK6MSHzetg8lCY8W9o0g7OHws/\nPTL+73UuzUp8zkbSIZI+kLRe0juSflZZgTmXLTK2/sxnE+H+bjDzMTj0crj0HU80rsoqrWdzH3AN\n8DpwCnAX0CvuoJzLJpU+8WHDSnh5MMwZBc1+Bmc+Dq1LLarrXFYrLdnUSlq35llJQ+IOyLlsVCkT\nH8zgo//ChGuj22dHDobDr4Y6XvPWVX2lJZvGkk4r7rOZPR9PWM7VMOu+hnF/gM8mQMsDoc+90Hzv\nTEflXNqUlmz+B5xczGcDPNk4VxFm8P4IeOX/ID8Pet4cLWxWq3bpxzpXhZRWQeDCygrEuRpn1Rcw\n9nfw1RvQ7nA45R5o+tNMR+VcLLwcrHOVrSAfpj0AU/8KtevCyXdHz86EArfOVUeebJyrTN/MhbGX\nQ+5M2PN4OOkOaNgy01E5FztPNs5Vhm1b4Y1/RK/6DeH0R2Cf070342qMlJKNpAbA1UBbM7tEUkeg\nk5mNizU656qDJTNhzEBY8Qns+0voPQx23DXTUTlXqVJdFvpRYAtwaPicC/w1lQMl1ZY0S9K48Lmp\npEmS5oefTZL2HSJpgaR5knoltR8kaU7Ydo/C6m2S6kl6JrRPl9Qu6Zj+4TvmS+qf4nW6GmL0rFy6\nD5tK+8Hj6T5sajxLV2/dCBOvg0eOhc1r4exn4PR/eaJxNVKqyWYPM7uNaA1AzGwjkGr//0rgk6TP\ng4EpZtYRmBI+I6kz0A/YG+gN3B/W0gF4ALgE6BhevUP7RcBqM+tAtATCreFcTYGhwCFAV2BoclJz\nNVtiyYDcNZswflgyIK0J58vX4YFDowrNB10AA6dBp96lHuZcdZVqstkqKYfo2Rok7UHU0ymRpNbA\nicC/kpr7ACPC+xFA36T2p81si5l9CSwAukpqATQ0s2lmZsDjhY5JnOs54JjQ6+kFTDKzVWa2GpjE\nDwnK1XAlLRlQYZvXRtOZR5wMqgUXjIeT7vyhQrNzNVSqEwSGAi8DbSSNBLoDF6Rw3F3AtWy/rHRz\nM1sa3i8Dmof3rYBpSfstCW154X3h9sQxiwHMbJuktcAuye1FHPM9SQOAAQBtfWXDGiO2JQPmTYBx\nv4f138AvfgdHDYEdGlTsnM5VE6kuMTBJ0vtAN6LbZ1ea2bclHSPpJGC5mc2UdFQx5zVJRaxpWznM\nbDgwHKBLly4Zi8NVrrQvGbDh26ie2Uf/hd32hn7/gVYHVjBK56qXlG6jSeoObDaz8UBj4E+Sdi/l\nsO7AKZK+Ap4Gjpb0JPBNuDVG+Lk87J8LtEk6vnVoyw3vC7dvd4ykOkAjYGUJ53IufUsGmMGHz8K9\nB8PcsdDjOhjwmica54qQ6pjNA8BGSfsBfwA+Jxo7KZaZDTGz1mbWjmjgf6qZnQuMBRKzw/oDY8L7\nsUC/MMOsPdFEgHfDLbd1krqF8ZjzCx2TONcZ4TsMmAj0lNQkTAzoGdqco+8BrbjltH1p1TgHAa0a\n53DLafuWrarz2iXwn7Pg+Ythlz3gt2/Ckdd6hWbnipHqmM22cMurD3CfmT0i6aJyfucwYFQ4fiFw\nJoCZfSxpFDAX2AYMNLPEKO5lwGNADjAhvAAeAZ6QtABYRZTUMLNVkm4C3gv73Whmq8oZr6uGyr1k\nQEEBvP8YvPIXsPzomZmuA7xwpnOlUNQRKGUn6X9EEwQuBI4guvX1gZntG294ladLly42Y8aMTIfh\nstnKz6OZZgvfhPZHRjXNmrbPdFTOZZSkmWZW6up+qfZszgJ+BVxkZssktQVur0iAzlUZ+dtg2n3w\n6t+gdj04+R448HwvNeNcGaQ6G20ZcEfS50WUMmbjXKpGz8qtvCWXy2rZR1HhzK9nQacT4cR/QMMW\n2+2S1fE7lyVSrY32HeGBTmAHoC6w3sz8STVXIYmn+RMPWSae5gcy+wt72xZ4/e/w5h2Q0wR++Rh0\n7vuj3kzWxu9clklpNpqZ7WxmDc2sIdEg/elEM9Scq5BYn+Yvr8XvwoOHw+u3RYUzB74Le59a5G2z\nrIzfuSyU6tTn71lkNFFJGOcqJLan+ctj6waYMBge6Rm9P+c5OPVBaNC02EOyKn7nsliqt9FOS/pY\nC+gCbI4lIlejpP1p/vL6/FV48XewZhEcfAkcOxTq7VzqYVkTv3NZLtWezclJr17Ad0RFMJ2rkLQ9\nzV9em1ZHa8080Rdq7wAXToAT/55SooEsiN+5KiLV2WgXxh2Iq5kSg+ipzuZK68yvT16E8VdHtc0O\n+z0cORjq1o81fudqqlQf6mwN/JOo3hnAG0TFOJcUf1TV4g91Zr/CM78g6kUUV2qm2MS0fjm8NAjm\njuYztecPmy9mdaPOniScK4d0P9T5KPAf4Jfh87mh7bjyhedc2ZU086twkih6SvKHtFk0hoM+uZX8\nrRu4p6Af9209gW3UAZ+y7FysUh2zaWZmj5rZtvB6DGgWY1zO/UhZZn4VTkwt+ZYHuIWDZg2BZp04\nr84/uHvrKVGiCXzKsnPxSTXZrJR0rqTa4XUuUSl/5ypNcTO8impPJCBRwHm1X+GVetdycK1PuT6v\nP1z4Mu+pKdd2AAAZlElEQVSs27XIc/mUZefikWqy+TVRdeZlwFKicv4+acBVqrLM/GrZOIef6mue\n3uGv3FT3Md4v6Eivrbcxaee+UKtWmRKXc67iUp2NthA4JeZYXA1R3hllRc386rFXM26fOI/fPzP7\nh3P9fDeGt3+dDnPvZTN1GZQ3gGfzjySnbh1uCYlpUK9ORU42SCQur3fmXHqVOBtN0j/5oSbaj5jZ\n7+IIKhN8NlrlKOuMsrKe68C6i/h30xE0XvsJuS168ttV/fhobf0iE0ZxCSWdMTpX3aVrNlryb98b\ngKEVisrVeGWZUVaWc9VjK1fUeYHf1nqRtWsbwpmP06pzH14s4fjiFlBLZ4zOuUiJycbMRiTeS7oq\n+bNz5ZHOWmKJYw7SPG6rO5w9ai3lufwj+GveuczuXP4CF17vzLn0S/U5GyjhdppzqUpnLbE9GsE5\nG0bQv/YrfM0unLd1MG8U/JxWFRzk93pnzqVfWZKNc+WSPDbSuEFd6tYSeQU//O1SrlpiC6YwttY1\n1K+9lBH5Pbl921lspH6R5yrrYH9pkwcqwiceuJqqxGRTaNG0BpLWJTYRrTbQMM7gXNVXeLB99cY8\n6tYWjXPqsnZTXtl/4W5cBa/8GWaPpMGue/L6QU/wrxk7smnNJloVMwmgrIubxVXvzBdaczVZSrXR\nagKfjRaP7sOmFnlLqlXjHN4afHTZTjZ3DIy/BjaujApnHjGo1MKZaf3+CsqmWJxLl3TXRnOuXNIy\n2P7dMnjpmqhK809+Duf+F1r8vPK+P02yKRbnKluZV+pMlaT6kt6V9IGkjyXdENqbSpokaX742STp\nmCGSFkiaJ6lXUvtBkuaEbfdI0fq8kupJeia0T5fULumY/uE75kvqH9d1upJV6El9M5g1Eu7rCp+9\nAsdeD5e8mnKiqfD3p1k2xeJcZYst2QBbgKPNbD9gf6C3pG7AYGCKmXUEpoTPSOoM9AP2BnoD90tK\n1CZ5ALgE6BhevUP7RcBqM+sA3AncGs7VlOiZoEOArsDQ5KTmijZ6Vi7dh02l/eDxdB82ldGzcsu0\nvSjlXlxs9UJ44lQYcxnstjdc+lZ066x2nZTiSOyTu2YTKrQtU4ubFfVvUbe22LBlW5n+TZ2rimK7\njWbRYND68LFueBnRCp9HhfYRwGvAH0P702a2BfhS0gKgq6SvgIZmNg1A0uNAX2BCOOb6cK7ngHtD\nr6cXMMnMVoVjJhElqKfiudqqr7TB6/IObpd5sL0gH977F0y+ASQ44e/Q5SKoVSulOIvaxwgzWqDI\nSQSVpfC/ReMGdVm/eRtrNuUBPmHAVW+xjtmEnslMoANwn5lNl9TczJaGXZYBzcP7VsC0pMOXhLa8\n8L5we+KYxQBmtk3SWmCX5PYijnFFKO2p+Yo8VV/ck/o/smIejL0CFk+HDsfCSXdB4zZlirO4fRKJ\nJtMD8cn/Ft2HTWX1xrzttnulAlddxZpszCwf2F9SY+AFSfsU2m6SMjYdTtIAYABA27ZtMxVGViht\n8DrWwe38PHjrLvjfbbDDjnDqQ/Dzs6KeTRnjjD3WNKoqcTqXDnGO2XzPzNYArxLdyvpGUguA8HN5\n2C0XSP4ztnVoyw3vC7dvd4ykOkAjonV2ijtX4biGm1kXM+vSrFnNXguutMHr2Aa3v54Nw3vA1L/C\nXifCwHdhv35FJprRs3KpVUR74TjKGmt5xqLSwScMuJokztlozUKPBkk5REtIfwqMBRKzw/oDY8L7\nsUC/MMOsPdFEgHfDLbd1krqF8ZjzCx2TONcZwNQwVjQR6CmpSZgY0DO0uWKUNpBf7oH+4uRtgklD\n4eGjYcNyOGsk/PIx2Gm3IndPjMPkF/FcWOE4yhJr4ry5azZh/DBuUhkJJ+3/ps5lsThvo7UARoRx\nm1rAKDMbJ+kdYJSki4CFRIuyYWYfSxoFzAW2AQPDbTiAy4DHgByiiQETQvsjwBNhMsEqotlsmNkq\nSTcB74X9bkxMFnBFK20gP61P1S98OxqbWbkADjgPet4EOSVPFixqHAagtvSj0v9liTWTFZ7jqlTg\nXDbyCgKBVxCoBFu+g8nXR7PNGu8OJ98Ne/RI6dD2g8cXWQlWwJfDTix3SHGd17mawisIuOwyfxK8\neBWsy4Vul8HRf44mA6QorkrMXuHZucpRKRMEXA22cRU8/xsYeQbU2wkumgS9bylTooH4xjd83MS5\nyuE9GxcPM/j4BXhpEGxeA0dcC0dcA3Xqlet0cY1v+LiJc5XDx2wCH7NJo3VLo8KZn46DFvtDn/vg\nJ/uUfpxzrsrxMRtXYWVe6MsMZj0BE/8M+VvguJui8Znalfd/M1+czLns5MnGFanMtdBWfQkv/g6+\nfB12PwxOuQd22aMyQ/bFyZzLYp5sXJFSfv6kIB+mPwRTbwLVhpPuhAMv+L5wZkWV1lNJ3l5L+tFD\nn15rzLns4Mmmhkn1NlNKdbuWfwJjLofcGSxrfiS/XXMuHzy3Iy0nv1YpyygX3l5UdYGSrsU5V3l8\n6nMNUpbSLCXW7dq2NSqa+eDhsOoLZhx0Gz2+vpTZa3dMa8mXknpXxW0vNmbnXEZ5sqlBSvvlnay4\n509u7roVhh8Fr94MnU+By9/jyo87simvIKXzQuqFL8tbibpwzP7MjHOZ57fRapCylLQv/PxJ+0a1\neKjNy3R8YwTs1Bz6PQV7nVDieXPXbKL94PHb3a4ryyB+aU/3F7c9oUmDugw9eW8fr3EuC3jPpgYp\nraR94R4HwFuDj+bL3zRk6o5/ouOCR6PCmQOnf59oSjov8KPbaunoXZVUiTrZ5kK9Ledc5niyqUFK\n+uVd1HjOX5+fzpePXQIjTgIrgPPHRlOa6zcq9byFJRJKWXtXt5y2L60a5yCilTaTKzwnby/pO51z\nmee30WqQkkqzdB82dbseR49as7i51iM0/2oNHHo59LgOdmhQ7Lnr1alV6mB94jvLUviytCWlE9uL\nq97sM9Gcyw6ebGqY4n55J34pN2Edf6n7BKfWfotPC9pw6daruHC3M+hbTKIpPAYDUXn+on7xJ5Jb\n4f3TMYjv1Zudy25+G80B0LJRfU6u9TaT6w3ixFrTuDPvdE7eejMfWIcSpzEXNQZjRAknWSKhlHZr\nrLy8erNz2c17Ng7Wfc1zTf5Jiy2vMrtgD67NG8Bn1ub7zSU9hV/cbSojSiTFrfqZ7hliXr3Zuezm\nyaYmKyiA90fApL/QIj+POfv8kdNm7EtBER3e4pJKcbevWjXO4a3BR6c95JLEkcScc+nht9FqqpWf\nw+OnwLiroMV+cNnb7HvGn2jRuOhFzYob+/DbV865VHiyqWkK8uHtf8ID3WHpB3Dy3dD/RWj6U6Bs\nySP5uZnaikZp0jUG45yrXvw2Wk3yzVwYMxC+fh/2PB5OugMattxul1THPooqgpk8CcA555J5sqkJ\ntm2FN/4Rveo3gjP+DXufBio8ZyySythHyksQOOccnmyqvyUzo97Mik9g319C71thx10qfNqyVAJw\nzrnYxmwktZH0qqS5kj6WdGVobyppkqT54WeTpGOGSFogaZ6kXkntB0maE7bdI0V/kkuqJ+mZ0D5d\nUrukY/qH75gvqX9c15m1tm6EidfBI8fClnXwq1Fw+r9SSjSpVGUubsKAQYmVnJ1zNVOcEwS2AVeb\nWWegGzBQUmdgMDDFzDoCU8JnwrZ+wN5Ab+B+SYmR6geAS4CO4dU7tF8ErDazDsCdwK3hXE2BocAh\nQFdgaHJSq/a++B88cCi8cy8cdAFcNg327FXqYZD6mjcl1UNL13o2zrnqI7ZkY2ZLzez98P474BOg\nFdAHGBF2GwH0De/7AE+b2RYz+xJYAHSV1AJoaGbTzMyAxwsdkzjXc8AxodfTC5hkZqvMbDUwiR8S\nVPW1aQ2MvSKa0qxacMH4aJnm+g1TPkWqVZm9CKZzriwqZepzuL11ADAdaG5mS8OmZUDz8L4VsDjp\nsCWhrVV4X7h9u2PMbBuwFtilhHMVjmuApBmSZqxYsaKcV5clPn0J7u8Gs56E7lfCpW9Du8PKfJqy\nVmV+a/DRPypNU9q5nHM1T+zJRtJOwH+Bq8xsXfK20FMpeuH4SmBmw82si5l1adasWabCqJj1K+DZ\nC+HpsyGnKVw8BY67EeqWrwBlaWvepOsY51zNEmuykVSXKNGMNLPnQ/M34dYY4efy0J4LtEk6vHVo\nyw3vC7dvd4ykOkAjYGUJ56o+zODDUXBfV/h0HPT4Mwx4DVodWKHTlqcigFcRcM6VJs7ZaAIeAT4x\nszuSNo0FErPD+gNjktr7hRlm7YkmArwbbrmtk9QtnPP8QsckznUGMDX0liYCPSU1CRMDeoa2Km/0\nrFxOvWUUU/9yFDx/Cavqt4HfvAFHDoI6O1T4/OWpyhxXJWfnXPWh6HdzDCeWDgPeAOYAifV5/0Q0\nbjMKaAssBM40s1XhmOuAXxPNZLvKzCaE9i7AY0AOMAG4wsxMUn3gCaLxoFVAPzP7Ihzz6/B9ADeb\n2aMlxdulSxebMWNGGq48PqPfX8wHo+/kD/oPtSngtm1n8Wyt47n5tP38F7tzLiMkzTSzLqXuF1ey\nqWqyPtms/JxZ957LATaXN/L3Yci2i1liuwGZqbDsnHOQerLxCgLZLn9b9LzMa7fw04JaDNo2gGfz\njyR5ebKiZn0limT62i7OuWzgySabLZsTlZpZ+gHsdRLnfXkqH6798QyvwrO+ChfJTDxkCXjCcc5l\nhC8xkI22bYGpf4XhR8G6r+GXI+CsJ/l170NTmvWV6oOZzjlXWbxnk20WvwtjLodv58F+Z0Ovv0GD\npkDq5f+9SKZzLtt4sskWW9ZHvZnpD0Kj1nDOf6HjsT/aLZXy/8Ut1ewPWTrnMsVvo2WDz6dGhTOn\nPwAHXwyXvVNkokmVP2TpnMs23rPJpE2r4ZU/R/XMdukAF06A3X9R4dOmervNOecqiyebTPlkHIy/\nGjasgMN+D0cOhrr103b6VG63OedcZfFkU9nWL4eXBsHc0fCTfeFXz0DL/TMdlXPOxcqTTWUxgw+e\nhpcHQ95GOPr/oqUAatfNdGTOORc7TzaVYc0iGPd7WDAZ2hwCp9wLzfbMdFTOOVdpPNnEqaAAZjwC\nk6+PejbH3x7NNqvlkwCdczWLJ5u4fDs/WqJ50Tuwx9Fw0l3QZPdMR+WccxnhySbd8vPg7XvgtVuj\n1TL73A/7/wpU3OLJzjlX/XmySaelH0SlZpZ9CJ37RLfNdm6e6aiccy7jPNmkQ95m+N+t8Nbd0GAX\nOPMJ6HxKpqNyzrms4cmmolZ/BU+eASvnw/7nQK+bIadJpqNyzrms4smmonZuCU1/CsffCh2OyXQ0\nzjmXlTzZVFSdHeCcUZmOwjnnspo/8OGccy52nmycc87FzpONc8652MWWbCT9W9JySR8ltTWVNEnS\n/PCzSdK2IZIWSJonqVdS+0GS5oRt90jR05GS6kl6JrRPl9Qu6Zj+4TvmS+of1zU655xLTZw9m8eA\n3oXaBgNTzKwjMCV8RlJnoB+wdzjmfkmJpSYfAC4BOoZX4pwXAavNrANwJ3BrOFdTYChwCNAVGJqc\n1JxzzlW+2JKNmb0OrCrU3AcYEd6PAPomtT9tZlvM7EtgAdBVUgugoZlNMzMDHi90TOJczwHHhF5P\nL2CSma0ys9XAJH6c9JxzzlWiyh6zaW5mS8P7ZUCilksrYHHSfktCW6vwvnD7dseY2TZgLbBLCef6\nEUkDJM2QNGPFihXlvSbnnHOlyNgEgdBTsUx9f4hhuJl1MbMuzZo1y2QozjlXrVX2Q53fSGphZkvD\nLbLloT0XaJO0X+vQlhveF25PPmaJpDpAI2BlaD+q0DGvlRbYzJkzv5W0sKwXlGRX4NsKHJ9NqtO1\nQPW6nup0LeDXk81SvZaU1k6p7GQzFugPDAs/xyS1/0fSHUBLookA75pZvqR1kroB04HzgX8WOtc7\nwBnAVDMzSROBvyVNCugJDCktMDOrUNdG0gwz61KRc2SL6nQtUL2upzpdC/j1ZLN0X0tsyUbSU0Q9\njF0lLSGaITYMGCXpImAhcCaAmX0saRQwF9gGDDSz/HCqy4hmtuUAE8IL4BHgCUkLiCYi9AvnWiXp\nJuC9sN+NZlZ4ooJzzrlKpGjoxFWU/0WTvarT9VSnawG/nmyW7mvxCgLpMzzTAaRRdboWqF7XU52u\nBfx6sllar8V7Ns4552LnPRvnnHOx82TjnHMudp5s0kTS7ZI+lfShpBckNc50TBUh6ZeSPpZUIKlK\nDnhK6h0Kuy6QNDjT8VREUYVtqzJJbSS9Kmlu+P/ZlZmOqbwk1Zf0rqQPwrXckOmYKkpSbUmzJI1L\n1zk92aTPJGAfM/s58BkpPNuT5T4CTgNez3Qg5REKud4HHA90Bs4OBV+rqseoXjX+tgFXm1lnoBsw\nsAr/77MFONrM9gP2B3qHZwOrsiuBT9J5Qk82aWJmr4QabQDT2L7yQZVjZp+Y2bxMx1EBXYEFZvaF\nmW0FniYq3lolFVPYtsoys6Vm9n54/x3RL7YiaxhmO4usDx/rhleVnXklqTVwIvCvdJ7Xk008fs0P\nD5+6zEi5IKvLrLAW1QFEVUKqpHDbaTZRCa5JZlZlrwW4C7gWKEjnSSu7XE2VJmky8JMiNl1nZmPC\nPtcR3SIYWZmxlUcq1+NcnCTtBPwXuMrM1mU6nvIKFU/2D2O1L0jax8yq3PiapJOA5WY2U9JR6Ty3\nJ5syMLNjS9ou6QLgJOAYqwIPMJV2PVVcccVdXZaQVJco0Yw0s+czHU86mNkaSa8Sja9VuWQDdAdO\nkXQCUB9oKOlJMzu3oif222hpIqk3UdfzFDPbmOl4HO8BHSW1l7QDUe28sRmOyQVhocNHgE/M7I5M\nx1MRkpolZp9KygGOAz7NbFTlY2ZDzKy1mbUj+m9majoSDXiySad7gZ2BSZJmS3ow0wFVhKRTQwHV\nQ4HxoZp2lREma1wOTCQafB5lZh9nNqryC4Vt3wE6SVoSitlWZd2B84Cjw38vs8Nf01VRC+BVSR8S\n/ZEzyczSNmW4uvByNc4552LnPRvnnHOx82TjnHMudp5snHPOxc6TjXPOudh5snHOORc7TzauSpK0\nvtDnCyTdm4E4ukmaHqbufiLp+tB+SjoqTUtqkc7KuxWV+HeX1FLSc2k43/WSrgnv/y7p6Iqe02Un\nryDgXMWMAM40sw9CpelOAGY2lvQ8RPoH4OE0nAdJtUNZlQozs6+BM9JxriT/JLrWqWk+r8sC3rNx\n1Y6kdpKmhrWFpkhqG9ofk3RG0n6Jv9JbSHo99E4+knR4aO8p6R1J70t6NtTxKmw3YClE9bHMbG44\n9vueVtJDi7MlbZJ0pKQdwxo174Z1Q4qrSH068HLSOZ+X9LKk+ZJuS7qWsyXNCfHfmnyNkv4h6QPg\nUElfSbolxDJD0oGSJkr6XNJvwzE7hX+398M5fxRb+Df+KLz/V9L1rZA0NLQPkvRe+N/hhqRjr5P0\nmaQ3Cck5/PstBHaRVFS9PlfVmZm//FXlXkA+MDvptQi4N2x7Eegf3v8aGB3ePwackXSO9eHn1UTF\nRwFqE1WC2JVoLZ8dQ/sfgb8UEcdfgNXAC8BvgPqh/YJEPEn7ngy8QVSC/m/AuaG9MdEaSDsW2r89\nMDPp8wXAF0AjorpVC4nqv7UM19+M6G7FVKBvOMaIel6Jc3wFXBre3wl8GK63GfBNaK8DNAzvdwUW\n8MMD4Il/s3bAR4Xi3Z2oWsPuQE9gOCCiP2rHAUcABwFzgAZAw3Dua5LO8TBweqb//+Wv9L/8Npqr\nqjaZ2f6JD6EIamJF0UOJFn4DeAK4jZK9B/w7FIYcbWazJR1JtOjaW1EZL3YgKhezHTO7UdJIol+u\nvwLOBo4qvJ+kjsDtQA8zy5PUk6jg4TVhl/pAW7ZfsKoFsKLQqaaY2dpwzrlEv9h3AV4zsxWhfSTR\nL/bRREn5v4XOkbi9NwfYyaL1ZL6TtCXU+NoA/E3SEURl5lsBzYFlRf3jhe+sDzwLXGFmCyVdEf5N\nZoVddgI6EiW2FyzUD5RU+FbjcqLk6aoZTzauJtlGuHUsqRZRAsHMXg+/WE8EHpN0B1FvZZKZnV3a\nSc3sc+ABSQ8DKyTtkrw93H4bBVxiZksTzUR/wZe0QN0moiSUbEvS+3xK/294s/14nCZxjoJC5ysI\n5zuHqKdzUEiMXxURR2EPAs+b2eTwWcAtZvZQ8k6SrirlPPWJrttVMz5m46qjt4kq1kL0i/ON8P4r\nots4AKcQ3c5C0u5Et5AeJlqd8ECi1Va7S+oQ9tlR0p6Fv0jSiQpdH6K/3POBNYV2+zfwqJm9kdQ2\nEbgicaykA4q4js+IbleV5l3gSEm7hkkKZwP/S+G44jQiWtMkT1IPot5TsSQNBHY2s2FJzROBXyfG\nuSS1krQb0a3JvpJyJO1MdGsx2Z5UzdL8rhTes3HV0RXAo5IGEd2GujC0PwyMCYPlLxPdLoLottcg\nSXnAeuB8M1sRbs09Jale2O/PRAkg2XnAnZI2EvWczjGz/ET+CYnsDGBPSb8Ox1wM3ES0IuKHoZf1\nJdFaSN8zsw1h4L6DmS0o7mLNbKmiadavEvUoxlvFFr8bCbwoaQ4wg9LL5V8D5ClaqRLgQTN7UNLP\ngHfCv8V6ojGq9yU9A3xAdMvsvcRJwm3MDuE7XTXjVZ+dy2KSTiW6nfXnTMcSt3CtB5rZ/2U6Fpd+\n3rNxLouZ2QuFx4CqsTrAPzIdhIuH92ycc87FzicIOOeci50nG+ecc7HzZOOccy52nmycc87FzpON\nc8652P0/s33i04bdH7cAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x868c0f0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot sample points.\n",
    "plt.plot(x[:,1],y,'o') # plot dots ('o') for each sample (house size, house price)\n",
    "plt.title('House Prices')\n",
    "plt.xlabel('House Size (normalized)')\n",
    "plt.ylabel('House Price ($)')\n",
    "\n",
    "# Plot line y = theta_1 * x + theta_0\n",
    "xx=np.arange(-2,4,0.1)\n",
    "yy=theta[0]+theta[1]*xx\n",
    "plt.plot(xx,yy)\n",
    "\n",
    "# Now we can use our model to predict the price of any house given it's size in squared feet\n",
    "hs = 2900\n",
    "hp = theta[0]+theta[1]*((hs-mu[1])/sigma[1]) # remember to standarize x!\n",
    "print 'Predicted price: '+str(hp)\n",
    "plt.plot((hs-mu[1])/sigma[1],hp,'or')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color='red'>Exercise</font>\n",
    "\n",
    "Reimplement the **GradientDescent** algorithm to use the original input data (X) with the two given features (size of the house, number of rooms) instead of using only the first one as in the previous example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from mpl_toolkits.mplot3d import Axes3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((47L, 2L), (47L,))"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('./P0data1.pkl', 'rb') as f:\n",
    "    (X, y) = pickle.load(f)\n",
    "\n",
    "# Our dataset is composed by input data (X) and output data (y) pairs.\n",
    "X.shape, y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAADuCAYAAAAOR30qAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsvXl4HHeZ7/upXtXd6tYuy5ZsyVps2Y6XeIntQA4BAiRc\nlhsOh0kIQxI4MAw7M3DJhRkYOAxJGM6cyXCH5wIJIZmFsGUIgQECITkMZHFiEgc7sWWp1dp39b5W\nddf5Q65ydauXaqm1WKnv8+hxUl316193V33rrff3fb+vIMsyBgwYMGBg7WFa6wkYMGDAgIEFGIRs\nwIABA+sEBiEbMGDAwDqBQcgGDBgwsE5gELIBAwYMrBMYhGzAgAED6wQGIRswYMDAOoFByAYMGDCw\nTmAQsgEDBgysE1jK3N8o6zNgwICB8iHo2cmIkA0YMGBgncAgZAMve/znf/4nO3fuXOtpGDBgELKB\n8tDR0cGvf/3rrG3f+c53eOUrX7lGM1qM73znO5jNZqqrq/F4PBw4cICf/vSnBfe/6qqrOHfu3CrO\n0ICB/DAI2cCGxPHjx4lEIgQCAd773vfyjne8A7/fv2g/SZLWYHYGDOSHQcgGKo6XXnqJq6++mtra\nWvbs2cNPfvIT9bWrr76au+++W/1/bXQtyzKf+MQnaG5uxuPxsHfvXk6fPg1AMpnkk5/8JNu2bWPT\npk184AMfIB6Pl5yLyWTiPe95D/F4nIGBAR5//HHa2tq48847aWlp4dZbb1W3KRgZGeFtb3sbTU1N\nNDQ08OEPf1h97dvf/ja7du2irq6ON7zhDQwNDZWcuwEDemEQsoGKQhRF3vzmN/P617+e6elpvva1\nr3HTTTfpSgk88sgj/Pa3v6Wvr49gMMj3v/99GhoaALjtttvo6+vj+eefp7+/n7GxMb74xS+WHFOS\nJO6++26qq6vp6ekBYHJykvn5eYaGhvjmN7+ZtX86neZNb3oT7e3t+Hw+xsbGuOGGGwB46KGH+PKX\nv8yDDz7IzMwMV111FTfeeGPJuRswoBuyLJfzZ+Bljvb2dtnlcsk1NTXqn8PhkF/xilfIsizLv/3t\nb+VNmzbJ6XRaPeaGG26QP//5z8uyLMuvetWr5G9961vqa/fee6967KOPPir39PTITz75ZNbxmUxG\ndjqdcn9/v7rtiSeekDs6OvLO8d5775XNZrNcU1MjNzQ0yEePHpV/9atfybIsy4899phstVrleDyu\n7v/YY4/Jra2t6riNjY2yKIqLxr322mvlu+++W/3/dDotOxwO2efzFZy7AQMXoItjjQjZQNn48Y9/\nTCAQUP++/vWvq6+Nj4+zdetWTKaLp1Z7eztjY2Mlx33Na17Dhz/8YT70oQ/R3NzM+9//fkKhEDMz\nM8RiMQ4dOkRtbS21tbVce+21zMzMFBzr2LFjBAIBZmdneeqpp7jmmmvU15qamqiqqsp73MjICO3t\n7VgsiyX6Q0NDfOxjH1PnUF9fjyzLjI2NFZy7AQPlwCBkAxXFli1bGBkZIZPJqNuGh4dpbW0FwOVy\nEYvF1NcmJyezjv/oRz/KyZMnefHFF+nr6+Pv/u7vaGxsxOFwcObMGfUmEAwGiUQiS5qjIBTW6G/d\nupXh4eG8i31bt27lG9/4RtbNKB6Pc+WVVxacuwED5cAgZAMVxdGjR3E6nXzlK19BFEUef/xxHn74\nYTUPe+DAAR588EFisRj9/f3cc8896rHPPPMMTz/9NKIo4nK5qKqqwmQyYTKZeN/73scnPvEJpqen\nARgbG+OXv/xlxed/xRVXsHnzZm677Tai0SiJRILf//73AHzgAx/g9ttv58yZMwAEg0F+8IMfFJ27\nAQPlwDhjDFQUNpuNhx9+mJ///Oc0NjbywQ9+kPvvv5/e3l4APvGJT2Cz2di0aRM333wzN910k3ps\nKBTife97H3V1dbS3t9PQ0MCnPvUpAO688066u7s5duwYHo+Ha665ZkW0w2azmYcffpj+/n62bdtG\nW1sb3/ve9wC4/vrr+fSnP80NN9yAx+Phsssu4+c//3nJuRswoBeCXF7XacPLwoABAwbKhy4vi3LN\nhQwYKApZlkmn08BCtFksX2vAgIFsGIRsoCLIZDKk02kkSSKZTKrbBUHAbDarf0pOWBAEg6wNGMiB\nQcgGloVMJoMkSWpULAiCSrhKOkwhai2U/cxmMxaLxSBqAwYwcsgGlgBZlslkMsRiMVWvqxCpLMuk\nUqmSpKoVw2sRiUSw2+24XC6DqA1sJBg5ZAOVhULESlri1KlTHDlyZElEWYhgZ2dn8Xg8WK1WUqlU\n1mva1IeS/jCI2sBGgkHIBkpCS8SZTEZNN8iyXHEyVAjWbDbnnUM6nV70vlqCzs1TGzBwKcEgZAMF\noSgmJElSSVBLmGWmu3RBm3vO3Z6PYIsRdSqVwmKx4HQ6DaI2cEnAIGQDi5CPiHOrzgRByCqPVhAO\nhxkfH6eqqgqXy6XmglcKxYh6amoKu91OU1NT1j5KJJ1P+WHAwFrCIGQDKmRZJhqNIssyFoslLxEr\nyCWvQCCA1+slnU7T0tJCKpViYmKCaDRKOp3GZrOpBO1yuXA6nVit1rzjViLyVojaZDJl3RCUsSVJ\nQhTFrGMMojaw1jAI2QCyLCNJEpIkMTIygt1uV82ASh03NzeH1+vFYrHQ1dWFx+NZRHSyLCOKItFo\nlGg0yuTkJNFoFEmSsFqtWUStyOdWCgq55pJsMaLWpmrsdrtK3AZRG6g0DEJ+GUNbzAHozg3Lsqxa\nYo6NjdHb24vb7VZfy4UgCNhsNmw2G3V1dVmvpVIplainpqaYm5tjamqKsbGxLKJ2uVx5I+pKoRRR\nB4NBJiYm6OnpUaP4YhG1QdYGlgKDkF+GyFfMoRCIyWTKmxuGBXKanJzE5/Ph8XhwOBzs27cv7356\nCSmXqIeGhqiqqqKuro5oNEosFmN6eppoNIooilgslkVEbbPZlvI16IKWqLXpDz1FLxaLxSBqA2XB\nIOSXCZQiDFEUVcLNRxD5FusymQzj4+MMDw9TV1fH5ZdfTlVVFU888UTF56lEn4Uiam3qY2ZmBp/P\nV5CoV0IFop2n9l8FyntmMpmsEnJlX6M60UAxGIS8waEoJkKhEA6HAyisTICFCFnJoabTaUZHRxkd\nHaW5uZnDhw/rjkaXqlEutahntVrVjh1aiKJILBYjEokwOzuLz+cjGo1isVgIBAI4nc6siLrcuen9\nPIWIWhlDSRM999xz7Ny5U424jaIXA2AQ8oaFtpgjnU5z6tQpjh8/XvICVwjZ6/UyPj7Oli1buOKK\nK3Tnb7Ul1KsJq9VKTU0NNTU16rahoSFsNhtOp5NoNMrc3BzDw8OkUinMZnPe1Eeh76cSRTBagpUk\nSSVfPUUv2tSHofzYuDAIeYOhmIa41EWcSqXUjsydnZ0cP358UcXcSqPSZG42mxcRNSwQopL6mJub\nY2RkhGQymUXUTqeT6urqFclRZzKZrN+l3KIXQ6K3MWEQ8gaBnmKOQkgkEvh8Pubm5qivr6e1tZWO\njo6VnfAaw2KxFCTqWCxGNBrF7/czOjpKMplUbxLDw8MqYdvt9iUToJ6IuxhRK+sBqVSKTCbD8PAw\n27dvN4j6EodByJc4FA2xNorSS8SxWIzBwUFCoRAdHR3s3LmT+fn5ot2cVxprke7QwmKx4PF48Hg8\nWdunp6eZmZnBarVmEbXJZFpU8FJVVVVWvrlc5BJ1KpUiEomov7tR9HLpwiDkSxSyLBOLxbKKFvQS\ncSQSwev1Eo/H2b59O7t3786SvZVDiCthMFQpVDr1UVVVxebNm7O2p9NpNaIOBAKMjY2RSCQwmUxZ\nC4lK49OV+K6U9MdSil4UYs4n0TOw+jAI+RKDtpjj5MmTHDx4UPeCWzAYxOv1IkkSnZ2d1NfX65K9\nFYISzVby4i3n/VcThT6n2WzG7XarhTEKtEQdDAYZHx/PIupkMsns7GxFiFqbj86HUkSdm+pSIu7m\n5maj6GWVYRDyJYJ8xRxms1kXefn9fmKxGP39/XR1dS2SjGlRrDAkF2udXljPKEXUgUBArf6Lx+OY\nTCYcDkdWRO1wOHSR33Ikhtp/FYiiyOzsLA0NDUbRyyrDIOR1jFLFHKWq6mZnZxkcHMRms2G32zl4\n8KAu2Ztekl0JQl6vJF+pJwGFqK1WK11dXep2pQNLNBolHA4zOTlJPB5HEIRFqY9coi4VIZeLTCaj\nkq0WRtHLysMg5HWIXEN4yL/ibjabF5nxKLaTPp8Pl8vFnj17cLlcPPnkk7pX9vVGyIVuCOuVVJeD\nlc6Vm0wmqqurqa6uztqej6gTiQQATqcTp9OJIAik0+mKEXOhcfQWveR2ejGZTEiShMvlMopeSsAg\n5HWEfJ05SlXVKYSYyWSYmJhgaGiI2tpa9u/fr1bmafctdcGudcpiI5L5clCMqOPxONFolNnZWcLh\nMCdPngTIm/ooh6jT6XTZxF5MopdMJnnppZdU3xPl5mYUvSyGQcjrAFoN8R//+Ee6u7t15Q/NZjOi\nKDI8PMzIyAiNjY0cOnQIu92+aF+9RFtOyqLQvolEQnV42yhYb2oSrdzObDZjt9vp6urKIupoNMr0\n9DTxeBzQT9RKyqISUG6wSs5ZgVH0kh8GIa8h8hVzKCdqqZNPkiTC4TAzMzO0traWLG/WS8jlqiy0\n+4bDYQYGBtRHakmSsozpq6urcTqdRTuIrNcIeb0RshbaJx8tUefuk0gksoyZYrEYsiwvImpRFCta\noZlOpxeNp7fo5eVG1AYhrwGKFXPkywtrIYoiQ0NDTE5OYrPZ2LlzJy0tLSXfU68iYykpi1AoxMDA\nAJIk0dXVhdvtVscQRZFIJEI0GmV8fFztIGK321WSVgoqFNP39UjI6xl6U1FKzrmpqUndLstyVkQ9\nOztLIBAAFtQ5uUUvS8lR5yPkQihG1HBRS33u3Dm6urrUBcSNQtQGIa8itJ05gCwiVlCIkJPJJD6f\nj9nZWbZu3crx48fx+Xy639tkMunqxlEOIUuSxJkzZzCbzXR2dqpWmUrxgZK2qK+vp76+Xj1OySsq\nJDA/P08sFlOJRSsBWyoJVBqVjJArfcNZzmKeouLQErViylRTU6M66M3Nzam/kbZfovZmWgjlEHKx\neWr/jUajWK3WLLOmfEUvd911F7fddtuKNjeoJAxCXgXk68xR6OLOJeR4PM7g4CCBQICOjg56enqy\nHk+Xq4jIhZ4INRgM0t/fTyQSoaenR1e7p9z3qKqqoqqqioaGBnW7LMuMj4/j9/uJx+PMzs4Si8WA\ni/lPJaLWq9GtFCpNyJWce6XHy2Qyardup9NJY2Nj1ntpUx/am2khoq4EIedC6diioFDRy49+9CM+\n+9nPVvS9VxIGIa8gFMXE2bNn6e7u1vUYpZzA0WgUr9dLNBqlo6ODXbt26ZK9FUI5hFwIfr8fr9eL\nIAh0d3czMjKyaPV/OVAiapfLlWVupF2oikQiWdKvXAvNlSpPriRWgkAr+RRRTGUhCAIOhwOHw1GQ\nqGOxWBZRw4JHiJawV9pFUBtYrPfzQQuDkCuMfMUcgUBA90WYSqWYnp7GbDazfft2Ghsbi8reKk3I\n+eD3++nv78disdDT06Ma74yNjRXsobdU5IvQiy1UKY/UueXJLpeLZDJJJpOhpqZmSab0WqznCFmJ\naCs5XrmEqSVqLWRZVru6iKLI2NgYsVgsax1BG1FX8nNo53apwCDkCqFYMYcSyRY72QKBAAMDA8Tj\ncRobG+nt7S35norsTQ+WQsjz8/MMDAxgtVqzGpkqWGvfiUIaXeUJY2hoiEgkwtmzZ1WvYyXlsRr9\n+ArhUoiQKyl7M5lMeDyerMXn3HWEfESt+FHnErXez6s0AbiUYBDyMqGnmKNQakGWZebn5/F6vVgs\nFrq7u4lEImWRbKUjZGXh8cSJE9jtdnbt2lUwLbFeC0PMZjMejwe3243T6aS5uRm4aEofiUSy+vFZ\nrdYsaZ7L5Vp088zNWS4HK1HqvJ7HyxeMFFtH0HYi1ypzlHSW3W5Xz9NiQU4wGFzkd73eYRDyEpFP\nQ1xosS6XkGVZZmZmBq/Xi9PpzCK9eDyu5kdLoZKELMsyc3NzDAwMIIoiBw8eLJkfLkae61G3W8iU\nXksAk5OTRCIRlQAUgo7H4zidzorM41KIkNcq4hYEAbvdznBQ4jdDYWxmF2/Y3cFmj139nfx+P6Io\ncurUqSyi1qY+rFYroVCoqJHWeoRByGVCuYP7/X5qamryStdyYbFYVOKenJzE5/Ph8XjYt2/footc\nr1643H0LEbJiQqTcHC677DJOnTqla7FuOXnpQlgLHXK+DtfaSE3JUc/OzjI2Npa1OKU8UpdDYOud\nkCtZqQflp0BOj4f59I9fIiktnFsPvTDF//cne2irdWC327FaraRSKXbt2qWu10QiEWKxGJOTk0Sj\nUf71X/+VZ599lnQ6zTe/+U3cbje33367+j15vV6++MUv8u53v5s/+ZM/wefz0dHRwfe//331PLj9\n9tu55557MJvN/OM//iNveMMbADh58iS33HIL8XicN77xjdx1110IgkAymeTd7343J0+epKGhge99\n73vq4rQgCDcDf3XhI35JluX78n12g5B1QlvMkUwm6e/v58iRI7qONZlMTE5O8uKLL9LQ0MDll19O\nVVVVwX1XYqEud19tlF5dXc3evXvLjgDXa8qiUvOw2+3Y7XZVQ22z2di0aVNB2Vc+aV6h0uRKEmgl\n0ymwthEywP1Pj5LOyNQ6FrTDwbjEvz8/xUeu7gDISlUU0rofPHiQBx98kIcffhhJkvj973/PPffc\nw5EjR0in07S2tnL99ddzxx138NrXvpbbbruNO+64gzvuuIM777yTF198kQceeIAzZ84wPj7ONddc\nQ19fH2azmT//8z/nW9/6FkePHuWNb3wjv/jFL7juuuu45557qKuro7+/nwceeIBPf/rTfO9732N+\nfh7g88BhQAZOCoLwE1mW/bmf3SDkEshXzGGz2Rb5xOZDOp1mZGSE8fFx6uvrOXLkSMlFpHKkbOXK\n3pTKwOnpabxeLx6PZ5EJUTlYL+S5GtBKqArJvrTSPK2HRK59pp7S+HKwEuOtZYQcF9OYTZqSaWFh\nm4JSuWP1OJOJyy67jA9+8INZ2x999FG6urpob2/noYce4vHHHwfg5ptv5uqrr+bOO+/koYce4oYb\nbsBut7N9+3a6u7s5ceIEHR0dhEIhjh07BsC73/1ufvzjH3Pdddfx0EMP8Td/8zcAvP3tb+fDH/4w\nsizzy1/+EuBXsizPAwiC8CvgWuC7uXM2CLkAihVzlCIixfBnYmKCLVu2sG3bNpxOp64V/ZXQFitz\nDgaDPPXUU9TU1HDgwIGCRKx8Pj3eyRshZaEXxb6PfBVvcFFDHYlEVPvMaDSKKIq8+OKLWaqPpTZN\nrXTEDZWVipVLyNfubuYfHx8kIWUWzkOTwGt3Xrz56VVPBIPBvDnkBx54gBtvvBGAqakptS1XS0sL\nU1NTwIKkUyFdgLa2NsbGxrBarbS1tS3arhyzdetW4OJ6xdzcnPL6iGYKo0DeaiqDkHOQrzOH3pMz\nlUoxNDTE9PQ0bW1tHD9+HLPZzPDwsK6IGiofISt56/7+fux2e9F0iQJFzlbqpC9EnuttMa8SWGre\nN5+GOhQKMTY2Rltbm9qLT2maajabs6Lp6urqrBLhfFiJFEglUa787I17mpAyGR56YQqLSeBdV7Ry\naNvFhdhSElIFgUBAJUgFqVSKn/zkJ9x+++2L9l8PHs0GIVO6M0cpJBIJBgcH8fv9bNu2jePHj2dd\nIBaLZVGHhUKoxEIdLFykygJifX093d3dxGKxkmSsHbfURVRuQ1Q9qHSEvB6LOWRZLtjiSZKkLP+I\n4eFhUqkUFotlkTRP8WdYiQi5kig3xy0IAm/d18Jb9+U3zZIkSdd6Rz6Vxc9//nMOHjzIpk2bANi0\naRMTExNs3ryZiYkJVSLZ2trKyMjFoHZ0dJTW1lZaW1sZHR1dtF17TFtbG5IkEQwGaWhoUF7X3hna\ngMfzzfllTchaDXFfXx+bNm3C4/HovvBisRher5dwOExHRwe9vb26ZG/FsNyUhdaovqGhQfVHnp2d\nJRKJ6B5XDymudWHIpYpiOV+LxYLH41GrIRWIopjlcRyJRJAkCavVSiwWY3p6Go/Hk1dDXS7WOkos\nBb055FAotEji+N3vfldNVwC85S1v4b777uO2227jvvvu461vfau6/Z3vfCd/8Rd/wfj4OOfPn+eK\nK65QNe5PPfUUR48e5f777+cjH/lI1ljHjx/nhz/8Ia95zWsQBEFRZ7xeEARFxvN64P/NN+eXJSEr\nGmKl7Y1CLJIk6ToZw+Ew8XicF154gc7OTvbs2VP0uKUsvumBdtxMJsP4+DhDQ0M0NTVx+PDhrJx1\nufnm5XgnK/n3pURt6zWHvNal01arldra2kURXyqV4rnnnkOWZSYmJhbZm2qleXrSBsrT4nqGXkLO\nzSFHo1F+9atf8Y1vfEPddtttt/GOd7yDe+65h/b2dr7//e8DsGfPHt7xjnewe/duLBYL//RP/6R+\nf1//+tdV2dt1113HddddB8B73/te/vRP/5Tu7m7q6+t54IEHABT1x/8Anrnwtl9UFvhy8bIi5GLF\nHIpWuBiCwSBerxdJknA4HFx++eV5u3PkolxCLrfJ6MjICMPDwzQ3NxdUcqyEM5zSK02BoioZHR1V\n55YrBVN6wJX6TJVAJYml0mNVitxtNhsmk4m2tjb1BqhoqBUf6tHRUaLRqOrIpl1IzNVQV1pCB5WP\nuMtZ1NNqy10uF3Nzc1n7NDQ08Oijj+Y9/rOf/Wxep7jDhw9z+vTpRdurqqr4wQ9+kHcsWZa/DXy7\n1JxfFoScj4hzT7pChCzLsupyZjKZ6OzspLa2lueee043yeoh+3KRTqcZGxsjGo2STCZLdgypVG5a\nCyVCzmQyjI6OMjIywubNmzly5Ij6PedKwWKxGIIgZJF0dXX1ss1/VgOVml+lZWrAIitKRUOdW5as\naKgjkUhee1OHw6FGyZWY40p8Vr2LermEfClgQxOyHkN4Bbmkqa1gq6qqYufOnVkLMCuVFy4FJQod\nGxujpaUFl8tFd3d3yeMEQdA9B71RqizLBAIBnnzySVpaWjh69CgWi4V0Oo0oigWlYOl0Wl248vv9\njIyMqAtXNpuNRCJBMBisSD60Uqh0ymItFuEKaai19qbBYJBkMsmzzz4LXNRQKzfPcu1NV8ILWW/K\nIhqNLnIIXO9YH2d7hSHLMpFIhGQyicvl0qWYUJQQsiwzNTXF4OAgbrebyy67LO+PWk7UWwlC1hLx\n5s2bVfJTdJOlUG7z0mIRspKvHhgYwOFwLIrO9Xg+51MYiKKolidrPSXsdntWNL1euogsFevN50Mr\nzXO73SQSCfbu3avamypEnWtvqs1RF3rCWQlC1rM+oZzrl9p5sqEIWVvMEQgE8Pv97Ny5U9exJpMJ\nv9/Pk08+SV1dXUm9rtlsXhVCliRJrfbbsmWLSsTlohI5ZEXB4fP5aG5uZseOHYTD4Yq1x7FarXg8\nHubn59XfTbFpVPKhSishWHjMVuw3V9qcfq0X9VYLWgLV2psqMjFlH+0TTq6GWpujXgkLTL2S1PWg\nKy4XG4KQ8xVzWK1WXSSo5GJ9Ph9ms5nDhw/rWqhTHs31oFxCFgSBVCrF6OgoExMTtLa2cuzYsYIn\ntp4LvJwccm7KQlnB9/l8NDY2qguHs7OzK74ir7VpzH3Mzhe9aQsr4vG4Lt21HlSSRNezbljP3Ipp\nqLVdrX0+H4lEgnQ6TV9fX1ZUvZybuJ5zbiVy16uBS5qQldXkfMUcVqu1qK+wEnkqKYB9+/YxNDSk\ni4yhvAi5nBNDaX9+4sQJtm7dWpSIobwijnL1zUqV3+DgIA0NDYukdGtpLlQoetN6HsdiMcLhMCMj\nI6pFozZ6Wyvz8kslQi4X+exN5+fnmZmZoampiWg0ytTUFNFoFEmSsmwzld+l1Hvr/e5CodCiG8al\ngEuakLXeErk/UqEcbyqVYnh4mKmpKTXytFgsJBKJspQQ5UTIeiCKIkNDQ0xNTWE2mzl48KCuaqRy\nCLmcCHlubo6+vj7q6urU4pLljLla0JJCKpWiurqaxsbGLCtNRZ2iOLRp89OFmqduhEU9PVgJb2XF\n2rSQvWlux5BiXa313jACgcAl54UMlzghQ2FSyI2Qk8kkPp+P2dnZguXN5RKy3nLoYtD6XyjzeuGF\nF3Qfr6RDSj0C6iETxQlubGwMj8fDwYMHiz7yXyr2m/msNOGiQ5uSn56amiIej2MymdT2QQpZV/IR\nuNLpj7U0AlrqeMV+E6W1UyQSWWRvarfb1RZdhexNobCx0HrHJU/IhaAQdTweZ3BwkEAgQEdHBz09\nPXl/xHLzvOWkLPIhlUoVvEGsVvNSBYo38sDAADU1NWzZsgWPx1My/7oSXharCa0sTwvtopXiJxEK\nhYhEItTU1BRt9aQH63mBcK3782nXDHI11PF4nLm5Ofx+Pz6fb9HirhJROxyOS7J9E2xgQo5Go8Tj\ncU6dOsX27dvZtWtXyQqxclBuykIpohBFEZ/Px9zcHO3t7XR3dy+6AMrVOC+VkBWt9cDAAG63W7Xk\n9Pl8yyqdVl5bCtZD6XS+RauXXnpJbdIZjUazypSVR2xtNaJys/pN3xwvTUbYUmPnuj3NOKzmipLe\nWhOonvEq0UhWuXmKokgsFlNVOFoNtWJvetddd/Hcc8/hdDr5yle+wmWXXcb+/fv52Mc+xunTpxEE\ngW9/+9vs3LlzVbqF3Hfffdxyyy3nL3yUgt1CYAMQcu6FHwqF8Hq9pFIprFYrR48eXZEFlHJTHIIg\ncPbsWTVS37FjR8F5rZT3hQJt0YvL5VrUSqocc6GVSC+sR8iyjNVqpbq6elEutFD12yMj8MS4iNVi\nRsbE074Af/uW3nWt2MhkMhUtxqk0wecWhWg10YpT29133829997L0NAQra2t/O53v+Pee+/l2muv\n5Yc//CGpVIpYLMaXv/zlVekW8oUvfAHgKCW6hcAGIGQFSnkzQGdnJ3V1dTzxxBNln/x699dLmoo1\nZyQSYctJ9QDiAAAgAElEQVSWLSUj9XLGhvJSFkoj0/7+fpxOZ8G2TXrNhVYqZbHWEXI5KFT9FkmI\n/O3J56hzWJHlDJKU5IWhBA89/gxbqxfWNCwWi+p3vFSsRISsV2mkd7xKE7Ke8aLRKHv27OGmm24i\nGAxy4MAB1ThI6aG4Wt1CXve61/GNb3yjZLcQ2ACEHAwGefHFF7HZbPT09GTZFlqtVlVeowd6FQtQ\nOkJWctfBYJDt27eTTCapr6/XTfblNDrVQ97K4sjY2FjB6kMFJpOpqGRQQTHiXmoUuJ4j5LJu7CyU\n6dvtirm8g4xFpLO7E1t4HJPJxMzMDIODg+o5mluNqOc8XIkIuZLj6S1zrvR4oVCIHTt2ADA4OEhT\nUxO33norp06d4tChQ9x1112r1i0kxyS/YLcQ2ACEbLFY2L17d94uyQpp6iVkZX89F0IhIozFYgwO\nDhIKhejs7FQj4tnZ2RVJQ5SKkP1+P/39/dhsNqqqqti7d29JYlluymK5FVLrMUIud07VdjMHt9Zw\ncjiIw2YmKWVorLays8XDWHKWxsZGNfWR6842MjKSpSzQqj1yZXmXQg55LXLSWmMhSZL4wx/+wNe+\n9jWOHj3Kxz72Me64446s/ddLVd8lT8hut7tgNFeqOCQXCiHreWTLJULFrD4SidDZ2cnu3buzfuCV\nMiMqtK9CxFarld7eXtxuN08++aSuSG8tUxbrYVGvEMq5YAVB4C+v6eTfnhnnxYkwm2vs3HJsK06b\nedFvUMydLbcfnyLLU/Kmlf4N1jvB6+0WopW9tbW10dbWxtGjR4GFlMIdd9yxat1ClLTIBRTsFgIb\ngJCLYSna4nKr76LRKF6vl1gsVtSsvlyS1atxzr0xBINB+vv7MZlMKhHn7lvqgivXfjMXoiji9/tV\n05mNgKWkYBxWM++9cuui7XrH0sryFLIAVB1uNBpldnaWUCjEiRMnsFqti6oRy00XrPcIuRxzeiVC\nbmlpYevWrZw7d46dO3fy6KOPsnv3bnbv3r0q3UI+85nPIOjoFgIbnJCVHLJelEPIkUiEeDzO6dOn\n6erqoqGhoWJdQ8r1Lk6n0yoRC4JAd3d3Xg1mOcbzeglZG51JkqR223a73QwNDSGKoqpO0D56F7op\nrOcIuVJYbhSqEIPH48Fut+NwOOju7s6qRsyV5eVWIxZ6/0shQtYzXm77pq997WvcdNNNpFIpOjs7\nuffee8lkMqvSLeSv//qvee9735u3W4ggCIKsOeEveUIuRoIWi2VJKYtiiEQiDAwMkEwmsVqtXHHF\nFbqinXLIvpwcciqVYnJyUr0oi4nhl0q0xcZT+hIODw8zNjZGa2srR48eJZ1Oq9+Lkh+NRCJqflSW\n5axquPUeTa/XYg4tgSrqgXyyPCU/PTMzQzweB1C/f4WolSq4ShJopcvE9ZrTh8PhrAX+AwcOqB7P\nWqxGt5D3vOc9vOc971FNywVBMMmynAGQcy60S56Qi6GShBwOhxkYGEAURbq6uqivr9edk4XK55DD\n4TD9/f3EYjHq6urYs2dPyXErHSErF3uuQb3SoUWBzWajvr4+q0RWEfRHIhECgUCWhWM8Hmd8fFy3\n4cxq4FIg5HzQyvK0TQIUt7xIJEIwGGRsbIxkMkkikcDr9eJ2uxd1t14P0JuyUDp7rzdciIgzgiC4\ngSuAeiAFzAETlzwhFzuxlY68epGPkEOhEAMDA6TTabq6urKij3JUHGazmVQqpWsexQg5EonQ39+P\nKIp0d3cjiiLBYFDXuJUiZK0LXCaTKdk+qtB7KHlOrVtbIpHg1KlTqgl+JBJZpDaorq5eUe/jSiCd\nkXlxIkxCytDZ6KTBdfEcWQ+FIVq3PC2eeeYZWlpaiMViWc5sStNUbX56LQyS9KQs1nPKS5ZlWRCE\nW4F2oA7oAaqBWmCd9MdZISzHMCgYDDIwMIAsy3R1deU1KlH8LPQS8nIi5Gg0Sn9/P8lkUs1TAczO\nzq5I89J8J7XW86K2tpZDhw5x8uTJikZQNpsNs9mcpffMVRtMTEyo3sfaR+7q6uoVa/lUDomK6Qz/\n81EvfxwPYxLAZjbx2Wt76GxcUAdU0hBoJbyVa2tr8zqzKWmn+fl5otGomnbSfv+5N8pKk6OeFIjy\n/a7jG7YAnAV+LcvyrPaFDUHIhXKeS5G9RSIRTp48iSAIdHV1Fc3JlmtSX06HEYU4o9EoAwMDxONx\nlYiXI6dbqnpCqfBzuVyq58VKIN9FVEhtoPU+npqaUp9klEWsWCyG1WqtSERazhjPDAU4NRZiU7Ud\nQRAIxkXufmKYL7+lVx1rvXpZ5PuchWR5uT4Syo1S+/STyWRIpVKruj4QiUTWtRfyhQ7UAAiCYF7Y\ntJBT3hCEXAjlRMh+v5/z58+TSqW4/PLLsxYEKjF+ueSdSqX44x//SCwWK6riKNfnWM8ctGNqC0tK\nVfhVCnqjqnyG6FpviWAwyNTUFBMTEwiCkBXJLbdkuRiCcQm4GKE5bWbmIhfTVes1H10u8vlIwEVZ\nXjAYJJ1Oc+bMmSy1TTmG9EtBMBjUdf2uFQRBMAG7gJkLfxZBEA4B4oYg5EIRsh7CnJ+fZ2BgAIvF\nwvbt25mamtL9Y65EsUc8Hqe/v1+t9GtsbCx6wVWiV16+/VKplPqkkKtnXkksl1y0i1ihUAiPx0Nj\nY6NKEooBkM/nQxTFrNxobgPVjCwTikvYraayiK+z0YlJgJSUwWoWmI+JXNmZnQKoFIlWyk1NQSXm\npcjybDYbfr+fffv2AWRVI2qbBChPNNpqxHxRv96ngfVuvXlhUe824OMXcso3saBNljcEIRdCIUKW\nZVklYpvNphJOIpFQa9OXM34+lCLkRCLBwMCASsShUChrVXyp4+buWyr6jEQi9PX1EYlEOHz48Lo+\nscuBVrurIDc3qm2gmrE4+HG/yEg4jSQLXF6XYv9+fWS1c1M1//3Kbdz39ChiPMPlbR5uPXYxJ17J\nNEOlZWWVzPnmSujyqW20srxIJML09LQqy9O2d1IWHy/1biHCwh3vncBVQK0gCO3Ap4C3AVMbgpAL\n3dVztytuZ0r7+lwPjHIXASsRISsyo2AwmFVyPTAwoGvcSkXISmQei8XYvn07kiRtGDIuhGK50f/1\n637OTIcZDS2sQZydkomKT/OnB5uyoulCBPHqHQ1c3VOPlJGxmrMJcz2oLPKh0ukPPZrmYrI8pRpR\n6Wwdj8cRRZFz585lpT5yF3IvgW4hFiAD/CnQAdiAVwIdG4KQS0Hr/+twOArmQsvtGlJuDlm7bzKZ\nxOv14vf7s0yIysVyvZO1N4Tu7m4aGxvJZDIMDg6WPZeNApPJxMB8krGwpK7Wp6UMjwxnePcrapCk\nJKOjo6rSIFeSZ7fb1eOs5pXtz7eSZvehhMTZyQhWs8BlW9yLbiylsJwiE5PJtKhJQDgcZnh4mE2b\nNi1ayFVST2fPnlUJW0FHRwdutxuz2YzFYuHZZ59lfn5+1czpv/SlL9Hf33+eC+b0giDcD2xlQeoW\nBD4vy/K/CYLQtaEJWZZlRFHk6aefprq6uqD/r4JyL5Jy+uopee5kMsng4CDz8/Ns376d3t7eZV2c\nS42QU6kUg4ODzM3NLboh6DUX2shwVy1cGhe/E7CYTYgWJ90tFxewtAUuwWCQ8fHxLEmeNjeqjeTW\nIyGn02l1rNFAnI//4EWiqTSyLNPd5OKrb9tFlVU/wa5E2bTNZqO2tjYrAtb24fN6vTzyyCNMTEzw\nox/9iJ6eHmRZ5rHHHsvyq77jjjtWzZz+2WefpaGh4Qo05vSCIHwd+K/AoCzLvxYEwQE0bwhCzpea\nmJ6exuv1kk6n2bNnz4osSpUTnYqiSCKR4OTJk3R0dLBz586KXJTlEnIqlaK/v5+pqamCnUteDn4S\npfDeK7fy6NlZpEwGkyBgEsBqEmitze4zWKjARRTFvL4SDoeDZDLJ9PR00S7XelHpCFkh0K897iOY\nkKi2L6w7nJ2K8PAfp/lvBzfrHm+1vJAF4WIfvo9+9KPE43H279/P9ddfz/nz51W/CS1W05y+vr6e\nCyT8K+BGQRBOyLL8LPAtZT6yLMeBJzcEISuQZZmpqSkGBwfxeDwcOHCAl156qeyTQu8jpZ6UhdLM\ndGZmBpPJxLFjx3RfQHqtMvUgnU4zNzfH/Pw8XV1di7puL2XMYvNcx6J8XdjRXM3/evsePvvwOTIZ\nGTISd/7fu/BU6TuXrFZr3kgukUjw3HPPZXW5NpvNWYtX5ZQrVzpCVgh5PJjEZr74xLSwLbHk8SoB\nvcZCSg7ZYrGoT37XXHMNZrOZP/uzP+P973//WprTtwPbBUE4DvQDA0CIhfZO0oYh5ImJCXw+HzU1\nNVx++eVqx+RyF+rK6RpSLEJWmplOT0/T3t7O8ePHeeqpp3RfPOXMoxgymQyjo6OMjIzgdrtpa2tj\n27ZtyxpTCyWaXu8ELMsyfdNR+meiuKssHN5Wi9NW/Lu9qrueRz96lLmoyOBLp7iiY3kLRcoCltVq\nZfv27ep2bYHL9PQ0kUhELVfWpj3yycFWKh+9v9XNL1+cwWoWkOWFue/dUt5TZqUleXqNhXIX9X73\nu9/R2trK9PQ0r3vd6+jt7c3af5Wr+uaAPwBvBvYBMSAJOIA9G4KQw+EwgUAgi4gVLNWkfqltnCRJ\nwufzMTU1xbZt2xZFouWaES2VkGVZZnx8HJ/Px6ZNmzh69Cjz8/O6fS/0olB6Y72R9ImhID96bgKb\nRUBMy/xhJMSfX9WO3VL8Bmk1m2jx2BkxrdxnyVfg8oR3nm89MYyUDvDmnhQ7PAsubUrVokLSkiSt\nSIT8wf/SwVQ4xfOjIQDefqCFV+9oKHZ40fEqgXLM6bWl34qJfHNzM9dffz0nTpxYU3N6WZZ/Dfxa\nEIRaYM+F7WHgcxuCkGtqati1a1fe15ZqUq+na4h2bEmSGBoaYmJiIi8Rw0WS1XOXL8cTWQslbeP1\nemloaODIkSNqlFJOvlkZS0/KREvISv5+aGhIFfy73e4lmaVXCrIs84sXp2ly21QCHg8k6Z+Jsmfz\n+iuxfcLr52M/eJGEtPBbnZ6M8w9v380rLqsnnU6rLm1zc3OEQiGef/75RZ7HSzH/0UbITpuZv7t+\nF5GkhNVsKnnjyoe1NKdXImSl+MTtdhONRnnkkUf43Oc+pxrKr4Y5vd/vp76+vo4L5vSCIJhlWU6z\nIHmrB34my3JEEITqDUHIxbCSXUMUfwqv18v4+Dhbt27l+PHjBU/Ccgi5HE9kuCjt6+/vp6amhoMH\nDy56WtBLyOGExL+fT/HtgeepdVj476/soLclP3Fpx5yfn+f8+fO4XC56enqQJCnvopZC0lqJ2EpD\nyshkKdAEeSE3vA5x39OjKhkDJKQM33lqlFd01WM2m7PkYOFwmH379pHJZLI8p6PRKEBez+lC37dW\nZaGg2r50ilgrQg6Hw+rTxtTUFNdff716/Dvf+U6uvfZajhw5smrm9EeOHAF4BvgiELhQqfcW4Drg\nz4B3Af8G3LHhCXmpKYtSSKfTDA8PEw6H2bx5c1EiVrBSffUkSeLEiRM4nU72799f8LFOLyF/63c+\nBoIZdrTZiKbS/P2v+/nSW3fT7F781CAIAuFwGJ/Ph8lkYs+ePbhcLtVqNNdnQisRUzx4LRaLShhu\nt3tF+vRdub2Ox/rmqHVaSKQyuGwW2htKP/6uBfJ9/ELfiBLVWiyWvJ7TSjSt9ZzWft9aT4lKrFlo\nsVbtoDKZjErcnZ2dnDp1atE+DQ0Nq2lOD9ANqpkQwAeA/wdIXPgD2LwhCLlU1xC9WmFl/2KEnE6n\nGRkZUXNILpdLFYGXQqUJORgMcv78eURR5MCBAyU9OPQQcjoj88fxEHV2MJkE3FUWYiGJobnYIkKO\nxWKqUX5vb28W+ebLLRdybRNFcVFHkWeeeSYrunO73ctaIHr9riacNjNnJsJ46qy8YVejbsXEauOW\nY608NxJUo+Qqi4lbNKXXWhRLKxXyPNZ+31pPCUEQsNls6ve+XM/ptYiQZVle75JNZXLihb8tgPfC\nto2TsqikBWc+Qs5kMioRb968mWPHjmGxWJiYmNA9dqUIORwOc/78eWRZZseOHZw7d25ReiIf9BCy\nSQCXzUwwCiAjy5CRwWW/eGElk0kGBgYIBoNFKx/1wmq1UlNby/mgwDx2oukw/+XQITW68/v9DA8P\nI4oiNpstK7rTmgEVg9kk8KqeBl7VU97C1Frgys56/uHtu/nOUwuLSDcfa+OVXfUF9y+XNK1WK3V1\ndYs8jwcHB0mlUnk9pwsVuBTDaumQtVA4YD0tKGuh2GwCP2Qhp3wYuEwQhNcA8xuGkAthuTlkrWxM\n26ZoqXNZjnexYlKfSqXo7u5WL6hKtmYSBIH3XNnO3/54jolgAhk43F5H7yY3kiQxODjI9PS0Wt2X\n73GwXMiyzD/8ZpCnBv3IQCyWwtUyy7V7mrOiu0JmQKtprblc6I3eXtFVzyuKkHClIQgCZrOZ2tpa\nWlpa1O1aSd7k5CSRSCTLc1orycslwUpHyHo017FYbFUsYpeDC22c/lkQhPcCJ4C/BJ4DPrRhCLlY\nhLyUriGZTIaxsTG1dr5QmyKlzFhXlFamSb1C3vF4nIGBAaLRKN3d3VlGOLn7FoNe4j64rZZb9zqo\nad1KjdPO7s3VDA8PMTo6ukhBUqi7SDk4PxPjaV+ABtfCgtOcGOM7T4/y2t7GLP+EQmZAWmvNmZkZ\nBgcHkSRJVSRkMpmKVMVVAutNDqhFvhxyMc9prUNbLBZTqxYVkl4N8/xcBAKBdW+KdcFy81WyLN8D\n3KNsf9moLMpJWZjNZvx+P5OTkzQ3N5fsF1duX71yjICSySQvvfQSgUCArq4umpqalmVSX47sbVO1\nhf0dtczMzPD0U6dpaWlR0zRaVML3IpZKYzJdFOebBZBkSIiZkoY2UkYmLQt5rTX7+vqwWCyLquK0\nkV11dfWqNsNcz4ScT2WRD4Uc2nJvjPF4nBMnThT1nK40LgGnNwV/e0FpEWYhr+wGvvWyIGQ9UWkm\nk2FiYoKBgQHMZnOWfrcYVkI5IYois7OzhEIhent7SxoQVTpCVkyZnnnmmUVa5lxUwvdie4MDu9lE\nKCHitFkIpWDvZifV9uJE+e/PT/Ivz4yRzsgc3lbDX7y2U62+EwQBq9WqGtQrUKR4ihxPaaKaKw9b\nKTleJfvpVRrLVVnkek4Hg0EOHz5c0HNaq5lWvvNC0HsjW+/m9ACCILiARlmW5zXbAsCODUPIxVaa\nixGGLMtq2XVDQwN79+5leHhY94p+ubrlYtG6UlwyOTlJTU0N7e3tar19MVQyQvb7/fT19ZFKpdi/\nf3/Wwk+hMZdLyDUOK597Yw//9Fsf0+EUXbUmPv26rqIX4LPDAe4/MYqnyoLFJPDscJB7nhjmI1dv\nL3gMLPxe+TwmlAXEXDmeopdOp9MVeQSvtKF8JVHpnC8U95xWomm/38/IyAipVEpt85TrOa13bpcC\nIQNm4KwgCO8EnmBBbbEXCGwYQi4X2lb29fX1HDp0CLvdTiKRWFGT+kRisUGLIqUbGxujra2N48eP\nqxGyHugtIilGnopyAxbE8f39/bqqFStl1dnZ6OR/vm03sNCKvtZZfFHu3NRC4YOS0nBXWXhhLLyk\n91YWBXMd27SRndLSCli0gFiOHK/S/fQqiUo7xxVDPr9jyP7OFQmkLMvY7XZSqRSzs7NFn2AuhZSF\nLMshQRC+BtwGHACagJ3AX2wYQtZ7kmsd4WpraxdVtK1kZV8+BYeycLh58+YsBcdKdJPOh3g8zvnz\n50kkEuzYsUM9mfUSbaGUxUpbeDa6rGQuaE4FQSAuptlaW1r6Vw60LYempqY4cuRIVmQ3NzfH0NCQ\n2psvN7LLd06uV3N60J9D1oOlpj/ytXnKZDLMz88zNDSUt6BIm/K4FAgZQJblRwVB+E/gNSzkkZ+Q\nZXlj99RTYDKZkCSJubk5vF7vIkc4LVaya4gytpImGRwcLLhwWE7pdLkeFbAQiQwMDBAIBNROIVqi\n0DvmWnknv3pHI4+fn+fcdBQTUGU18YGr2lf8ffNFdoocLxwOqwtaWtWBkvZQlAeVJORK5qMrWalX\nSQ2yyWRSUxldXV3q9lzP6c985jP09fXR0tLC9PQ0+/bt401vehNHjhyhtbWVn/70p6veKQTgr/7q\nr7j55psRBOEgC2ZCvwI+CZwFpgAr8ApBEEY2DCEXOjFlWSaTyXDixAlqamo4cOAADoej7HEKoVzl\nRCQS4amnnqKurq7oYtlKlVnLsqwa1BfrWKI3N7yUm0ElYLOY+B9v3snp8TBJKc3O5uqSaY6VgjZP\nql1AVFQH4XBYbTckiiLpdBqv16sS9VIr4mRZrnhp8ko4x1UC+Qg+13P6wQcf5DOf+QwHDx6kpqaG\ns2fP4vV62bVrl5r+W+1OIYIgcOjQIW655ZY6FnyPR4AqFnyR21gwF3Je+O9/3zCEnAvFbEe5CPbs\n2ZN1sVQKeiJkpblqX18foihy9OjRkpV15RJ9KVJUKg2j0ShWq7WoQT0sP2WxGrCYBA60FS8XX0vk\n63QdDocZHBzE7XbnrYhTSFrxlyiG9Rwhr5U5fTgcZvfu3Rw9epTR0VFuvvlmPvvZz/L3f//3wOp3\nCgF43etexze/+c1rZVn+LtAvCMJR4G9kWT6TO/8NR8jaztJOp5N9+/YxODi4YtaPZrNZNdLJB7/f\nz/nz56mqqmLXrl14vV5dZc7l5IXNZnNBvw5temTTpk24XC7a20s/2q91ymKpudbpcJLf9s8TSaap\nluJcVb3+7DWtVitNTU1ZGl6tHE/rL6HI8RSi1rq1VTqHDJUrOV4JYyE913AoFFJVFh//+Mf5yle+\nQjh8cbF3LTqFXDi+VRAEiyzLEnAVUAP8de78NwwhC4KgEnFVVVWWv0K5xSEKGek52QtFyMFgkP7+\nfkwmE7t27cLtXig91ptvXm4OWWvHWVtbq6ZHpqenlzym3v0ymQzj4+MAuN3ugoUAsiwTTqaxW5bm\nt5uLUELiB3+YwGwScFhNnJpMYndGeXNzU+mDVwmFZG/55HjaBqr5pGFWq7VicjxlbpXCSjU4LQXF\nnP6nP/0pzc3NHDp0KNckXsUqdwqBi8ZCaeDVgiB8ADjDQiojAoxtGEIWRZGxsTF27969yN1qKeXT\nysp5KeSWQ0ciEc6fP086naanpydLE7lSeeFcUgwEAvT19VFVVbUoZ6631Ftv5KvdT5ESer1eGhsb\nMZlMDA0NEY1GFy1wZSxV/P+/G2VwLoYgwDsObuaa3qasMcu9WCaCCVJShi0X1BaNThNnp+O8uaxR\nFqOSRFVOmqFQA1VFGjY7O0ssFssrx3O73Wvq57GW3UJqa2v5/e9/z09+8hP+4z/+g0QiQSgU4l3v\neteadAq5cPwYoPzwZ4BXAzeyQM4moBP4yIYhZJvNxr59+/K+ttJdQxTPhP7+fhKJhGpUnYtyCKac\nxTKFvCORCH19fQBqVF5o3FKEXG7KYn5+nr6+PjweD4cPH1bnpHzm3AWu7/xhjgF/hkaXGcFi5f4n\nh9lcbWFPW92Soxar2YR2xmJGxm5dX1VxlZC9KdIw5TfcsWPHIjmez+fL25evkBwPKuuQtlY5ZCWS\nvv3227n99tsBePzxx/nqV7/Kv/zLv/CpT31qVTuFADzyyCMAv5RlWRIEoQ3IAJ+UZbkPQBAEK2AD\nkhuGkKG4wVA8Htc9TjkEnk6nmZ+fJxqN0tXVtUg+tlSUM4YoiszMzBAMBtmxY0fR6rpK+16kUimm\npqZwuVzs3btXTRPlfn+5C1yRU6fp2GxCkGUkSWQ+luDpM/1ExjKqvafH4ylJIgDBuMhsJIVJgM0e\nOyOBBBYBgkmZt+xcfg650sUcK6FDLiTHSyaTam5aK8fTRtLKAmKlUxZrZb1ZDLfddttqdwrhc5/7\nHLfeeuu8IAjXAv+NhdSFQxCEfwN+Lsuy4o28cXLIxbASxR6pVAqv18vc3BwWi4WjR4+uukeBMofZ\n2VnsdjtHjhwpOYdKLdYlEgn6+/uZn5+nqampYE/DQmh225kIJalzWjFZzDhSAgf3dLC/1cPJkyep\nqakhFosxPT1NPB4njQmT3Umdx01LQ41KIpOhBA+/ME1aXmjJtK3ewf5WD0kpQyog0llf2WKR5WI1\ndciCIFBVVUVVVdUiOZ5C0lpLzUQiweDgoErWyzGolyRJ1+K1XpRD8Llzvvrqq7n66quBNekUosVf\nAr8EfgMcZEGLfAbwCYJgkmU587Ig5KXkkAvlb0VRxOfzMT09TUdHB93d3arecLWg9bzo6OigtbUV\nr9eraw7LjZAVT+SZmRm6urqoq6sra8FUwa3H2/jqr73MhlOkZZkrt9exr9WDIAiYTCbq6+vV/N5s\nJMUvz0wRnk+QmAjS4/GzyZYiIWb43xMCNrud1joXVdVVDPvj7G31sKfewYA4W/a88qHSEXKllBFL\nHctsNi+y1BRFkeeffx6Xy5UlxyvU7qkU1qJbSCKRKFpjsA5gl2X5qxf++w+CILwLmIWLxvUbipAL\nRXXlqizy7Z9Op9Wu0kozU6V4opzCiHL8k3OhNctvbW1V5xCPx3XPoRxC1t6UlPceHh5m27ZtHDt2\nDJPJxMTERMHxihFZW62DL715J6P+BFVWE+31F72Kc3/HX7w4jQxsa64hI8tMh1Ps3FbH77zz+GJz\nmGIpIgmRBqvETFTitMlPestChO1wOJZNqJdCyqISY9lsNpqbmwu211LkeLIs53XH02ItcsiBQKBk\nG7M1xlFBEP4ZGAYmgcuAawVB8AJhWZbPbyhCLoTlpCyUggqFBI8dO5Z1Yiy1sk/vhaQQk2KE1NTU\ntKhryUqUWSvl5rIsMz09zcDAAE1NTYs8kZejQ662W+htqS74ekaWecoX4KnBAB6HhZpgkl0t1QjA\no+fmqHFa2NniYcQfJ5gRaKuvpq0eDu6uw5JO4vf7GR8fZ3R0VI30tIUXhX6DQExkNJDAYhLoaHBg\nqdiYJooAACAASURBVODDz3ol5ELnZL52T9rmqVo5nra9ViKRqPhTY6nPegn4WBwEmoFGFgyF7mTB\ny+JGYIsgCK/cUIRc6ARYSl+9RCLB6OgoQ0NDBc3ZlwKFkPVIkkwmE9PT06r/huJIl2/MSkfIgiAQ\njUY5ceIELpdrkQmTdryV6BItyzIj83H6JiO01NiIpzKcnYoyMBOjo9GBx26h2m6hu3lhEfH8TJRU\nWuat+zexuXbhsTUUClFbW0tDQwOiKKpeE8PDw6onr3Zxq7q6mrlYmu+eHCclZZBlmWa3nf+6P39j\ngKWg0jnktTADKtQ8VbuAGIlEeOmllzCbzavWXmu9W29eqMxbVJ2nxYYi5EIoR0ImyzKhUEjVG+o1\nqtcLvfriYDBINBplfHyc/fv3F9VgVjpCjkajDA0NkUqlOHjwoHrhpaQMj52boX8mSlO1jTfs2aS7\nxFqWZZ4dDvL0YAC71cTrdzXRXp8/36cQVjAhYbea6N3k5p9PjDIfFTELAqGExIE2D7MRkcZqK9sb\nnNQ5rbz98s3UOKyEEhJPDfrpG4rQ2WzmtTV12K3WRS5i2g4XitfEb4aShCQzTR4ndruNiUCMlybC\n2NdhDlnb7r4SYy13XlrfY7/fz+7duzGbzXnba+npyVcuAoHAeo+QS+JlQch6fmhZlpmZmWFgYACH\nw0FjYyM7d+7UPf5yK/sUKIUlmUwGt9tNb29vyYWKctIGxQhZkZuFQiGam5tJp9NZUdAPnxvjD0NB\nap1WRv1xhubjvHNfTUH7TS2e9gX4l2fGcNvNSBmZs5MRPnlNJ5tr8q/Ey7JMrcNCUsoQSoggQ4vb\nTo3TQq3DSv9sjN6WaiaDSaxmgWt2NlLjsJKSMvzs9BQJMYNJEOibTWDum+UNu5sXvUc+r4nT4gjO\neBJzRiKRSBD0xzh9dpYed5r+/n41mi4lxSuE9ZyyqHSps9lszvsd5/bkK9VeS+9TxSWQsiiJDUXI\nSz3R5+bm6O/vx+VyceDAAdLpNAMDA7qPLycvXChCVmRk0WiUnp4e6uvrOXXqlK7It9yCk/FAgvFE\nEJfdQmejU12wnJycVLtJz83NMTc3Byz4Qzx2boZ/f26c7mYX7ioL7ioL44EEk2ERVw4hy7KMJEmc\nGJxnYC5OrdPKc6Nhah0LaQZYqKo7NRbKS8jK59la5+CyzW5++dIMUkamzmWmwXXxaeXNe5sR0zJW\nswmzaeGY+ZhIOJlms8fOVFSg1mlheD5OUsroKs/evdnNo2eTNHtcmOwO6s3VvPZAE/6xAerq6giH\nw2qFnPLorrXXLEVqlY6Q1+NYpcYr1JNP6XCtqDyU9lpVVVUbxpy+FDYUIRdDvig2EAhw/vx5rFZr\nlvdFuV1DlKhXT24sl5BFUVT1zF1dXezZs+dis88yvZn14NxsiidGw7hcUdJyhu3VGTpM81nKEbgY\nSc9HU3zjt4MkpQyRZJrTY2FkWaDFY0dGxmI2kREXIm4lUk6n0/zq7Az/cXoGp81EQswwGkjQXl+F\nw2Ja+C3khXrRQlAiycPttbTWVjHqj5POyCSlDP5Yin2tHv73+XlsFhN7Nrupu2C9aTEJyBn54lzk\nhd/epPOedWhbDWJa5vnREDazwNsObGJzjZXQpJmGhoasNkSFCETJmWoNgRRU0lGtktF2pSPkpcyt\nUIdrv9+P1+st2F5L0UyHQiFaWloq9hnWAhuKkIudANru0Np2Rb29vYtKjFeya4hCsloZXXt7Ozt2\n7Fg0/0oTspjO8MRInHqXjSqrxPT0DKeCVbzm2stprc9eoFEW6/qmIsRTaVrrHPRuqubsZJizk2HS\nGZmeZhettVVMTciq/E9Jh/ymb54ttVVYzSZkZCKpNLMRESkDmYyM02pi/5ZqRFFUo0blZiBlIJ5K\no6whbq6p4q/fuIPv/2GC+UiKHc0unFYT4YSEmJYZDyS4bk8z7ioLDS4rOza5ODsZIRhLY0uLXHNZ\nc8nu1ernFgSu7Kzjys6LqoJ8bbcgP4FoFQj5OookEgk8Hk9FyHQ1VBZrDUFYaFbrcrkWmdNHIhHC\n4TAjIyN86lOfYnh4mO3bt+Pz+bj//vux2+1IksTb3/52vvCFL6ypOb1ebChCLgaLxaIu1omiSE9P\nT8HHm3KJsFwjICVXvWXLlkUyuuXMoxBCcZERfxyrWUCUJEKBMKLDwdatW5mNpcnkiVW1i3VKQqK7\nuRqTSSAhZnjr/hau6KgnHg2TSCSIx+NYrdaFfLZmDAABAU+VldfvbiacELFbzBzfXktjtY1MJqOS\nuShJPHRqkof/EMJ15jxXbK/jnYe3YLOY2VJTxcdfvdDA9Genp5FlGYd14XubDCeZDCVwV1UjCAKv\n6mmgo8HJuYEYW5tr2dO6PG1qOeSpVSAo0ZpSwhwOhwmFQkxNTTExMYHZbFajPKWEuRxSXCuVhR5U\nUvJWyJxeK8f72c9+xkc+8hFe+9rXAnDjjTfy+c9/HlEUeeUrX8l1113Hgw8+uCbm9G95y1tKNgtW\nsKEIudBJEI/HCYfDnDt3jt7e3qzHznLGKQS9JvVTU1MMDQ1RXV2dt21TLpbTK0/BuakIn3/4JRIp\niWg8ToNdZlu9i63NLYQTEk6bmcbqxVI6JWXR2+Lm0XMzTIYSWM0mrGYTNx/bxt5WD5lMBrvdTlVV\nFS+88ILqyOXxeLi8xc6Tw1FqnTbiYoZ6l41X9TTgtJkXvQ8sEMIjz57nxycn2dpQjcvl4JnhEA1O\nK2/c06S2vjKZTCREialQEodtISKWZTBr208JAtsbnGQCdmrdy1fILDea1ZYwRyIRqquraWpqyiq6\nUJoHAIvaPhU6TypNyJVSbFRaBlmOOf2hQ4fYvXu3uk0URURRRBCENTOn/8UvfsGNN96o67NuKELO\nhaIaCAaDVFdXs23btpJkvBSUimTn5uY4f/48Ho+Hzs5O3fnmcuRshZQed/z8LJFoDJuQpsntZD6W\nYneVhXBCorHaxnV7NuGwLT7ZFUKudVr5wFXbedI7T1xMs3eLm55ml0qQFouF3t5eYOFCjMVihMNh\nDjaESATjeP0htjitXN1eR9g/i+DxLPJIUDq7DMzZ2NRYR61nIVdR5wSvP4nNZlOj6GBcZC6cpH8m\nitm0oC7Z3+am2W1dEcP2SkOrGChUdKHkpbUyMYfDkZWXttvtFU9Z6HE31DvWapdNw0UvZGUOhw4d\nor+/nw996EMcPXp0zczplWP0YEMSciqVYnBwkLm5ObZv36526qj0ApmCYib1yqLhvn37cDqdTE9P\nEwwGdY27FE9k5QJNp9N4B30MzQRpqXFgt7sRBDAnJLob7LzzVd1Fx9NK6Rqqbfxfezep4yqkkksG\ngiCo/r0tLS307rz4uB4KhQiHw4yPj5NIJLDZbNjtdsLhMFVVVezbt4+gL8zZ5ybUiDSaTHPZFrv6\nPmazmeHJKE01TjbXOpiJpIimJNrrXVRZTGpuHi72UlT+XQ5xraaXRSHXNsWoXruwlUwmEQSBuro6\n1RVvqZ9zvUvo9HYLUdKQZrOZ559/nkAgwPXXX7/IHGgNzOl1YcMR8sDAAJOTk7S3t9PT06OeZCvZ\nNSTXpD4ajXL+/HkkSWLHjh1ZGsxyTer1zlkZ12w2Mz4+js/nY8uWLexua2QsEKdKWFjUA2hxLVws\nsiwzHkwQT6VpqLZR57z4eK/VK2sX7JQTWe/JrH1cVzwSUqkUfX19hMNhamtrEUWRU6dO4cJEnVnG\nN5XEarXQXOPkzXuzV83TGRmTIFDjtFHjtBOMizjsNjVihIULeGxsjPn5eTZt2qQSde7i4VpE00sh\nd0EQcDqdOJ3OLJ+J559/ntraWpLJJHNzc8RiMQRBWGStqYfMKkmia2VOn0qlFlWT1tbW8upXv5pf\n/OIXa2ZOrzjN6cGGImTlZMzXwHMpBkN628ZYLBbi8TiJRIKBgQEikQjd3d150yPlEnKhXnm5+D/s\nfXl8U1X6/pOl6Z5u6d7SPW0plKUti9sPFdBBwH2bcWfEcQFcZhRlGHVGQBQRRUVFhxFHcFC/7qgj\nKK5QdmRpk+5b2qZtmn1Pzu+PzjnepEmatAli7fP59ANt05uTm9z3vud5n/d5+Xw+ent70dra6jLR\n+uEEE/787gnUKfUAAS4pjUd6zICC4otT3dh2oB1qox1xkUI8NLcIEzLj2PGcTifLiIGRZxVOpxOt\nra3MpY4r8QMGLjxpoRa1ChW0Oj3EPDVkxw+z4CIWizEuPgL1SgO05oFiodHmQL4kiq1ZpVKhvr4e\nCQkJmDZtGoRCIbuZEDKgBqHnn/7Lvcl4GzV1JpoL8Xg8JCUluXxGudaaXCkedzZfbGzsoM/1mZwh\nB+qF3NPTwyZSm0wmfPnll3jooYeYofzpNqenRvn+YFQFZGCAG/JmUk/9C/wBDeD+tk0rlUrWWDF+\n/HivF10oxjhptVqo1eoBPtWtzdrpJKjMice5hQMFtY4+LWQ9ZsSlmfHyt83oNVgRxuehz2DFQ++f\nwv/dMQ2i/zVQGAwGnDp1inVaxcTEDLs7TalUskGrVVVVHi9YoVCIZEkikiWu7c1U3tTZ2TmQVRsd\n6FWHISIyCpOzEpAQMeB4J5fLQQjBhAkTXM4BNyv++bw4B9Ea9PloJk2D9JnaXeepg82TtSadzafT\n6dDf34/W1lYXKV5sbCwsFkvQ1uUv5+svAvVC7uzsxM0338ySiWuuuQbz58/HzJkzfxFzek/Tg7xh\n1AVkXxacwdYWOxwOtLa2orW1FZGRkZg2bdqQH2pfXsvuGKqoZzKZUFdXB4vFgri4OBQVFQ3a2rWq\nTIgMEyA5dqBgkxAVhvo+E/gtaijUZsRHCSHg8xEuJFAZrZB3a1GaFos+gw1xuRNgsZphs1vR1tYG\nvV4PYMCQRywWs2zLVzak1Wohl8sRFRWFKVOmBFw48hZgjEYjtFotNBo19u2rh9VqRWxsLCQSCZvh\n562jC4ALL809rqcgrdVqQQiBzWZjgXq4weuXCO7c2XzcdVitVuh0OvZ18uRJNkCV64oXaLb7S1hv\nWiwWljyVl5fjyJEjgx7zC5vT+4VRF5C9YTiOb94CMp2q3NLSgoyMDJSXl6O9vX1YfPNQj/UUkGl3\nn0qlQmFhISQSCWpqajxK5CJFAtg4P9dZnKhXWWAQqmFxDKgW4iJEsDmdEAl4EPB4aOkz4t0jnXCS\nAQvMopRoXDZpPAR8nkvGqlAo2JaYux2OjY2FwzHg/WC1WlFcXOxxvt9wQQOMTqeDWq1Gbm4uMjIy\n0NGnw2s/tKC5rxvxYQ7MyeYjMz6C0R2xsbE+TWzcg7TJZIJMJgMAFBQUuPhfc4uHAoGAHXOoz8CZ\nkm3zeDxmBiSRSKDValFSUgKBQMDeX67/sXv3oS+V0C9BWZzpTm/+4jcTkIeTIbsHQ643sEQiYVpi\ng8EQkmnS7o+lHGxzazvycnMwY4bUJRB4Om5pWixOderQoTaBx+Oh12BFvIiP4tQY5CVGoaFXD5vD\nimiRAFOy45CfHIUtP7ZDHCFEdLgQXVoLPjnejW6dBQsmpiE7IRIxsWJEx8YySZDT6YRKo8NPLb3o\na25FuFWLaL4NYrEYEokEVquV+eUGAxqNBnK5HLGxsaisrByYCON04tV9Xei38JCTmgC10YZv1Xw8\nUpULu9nIKA/awEJvHGKxeJA6weFwoLm5Gb29vR5rAfTGx+WmuZQH4J2XDmZAps8TDNDgLhQKER8f\n79I0xR2g6smxjQZpKmf8pQLyr93HAhiFAdmXJ/JIKAtqQBQTEzPIGziQYB/oNGl6wXd2dkJW34hv\ne8LRqI2AoLkHl08WYv7ENHbRe8uQr6nIRFu/CU5CcLhJgJpmA04odJDEimBxRMJqJ5glTcK9FxQg\nTCCAwWpHUrQI3VoLjndoYXM4odRasP1AO8anDwR4JwGmZsfhghIJnE7g/ZNqyBUaGPV6xERH44+z\npMiI4UOr1aKvrw9NTU2w2WyIjIxk2apYLA6IwqAGTFarFaWlpS5OdCqDDX0GK1L+R80kRovQo7NA\nY+UhNznZxcSGu1VvampiFAflyFUqFTIyMlBVVeUxA/Wm0vDFSwNgwepMhK8gypXiUR0v17HNfeQT\nAERFRUGv149Iikfhz05Aq9We6dNC/MKoC8jeEAhVAPxc1KMcqFAodDEgcn+sv8cOJKMRCAQwmUyo\nrq6GWCxGiyATDVoV0uPCYXcS7DikQGZ8JKaOi/eZeYuEfBQkD6zbZDLj25N2mM1GiCOEyIqPRFla\nDPh8PsIEA2srTYvFsQ4NOjVm8HhARJgAmfGRaO4zYOdJJapy4sHnAQdb1RBHCEFsJhyRt2JcYiSK\nirJhtDqxW9aHP52Xi+joaJeLmBaX1Go12traYLFYEBHhSiu4N45Qrr67u9vrZO8o0cDUZJvDiTAB\nHw4ngYMQRHtoehGJRIOMgjQaDWpra8Hj8SAWi6FUKqFUKhEdHe3Cl/vK1Hzx0g6HAwqFgrnEcT08\nfCk8ThcCNc735thmtVrR0NAAp9OJlpYWJsVz7z4MtOg31NpGgxcyMAoDsrc3LtCtncPhQFtbG5RK\nJaRSqU9+KpCs11/QVm+TyYTp06cjOjoaW98/ifioAb+IMAEPQj4P9T0GTB0XP+QaaMZWnBaLKRlR\n+FSmgSmMj5wEESKcAmisAtjtDoQJ+LigWAIBH3i/zwg+byATjhIJoDHZERnGh/B/1mlRQuC7Y3Lk\nxAmRlJSAZMkATxwuBEy2wTcHrp42NTWVrYvbONLR0cEaR2JjY0EIQV9fH9LT030WTWPChbhiSjre\nPdwJHg8gBPhdWTIrZnqD3W5nHtAlJSWDioc0A+zu7kZ9fT0cDgeioqJcbiC+qBg+f2CXIJPJEBcX\nh+nTp7tw0Z4yafp3pzNIB6tRQiQSQSQSsUktwM+DAOh5bGhoYOfR3RVvuGsYoyxGKWi7dV9fH2Jj\nYzF58uQh/yaYnCDdlhuNRhQWFkImk7GsPFUcjhMdWpYN2p1O5g/sy/eCZmi0APWH8ydDKO6AzmxD\nGOxQqo1Ii7Ti+LEj6DM5YEIEEsTR+NOMVHxRp4PZ7oBC40BshBDh/+uIU6lUaFcZcFZxJmaVZaHu\nu2ZoTDaEC/lQ6q04r9C/FnVPjSPAAEUkl8vB5/MRGRmJrq4u9Pb2utAd7tvhuaUpKEyORrfWAkmM\nCIXJg3cz3HPS2dmJlpYWjBs3zqPbHp/P92iuToNLX18fmpubYbVaERkZ6cJLU6cx+l66Uyz0PaNw\npztGWjwMFMH0n3CXqXkzqaeueHS3xJ3LR4O0v1OkxwLyGYqhgqO3ogp3vH1+fj7r5gkV3Ndht9vR\n2NjICknJycmDJHy/r8rC6j45unUWECdBWboY5/4v8Hkq6rlbYtIsSCTkYf7EVHxwtBNKnQMl2cmY\nX56Gtn4T9h3vhsNuw6lWI+IUOpRFW9GpdyI6KhIXSsXY06DBYVk34sRiFOdm46JJWYiNEOLG6dn4\nskYJg9WBWVIJzi30X3vJBd3yGgwGlJWVuVzEdC6eVqt14X652WpuYgzyJd4DMfBzUVAsFrOioL/g\ndsK586l0bR0dHdDpdLDb7UhISEBGRsaQemZvdAf91z2btlgscDqdsNvtZwTlwYU/MjVumz3dLQE/\nz+WjXh4GgwFmsxkymcyl+9D9+BqNBvn5+SF5PacToy4g+wLlWbl3b0pNdHR0uIy31+v1AXHOgYDS\nCzSrpVOtc3Jy2PN7Qqo4AqsWjkdjnwEiAR+FydEQCn4uMFFZn7dATOEkBAea1bA4CFLE4VDqrDjR\nocNxhRaJ0eGIFEUBiENbvxGlE9JxcWIEOjo60NzcjOlJQkhjw+FwmJAlFqCvqx02sRjpsbFYdHbO\nsM8JPQ8KhQJ5eXkoKSkZFLzCPMzFs9vt0Ov10Gq1LlppbrZKJ3lYrVbU1dXBbDZ7zFiHCy6fGhER\ngf7+fqSmpiIrK4tx5kqlEkajkRmr0xuIL8tNT8VDbmafm5vrYpPqb+eht9cQLIxEZcGdywcMBOia\nmhqkpqZCr9e7SPEodUQTmalTp6KtrQ033XQTuru7wePxsHjxYixbtuxX4YUM/MYCMtUiC4VCEELQ\n0dGBlpYWpKenD/IlDrTV2pvbmifQAmNPTw8aGxuRkpLi91TrmAghyjMH89lc4/uhWp1VBhua+ozI\niBtonHA4CQ63aQbGJEUJYLLZcUKhg0JjhtlixbR4I5JjRKisrGRbSG5zBvV3dudXxWKxX9lnb28v\n6uvrkZKSwlpW/YUnmRZXK02zVYvFAofDgZSUFOTl5XmcoD0S2Gw2NoKLG+yjo6MhkUhcHkcVHi0t\nLdDr9S6joLg3EHcYDAbU1tYiKipqUGbPzaC5N2N/gnSw7TKDKXujzoiepHjUWfDjjz/Gnj178Nln\nnyEvLw+TJ0/Gs88+y+w458yZg3/9619nvBcyMAoD8lBTQ+gFUV9fj6SkJK9TpQPVLQcyV8/pdOLQ\noUOIi4tDRUXFiG0P6VZYpVIxVYCvgOMkAwY9jI/kDZjQl6bF4Gi7Bi19RugsdgjsZhjURtSIknFW\nZSEiwn6+yDyNgqe8oFarRW9vL5O6UY9kGnDo+TYYDEzBMnny5KAFSW53n0qlgk6nQ0ZGBuvio1py\negPhri3QEfWEECgUCrS2tiI3N9djZs+Fpyyf3kAo3eHebBMdHY2+vj5oNBoUFxd7LDAH0nlIaQ76\nd8G2LQ12QPaUqHA/f8888wwWLVqEv/71r4iPj2eGQLGxsSgtLUVHR8evwgsZGIUB2RccDgeOHz+O\nuLi4QVpidwQ6rcOfuXp6vR5yuRwmkwmlpaUu3NlwQQt2cXFxyMzMhEajQXt7O5OT0UyVysmAgfbp\nxGgRlFoLosMFUJvskKZE49zCJBBCcKRRiQieFZPGSZApSUCn1gKt2e4SkD2Bywty+VWayahUKrS0\ntDD+0+l0Yty4cUhPTw+aFy8FbSsnhKC8vJxl9twsi1ugc7+BcGkFb2uj6onhcNFceGsPNxgMUCgU\naGpqglAohEAgQHNzs4sMz9d588VL0yIv7TqlyUowHPGCGeAD8UJOTExEVlYWxo0bBwBobm7GkSNH\nfjVeyMAoDMieshOdTge5XA6j0YicnBz2hgV6HF/wFcAtFgvq6+uh1+shlUqhUCj8DkDeqBB3nlgg\nECAtLc1lbJDZbP6f34PGRfMrFotRlRqDOrUIRjsPBZJoTMoSo6+3B2F9DShJDse49CxEioSwOwe2\nsxFhw2/R5RZvOjo60NraiqysLERERDAPBe7auEqF4cgVW1paoFQqUVRU5HMggbcCnbsRj/vaIiMj\n0d7eDoPB4HEmYzBgtVrR2NgIHo+HGTNmICIiwuXm1t/fj5aWFmY56UvHzQU30Or1etTW1iImJgYT\nJkxg2TJXlUOPEygvHSxO2t9sm2tODwwkP1deeSU2bNgwqGHkTPVCBkZhQAZ+NhgyGo2or6+HxWJB\nUVER1Gp1UF2ouPBEcdjtdjQ3N7OGBuoC193dHVD7NDcgD1Wwo+AWmriaXxqkdTodEsxaRFksgEWI\n6jYzoqKiUD6hDKlmPr6s6YHGZIeTAGcXJEIcMbzsj8KTLSaAQTcQqlSgWX54ePigLN/T6yWEME5+\nKM2yL/jSSms0GigUCqhUKoSFhSEqKgpdXV0wGo1D+mT4C1rc7OzsHHRD4d7c3Of1edNxcyWCdG0O\nhwNNTU1QqVQoKSnx2uHmT+dhqBUe/mbIZrOZ7YJsNhuuvPJK/OEPf8AVV1wBAL8KL2RglAZkqiVW\nq9XsQ83j8aDX6wMq1AUCbieg0+lkBcPs7OxB/szD8bOggXkk3sTuQZqreR43bhwMZitq5XVw2KyY\nECECLzwayfFi5CaHD9uDgdpiAhhki+ltbfRicR8Oyg023CDtdDohl8sRHh4+LEe5ocDj8WCz2dDa\n2gqxWIxzzz0XYWFhLmvr6uqCyWTyqKLw97yp1WrI5XJW2/AnM/Sm46bt4bToajQaIRAIEBYWBq1W\ni9TUVEydOnXYnYe+iofBmNJC4Y8nOS1K0udetGgRSktLcf/997PH/Bq8kIFRGpBlMhkkEsmgAotQ\nKPTb8B0IrOBBM2SlUon6+npIJBKvyolAvS+sVisEAgELivQ1EUJwqE2Db+V9EAp4uGh8CopTh5Zy\ncbf1BQUFiIlLwKvft+Bgqw4CHh+XT87FXGk8u6BPneoa1OI8FKVAdwd9fX0oKioKyBOWghtsuO25\nNCOkI7IsFgtzI+vv70dsbKxLRjgS2Gw2NnTAnZ7guqVxH0+zVRoIaYMJPXfuUjebzcbkeGVlZR7b\n8wOFe3u41WpFbW0tzGYzMjMzYTKZcOjQIQCB2an6CtJms5nNjnQfpzXcphZ/p4UAA5+X77//Hm++\n+SYmTpzImrpWr16N5cuXn/FeyADAC1DyElx9TIhgs9k8dq319fWhp6eHDeUcCgcPHkR5eblfLmU1\nNTXo7e1FQkICCgsLfRYMW1tbwePxXAoAnkAIwalTp2A0GpGQkIC4uDjExsay4tGhVjU2f98CcYQQ\nTjLQrnz/hQXMt8LT8bq6utDc3IzMzExkZWWBz+fjjX1t+LJGiTRxOBxOAqXehvsvzEfFONcCGA2E\n9MudUqAKiu7ubjQ3NyMrKwuZmZkh6SqjWtzs7GxkZma6BEKtVjvibJX7HDk5OUhPTx92gLfb7Uzq\nptVqodfrwePxWHavVqtZM1KwuU3u6ygoKGANRxRciSD9cjqdzHuCnjtfBUvuc9CmJvfiIf3i/o0/\nxcO6ujpIJBKf0jGbzYa5c+fi4MGDAZ2b0wy/3thRmSF7w3BN6n0FZDo/T6/XIzU1FVKpdMjjDjUr\nj7slLCoqYkUmd73vpw12iHhATPjAIFCrzomDLWqPAVmtVrPsxV0RcFKhRXzkgEeGUMCDkA/UKQ0u\nAdnT1tg9SDc3N0On00EkEiE1NRXh4eGwWq3DKs55A1U2cK03gYGMUCKRDJmtCgQClyzfkxsZL2CR\nfAAAIABJREFU9RGJiYkZkXqCQigUDpourdFoUFNTw3TUbW1taG1tHZStjqTmQXXL0dHRqKqq8nis\noQYAUF6eqz7hKjxMJhPTRnOfw1Og5XYeureGe+Ol/bXeHA1Ob8AoDcjBtOD0Fjhpi69Go0FRURGs\nVitMJpNfxxUIBDCbzYN+7qlgx9VbcpUABoMB4W2NMPYb0GMxghAnNDY+NGqgvz+KXcxU/uV0OjF+\n/HiP2+FUcQROKLSIZB4ZBJKYoXcFNEgDAyOshEIh84imQZryvu6ZdKBB2mq1or6+HiaTyW9lQ1hY\n2CBXN25jhnv7dXR0NNRqNevkC4V6wuFwoLGxEf39/Rg/frxLIOGaGXV1dbH3LZBslR6HTl33plv2\nBW8ac3enPp1Ox5ptEhISGLXmj8LDfb3e9NL0mnI4HF6Lh6PFxwIYpQHZG4Y76JQLal7e1dXl0uKr\nVCqHLNRZ7U60q03o19oQ4fj5uP4qJyioXOuqaQXY8FUjgIEPdWIkwYwcMbq7uyGXy2E2m0EIQWpq\nKtLT073SKH+YloknPjOiR2eFgxCMT4v1yxzIly2mt0yaqwTwR0HhdDrR3t6Ojo4O5OfnIyUlZUTZ\ntqfGDJvNhubmZjQ1NSEyMhJOpxM1NTUsW6XdcyOlXugOJzMzE1VVVT7NjGhF31NHJOVVPTXb9Pf3\nQyaTIS0tDZWVlUGji7jqk6ioKPT39zM1AtcT2Z8BAO7wxEtTSR5Vr/hyxBst00KAUcohOxwOj5mw\n0+lEdXU1Zs6c6ddx6uvrIRaLkZKS4tJqnZmZiXHjxrl8yFQqFbq7u1FaWurxWEarA2/sbUWn1gyr\nxYIIYsEDC6YiKozP+LXhKCdaVSYcaukHX8DHzLwESKLDoFAo0NbWhuzsbIjFYsZdUn6QG2hoEUdv\nsaOhx4AwAR9FKdEIE3i/gOjklMbGRmRkZCA7OzvgC989SGu1WpcgTa03m5ubIZFIkJeXF9QpFBRc\neqKgoIBlnw6Hw+W8+fLIGAp0FJRAIIBUKg1KZybVI9M1WiwW2O128Pl85ObmQiKRBJUqAn7O7tVq\ntU8vEPdZfdxdCP3ydu4IIWwyuSc7VHdHPEII1q5diyNHjuC///1v0F5rCODXG/GbCsgA8OOPP+Ks\ns87y6zjNzc0ICwuDSCRCfX09EhMTkZ+f73HLqNVq0dLSgokTJ3o81u5aJb6V9yEzIXKgGq3ow++m\n5mFOSXLQhOp0qklSUhJyc3M9cm90W0wpBb1eD0LIoGzQW6ChTTaRkZEoLCwM2lgmCrPZjL6+PrS0\ntDDfEfcpI74aH/wFnUtIfZD9oSe4Lc402ABw8aHgqhToyK2urq4hm1SGC1qobWpqQlZWFsLDw9na\n6A3O35mCvqBSqSCXy9kNONBjcE2guDc47gAAPp8PuVyOhIQE5OfnD3mTVyqVeOCBB8Dn8/H3v//d\nazJ0hmCsqDdS2Gw2tLW1IS4uDpMnT/bpzTqUtrjfYGMcLV/AB+wW/CRrQmG4HmKxGHFxccMObnq9\nHnV1dRAIBC5twp7gyeOXG6SpGQ8AlyAdHh6OxsZGGI1GSKXSkBRR6FQNOhSAFui4DSMKhWJQoAkk\nSHOVJjk5OR59kL3BWwGMnjvu0FeRSAS9Xo+kpCRMnTo16DcuADAajaitrUVERASqqqpYouBuZ0mD\nIJdScPeV9nYOqCTPYrFg0qRJfvsTu8PbrD567mhhPDw8HEajES0tLV4HABBC8N577+Hpp5/G448/\njssvv/yM7bwLFKMyII/0zTEajewDkpiYiLKysiH/ZigFR25SJA63qhEj4gPgIyYxBdOLExATw4Na\nrUZrayusVivjBv1xS6PttTqdDkVFRcMubHgK0jQbpKON6MWSmJgIrVYLAEHhVYGBC6y3txcNDQ1I\nS0sb1GXnS4scSJDmtgoHQz0BDD53VquV+ZVkZ2fDYrHg6NGjwyrOeQMdj6RUKlFcXOzzfQ8PD0ey\nl5mC7k0j7hJBqrDIzc1FWlpa0IMe5YAVCgUkEgkqKyvB4/G8DgD47rvvEB4ejs8++wwSiQRff/21\ni6pmNGBUUhZOp9Nr8W7fvn1eu6BogOvv70dRUREAsEr1ULDb7Th06BCmT5/u8nPKedkdDnwl68Xe\nJjUAYHpePOaWpoDP+ZDTSjZtetBqtbDb7WxbR7/4fD7zDg7VxQK42mLm5Ax4HXvT09K1+fL39QTq\n+BYWFoaioqIR8avuOmna1RcTEwOj0QiLxYLS0tKQFIBojaGtrc1j8ZFbnKMZ63Dc5tRqNWQyGXtP\nglW046pP1Go1VCoVeDweJBIJ4uPjffK+wwFVgqhUqiG9qSlnvm7dOuzatYt1Tubm5uKDDz74tWTH\nv10OmRACq9Xq8XeHDh3ChAkTXC58qhagAY5OeFCr1VAoFBg/frxfz7l3717GT3PH8HALdvb/VYuF\nfl5IVOJGL+S+vj6YzWbmZ0AvlmAWvLhBcqgmF27xiwZp7hQPb0GaTmjp7++HVCoNiWyJEIK2tja0\ntLSwIqF763UwOGmdTofa2lqIxWIUFBT4rR3mus3RIO3NrpT6LVPZn7/da4GA3lTa29tRWFiIhISE\nQTdgwDtn7i/oris1NRU5OTlDnvuuri7cd999EIvF2LBhA+PiVSrVsDpAfyGMBWRPOHbs2EC7cEwM\n6zBqampCeno6cnJyBklvGhsbUV5e7tfz0oKhu1tWMO7glGeLiIhAQUEBa3rgqie4QZAWSQIBt9A1\nEgqEdqbR9RkMBrYljo2NhdVqRWdnJ8aNG4fMzMyQZDh6vR4ymQxRUVEoLCx0yTy5mbROp4PJZBrk\nj+FP8YuO3dJoNEFzfeMqKOgaTSYTbDYbkpKSkJGRwTj9YII2kVC1ibebCpf3de/s4za0eMr0qUpD\no9GgtLR0yBZxp9OJHTt24Nlnn8WqVauwYMGCX0s27AljAdkTTp48iczMTDgcDsjlcsTHx6OgoMBj\n0cVsNuPUqVOYOnWqX8/7ww8/YMaMGSMyAHIHte40m82QSqVeL3p6oVCqgyvTokVDb3QCd7s90jZh\nb7Db7ayYRgX+QqFwkMfDSJ+XGySLi4v9Lj66S/B8BWmu7I+2bociUNAuuLCwMGRnZ7s49QXLrpTL\nR7vLzAI5BqVj6Plznx7jdDqZBjsrK2vIdXZ2duLee+9FYmIinn322V9TJuwNv92ADMCridDx48eh\n0+kQFRUFqVTqc+vnjRd2B6UmfvrpJ5hMJr+C4FDgGgDl5+cP8iDw9xie6ATudp0G/MTEROTl5YXE\nnpR2NVKFBr2pcIeWarVaVlzirs9fkyBCCLq7u9HU1IRx48Yx2mkk4Lq50SBNZ/OFh4cztUmwgzFX\nLieVSj0GI3e7Uk/eIkN5I2s0GmbElZubG1TPEZrpazQatLa2MqrIk/m/O9e+fft2bNy4EatXr8Yl\nl1zya86KufhtB2Sr1epiZkKdqFQqFbKyslBQUDDkMdx5YU+/d++wczqdg4Ig3a7HxcUNGWS4sqzh\nNl34AqUT+vr6oFAoWCYTHx8fcBAcCtSGtL29HXl5eUhNTR3yuO5UjMFgYDIt+uVOJ/iiJ4IFmkl2\ndXUhPT0dTqfTJZPm0kUj8UWmQTIpKQl5eXkBvfe+mm24QVAkEjFqyh/qYLig2uXs7GxkZGQAgEuW\nz72J7NmzBxEREfj000+Rm5uL9evXBzSL7leAsYBMCGG8aF9fHwoLC2EymcDn84d0WqPw1EjirWDn\nDXa7nQUZjUYDo9GIsLAwFqBpYYnaScbGxiI/Pz8k2lVPtpjcIEgzVU/rCyTI9Pf3M2/fkXbZcWVa\ndH3u6omSkpKQ+RnQwEKLUO5B0mq1utxE6PoCCdJ2u50NSS0pKQlqkOQG6d7eXuh0OiYjpEE6WDdh\nwFW7XFpa6rMoTDP9tWvX4quvvkJYWBhMJhOysrLw8ccfj5bsGPitB2SLxYLm5mZ0dHSw4hGfz2fi\n+Pz8fL+O4x6Qg1WwoxexRqNBf38/dDodeDweUlNTIZFIgl644Vok+mOLyQ0yNBP0xyCI7kQcDseQ\nlNBIXktHRweampoY/eHO+QZDPWGxWFBXVwe73Y7i4uKAmiI8BWlvmT4duhoq/h4YCJJyuRw2mw0l\nJSXg8/kumSpdn7sTXqBr6e3tRV1dnd9yzI6ODixduhSZmZlYt24du6mq1epRYxj0P/y2A/KRI0cQ\nHh6OnJwcF160p6eHSa38AVc5MdKJHe6g0i+VSsWUH1wNcqCNIt6g0Wggl8shFou9tn77A8pZ0vW5\nF5Y0Gg3biYRKsE/pCU+t254Kc4FwqhSEELS3t6O9vR0FBQUukzhGAhqk6foMBgOsVitEIhHGjRuH\nhISEoGaqgCu3PpQ5k7untL92pfRvZTIZHA4HSkpKhkwmnE4n3nzzTWzatAlPPfUULrrootGUDXvC\nbzsgu3PIFIFoiwFg7969qKysZN8HIxBzVQ2+qvS0UYQGQNoowm1pFovFXukAOqLJarVCKpX6FN8P\n93WYzWa0t7dDoVBAIBBAIBAE5SbiDqqeUKvVAVlKemoW8ZXpU7/l+Ph45Ofnh8TQiOqjFQoFo3Pc\nM1V/W5t9wWw2o6amBiKRCFKpdFjvA7fwyjUK4q7PaDSisbER+fn5fk1Sb29vx5IlS5Cbm4unnnpq\n1Di1DYHfdkC22+0evSX0ej0aGhowadIkn39PM+JTp05Bp9MhLi4OgsgYaBwiJMRGIV8yPIkWNQAa\nrqqBSoxokPakQY6KikJbWxsb0cS1xQwmPDWQcLsN6ZfNZhvUbRhI8wTN8IIhMfM1+cRms7EML1Qy\nK61Wi9raWvb+e+sYdefMAwnSNMPv6OgIiakRLQz39/ejo6MDdrudGUDRL09t9U6nE2+88QZeffVV\nrFu3DrNnzx7tWTEXYwHZU0C2WCw4ceIEKioqPP6dp4Kdw+FATVsP1n3VAoPZBpvDgQnJYbipIhkJ\n/1MnDKUBpcFLIBCgqKho2CYtnsBVdnR3d0Oj0bCZatQMJxgaX4pAu+zcuw1p2zAN0nQ0lXtwMhgM\nkMlkiIiICImzHF1bV1cXGhsbkZSUxLJVLh3DzaSHC7vdjoaGBuh0OpSUlAS8W/EWpN1vxLTBg3YN\nhirDp0ogOrLJl13p7t27kZmZiX/+85+QSqV4+umnQ2L+f4bjtx2QvVlwOhwOHDhwADNmzBj0O18F\nu5Uf10KhNiMhKgyEECg0ZiyqlCAvxgGNRgOLxcKyBKpOCAsLC5oB0FBwt8Xk8/keu+XoxRsXFxew\nPIt7IWZlZfkl8PcGp9M5KEjTTD8mJoZtj4fbrOAPuG5p7gGf0jHcTJqa3HCDtD83iZ6eHtTX1we9\nicQ9SKvVajgcDiQnJ7PCcLA5abPZjNraWohEIhQVFfmkQRyOgWvjkUcewYEDB0AIQWxsLC666CI8\n8cQTQVsTRW5uLruxC4XCQTP2CCFYtmwZdu7ciaioKPzrX//yu+krCBiz3/QEOkmaC38Kdt1aM2LC\nBez3fD4fCI9GQUEyO4bZbGaFrcbGRphMJjgcDkgkEhQWFoYkK+CONnLv5HOf48aVtymVSibP4srb\nvGX61Mg9OjoaFRUVI85WuYbldDoG9RRpaWlhOwg6NcLXVjhQ0Kkvvb29Xt3SeDweIiMjERkZyXhR\nbpDu7+9HS0vLoMIr1y7SbDZDJpOBx+Nh6tSpQW93prsggUAApVLJVBq0tZm+x8FQTxBC2OADf2mQ\n9vZ23HPPPSgpKcGBAwcQExMDi8UChUIx3Jc8JHw5wH322Weoq6tDXV0dqqurceedd6K6ujpkaxkO\nRm1A9vaB4/48EOVESWosjrZrkBwjgt1JwAOQFf8z7UAv4IiICPD5fKhUKmRmZkIikUCv16Orqwty\nuRyAf+3MQ8HpdKKtrQ2dnZ3Iy8vza7SRpxlzlE/VaDRob29nW3UapCMjI9Ha2gqDwYDi4uKQbTW5\n9MSMGTNYUONuhVtbWz12GwZCx/T19aGurg7p6emoqqoK6Nx7C9KUM+/r60NTUxNsNht4PB4sFguy\ns7ORnZ0dkmYVrnZ5woQJTLscHh4+aI4gvRFz7Tb97Yg0mUyoqalBVFQUKisrh+T/nU4nXn/9dWzZ\nsgXPPvssZs2axY4dHh6OvLy8IJ2BwPDhhx/ipptuAo/Hw4wZM6BWq9HZ2clmVZ4JGLUBeSg4HI6A\nJGy3nT0OG3Y3oqHXAD6PhxumZaEkzZUHpLRBREQEpkyZwjKiuLg4lyyQBpiWlhbWyRdIuzCdrZaS\nkuLVStRfuPvlcjP9lpYWaDQaVlDq6+uDzWYLqCg3FLhDPz2pJwQCwSBjc655ER1UOtQ5tFgskMlk\nIIRg8uTJPpsVAgF31lxaWhp0Oh2bxxcXFwe9Xo9jx465zMGjmfRIgjTV+44bNw7FxcU+Py/ehr1S\nqoi2tXPPIW0W6ejoQEdHB4qLi/3qnGtqasKSJUtQVlaGH374IWRdgJ7A4/Ewe/ZsCAQC3HHHHVi8\neLHL7zs6OlwawrKystDR0TEWkH8p0IKdUCjEyZMnAyp4xUeG4dFLpNBZ7IgQCiAS/pxZcQ2AioqK\nfBraeAow/lAJERERLqqGYAYVLng8HsxmM1pbW5GQkIBJkyZBIBAwZYdSqUR9fb3X2Xz+gmvQk5WV\n5XHopzcIhUKPdAzVSHO36mKxmOl/CwsLg6Ypdof7vDn3nQT1dqBZakNDAytscumEoW50VqsVMpkM\nTqdzRDTIUEG6s7MTarUaQqEQycnJbNKIt2vF4XDgtddew9atW7Fhwwacd955p11B8f333yMzMxNK\npRJz5sxBSUkJzjvvvNO6hpFi1Bb13B3fuAU7QsggZzSaHdAg7U8DQTAMgLzBYrGw9anVala1TklJ\nQWpqatD0vVzQLju73Q6pVOozu6FFOa78DvCPjqH0RHh4OIqKikKingAGskiZTAaRSAShUDikBnkk\nz1NfX++3kxnFUOoTeqMTCoUuBdVgNqt4WhN3yGhUVJRXAygqwYyNjcV9992HSZMmYdWqVac1K/aG\nxx57DDExMfjzn//MfnbHHXdg1qxZuP766wEAxcXF2LNnz+nKkMeKeoBnnpjP5w+ajcbNUru6umAy\nmZjsiWaqNHC4GwC5jxwKBiiVYLVaYbPZmEJDq9Wit7cXjY2N7OKl6xuuUT01zunu7kZBQYHLuB9v\n4BblKLzRMdwpzd3d3SE1pQfAzNyNRiMmTZrkIjHjapA7OjpgNpuHLW+zWCyQy+VwOp3D2rHweDzE\nxMQgJiaGme9wrSy7u7vZDdJmsyEyMhJFRUUhM93R6/WoqalBQkKCC7+emJjoosumu5HDhw9jzZo1\nkMvlyMzMBCEEx48f96hgCgYcDgcqKyuRmZmJTz75xOV3n332Ga699lrk5+czdcerr77q8piFCxfi\nhRdewHXXXYfq6mrExcWdUXQFMIoDMt2+xsfHM444EJ6NNhBQr4nm5mbYbDaIRCIYjUaIxeIhB5+O\nBCqVCnV1dUhKSkJVVRXbykZHR7MPETdLVSgULlkqDdJDFQ1pdpeamjriG4s3vlej0aCzsxMymYx1\n8vX09MBisQTFc4KCe6PMzc1FSUnJoON648zpToTONuRKGN0HbXI7LakON1jg8/ksSKenp6OtrQ0d\nHR0oKCgAj8djc+5GShlxQW/IPT09KCkp8Um5AQPXikqlwjPPPIOZM2di9+7dsNlsOHz4cEi8Syie\ne+45lJaWQqvVDvpdf38/gJ8TsDvuuAMXX3wxXn75ZQDAn/70J8ybNw87d+5EYWEhoqKisGXLlpCt\ndbgYtZTF/v378cADD7BpDhUVFaiqqhr25FyTyQS5XA6r1QqJRMJ8Hai2MlgNGPR5AEAqlQa8Vm6W\nqtFoXApeXOWEyWSCTCaDUChEUVFRSPhowDM9wTXe0Wg0jEpwl98F+jy1tbVBseD01W0YHh4OlUqF\n+Ph4FBUVhcQ/GnDNVj119HEnd1DajRDiEqT9mYFHi5D+eiI7HA5s2rQJb7/9Np5//nmcc845I36t\n/qC9vR0333wzVqxYgfXr1w/KkPfs2YN169YN+nmgoD0IIcBvuzGEwmaz4eTJk9i3bx8OHDiAo0eP\ngs/nY8qUKZg6dSqqqqoglUq9fnC5BkCFhYWD9JeeAiCdhEEDjD8ZoCdbzGCeA27zgEajgdPpdOGj\ng62RdTgc7LwNRU9w25kDNVbiPk8gHheBwm63Qy6XQ6VSIS4uDmazeRDf68tXxF84nU40Njay4Z+B\nyAzdg7Q7r08pJqrF93fIKIVcLsfSpUsxbdo0/OMf/wjZ7tATrrrqKjz88MPQ6XQeA++ePXtwxRVX\nMCfDdevW+TUtnguHwxGSzsb/YSwgewIt6B06dIgFablczsaQV1RUYNq0aUhKSsLOnTuRkpLCDLb9\n3c7bbDYWWGgGSLW93C4+uh5qixnKcUBcT4isrCwkJSW5FDa5AZCucTjZHyGEqQhG0s3nyVjJPQBS\ni9WMjAxkZWUFncen4GqXuQMDfHUbBpKlUvT390MmkyE9PR3jxo0LyufA4XAMyqQdDgesVisSEhJY\nd5uvc2e32/HSSy/hnXfewcaNG70ObAgVPvnkE+zcuRMvvfSS10xYq9Uyumfnzp1YtmwZ6urqhvV8\nTzzxBFJSUgbJ5kaIsYDsL2hQ3L9/P/bt24cvvvgCra2tmDhxIubMmYOKigpMmTIFMTExww4ulKek\nAcZutzMzbrFYDKlUGjLawL2t2pOqgSvL0mg0rOJPt8BxcXFDBhej0chUDaFQT1BVQm9vL9rb22G3\n2wepJoYz3NUbrFYr5HI57HY7SkpK/Hp/hspSPXUb2u121NXVwWQyobS0NGSZp8PhQENDAzQaDbKz\ns9nOSa/Xg8fjDermEwgEqK2txdKlS3H22Wfj8ccfD9ln1BcefvhhvPnmm0wpo9VqccUVV+Df//63\n17/Jzc3FwYMHA7KBbW1txeWXX46pU6dixYoVyM3NBRA0GmMsIA8H7777Lt555x2sWrUKVqsV1dXV\nqK6uxpEjR2Cz2VBeXo6KigpUVlZi/Pjxw+IqqbzMbDYjKSmJ8dH0ouAW5EbyQbDZbMzQJpCBnxTc\n4KLRaAZduJQz525/Q6me4HYn0vZd95FZOp2OKUC4awzUs4PuWqjqZCTvg7fZhrGxseDxeOjt7UVe\nXl5Q5gB6A82+6Vgw9+dxX+Ojjz6KhoYGqNVq3HHHHbjmmmtQVlYWsl2ILwUF14PC6XQiMzMT3333\nnctjurq62Iiw/fv346qrrkJLS4tPRzzAtXN327Zt+Omnn/Dkk0+y93+4NScPGAvIw4Gvu6HRaMSR\nI0ewf/9+7N+/H6dOnUJsbCwL0FVVVT63zlzdsid5Gb0oNBoNG/UkFApZgKb6aH9eA1UB+Du5wV84\nHA6XgpxGo2Hde5mZmYiPjw+aaoILtVrtMpDTV6bOXWOgxkq0OBgdHY3CwsKQFe2MRiNOnTrFrCvp\nANVAOjb9Abe9evz48X4Fl5qaGixZsgRnnXUWfve73+H48eM4fPgwXnvttZBpxtevX4+DBw9Cq9UO\nCsg7d+7Exo0bsXPnTmzatAmPPPII1Gq1i4LihRdewKZNmyAUChEZGYn169f7Ra1UV1eju7sbCxcu\nxKlTp3DuuedixowZyMnJwTfffIOsrCx88cUXwXiJYwE51CCEoK+vD/v370d1dTX279+PtrY2jBs3\nDlVVVaioqEBFRQXEYjG++eYbREZGBjy4lDvqiRqsR0ZGuigSuFm6Wq2GXC5n1flQBhSZTIawsDDk\n5ua6UDLu2t64uLhhX8h0PpvZbEZxcfGwmw48zQ3kdkPGxMRAoVCEvDjIzb7dJXPuU7j9GfDqC5T7\npjUQfwrLzz33HD766CO89NJLqKqqGtFr9RdDKSiC1dDhnhXv2LEDjz76KAoKCiCVSvHXv/4VnZ2d\nAMAK/RdffDG2bt2KtLS0kb7MscaQUIPH40EikWDevHmYN28egJ+r5NXV1di1axdWrlyJzs5OFBYW\n4uqrr4ZQKERaWprfqgaRSASJRMK4MK7XBLdBJDIyEmazGTweL+DqfCDgOqVJpVLWpBATE+OyRqrh\n5mp7AzGp5wYufydW+4IvY6Xu7m6cOHGCZahcz45gZoTU7jMqKspFW85do3sTBlciSBuW6OxAejNx\n7zbkDhn1t2Hl1KlTWLJkCS644AJ8//33QVfd+MK9996Lp556ivHt7giGB4XT6WRJ0I8//giFQoGf\nfvoJP/zwA6xWK15++WWsW7cOjz/+OMLCwvDZZ59h/fr1KC0tDdk4Mk8YC8hBBp/PR2FhIQoLC5GW\nloaTJ0/itddeA4/HQ3V1NV5//XUcP34cIpEIU6ZMQWVlJSorK5mH8VDgOo6lpaUxUb9CoUBiYiKb\ncsLj8Vwy1GBsf6l6IiMjw6dTGo/HQ0REBCIiIlxc0Tz5YXALSbQgp9frUVtbi5iYGFRWVobEKQ0Y\neK+USiVsNhtmzpyJiIgIj/aaw512QuF0OtHa2oru7m6vdp/e4H5DBnx3G1I1TX5+vl9Ulc1mw4YN\nG/Dpp5/ipZdechlXdjrwySefICUlBRUVFdizZ0/Inod+Vp9//nm8+uqrOOuss/DWW2/hT3/6E7Ky\nsvD//t//w/vvv48dO3Zg/vz5+Mc//oG7774bf/jDH0K2Jk8YoyxCCIfDAT6fP+iiIIRAq9XiwIED\njOpoaGhAeno646MrKyuHLCbRLruUlBTk5OS48KrUEY1SHQaDwWWLHhcX53cWxKUnioqKgpY90aIh\n1w+DzkLMyspCampqUCedUHA7+obKvrk3Eq7fhL9dcrTxIikpCXl5eSEpilEpp0wmg8VigUgk8jhW\nyT3bP3HiBJYuXYq5c+dixYoVpzUrpvBHQTFcyoKrKzYajXj99dexefNmfPXVV5BIJFiwYAFKSkrw\n9NNPw2AwYNu2bdi1axdefvllxMTEBDsRGOOQf02gc9D27dvHioZUtUAD9OTJkxEVFYUg527MAAAg\nAElEQVS6ujro9XoIhcKA5HLuzRcWi8Wn9tgbPREKKJVKln3HxsYO4lHdne+GG6S5U0KGmnjhDf4Y\nK0VERKC5uZm5vwV7wCwX3d3dg4aMeus2rK+vR21tLTQaDY4dO4bNmzeHbGqG2WzGeeedB4vFArvd\njquuugqPP/64y2P27NmDSy+9FHl5ecz7RCaTuTzm008/xQsvvICdO3eiuroaS5cuxf79+30+N7c4\nT3cQ33zzDR566CHcfvvtWLRoETo6OnDBBRdg06ZNuOCCC9DY2Aij0YgJEyYE90QMYCwg/9pht9tR\nU1PDGlgOHjyI3t5eCIVC3HPPPTj33HNRXFw87MKdt+YLSh2oVCpkZWUFVIQMFO4t3J6yNE+FTW/G\nT95AqR2lUhkwbeAPuLKxnp4eqNVqhIeHs1FKwaKNuLBaraitrQWPx0NxcfGQ54AQgl27dmH9+vVw\nOBzg8XjQ6/X4xz/+gfnz5wdtXdznMxgMiImJgc1mwznnnIPnnnvOxXyI2+jB/T9XQUEIwT333IPP\nP/+ceVD4Q618++23WLZsGUpLS5GYmIinn34ab775Jn766ScsWbIExcXF2LBhAzZu3Mg+gyHEWEAe\nTdBoNJg1axZuvPFGTJo0CYcPH8b+/fshk8mQkJDgIr0biZ7VYDDg1KlTcDqdiIyMHGRcPpxZfJ5A\nedWuri5IpdKAWsU9NdpwJ1vTIE23q1QyR6mdUN1cuMW0kpIShIWFDVJ2uKsmhpPtexoyOhSsVivW\nrVvHtuSTJ08GMHDTp12aoYTRaMQ555yDTZs2Yfr06eznofKgqKurw6JFi7By5UpUVlZiwYIFWLRo\nEaZOnYp3330XAoEAjz32GIABH+XT4MkxFpBHG/R6/aCtLzV6pw0sBw4cYGOdqKHSlClTIBaLfV74\nvugJu93uEvyMRiPrkKPt4IGoEfr7+yGXy5GcnOyXoY0/4HoLczsNnU4neDweCgsLIZFIQhaMKeUy\nlO57pMZKdMhoWFgYpFKpX5TLsWPHsGzZMsyfPx/Lly8PmZbYExwOByoqKlBfX4+7774ba9eudfl9\nqDwo5HI5HnnkEWzevBkJCQnYt28f7r//fnz44Yf46aef8Prrr2Pp0qUhswr1gLGA/FuF0+lEXV0d\n46MPHz4Ms9mMCRMmsCBdVlbGLsyWlhY2W8xfeoI7kVmj0bioEahlpfsW0Gq1oq6uDlarFcXFxSHL\nyqjSoLGxEampqSxT5c7jowFwpDQCHWQqEAgglUoDDnZcYyXuuXQ3VhIKhVAoFGhtbYVUKvVryKjF\nYsHTTz+Nr7/+Gq+88grKy8uH+zJHDLVajcsvvxwbN2504WiD5UFBCMGf//xnpKamoqysDBMmTMCD\nDz6Ixx9/HAUFBQgLC8OCBQuwePFiXHDBBVAqlad7tt9YQB7Dz7BYLDh69Cjjo0+cOAE+nw+r1Yry\n8nI8+OCDfkvvPMFdjaDVal3sIC0WC+tQ9Gcg63BhMplcxtS7B0ia7dPgRxtEfOl6vb3ejo4OtLe3\ns+w7WOAW5KiW22AwIDw8HBkZGYiPjx/SWe7o0aNYtmwZLrvsMjz44IMhkw4Ggr///e+IiopymeLh\nDn88KNwzYoVCgdtvvx05OTk455xzsGLFCmzbtg3btm2DQCDAnDlzkJGRgbvvvhtbtmxBcXFxUF+X\nnxgLyGPwjtdeew0vvvgirr/+ephMJhw4cADNzc3Iyspiqo6KigokJiYOO3g6nU50d3ejoaGByf+4\n1qT+jsry97mGy0m7Z6gWi8XFoN69G5K2V8fExKCgoCBkxSCqvOno6GDZN5eS4eq4IyMj2TCCtWvX\n4rvvvsPLL7+MiRMnhmRt/igolEolVq5cid27dyMiIgJhYWGDCoiBelC8/fbbeO+99zB37lxceOGF\nyM/PR0NDA+677z589NFHAAY+2z/88ANWrVqF9957Dz/88AM6Oztx11134dprrw3J+fADv72AfNtt\ntzGh+YkTJwAMTN649tpr2RSJHTt2MH50zZo1eP311yEQCPD888/joosuAgAcOnQIt9xyC0wmE+bN\nm4fnnnvutA9sDDVaWlqQkZHhEmioEoHLR+t0OheD//Lycr/8EOx2OxoaGqDValFSUsI6B7ktzBqN\nxueoLH+h0Wggk8mCpvXldkPSddrtdkRHR8PhcMBoNLLKfahgNBpRU1OD2NhYFBQUeMyGucZKx44d\nw4oVK6DVapGXl4c//vGPmDVrFgoLC0OyPn8UFC+++CIefvhh5ObmwmAwwGKxoL29fVgeFHq9Hjff\nfDOsVisuu+wyfPzxx5g8eTL+/Oc/4+TJk3jhhRfw2GOPsfe/vLwcq1atwoIFC6DT6RAeHn5auXMP\n+O0F5G+//RYxMTG46aabWEB+8MEHkZiYiOXLl+PJJ59Ef38/1q5di1OnTuH666/H/v37oVAoMHv2\nbMjlcggEAkybNg3PP/88pk+fjnnz5mHp0qX43e9+9wu/ul8GNpsNJ06cYHz0Tz/9BIFA4GLwX1RU\nxAKG0+mEUqlEU1OTX/7O3DZrGvyoYoJm0d4aL7jGOSUlJSEdrqlWq1FTU4PIyEiIRKKAhroGAjpk\ntLOzEyUlJX7J88xmM9asWYO9e/di/fr1sFqtOHjwIJKSknDjjTeOaD3+wJuCIphDRXfv3o05c+ag\nq6sLKSkp+OSTTxgtYbfbcf3112PWrFlYuHAhsrOzccMNN2D58uWYMGFCKKeABILfXkAGgObmZsyf\nP58FZO6HoLOzE7NmzYJMJsOaNWsADHQKAcBFF12Exx57DLm5uTj//PNRW1sLANi+fTv27NmDV155\n5Zd5QWcYCCHQ6XQuBv91dXVITk5GYWEhDh8+jDvuuANXXXXVsDu/PCkmuKOyxGIxjEYjGhsbkZOT\ng/T09JBdcNRDWKvVorS01CXoe7LVHIljG3dsU35+vl/B/eDBg7jvvvtw7bXX4v777w+1ltYFQyko\nqKqDSsouvPBCrF27NuD2bBpQH3zwQbS1tWH79u0AgLlz52LVqlWoqqpCTU0NNmzYAKVSyYb1btmy\n5bSejyEwZi4EDHQx0TtyWloauru7AQwYlnC3V9SwJCwsDFlZWYN+PoYBUI+M888/H+effz6Agaz4\n4YcfxnvvvYdzzjkH27dvx/PPP4+ioiLmeDd16lS/26A9TWOmwa+vrw8NDQ1seoher4dSqQzqsFQK\n6paWmZmJoqKiQcf2NNSVS8kolcpBrnJ0nVwEOmQUGMiKV69ejerqavz73/9GaWlpcF50ABAIBDh6\n9ChTUJw4cSJoXW5btmxBdnY2Zs+ezc77gw8+iCuuuAKPPPIIvvvuOxiNRtxwww246KKLsGzZMmzY\nsAE1NTWwWCyYOXNmUNZxujHqAzIXQ02e5oLy0TExMRg/fjwA4LHHHsPmzZuZEH/16tXM5e23zEfz\n+XycddZZeOKJJxgn7XA4UFtbi+rqanzwwQf429/+BofDMcjg398Mhs/nsy64srIyJCUlseBHp277\nGpUVCGw2G+RyOWw2m99uaRTeXOUoJdPe3s6KhnFxcQgLC0NHRweSk5NRWVnpV1ZMB/hef/31+Prr\nr3/xLDA+Ph7nn38+Pv/8c5eAnJmZiba2NvZ9e3s7MjMzhzzem2++iRdffBHvv/8++xkhBBKJBMuX\nL2fyucWLF2Pv3r347LPPMG/ePHz88cchawM/XRj1ATk1NZVpbDs7O5GSkgLA+4clMzMT7e3t+Nvf\n/oZ77rkHl156qcuH6L777hsk2zl16hTefvttnDx5chAffeedd2Lz5s2Mj/78889HJR996aWXunwv\nEAhQVlaGsrIy3HbbbQAGuEbaYUizGbFY7NJlmJmZOSgoabVa1NbWIjExEVVVVYxPdg9+3A6+vr4+\nNDU1sWIcDdC+jIC4cwfz8/ODJs8LDw9HSkoK++xRSqa+vh5arRYikQhKpRIGg8HnOk0mE5544gkc\nPnwYb731FkpKSka8tuGip6cHYWFhiI+Ph8lkwpdffomHHnrI5TELFy7ECy+8gOuuuw7V1dWIi4vz\niz9WqVS4+uqr8Z///Afjxo3DNddcw96HefPm4a677sLBgwexePFizJw5EzNnzsR1110HqVQaktd6\nOjHqA/LChQvxxhtvYPny5XjjjTdY4Fi4cCF+//vf4/7774dCoUBdXR2mTZvGOECRSISEhASo1epB\nwcYdH374Ia677jqEh4cjLy8PhYWF2L9/P3Jzc6HVahk1ctNNN+GDDz4YlQHZH0RFReGcc85hnCIh\nBL29vczg/9///jfa29uRk5PDMuhPPvkEl156Kc4777whDXq41qTUZIc7iFShUHgdlWWxWFBTUwOR\nSBRSy0/g5xtMamoqysvL2RRoalhE1wkMZOtUl7tp0ybceOONWLduXcimI7e1teGmm25Cd3c3eDwe\nFi9ejGXLlrk8Zs+ePZg/fz4cDgcAQCwW4+6778b8+fNdFBTz5s3Dzp07UVhYyDwovIF2VFJK7OGH\nH8Yll1yC+++/f9BjH3nkEVx44YV47bXX8Mc//hEA2C72145RFZCvv/567NmzB729vcjKysLjjz+O\n5cuX45prrsHrr7+OnJwc7NixAwBQVlaGa665hm2bX3zxRfYhf+mll3DLLbew7IUbQDdu3IitW7ei\nsrISzzzzDBISEsb46GGCx+MhOTkZl1xyCS655BIAAxdmQ0MDXnnlFdxzzz0oLi7GE088gfHjxzN9\n9MSJE/0uGNLZdbGxsWynwx2V1djYCLVaDbvdznyHHQ4HhEJh0KklboFwwoQJLgVC7jq5j5fJZKiu\nrsaJEycQHh6ODz74ANHR0bjjjjuCujYKoVCIZ555BlOnToVOp0NFRQXmzJkzKODNmjXLo//En/70\nJ/Z/Ho+HF1980efz0YIdn89nDR9ZWVlYvHgxs2aNj493eS+Sk5Oxfv36kPtv/BIYVQGZVl/dsXv3\nbo8/X7FiBVasWDHo55WVlThx4gRTbNAPw5133omVK1eCx+Nh5cqVeOCBB/DPf/4zeC9gDODz+cjN\nzYXD4cDx48eRlpYGq9WKY8eOobq6Gps3b2bBiWvwX1BQ4LfkjBbjhEIhenp6kJaWhuzsbJcMdahR\nWYGCO2TUU4HQE6qrq/GXv/wFN998M/7v//4PAoEA/f396OvrG/Y6hkJ6ejqjFWJjY1FaWoqOjo6Q\nZaDccUrbt29HcXExrr32WmzYsIElUkuWLBl0A549e3ZI1vNLY1QF5FCDboMB4Pbbb2cdR0Px0XQb\nWFdXB6PRiOeeew7Lli0ba1rxgrCwMDz77LPse5FIhKqqKlRVVeGee+4BIQQajYYZ/K9cuRKNjY3I\nyMhg2ujKykpIJBKP58bpdDIjJa6qITIycshRWTExMS4871A3Aa5W2t8JxgaDAX//+99x4sQJ7Nix\nA0VFRex3CQkJIfWl5qK5uRlHjhxx0RZT/PjjjygvLx+2IRAX//3vf7F27Vo8++yzbJ7f1VdfzaZ2\nzJw5E2efffZIXsqvB4SQQL5+U2hqaiJlZWXse4VCwf6/fv16cu211xJCCDlx4gQpLy8nZrOZNDY2\nkry8PGK32wkhhFRVVZGPP/6YHDx4kFx88cXknXfeIUVFReTkyZPkL3/5C1mzZg0hhJA1a9aQBx98\nkBBCyMmTJ12Ol5+f73K8vXv3EqfTSS6++GKyc+fO03IuznQ4nU7S3NxM/vOf/5AHHniAnHfeeWTi\nxInkyiuvJGvWrCFffvkl6enpIR9//DF57733yMmTJ4lOpyMGg8HvL51OR7q6uohMJiP79+8nX331\nFfn666/JoUOHSF1dHVEqlUSv17PHt7a2kl27dhGZTObyc29fer2efPHFF2Ty5Mnk+eefZ+/5LwGd\nTkemTp1K3nvvvUG/02g0RKfTEUII+fTTT0lhYeGInuvxxx8nK1euJIQQolKpyPbt28l9991HCCHk\nb3/7G7niiiuIRqMZ0XOcAfArxo4FZC+47rrrSFpaGhEKhSQzM5O89tpr5IYbbiATJkwgEydOJAsW\nLHAJ0E888QTJz88nUqnUJUgeOHCAlJWVkfz8fHL33XcTp9NJFi5cSP773/8SqVTKjqFQKIhUKiWE\nELJ69WqyevVqdoy5c+eSH3/8kSgUClJcXMx+vm3bNrJ48eJQn4pfLWw2Gzl69Ch55ZVXyM0330xS\nUlJIfn4+ufXWW8mmTZvIgQMHiFarDSgou39pNBrS3t5OTp48Sfbu3Ut27dpFvvnmG7Jr1y6ye/du\n0tnZ6ddxlEolueuuu8gFF1xA6uvrf9HzZrVaydy5c8kzzzzj1+NzcnJIT0+Pz8c4HA72f6fTyZ6H\nEEI++ugjcsUVVxClUkkIIWT37t1k/vz57PHHjh0LaP1nKPyKsWOUhRd44qMXLVrk9fFD8dEU3G3g\nWNNKaCEUCjFp0iRMmjQJn376KVatWoVrr70Whw4dwv79+7F27VrIZDIkJia6SO8C6fwTCoUuNEJP\nTw/kcjkkEgkEAgHq6+t9jsoihOC7777D8uXLcfvtt2Pjxo0h8Wz2Rz1BCMHSpUvxr3/9C+Hh4ayb\n1R3uhkBOp3NIO1D6mnbs2IFPPvkEW7duZZx8YWEhpFIp1qxZg/Xr1zOPaLVajfj4+F/UNvR0Yywg\nn0bo9XpceeWV2LBhw6BurECaVujF1dTUhL6+PsZJjzWueMf777/PgsKsWbMwa9YsAD9rj6mh0j//\n+U90dXUhPz/fxeA/NjbW53mx2WyQyWRwOByorKx0KUIRzqgsOrm7uroa33zzDWw2G9RqNXbs2BFS\nHa0/6onPPvsMBw4cgF6vR2pqKs4991wUFRVh9erVaG1tBTCgonj33XddDIHefvttj+eGcDwk7HY7\nnn76adbqzf19SUkJbr31VixatAhXXnklTp06hc2bNwd9zNavAv6m0uQ3RlkEG562gcOlLBQKBTl0\n6BDZtm0bueWWWxgn/eijj5Knn3560HOPcdKBweFwkJqaGrJlyxZy5513kunTp5MpU6aQG2+8kWzY\nsIH88MMPRK1WM7rh+PHjZNeuXaSxsdEvekKv15MPPviAzJgxg/z+978nN998M6moqCAbN248ba+R\n0mZcLF68mGzbto19z/18Bgp3/ruzs5MsWLCAnHvuuYSQn2kLLnQ63S9O14QQY5TFmQJCCBYtWoTS\n0lIXoftwm1ZaWlowffp0rFixAkuWLIFKpfJJXYw1rgQGPp+PkpISlJSU4JZbbgEw4B1BDf5ffPFF\nnDx5EmFhYbDZbCgtLcXDDz/s12w7nU6HlStXorm5Gdu3b0dubi77HQnM6GvY8Kae6OjoQHZ2Nvue\nUmLDcWejEr1XXnkFUqkU8+fPx4oVK/DQQw9h165dmD17NpxOpws9Q/1LfssIzYCxMbjghx9+wJtv\nvomvvvoKkydPxuTJk7Fz504sX74cX375JYqKirBr1y4sX74cgGvTysUXXzyoaeWPf/wjCgsLUVBQ\ngNLSUpeLa+PGjSgvL8dtt92G/v5+AN4vtI6OjjFO2k9ERERgxowZuPfee/HWW2/hySefhNlsxmWX\nXYbi4mI8/PDDmDFjBq688kqsWbMGX375JVQqFQuyhBB8/fXXmDt3LiorK/H555+7BGMAp4Uq8kWb\nBRPffvst5syZg9jYWGzZsgVLly6FRCLBDTfcgFdeeYWN03I6nSFbw68S/qbSZIyyOOPgLk3q6uoi\ndrudOBwO8sgjj5Bbb72VEELI3XffTd588032d7fddht55513yIEDB8isWbNIVVUVKS8vJ7m5uaSo\nqIgQQkhfXx+ZPXs2KSwsJLNnzyYqlYr9/erVq0lBQQGRSqXk888/Zz8/ePAgmTBhAikoKCBLlizx\nuC0dLWhubib9/f0uP3M4HKShoYG89dZbZNmyZeTss88m5eXl5KqrriLnnXceueiii0hLS8svtOKh\n1RMjoSzc3+sXX3yRfPjhh0SlUpEpU6aQv/71r4QQQlpbW8mdd95J7r333mG+il8txmRvoxlDXVxc\nDbUvTloqlTJN6datW0lycjLZu3fvmEY6SLBareTgwYPk0UcfdZF+nW44nU5y4403kmXLlnl9zCef\nfEIuvvhi4nQ6yd69e0lVVZXPY/b29pKDBw+y72tqaph07cknnyTnnnsuqaqqIh999BEhhBCLxUKM\nRiP59ttvyeuvvx6EV/WrwlhAPlPR0tJCXn755WH/vbeLa7iNKzSIzpkzhxQUFJB9+/aNaaR/Rbj1\n1ltJcnKySxMTF19//TWJiooiAEhERARJS0sjkyZNIp9++inZtGkT2bRpEyFk4HN11113kfz8fDJh\nwgRy4MABn8+7cuVKsmTJEnL06FFy9dVXk6qqKjJlyhTy1Vdfka1bt5ILL7yQHD9+nBBCiF6vJwsW\nLHDZUf3GMBaQz1Q89dRTZOrUqez7/v5+8sEHH7AP71D47rvvCAAyceJEMmnSJHZxDbdxZfz48UQk\nEhGhUEj+8pe/EEIIiYuLY49xOp3se1/0x4UXXsh+/u2335JLLrkkwDMzhuHgm2++IYcOHfIZkIP5\nXtCbeVtbG7ntttvI3XffzXZQW7duJZdddhk5cOAA+f/tnXlQVXeWxz9He2JHAtqT0YpLlnEJJgoa\naOg2artMkQjqyERrhhiDaCcVRSVOiONoZTGTUN1lEmtSMWpXYNLYbY8Lnak2KphkjIILELfgmjG4\nEBsENGIKMWz5zh/3vRsw7gso3k/VK97v936/d+/vAt933rnnnF9SUpLGjh2radOmKTQ01B1zm+IJ\n8s3KE088oYULF7rt3bt3a+LEiVq6dKkkxxdZV1fX5Kmzp06d0pAhQ7R79+5GgixJ7du3l3RpQT57\n9qwiIiLUvXt33XXXXXrllVckSa+++qo6d+7c6APEj+eTvnbOTfNvyPUS5IbXv7a2VpKUmZmpgQMH\nNhLbqVOnKjk5WZK0fft2LVq0SPn5+dd8/Fucy9JYL8qiGdizZw9ff/01x44dc/sCAgLc3RZatWpF\n69atf1Tz9kbfkW6484O/sD9wRYX927Rpw/r163n99deJi4sjKyuL3NxcwCnuv2vXLnbt2uUmrDQs\n7p+VlUViYqJbZ9df3P/gwYMcPHiQrKysG7r+loy/GFB0dDR79+694vnSD0keqampTJgwgczMTIYP\nH87o0aOpr69n3759ACQnJ7Nz504qKysJCwtj8uTJREREXNf1tFQ8QW5iampqOHr0KGVlZYwbN47w\n8HCeeuopzpw5w4MPPkhxcTEJCQmMHj2a+fPnc/bsWXfuhVJqdQ3xq+Xl5VRUVAC4Oz/06tXLjZEG\nfhQjvWzZMqqrqzl8+LAbI92pUyeCgoLIy8sjICCAJUuWEB0dTW1t7UXDuS4UI11SUuLGSJuZGyPt\nceWEhYVRVFREQUEB06dPJzY29rLn+o0AM6OmpobNmzfzySefEB4ezooVK1iwYAGTJ0+mrKyMefPm\nkZOTw6xZs3j44Ydv6C7gLRVPkJuY3NxcunXrRmpqKtnZ2Xz88ccMHTqU+++/H4C4uDj69+9PYmIi\nW7ZsYevWrQDMmjWLjRs3uv8gfiuyoeVyNZSUlDB06FBCQ0OJiIggKirK3S34amOku3fvzueff058\nfDxRUVFejHQzExQU5CZcxMTEUFtby4kTJy45T5JrBLz//vskJCTwzjvvMGXKFF544QUmTZpEZmYm\nhw4dYsaMGRQUFJCWlkafPn149913W2QK/o3Gy9RrYj777DMeeeQRt/3999/Trl07goODycvL48yZ\nM+5uECdOnGDhwoUMGDCAQ4cOucH0J0+epFu3bpw+fZq6ujrWrVtHr1696NGjx3mPeW5GVENCQ0PZ\nuXPnj/rvvvvuqy7s76fhbsRecf/m40qLAfl37jAzKisrSUlJoaioiPj4eGbMmMGePXsYPHgw4eHh\nREdH88Ybb7Bs2TISExOJiopyjQuPK8ezkJuYjIwMIiMj3XZxcTHl5eWEhISwdevWRr628vJyAgMD\nKS0tJSAggM6dOwOwadMmt2j5woULWbRoEQkJCQwePJjCwkLAEcPy8nLgwq4Ov9+vrq7uhqz1XJ90\n69atadWqFc8++yz5+fnApX3SfoqKisjNzXU3Bfjmm2+IioqiZ8+eREVFuRY3OIWUevToQXBwMOvW\nrXP7t2/fTkhICD169CApKanJUpWvlkmTJtGxY8dGOzk3RHKqswUGBtKzZ08OHDhA165dSUtLY/Hi\nxe7+dhkZGfTp04e+ffuSlJR0wWJAfhreu8jNzSU1NZVRo0YxfPhwUlJSWLt2LXv37qVt27aMHDmS\nO++8k7y8PJ555hlPjK+Vy737Jy/K4rqQnZ2tkydPuu01a9ZozJgxKioqUnp6eqMMpscee0wrVqzQ\n2rVrlZycrMLCQklSUlKSpkyZIkmKj493IzZKS0v13Xffad++fXr55ZcVGRmp/v3764MPPvjReVRU\nVGjOnDkaOHCg+vXrpwEDBmjlypWSpG+//faq11dWVuZmsFVVVWngwIH66KOPrjlGOjg4WIMHD3aj\nBW6HxJVLhbOtWbOmUSJHZGTkVR/r3KSV8ePHu383M2fOVGxsrPva888/r8TERFVUVEhyYow9LokX\n9nYrcObMGe3YscMNc4uNjdWgQYMUExOjOXPm6PTp01q6dGkjoe7du7eb4pqamqqxY8dqyZIlkhwR\nHD9+vJ577jlJTszy3LlzVVpaKumHGNL09PRGsdA5OTluyNny5cv13nvvSXLCm64k3OyLL75Qv379\nFBISot69e+u1116TpGsq7n/fffepS5cu+vTTT11Bvl0SVy4WznY9qrOduxNHVVWVJCkjI0MdOnRQ\nVVWVjh8/rqefftrNCi0rK9OgQYNaSuH4psIT5FuVgoICrVq1yhXC4uJihYWF6cUXX9TixYtlZior\nK3NjQY8ePapHH31UCxYsUHZ2tmJiYtS/f39FRESob9++ioiIUG5urqQfBDktLRIWVtsAAAViSURB\nVE3Dhw9XSUmJe9zq6mrV19erpqbmvCLcXHHAY8aM0bZt2xrF094uiSsXE+QRI0YoJyfHbQ8bNuyS\n2XUN2bFjh6ZPn65NmzZJctKdMzIyVF1dLUkaN26caxlnZWUpKipKW7ZskSTXOva4bLw45FuVkJAQ\nRo0a5fr5OnXqxIIFC7j33nupqakhNDSUDh06kJ6eTkpKCmVlZcTFxbF//37uueceKioq2LJlC/n5\n+Sxfvpy33nrL3XXB7x8cMWIEgYGBjBw50q1cdscdd7B161YmTpxIYWEhK1euZPbs2WzevJnKysoL\nFiG/kaxevZqOHTsSHh5+wTFXUtzfT319PTk5Oa5Peu7cuXTp0qVRNT4/LcUn7ccfodO2bVsCAwPZ\nsGGDG72zceNGN0556dKlbNiwgQ8//JDHH3+cAQMGcPz4cQDatWvXPCff0rlc5fYeN98D6AfMBzKB\ndKALEASsAYY1GNf6Iu8xGlgPpPnajwO/AzoD4cDbwErge+BF35hgoDdgTbDG3wDHgCPAcaAK+CPw\nJdDJN6YT8KXv+WxgdoP564D+vjEHGvT/ATgIrPa15/rXd87xHwa+ANoAfw8U+q8nkA/8EjDf7yD6\nBqz/AWDPBV77HfBkg7Z7TS7yfrOBOUAbX/tXwCLgn4A7fO85Dejsez3d97v/G+Anzf0339IfXtjb\nLYykXcCuc/vNbB7wGzNrDxQAv204zsx+CrSXdFzSX8xsP7DezH4CdAWqJRUDxcB2MxsLdARWmNkw\nYCDOh8FDZvaSpD/fwDXOxhERzGwIjmiON7M3gQm+tU0A/uKbsgr4k5nNx/lQ6QnkS6o3s2/N7Jc4\nAh8NvOVby8UYDSyTVA0cNrOvgEgzOwIEScr1ndsSIBZHmJuKVcA0M1sG/AI4LankEnMmAd2Bn5nZ\nl5JSzSwYR5h34wjyDOAOMzsLlABTJdXesFV4uHiC3AKRtBF41MzaAPfhCCtm1lpSPY7lO9HM/gDs\nBAYDpyTVmdk9QKlvfAfgJSAMGAq0BpYAucBE3+H+bGYfSappsgU6/BbnA+LXwFHgnwEk7TWzFcA+\noA5HTOp9cxKB3+NYnWtw1tFQkKebWTywDUiWdArnW0dugzHHfH21vufn9l83zOy/gSHA35nZMeBV\nHEsVSYuBtUAM8BXON4eJ53+nRjyGY/F/BUSZ2UM43xSOAE8CbwD/BYwCfg5MkbTvui3K46J4gtyC\n8Vl1Bxu0/cJ0FCgCZuKIUy6Q4BPgB4APzewBYDpwQtIgADP7FdAexy2SBfwtEEATxbNL2gBs8D0/\nCfzDBcalACnn6d9mZv8OxEhK9FncfhYBrwPy/Xwbx5psNiQ9eYnXBUy9wvc8bGb/CfyLpGFmlgD8\nGqjBcU3ESvoffNfZo2nxBPk2RNIx4D/8bTNrK6nKzLoBX+NY1G/jWGf/ZmYPSDqCI8Q7JSX65gUB\nHYB6bh0GAP9oZjHAT4EgM/ujpPH+AWb2PrDa1/wrcG+D+V19fX/1PT+3/6ZH0itmts/MZkp608zy\ncKzjl4ACM1sD1PoE36MJMe+ae5wPM+uN83U+DOem2Js4ftp5wP8B7wGBwDeSbsmN0Rr4pEeaWSe/\n/9XM/hX4haQ433X4ExCJ45P+X6CnzyedDyQBeTjug3clrT3fsW42zCwEyAZ+LqnQ1/eQpP3Ne2a3\nN56F7HFeJO0F3DqNfv+zz685FedGWiUwFufGz63OPDPrh+OyOAI8B5ftk74T52ZeU97QuyYk7Taz\nVOBTnOgRgAPNeEoeeBayxzVgZh0klTf3eXhcPWY2SZJX5ekmwRNkDw8Pj5sEL1PPw8PD4ybBE2QP\nDw+Pm4T/B3WPicWl+IbdAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x847d630>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot sample points.\n",
    "# plot dots ('o') for each sample (house size, house price)\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "\n",
    "ax.scatter(X[:, 0], X[:, 1], y)\n",
    "\n",
    "ax.set_xlabel('House Size')\n",
    "ax.set_ylabel('Number of Rooms')\n",
    "ax.set_zlabel('Price($)')\n",
    "ax.set_title('House Prices')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def GradientDescent(X, y, max_iterations=100, alpha=1):\n",
    "    m = X.shape[0]  # number of samples\n",
    "\n",
    "    # y must be a column vector\n",
    "    y = y.reshape(m, 1)\n",
    "\n",
    "    # initialize the parameters to zero\n",
    "    theta = np.zeros(shape=(3, 1))\n",
    "    # three parameters in this particular case\n",
    "\n",
    "    # Repeat for max_iterations (it would be nice to also check convergence...)\n",
    "    for iteration in range(max_iterations):\n",
    "        grad = np.dot(X.T, (np.dot(X, theta) - y)) / m\n",
    "        theta = theta - alpha*grad\n",
    "    return theta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# prepare x with a column of ones (this is the x_0 for the bias term)\n",
    "x = np.ones(shape=(X.shape[0], 3))\n",
    "x[:, 1] = X[:, 0]  # include the 1st feature (size)\n",
    "x[:, 2] = X[:, 1]  # include the 2nd feature (number of rooms)\n",
    "\n",
    "# Scale features and set them to zero mean (standarize)\n",
    "mu = np.mean(x, 0)\n",
    "sigma = np.std(x, 0, ddof=1)\n",
    "x[:, 1] = (x[:, 1] - mu[1]) / sigma[1]\n",
    "x[:, 2] = (x[:, 2] - mu[2]) / sigma[2]\n",
    "\n",
    "theta = GradientDescent(x, y)\n",
    "print 'theta result:'\n",
    "print theta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting plane should be plotted like in the following example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAADuCAYAAAAOR30qAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsvXd4I+d5LX7QO0GwF7AB7FxuIbdK1kqKrchFkVXiEkeW\nbcU3uY7t2Ck3bjfPlW9u7J8dXV/bsR3HsZPYcdFPipuapWvFUWRLW7R9l7ssqATBBqL3wQzm/oGd\n0QCYAdHI5e7OeR49axPghxkQOPPO+51zXglN0xAhQoQIEdce0mt9ACJEiBAhIgeRkEWIECFih0Ak\nZBEiRIjYIRAJWYQIESJ2CERCFiFChIgdApGQRYgQIWKHQCRkESJEiNghEAlZhAgRInYIREIWIUKE\niB0CeYXPF219IkSIEFE5JOU8SayQRYgQIWKHQCRkESJEiNghEAlZhAgRInYIREIWIUKEiB0CkZBF\niBAhYodAJGQRIkSI2CEQCVmECBEidghEQhYhQoSIHQKRkEWIECFih0AkZBEiRIjYIRAJWYQIESJ2\nCERCFiFChIgdgkrDhUSIKAmapkFRFABAJpNBIikrU0WECBEQCVlEnZDNZkFRFEiSRDqdZn8ukUgg\nk8nY/6RSKaRSKSQSiUjWIkQUQCRkETUhm82CJEm2KpZIJCzh0nQurZUh6sLfSyaTMJlMkMvlIlGL\nEAGRkEVUAZqmQdM0MpkMstksALBEypAw8zPuv1wQBAGXywWdTgeCIPJ+RyqVQiaTiUQt4qaDSMgi\nygZN02xFXEjElYL5HZlMxvsaFEXlETXz3ML2h0jUIm4kiIQsYlMUEjFDgrUQYWE1zf0537pcoqZp\nOu85XIIu7FOLEHE9QSRkEYJgFBMkSbIkWK+KVIiQSz2/UqJmCJpvQ1GEiJ0IkZBFFIGPiKXS+krW\nKyXkUusIETXT5yYIIu85BEFAJpNBp9OJRC1iR0EkZBEsaJoGSZLw+/2IxWIwm80VE3EikcDy8jLU\najV0Oh20Wm3dybwclCJqv98PAOjs7Mx7TKyoRVxriIQsgiViRppGkiQSiURFRBSLxeBwOJBIJNDR\n0YFkMomNjQ0kk0lks1loNBrodDr2P6VSWZcKuVIw58SQLwPmWEiSRCaTKfqdQtWHaHoRsRUQCfkm\nBtfMAbxeVUqlUlZFsRkikQjsdjtIkoTFYoHJZCrSHNM0jVQqhXg8jng8Dr/fz/7vixcvQqvVskSt\n1WqLlBfbASGJXqGWmmnh0DRdsqIWyVpENRAJ+SYEn5mjcDNss+o1GAzC4XAAAKxWKxobGwG83rvl\nrieRSKDRaKDRaNDS0gIgR3BnzpzB4OAgS86BQACJRALZbJZteVwvRF34O1KpFHK5XCRqERVBJOSb\nBKXMHIUQqpCZ/qvD4YBSqcTQ0BAaGhqqOh7mtQuJmnkdbkV9PRJ1NpvNs5Bzn0tRFPR6vUjUIoog\nEvINjmrMHIUVMk3TWF9fh9PphE6nw8TEBHQ63aavW+o1SqksNiPqRCKBeDyOYDCIeDzOEjW39cEo\nKISObatQyp1I0zRisRjcbjfGxsbyHhNNLyIAkZBvWNRi5pBIJMhms8hms1hdXYXb7YbRaMSePXug\n0Wg2/d16Sdr41maIurm5mf05TdNIp9NsRe31epFIJEBRFFQqVR5JMxeSa0F0XHdi4YbiZqYXbutD\nVH7cuBAJ+QZDvcwcsVgMx48fR3NzM6ampqBSqep6nPUkFIlEArVaDbVaXRZRJ5NJKBQKxOPxvNaH\nXL71Xwe+OwfR9CKCgUjINwjqYeagKAoejwcejwcAcOjQISiVyq043G2BEFEvLi4CAHQ6HRKJBG9F\nrdVqodfr607Um7VyCo+/UtOLSNTXN0RCvs7BaIhjsRg8Hg+Gh4crJuJMJoPFxUWsrq6iq6sLU1NT\nuHLlynVNxptBoVCgubm5qKImCIKtqJeXlxGPx0FRFJRKZVHroxqiroSQhVCKqAF+LbVI1NcHREK+\nTlFo5gBQsZmDIAi43W6sr6/DbDbj8OHDkMlkeUqMGxV875NEIoFKpYJKpUJTUxP780KiXllZqZqo\nGf3yVp6TkPKDJEksLS2Boih0dXUBAEvMfBI9EdsPkZCvMwiZOWQyWdkkmkql4HQ6EQwG0dvbiyNH\njuSRRK2bcvF4HDKZDCqV6ob4Ypci6kwmg1gshkQigdXVVcTjcZAkCYVCwbY8mH8VCgW7wbrdx8/8\nyzW0iKaXnQeRkK8TlGPmYB4TQiKRgNPpRCQSQX9/P0ZHRyvSIZcCTdMIBAKw2+3s7zMhPnq9vsg2\nfSNAIpFAqVSiqamJl6iZippL1MzvcfXUCoVi246Zoij29UTTy86DSMg7GPUwcwCv50wkk0kMDAxg\nfHy8ao0w3zFubGzA4XBAo9FgfHyczamQSCQgSZIlpo2NDbhcLmQyGcjlcqRSKSwtLbGEvZ3EtJVg\niFqpVMJkMuU9trKygmAwCJqmsba2lldRF7Y+tuL9yGazmxppqjW9iJNeaodIyDsQ1Zo5Cgk5HA7D\n4XCwORNNTU1l65DLOUafzweHwwG9Xo/JyUlotVoAuU1C5ssrl8thNBphNBrzfj+TyeDUqVOQSCRY\nX19HLBYrIiaGqOstR7sWoUYMZDIZtFotzGZz3s+5PWqh94Npf9RC1Nlstuoe9mamF3HSS+0QCXkH\noVYzB4NgMMi2DiwWC5szUa9jXFtbg9PpRENDA3bv3s0ScSVQKBSQy+Xo7u7O+zmXmLi3+tzNs3rI\n0a4VGQgRolBFzW19+Hy+vDuMwoq6nFZQLYRcCuVoqZn/73K5MDAwIJpeeCAS8g5APcwcjOri5MmT\nUCqVGBkZgcFgqOsxrq6uwul0wmQyYd++fVCr1XVbnwEfMZWSo/E58a5FtkW5qFT2plAo0NjYWHRR\nrZaot4qQhVD4Oc5kMohEIuwdnWh6yYdIyNcQ9TBzcHMmMpkMpqenN82ZqATZbBYrKytwu91oamrC\n9PR03V17m6GUyqHQicfNthAKIbqWLYt66JCB6ok6Go2y74dCodh2kqMoCnK5XDS9CEAk5GsApppd\nWlqC0WisaqoGkzPhcrnQ2NiIPXv24OzZs3UjY5qm4fF4sLi4iNbWVuzfv3/HqSNKWaaF0uI0Gg0y\nmQwMBsM1mWhSL0IWQimiTiQSiEQiCIfD8Pl8eaOsCivqrTpGiqJK3sHc7KYXkZC3EYVmjlAoxFZy\n5SKbzcLr9WJxcREtLS11r1gpimIrzVQqhQMHDuw4It4MpdLikskknE4na4pJJBIAUDTRRKPRbAlR\nb6UxpBQUCgWMRiPUajUGBgbYkCiuCsbv92NxcXFLiXozQhZCOaYXPqL+yle+gk9+8pPXjYJHJORt\ngJCZQy6Xb6odZsBU1F6vF+3t7Th48CDvh6zaCozJsfB6vejo6IBOp8PQ0FDF6+xkSCQSNqJTq9Wi\nra0NQO7vk0qlEIvFWJVDMpkEgKJIT41GUxMpbXWFvBkKe8hCKphCovZ4PEin0zUTNUmSde3xb0bU\nP/7xj/GZz3ymbq+31RAJeQuxmZlDJpNtSsiFOROHDh0SVBcwOcaVfOFJkoTH48Hy8nLe+mtra2Wv\ncb1DKpVCq9UWqUWy2SySySTi8Tii0ShWV1eRSqUAgJWgMaSkVqvLet+3e1OtEBRFlfX6pYiayaMO\nBAJ5RF148eJzajI95K0GV0t/PbUuREKuMyoxc5QiZIIg4HK54PP50NPTw+ZMlALj1ivnC8cl+u7u\n7qL1mQ90NR/mbDZ7zSvBUih3U08qlbLkwlTTQO78EokEYrEYwuEwlpeXkUqlWGLnEnUhKV3r96XW\nC4JcLkdDQ0PRpBguUQeDQSwtLSGdTue9hzqdjn2fthM79XPIB5GQ64RqzBx8hMzNmejr68Pg4GDZ\nH+ByLM+ZTAYulwvr6+sliV4ikZTl6uJbm6muJRJJ3pdRr9fvmHyLWo5BKpVCr9dDr9fn/ZyiKF5S\nYqpHvV6PRCKR52TcbmxVD1uIqCmKYlsfwWAQfr8fJEnC5/MVqWDKvcsoF/Vuj2wHREKuEbWYOWQy\nGWtBTSQScDgciEajGBgYEMyZ2Gy9cipuvkChQjDkXs4Hmpsa19vbi4MHD7J3CnyVJNOHJAgCgUAA\ner3+uts45INMJoPBYCjSf3OrR+bfpaWlvH4sU1Vvx/uwnRcCmUyWR9SMzry1tZX9bIRCIXi93ry7\nDC5ZV0vU4XC4qOWy0yEScpWoh5lDJpMhmUziwoULSCaTsFgsmJiYqPoLw1chp9NpOJ1OBAKBiiru\ncvIsuETc19fHkjyjJBGqJBmCCgQC2NjYgNvtRiaT2Rbb9LUAt3qMx+NsGFGpnI/CO4vrRSWwGUiS\nZDXhfBcv7l0GXzuoEqKORCJ1daluB67/T/s2ox5mDiB39XY6nUin09i1a1fZOROlwCVkbuujv78f\nIyMjFa1fqv3Brba5RFwuGIJSKpV5FwgmypKxTcdisSI3HmObruVWdKcYQ0rlfHDNHU6nc9tyPrYa\nm8ne6knUoVBIrJBvVDAaVuD13nA1RBwIBOBwOCCVStHd3Y1oNJpnaqgFTMXt8XgQiUSqbn0A/IRc\nKxEXorAKVygUMJlMvLZphqg9Hk+eyYNLTpVoh69VH7uc3rGQuaOcnI9applsB6rVIZci6mQyiVgs\nhkgkgpWVFSSTSfzgBz+Aw+FANpvFU089hfHxcTY/o9bj379/P7q7u/HMM8/UtBYfduZfbQeBa+aY\nmZlBf39/0cZFOWtsbGzA6XRCpVKxORORSAShUKgux8m0AAKBAIaHhzeN2NwMXLKsNxFzsVm1yrVN\nF7rxGEmakHaYIep6bxbVglo288rJ+eBOMxGauH0tUS0hC4HJ2y5si01OTuKf/umfcPr0aZw6dQrf\n+9738Fd/9VfYs2dPTa/3la98BWNjY4hEIjWtIwSRkAXAZ+aoxMgBvJ6M5nK5oNfrMTExkfelKCdU\nfjNws471ej26urryJFrVQiqVIp1OY2lpCX6/v2LFRzmo9YLBaIdbW1vZnzOSNL5bW4akk8kkNBrN\nNVE61Ps1S+V8cO8sGPdlPB7HhQsXBHM+thokSW5L9a7RaKDX63Hrrbfi4x//eF3WXFpawrPPPovP\nfOYz+NKXvlSXNQshEnIBSpk55HJ50QQFoTUKcyYYqyoXlRI8F9FoFHa7HQRBwGq1oqmpCXa7vS79\nUYIgEI1GMTMzA4vFgqGhoYqIuFzCqSQIv1xwNxLb29vZn3PlV0xbx+PxQC6XF0002coNtO0yhvDd\nWVAUhbNnz2JoaEgw56OQqOt9rPWukEshFAqhp6enbut9/OMfxxe/+EVEo9G6rVkIkZBRvpljM2dd\npTkT1YxKikQisNvtoCiKDZ0v9/g2A0EQcDqd8Pv9UCgUGBoaKsrnrSe2gpCFwJVfJRIJGI1GtLS0\n5G2gcSd4KJXKIqKuB5FcS2MII2MslfPBVX0kEgnQNM1rH6+WqLeTkOupsnjmmWfQ1taG6elpvPTS\nS3VZkw83NSFXauYQqpDLzZkodz0+hEIh2O12AIDVauX9oFVD8EBOGudyueD3+9Hf34+hoSHMzc1t\nOVley74u89p8G2jcvmwsFiuK9eQSdaVV5LUk5FJkWKoFxCTnxWKxvF4936bqZue2necfiUTqprJ4\n5ZVX8NRTT+G5555DKpVCJBLBQw89hO9///t1WZ/BTUnIjHSNoqiKzBxyuTwvUaqSnAk+lFMhcqd/\nDA4OlvyAVUrIXI1yf38/hoeH2feAycWoBswdR7nP3Wko1ZflkhNTRQL5IUR6vV5wI/FaV8iVVrbc\nnI9Coi7M+Ugmk3kBTnybqtt57uFwuG4V8uc//3l8/vOfBwC89NJLeOyxx+pOxsBNRsi1mjlkMhlS\nqVRVORN8EHpdZoKzw+GAQqEoe/qHTCYrmmfGh0Ii5tMoM9bpSsCE2btcLtaiy812YBx53C/nTiRk\nIQjFegqRU2GOg16vv6bhQvV87c1yPoT0wgRBwO/3C4YP1RPhcHhLW25bgZuCkOtl5shms1hbW8Pa\n2tqWqA4KJziPjY0VyXlKYTPVRjlEzF2rXEKmaZol4paWFkxNTbFky9hjg8EgPB4PCIJgN9KYWz+5\nXL6tutmt2EjkIyfGzMCc/9LSEkKhEC5evMgG5G+nE287LgabbaqGQqEiq3ThRateAflbRch33HEH\n7rjjjrqvC9zghFwYCF8tETM5E6FQCCqVCvv376/rlZ07wVmn02HXrl1VaUaFSDSdTsPhcCAYDGJg\nYKAs1145LQvunL3m5mZ2qghFUchkMoJifsaRFwqF4PP54PF4QFFU3tglxpG3VQSyHbfOfOd/9uxZ\njI6OspI0rhOvcJBrvTYSGZSbBLgVYHI7VCoVrFZr3jFxw4e4gUx89vFK/m7xeHxHaK8rwQ1JyDRN\nIxaLIZ1OQ6fTVZwxwSAajcLhcCCVSsFisaC3txdOp7NuX2ZG2XH8+PGaJjgzkMlkeYTMtU9X6tor\n1bJg9NUOh4N3zl45TjSTyQStVouBgQFotdqi/qzf7y/qz+5Eo0eloGkaCoUCGo2maD+AqxvmDnKt\n14Wq0vS+eoNvU7EwfIhBYUA+k3dSbs4HU0xcy+zpanBDETLXzBEKhRAMBjEyMlLxOuFwGHa7Hdls\nFhaLBSaTCRKJBMlksmYjB5BfWVIUhT179tRExAyYlkUtRMxdq5CQmYGqDocDjY2NmJqaqmnydGGI\nuFB/tlRiHLc/fT0E8JTa1FMqlWzwEPf53PmA3AtVpSqHax2OX4kppNycDyaQiZvzQdM0e9G+3i7c\nNwQh85k5FApFxeTJzZmwWq1FH4ZKZGpCx8md4Dw1NYULFy7UjUhIkkQgEEA4HK4pxwLIr5CZlord\nbofRaMS+fftqImIuNmuLlEqMY6rpwtt+LklvpwutHFSqsihnIzEWi+VtJAqF5F9rQq6HBrmcnI9X\nXnkFf/d3f4fFxUW8+c1vxsTEBB588EHccsstVb1mKpXC0aNHkU6nQZIkfvd3fxef/exnazoPIVzX\nhMzoRfk0xAqFomjoodAazEaaWq0uqWio1niRzWaxvLwMt9uNlpaWvAnOzJq1kHIqlYLD4UAgEIBC\nocDBgwfrkhzHBInb7XYYDAbs3buX13FYLWo5Rr4KarMgIobYCYK4ZuqOesneuJthXHBT0UKhUF5P\nlilUgsHgtmUvFx7bVl0cuTkf73rXu/Dbv/3beOihh/Av//IvmJmZqamXrFKp8Ktf/Qp6vR6ZTAZv\neMMb8Ja3vAWHDx+u4xnkcF0TMlc+Vfgh36yaZfqgTqcTKo0GnT196OlsL/llqVSbm81msbS0BI/H\ng7a2Nt4JzrW46xgiZirivr4+zM/P1/yFp2maTRRrbm6uubcthHrL3jYLIorFYohGowiHwwiHw7xx\nlvXa4d/sOLcKpULyXS4X0uk0760+t6LeKsXLdtumTSYTOjo60NHRUdNaEomEvUPLZDLIZDJb9je8\nrgkZEFYWCFXI3LaByWRiq77nXzmL9/2PbyAQiaLNZESLqQFtTUa0mYxovfq/W01GLPtCCEXjaDQI\nX3ELJziXcu5VQ8iFRDw2NgaJRIJUKlWVU48Bo3+22WyQSCRob2/H6Oho1etthu3SIXNdaG1tbaAo\nCi0tLWxgPN/GEVNN7/Q4y3Ihl8tZFUdnZyf7882S4pj3oB6tn+0k5HqaQoDcsU9PT8Nms+HDH/4w\nDh06VLe1ubi+P2UlUEjUFEXB6/XC4/GgtbW1SBnw5lv34a7De/DDX7yMv/3ez3HRtsi7rlIuw/R/\nXsCsexlapRKtTcYcWTc2oKXRALmEAjJpDA30YmJkEO2tTSXbEZUQcjKZhNPpRDgchsViYYlY6Jwr\nAUPEarUau3btYqvJrcS13nAR2uHn2qa5JMWoHRiyriXT4VqAT2UhFOmZTqcFA4i4F6pK3oPtSnoD\n6j++SSaT4dy5cwiFQrj//vtx6dIl7Nq1q27rM7hhCZn5snPH3G9erUrx3nvuwANvPIJv/tsL+MYT\nv0A8mc57DkFSOHZhHqYGHXo6WnDqsg0kxU+Cw31dUMhlcHrX0WpqyKu0mf9NJmMYiKZg7aPQ1mSE\nWll8bMlkEg6HA5FIhJeIXz9+WcWEHAwGYbPZoFQqMT4+zt6aMV/AapDNZpHJZEoGKzG4Fr3czV5T\niKS4sjyfz5eXv8zdSNwpg1wLUe6mnkQigVqthlqtFsygLsy1KNxI5JMmUhS1bYS8VeObGhsbceed\nd+L5558XCZkPQh/8TCaDdDqNEydOoLu7u6KcCZ1GhT9/7714+J7b8dj3nsIPnnsZZEEVG4zEcfzi\nPNqbGtCgVWFhyVe0zrx7GQCwZ7gfkVgCpy7beV+vUa/F6IAZpy7boFWrcuTdZERzgx5KKQ2tUoah\ngT6MWPqwGk4iKwuhpdEARcH5VJKvzGRkyOVyjI6OFvUcq2knMI49p9MJmUzGjh1iKspC1cO1tE5X\nSpjlyvK8Xi+7icYl6a2O9SwHtaosNsugFpImMv8RBFHWRboeqGeF7PP5WHVHMpnEL3/5S3ziE5+o\ny9qFuO4JuRBMctnGxgYkEknFgT9ctJqM+MLH3os/fPAu/M23/w3P/eZM0XPWAhGsBXLVsFwmxWXH\nUtFzzs+7IJVKcHBiCK7ldawHw3mPh2IJHL84D3NbM9qbjTh9xQH70lrBKqegVauwZ7gf5+edSKYz\nMBl0aG26Wm2bjGhtakAyEsJ6Wo5Gg5atwpuNBvaLGA6HYbPZIJVKSypKKrVOFxpFmI1WRvUQi8Xy\nVA9arRbpdBpKpRIqleq6NXuUI8vjxnomk0nYbLY8ot6utsdWyd42G2bLGH18Ph/W1tawtLRUdLGq\nd+UcCoXyrNu1YGVlBe973/vYMLJ3vvOduOeee+qydiEkFVYoOy4JhjGCcM0QfX196OzsxOnTp7Fn\nz566yXteOXsZn/7q9zDnWRd8zt6RfgQjcbhXiitmAFCrFNg3MoCLC4uIJVO8z+lpM0GjVmF+cZX3\n8WajAYO9nTg1YwPFQ5o6jRq7h/pwbs6JZJqATCqFqUEHg0YJk0GL3q52DPZ2o9GgRVODAW1NVzcx\nTUY0GXNfqlAohOXlZYyPjwuea6E+2WKxQK1WI5vNgiAIwS8/o6G12WxsSD9TUXG/qHq9fktucefn\n59HW1rbtE4lpmsbJkycxODjIknU8Hs/LHK4kyrJSXLlyBWazuaygqq3A3NwcOjo6oNVq2XNn3od6\nD7P93Oc+h/379+PBBx+s81lUjbL+mNd9hZxMJjE3N4dYLIb+/v48MwQjfauVkBmylyRD+NsPPQBX\nIIWvPP4LnioWODfngkwqxaFdQ7B7V7ERzN8YS6UzbA/68OAwXuMhVc96EIAwufvDUfgvRtHT0YJW\nUwPOXHHkPR5PpnDswhxaGg2YsJhxZtaJjVAUGyHAueLHmXkPgFNoazJioKsNr83YkL16YVYq5Ghp\nNMDUoEODWomDe65AAgmrOmF64XJQWFv2QqfTFemTN3NIcTW0XGdaYVXJBPGrVKq8tsf1tpnGgMlS\naW5uzuvNbpYWV5iWVy12ijGknGG2S0tLbAZ1NcNsr8ekN+AGIGSKotDe3o6JiYkiEijXHCIE7mYa\n43ybnZ3FG61WPHDXrfjXZ/4T//tfn8JGKH/gIZXN4sSlBWhUShzZPYwLC+6izUGmB93aqIdRp4bN\nu1H0+gy5H9w1BKd3Db5g/ut4VjfgWd3AaH83pBIJLjvz2yU5Eo6ip70JrSYjzsw68x5fD4SxHgij\nv6sNjQYdzs05QWRILPuCWPblLgrHZhzYZe1FOpPBwuJK3u+rlQq0NhkxNWqBXCaFTqNmCdtk0LEE\n3mZqgFpVTCSFPWQhs0c6nWbbHtzNND6yKqeqvJaZxHwolRbHzfZgZHlMX77SEKKdQsh8KKUhT6VS\nLFGXO8y23rK37cJ137JgdvT5MDs7i9bW1rw/cDlg0t1isRgGBgbQ1tbG/qEXFhZgNBrZL048mcLX\nn3gez7x8CuFoHMFIDASZv7HWZNRjuLdrU0WGXCbDZYeH93GNSom9I/0lWx37RgbgXfNhPRTjfXxs\nwAwaNGadXt7Hxy09oCgKc1c3I7mQSCQY6WnDRjiOjXDx+gatBrsGe3F21oEUUfz3uGXPCD71gQew\nZ7iP/RnT6uBukpWLbDabd9sfi8VAEERZZMXcOtdTFlUuXnvtNRw4cKCmNbiVJPNv4TQTvjuJ8+fP\nY3R0dNs21gpx7tw5jI+P16WFyM1dZt6HVCoFkiTxjW98A8FgEB/84Adx1113oaOjo+oLsMfjwcMP\nP4y1tTVIJBL84R/+IT72sY9Vs1RZB3BDE7Ldbi/KZS0F7gRni8WClpaWoj+k0+mESqVCV1dX3s/X\n/CH87Xd/hsdfeAVqlQKNBh30GjU0aiWUcjmkUikUMglCkRhiiSQyWSCRSiMYibPtAgDYM9KPcDQB\n1zJ/n9rUoMNIX7cguUslEuwd6cfi2kZRu4TBvtEB+ENRLK4WV+UAMD1mxao/AO/V1gkXSrkc0+MW\nXHF6EYrGix5vbzKir6sVr83YUPjRkkgkuOe2afzl+96Ovs5WOBwONDQ0VEXIQuBqiLlkxbVOr6+v\no7u7+5r0kE+dOlUzIQutzVSSzHlz7yR0Oh1WV1fZyefX4g7h1KlT2Ldv35aaQwiCwMmTJ/Hoo4/i\n4MGDsNlsaGtrw3e/+92q1ltZWcHKygqmpqYQjUYxPT2Nn/3sZyX3VgRwcxAy03vig9vthkwmg9ls\nLrkGd4KzxWJBc3Oz4Ad2cTFnGOnt7eV9fN69jP/17X/D/z12TvD1uC0GiUQCvUaFJqMBOo0aapUC\nSrkcapUCRIZEKp1BiiAQT6YQiiYQjuWSvnKKjEacvsIvpctV1AO4sOAqapcAgFwmw/SYBbalVfhD\nxcQtl8kw2NWMZX8EkURxRa7XqrF7sA9n5hxIpYsviP2drTDqtTi/4C56TCGX4fffehT337IL5s62\nPAnVVoBrnWaCeJhcB64jb6sT47LZLM6cOYP9+/dv2WvwvSZTRdpsNjbLg4mx5J77VmuEX3vttbpn\niQvh6NHBB7tIAAAgAElEQVSjeO211+p+Tm9/+9vxkY98BHfddVelvyoSstfrRSaTQX9/P+/jzARn\nkiRhtVrzYg+FsLy8jHQ6jYGBgZLPO3ZhDo/+/eO8hMRg3+gANoIReNb8vI/ztSnkMhkaDVroNWrI\npbmKtdnUADJLI5ulEQiFIZXJEUumEIrGoVIqSrZLtGoV9oz04/ycE4lU8fuo06ixe7gP52Zzio1C\ntJoaYOluz9sY5GJswAySpLDgWSl6TKtW4uG3vgF/+tB90Ki3L+hmdnYWXV1d0Gq1RW0PkiTzbMP1\nDMqnKArnzp3D9PR0Hc6icnAJkYmx5J57oWW63kMC6tGuKRe33XYbzp49W1fyd7lcOHr0KC5dulTk\n7iwDN4fKotQbrlAo2OxYLpi8Y5qm2bzjciGXyxGPF9+qcxGPx6FDGv/z/W/GwnoMX3vyl7wtiLOz\nTsikUkz0d2I1FCuqVJNpIk+RkSNVit2sY+Fewd6RfoR4Wh0kScHpXcNQbye62poRjSUglUkhQW7z\nMU2Q2AhG0NXaBK1ShksOL7IcXo0nUzh2PqfY2D3ch1Mz9jxViC8YgS8YQV9nK5oa9Dg7l79xeOXq\nRuPU6ADWA2EsrQfYxxIpAt/8ya/wk5dO409//x68+7dvhUy2fZtOmyXG8QXlc9UelQYR7YTNxM2m\nbTOW6cJz51qmSw1x3QnYCrNRLBbDgw8+iC9/+cvVkHHZuO4rZCBnBuFDIBDA2toaxsbGAORPcObL\nOy4HhWtyEY/HYbfbkUwmYbVa2dZHhiTx3adfwpe+/xQCPBtiAPJMH3yVKgC0mQxo1Kkxz+MKBACZ\nVIrR3nashWL5hM3BSH8XZFJ+AwsA9LS3oL+rFeuBMLJkBiZTI2RXU+5IKguNSoFEOg33ig+hSLyo\n6h4fMIPKUphzF1fECrkMU6MWLCwuIxApvqgN9nTgE++7D3ffspf32OoFpkKu5IvFdaMxVWU6nWaD\niLhVpVCPlCAIXL58GXv3bu35CaHaCpWR5XGrae48PG7bp9SG3XZVyBRF4c4778TZs2frsl4mk8E9\n99yDu+++G3/2Z39W7TI3R8sCgGC+LTOCyWw2sxOcLRZLTVe4cDgMj8eT52MXIuKi44kn8XePP4d/\n/MkveW//gaumj54OnLps5zV9AKVdgUB5iow9w33wBcJY3gjxPj7S341UMgn3WoD3cUaxseILwqjX\nsv3vLEWCzGSg06gRjSeRSBNIEyTi6QxiiRRo5Kzpk4N9ODfnQIoojkidHrPg0488gAMTg7yvXSuq\nIWQhMPMBuRuJhdpZRvFAEARmZ2exZ8+eOpxF5Th58iQOHjxYt/W4TjzmPRCS5Ukkkm3rn4fDYbzn\nPe/Byy+/XPNaNE3jfe97H5qamvDlL3+5lqVubkKmaRrLy8uYnZ1FS0sLrFZrRROchRCPx7GwsIC9\ne/eyqoxUKsX2oMu5jVvZCOIL//xTPPHLV5DN8r+lPe3N0KkVmHXzu/WAnMwtEIkJugI3U2TIpBJM\nj1vh9K4XaZzZ1xgdQCAchXuFX5ExNWrBRijCq9iQy6SYGrPAtriCQCQOmVSSU56oFFAr5VAp5NCp\nVZDIZIBEAoqikSQIxBMphONJ3LJ7GJ94//0Y7Kktz7YQV65cQXd395bdehaG8MRiMVbxkMlk0Nvb\nW1ZFWW/Um5CFwCfLoygKBEGgu7s7L9JzK9oei4uL+MQnPoFnnnmm5rV+85vf4LbbbsPk5CTbS//c\n5z6Ht771rZUudfMQciaTKRo35HQ6oVarkUgkcOTIkbq9VjqdxtmzZ6HRaJBOpysi4kJccS7hr//x\nSfzq5EXB54z2dwMSCGqHZVIp9o9beV2BDMxtzWhrMuLMrIP38c0q6s1eQyqRYJe1G15fCH6eloxW\nrcLuoV5cWHDztmM6mo0watWY8+Q7HxVyGRoNOrzp0G782UO/g/am+uiGr5WFOBaLYX5+Hh0dHXkV\nZeHYqa3Kt9jOTbVCJJNJzM7Owmw2s+eeSCQgkUiqNvgI4eLFi/j617+OH/zgB3U8g5pxcxEyRVHs\nAE6DwQCLxQKNRoNjx45VPUurELFYDDabDRsbG9i3b1/VRFyIf/j+v+HFC04setcRTSSRJDJIFkjJ\npkYt8AXDmygyBnBuzlH0uwwGulqhUasEWx2bVdSbSekYxcaFef7Hm416WM0dOH3FDornzmCkrwuQ\nAHOuYmOKSiHHO37rAD54353oaG2pKevhWhFyPB6H0+nMa3cVWoaZfwHkaadrjfXcSg10OYjH43C5\nXJiYmMj7OTNyqtDgU9ibrySA6Ne//jWeffZZfP3rX9+KU6kWN4fKAsjF483Pz8NoNNZ97huQI2K7\n3c5WxIlEomL3nxAIgsBQVxMm+jswtxrFP/zsPxCIbkClkKNBr4VBp4FWrYJEKkF3WzMs5g6QJAUi\nQyJNZJBIpxFNpBCJJXDswhyMOg3GB8w4v+AuIlXncq61IWQ+Yezc3NQ5LnKqjzkYtGpMWrpxxb2S\n9xqJVBrHzs+hyajHhKUHZ2YdeY/7wzH4wzb0tDejudGAc3OuvPUZh+De4X4EIrG8Nkg6Q+L7LxzD\n06+cw7t/axp37LZCpcyP9twOLW0toGm6qPIVsgxzN9L4Yj25G2nlnPNOsE3zHafQyClub351dTVP\nllc4xLbwvK5X2zRwgxCyQqGo6yRkBkxFnMlkWMNIvZDJZOByubC+vg6pVIpbbrkFb5BK8d6334V/\n/vmv8OUfPM1Kygqh06iwe7gfl+yLeZuDKqUCMpkUa4EQ+tqb0NpkBA0ppDIp6CwNksoReTyRQiZD\n4sjuEdg9q0VxoEvrfiyt+2FubUSDwVBk544mUrjo8MLc3oz2pmLiDoRjOBm2obutCe1NxRkanjU/\nPGt+WLpaIZFKYF/KvzCcm89leByYGIRzeS2vTRKOJfEPT/0Gvzg5i7947+/gjsEONuPA4XCAJEnW\nQszdUONWltfDgFNuvgXXacrVD3OJinvOfAE815qQSZKsyKEnFEDE5JowY7e4kkSdTge32w23231N\nbPH1wA3RsiBJUjCY/dVXX8WRI0cqutVjnHuZTIbXMPLqq69W3QZhhk2ura2ht7cX3d3dOHHiBA4d\nOpT3hQnHEvjqj57Fd376Im82BAC0NBpgNXcImjKAq65AqUSwTaHTqDE1OoBAOAapVAKlQgG5XAqp\nRIpQOAy1Wg0ZaPhDMcSJDCKJZJEzb7McjuHezpwqRKAPvme4H8FoDIs8G4calRJ7hvtwybaIGE8b\nZPdQHz71gftx697c7L/CMKJYLIZEIpEn0fL7/ejv79/2KioSiWB5ebnucwoLA3i458xop5VKJVZX\nV6+Z5M7n87HZMPUGN9fiO9/5Dl544QX4fD709vZi165d+NKXvlT11OlHHnkEzzzzDNra2nDp0qVa\nDvPm6SEzmch8OHHiBKanp8u6rduMiBlUQ8gkScLtdmN1dRU9PT0wm80sAZ88eRJTU1O8x+hdD+D/\n++ef4Mf/fkxQkWFua4JaIYPNy6+2AOqhyJBietwKl3cdkUQCRp0Weq0GWrUSKqUCcpkMapUSGYpE\nOpVBkiCQTBGIJVIIx+NIEyT2DPchFE3wHoNMmlNkuLzr8IWK7wpMBh2G+7pw5ooDGZ6L7x3TE/jU\nI/djbIDfJs9NTltcXIRMJgNN00XRnlu18w/kbqVXV1cxMjKyJesXgtufDYVC2NjYgEqlKjnFZauw\nuroKgiAEIwfqiS984QvYs2cP3vzmN2NmZgYHDx6s+u7g5Zdfhl6vx8MPPywScrkoRchnzpzB2NhY\nyb5yuUTM4NixY0UVrRBIksTi4iKWl5dZIi788J8+fRoTExMlWy4z9kX89beexEunZwSfM24xI5VK\nw7EsbBzZTJHR3doEo06Fy65iYweQC9ifGrHgwoJbUJExPWbNxYVyiFWtVMCo10KvVaOtyQipRIp4\nMgnQNJRKJUgqi/TVzdnWRiPOzjl4K+KuVhM6W0xFbRIAkEoluP/OQ/iL996L7jbhv+Hly5fR29sL\nnU5XVE0nk8m8nX/mv3pkXASDQWxsbGBoaKjmtSoFcyEaHx/PcyIy1TQzxYV73vV043m9ubuj7u7u\nuqxXCp/+9Kdx7733VpM3wQuXy4V77rlnWwj5hughl/rQMCH1fIhGo7DZbCBJEoODg2VbqJl5caU0\npBRFYXFxEV6vF2azGUeOHBGsQsqZPD1h7cXjX/hzvHTqEh795uOY5VEiMG2JCUs3wvEklgpMHUxO\ns1atwpE9I7xqCK8vAK8v14Yg0im4VvPXSKUzePXCHGvnPn3FjgwnbpTKZnFyZgFqpQKHJ4dxyeZG\nLJlGisggFQhjLRCGfWkNaqUC4wNdmHWvIlFgkpnFMkwGHW7Z3Q/3qg9qpQIa1euVOCS5ijiRSiGe\nIpBMpRGNpxCOJ/Djfz+OZ14+hfffeyc+/M43o9EgfKvKHebJTZzjVtOMhJLJuKglKP9aWqe5PWSl\nUpk3GIB5nBuSv7KyUtcpLsz7tx0QN/V2MPhC6hkipigKVqu14skCzNghPlAUBY/HA6/Xi66uLhw+\nfHjTD3C506JDoRD0SOFLf/wgZrxBfPlHv4B3vVgGN+PwQiGX4dDkMObdXgQLbMpcNcQuay9OXbYV\nydC4A1rDMWFFRndbEzqaTUWpcykig+MX52HUa3Bg3Ipzc668VkOKyODMnBsNOg0ODQ0VtSKC0The\nvTCHzhYTGg063ooYAHYP9oIG2Gpcq1bCoNPixRMXcfKSDW+5dR/ef++dedO8N7srlMlkaGhoyDOO\ncOVp0WgUGxsbrI5Wq9XCYDBsOtXjWhIyRVElLx5CIfnlTHFhTB6l1t/OidPX67QQ4AYh5HIrZCbd\nrVoi5luTQTabhcfjwdLSEjo7OysarspU3EKIRqNYWFgATdMYHh5GQ0MDpqaAd959G7790xfx1R89\ny8ZyMsiQFE5cnIdOo8Ytu0d4YzID4RhOhBfQ2WxEZ2szr3Hk/PzrI6kcPFNLvOsBeNcDGLq6cXel\nYOMuHEvitct2tDTo0GLUY7bA/BGJJ3Hi0gI6W0zoai1uRaxsBLGyEcRgTwdUSgVm7Pkbhxdsi5BK\nJdg/boVndQNrgTASKQLMq5ydc+Jfnv4P/PlDv4MH33iYJY1qpk7zydO4fVruVA/G7MHt017rCrma\nPnE5U1yYixMgPMWl1LSQeuN6rpBviB5yqQhOp9OJbDaLSCRSMxEzuHz5Mjo7O2EymZDNZrG0tASP\nx4OOjg709fVVXAnMzc2hpaWlSFbHld0NDQ0JfsiCkRi++qPn8Mq5K4jGEqy5hNuO2CwmE4DgKCgG\n5WRkTFjMSKQIOAUC9ge6WiGXSrCwxP+41dwOtVKBGQFVyORgL+LJFBze4t9XKuSYGh3AZccSIvEk\n7/l98gP3oV0rRX9/f9U775uhMDGO6dMy0q+Ojo68xLjtwFaqHBgITXGRy+XIZDJobW1FS0tL2SOn\nqsUdd9yB48eP1y3bejt7yDc0IUciEVy8eBHZbBaTk5N1u2rOzc3BZDKBIAi43W60t7ejr6+v6g9A\n4VioRCIBm82GVCqFwcHBsnKaAWBxdQOf/eaP8NwrZ0HTuexkg059VQ2hgkwCIEvBoNdCIpUBkIDK\nZlmTSTJNIJZIoq+rHf5QpGpFhlQqwdSoBYsrG0UaZwaTg70IRWPwCIQXDXa3IkmQ8PqKp5Yw6zMV\ncSEadBqMDZhx9uqMwELstnbjv3/wd3F4T3Fi31ZidXUV4XAYDQ0NLFELVdP11gyvra0hlUqhr69v\n8yfXGQRB4NKlSzAajexEl8IpLvWctH306FGcOXOmLmv93u/9Hl566SVsbGygvb0dn/3sZ/EHf/AH\n1Sx18xAykB/BGYlEYLPZQNM0S8JWq7Uur5PNZnH+/HmEw2F0d3ejv7+/5isxk7thMplgt9sRjUYx\nODhYcnKJEEKhEP7j+Bn84N9P4zdnrwg+b8LagwxJsb3iQug0alg6myGRyZDN5qpPhVwG6dUozkQy\nCWQpyORyhBNpxJNpxBLJvKpcrVRg70g/ZuweRHmmjkglEowPdMEXivESK/P48kYQgUhxrrVKqcC+\nkX7BirijuRHm9macvmLnHSf1tjdM4S/ffx/6O7d2YgkDIekXn24ayL/9NxgMNVXTKysrIEkSPT09\nNZ1DtSic51c4xYUZOSWVSmua4kLTNI4ePVr3cPo64OYiZIIgEAqF2OB5q9WKxsZG+P1++Hy+msX4\nNE1jZWUFLpcLCoUCra2tgpNIKoXdbkcgEABJkrBYLHlDVStFNBqF0+nE7t278bN/fwWf/+efwr3K\nX4UCuajLVX8I3nX+5zD5FaVymrnGEIVcBoNWA51GBY1aCbVSCZVSDq1KhXSGvJqrTCFDUognUogn\nU4BEgv6uNsw4PIjyECtDvELErlMrMWRux4xrOU/xwWCguw0GrRoXFhaLHlPIZfj9t9yGj73nbWg2\nbm22RSWkWJi/XDjElUta5VTTS0u5cWHbITvjw5kzZ7B79+5N23nMJiI3Ka+SKS7ZbBa33347zp0T\nHqF2jXBzEfKpU6dAEAQGBwfzNiD48osrAZeIm5ubMTAwgPX1dVAUVfPtH0EQcDqdWFlZQVNTEyYn\nJ2u+qsfjcVy+fBlSqZQN4n/21fP44r/8DCsbxbf/QI6UpsesmHXxDy4FAJNBi44mA+YW1wV70EJT\nSxgIbdwxaDbqMWHtwYo/BIVMBhXHNSiRgD2nQCSGVCo3ZzCWSCGeylXlLUY9Whv1mHWv8n5QJyw9\nSGcysHmKI031GjX+6MG78F8eeBO06q2RZy0vLyObzW4647EUCnvTTBDRZtNMFhcXoVQq0dFR3yjT\nclHLPD2hnjyQf94qlQrZbBbvfOc78corr9T7FGrFzUXIsViM99YmkUhgfn6+YssoTdNYW1uDw+GA\nyWSCxWJhb7dWV1cRj8erboNwcyz6+/shlUrZSde1IBaLYW5uDuFwGNPT03kXpmSawD/++Jf4u8ef\nQzRRXIUCucGlk4N9ODvrELRr93S0oLWxQTDKsxzzyWBPB1QKueDGXUdzI7rbmgSJu6O5Eea2Jpye\ndYCmcxcUvVZ9dcq3CjqNCkadFtF4HBSVBUmRIEkKGZIEmaVh1KrhC8exHowUyf1aTQ25cVJ335rT\nPNcRW2WOKKeajkQiMBqN15SQ6500V3jeTz/9NL773e8iHo/jHe94B3bv3o23vOUtNV0An3/+eXzs\nYx8DRVH44Ac/iE9+8pPVLnVzETI3E5kLgiBw/vz5sj8MDBE7nU40NjZiYGCgyEHn8/kQCAQqtsAy\nrr2VlZU8+/T6+jrC4XDVDq5kMgmbzYZEIoH+/n54PB7ByQz+cBT/5/tP47tP/wfv7T0AtDUZYW5r\nKgoF4iJnUaaLZG4MNsugAIDxgW6Eo3F4BaaWWM3t0KiUuGTnz8iwdLdDp1Hhoq24FQFcrYiJDGxL\nxRWxXCbFaE87VgMRZGn6aotFBbVKCZVCgWajAQ+88RDefMs+3rWrgcfjgUwmQ1dXV93WLAVuVbmy\nsoJsNgu5XF7zbMBqsF1ZzDMzM3jsscfwJ3/yJ7h48SKOHj2KycnJqtaiKArDw8P45S9/CbPZjAMH\nDuBHP/oRxsfHq1nu5nHqlUIppx4XTLC93W6H0WgsmR5XyhjCB8YssrS0BLPZjMOHD+fJfspx6vEh\nnU7D4XAgFAphcHAQLS0toCgKLpdL8HeajQb8rw+/Bx+8/0343Hd+jKf+87Wi56wHwlgPhNHVYoSp\nwcBbyTLDS/eNDsAfihZNDEmmCRy/uACTQYeDu3px5oq9SJFx2emFVJLTEPMpMuxLOTVxTuqWhsOb\nr2Fm/v8uaw+SaYJ9PoMZhwcSiQTTYxYs+4J5LRuSyuKSawV6rRrjFjPOz7uQJvJbOs8fO4ddlm78\n+e+/DUf3T9a8ecsXv7mV4DryCIJAU1MTGhsb2aoyGAxiaWmJnQ3INbdstTRtqxAOh9Ha2oqjR4/i\n6NGjNa118uRJDA4Osneu7373u/Hzn/+8WkIuCzcMIQtd4RlVgBBomsbGxgbsdjsMBkNZecrlkjxX\no9zZ2Sno2quUkDOZDJxOJzY2NjAwMIDR0VH2/KVSaVmuv/6uNnzrrz6E//qOu/E/v/UEjl+YL3rO\n8kYYyxth7LL2Ip3JYGGxON/i7KwTcpkMh3YNweZZKZoYEozGcfLSgmAUZ5amceqyHSqlAocnh3g3\n7i7aFlli9a77serPJ+5Ldi7xBrDCqbhpmsbpKw4o5HIc2jWEOdcyQrHX++SxRAonL9nQampA/1Ab\nTl+25/XILzm8+MBffwsHR/vwnjdOw2purzqMaCcYQ7gqBi4YOVosFoPX62WlaYXVdDUh+dsZdxoO\nh+sWven1evM2YM1mM06cOFGXtYVwwxBypWCI2OFwQKfTYffu3dBqtWX97mbOumw2y24Etre34+DB\ngyWrq3IJmaIouN1urKysoLe3F4cPHy6quDa7ABVi0tqDL/7XB/CL35zGi+fsiCbSSKYJJJJpROMJ\npDIkLtlzhLh/3IplXxDLvnxFBklROHFpATqNCocnh3Buzlk0vJTr6FPwRHGmiQyOX1yAUa/FoV3F\nVmqGWJUKOQ5PDhVJ3QqJ94rLiwjHvZghSZy4tAC9Vo1dA11YWFpHmqNRZrKnezta0NSgx7l5V97x\nnZx14/S8Bw/81kF88HeOIplMwufzsTGX5QTl72TrtFKphFKpzDNNCYXkM9M8uEqPUtX0dmYxX88u\nPeAGIuRKqpRAIACbzQaNRoPJycmyiZiBUMuCpmmsrq7C6XSiubkZBw4cKEs7uhkhcyvt7u7uopZH\nNWDCj5aXl9HX14ePvv9d+AgN/Oj5X+Nvv/czNhtCLpNCf1XGFoom0NSgx2BPB7LZLDsVmwZAUbkN\nllWfH5buNhj1esw4PIglUnkVJ1NlD/e0IZZMY3kjv9oNxxJXrdSN6Got3tgjMiSOX1zIZWDsGioy\nfzDEa9CqcXgy93iac3GIJVK45FxGa6MB/V3tOH0lvyJeXN3A4uoGRvu7QdM0O8UEyAUnPfnicTzz\n6zP4g/t+Cx96x90waDWgKIrt1XLzHgpHMGWz2R0RLlQuSoXkM+dbTjW93TkW9SLk7u5ueDyv718s\nLS1tuWzwhiHkUpBKpaAoCuFwGDabDWq1Grt27araOlvYsqBpmp1Y0djYiOnp6YqSrYQImZmczVTa\nleRjCCGbzcLr9WJxcZENP+KS+0Nvux0PvPEwvvlvL+Brjz+HRIpAKBovksMZtBrsGuzBmSuOvEqT\ni572ZljN7Zhzr0CnUUGnVkGlVECllEMKCdRKJVoatKAhubrBKAFNA2Q2C4Kk4F71YWygG1KptCjD\ngsnAEDJ/RBMpHL+4gLYmI/o6WnCq4HFfKApfKIr+zlYYDTqcL6iIZ125Cn7fSD82QtG8WYbJNIGv\n/f/P44e/+A0++u634L1vu50374GpLpn0tHA4DKVSiWAwWHZ1WS/Us0oVmubB9KYLq2mVSgWCIBCJ\nRLb8fMPhcN3MLwcOHMDCwgKcTie6u7vx+OOP44c//GFd1hbCDUPIpSoPZsCjSqXC+Ph4Uf+sUjBt\ngWr6z3woJGSG4O12e0WVdilwZXwtLS0l2yhatQp/9tC9OLprAN/7xav4yUunQBZcMKKJJI5dmEeT\nQYshcxtmXMtFbjhmVBNftcmFWqnAnqsjqQrjQDeCUcikUoz3d0GpVCCaSOVIXSGHTCaD7CrJ3D41\ngTRBIEVkQGRIpAgSyXROq3xm1oG+zlY06LS4sODOW9+14gNWfBgbMIOiKMwX9MnPzrkgl0lxcGIQ\n9qXVvB55IBLDZ7/1JP75qf/Af3v47bj39td1tkwKnFarZS3xNpsNRqMRCoUC0WiUrS5pmmaraYPB\nsCXKh61uGzAZ0nzVtM/nQzKZzKumC+8e6pW9XM8KWS6X42tf+xruvvtuUBSFRx55pGhIa71xw8je\nstlsUcxmMBiEzWZDPB7H2NhY3gelVrz88stslu7g4GDFbQ8uaJrGsWPHcOTIEfj9fthsNhgMBlit\n1qrmBBZONPH7/VhYWEBDQwMsFkvZazKuxBQtw+e+/WM8+5vTgs+1mNuh16pxYd4t+Jy9I/3YCEax\nxBMZCuSmggz1dfEqMoCclXrI3Ir1UBTBKL+WesLag1Q6A3uB1E2jUkKvVaOzpRGSLAUK0ty4Kpk0\nRwQSADQNnUYNfyiKFJGbepJI5mzhyTQBrVqJ3UN9uLiwyJpRuJgc7MWnHnkAb9jL7wpdWFhAS0tL\nUbgV06uNRqOCOmKDwVBTxsWZM2cwOVm7UqQahEIhrK+vY3h4GECxbToWi+VlL1c6wJWLj370o/jQ\nhz6EQ4cObcWp1IKbV/YWCoVgs9kgk8kwOjqKxcXFuoVjh0IhLCwsgCAITE1N1VxtA7nqgiRJtoqv\nZIOxFMLhMBYWFqBQKDA5OVlxi0YikeRs6OYOfOfRD+M3Zy7hf/z945hxFle6Do5ELZkmeN1w55hq\nc9cg7J7VqhQZc551qJQK7B+zYNblLdI4z1xVXOwb6ceKP4TVq4qLZJpAMk3AF4xAIgH2DPVh1R8q\n2qAEcn3z6TELfIthBK5mSctlUigVciyubqCrrQnmtiaE4wlIJVK2SgeA//P9p/Hsr0/j4XtuLxon\nJbSpx+3VcsHVEbvd7qKMC4aoyyHZauM364HC6E2+uwcgp9Pn6qbj8ThvL75UNc0YYK5X3DAVMqMj\nttlskEgksFqt7B9mbm4OTU1NaG2tPkQmEolgYWEBEokEQ0NDuHjxYsXDU/nAZB0HAgEcOnSoaBx6\nNfj1r38Ng8EAkiTZ/ORq4PF4QNM0Ojo6YLfbEQqFMDQ0hBNX3Pj8P/2Yl3SB3BduanQAXl+AJcRC\naNUqTFi6cWFhUbAHLaTIYGDUazE20I0zVxwgeEwucpkUIz3tcK8FeM0pCrkMU6MWzLuXEeSxjGvV\nKsGAnU4AACAASURBVOwe6sX5eXfedG8GQhcOIJdId98dB/EX770X5vZcrOrc3Bw6OjpqIgxuxCU3\nMW6z2YC1WJdrxdraGpLJZFXZL5VW0/fddx9++MMf1vVuuE64uZx6BEHgzJkzsFgsRR94h8MBjUaD\nzs7OitdlMomZMU9Mf+r48eM4cOBA1VVHPB6HzWYDQRAYGhrC5cuXq55kzSCVSsFut2NlZQV79uyp\n6QIE5Ah5fX0dqVQKAwMD6OzsZL/QJEXh+8++jP/9rz8vCq1noFTIsW9kAFec/GlsAGDUazDc140z\nV+xFNmYGe4b6EIrxD0cFGKt1c9HUEgZ6jQqD5nbMOJaQ4WmF6DVq7Brswbk5F69lvNlogNWcU2Tw\nHeNQbycUchnvZG+VQo73/c4d+Mi73oKVpUWYzea6XHS54Ju0zcwGZAjL4/HgwIED26Z24KIeGR6F\nYEKIotEo4vE4FhcX8Zd/+ZfIZrN4z3veg6mpKezfv78uQ1WffPJJPProo7hy5QpOnjwp6ILdBDcX\nIQP5EZxcLC7mdLSV7L5ulkl86tQpTE5OVtwKYUgzGo1iaGiIDaWvZpI1A65RxGq1wu12Y+/evVVv\nBGazWSwvL7O97H379gn2LuPJFL7+xPP45pMvIMHTVwU4+cSz/JUsAEH9LwOZVIqp0QG4ln28U6mB\nza3WbU1G9Ha04PRlO+8H2aTXwtzehBmHlzdAqae9GS2NDTg7x28p3z3Uh0gskdsoLECDXot33L4P\nH/m9e9DctD3jhbhyPIfDAa1WW3ELoB7weDyQy+VVFUSVgKIoHD16FI8++iguXryItrY2fOhDH6p5\n3StXrkAqleKP/uiP8Nhjj4mEXC4IguA1RaysrJQd3pNMJmG32xGLxUpmEp87dw5DQ0Nl92UJgoDD\n4UAgEIDVai2K2Hz11VcrboEUaom7uroglUrLmmLNB27bp7m5mZ3KXE6I0nogjMf+9ed47dICUukM\nEuk0klcT2ZiqkgkFOiUQGgSgLEXG3pH+khkZu6w9SKSIIqs1g65mI5oaGwSJu6ulEXq1EvMCU01G\n+rogkUhYaRwXUqkE06MWuAXC+TubG/Hn770Xv/umYlPPVoLJkqBpGqlUKm8DMZVKVWz2qAQul6uo\nX7wV2Oos5DvuuGPLCfmG3NQrRDlW51QqBYfDgXA4DKvViomJiZJ/1M3cegxIkoTL5cLa2hoGBgYw\nMjLCuy4jfSvnlnIzLXG59mkuQqEQ5ufnodVqMTU1BbVajfX1dSST/K2GQrQ1GfHFjz2MhcUV/PW3\nnsD/PX6efUylkEOrUUEmkyIQjWP3UC+ajAaEo3Gk02k06PXslT5L08hms/itA7uw5g8hmc5NMkml\nCcRT6avDU3MZGYd29eI0jyKDsVLvH7diaW2jyGq97A9j2R8WjONcvtr3Hh3oRpaiML+Y/zhzsRjr\n70I0nsISZ2Mwm6XxGtcK7ljKy3he8YfwF1/+Hv7xpy/ikx+4H288WF3wTbWQSCTQaDTQaDSCG2p8\nZo9a5HjM6Krtwg4Lpq8INwUhKxQKQfJkMon9fj8sFgvGxsbK+oNuFjDErV57enpw5MiRkhVROYRc\nrpa4kmyMWCzGDlAdGxvL628yKotykc1moZaQ+KO37Md9t+3BN37877jk8CKdIZHOkEXTrycsZoRj\nSVx2FWdkALlNualRC2yeFXbTTSGXQXvVYLIWCGGkrwvtzY0IRmKQX81qkEgkoJHTifd1tmJswIz1\nQBjpDIlUOoNYPAGCpPLCh1Z8QSwX5EXPOhlzyAB8wUiRXO+KaxlSqQTjfR3w+kIIczI4Cq3ghe2a\nOfcyPvDo13Fk9zA+9YH7sXdk62bdlQO5XI7GxsY8DS/XOh0KheDxeKqS422XUy+VSlXlAwCAN73p\nTVhdLd6k/pu/+Ru8/e1vr/XQysYNRchCBMIMWeSiMJN4eHi4oiurUNW9WfUqhM1IlNESGwwGtoIV\nQjkVMtPLjsViGBoa4p3bV26lzW11tLS0sBrQt95+GM+9chZf/O7P4Fou7qvOOHJTLCb6u7AejMIX\nzs9PJqksTs7YoFWrcHhyGOfnXUimCYRjibwp2zOOpZIbawB/H1sqlUCrVsGzugG1Solb944glSZB\nZSnIr5pOJBKApgFzezOGejsQjiVAZMic+SSTQTqdgccXApXN4uCEFZfsnrzJKowVvNVkQLNBi9nF\n/DbKsQvzuPdPv4C3vWEKn3j/fejvqv9tfbXhPkLWaa4cb3FxkQ3JF5LjbdfE6VAoVLWi6MUXX6zz\n0VSHG4qQhcAlT5Ik4Xa7sbq6it7e3k0rVyEUtiy4k0VaW1s3DRTiW4+PkBktsVwuL1tLXIpISZKE\n0+mEz+eDxWLB+Ph41Ul5zPHNz89DrVbnXSiY9+ae26Zx95G9+P5zL+OrP3oO/gLSpWkaM65lKBVy\nTFq6YfeuI5HOv3gmUmkcvziPZqMBuwZ7ixLZgNczMoQUGVyrtUmnxqxnFdksjVgiN3UEyOVY6LVq\n7LL2FGVgMGCkcBcW3EUjrS7aPGg06HBgfBCr/hBkMmnOeJI7URBEGnusXaCzNK5eEnI5IFkal51L\n+P3PfBVvvXUf/vDBu9Bqqo5Y+FBvlx431pP7Gowcz+/3w+12s3K8RCKBYDDI6o+3qqVwvQcLATfY\nph5JkrykRlEUTp48ic7OTni9XpjNZpjN5pqu2oxyw2w2sznKjY2NeZNFKsHMzAy6u7vZD1Q8HsfC\nwkJVWuK5uTm0tLSwCg4g94VhMpm54filEAqF4PV6ee2iiUQCCwsLyGQyvMdHkiRIksx7jWgiib9/\n8gV852e/4tX1Ajlt8UhfF87OOQUD9LtbTWhuNPDOyAOuKjLGBuDyCisyBrraoNeqBcPt20wN6Ots\nKwofYtBs1MNq7hCUwnW1mtDRbBKcrDI20I1EMgX3arFrUadW4ZG334k/fuebodNU7tQsBEEQmJmZ\nwb599QvbLweMHO/ChQtobm5GMpkskuOVSserFCdOnMATTzyBb33rW3U4+tfx05/+FB/96Efh8/nQ\n2NiIvXv34oUXXqh0mZtPZcFHyAwRzc3NYWhoCL29vXW5fVpeXkYwGEQsFoNOp4PVaq26fwUAs7Oz\naG1thU6nK1J5VIqFhQUYjUa0tbWxCXQOhwPt7e3o7+8v+8MfiUTgdrvzJi5kMhnY7XYEg0EMDQ2h\npaWF93f5CJnBqj+EL/3r03jyxWNsYlwhhNLeuOhrN4GmgcV1/lmB5Soy+MLtGQiFDzHYTAqXG1el\nwIyjWNHBGGgKM5wZNOo1eNedU3jgjgMwGhvYAPlKq91UKoX5+Xns3r27ot+rF06dOoWpqSn2uCmK\nYjXETOujHnK8F154ASdPnsQXv/jFrTqVWnBzEzKjpXW73Whvb8fa2hpuvfXWurxOOBzGpUuXQNM0\n9u7dWxf79OzsLFtB8MniKoHdbodOp4NCoWAzLKxWa8WVO6Nf3b17N7LZLBYXF+H1etHX14fu7u6S\nx0dRFDKZTEnyuOzw4DNf/R5Oz/PLzwDAau6AWqUoSnvjYtLag/VgGGsB/mq40aDDcF8nzlxx8GZk\nbEaMQG5kFUlSWPDwb0CWksIBOY2yzx/ECs8xKuRyTI0OYNblzeuNM+jrbMF/ufd2HBg2s2FEldin\nE4kEHA5H1YN+a0U5LkFGjsek4xU68hiVRyk53hNPPIG1tTV8+tOf3qpTqQU3HyEzJMD0cltaWtDf\n3w+lUlmT8YIBo0jIZrNoa2tDPB7H6Ch/kEwlx7y4uMj2nsfHx2vu983OzsLv90On02F4eLjqXIx4\nPI75+Xl0dHTA4XCgo6MD/f39m95h0DSNTCYDgiAgk8kgkUh4z4miKJw9exaEwoDP/dNPipLYuJgc\n7EUsmYLTy68N5ioyAhH+ydltjQa0NTXgkoOfNFlnYUG4PRf7RgewHgjDu16cgQEAe4f7EYjEikZa\nAWAVHZ7VDawFijXKBq0aEyX619NjFnzqkQewf8yCRCLBElc0GgVJklCpVCxx6fV6aDQaSCQSdvNt\nK0cPlUIt8/S4cjzGlSckx/v2t78NtVqNP/7jP67zGdQFNx8hBwIBXLhwgbeXe+zYMRw6dKgqsuO6\n9oaGhmAymRAOh+HxeKquOrgVPDP0UqFQ1GQvTSaTWFhYQDgcRnt7O5uuVS1WVlZw+fJldP6/9r48\nuq3q3nprtGXZlud4kDxbtjN6iCF5JVAKJa+BfowtUGhDIY9v8UpIH6UPKAvKPBRoSKEMKSS0lJYy\ntRTol5YxQIrjOIlDnHiU53mWZMka7vD9IZ+bK/leW5IlS8R3r5UFSWzpWJH2Pfd39pCV5dcOm7yX\naJoGTdNgGIaLKeV/DSFphmHQ0NCA9evXg2VZvPtZPZ5/8wNMWqfhcLnhnA0EIjNauVyGqrJCwf49\nAq0mBmuKxfMnACAnTQeVQoGuYWFSTdBqsLJQj6PN3uH3BCqlApVlBWjrHhTMwFDIPeFEpv5hjE/N\nbd6OUatQYczHyU5vjTJBRnIi8rLTcfhkh+D8+tsb1uKO6y5FSe4p5xvfPk2IemZmBgqFAjExMXA6\nnZyRaalDhkJdcMqX45Ff9957LwYGBpCfn48rr7wSFRUVWLt27aIPEH/+85/j3XffhVqtRlFREfbu\n3RvsweHyI+SZmRm4XC5BSdihQ4ewbt26gOzETqcTJpMJZrOZKxEl/8Ak46KioiKgNfpqiQsKCqBS\nqdDX1weappGXlxfQ4wGnXICTk5MoLi6Gw+EAwzBBPRbg+dlaW1s9rSA07VeUITtr6OC3YpD/EsUH\nTdMcQbMsC6vVCpPJhHXr1kEul0Mul8PlpjyKjNf+HxeKr1IqoYlRzTZCKxGjViErPRkzDidomoVC\nIYdcLoN89vlYFlCpFGBZYGbGCYphQFEUXBQNl5vCtM0OmgUKczIwbraiR+BgDQDSkhJgWJGKhtau\nOVnPwMLkr4lRY11JHo63C8d1LhSOlJ+VDl18HI4J3Dko5HJ87/yNuPXai5CZJm7FpigKg4ODGB0d\nhUaj4UYecXFxXrvpxeZtz4elapy+/fbbodfrERMTg/b2djz33HOLfsx//etf+Na3vgWlUonbb78d\nAPDYY48F81DLj5CFMpEJjh49itLSUr9u391uNzo6OjizyIoVK+ZcaR0OB06cOIHq6mq/18fXEvtm\nHQdi7ybgd+zl5+cjO9szxxwYGIDT6URBQWBmA6fTifb2dkxPT8NoNEKr1eLYsWPzfpj4RAx4SHih\nXYnD4eAiTIuKiqDVarndNIFl2o7n3voAv39vP5wCgT8AkJTgUWQcaer06t/jIzczDcmJWhwTyWkm\no44JyzRmnC6oVUqoFArIZIB8NiNZDkCtVoJmAIVC7tlhyjySQIqmoVapoFLIMTA6CZebgtPt5uI+\nWRZITtTCmJuN+pPtgoqMhcKR5ptfx8aocMPF5+Gm721Golb4UHl8fBxTU1OcBZ5hGK7dg+ym+Ylx\nhKjJyGMxYBgGR44cCdZuHBB27NiBbdu2LXo0KYa//vWvePPNN/Hqq68G8+3Lzzo935tHpVKJkjUB\nX6Ocl5eHkpIS0RGHv83TgH9a4kDcdSzLor+/nxt3+JpPFApFQNZpmqY5ezdfm0xRlOjjBEPExEZO\nLnT8Ow4CQszJugTc+eNL8cMtZ2Pnn97D3z49BMaHzKasdhxsbEdWWjKy0pNwpGmu0oF05ImRGjGf\neBQZBWhs7xZVZJTnZ8NsmxGUqgEexUVGis4rIClGrQLLsugeHEVWqg5FuTkYnZiCUqmEUuHJUpbL\nPVrls6tWgmEYuNzUbGehh/DdFA0GLDZVlqF3aALjFiscDhfcNA2H043fvr4Pf9r3ObZftQU/uvAc\nqFXeH2tfHTK/lDUzMxPA3JEHiczkx1wSlUcgI4+lMoUA4dch79mzB1deeWXYHh84zQh5PsxHoHyN\nrl6v98ss4g+BBqIl9ufxiCPOZDIhJSVF1HxCOgQXAp/YSXmq7wfX9w6KEDEZO/hDxKQbsKfHEz9Z\nU1Mj+vqSPycf4kJDFnb9/Ab838svwEMvvY3Pjpyc8z2DY5MYHJtEsSETapVS0K3X1On5s8qyAvQN\njWJ0yjsc35OR0YrkBC3OWG3AUYFdd1PXAHcwJ6TIIJVV+ZkpkMnk6Bwcg9Plnt3hew4J+0anPLNf\nGS16iCkmxTP1Ds1mOBegrWcQFtsMNDHq2VGOCn98fz/e3X8IP/4/38LF36zxGhkt9H6WyWRcAw5f\nyiiUcUFGHnyiFht5LDUh+7ax+AN/bNMPPfQQlEolrrnmmkWvcz6cViMLlmXhcgkf5JhMJsTHx3tZ\nQPkHa5mZmcjLywtIoC6m3ODbkv3VEk9NTWFgYED0JJyE/2g0GhQXF8+reR4fH8fo6KioAoR0Aba3\ntyMlJQWFhYWCxE6qpf7jP/7D68DOXyImazGZTEhOTkZ+fv6iK4Q+O3ISD730FhpFDB2AR5Fhmbaj\nW0DpAHjGFJVlBWjvGRI8lAMWNnZ4pGr5aO4aEJSqAcDK/CyMmacxMjn3YA+YP65zISneQuH5q4tz\nceePL8WmynIMDg6CoqiQlX/yRx5kR+0bkp+QkACNRgO73Y6urq6wd9EBwHnnnYf9+/cvyg8ghJdf\nfhkvvPACPvroo8U0+Sy/GTIgnonc3d0NhUIBvV7vdbCWmpoqSkgLwZeQfXOJA9ESW61WdHZ2zhHv\nE+kZwzAwGo1+hZvPR+4WiwUtLS2IiYlBSUnJgm9eEgsqdmAnBrJuhULh1/MEApZl8fbHB/HYy38V\nlZ/J5TJUlhagZ3BM1K2n1cRgTVEuGlqFg+kBzLvrBhaWqi0kx5PLZFhdpMfA2BTGBBQZC2mUFwrP\nP7tqJa6/8D9QmpsV1gp7shnimz3sdjt3Idfr9RxZh2vHvGnTppBHb+7btw+33nor9u/fv9jCh+VJ\nyGKZyP39/XC5XIiPj4fJZAq48FMIhJDFcokDgd1uR0tLC2dvJQdsVqsVRqNRMPxHDEIOOyKJczqd\nKC0t9cuKzbIsPv/8c66pW6lULvhmJ4oPEsAfzpme0+XGnnc+xtOv/UN0l+qPWy8tKQH52Rk40iQs\nMwPm380CvPD7JpOgIiMuNgYFK5LROTwxJwMD8GigS7LT0DE0jhnn3IvDQsQ/n2NQJgP+c8Na3H3j\nlVyd1FJhdHQUIyMj0Ol0HFETHTE5PJxv5OEvwpWFXFxcDKfTyd3lbtiwAc8//3wwDyURMh8dHR3o\n6elBSkrKoluiCQ4cOACDwYDe3l5kZ2cvypbtdDrR2NiIdevWeYX/CCk8FsL09DQnJyOKkYmJiTnS\nPTHwD+yGh4c5izjDMNwHiPwidxbEyUcUH5mZmUuWSztpseE3f34fv3/3E9F+Pl18HHJSE9DSOyJa\nFZWbmYakhDjRjIyFwucBT0ZGQlwsvhIZqaTq4lFkyMThk8I72qT4OBQbMtHQ2iXoKkxLikdeZjqO\ntHQKEv98jsEYlRI/vPAc3HL1FiQlBFZ4GyxGR0cxPT3tpfgRa9lWq9VeUrxAgogIITc0NITrR1ks\nlichu91uL2UAKSelKAqxsbFYt27dop+DjDyOHz8Og8GAoqKiRc9GXS4XvvzySygUCuTm5voV/iMG\nu92O5uZmpKSk+G11BhY+sCOJXlarFRaLhXOIKRQKzMzMIC0tDYWFhSGf4fmLvuFxPPbyX/Fh3XG4\nXG443e45pDVfMSlBeUHOrCJDuMQ1Vq3COmM+Tnb0wsrLQOZDLPyeYKEMjKy0JOSki7erZKYkIF4T\ni/Z+4R37fI7BRK0GN31vM264+FuIjQmf/hgAhoaG4HK5Fuy2IyMPvhTPbrdzihA+UQttehwOBy68\n8ELU1dWF60dZLJY3IfMVDkS+JjSjDRR8LbHNZsOaNWsWRUCE3E0mE1wuFzZt2rSo5CuWZdHX14eW\nlhYUFBT4fVDJsmzAB3YkepPEMZJdj8vlgkajQUJCAhITPaE4MTExS7ZjPt7WjQd+9wYOHGuBUiFH\njFqFWLUaKpUCaqUSKqUC8XGxSEqIx5TVBrlc5pGfyWSQy2WQQQZWBsSqVHBTlEd2xjKgaZYzy1AM\nC5VSgYS4WHQPjsFNeUL4XW4KDpdHf0wO5oTC7wkWysAo0mdCE6MSrZsy5mbC5abQNTiXeBVyz+Gl\nqXcQk9a5I52stGTceu1F+N75wUXQ+oO+Pk/mdbDza34vICFqX+t0bGwsLBYLfvKTn+CDDz4I8U8Q\nMixPQiY7YpvN5qVwmJmZQVNTE6qqqoJ+3NbWViiVSs6CeuzYMRQVFQUdLkTInYT/HD58eFGi9snJ\nSa6GyWazYcOGDQt+T7DGDtKYbTQa5/z8JCiG7KItFgucTidiYmI4gk5MTAxLuSbJpe7u7kbvlAO7\n3/mMk7wJYZ0xH1PWaXQLEBrgsUlXlReipWuAcw76IistCZmpSTja0uX1fTEqFdQqJdQqJYr0KzA4\nPAZlTAxUCgUUChkUcoXHYSiXQa1SQilXzI5cWLAsuDsVhmWRqNVg0mKDy+2Gm2LgpmmP+9BNw01R\nKMnNQt/IOAZG5xK/WqmA0ZCBjoFx2AUUGca8bNxx3SU4/8zQp8F1d3cjNjbWS920WLAs62Vsefvt\nt/Hqq6/C7XbjBz/4ASoqKrB58+aAzl3EcPfdd+Odd96BXC5HRkYGXn75ZS7qIEAsT0Lu6OiASqVC\nenq614edoigcOXIEZ5xxRkCP57vT1ul03N/5Zhj7C6vV6qVAIEaRYAOQiKKBZVkuTKiurm5eQl6s\nsaOoqEg0elPs+ZxOp9e4Y2ZmhpsbEpJeTIA52bETJ6RKpQLDMHjjwy/x+O/fwaDILpVkT3T0Dwsq\nHQDMBtfnzh6qCSsySgyZUItEbQJAXIwKa0ry0NDaJXgw58lx9qxDKAOD7Lj7hscFw4nIxaN3eBwz\nDidiVCqolAqwDA21Wg2lQo6keA0sVjsYloVS6XEdqhRKKFVK5Gam44cXnoPKstDVSXV0dCAxMTGg\n90owqK+vx+7du3HdddehoaEBl19+OYqLixf9uBaLhTsA/81vfoOTJ0+G9VDvtDOG5ObmCpoi/C0l\nJSBaYqIWENISB+LWAzy79Pb2dszMzMBoNC5ageByudDe3g6LxeKlxCBkK4RgiDgQY4cY+MYDvnyI\nSKUsFgtGR0dht9uhUCi8xh0LZQATRYrT6UR5ebnXjl0ul+PKC76B/3NODV7620f47V/+Hyw+gT40\n43HraWLU2LDGiOPt3bD5KDKm7Q7UHm9FRrIOeSXpwq0ls/PitSV5sNpm0DngnUxnd7pxsLEd6cmJ\nyCtKw+HmDq8ZN80wOMRfR1u3VwYGy7I43NQBtUqJDWtKcLKjz+tncVM0Dh5vWzBjQ8zZWHfChDc/\nqsXZ60pw8/cvwKri/KDyl/lYyvqm7OxsbN68GZs3bw7Z4/LVSDabLexjt9Nuh0zP3soJwZ8dKF9L\nvJDKob29HQkJCQvejvmrdPB3h0yszkNDQygsLBRUNPg+VjAOO8AzViEGklAYO/yB2+2G1Wrlfk1P\nT0Mul3vtpAnp9vT0cK+D712RECYs03jq1ffwyvv7BZPcACBFF48SQ5ZgozVBfnY6dFrh4B/glA66\no29IcH4LeMKDErVx+Kpd+DE8GRhZOCyS48z1BLYIp9IRjXK9wMUDmA3PVwtnTSvkcnxnwypc9o01\n0MVrvAwfgTR8NDU1IScnJ+iuO3/x1ltvobe3F3fffXfIH/uuu+7CH/7wB+h0OnzyySfB6pGX58gi\nWEIORkvc1dUFlUolemDh+5gLKR2ICUPsa8hOtaurCzk5OcjNzRVdI/9nDebAjmQ/h8PYEQxomuYI\n2mKxYGpqCg6HA1qtFllZWdDpdEhISPB7N9Y9OIpH9ryN9z4/LNobqF+RivTkRBydR5GxqtAAN0Wh\ntUc4uD5GrUJVWQEa23tEFRkrC/VwUzTXC+iLnPQUpCcnoEEkICkzNQn6FamiGugVyQnITEsRvXis\nKc6FbcaBDoGsaa0mBv916Xn4wQUbwVBu7gJJGj74WmKheNbGxkYUFhaGRGY6H/bs2QO5XI7t27cH\n/L3+Nk4/8sgjcDgcuO+++4JZ4vIk5PkS34QIzzeXOBAtsVhkJp84s7KykJeX59dj1tbWoqamRvBr\nx8bG0NbWhuTkZL9kdr4OO8C/8cRSGjuCgc1mQ0tLC9RqNQoLC0FRFDeTtlqtYBgGWq2WG3cs1KZx\ntLkTD730Jr78qlX0a0rzcyAH0CSihODUFGNTGBgVdg4mJWhRlp+Dw00mwa5AmUyGyrICDM3zGCWG\nTCiVCjR1Cq+jICcD8RrxnsDSfM9hVEvXwJy/I1nTYuH56cmJ+OkPLsTV/3kWlAoFWJb10hITdQ1f\nS5yQkIDW1laUl5cH1TMZCHbu3InCwkJce+21YXuOnp4ebNmyBY2NjcF8u0TIvjh48CCqq6uhVCpF\nc4kDgW9kJj8jIjk5GYWFhQE5kIQym61WK1paWqBSqVBSUuLXToNlWRw4cADFxcV+S874xo6CgoKg\nDCnhBBn7WCyWeS8UfK002U3TNM05w8jIw/ff5V+1x/DIS2+J7nQBoKK0AONTVvQOi2dklOVmond0\nStQ5mJ2ejKy0ZNGuQH9UHR5liG1OszbBykI9XC4K7X3CGuj5NMpqlceq7TufJijMWYH/3Xoxtpw1\nV63ka5+2Wq2YmJiAVqvlXnsy+gi1zO7ee+/Fueeei4suuiikj9vW1oaSkhIAwNNPP439+/fjzTff\nDOahJEL2xZEjR1BeXs41gMTHx8/JJQ4Eo6OjmJychNFo5E74/c2IEMLRo0dRVlYGjUbDZQY7HA4Y\njUYvdYcY+Ad2ExMTmJiY4HYuhJASExO9CIllWYyMjKCzsxMrVqwIWQlsqEAS6Xp7e5GXl4esrKyA\nLxREJsVXeAhppZVKFf7yrwN48pW/C+4SgdlsivJCmPqElRCAn4qM3CyoVUrRrkDyGA0tnYI5Pl4W\nsQAAIABJREFUG0QZ0jkwgtHJuVkdMpkMlaX5GBibxJBAONFCrSYLtaZUlRXgzusvw5mrSwTXT3Do\n0CFUVFR4XSRtNs+FhhA1IenFnE/ceuut2Lp1K84666ygH0MIl19+OVpaWiCXy5GXl4fnn38+WE31\n8iTk+RLfDh8+zHWP8eVmwWJychI9PT3cjNbf8B8xHDt2DHl5eRgZGcHY2BiKi4v9Oqha6MCO3F7y\ndcEulwtKpZKbwwrpiSONyclJbkxTUFAQkqp4gvm00qoYDf5x8ATe+vgQXBQFpULhyS4m/5XLoZiV\nkI1PTkGuUEETG+NpPZHJAJkMMnh2uwqFHC4XNfvBmdUX83TGifFxsFhtcFE0GIYFzTCgGRo07XlP\naeNikRinQVvPANw0C4qm4HZTXMNIrFqFyrICHG/rwfTM3Bm1UiHH6kI9OgfHBHftC7WarJjN6KgX\nmU+ff+Za3PHjS2DMFdbmirWF8O9kiPGDoiiueZrsqNVqtV8X4BtuuAF33323V35LlEEiZAKiJZ6a\nmkJxcfGieusInE4nmpqaMD4+joqKCr8iNucDwzCoq6uDy+VCQUEBcnJy/LqtC+bAjh80lJGRAZfL\nxZF0bGwst4tOTEwM++xPCKS2nmGYkFw4A4HT6eRIundgGH/Y9yU+PNIiqrZISYxHcW7WbDaF8NcU\nZGcgIV6Dr0QO5eRyGarLi9A9MCqakZGXlY7kRC0aeOYThVwOlVIBlVKJ+LgYFBuy0D04CoZlwdAU\nFHI5NLGxUCoVUCoUSNHFw2afAc14AodksxcOmQzcxcblnr14zH7SWQAsy0CricWM0wWni+LuwiiG\nAUMzYAGcf8YaXH/JechM9R4lBVLfRDYOfFee0+mESqXymksLadWvuOIKvPTSSyH5bIcJy5OQgVMR\nnPzEtJKSEkxMTECn0yEjIyPoxybmiOHhYej1eoyPjwft/gNOjQxMJhNkMhm3K/bn+0Jt7ODvGgkp\nOZ1OL5ImVtVwgMj5RkdHOXlgNKClqw8Pv/QWPqwTP8wxZKYhTRfv5dbzxaoiA1xuN9p6xDMyKv1Q\nZFAUg9aeuQdzALAiJRG6uFi09gm3c2ck65CfnS4qhctOT8aKlCTRjI2VhXq43JRgRodQnVQo+vTI\nXJqfcUGaTNRqNXp7e/Hggw/iww8/DMvF+8knn8Rtt92G0dHRxbwnly8h22w2dHR0zNESd3V1Qa1W\nB2V9ZBgGfX196OnpgcFggMFgAEVRC3bOzYepqSm0tLRAq9WiuLgYPT09SE5OnpeQgzV2kDmswWAI\nKB5U7NY+NjbWaya9mKwKclHq6OhAdnY2DAZD2LIVgsX4+Dje+/gL/PmTBhw3iVuxC7LSIJPL0CES\n+iOTeRLjBsYmg1ZkAEBVeSGGx6dE86AXmlHnZ6dDF6/FMV7dFB/G3CwolQrBHGiZTIYKYx6Gx82C\nGR3JiVpsv2oLfrjlbHx1rCEsBaekyaSvrw+PPvooamtrYTAYUF5ejiuvvDJkh3u9vb3Ytm0bmpub\ncfjwYYmQAwXLsqitrUVGRsYc4gmm2ZmvxkhPT/eaZdI0jbq6OmzcuDGgNdrtdrS2ts6ZO5tMJmi1\nWq7nzHcdgRIx4G3sCNUcltigyU7aN6siEJImNnLShBLO9uNgQMYnxJau0Wiw799H8fBLb8MkomIA\ngLK8TEyYpzHiUxVFoFYqUVlWMNs4It5aspAio7KsAM2dfbDYhHfUa42zrkEBjTGw8I57nTEflmn7\nHNch9/ylBWjtGRRUhBhWpOLyTWvxPz/+flgVOyR6s66uDs3NzZDJZCGbJV9xxRW4++67cfHFF6O+\nvl4i5GAglok8NDQEm83Gte8uhImJCbS1tSE+Ph7FxcVz5qn8iiN/12UymWA2mwXt2EJGk2AddsTY\noVQqF6x8CgV8SdpqtcLhcIgGCpHXwmazLdg3GAkQGeDQ0JDgvxVF03j1H59j56vvCqocgFOKjLbu\nAVG3XrwmBqtmW0uCVWTExaqxriQfR0UUGf7MqOfbccvlMqwvL0KXyPd7rNp5ooqQlYV63HHdJTi7\nauXs44X27idcWcjvvPMOPv74Y+zatQv5+fkSIQcL30xkgoW65gjIrk0ul6OkpGRe9YE/dmeaptHd\n3c1pfMWkW729vWBZlsuODebAjhDd9PR0VBg7HA4HN+qwWCxwOBycNDErKwsGgyEkdfOhBEnhW7Fi\nBfLy8uYlENuMA8+98U+88NYHsAuoFABAGxuDNSV5ONbaJZgtAQDJ8RpkpenQ1DUk+iETy8ggSEtK\nRJE+A4dOCM+HY9QqVJUWoNEkPKMmGui27gHBuqlYtQqV83x/WlIiCvUrUH+iXfD5z6oow+1bL8bK\nQk+Nmny2cRtYHEm73W5ccMEFqK+vD/h753PpPfzww/jXv/4FnU4nEfJiIEbIZrMZvb29WL16teD3\nEe1vIOE/8xEyiYLs7Oz0ywU4MDAAp9OJ/Pz8gMcTNE2jt7c3ao0dwCmiS05ORlJSEneaTlLf+AeH\nkSBpofGEvxiZMOOJV/6O1/Z9Iaq2WIiwACB3RSo0ahVaRILt5TIZjIYMjFnsosl0eVnpSE7QokFk\nPqyLj0NZQQ6ONHUIzqi1mlisLfbs2oUuILr4OJQX6HGkycTJ7/hYkZyAVF0CTgo4AmUyGS4+pwY/\n++F3kZOePOdOlrzXZTKZ3yQ9NjaGG264AR9//LFfX+8Pjh8/jvPOO48zYvX19SE7Oxt1dXWCI0U/\nsHwJmaIowcQ3MrutqKjw+nPiAhsfH/db+0sglj8xPj6O1tZWJCUloaioyK/ZKKlLKiws9Hpjzgcy\n4+7s7OR2nNFk7ABOve7kjkOI6PhyM4vF4kXSZNwRLpLmjycWq+5o7xnEw3vexj+/FL99XogwAWB1\nkQEOl3jjSIxKiVVFerR2DwnqjwFgZYEeFEOjtVvYfZiVloyc9GTRVpK0pAQU6lfg8MkOwYtMVloy\ncjJSUH/SJPj9Zfme0ZtYndSPvvtN3HL1hUiKj/MazfE3U2QnPR9Jm0wm3H///Xj77bcF1xEKSDvk\nRUCMkF0ul5cqgmEYdHd3+x3+IwTf/Al+1jHJJl4I5E1IiMt3xygW5k7cgcRxGG0HYhRFobOzExMT\nEygpKQk4MJzoowlR2+12qFQqr530YvKTgcDGE4GgvXcI5mnbKbMHPfuL8fxiGAZymQwzLs/dHEWf\n+nNq9ms9XwN09XoKelNSUyFXKOF0uTAzM4MZhxMM5YbD4YSLYaFUKKFQKiFXKCCXK7jHS0/WoW9k\nnHt+3+fLSNHBbLVjymoDPft3DHNqrRnJOigVCpj6h079DLP/pWgG+vRkMDSFjsFxwdeissxjORer\nk/rJld/BDZecBw2vTopPzmJxsoSoGxoa8PLLL2Pv3r0h+bcTwlIR8mmXhwyIV9ST/GLf8J8NGzYE\nvatUKBSgaRputxvt7e2w2+1+jzt8D+w0Gg23e+eTEcnMIAdksbGxGBsbA8Mwc/J/owH81g6DwYAz\nzjgjKNJUq9VIS0vz+hCQ18VqtWJ4eNiLpAMJueePJ9atWxfyQ89iQ1C3tRxYlsXQ0BC6urpwQc35\nC5bG8rW6FovFK1eavC6LzTb2hdPpRHNzM+RyOUpLS6FWq3mETYPmnIceAwnFEfmsE5H3deZpuxch\nk3XyP5e+JE3+/x//+Af6+4UDl0KFrq6usD4+wWm5QxaL4CS19iqVKqjwHyEcPnwYGo0GU1NTKCoq\nQkZGhl/kE8yBnc1mQ3t7O6amphAbGwuGYaLCWceHUGtHuMEPuSdkRNxd5HUhJB3K8US4YLPZ0Nzc\njLi4OBQXFwf9GlIU5ZXfQXKl4+PjuQuYWGnofOAXFpSUlETsNRwZGcHPfvYzyOVy3H///SgvL4/I\nOvzE8h1ZCBEyIQqLxYKNGzcuOp+VGEVaW1uRm5uL4uJiv63OgR7YkYhQYuwgoxUxPbBGo/Ei6aUY\nZTidTrS1tYn27C013G73nHEHeb2IJjs+Pj6qDj4ZhuHKEUpLS8OikOHnSpNfAOZElorp1e12O5qa\nmjgzUyjzRfwFy7J466238Pjjj+O+++7DpZdeGlX/jiJYvoTMT3wjc1mKomA0GtHY2LioIlGWZTE6\nOor29nakp6fD6XQiOzt7wflosMaOsbExmEwmpKamIj8/f8EPgK/9mWRUxMXFeZF0qHauZA4/PDzs\nd2vHUoOMJ2iaRmZmJncR862LIrf1kVj/xMQEWltbkZmZOW/xQDjAMAyneCEXMJIrzR93DA4OYnh4\nOGwXC38wPDyMW2+9FVqtFk899VRU3uGIYHkTMrm9N5vNMBqNnLB/vhD4hWA2m9HS0sLdSsbGxqK1\ntXVeu3OwRDw9PY3W1laoVKpFGzv4aW/kF0VRXiS9UJC70GOSi0WoD8RCBX/GE6Quij/u8CXpuLi4\nsP1sLpcLra2tcLvdXPRqNIAcMlssFk6/r1AooNPpuHYWsZaQcK3nzTffxJNPPokHHngAF198cdRd\n+BfA8iVku92O2tpawb65w4cPY/Xq1QG9kcSszoCnVVej0SArK8vre/iHDsEaO/zNQQ4GJCOYT9I0\nTXO3rqcygufuyMnFQq1WcxemaMNi1BP8Tj+LxQKbzQa5XO51cLjYAzJ+vkhRUVFU3lnQNI2Ojg5M\nTU2hvLwcWq2Wy5Xmx7iScwzy2iwm10QIQ0ND+J//+R8kJibiqaeeWnSyYoSwfAmZ3LYLfWCOHTuG\noqIiv2acpM6IhNALvRG6u7shl8thMBi8nj/QAzvSvzc8PIyCggK/DwdDCZZlYbPZvGavNE1zh0Bx\ncXEYHR2F1WoNSWt2OLAYc8d84B+Q8Unad9zhD0lPT0+jubmZO/iMxBx2IUxOTqKlpYULe5qv55Hv\nxuRb5vnh/8FoyBmGweuvv46dO3fioYcewne/+92ou2gFgOVNyGIh9SdOnEBOTs68ZMIvJ83Pz0d2\ndrboG6G/vx9utxv5+flBJ7HxjR1LPT9cCGS+2NPTg9HRUahUKiiVSi8iCuakPhzrXGr1BCFpQkZi\n7djk35PsOCcnJ1FWVhZ1+R2A5+6ANNWUl5cHfUHjG334GnI+Sc83rx8cHMRPf/pTpKSkYOfOnQFr\n2KMQy5eQgVOZyL6Yb+bLtzr7W046PDwMi8WCwsLCgOfEU1NTaGtrQ0JCQkgkeOGAUGsHIWmyW5ye\nngbLsl5ElJCQsGQXFnIgFg2zbKJi4L82crkcSqUS09PTyMzMRFFRUcQvYEIgudzBVmUtBCGtNH8U\nZLVakZubi7/+9a94+umn8fDDD+PCCy/8Ou+K+Vi+xhAAnCzMF8Qc4gsyc9TpdKipqfGLHFmWhUKh\nwOjoKOes80dKRRo7aJqOSmMHcGqNDMNg9erVXjJB8iHi7/BomuZIuq+vD9PT05DJZNy4Q6fThdyY\nEG5zRzBQKBRISkri7sCIecLtdkOv12NmZoYLweFfvJbyAuYLp9OJlpYWT1ZzdXXYNgZqtRqpqale\noz/+XcbOnTvx2WefwW63Y8uWLWE3e0QjTltCFoMvIU9PT3MlhmvWrPGrcYB/YJeYmAij0QiLxYLu\n7m7YbDYoFAoviRkxJfCtxMXFxVF5OBFsawf/BJ7/WGQ31N3dze0W+YeGwZD018HcwbIs+vr60N/f\nzx3a8cHXA5MLGIA5445w7qT5jkp/m2pCDaVSCZ1Oh/feew/Hjh3Dc889h3POOQeNjY1obW09XXbH\nfuO0HVmIJb4RG3JOTg7a2tq4PN7k5GS/HtefAzu+KYHcmhFtdEZGBgoKCgSzKSIJ/iw7JycHer0+\nLDs2ocOxQLTA0TSeEIPFYkFLSwuSkpJQWFjoN6ny7zLIuAMAd5cRSpK22+1ebsBIHSz29/fjlltu\nQU5ODp544omoPCgOEZb3DFmMkIeGhtDd3Q2apgO2Oi/G2JGUlASdTgebzQaz2ezVVafT6ZbMUScE\nq9Xqpa9e6nWIaYH5dxlyuRxtbW0hV0+EEhRFwWQywWKxhGwUxSdpvrOOb39OSEjwm6RZlkVPTw8G\nBwdRVlYWMQJkGAavvPIKnnvuOfzqV7/C5s2bo2qDEgZIhMwnZIZh0N/fj87OTsTExKCmpiZsVmdg\nYa0u31FnNpthsVjgdru9dMCJiYlh3bm4XC4uEKm0tNRLXx1pkLsMs9nMhQjFxcUhNTWVe22iKdie\nHIjxre3hgtChKsMwXjtpIZK2Wq1oamrirOOROljs6+vD9u3bkZ+fj1/96ldh09pHGZY3IZMITr7V\nOS0tDRkZGeju7sa6devm/f5gidjpdMJkMsFut6OkpCSgN5uvDthisXh90HQ6XUhuWUkOR39/f9SG\n2QOnxhMZGRnIz88HRVFer42/MaXhxMzMDFpaWqBUKmE0GiN2lzMfScfHx3NtzeXl5RG78DIMg9//\n/vfYvXs3nnjiCZx//vlR+b4LEyRCJh/o2NhYlJSUIDY2Fk6nE42Njaiurhb8vmAdduEydvh+0Mgt\na0JCAjfqCORgjKhJ0tPTkZ+fH5Xyq0DMHb7hSvwev1A0YouBf7BoNBqjUifLMAwGBwdhMpm4OzRC\n0nyFx1LMj3t6erB9+3YUFRXh8ccfj6q7sSXC8ibknp4e9Pb2zinQpGkahw4dwoYNG+Z8TzAOO35u\n7VIZO/jqBbPZPK+yg8Cf1o5II1TqCd9wJf68PhQxpSTThAQ+ReNFjaIoro6srKyMky2SnBf+BZ4E\nCfHtz6EiaYZhsHfvXrz44ot48skncd555y2nXTEfy5uQxQ71hJqigx1PRJOxQ0jZoVarER8fj5mZ\nGTgcjqjdyQFzxxOhvKj5JuBZrdagYkr5JQRlZWV+SSQjATKi89fgQUiab38muSb8nXSgCYHd3d24\n+eabUVZWhsceeywq9fZLiOVNyPwITl+QYtJgiZhv7DAajVH5wWRZFr29veju7kZ8fDyXBRwtyg6C\ncGVPLIRAYkr5ksBwudhCAZfLhebmZgBAaWnpou4C+GlvfJL2JyGQYRi89NJL2Lt3L3bu3IlvfvOb\nUfl6LTEkQp6PkM8888yAidjtdqOrqyuqjR3AqTD+xMREFBYWch8aXxIym81LruwgiEZzh28CntVq\nhcvlgtvtRmxsLAoKCpCcnLwkLSiBgG/wIFLOcD0PGXeQ3TQhaZVKhdbWVhgMBtxzzz1YtWoVHn30\n0ajcrEQIEiH7EjI5sKuvr0dsbCznLFvoUIxI5vr6+pCbmztv2FAkEUxrByEhIr0Ll7KDj4mJCe5g\n0Z+8kEiAYRh0dXVhZGQEubm5kMlkAceULgVmZmbQ1NTEHVwv9cWCkHRnZyfuuecefPXVV9Bqtaiq\nqsJVV12Fyy67bEnXE8VY3oTsm/jGP7BjWRbT09McCU1PT3OHYoSkyak0MXakpaX51dgRCYS6tUNI\n2UFyKYJRdhBEajwRKEj0pJgb0Pd23vciJqYDDiXISGpgYAClpaV+O03DgY6ODmzfvh3r1q3DQw89\nBI1Gw4Xur1mzJqzPvXPnTrz44ouQyWRYs2YN9u7dG5X53JAI2UPI/s6J+YdiRLngdrsRExODvLw8\npKWlRXze6guise7o6Ai7lTgYZQdBNI4nhOByubg7jEDbO8TUC+GIKZ2enkZTU1PA1uxQg6Zp7N69\nG6+++iqeeuopnH322Uv6/P39/TjrrLNw8uRJaDQafP/738eWLVtw3XXXLek6/MTyTnuzWq0wm81I\nSkriSHi+XaNKpUJqaiqnSnC73TAajWAYBhaLBQMDA9y8lV9jE6kPA98JWFlZGfYqHd8UM+CU5dls\nNmNkZIRTdhACIlZx0j8YbHVWuMGfwRYWFgalIefnIOfk5ADwvtPo7+/30pDzSdrfiygpQR0fH4+o\nwQMA2tvbsX37dlRXV+OLL75YdGlwsKAoCjMzM1CpVLDb7cjOzo7IOkKF03aHXFdXh5/97Gcwm80o\nKytDdXU1ampqRGMa/TF28A81zGYzrFYrlwPMn0eHc77sdrvR0dEBi8US1oqnYEGMGhMTExgaGuJ2\nicnJyVGj7ODDZrOhubkZWq0WRUVFYZ/BCgUIyWSyBZtHpqamuDFKJEsMaJrGc889h9deew2/+c1v\ncNZZZ0VkHQS7du3CXXfdBY1GgwsuuACvvvpqRNczD5b3yILA7XbjxIkTqK2txaFDh9DQ0AC5XI7K\nykpUVVWhqqoKX3zxBVasWIGqqioYDIaA3uxCt/JKpdJLWhYKOy+/gy2apVe+44nU1NSoUXbwQdM0\nF4VaWloa0QubWKg9IefJyUm4XC6sXLkyYjtRwFPucMstt+CMM87AAw88EPEzgMnJSVx++eX4y1/+\ngqSkJHzve9/DFVdcgWuvvTai6xKBRMhCIAd6hw8fxmuvvYY333wTer0eqampqKqqQnV1Nc4444xF\n5Tu43W7uwNBsNsPhcHipOoi+1V8ItXZEI/xVTyyk7Aj3oRixj2dlZQV8AV4qUBSFvr4+dHd3IzY2\nlitD8GdmH461PPvss3jjjTfw9NNPe5mqIok33ngD+/btw0svvQQA+MMf/oDa2lo8++yzEV6ZIJb3\nDFkM5PZw48aNePnll3HgwAEYjUYMDg6irq4OtbW1+N3vfoeRkREUFxejuroa69evR2VlpV9tIIBn\nHp2WlsYdXvH1v+Pj4+js7ARFUdw8WoyA5mvtiCbw1RNr165dcOckk8mg1Wqh1Wq5mR9/3jowMDBn\n3hqKxhGn04nW1lYwDIOKiopoPY2Hy+VCS0sLGIbBhg0buPMBfkypyWSCzWbz6qkLB0k3Nzfjlltu\nwTe+8Q0cOHAgql6z3Nxc1NbWwm63Q6PR4KOPPsL69esjvaxFYdntkP0FTdNoaWnBwYMHcfDgQRw9\nehRutxtr167lSHrlypVBzxz5p/JkHk0uFqRfbGpqSrTtOhoQbvWE0K18MLtE0t7R19cXsWYMf8DP\nRfHX4CFmmV9sAh5FUXjmmWfw9ttv47e//S3OPPPMYH+ssOKXv/wl/vKXv0CpVKKyshIvvvhi2A+4\ng4Q0sgg17HY7jh49irq6OtTV1eHkyZNISEjgCLqmpmZRTRsURaG7uxt9fX2IiYkBy7JQqVTcLpqv\nj440ImXu4Cs7hAhIp9N5pbtZrVY0NzdDp9NFbbko4Lkbam5uRkxMzKINHi6Xa05MaUxMjNdOej6S\nbmpqwvbt23HOOefgl7/8ZdS8577mkAg53GBZFuPj46irq8PBgwdRV1eH3t5e5ObmoqamBtXV1aiu\nruakd/PBYrGgtbWVO+0nSgTy4SIE5HA4oNFoOJIOdB69WDgcDi7Ho7S0NOIHO4DnNeLPo0kEJ0VR\noCgKZWVlURuqxDd4hDP8aaGYUrvdDr1ej127duHvf/87nn32WdTU1IRlLcsUEiFHAgzDoKOjgxt1\n1NfXw2azYeXKlVi/fj3Wr1+PtWvXcrdVNpsN3d3dfrd2kHm02WzmSIimaS8XXTgajL8u5g7A097R\n1taGpKQkqFQqLpMiLi7O60IW6cPRSBo8SNiUxWLB2NgYbrrpJvT09CAxMRFXXXUVzj33XHzrW99a\nsvUsA0iEHC1wuVz46quvOJI+fvw417bb39+PF154AZWVlUF/IMk8mpA00bbyb+MXc9jzdcieADy7\nd9IgbjQavWaJ/OAgMrMnF7KlsjsT8A0eZWVlXnndSw23242nnnoK77//Pn77298iIyMD9fX1sFqt\nS+J4m5qawrZt29DY2AiZTIY9e/Zg48aNYX/eCEAi5GjFiRMncO2112Lt2rUoKCjA0aNHYTKZkJWV\nxc2j169fv6hMCtLuTHbRNpsNarXaax690OFHNI4nhMAwDHp7ezE4OIiSkhK/D0H5FzLfNpZQKTt8\nYTab0dzcjIyMjIi3Zjc2NuKWW27BBRdcgLvuuisih2Fbt27Fpk2bsG3bNrhcLtjt9tO1eVoi5GgF\n2cnm5uZyf0aUALW1tdyh4cTEBIxGI0fQFRUVi9rpkltUQkBOp5PLtyVErVQqvcYTRUVFUatKAE61\nd4SquDNUyg6hx21vb8f09HTEw+3dbjd+/etfY9++fXjuuedQVVUVkXWYzWZUVFSgo6MjKk1OIYZE\nyF93UBSFpqYmzmV49OhRsCyLdevWcSRdWloa9CyUZVnMzMx4HYiR/F+dToeCggIkJiZGpXHC7XbD\nZDJxBBfONgq+/tdsNnPZCf5Ky4gRRa/Xh72ReiEcP34ct9xyC77zne/gF7/4RURt7A0NDbjxxhux\ncuVKHDt2DNXV1di1a9fpmqEsEfLpBpKlcfjwYW4X3dLSguTkZC/pXTB5zWQ8QVEU9Ho9Z2SxWq1e\nO0SdTgeNRhMxUmFZFiMjI+jo6IiohVxI2cHv7SNW7NbWVk7pEUn5mMvlwhNPPIEPP/wQzz//PCoq\nKiK2FoL6+nps2LABBw4cwJlnnokdO3YgMTERDzzwQKSXFg5IhLwcQAiKHBgeOnQIg4ODKCgo4AKV\nKisrkZiYuGA0pth4gqIor1GH3W7nJFPEDr4UOy2i1VWr1SgpKYmqkCK+G9NsNmNsbAwzMzNISEhA\nenq610hoqXHs2DHs2LEDF110Ee64446oed2GhoawYcMGdHV1AQA+//xzPProo3j//fcju7DwQCLk\n5QqGYdDW1sbNo48cOQKHw4HVq1dzJL1q1Sp8+OGH0Gq1MBgMAasnfAODXC4XFxhEoklD2VxMAvgj\nHca+EBwOB5qamriLBnHSRULZ4XQ68fjjj+OTTz7BCy+8gLVr14bleRaDTZs24cUXX0RpaSnuvfde\n2Gw2PP7445FeVjggEbKEU3A6nWhoaEBtbS3279+Pzz77DCtWrMA555yDM844A+vXr19U27NQYBDL\nspw+OljFAomdTE9PD3kbdShBDmX7+/vnVXr4Kjump6e5CNdQKjsaGhqwY8cOXHLJJfjf//3fqOsB\nJGhoaOAUFoWFhdi7d29UX3AXAYmQAWDfvn3YsWMHaJrGtm3bcMcdd0R6SRHFzMwMvv3uXNOZAAAK\n1ElEQVTtb+POO+/Exo0bcejQIe7QsKurC3q9njswrK6uRkpKStAzWoZhvA7DxKqyxBpc2tra4HA4\nUFZWFrXBSoDH3NPU1ITExMSg7NmhVHY4nU489thj+Pzzz/H888+HvUJJgt+QCJmmaRiNRnzwwQfQ\n6/WoqanBn//8Z6xcuTLSS4soWJYVnSd3d3d7zaOtVqtXwL8/aW7zwbcqa2ZmhjsMI6OOiYkJdHV1\noaCgYFExqOEGeb1GRkZQVlYW0kxlMrcnr5M/oUFHjhzBT3/6U1x++eW47bbbonZXvEwhEfKXX36J\ne++9F//85z8BAI888ggA4M4774zksr5WcLvdaGxs5ObRX331FRQKBRfwX1NTg5KSkqDnoMTCazab\nMT4+juHhYchkMiQnJyMpKSniVVliiITBQ0jZceLECZw8eRJTU1Po7OzEnj17sGrVqrCvRULAkPKQ\n+/v7YTAYuN/r9XocPHgwgiv6+kGlUqGyshKVlZW46aabwLIsrFYrDh8+jNraWjz44IOcrZovvfN3\nZyuTyaBSqTA9PQ2r1YrKykqui89isWBwcJDLWiZVWaSLLhI7Z5qmYTKZYLFYsHr16iXVzKrVaqSn\np3NKGPJv8cYbb0CtVkOr1eIHP/gB7rvvPlxyySVLsiaaprF+/Xrk5OTgvffeW5LnPJ1xWhOyhNCD\nZGSce+65OPfccwF4iGFgYIAL+H/hhRcwOjqKkpISLvGuqqpKsG9wYmICra2tyMzMRE1NDbfTjI+P\nR3x8PBdgz5+zdnV1ha0qaz4Qg0dOTg5KSkoiOkpxOBx4+OGHcfDgQezevRvl5eUAPGMUl8u1ZOvY\ntWsXysvLYbFYluw5T2ec1oSck5OD3t5e7vd9fX1cI7CE0EEmkyEnJweXXnopLr30UgAeAm1ubsbB\ngwfxt7/9Dffccw9omuYC/gsKCvD666/jhhtuEC2e5UOs9ZrMWAcGBhZdlSUGt9uN1tZWuN3uqGga\nIQW+V199NT755BMveaFcLl+y9fX19eH999/HXXfdhV//+tdL8pynO07rGTJFUTAajfjoo4+Qk5OD\nmpoa/OlPf5JmbBGC3W5HfX09nnnmGXz00UcoLy+HTCbzGnXk5OQsSnrHN2dYLBa/qrLmezziCoyG\nA8aZmRk8+OCDOHLkCF544QWUlZVFbC0AcMUVV+DOO++E1WrFE088IY0s5oc0Q1YqlXjmmWewefNm\n0DSN66+/fsnIuLe3Fz/60Y+4Q6obb7wRO3bsWJLnjlbExcVBr9fDYDCgq6sL8fHxGBsb4wL+//jH\nP6Kvrw95eXle0judTuf3PFqj0UCj0WDFihUAvKuySFcfqcoiJC00SnE4HGhuboZSqUR1dXXE3W21\ntbW47bbbcO211+KJJ56I+CHne++9h4yMDFRXV+PTTz+N6FpOJ5zWO+RIYnBwEIODg6iqqoLVakV1\ndTX+9re/LXvJ3UJgGAYmk4lrYKmvr4fdbvcK+F+zZs2ioiLJPJqk7tntdi6fmrRnDA0NRUWfod1u\nxwMPPICGhgb87ne/g9FojOh6CO6880688sorUCqV3F3JZZddhj/+8Y+RXlq0QpK9RRMuvvhi3Hzz\nzfj2t78d6aV87eByuXDs2DFOH93Y2IiYmBhUVlZyJF1UVLQo6ZnL5cLIyAi6urrAsiyUSmVEq7IA\n4N///jd+/vOfY+vWrdi+fXvEd8Vi+PTTT6WRxcKQCDla0NXVhbPPPhuNjY0RbYc4XcCyLMxmMw4d\nOsTtpDs6OpCdnc1po9evX4+0tDS/Rh1CBo9IVWUBHuff/fffj8bGRuzevRslJSUhf45QQiJkvyAR\ncjRgenoa55xzDu666y5cdtllkV7OaQuWZdHT08MR9KFDhzA5OTkn4N83OtRisaC5uRlpaWkLZmWE\nuyqLZVkcOHAAt99+O66//nr893//d9TuiiUEDImQIw23242LLroImzdvxq233hrp5Sw7UBSFEydO\ncDbwo0ePQiaTYd26dVi7di3+/e9/4xvf+AauvPLKoAPuQ1GVBXh2xffeey+am5uxe/duFBUVBbUe\nCVELiZAjCZZlsXXrVqSkpOCpp55a8ueXHFRzQQL+X3zxRTz++OMoKSnBxMQEUlJSvKR3iw2997cq\ni6zp888/xx133IH/+q//wk033RS1iXYSFgVJ9hZJHDhwAK+88grWrFnDtTM8/PDD2LJly5I8v+Sg\nmguZTIb4+HjQNI0vv/wSubm5YFkWw8PD3IHhnj17MDQ0hMLCQq+A/4SEBL9JOiYmZo7FmVRljY6O\nciqS/fv3w+12Y2pqCq+//nrUKCgkRA7SDvk0RF9fH7Zu3co5qKQdcmBgGAatra1eAf8ul2tOwH+w\nqguWZfHhhx/i0UcfRWFhIVQqFRobG3Hdddfh5ptvDvFPMxeSRj4ikEYWyxWSgyr0cDgcXMD/oUOH\ncOLECWi1WlRVVXGHhv6kvlmtVtx9993o6urC7t27kZ+fz/2dWCxqqCFp5CMCaWSxHCE5qMKD2NhY\nbNiwARs2bADgIc+JiQku4P/1119Hd3c3DAaDl8swOTkZMpkMLMvi008/xS9+8Qv85Cc/wfPPPz+H\nvJfKlp2VlYWsrCwAQEJCAsrLy9Hf3y8RchRA2iGfZpAcVJEDwzDo6uriRh319fWwWq0wGo0YGRmB\nRqPB7t27kZubG+mlcpA08ksGaWSx3CEJ9iMPt9uNr776Cu+++y7uueeeqFJQSBr5JYU0spAQOUxN\nTWHbtm1obGyETCbDnj17sHHjxkgva8mhUqm4TOhogtvtxuWXX45rrrlGIuMogrRDlhAWbN26FZs2\nbeIahe12u1eWsYTIIdIa+WUKaWQhITIwm82oqKhAR0dH1BaULmd88cUX2LRpE9asWcONUJZSI79M\nIRGyhMigoaEBN954I1auXIljx46huroau3btWtL+OQkSogx+EXL0nDBIOG1AURSOHDmCm266CUeP\nHoVWq8Wjjz4a6WVJkBD1kAhZQsih1+uh1+tx5plnAvAYVY4cORLhVUmQEP2QCFlCyJGZmQmDwYCW\nlhYAwEcffSSZDiRI8AMSIUsIC55++mlcc801WLt2LRoaGvCLX/xiSZ9/586dWLVqFVavXo2rr74a\nDodjSZ8/XNi3bx9KS0tRXFwsjYFOQ0iHehJOO/T39+Oss87CyZMnodFo8P3vfx9btmzBddddF+ml\nLQo0TcNoNOKDDz6AXq9HTU0N/vznP0t3H18PSId6EpYvKIrCzMwMKIqC3W5HdnZ2pJe0aNTV1aG4\nuBiFhYVQq9W46qqr8M4770R6WRJCCImQJZx2yMnJwW233Ybc3FxkZWVBp9PhggsuiPSyFo3+/n4Y\nDAbu93q9Hv39/RFckYRQQyJkCacdJicn8c4776CzsxMDAwOw2WxSuJKErwUCnSFLkBD1kMlk3wPw\nnyzL3jD7+x8B2MCy7H9HdmWLg0wm2wjgXpZlN8/+/k4AYFn2kYguTELIIO2QJZyO6AGwQSaTxck8\n3u3zADRFeE2hwCEAJTKZrEAmk6kBXAXg7xFek4QQQkp7k3DagWXZgzKZ7E0ARwBQAI4C2B3ZVS0e\nLMtSMpnsZgD/BKAAsIdl2RMRXpaEEEIaWUiQIEFClEAaWUiQIEFClEAiZAkSJEiIEkiELEGCBAlR\nAomQJUiQICFKIBGyBAkSJEQJJEKWIEGChCiBRMgSJEiQECX4/47nJ4ZuxoMFAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7bfd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "point  = np.array([1, 2, 3])\n",
    "normal = np.array([1, 1, 2])\n",
    "\n",
    "# a plane is a*x+b*y+c*z+d=0\n",
    "# [a,b,c] is the normal. Thus, we have to calculate\n",
    "# d and we're set\n",
    "d = -point.dot(normal)\n",
    "\n",
    "# create x,y\n",
    "xx, yy = np.meshgrid(range(10), range(10))\n",
    "\n",
    "# calculate corresponding z\n",
    "z = (-normal[0] * xx - normal[1] * yy - d) * 1. /normal[2]\n",
    "\n",
    "# plot the surface\n",
    "plt3d = plt.figure().gca(projection='3d')\n",
    "plt3d.plot_surface(xx, yy, z)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color='red'>Exercise</font>\n",
    "\n",
    "Try different values for the learning rate $\\alpha$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color='red'>Exercise</font>\n",
    "\n",
    "Plot curves of the cost function $J$ as a function of the number of iterations for different $\\alpha$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
